--- a/jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpSenderOverHTTP.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpSenderOverHTTP.java
@@ -71,7 +71,7 @@ public class HttpSenderOverHTTP extends HttpSender
             String query = request.getQuery();
             if (query != null)
                 path += "?" + query;
-            metaData = new MetaData.Request(request.getMethod(), HttpURI.from(path), request.getVersion(), request.getHeaders(), contentLength, request.getTrailers());
+            metaData = new MetaData.Request(request.getMethod(), HttpURI.from(request.getMethod(), path), request.getVersion(), request.getHeaders(), contentLength, request.getTrailers());
             if (LOG.isDebugEnabled())
                 LOG.debug("Sending headers with content {} last={} for {}", BufferUtil.toDetailString(contentBuffer), lastContent, exchange.getRequest());
             headersCallback.iterate();
--- a/jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java
+++ b/jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java
@@ -519,7 +519,7 @@ public class HttpGenerator
     {
         header.put(StringUtil.getBytes(request.getMethod()));
         header.put((byte)' ');
-        header.put(StringUtil.getBytes(request.getURIString()));
+        header.put(HttpMethod.CONNECT.is(request.getMethod()) ? StringUtil.getBytes(request.getURI().getAuthority()) : StringUtil.getBytes(request.getURIString()));
         header.put((byte)' ');
         header.put(request.getHttpVersion().toBytes());
         header.put(HttpTokens.CRLF);
--- a/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java
+++ b/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java
@@ -1072,7 +1072,8 @@ public interface HttpURI
                         switch (c)
                         {
                             case '/':
-                                mark = i;
+                                pathMark = mark = i;
+                                segment = mark + 1;
                                 state = State.HOST_OR_PATH;
                                 break;
                             case ';':
@@ -1110,6 +1111,8 @@ public interface HttpURI
                                 pathMark = segment = i;
                                 state = State.PATH;
                                 break;
+                            case ':':
+                                throw new IllegalArgumentException("Bad Scheme");
                             default:
                                 mark = i;
                                 if (_scheme == null)
@@ -1130,7 +1133,7 @@ public interface HttpURI
                         {
                             case ':':
                                 // must have been a scheme
-                                _scheme = uri.substring(mark, i);
+                                _scheme = URIUtil.validateScheme(uri.substring(mark, i));
                                 // Start again with scheme set
                                 state = State.START;
                                 break;
@@ -1235,8 +1238,19 @@ public interface HttpURI
                                 encoded = true;
                                 break;
                             case '#':
+                            case '?':
                             case ';':
-                                throw new IllegalArgumentException("Bad authority");
+                                if (encodedCharacters > 0)
+                                    throw new IllegalArgumentException("Bad authority");
+                                _host = uri.substring(mark, i);
+                                if (_host.isEmpty())
+                                    throw new IllegalArgumentException("Bad authority");
+                                encoded = false;
+                                pathMark = mark = i;
+                                segment = mark + 1;
+                                state = State.PATH;
+                                i--;
+                                break;

                             default:
                                 if (encodedCharacters > 0)
@@ -1260,8 +1274,13 @@ public interface HttpURI
                             case '/':
                                 throw new IllegalArgumentException("No closing ']' for ipv6 in " + uri);
                             case ']':
-                                c = uri.charAt(++i);
-                                _host = uri.substring(mark, i);
+                                i++;
+                                String host = uri.substring(mark, i);
+                                URIUtil.validateInetAddress(host);
+                                _host = host;
+                                if (i == end)
+                                    break;
+                                c = uri.charAt(i);
                                 if (c == ':')
                                 {
                                     mark = i + 1;
@@ -1276,7 +1295,7 @@ public interface HttpURI
                             case ':':
                                 break;
                             default:
-                                if (!isHexDigit(c))
+                                if (!isHexDigit(c) && c != '.')
                                     throw new IllegalArgumentException("Bad authority");
                                 break;
                         }
@@ -1444,6 +1463,7 @@ public interface HttpURI
                     break;
                 case SCHEME_OR_PATH:
                 case HOST_OR_PATH:
+                    checkSegment(uri, segment, end, false);
                     _path = uri.substring(mark, end);
                     break;
                 case HOST:
--- a/jetty-http/src/test/java/org/eclipse/jetty/http/HttpURITest.java
+++ b/jetty-http/src/test/java/org/eclipse/jetty/http/HttpURITest.java
@@ -26,6 +26,7 @@ import org.junit.jupiter.params.provider.Arguments;
 import org.junit.jupiter.params.provider.MethodSource;

 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.empty;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
@@ -207,6 +208,20 @@ public class HttpURITest
         assertEquals("/@foo/bar", uri.getPath());
     }

+    /**
+     * Test of an HttpURI of just a "/".
+     * The {@link HttpURI#from(String)} is used by HttpServletResponse.sendRedirect(String).
+     */
+    @Test
+    public void testFromSlash()
+    {
+        HttpURI uri = HttpURI.from("/");
+        assertThat("has no violations", uri.getViolations(), is(empty()));
+        assertNull(uri.getScheme());
+        assertNull(uri.getAuthority());
+        assertEquals("/", uri.getPath());
+    }
+
     @Test
     public void testParams()
     {
@@ -721,6 +736,8 @@ public class HttpURITest

             // Simple IPv6 host no port (default path)
             Arguments.of("http://[2001:db8::1]/", "http", "[2001:db8::1]", null, "/", null, null, null),
+            Arguments.of("http://[0:0:0:0:0:ffff:127.0.0.1]/", "http", "[0:0:0:0:0:ffff:127.0.0.1]", null, "/", null, null, null),
+            Arguments.of("http://[::ffff:127.0.0.1]/", "http", "[::ffff:127.0.0.1]", null, "/", null, null, null),

             // Scheme-less IPv6, host with port (default path)
             Arguments.of("//[2001:db8::1]:8080/", null, "[2001:db8::1]", "8080", "/", null, null, null),
@@ -762,7 +779,8 @@ public class HttpURITest
             assertThat("[" + input + "] .param", httpUri.getParam(), is(param));
             assertThat("[" + input + "] .query", httpUri.getQuery(), is(query));
             assertThat("[" + input + "] .fragment", httpUri.getFragment(), is(fragment));
-            assertThat("[" + input + "] .toString", httpUri.toString(), is(input));
+            if (!input.contains(":ffff:127.0.0.1"))
+                assertThat("[" + input + "] .toString", httpUri.toString(), is(input));
         }
         catch (URISyntaxException e)
         {
@@ -795,7 +813,8 @@ public class HttpURITest
         HttpURI httpUri = HttpURI.from(javaUri);

         assertThat("[" + input + "] .scheme", httpUri.getScheme(), is(scheme));
-        assertThat("[" + input + "] .host", httpUri.getHost(), is(host));
+        if (!input.contains(":ffff:127.0.0.1"))
+            assertThat("[" + input + "] .host", httpUri.getHost(), is(host));
         assertThat("[" + input + "] .port", httpUri.getPort(), is(port == null ? -1 : port));
         assertThat("[" + input + "] .path", httpUri.getPath(), is(path));
         assertThat("[" + input + "] .param", httpUri.getParam(), is(param));
@@ -890,6 +909,24 @@ public class HttpURITest
         assertEquals("//host", uri.asString());
     }

+    public static Stream<String> badSchemes()
+    {
+        return Stream.of(
+            "://host/path",
+            "\t://host/path",
+            "  ://host/path",
+            "unknown^://host/path",
+            "http^://host/path"
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("badSchemes")
+    public void testBadSchemes(String uri)
+    {
+        assertThrows(IllegalArgumentException.class, () -> HttpURI.from(uri));
+    }
+
     public static Stream<String> badAuthorities()
     {
         return Stream.of(
@@ -909,9 +946,11 @@ public class HttpURITest
             "https://user@host:notport/path",
             "https://user:password@host:notport/path",
             "https://user @host.com/",
-            "https://user#@host.com/",
             "https://[notIpv6]/",
-            "https://bad[0::1::2::3::4]/"
+            "https://bad[0::1::2::3::4]/",
+            "http://[normal.com@]vulndetector.com/",
+            "http://normal.com[user@vulndetector].com/",
+            "http://normal.com[@]vulndetector.com/"
         );
     }

@@ -921,4 +960,40 @@ public class HttpURITest
     {
         assertThrows(IllegalArgumentException.class, () -> HttpURI.from(uri));
     }
+
+    public static Stream<Arguments> authoritiesNoPath()
+    {
+        return Stream.of(
+            Arguments.of("http://good.com#@evil.com", "good.com", null, "@evil.com"),
+            Arguments.of("http://good.com?@evil.com", "good.com", "@evil.com", null)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("authoritiesNoPath")
+    public void testAuthorityNoPath(String uri, String authority, String query, String fragment)
+    {
+        HttpURI httpURI = HttpURI.from(uri);
+        assertThat(httpURI.getAuthority(), is(authority));
+        assertThat(httpURI.getPath(), is(""));
+        assertThat(httpURI.getQuery(), is(query));
+        assertThat(httpURI.getFragment(), is(fragment));
+    }
+
+    public static Stream<Arguments> connectURIs()
+    {
+        return Stream.of(
+            Arguments.of("localhost:8080"),
+            Arguments.of("127.0.0.1:8080"),
+            Arguments.of("[::1]:8080")
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("connectURIs")
+    public void testConnect(String authority)
+    {
+        HttpURI httpURI = HttpURI.from(HttpMethod.CONNECT.asString(), authority);
+        assertThat(httpURI.getAuthority(), is(authority));
+    }
 }
--- a/jetty-server/src/test/java/org/eclipse/jetty/server/handler/NcsaRequestLogTest.java
+++ b/jetty-server/src/test/java/org/eclipse/jetty/server/handler/NcsaRequestLogTest.java
@@ -353,9 +353,9 @@ public class NcsaRequestLogTest
         setup(logType);
         testHandlerServerStart();

-        _connector.getResponse("GET http://[:1]/foo HTTP/1.1\nReferer: http://other.site\n\n");
+        _connector.getResponse("GET http://[::1]/foo HTTP/1.1\nReferer: http://other.site\n\n");
         String log = _entries.poll(5, TimeUnit.SECONDS);
-        assertThat(log, containsString("GET http://[:1]/foo "));
+        assertThat(log, containsString("GET http://[::1]/foo "));
         assertThat(log, containsString(" 400 50 \"http://other.site\" \"-\""));
     }

@@ -366,9 +366,9 @@ public class NcsaRequestLogTest
         setup(logType);
         testHandlerServerStart();

-        _connector.getResponse("GET http://[:1]/foo HTTP/1.1\nReferer: http://other.site\nUser-Agent: Mozilla/5.0 (test)\n\n");
+        _connector.getResponse("GET http://[::1]/foo HTTP/1.1\nReferer: http://other.site\nUser-Agent: Mozilla/5.0 (test)\n\n");
         String log = _entries.poll(5, TimeUnit.SECONDS);
-        assertThat(log, containsString("GET http://[:1]/foo "));
+        assertThat(log, containsString("GET http://[::1]/foo "));
         assertThat(log, containsString(" 400 50 \"http://other.site\" \"Mozilla/5.0 (test)\""));
     }

--- a/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java
+++ b/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java
@@ -13,6 +13,7 @@

 package org.eclipse.jetty.util;

+import java.net.InetAddress;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.nio.charset.Charset;
@@ -1404,6 +1405,67 @@ public final class URIUtil
         return URI.create(buf.toString());
     }

+    /**
+     * Validate an IPv4 or IPv6 address.
+     * @param inetAddress the address to validate
+     * @throws IllegalArgumentException if the address is not valid
+     */
+    public static void validateInetAddress(String inetAddress)
+    {
+        try
+        {
+            InetAddress ignored = InetAddress.getByName(inetAddress);
+        }
+        catch (Throwable e)
+        {
+            throw new IllegalArgumentException("Bad [IPv6] address", e);
+        }
+    }
+
+    /**
+     * Validate and normalize the scheme,
+     *
+     * @param scheme The scheme to normalize
+     * @return The normalized version of the scheme
+     * @throws IllegalArgumentException If the scheme is not valid
+     */
+    public static String validateScheme(String scheme)
+    {
+        if (scheme == null || scheme.isEmpty())
+            throw new IllegalArgumentException("Bad scheme");
+
+        //  scheme = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+        StringBuilder toLowerCase = null;
+        for (int i = 0; i < scheme.length(); i++)
+        {
+            char c = scheme.charAt(i);
+            if (c >= 'A' && c <= 'Z')
+            {
+                if (toLowerCase == null)
+                {
+                    toLowerCase = new StringBuilder(scheme.length());
+                    toLowerCase.append(scheme, 0, i);
+                }
+                toLowerCase.append(Character.toLowerCase(c));
+            }
+            else if (c >= 'a' && c <= 'z' ||
+                (i > 0 && (c >= '0' && c <= '9' ||
+                    c == '.' ||
+                    c == '+' ||
+                    c == '-')))
+            {
+                if (toLowerCase != null)
+                    toLowerCase.append(c);
+            }
+            else
+            {
+                throw new IllegalArgumentException("Bad scheme");
+            }
+        }
+
+        return toLowerCase == null ? scheme : toLowerCase.toString();
+    }
+
     /**
      * Combine two query strings into one. Each query string should not contain the beginning '?' character, but
      * may contain multiple parameters separated by the '{@literal &}' character.

diff --git a/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java b/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java
--- a/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java
+++ b/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java
@@ -627,6 +627,8 @@ public interface HttpURI
             _uri = null;
             _scheme = baseURI.getScheme();
             _user = baseURI.getUser();
+            if (_user != null)
+                _violations.add(Violation.USER_INFO);
             _host = baseURI.getHost();
             _port = baseURI.getPort();
             if (pathQuery != null)
@@ -638,6 +640,8 @@ public interface HttpURI
             _uri = null;
             _scheme = baseURI.getScheme();
             _user = baseURI.getUser();
+            if (_user != null)
+                _violations.add(Violation.USER_INFO);
             _host = baseURI.getHost();
             _port = baseURI.getPort();
             if (path != null)
@@ -664,6 +668,8 @@ public interface HttpURI
                 _host = "";
             _port = uri.getPort();
             _user = uri.getUserInfo();
+            if (_user != null)
+                _violations.add(Violation.USER_INFO);
             String path = uri.getRawPath();
             if (path != null)
                 parse(State.PATH, path);
@@ -1045,6 +1051,10 @@ public interface HttpURI
         public Mutable user(String user)
         {
             _user = user;
+            if (user == null)
+                _violations.remove(Violation.USER_INFO);
+            else
+                _violations.add(Violation.USER_INFO);
             _uri = null;
             return this;
         }
diff --git a/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java b/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java
--- a/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java
+++ b/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java
@@ -123,7 +123,8 @@ public final class UriCompliance implements ComplianceViolation.Mode
             Violation.AMBIGUOUS_PATH_SEPARATOR,
             Violation.AMBIGUOUS_PATH_ENCODING,
             Violation.AMBIGUOUS_EMPTY_SEGMENT,
-            Violation.UTF16_ENCODINGS));
+            Violation.UTF16_ENCODINGS,
+            Violation.USER_INFO));

     /**
      * Compliance mode that exactly follows <a href="https://tools.ietf.org/html/rfc3986">RFC3986</a>,
