--- a/src/main/java/com/mysql/cj/conf/PropertyKey.java
+++ b/src/main/java/com/mysql/cj/conf/PropertyKey.java
@@ -72,7 +73,6 @@
     authenticationFidoCallbackHandler("authenticationFidoCallbackHandler", true), //
     authenticationPlugins("authenticationPlugins", true), //
     autoClosePStmtStreams("autoClosePStmtStreams", true), //
-    autoDeserialize("autoDeserialize", true), //
     autoGenerateTestcaseScript("autoGenerateTestcaseScript", true), //
     autoReconnect("autoReconnect", true), //
     autoReconnectForPools("autoReconnectForPools", true), //
--- a/src/main/java/com/mysql/cj/conf/PropertyDefinitions.java
+++ b/src/main/java/com/mysql/cj/conf/PropertyDefinitions.java
@@ -518,9 +519,6 @@
                 //
                 // CATEGORY_BLOBS
                 //
-                new BooleanPropertyDefinition(PropertyKey.autoDeserialize, DEFAULT_VALUE_FALSE, RUNTIME_MODIFIABLE,
-                        Messages.getString("ConnectionProperties.autoDeserialize"), "3.1.5", CATEGORY_BLOBS, Integer.MIN_VALUE),
-
                 new MemorySizePropertyDefinition(PropertyKey.blobSendChunkSize, 1024 * 1024, RUNTIME_MODIFIABLE,
                         Messages.getString("ConnectionProperties.blobSendChunkSize"), "3.1.9", CATEGORY_BLOBS, Integer.MIN_VALUE, 0, 0),
 
@@ -784,8 +782,9 @@
                 new BooleanPropertyDefinition(PropertyKey.gatherPerfMetrics, DEFAULT_VALUE_FALSE, RUNTIME_MODIFIABLE,
                         Messages.getString("ConnectionProperties.gatherPerfMetrics"), "3.1.2", CATEGORY_DEBUGING_PROFILING, 10),
 
+                // TODO currently is not used !!!
                 new IntegerPropertyDefinition(PropertyKey.reportMetricsIntervalMillis, 30000, RUNTIME_MODIFIABLE,
-                        Messages.getString("ConnectionProperties.reportMetricsIntervalMillis"), "3.1.2", CATEGORY_DEBUGING_PROFILING, 11, 0, Integer.MAX_VALUE), // TODO currently is not used !!!
+                        Messages.getString("ConnectionProperties.reportMetricsIntervalMillis"), "3.1.2", CATEGORY_DEBUGING_PROFILING, 11, 0, Integer.MAX_VALUE),
 
                 new BooleanPropertyDefinition(PropertyKey.logXaCommands, DEFAULT_VALUE_FALSE, RUNTIME_MODIFIABLE,
                         Messages.getString("ConnectionProperties.logXaCommands"), "5.0.5", CATEGORY_DEBUGING_PROFILING, 12),
--- a/src/main/java/com/mysql/cj/LocalizedErrorMessages.properties
+++ b/src/main/java/com/mysql/cj/LocalizedErrorMessages.properties
@@ -154,7 +154,6 @@
 Connection.CantDetectLocalConnect=Unable to determine if hostname ''{0}'' is local to this box because of exception, assuming it''s not.
 Connection.NoMetadataOnSocketFactory=Configured socket factory does not implement SocketMetadata, can not determine whether server is locally-connected, assuming not"
 Connection.LoginTimeout=Connection attempt exceeded defined timeout.
-Connection.WarnAutoDeserialize=This connection is using the auto-deserialization feature (''autoDeserialize=true'') which is deprecated and will be removed in a future release of MySQL Connector/J.
 
 ConnectionGroup.0=Cannot remove host, only one configured host active.
 ConnectionGroup.1=Host is not configured: {0}
@@ -470,9 +469,6 @@
 ResultSet.Invalid_value_for_getShort()_-____217=Invalid value for getShort() - ''
 ResultSet.___in_column__218='' in column 
 
-ResultSet.Class_not_found___91=Class not found: 
-ResultSet._while_reading_serialized_object_92=\ while reading serialized object
-
 ResultSet.Invalid_value_for_getShort()_-____96=Invalid value for getShort() - ''
 ResultSet.Unsupported_character_encoding____101=Unsupported character encoding ''
 
@@ -809,7 +805,6 @@
 ConnectionProperties.authenticationPlugins=Comma-delimited list of classes that implement the interface ''com.mysql.cj.protocol.AuthenticationPlugin''. These plugins will be loaded at connection initialization and can be used together with their sever-side counterparts for authenticating users, unless they are also disabled in the connection property ''disabledAuthenticationPlugins''.
 ConnectionProperties.authenticationFidoCallbackHandler=Fully-qualified class name of a class implementing the interface ''com.mysql.cj.callback.MysqlCallbackHandler''. This class will be used by the FIDO authentication plugin to obtain the authenticator data and signature required for the FIDO authentication process. See the documentation of ''com.mysql.cj.callback.FidoAuthenticationCallback'' for more details.
 ConnectionProperties.autoClosePstmtStreams=Should the driver automatically call the method ''close()'' on streams/readers passed as arguments via ''set*()'' methods?
-ConnectionProperties.autoDeserialize=Should the driver automatically detect and de-serialize objects stored in BLOB fields?
 ConnectionProperties.autoGenerateTestcaseScript=Should the driver dump the SQL it is executing, including server-side prepared statements to STDERR?
 ConnectionProperties.autoReconnect=Should the driver try to re-establish stale and/or dead connections? If enabled the driver will throw an exception for queries issued on a stale or dead connection, which belong to the current transaction, but will attempt reconnect before the next query issued on the connection in a new transaction. The use of this feature is not recommended, because it has side effects related to session state and data consistency when applications don''t handle SQLExceptions properly, and is only designed to be used when you are unable to configure your application to handle SQLExceptions resulting from dead and stale connections properly. Alternatively, as a last option, investigate setting the MySQL server variable ''wait_timeout'' to a high value, rather than the default of 8 hours.
 ConnectionProperties.autoReconnectForPools=Use a reconnection strategy appropriate for connection pools?

--- a/src/main/java/com/mysql/cj/jdbc/ConnectionImpl.java
+++ b/src/main/java/com/mysql/cj/jdbc/ConnectionImpl.java
@@ -426,10 +428,6 @@
                 this.propertySet.getProperty(PropertyKey.socketFactory).setValue(SocksProxySocketFactory.class.getName());
             }
 
-            if (this.propertySet.getBooleanProperty(PropertyKey.autoDeserialize.getKeyName()).getValue()) {
-                this.session.getLog().logWarn(Messages.getString("Connection.WarnAutoDeserialize"));
-            }
-
             this.dbmd = getMetaData(false, false);
             initializeSafeQueryInterceptors();
         } catch (CJException e) {
--- a/src/main/java/com/mysql/cj/jdbc/result/ResultSetImpl.java
+++ b/src/main/java/com/mysql/cj/jdbc/result/ResultSetImpl.java
@@ -29,11 +29,8 @@
 
 package com.mysql.cj.jdbc.result;
 
-import java.io.ByteArrayInputStream;
-import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.io.ObjectInputStream;
 import java.io.Reader;
 import java.io.UnsupportedEncodingException;
 import java.math.BigDecimal;
@@ -384,35 +381,31 @@
 
             if (this.rowData.size() == 0) {
                 b = false;
-            } else {
-                if (row == 0) {
+            } else if (row == 0) {
+                beforeFirst();
+                b = false;
+            } else if (row == 1) {
+                b = first();
+            } else if (row == -1) {
+                b = last();
+            } else if (row > this.rowData.size()) {
+                afterLast();
+                b = false;
+            } else if (row < 0) {
+                // adjust to reflect after end of result set
+                int newRowPosition = this.rowData.size() + row + 1;
+
+                if (newRowPosition <= 0) {
                     beforeFirst();
                     b = false;
-                } else if (row == 1) {
-                    b = first();
-                } else if (row == -1) {
-                    b = last();
-                } else if (row > this.rowData.size()) {
-                    afterLast();
-                    b = false;
                 } else {
-                    if (row < 0) {
-                        // adjust to reflect after end of result set
-                        int newRowPosition = this.rowData.size() + row + 1;
-
-                        if (newRowPosition <= 0) {
-                            beforeFirst();
-                            b = false;
-                        } else {
-                            b = absolute(newRowPosition);
-                        }
-                    } else {
-                        row--; // adjust for index difference
-                        this.rowData.setCurrentRow(row);
-                        this.thisRow = this.rowData.get(row);
-                        b = true;
-                    }
+                    b = absolute(newRowPosition);
                 }
+            } else {
+                row--; // adjust for index difference
+                this.rowData.setCurrentRow(row);
+                this.thisRow = this.rowData.get(row);
+                b = true;
             }
 
             setRowPositionValidity();
@@ -1130,33 +1123,6 @@
                 // TODO Field sets binary and blob flags if the length of BIT field is > 1; is it needed at all?
                 if (field.isBinary() || field.isBlob()) {
                     byte[] data = getBytes(columnIndex);
-
-                    if (this.connection.getPropertySet().getBooleanProperty(PropertyKey.autoDeserialize).getValue()) {
-                        Object obj = data;
-
-                        if ((data != null) && (data.length >= 2)) {
-                            if ((data[0] == -84) && (data[1] == -19)) {
-                                // Serialized object?
-                                try {
-                                    ByteArrayInputStream bytesIn = new ByteArrayInputStream(data);
-                                    ObjectInputStream objIn = new ObjectInputStream(bytesIn);
-                                    obj = objIn.readObject();
-                                    objIn.close();
-                                    bytesIn.close();
-                                } catch (ClassNotFoundException cnfe) {
-                                    throw SQLError.createSQLException(Messages.getString("ResultSet.Class_not_found___91") + cnfe.toString()
-                                            + Messages.getString("ResultSet._while_reading_serialized_object_92"), getExceptionInterceptor());
-                                } catch (IOException ex) {
-                                    obj = data; // not serialized?
-                                }
-                            } else {
-                                return getString(columnIndex);
-                            }
-                        }
-
-                        return obj;
-                    }
-
                     return data;
                 }
 
@@ -1232,38 +1198,6 @@
             case MEDIUMBLOB:
             case LONGBLOB:
             case BLOB:
-                if (field.isBinary() || field.isBlob()) {
-                    byte[] data = getBytes(columnIndex);
-
-                    if (this.connection.getPropertySet().getBooleanProperty(PropertyKey.autoDeserialize).getValue()) {
-                        Object obj = data;
-
-                        if ((data != null) && (data.length >= 2)) {
-                            if ((data[0] == -84) && (data[1] == -19)) {
-                                // Serialized object?
-                                try {
-                                    ByteArrayInputStream bytesIn = new ByteArrayInputStream(data);
-                                    ObjectInputStream objIn = new ObjectInputStream(bytesIn);
-                                    obj = objIn.readObject();
-                                    objIn.close();
-                                    bytesIn.close();
-                                } catch (ClassNotFoundException cnfe) {
-                                    throw SQLError.createSQLException(Messages.getString("ResultSet.Class_not_found___91") + cnfe.toString()
-                                            + Messages.getString("ResultSet._while_reading_serialized_object_92"), getExceptionInterceptor());
-                                } catch (IOException ex) {
-                                    obj = data; // not serialized?
-                                }
-                            } else {
-                                return getString(columnIndex);
-                            }
-                        }
-
-                        return obj;
-                    }
-
-                    return data;
-                }
-
                 return getBytes(columnIndex);
 
             case YEAR:
@@ -1414,18 +1348,6 @@
                 return (T) this.thisRow.getValue(columnIndex - 1, new DurationValueFactory(this.session.getPropertySet()));
             }
 
-            if (this.connection.getPropertySet().getBooleanProperty(PropertyKey.autoDeserialize).getValue()) {
-                try {
-                    return (T) getObject(columnIndex);
-                } catch (ClassCastException cce) {
-                    SQLException sqlEx = SQLError.createSQLException("Conversion not supported for type " + type.getName(),
-                            MysqlErrorNumbers.SQL_STATE_ILLEGAL_ARGUMENT, getExceptionInterceptor());
-                    sqlEx.initCause(cce);
-
-                    throw sqlEx;
-                }
-            }
-
             throw SQLError.createSQLException("Conversion not supported for type " + type.getName(), MysqlErrorNumbers.SQL_STATE_ILLEGAL_ARGUMENT,
                     getExceptionInterceptor());
        }
--- a/src/main/java/com/mysql/cj/jdbc/integration/c3p0/MysqlConnectionTester.java
+++ b/src/main/java/com/mysql/cj/jdbc/integration/c3p0/MysqlConnectionTester.java
@@ -28,13 +28,12 @@
 
 package com.mysql.cj.jdbc.integration.c3p0;
 
+import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.sql.Connection;
 import java.sql.SQLException;
 import java.sql.Statement;
 
-import com.mchange.v2.c3p0.C3P0ProxyConnection;
-import com.mchange.v2.c3p0.QueryConnectionTester;
 import com.mysql.cj.exceptions.CJCommunicationsException;
 import com.mysql.cj.jdbc.JdbcConnection;
 import com.mysql.cj.jdbc.exceptions.CommunicationsException;
@@ -44,9 +43,12 @@
  * ConnectionTester for C3P0 connection pool that uses the more efficient COM_PING method of testing connection 'liveness' for MySQL, and 'sorts' exceptions
  * based on SQLState or class of 'CommunicationsException' for handling exceptions.
  */
-public final class MysqlConnectionTester implements QueryConnectionTester {
+public final class MysqlConnectionTester {
 
     private static final long serialVersionUID = 3256444690067896368L;
+
+    public static final int CONNECTION_IS_INVALID = -1;
+    public static final int CONNECTION_IS_OKAY = 0;
 
     private static final Object[] NO_ARGS_ARRAY = new Object[0];
 
@@ -60,7 +62,6 @@
         }
     }
 
-    @Override
     public int activeCheckConnection(Connection con) {
         try {
             if (this.pingMethod != null) {
@@ -69,9 +70,11 @@
                     // We've been passed an instance of a MySQL connection -- no need for reflection
                     ((JdbcConnection) con).ping();
                 } else {
-                    // Assume the connection is a C3P0 proxy
-                    C3P0ProxyConnection castCon = (C3P0ProxyConnection) con;
-                    castCon.rawConnectionOperation(this.pingMethod, C3P0ProxyConnection.RAW_CONNECTION, NO_ARGS_ARRAY);
+                    // Assume the connection is a C3P0 proxy, without requiring c3p0 at compile time.
+                    Field rawConnectionField = con.getClass().getField("RAW_CONNECTION");
+                    Method rawConnectionOperation = con.getClass().getMethod("rawConnectionOperation", Method.class, Object.class,
+                            Object[].class);
+                    rawConnectionOperation.invoke(con, this.pingMethod, rawConnectionField.get(null), NO_ARGS_ARRAY);
                 }
             } else {
                 Statement pingStatement = null;
@@ -91,7 +94,6 @@
         }
     }
 
-    @Override
     public int statusOnException(Connection arg0, Throwable throwable) {
         if (throwable instanceof CommunicationsException || throwable instanceof CJCommunicationsException) {
             return CONNECTION_IS_INVALID;
@@ -112,7 +114,6 @@
         return CONNECTION_IS_INVALID;
     }
 
-    @Override
     public int activeCheckConnection(Connection arg0, String arg1) {
         return CONNECTION_IS_OKAY;
     }
