diff --git a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpMethods.java b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpMethods.java
new file mode 100644
index 0000000..7961c16
--- /dev/null
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpMethods.java
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2025 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.http;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Utility functions related to HTTP methods.
+ */
+final class HttpMethods {
+
+    private static final Set<String> STANDARD_HTTP_METHODS = new HashSet<>(9);
+
+    static {
+        STANDARD_HTTP_METHODS.add("GET");
+        STANDARD_HTTP_METHODS.add("HEAD");
+        STANDARD_HTTP_METHODS.add("POST");
+        STANDARD_HTTP_METHODS.add("PUT");
+        STANDARD_HTTP_METHODS.add("DELETE");
+        STANDARD_HTTP_METHODS.add("CONNECT");
+        STANDARD_HTTP_METHODS.add("OPTIONS");
+        STANDARD_HTTP_METHODS.add("TRACE");
+        STANDARD_HTTP_METHODS.add("PATCH");
+
+    }
+
+    private HttpMethods() {
+    }
+
+    /**
+     * Checks if the given method is a well-known HTTP method defined in
+     * <a href="https://www.rfc-editor.org/rfc/rfc9110.html#name-methods">RFC 9110</a> and
+     * <a href="https://www.rfc-editor.org/rfc/rfc5789.html">RFC 5789</a>. This is
+     * case-sensitive and known methods are all uppercase.
+     * @param method HTTP method to check
+     * @return {@code true} if the given method is a standard HTTP method
+     */
+    static boolean isStandard(String method) {
+        return STANDARD_HTTP_METHODS.contains(method);
+    }
+
+}
diff --git a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpRequestTags.java b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpRequestTags.java
index cac0eca..0977f88 100644
--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpRequestTags.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpRequestTags.java
@@ -15,6 +15,7 @@
  */
 package io.micrometer.core.instrument.binder.http;
 
+import io.micrometer.core.lang.Nullable;
 import io.micrometer.core.annotation.Incubating;
 import io.micrometer.core.instrument.Tag;
 import io.micrometer.core.instrument.util.StringUtils;
@@ -46,8 +47,13 @@ public class HttpRequestTags {
      * @param request the request
      * @return the method tag whose value is a capitalized method (e.g. GET).
      */
-    public static Tag method(HttpServletRequest request) {
-        return (request != null) ? Tag.of("method", request.getMethod()) : METHOD_UNKNOWN;
+    public static Tag method(@Nullable HttpServletRequest request) {
+        if (request == null) {
+            return METHOD_UNKNOWN;
+        }
+
+        String method = request.getMethod();
+        return HttpMethods.isStandard(method) ? Tag.of("method", method) : METHOD_UNKNOWN;
     }
 
     /**
diff --git a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/HttpMethods.java b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/HttpMethods.java
new file mode 100644
index 0000000..82e9d28
--- /dev/null
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/HttpMethods.java
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2025 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.jersey.server;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Utility functions related to HTTP methods.
+ */
+final class HttpMethods {
+
+    private static final Set<String> STANDARD_HTTP_METHODS = new HashSet<>(9);
+
+    static {
+        STANDARD_HTTP_METHODS.add("GET");
+        STANDARD_HTTP_METHODS.add("HEAD");
+        STANDARD_HTTP_METHODS.add("POST");
+        STANDARD_HTTP_METHODS.add("PUT");
+        STANDARD_HTTP_METHODS.add("DELETE");
+        STANDARD_HTTP_METHODS.add("CONNECT");
+        STANDARD_HTTP_METHODS.add("OPTIONS");
+        STANDARD_HTTP_METHODS.add("TRACE");
+        STANDARD_HTTP_METHODS.add("PATCH");
+
+    }
+
+    private HttpMethods() {
+    }
+
+    /**
+     * Checks if the given method is a well-known HTTP method defined in
+     * <a href="https://www.rfc-editor.org/rfc/rfc9110.html#name-methods">RFC 9110</a> and
+     * <a href="https://www.rfc-editor.org/rfc/rfc5789.html">RFC 5789</a>. This is
+     * case-sensitive and known methods are all uppercase.
+     * @param method HTTP method to check
+     * @return {@code true} if the given method is a standard HTTP method
+     */
+    static boolean isStandard(String method) {
+        return STANDARD_HTTP_METHODS.contains(method);
+    }
+
+}
diff --git a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/JerseyTags.java b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/JerseyTags.java
index 1dc06c0..d8aecde 100644
--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/JerseyTags.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/JerseyTags.java
@@ -15,6 +15,7 @@
  */
 package io.micrometer.core.instrument.binder.jersey.server;
 
+import io.micrometer.core.lang.Nullable;
 import io.micrometer.core.instrument.Tag;
 import io.micrometer.core.instrument.binder.http.Outcome;
 import io.micrometer.core.instrument.util.StringUtils;
@@ -62,8 +63,13 @@ public final class JerseyTags {
      * @param request the container request
      * @return the method tag whose value is a capitalized method (e.g. GET).
      */
-    public static Tag method(ContainerRequest request) {
-        return (request != null) ? Tag.of("method", request.getMethod()) : METHOD_UNKNOWN;
+    public static Tag method(@Nullable ContainerRequest request) {
+        if (request == null) {
+            return METHOD_UNKNOWN;
+        }
+
+        String method = request.getMethod();
+        return HttpMethods.isStandard(method) ? Tag.of("method", method) : METHOD_UNKNOWN;
     }
 
     /**
diff --git a/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/jersey/server/JerseyTagsTests.java b/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/jersey/server/JerseyTagsTests.java
new file mode 100644
index 0000000..cf79f9f
--- /dev/null
+++ b/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/jersey/server/JerseyTagsTests.java
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2026 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.jersey.server;
+
+import io.micrometer.core.instrument.Tag;
+import org.glassfish.jersey.server.ContainerRequest;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for {@link JerseyTags}.
+ */
+class JerseyTagsTests {
+
+    @Test
+    void nullRequestShouldResultInUnknownMethod() {
+        assertThat(JerseyTags.method(null)).isEqualTo(Tag.of("method", "UNKNOWN"));
+    }
+
+    @Test
+    void unknownMethodShouldResultInUnknownTagValue() {
+        ContainerRequest request = mock(ContainerRequest.class);
+        when(request.getMethod()).thenReturn("WELL");
+        assertThat(JerseyTags.method(request)).isEqualTo(Tag.of("method", "UNKNOWN"));
+    }
+
+    @Test
+    void knownMethodShouldResultInTagValue() {
+        ContainerRequest request = mock(ContainerRequest.class);
+        when(request.getMethod()).thenReturn("GET");
+        assertThat(JerseyTags.method(request)).isEqualTo(Tag.of("method", "GET"));
+    }
+
+}
diff --git a/settings.gradle b/settings.gradle
index 126dc94..ab05283 100644
--- a/settings.gradle
+++ b/settings.gradle
@@ -7,24 +7,10 @@ pluginManagement {
     }
 }
 
-plugins {
-    id 'com.gradle.enterprise' version '3.11.1'
-    id 'io.spring.ge.conventions' version '0.0.11'
-}
-
 rootProject.name = 'micrometer'
 
 enableFeaturePreview("STABLE_CONFIGURATION_CACHE")
 
-buildCache {
-    remote(HttpBuildCache) {
-        url = 'https://ge.micrometer.io/cache/'
-    }
-}
-gradleEnterprise {
-    server = 'https://ge.micrometer.io'
-}
-
 include 'micrometer-core'
 
 ['core', 'boot2', 'boot2-reactive', 'spring-integration', 'hazelcast', 'hazelcast3', 'javalin', 'jersey3'].each { sample ->
diff --git a/build.gradle b/build.gradle
--- a/build.gradle
+++ b/build.gradle
@@ -1,11 +1,15 @@
 buildscript {
-    dependencyLocking {
-        lockAllConfigurations()
-    }
+    configurations.all {
+        resolutionStrategy.eachDependency { details ->
+            if (details.requested.group == 'org.ajoberstar.grgit') {
+                details.useVersion '4.1.1'
+            }
+        }
+    }

     repositories {
         mavenCentral()
         gradlePluginPortal()
     }

     dependencies {
