diff --git a/validate.go b/validate.go
index e64891c..5056bb9 100644
--- a/validate.go
+++ b/validate.go
@@ -306,9 +306,10 @@ func (ctx *ValidationContext) validateSignature(el *etree.Element, sig *types.Si
 	var ref *types.Reference
 
 	// Find the first reference which references the top-level element
-	for _, _ref := range signedInfo.References {
-		if _ref.URI == "" || _ref.URI[1:] == idAttr {
-			ref = &_ref
+	for i := range signedInfo.References {
+		if signedInfo.References[i].URI == "" || signedInfo.References[i].URI[1:] == idAttr {
+			ref = &signedInfo.References[i]
+			break
 		}
 	}
 
