--- a/lib/json5.js
+++ b/lib/json5.js
@@ -456,7 +456,13 @@

                     white();
                     next(':');
-                    object[key] = value();
+                    var val = value();
+                    Object.defineProperty(object, key, {
+                        value: val,
+                        writable: true,
+                        enumerable: true,
+                        configurable: true,
+                    });
                     white();
                     // If there's no comma after this pair, this needs to be
                     // the end of the object.
@@ -524,7 +530,12 @@
                     if (Object.prototype.hasOwnProperty.call(value, k)) {
                         v = walk(value, k);
                         if (v !== undefined) {
-                            value[k] = v;
+                            Object.defineProperty(value, k, {
+                                value: v,
+                                writable: true,
+                                enumerable: true,
+                                configurable: true,
+                            });
                         } else {
                             delete value[k];
                         }
