--- a/src/xregexp.js
+++ b/src/xregexp.js
@@ -33,9 +33,9 @@ const nativ = {
 // Storage for fixed/extended native methods
 const fixed = {};
 // Storage for regexes cached by `XRegExp.cache`
-let regexCache = {};
+let regexCache = Object.create(null);
 // Storage for pattern details cached by the `XRegExp` constructor
-let patternCache = {};
+let patternCache = Object.create(null);
 // Storage for regex syntax tokens added internally or by `XRegExp.addToken`
 const tokens = [];
 // Token scopes
@@ -772,10 +772,10 @@ XRegExp.cache = (pattern, flags) => {
 XRegExp.cache.flush = (cacheName) => {
     if (cacheName === 'patterns') {
         // Flush the pattern cache used by the `XRegExp` constructor
-        patternCache = {};
+        patternCache = Object.create(null);
     } else {
         // Flush the regex cache populated by `XRegExp.cache`
-        regexCache = {};
+        regexCache = Object.create(null);
     }
 };
 
--- a/lib/xregexp.js
+++ b/lib/xregexp.js
@@ -65,8 +65,8 @@ var nativ = {
 
 var fixed = {}; // Storage for regexes cached by `XRegExp.cache`
 
-var regexCache = {}; // Storage for pattern details cached by the `XRegExp` constructor
+var regexCache = Object.create(null); // Storage for pattern details cached by the `XRegExp` constructor
 
-var patternCache = {}; // Storage for regex syntax tokens added internally or by `XRegExp.addToken`
+var patternCache = Object.create(null); // Storage for regex syntax tokens added internally or by `XRegExp.addToken`
 
 var tokens = []; // Token scopes
@@ -845,10 +845,10 @@ XRegExp.cache = function (pattern, flags) {
 XRegExp.cache.flush = function (cacheName) {
   if (cacheName === 'patterns') {
     // Flush the pattern cache used by the `XRegExp` constructor
-    patternCache = {};
+    patternCache = Object.create(null);
   } else {
     // Flush the regex cache populated by `XRegExp.cache`
-    regexCache = {};
+    regexCache = Object.create(null);
   }
 };
 /**
