--- a/lib/path/evaluation.js
+++ b/lib/path/evaluation.js
@@ -346,7 +346,11 @@

       if (object.isIdentifier() && property.isIdentifier() && VALID_CALLEES.indexOf(object.node.name) >= 0 && INVALID_METHODS.indexOf(property.node.name) < 0) {
         context = global[object.node.name];
-        func = context[property.node.name];
+        const key = property.node.name;
+
+        if (Object.hasOwnProperty.call(context, key)) {
+          func = context[key];
+        }
       }

       if (object.isLiteral() && property.isIdentifier()) {
