--- a/lib/socket.ts
+++ b/lib/socket.ts
@@ -16,6 +16,8 @@ import type { ParsedUrlQuery } from "querystring";

 const debug = debugModule("socket.io:socket");

+function noop() {}
+
 export const RESERVED_EVENTS = new Set(<const>[
   "connect",
   "connect_error",
@@ -115,6 +117,9 @@ export class Socket extends EventEmitter {
     this.connected = true;
     this.disconnected = false;
     this.handshake = this.buildHandshake(auth);
+
+    // prevents crash when the socket receives an "error" event without listener
+    this.on("error", noop);
   }

   /**
@@ -416,12 +421,11 @@ export class Socket extends EventEmitter {
    * @private
    */
   _onerror(err): void {
-    if (this.listeners("error").length) {
-      super.emit("error", err);
-    } else {
-      console.error("Missing error handler on `socket`.");
-      console.error(err.stack);
-    }
+    // FIXME the meaning of the "error" event is overloaded:
+    //  - it can be sent by the client (`socket.emit("error")`)
+    //  - it can be emitted when the connection encounters an error (an invalid packet for example)
+    //  - it can be emitted when a packet is rejected in a middleware (`socket.use()`)
+    super.emit("error", err);
   }

   /**
