diff --git a/src/main/java/org/apache/maven/shared/utils/cli/shell/BourneShell.java b/src/main/java/org/apache/maven/shared/utils/cli/shell/BourneShell.java
index 8116707..e3af665 100644
--- a/src/main/java/org/apache/maven/shared/utils/cli/shell/BourneShell.java
+++ b/src/main/java/org/apache/maven/shared/utils/cli/shell/BourneShell.java
@@ -23,7 +23,6 @@ package org.apache.maven.shared.utils.cli.shell;
 import java.util.ArrayList;
 import java.util.List;
 import org.apache.maven.shared.utils.Os;
-import org.apache.maven.shared.utils.StringUtils;
 
 /**
  * @author Jason van Zyl
@@ -31,17 +30,16 @@ import org.apache.maven.shared.utils.StringUtils;
 public class BourneShell
     extends Shell
 {
-    private static final char[] BASH_QUOTING_TRIGGER_CHARS =
-        { ' ', '$', ';', '&', '|', '<', '>', '*', '?', '(', ')', '[', ']', '{', '}', '`' };
 
     /**
-     * Create instance of BournShell.
+     * Create instance of BourneShell.
      */
     public BourneShell()
     {
+        setUnconditionalQuoting( true );
         setShellCommand( "/bin/sh" );
         setArgumentQuoteDelimiter( '\'' );
-        setExecutableQuoteDelimiter( '\"' );
+        setExecutableQuoteDelimiter( '\'' );
         setSingleQuotedArgumentEscaped( true );
         setSingleQuotedExecutableEscaped( false );
         setQuotedExecutableEnabled( true );
@@ -57,7 +55,7 @@ public class BourneShell
             return super.getExecutable();
         }
 
-        return unifyQuotes( super.getExecutable() );
+        return quoteOneItem( super.getExecutable(), true );
     }
 
     /** {@inheritDoc} */
@@ -107,50 +105,34 @@ public class BourneShell
         }
 
         String dir = getWorkingDirectoryAsString();
-        StringBuilder sb = new StringBuilder();
-        sb.append( "cd " );
 
-        sb.append( unifyQuotes( dir ) );
-        sb.append( " && " );
-
-        return sb.toString();
-    }
-
-    /** {@inheritDoc} */
-    protected char[] getQuotingTriggerChars()
-    {
-        return BASH_QUOTING_TRIGGER_CHARS;
+        return "cd " + quoteOneItem( dir, false ) + " && ";
     }
 
     /**
      * <p>Unify quotes in a path for the Bourne Shell.</p>
      * <p/>
      * <pre>
-     * BourneShell.unifyQuotes(null)                       = null
-     * BourneShell.unifyQuotes("")                         = (empty)
-     * BourneShell.unifyQuotes("/test/quotedpath'abc")     = /test/quotedpath\'abc
-     * BourneShell.unifyQuotes("/test/quoted path'abc")    = "/test/quoted path'abc"
-     * BourneShell.unifyQuotes("/test/quotedpath\"abc")    = "/test/quotedpath\"abc"
-     * BourneShell.unifyQuotes("/test/quoted path\"abc")   = "/test/quoted path\"abc"
-     * BourneShell.unifyQuotes("/test/quotedpath\"'abc")   = "/test/quotedpath\"'abc"
-     * BourneShell.unifyQuotes("/test/quoted path\"'abc")  = "/test/quoted path\"'abc"
+     * BourneShell.quoteOneItem(null)                       = null
+     * BourneShell.quoteOneItem("")                         = ''
+     * BourneShell.quoteOneItem("/test/quotedpath'abc")     = '/test/quotedpath'"'"'abc'
+     * BourneShell.quoteOneItem("/test/quoted path'abc")    = '/test/quoted pat'"'"'habc'
+     * BourneShell.quoteOneItem("/test/quotedpath\"abc")    = '/test/quotedpath"abc'
+     * BourneShell.quoteOneItem("/test/quoted path\"abc")   = '/test/quoted path"abc'
+     * BourneShell.quoteOneItem("/test/quotedpath\"'abc")   = '/test/quotedpath"'"'"'abc'
+     * BourneShell.quoteOneItem("/test/quoted path\"'abc")  = '/test/quoted path"'"'"'abc'
      * </pre>
      *
      * @param path not null path.
      * @return the path unified correctly for the Bourne shell.
      */
-    private static String unifyQuotes( String path )
+    protected String quoteOneItem( String path, boolean isExecutable )
     {
         if ( path == null )
         {
             return null;
         }
 
-        if ( path.indexOf( ' ' ) == -1 && path.indexOf( '\'' ) != -1 && path.indexOf( '"' ) == -1 )
-        {
-            return StringUtils.escape( path );
-        }
-
-        return StringUtils.quoteAndEscape( path, '\"', BASH_QUOTING_TRIGGER_CHARS );
+        return "'" + path.replace( "'", "'\"'\"'" ) + "'";
     }
 }
diff --git a/src/main/java/org/apache/maven/shared/utils/cli/shell/CmdShell.java b/src/main/java/org/apache/maven/shared/utils/cli/shell/CmdShell.java
index 7c27ac1..04aa6de 100644
--- a/src/main/java/org/apache/maven/shared/utils/cli/shell/CmdShell.java
+++ b/src/main/java/org/apache/maven/shared/utils/cli/shell/CmdShell.java
@@ -23,9 +23,7 @@ import java.util.Arrays;
 import java.util.List;
 
 /**
- * <p>
- * Implementation to call the CMD Shell present on Windows NT, 2000 and XP
- * </p>
+ * Implementation to call the CMD Shell present on Windows NT, 2000, XP, 7, 8, and 10.
  *
  * @author <a href="mailto:carlos@apache.org">Carlos Sanchez</a>
  * 
diff --git a/src/main/java/org/apache/maven/shared/utils/cli/shell/CommandShell.java b/src/main/java/org/apache/maven/shared/utils/cli/shell/CommandShell.java
index dcee046..ba85633 100644
--- a/src/main/java/org/apache/maven/shared/utils/cli/shell/CommandShell.java
+++ b/src/main/java/org/apache/maven/shared/utils/cli/shell/CommandShell.java
@@ -21,13 +21,12 @@ package org.apache.maven.shared.utils.cli.shell;
 
 
 /**
- * <p>
  * Implementation to call the Command.com Shell present on Windows 95, 98 and Me
- * </p>
  *
  * @author <a href="mailto:carlos@apache.org">Carlos Sanchez</a>
- * 
+ * @deprecated Windows ME is long dead. Update to Windows 10 and use {@link CmdShell}.
  */
+@Deprecated
 public class CommandShell
     extends Shell
 {
diff --git a/src/main/java/org/apache/maven/shared/utils/cli/shell/Shell.java b/src/main/java/org/apache/maven/shared/utils/cli/shell/Shell.java
index 5809610..0268108 100644
--- a/src/main/java/org/apache/maven/shared/utils/cli/shell/Shell.java
+++ b/src/main/java/org/apache/maven/shared/utils/cli/shell/Shell.java
@@ -27,14 +27,13 @@ import java.util.List;
 import org.apache.maven.shared.utils.StringUtils;
 
 /**
- * <p>
  * Class that abstracts the Shell functionality,
- * with subclases for shells that behave particularly, like
+ * with subclasses for shells that behave particularly, like<p>
+ * 
  * <ul>
  * <li><code>command.com</code></li>
  * <li><code>cmd.exe</code></li>
  * </ul>
- * </p>
  *
  * @author <a href="mailto:carlos@apache.org">Carlos Sanchez</a>
  * 
@@ -50,6 +49,8 @@ public class Shell
 
     private boolean quotedArgumentsEnabled = true;
 
+    private boolean unconditionalQuoting = false;
+
     private String executable;
 
     private String workingDir;
@@ -65,9 +66,9 @@ public class Shell
     private char exeQuoteDelimiter = '\"';
 
     /**
-     * Set the command to execute the shell (eg. COMMAND.COM, /bin/bash,...)
+     * Set the command to execute the shell (e.g. COMMAND.COM, /bin/bash,...).
      *
-     * @param shellCommand The command
+     * @param shellCommand the command
      */
     void setShellCommand( String shellCommand )
     {
@@ -75,9 +76,9 @@ public class Shell
     }
 
     /**
-     * Get the command to execute the shell
+     * Get the command to execute the shell.
      *
-     * @return  The command
+     * @return the command
      */
     String getShellCommand()
     {
@@ -86,7 +87,7 @@ public class Shell
 
     /**
      * Set the shell arguments when calling a command line (not the executable arguments)
-     * (eg. /X /C for CMD.EXE)
+     * (e.g. /X /C for CMD.EXE).
      *
      * @param shellArgs the arguments to the shell
      */
@@ -99,26 +100,39 @@ public class Shell
     /**
      * Get the shell arguments
      *
-     * @return  The arguments
+     * @return the arguments
      */
     String[] getShellArgs()
     {
-        if ( ( shellArgs == null ) || shellArgs.isEmpty() )
+        if ( shellArgs.isEmpty() )
         {
             return null;
         }
         else
         {
-            return shellArgs.toArray( new String[shellArgs.size()] );
+            return shellArgs.toArray( new String[0] );
         }
     }
 
+    protected String quoteOneItem( String inputString, boolean isExecutable )
+    {
+        char[] escapeChars = getEscapeChars( isSingleQuotedExecutableEscaped(), isDoubleQuotedExecutableEscaped() );
+        return StringUtils.quoteAndEscape(
+                inputString,
+                isExecutable ? getExecutableQuoteDelimiter() : getArgumentQuoteDelimiter(),
+                escapeChars,
+                getQuotingTriggerChars(),
+                '\\',
+                unconditionalQuoting
+        );
+    }
+
     /**
      * Get the command line for the provided executable and arguments in this shell
      *
      * @param executableParameter executable that the shell has to call
      * @param argumentsParameter  arguments for the executable, not the shell
-     * @return List with one String object with executable and arguments quoted as needed
+     * @return list with one String object with executable and arguments quoted as needed
      */
     List<String> getCommandLine( String executableParameter, String... argumentsParameter )
     {
@@ -126,13 +140,13 @@ public class Shell
     }
 
     /**
-     * @param executableParameter Executable.
-     * @param argumentsParameter The arguments for the executable.
-     * @return The list on command line. 
+     * @param executableParameter Executable
+     * @param argumentsParameter the arguments for the executable
+     * @return the list on command line
      */
     List<String> getRawCommandLine( String executableParameter, String... argumentsParameter )
     {
-        List<String> commandLine = new ArrayList<String>();
+        List<String> commandLine = new ArrayList<String>();
         StringBuilder sb = new StringBuilder();
 
         if ( executableParameter != null )
@@ -145,15 +159,11 @@ public class Shell
 
             if ( isQuotedExecutableEnabled() )
             {
-                char[] escapeChars =
-                    getEscapeChars( isSingleQuotedExecutableEscaped(), isDoubleQuotedExecutableEscaped() );
-
-                sb.append( StringUtils.quoteAndEscape( getExecutable(), getExecutableQuoteDelimiter(), escapeChars,
-                                                       getQuotingTriggerChars(), '\\', false ) );
+                sb.append( quoteOneItem( executableParameter, true ) );
             }
             else
             {
-                sb.append( getExecutable() );
+                sb.append( executableParameter );
             }
         }
         for ( String argument : argumentsParameter )
@@ -165,10 +175,7 @@ public class Shell
 
             if ( isQuotedArgumentsEnabled() )
             {
-                char[] escapeChars = getEscapeChars( isSingleQuotedArgumentEscaped(), isDoubleQuotedArgumentEscaped() );
-
-                sb.append( StringUtils.quoteAndEscape( argument, getArgumentQuoteDelimiter(), escapeChars,
-                                                       getQuotingTriggerChars(), '\\', false ) );
+                sb.append( quoteOneItem( argument, false ) );
             }
             else
             {
@@ -211,7 +218,7 @@ public class Shell
     }
 
     /**
-     * @return false in all cases. 
+     * @return false in all cases 
      */
     protected boolean isDoubleQuotedArgumentEscaped()
     {
@@ -273,7 +280,7 @@ public class Shell
     public List<String> getShellCommandLine( String... arguments )
     {
 
-        List<String> commandLine = new ArrayList<String>();
+        List<String> commandLine = new ArrayList<String>();
 
         if ( getShellCommand() != null )
         {
@@ -285,7 +292,7 @@ public class Shell
             commandLine.addAll( getShellArgsList() );
         }
 
-        commandLine.addAll( getCommandLine( getExecutable(), arguments ) );
+        commandLine.addAll( getCommandLine( executable, arguments ) );
 
         return commandLine;
 
@@ -354,7 +361,8 @@ public class Shell
 
     /**
      * Sets execution directory.
-     * @param workingDirectory The working directory.
+     * 
+     * @param workingDirectory the working directory
      */
     public void setWorkingDirectory( File workingDirectory )
     {
@@ -365,7 +373,7 @@ public class Shell
     }
 
     /**
-     * @return The working directory.
+     * @return the working directory
      */
     public File getWorkingDirectory()
     {
@@ -398,4 +406,13 @@ public class Shell
         this.singleQuotedExecutableEscaped = singleQuotedExecutableEscaped;
     }
 
+    public boolean isUnconditionalQuoting()
+    {
+        return unconditionalQuoting;
+    }
+
+    public void setUnconditionalQuoting( boolean unconditionalQuoting )
+    {
+        this.unconditionalQuoting = unconditionalQuoting;
+    }
 }
