--- a/src/main/java/org/apache/commons/fileupload/FileUploadBase.java
+++ b/src/main/java/org/apache/commons/fileupload/FileUploadBase.java
@@ -145,12 +145,19 @@ public abstract class FileUploadBase {
      * The maximum length of a single header line that will be parsed
      * (1024 bytes).
      * @deprecated This constant is no longer used. As of commons-fileupload
-     *   1.2, the only applicable limit is the total size of a parts headers,
-     *   {@link MultipartStream#HEADER_PART_SIZE_MAX}.
+     *   1.6, the applicable limit is the total size of a single part's headers,
+     *   {@link #getPartHeaderSizeMax()} in bytes.
      */
     @Deprecated
     public static final int MAX_HEADER_SIZE = 1024;
 
+    /**
+     * Default per part header size limit in bytes.
+     *
+     * @since 1.6
+     */
+    public static final int DEFAULT_PART_HEADER_SIZE_MAX = 512;
+
     // ----------------------------------------------------------- Data members
 
     /**
@@ -171,6 +178,11 @@ public abstract class FileUploadBase {
      */
     private long fileCountMax = -1;
 
+    /**
+     * The maximum permitted size of the headers provided with a single part in bytes.
+     */
+    private int partHeaderSizeMax = DEFAULT_PART_HEADER_SIZE_MAX;
+
     /**
      * The content encoding to use when reading part headers.
      */
@@ -1042,6 +1054,7 @@ public abstract class FileUploadBase {
                         format("The boundary specified in the %s header is too long", CONTENT_TYPE), iae);
             }
             multi.setHeaderEncoding(charEncoding);
+            multi.setPartHeaderSizeMax(getPartHeaderSizeMax());
 
             skipPreamble = true;
             findNextItem();
@@ -1534,4 +1547,26 @@ public abstract class FileUploadBase {
         listener = pListener;
     }
 
+    /**
+     * Obtain the per part size limit for headers.
+     *
+     * @return The maximum size of the headers for a single part in bytes.
+     *
+     * @since 1.6
+     */
+    public int getPartHeaderSizeMax() {
+        return partHeaderSizeMax;
+    }
+
+    /**
+     * Sets the per part size limit for headers.
+     *
+     * @param partHeaderSizeMax The maximum size of the headers in bytes.
+     *
+     * @since 1.6
+     */
+    public void setPartHeaderSizeMax(final int partHeaderSizeMax) {
+        this.partHeaderSizeMax = partHeaderSizeMax;
+    }
+
 }
--- a/src/main/java/org/apache/commons/fileupload/MultipartStream.java
+++ b/src/main/java/org/apache/commons/fileupload/MultipartStream.java
@@ -25,6 +25,7 @@ import java.io.OutputStream;
 import java.io.UnsupportedEncodingException;
 
 import org.apache.commons.fileupload.FileUploadBase.FileUploadIOException;
+import org.apache.commons.fileupload.FileUploadBase.SizeLimitExceededException;
 import org.apache.commons.fileupload.util.Closeable;
 import org.apache.commons.fileupload.util.Streams;
 
@@ -173,7 +174,10 @@ public class MultipartStream {
     /**
      * The maximum length of <code>header-part</code> that will be
      * processed (10 kilobytes = 10240 bytes.).
+     *
+     * @deprecated Unused. Replaced by {@link #getPartHeaderSizeMax()}.
      */
+    @Deprecated
     public static final int HEADER_PART_SIZE_MAX = 10240;
 
     /**
@@ -266,6 +270,11 @@ public class MultipartStream {
      */
     private final ProgressNotifier notifier;
 
+    /**
+     * The maximum permitted size of the headers provided with a single part in bytes.
+     */
+    private int partHeaderSizeMax = FileUploadBase.DEFAULT_PART_HEADER_SIZE_MAX;
+
     // ----------------------------------------------------------- Constructors
 
     /**
@@ -412,6 +421,28 @@ public class MultipartStream {
         headerEncoding = encoding;
     }
 
+    /**
+     * Obtain the per part size limit for headers.
+     *
+     * @return The maximum size of the headers for a single part in bytes.
+     *
+     * @since 1.6
+     */
+    public int getPartHeaderSizeMax() {
+        return partHeaderSizeMax;
+    }
+
+    /**
+     * Sets the per part size limit for headers.
+     *
+     * @param partHeaderSizeMax The maximum size of the headers in bytes.
+     *
+     * @since 1.6
+     */
+    public void setPartHeaderSizeMax(final int partHeaderSizeMax) {
+        this.partHeaderSizeMax = partHeaderSizeMax;
+    }
+
     /**
      * Reads a byte from the <code>buffer</code>, and refills it as
      * necessary.
@@ -546,9 +577,6 @@ public class MultipartStream {
      * trailing <code>CRLF</code> marker. Parsing is left to the
      * application.
      *
-     * <p><strong>TODO</strong> allow limiting maximum header size to
-     * protect against abuse.
-     *
      * @return The <code>header-part</code> of the current encapsulation.
      *
      * @throws FileUploadIOException if the bytes read from the stream exceeded the size limits.
@@ -569,10 +597,12 @@ public class MultipartStream {
             } catch (IOException e) {
                 throw new MalformedStreamException("Stream ended unexpectedly");
             }
-            if (++size > HEADER_PART_SIZE_MAX) {
-                throw new MalformedStreamException(
+            size++;
+            if (getPartHeaderSizeMax() != -1 && size > getPartHeaderSizeMax()) {
+                throw new FileUploadIOException(new SizeLimitExceededException(
                         format("Header section has more than %s bytes (maybe it is not properly terminated)",
-                               Integer.valueOf(HEADER_PART_SIZE_MAX)));
+                                Integer.valueOf(getPartHeaderSizeMax())),
+                                size, getPartHeaderSizeMax()));
             }
             if (b == HEADER_SEPARATOR[i]) {
                 i++;
