--- a/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/system/ApplicationTemp.java
+++ b/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/system/ApplicationTemp.java
@@ -18,13 +18,17 @@ package org.springframework.boot.system;

 import java.io.File;
 import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.nio.file.FileSystem;
 import java.nio.file.Files;
+import java.nio.file.LinkOption;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.nio.file.attribute.FileAttribute;
+import java.nio.file.attribute.PosixFileAttributes;
 import java.nio.file.attribute.PosixFilePermission;
 import java.nio.file.attribute.PosixFilePermissions;
+import java.nio.file.attribute.UserPrincipal;
 import java.security.MessageDigest;
 import java.util.EnumSet;
 import java.util.HexFormat;
@@ -54,6 +58,8 @@ public class ApplicationTemp {

 	private final Lock pathLock = new ReentrantLock();

+	private final UserPrincipal directoryOwner;
+
 	private volatile Path path;

 	/**
@@ -69,6 +75,22 @@ public class ApplicationTemp {
 	 */
 	public ApplicationTemp(Class<?> sourceClass) {
 		this.sourceClass = sourceClass;
+		this.directoryOwner = directoryOwner();
+	}
+
+	private static UserPrincipal directoryOwner() {
+		try {
+			Path tempFile = Files.createTempFile("application-temp", "-owner");
+			UserPrincipal owner = Files.getOwner(tempFile);
+			Files.delete(tempFile);
+			return owner;
+		}
+		catch (UnsupportedOperationException ex) {
+			return null;
+		}
+		catch (IOException ex) {
+			throw new UncheckedIOException(ex);
+		}
 	}

 	@Override
@@ -111,8 +133,30 @@ public class ApplicationTemp {

 	private Path createDirectory(Path path) {
 		try {
-			if (!Files.exists(path)) {
-				Files.createDirectory(path, getFileAttributes(path.getFileSystem(), DIRECTORY_PERMISSIONS));
+			FileSystem fileSystem = path.getFileSystem();
+			if (!Files.exists(path, LinkOption.NOFOLLOW_LINKS)) {
+				Files.createDirectory(path, asFileAttributes(fileSystem, DIRECTORY_PERMISSIONS));
+			}
+			else {
+				if (supportsPosixView(fileSystem)) {
+					PosixFileAttributes attributes = Files.readAttributes(path, PosixFileAttributes.class,
+							LinkOption.NOFOLLOW_LINKS);
+					Assert.state(attributes.isDirectory(),
+							() -> "'" + path + "' already exists but it is not a directory");
+					Assert.state(DIRECTORY_PERMISSIONS.equals(attributes.permissions()), () -> "Existing directory '"
+							+ path + "' does not have the permissions " + DIRECTORY_PERMISSIONS);
+					assertDirectoryOwnership(attributes.owner(), path);
+				}
+				else {
+					Assert.state(Files.isDirectory(path, LinkOption.NOFOLLOW_LINKS),
+							() -> "'" + path + "' already exists but it is not a directory");
+					try {
+						assertDirectoryOwnership(Files.getOwner(path, LinkOption.NOFOLLOW_LINKS), path);
+					}
+					catch (UnsupportedOperationException ex) {
+						// Ownership check not supported. Continue.
+					}
+				}
 			}
 			return path;
 		}
@@ -121,13 +165,22 @@ public class ApplicationTemp {
 		}
 	}

-	private FileAttribute<?>[] getFileAttributes(FileSystem fileSystem, EnumSet<PosixFilePermission> ownerReadWrite) {
-		if (!fileSystem.supportedFileAttributeViews().contains("posix")) {
+	private void assertDirectoryOwnership(UserPrincipal owner, Path path) {
+		Assert.state((this.directoryOwner == null) || this.directoryOwner.equals(owner),
+				() -> "Existing directory '" + path + "' is not owned by " + this.directoryOwner.getName());
+	}
+
+	private FileAttribute<?>[] asFileAttributes(FileSystem fileSystem, EnumSet<PosixFilePermission> ownerReadWrite) {
+		if (!supportsPosixView(fileSystem)) {
 			return NO_FILE_ATTRIBUTES;
 		}
 		return new FileAttribute<?>[] { PosixFilePermissions.asFileAttribute(ownerReadWrite) };
 	}

+	private boolean supportsPosixView(FileSystem fileSystem) {
+		return fileSystem.supportedFileAttributeViews().contains("posix");
+	}
+
 	private Path getTempDirectory() {
 		String property = System.getProperty("java.io.tmpdir");
 		Assert.state(StringUtils.hasLength(property), "No 'java.io.tmpdir' property set");
--- a/spring-boot-project/spring-boot/src/test/java/org/springframework/boot/system/ApplicationTempTests.java
+++ b/spring-boot-project/spring-boot/src/test/java/org/springframework/boot/system/ApplicationTempTests.java
@@ -23,15 +23,20 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.attribute.PosixFileAttributeView;
 import java.nio.file.attribute.PosixFilePermission;
+import java.util.EnumSet;
 import java.util.Set;

 import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assumptions;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.DisabledOnOs;
+import org.junit.jupiter.api.condition.OS;

 import org.springframework.util.FileSystemUtils;

 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalStateException;

 /**
  * Tests for {@link ApplicationTemp}.
@@ -85,6 +90,35 @@ class ApplicationTempTests {
 		}
 	}

+	@Test
+	@DisabledOnOs(OS.WINDOWS)
+	void whenDirectoryExistsWithWrongPermissionsGetDirThrows() throws IOException {
+		ApplicationTemp temp = new ApplicationTemp();
+		Path path = temp.getDir().toPath();
+		Files.getFileAttributeView(path, PosixFileAttributeView.class)
+			.setPermissions(EnumSet.allOf(PosixFilePermission.class));
+		assertThatIllegalStateException().isThrownBy(new ApplicationTemp()::getDir)
+			.withMessageContaining("does not have the permissions");
+		FileSystemUtils.deleteRecursively(path);
+	}
+
+	@Test
+	void whenSymlinkExistsInDirectoryLocationGetDirThrows() throws IOException {
+		ApplicationTemp temp = new ApplicationTemp();
+		Path path = temp.getDir().toPath();
+		FileSystemUtils.deleteRecursively(path);
+		Path linkTarget = Files.createTempDirectory("application-test-tests");
+		try {
+			Files.createSymbolicLink(path, linkTarget);
+		}
+		catch (Exception ex) {
+			Assumptions.abort("Symlink creation not supported");
+		}
+		assertThatIllegalStateException().isThrownBy(new ApplicationTemp()::getDir)
+			.withMessageContaining("already exists but it is not a directory");
+		FileSystemUtils.deleteRecursively(path);
+	}
+
 	private void assertDirectoryPermissions(Path path) throws IOException {
 		Set<PosixFilePermission> permissions = Files.getFileAttributeView(path, PosixFileAttributeView.class)
 			.readAttributes()

--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/security/reactive/EndpointRequest.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/security/reactive/EndpointRequest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012-2024 the original author or authors.
+ * Copyright 2012-2025 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -185,6 +185,7 @@
 		}

 		private PathPatternParserServerWebExchangeMatcher getDelegateMatcher(String path) {
+			Assert.notNull(path, "'path' must not be null");
 			return new PathPatternParserServerWebExchangeMatcher(path + "/**");
 		}

@@ -304,12 +305,19 @@
 			List<ServerWebExchangeMatcher> delegateMatchers = getDelegateMatchers(paths);
 			if (this.includeLinks && StringUtils.hasText(endpoints.getBasePath())) {
 				delegateMatchers.add(new LinksServerWebExchangeMatcher());
 			}
+			if (delegateMatchers.isEmpty()) {
+				return EMPTY_MATCHER;
+			}
 			return new OrServerWebExchangeMatcher(delegateMatchers);
 		}

 		private Stream<String> streamPaths(List<Object> source, PathMappedEndpoints endpoints) {
-			return source.stream().filter(Objects::nonNull).map(this::getEndpointId).map(endpoints::getPath);
+			return source.stream()
+				.filter(Objects::nonNull)
+				.map(this::getEndpointId)
+				.map(endpoints::getPath)
+				.filter(Objects::nonNull);
 		}

 		@Override
--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/security/servlet/EndpointRequest.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/security/servlet/EndpointRequest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012-2024 the original author or authors.
+ * Copyright 2012-2025 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -321,13 +321,20 @@
 			List<RequestMatcher> delegateMatchers = getDelegateMatchers(requestMatcherFactory, matcherProvider, paths);
 			String basePath = endpoints.getBasePath();
 			if (this.includeLinks && StringUtils.hasText(basePath)) {
 				delegateMatchers.addAll(getLinksMatchers(requestMatcherFactory, matcherProvider, basePath));
 			}
+			if (delegateMatchers.isEmpty()) {
+				return EMPTY_MATCHER;
+			}
 			return new OrRequestMatcher(delegateMatchers);
 		}

 		private Stream<String> streamPaths(List<Object> source, PathMappedEndpoints endpoints) {
-			return source.stream().filter(Objects::nonNull).map(this::getEndpointId).map(endpoints::getPath);
+			return source.stream()
+				.filter(Objects::nonNull)
+				.map(this::getEndpointId)
+				.map(endpoints::getPath)
+				.filter(Objects::nonNull);
 		}

 		@Override
@@ -429,6 +436,7 @@
 		RequestMatcher antPath(RequestMatcherProvider matcherProvider, String... parts) {
 			StringBuilder pattern = new StringBuilder();
 			for (String part : parts) {
+				Assert.notNull(part, "'part' must not be null");
 				pattern.append(part);
 			}
 			return matcherProvider.getRequestMatcher(pattern.toString());
