--- a/deps/dicer/lib/HeaderParser.js
+++ b/deps/dicer/lib/HeaderParser.js
@@ -21,7 +21,7 @@
   this.maxHeaderPairs = getLimit(cfg, 'maxHeaderPairs', 2000)
   this.maxHeaderSize = getLimit(cfg, 'maxHeaderSize', 80 * 1024)
   this.buffer = ''
-  this.header = {}
+  this.header = Object.create(null)
   this.finished = false
   this.ss = new StreamSearch(B_DCRLF)
   this.ss.on('info', function (isMatch, data, start, end) {
@@ -47,7 +47,7 @@
 HeaderParser.prototype.reset = function () {
   this.finished = false
   this.buffer = ''
-  this.header = {}
+  this.header = Object.create(null)
   this.ss.reset()
 }
 
@@ -55,7 +55,7 @@
   if (this.buffer) { this._parseHeader() }
   this.ss.matches = this.ss.maxMatches
   const header = this.header
-  this.header = {}
+  this.header = Object.create(null)
   this.buffer = ''
   this.finished = true
   this.nread = this.npairs = 0
