--- a/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveConstants.java
+++ b/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveConstants.java
@@ -26,6 +26,7 @@
      * The type of compression.
      */
     public enum COMPRESSION_TYPE {
+        UNKNOWN(-1),
         ZLIB(0),
         BZLIB(1),
         LZO(2);
@@ -37,7 +38,7 @@
                 }
             }
 
-            return null;
+            return UNKNOWN;
         }
 
         final int code;
--- a/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveUtil.java
+++ b/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveUtil.java
@@ -84,6 +84,9 @@
      */
     static String decode(final ZipEncoding encoding, final byte[] b, final int offset, final int len)
         throws IOException {
+        if (offset > offset + len) {
+            throw new IOException("Invalid offset/length combination");
+        }
         return encoding.decode(Arrays.copyOfRange(b, offset, offset + len));
     }
 
@@ -104,6 +107,9 @@
      * @return Whether the buffer contains a tape segment header.
      */
     public static final boolean verify(final byte[] buffer) {
+        if (buffer == null) {
+            return false;
+        }
         // verify magic. for now only accept NFS_MAGIC.
         final int magic = convert32(buffer, 24);
 
--- a/src/main/java/org/apache/commons/compress/archivers/dump/TapeInputStream.java
+++ b/src/main/java/org/apache/commons/compress/archivers/dump/TapeInputStream.java
@@ -311,6 +311,9 @@
                 + " records found, must be at least 1");
         }
         blockSize = RECORD_SIZE * recsPerBlock;
+        if (blockSize < 1) {
+            throw new IOException("Block size cannot be less than or equal to 0: " + blockSize);
+        }
 
         // save first block in case we need it again
         final byte[] oldBuffer = blockBuffer;
