--- a/esm/index.js
+++ b/esm/index.js
@@ -26,7 +26,7 @@
     const k = ke[y];
     const value = output[k];
     if (value instanceof Primitive) {
-      const tmp = input[value];
+      const tmp = input[+value];
       if (typeof tmp === object && !parsed.has(tmp)) {
         parsed.add(tmp);
         output[k] = ignore;
--- a/cjs/index.js
+++ b/cjs/index.js
@@ -27,7 +27,7 @@
     const k = ke[y];
     const value = output[k];
     if (value instanceof Primitive) {
-      const tmp = input[value];
+      const tmp = input[+value];
       if (typeof tmp === object && !parsed.has(tmp)) {
         parsed.add(tmp);
         output[k] = ignore;
--- a/index.js
+++ b/index.js
@@ -38,7 +38,7 @@
         var k = ke[y];
         var value = output[k];
         if (value instanceof Primitive) {
-          var tmp = input[value];
+          var tmp = input[+value];
           if (_typeof(tmp) === object && !parsed.has(tmp)) {
             parsed.add(tmp);
             output[k] = ignore;
