--- a/java/org.apache.derby.engine/org/apache/derby/impl/jdbc/authentication/LDAPAuthenticationSchemeImpl.java
+++ b/java/org.apache.derby.engine/org/apache/derby/impl/jdbc/authentication/LDAPAuthenticationSchemeImpl.java
@@ -183,7 +183,56 @@
 		}
 
 		throw getLoginSQLException(e);
-	}
+	}
+
+	/**
+	 * Given an LDAP search string, returns the string with certain characters
+	 * escaped according to RFC 2254 guidelines.
+	 * Cribbed from org.apache.catalina.realm.JNDIRealm.
+	 *
+	 * The character mapping is as follows:
+	 *     char -&gt;  Replacement
+	 *    ---------------------------
+	 *     *  -&gt; \2a
+	 *     (  -&gt; \28
+	 *     )  -&gt; \29
+	 *     \  -&gt; \5c
+	 *     \0 -&gt; \00
+	 *
+	 * @param inString string to escape according to RFC 2254 guidelines
+	 *
+	 * @return String the escaped/encoded result
+	 */
+	protected String doFilterEscaping(String inString) {
+		if (inString == null) {
+			return null;
+		}
+		StringBuilder buf = new StringBuilder(inString.length());
+		for (int i = 0; i < inString.length(); i++) {
+			char c = inString.charAt(i);
+			switch (c) {
+				case '\\':
+					buf.append("\\5c");
+					break;
+				case '*':
+					buf.append("\\2a");
+					break;
+				case '(':
+					buf.append("\\28");
+					break;
+				case ')':
+					buf.append("\\29");
+					break;
+				case '\0':
+					buf.append("\\00");
+					break;
+				default:
+					buf.append(c);
+					break;
+			}
+		}
+		return buf.toString();
+	}
 
 	
 
@@ -394,6 +443,10 @@
 	private String getDNFromUID(String uid)
 		throws javax.naming.NamingException
 	{
+
+		// Escape the uid as a defense against LDAP injection. See DERBY-7147.
+		uid = doFilterEscaping(uid);
+
 		//
 		// We bind to the LDAP server here
 		// Note that this bind might be anonymous (if anonymous searches
