From 26a39a4dd2cb76a75ea71dd5244293f5cf0a2bb1 Mon Sep 17 00:00:00 2001
Date: Tue, 4 Nov 2025 11:46:54 +0530
Subject: [PATCH] patch

---
 src/cairo-bentley-ottmann-rectangular.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/cairo-bentley-ottmann-rectangular.c b/src/cairo-bentley-ottmann-rectangular.c
index 65f95d797..b54d7ab23 100644
--- a/src/cairo-bentley-ottmann-rectangular.c
+++ b/src/cairo-bentley-ottmann-rectangular.c
@@ -848,6 +848,9 @@ _cairo_bentley_ottmann_tessellate_boxes (const cairo_boxes_t *in,
 		rectangles[j].left.dir = -1;
 	    }
 
+		if (unlikely (rectangles[j].left.x == INT32_MAX)) rectangles[j].left.x = INT32_MAX-1;
+	    if (unlikely (rectangles[j].right.x == INT32_MAX)) rectangles[j].right.x = INT32_MAX-1;
+
 	    rectangles[j].left.right = NULL;
 	    rectangles[j].right.right = NULL;
 
-- 
2.43.0

