--- a/pypdf/generic/_data_structures.py
+++ b/pypdf/generic/_data_structures.py
@@ -1395,6 +1395,8 @@ class ContentStream(DecodedStreamObject):
         settings = DictionaryObject()
         while True:
             tok = read_non_whitespace(stream)
+            if not tok:
+                raise PdfStreamError("Unexpected end of stream.")
             stream.seek(-1, 1)
             if tok == b"I":
                 # "ID" - begin of image data
--- a/tests/generic/test_data_structures.py
+++ b/tests/generic/test_data_structures.py
@@ -10,7 +10,7 @@ from unittest import mock
 import pytest

 from pypdf import PdfReader, PdfWriter
-from pypdf.errors import LimitReachedError, PdfReadError
+from pypdf.errors import LimitReachedError, PdfReadError, PdfStreamError
 from pypdf.generic import (
     ArrayObject,
     ByteStringObject,
@@ -458,3 +458,14 @@ def test_content_stream__parse_content_stream__limits() -> None:
             match=r"^Read stream length of 240 exceeds maximum allowed length of 128\.$"
     ):
         content_stream._parse_content_stream(stream)
+
+
+def test_content_stream__read_inline_image__end_of_stream() -> None:
+    # Broken content stream, for example due to filter errors.
+    # Specific example:
+    #   Error -3 while decompressing data: invalid distance too far back
+    #   b'q 0.1 0 0 0.1 0 0 cm\n/R7 gs\n0 g\nq 4.8 0 0 -135.6 2155.08 7150.48 cm\nBI\n/IM true\n/W001'
+    content_stream = ContentStream(stream=None, pdf=None)
+
+    with pytest.raises(expected_exception=PdfStreamError, match=r"^Unexpected end of stream\.$"):
+        content_stream._read_inline_image(BytesIO(b"\n/IM true\n/W001"))
