diff --git a/lib/patch/parse.js b/lib/patch/parse.js
index b65d5c6..d7a456f 100644
--- a/lib/patch/parse.js
+++ b/lib/patch/parse.js
@@ -64,17 +64,18 @@ function parsePatch(uniDiff) {
 
 
   function parseFileHeader(index) {
-    var fileHeader = /^(---|\+\+\+)\s+(.*)$/.exec(diffstr[i]);
+    var fileHeaderMatch = /^(---|\+\+\+)\s+/.exec(diffstr[i]);
 
-    if (fileHeader) {
-      var keyPrefix = fileHeader[1] === '---' ? 'old' : 'new';
-      var data = fileHeader[2].split('\t', 2);
+    if (fileHeaderMatch) {
+      var prefix = fileHeaderMatch[1];
+      var data = diffstr[i].substring(3).trim().split('\t', 2);
       var fileName = data[0].replace(/\\\\/g, '\\');
 
-      if (/^".*"$/.test(fileName)) {
+      if (fileName.startsWith('"') && fileName.endsWith('"')) {
         fileName = fileName.substr(1, fileName.length - 2);
       }
 
+      var keyPrefix = prefix === '---' ? 'old' : 'new';
       index[keyPrefix + 'FileName'] = fileName;
       index[keyPrefix + 'Header'] = (data[1] || '').trim();
       i++;
-- 
2.43.0

