--- a/index.js
+++ b/index.js
@@ -68,20 +68,26 @@ function pathToRegexp(path, keys, options) {
   path = path.replace(
     /\\.|(\/)?(\.)?:(\w+)(\(.*?\))?(\*)?(\?)?|[.*]|\/\(/g,
     function (match, slash, format, key, capture, star, optional, offset) {
-      pos = offset + match.length;
-
       if (match[0] === '\\') {
         backtrack += match;
+        pos += 2;
         return match;
       }

       if (match === '.') {
         backtrack += '\\.';
         extraOffset += 1;
+        pos += 1;
         return '\\.';
       }

-      backtrack = slash || format ? '' : path.slice(pos, offset);
+      if (slash || format) {
+        backtrack = '';
+      } else {
+        backtrack += path.slice(pos, offset);
+      }
+
+      pos = offset + match.length;

       if (match === '*') {
         extraOffset += 3;

--- a/index.js
+++ b/index.js
@@ -84,6 +84,7 @@
       backtrack = slash || format ? '' : path.slice(pos, offset);

       if (match === '*') {
+        backtrack = '';
         extraOffset += 3;
         return '(.*)';
       }
@@ -113,6 +114,7 @@
         + ')'
         + optional;

+      backtrack = '';
       extraOffset += result.length - match.length;

       return result;
