--- a/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveConstants.java
+++ b/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveConstants.java
@@ -67,7 +67,8 @@ public final class DumpArchiveConstants {
     /**
      * The type of compression.
      */
     public enum COMPRESSION_TYPE {
-        ZLIB(0),
+        UNKNOWN(-1),
+        ZLIB(0),
         BZLIB(1),
         LZO(2);

@@ -85,7 +86,7 @@ public final class DumpArchiveConstants {
                 }
             }

-            return null;
+            return UNKNOWN;
         }
     }
 }
--- a/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveUtil.java
+++ b/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveUtil.java
@@ -55,6 +55,9 @@ class DumpArchiveUtil {
      *
      * @param buffer
      */
     public static final boolean verify(final byte[] buffer) {
+        if (buffer == null) {
+            return false;
+        }
         // verify magic. for now only accept NFS_MAGIC.
         final int magic = convert32(buffer, 24);

@@ -114,6 +117,9 @@ class DumpArchiveUtil {
     /**
      * Decodes a byte array to a string.
      */
     static String decode(final ZipEncoding encoding, final byte[] b, final int offset, final int len)
         throws IOException {
+        if (offset > offset + len) {
+            throw new IOException("Invalid offset/length combination");
+        }
         return encoding.decode(Arrays.copyOfRange(b, offset, offset + len));
     }
 }
--- a/src/main/java/org/apache/commons/compress/archivers/dump/TapeInputStream.java
+++ b/src/main/java/org/apache/commons/compress/archivers/dump/TapeInputStream.java
@@ -74,6 +74,9 @@ class TapeInputStream extends FilterInputStream {
             throw new IOException("Block with " + recsPerBlock
                 + " records found, must be at least 1");
         }
         blockSize = RECORD_SIZE * recsPerBlock;
+        if (blockSize < 1) {
+            throw new IOException("Block size cannot be less than or equal to 0: " + blockSize);
+        }

         // save first block in case we need it again
         final byte[] oldBuffer = blockBuffer;

--- a/src/main/java/org/apache/commons/compress/harmony/pack200/NewAttributeBands.java
+++ b/src/main/java/org/apache/commons/compress/harmony/pack200/NewAttributeBands.java
@@ -279,7 +279,7 @@
         stream.mark(2);
         stream.read(); // '('
         char next = (char) stream.read();
-        if (next == ')') {
+        if (next == ')' || next == (char) -1) {
             stream.reset();
             return null;
         }
@@ -878,7 +878,11 @@
         final StringBuffer sb = new StringBuffer();
         int foundBracket = -1;
         while (foundBracket != 0) {
-            final char c = (char) stream.read();
+            int read = stream.read();
+            if (read == -1) {
+                break;
+            }
+            final char c = (char) read;
             if (c == ']') {
                 foundBracket++;
             }
--- a/src/main/java/org/apache/commons/compress/harmony/unpack200/NewAttributeBands.java
+++ b/src/main/java/org/apache/commons/compress/harmony/unpack200/NewAttributeBands.java
@@ -297,7 +297,7 @@
         stream.mark(2);
         stream.read(); // '('
         char next = (char) stream.read();
-        if (next == ')') {
+        if (next == ')' || next == (char) -1) {
             stream.reset();
             return null;
         }
@@ -872,7 +872,11 @@
         final StringBuffer sb = new StringBuffer();
         int foundBracket = -1;
         while (foundBracket != 0) {
-            final char c = (char) stream.read();
+            int read = stream.read();
+            if (read == -1) {
+                break;
+            }
+            final char c = (char) read;
             if (c == ']') {
                 foundBracket++;
             }
@@ -920,7 +924,11 @@
         final StringBuffer sb = new StringBuffer();
         int foundBracket = -1;
         while (foundBracket != 0) {
-            final char c = (char) stream.read();
+            int read = stream.read();
+            if (read == -1) {
+                break;
+            }
+            final char c = (char) read;
             if (c == ']') {
                 foundBracket++;
             }
