--- a/lib/websocket/driver/draft75.js
+++ b/lib/websocket/driver/draft75.js
@@ -44,6 +44,7 @@

         case 1:
           this._length = (octet & 0x7F) + 128 * this._length;
+          if (this._length > this._maxLength) return this.close();

           if (this._closing && this._length === 0) {
             return this.close();
--- a/lib/websocket/driver/hybi.js
+++ b/lib/websocket/driver/hybi.js
@@ -452,7 +452,12 @@
       if (error) return this._fail('extension_error', error.message);

       var payload = message.data;
-      if (message.opcode === this.OPCODES.text) payload = this._encode(payload);
+
+      if (payload.length > this._maxLength)
+        return this._fail('too_large', 'WebSocket frame length too large');
+
+      if (message.opcode === this.OPCODES.text)
+        payload = this._encode(payload);

       if (payload === null)
         return this._fail('encoding_error', 'Could not decode a text frame as UTF-8');
