--- a/debug.go
+++ b/debug.go
@@ -139,6 +139,6 @@
 		copy(tv[:], v)
 		return tv, err
 	default:
-		panic(fmt.Sprintf("unknown type, %s, %T", typ.String(), val))
+		return nil, fmt.Errorf("unable to decode unknown type, %s, %T", typ.String(), val)
 	}
 }
--- a/header.go
+++ b/header.go
@@ -151,7 +151,7 @@
 		err = tv.decode(r)
 		v = tv
 	default:
-		panic(fmt.Sprintf("unknown value type %d", raw.Type))
+		return nil, fmt.Errorf("unable to decode header of unknown value type %d", raw.Type)
 	}
 
 	// Error could be EOF, let caller deal with it
--- a/header_test.go
+++ b/header_test.go
@@ -1,6 +1,7 @@
 package eventstream
 
 import (
+	"bytes"
 	"reflect"
 	"testing"
 	"time"
@@ -64,3 +65,16 @@
 		t.Errorf("expect %v headers, got %v", e, a)
 	}
 }
+
+func TestHeaders_UnknownType(t *testing.T) {
+	buf := bytes.NewBuffer(nil)
+	buf.WriteByte(3)       // header name length
+	buf.WriteString("foo") // header name
+	buf.WriteByte(0xFF)    // unknown value type
+
+	_, err := decodeHeaders(buf)
+	// This shouldn't panic and should return an error
+	if err == nil {
+		t.Fatal("expect error for unknown header value type, got none")
+	}
+}
