--- a/index.js
+++ b/index.js
@@ -37,7 +37,7 @@ function fresh(req, res) {
   if (cc && cc.indexOf('no-cache') !== -1) return false;  
 
   // parse if-none-match
-  if (noneMatch) noneMatch = noneMatch.split(/ *, */);
+  if (noneMatch) noneMatch = parseTokenList(noneMatch);
 
   // if-none-match
   if (noneMatch) {
@@ -55,3 +55,40 @@ function fresh(req, res) {
 
   return !! (etagMatches && notModified);
 }
+
+/**
+ * Parse a HTTP token list.
+ *
+ * @param {string} str
+ * @return {Array}
+ * @api private
+ */
+
+function parseTokenList(str) {
+  var end = 0;
+  var list = [];
+  var start = 0;
+
+  // gather tokens
+  for (var i = 0, len = str.length; i < len; i++) {
+    switch (str.charCodeAt(i)) {
+      case 0x20: /*   */
+        if (start === end) {
+          start = end = i + 1;
+        }
+        break;
+      case 0x2c: /* , */
+        list.push(str.substring(start, end));
+        start = end = i + 1;
+        break;
+      default:
+        end = i + 1;
+        break;
+    }
+  }
+
+  // final token
+  list.push(str.substring(start, end));
+
+  return list;
+}
