--- a/core/src/main/java/org/bouncycastle/crypto/modes/G3413CTRBlockCipher.java
+++ b/core/src/main/java/org/bouncycastle/crypto/modes/G3413CTRBlockCipher.java
@@ -13,8 +13,6 @@
 public class G3413CTRBlockCipher
     extends StreamBlockCipher
 {
-
-
     private final int s;
     private byte[] CTR;
     private byte[] IV;
@@ -24,7 +22,6 @@
     private int byteCount = 0;
     private boolean initialized;
 
-
     /**
      * Basic constructor.
      *
@@ -76,7 +73,6 @@
         CipherParameters params)
         throws IllegalArgumentException
     {
-
         if (params instanceof ParametersWithIV)
         {
             ParametersWithIV ivParam = (ParametersWithIV)params;
@@ -95,7 +91,7 @@
             {
                 CTR[i] = 0;
             }
-
+    
             // if null it's an IV changed only.
             if (ivParam.getParameters() != null)
             {
@@ -184,27 +180,33 @@
         if (byteCount == s)
         {
             byteCount = 0;
-            generateCRT();
+            generateCTR();
         }
 
         return rv;
 
     }
 
-    private void generateCRT()
+    private void generateCTR()
     {
-        CTR[CTR.length - 1]++;
+        int start = CTR.length - 1;
+        while (++CTR[start] == 0)
+        {
+            start--;
+            if (start == IV.length - 1)
+            {
+                throw new IllegalStateException("attempt to process too many blocks");
+            }
+        }
     }
 
 
     private byte[] generateBuf()
     {
-
         byte[] encryptedCTR = new byte[CTR.length];
         cipher.processBlock(CTR, 0, encryptedCTR, 0);
 
         return GOST3413CipherUtil.MSB(encryptedCTR, s);
-
     }
 
 
--- a/core/src/test/java/org/bouncycastle/crypto/test/GOST3412Test.java
+++ b/core/src/test/java/org/bouncycastle/crypto/test/GOST3412Test.java
@@ -1,5 +1,6 @@
 package org.bouncycastle.crypto.test;
 
+import org.bouncycastle.crypto.StreamBlockCipher;
 import org.bouncycastle.crypto.engines.GOST3412_2015Engine;
 import org.bouncycastle.crypto.modes.G3413CBCBlockCipher;
 import org.bouncycastle.crypto.modes.G3413CFBBlockCipher;
@@ -7,6 +8,7 @@
 import org.bouncycastle.crypto.modes.G3413OFBBlockCipher;
 import org.bouncycastle.crypto.params.KeyParameter;
 import org.bouncycastle.crypto.params.ParametersWithIV;
+import org.bouncycastle.util.Arrays;
 import org.bouncycastle.util.encoders.Hex;
 import org.bouncycastle.util.test.SimpleTest;
 
@@ -87,8 +89,64 @@
     {
         super.performTest();
 
+        ctrTest();
 //        cfbTest();
 //        ofbTest();
+    }
+
+    private void ctrTest()
+        throws Exception
+    {
+        StreamBlockCipher sb = new G3413CTRBlockCipher(new GOST3412_2015Engine(), 128);
+
+        sb.init(true, new ParametersWithIV(new KeyParameter(Hex.decode("8899aabbccddeeff0011223344556677fedcba98765432100123456789abcdef")),
+            Hex.decode("0001020304050607")));
+
+        byte[] block = Hex.decode("000102030405060708090a0b0c0d0e0f");
+        byte[] output = new byte[16];
+        byte[] last = new byte[16];
+
+        sb.processBytes(block, 0, block.length, last, 0);
+
+        for (int i = 1; i < 255; i++)
+        {
+            sb.processBytes(block, 0, block.length, output, 0);
+            if (Arrays.areEqual(last, output))
+            {
+                fail("cipher text repeats 1");
+            }
+        }
+
+        sb.processBytes(block, 0, block.length, output, 0);
+        if (Arrays.areEqual(last, output))
+        {
+            fail("cipher text repeats 2");
+        }
+
+        sb = new G3413CTRBlockCipher(new GOST3412_2015Engine(), 128);
+
+        sb.init(true, new ParametersWithIV(new KeyParameter(Hex.decode("8899aabbccddeeff0011223344556677fedcba98765432100123456789abcdef")),
+            Hex.decode("0001020304050607")));
+
+        sb.processBytes(block, 0, block.length, last, 0);
+
+        for (int i = 1; i != ((1 << 15) - 1); i++)
+        {
+            sb.processBytes(block, 0, block.length, output, 0);
+            if (Arrays.areEqual(last, output))
+            {
+                fail("cipher text repeats 3");
+            }
+            byte[] tmp = last;
+            last = output;
+            output = tmp;
+        }
+
+        sb.processBytes(block, 0, block.length, output, 0);
+        if (Arrays.areEqual(last, output))
+        {
+            fail("cipher text repeats");
+        }
     }
 
     public static void main(
