--- a/src/types.ts
+++ b/src/types.ts
@@ -31,5 +31,6 @@
 
 export interface AbortSignal {
+	aborted: boolean;
 	addEventListener: (name: string, listener: (...args: any[]) => any) => void;
 	removeEventListener: (
 		name: string,
--- a/src/index.ts
+++ b/src/index.ts
@@ -15,19 +15,29 @@
 ): Promise<EventListenerParameters<Emitter, Event>> {
 	return new Promise((resolve, reject) => {
 		function cleanup() {
-			signal?.removeEventListener('abort', cleanup);
+			signal?.removeEventListener('abort', onAbort);
 			emitter.removeListener(name, onEvent);
 			emitter.removeListener('error', onError);
 		}
 		function onEvent(...args: any[]) {
 			cleanup();
 			resolve(args as EventListenerParameters<Emitter, Event>);
 		}
 		function onError(err: Error) {
 			cleanup();
 			reject(err);
 		}
-		signal?.addEventListener('abort', cleanup);
+		function onAbort() {
+			cleanup();
+			const err = new Error('The operation was aborted');
+			err.name = 'AbortError';
+			reject(err);
+		}
+		if (signal?.aborted) {
+			onAbort();
+			return;
+		}
+		signal?.addEventListener('abort', onAbort);
 		emitter.on(name, onEvent);
 		emitter.on('error', onError);
 	});
