--- a/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/RabbitConnectionFactoryBean.java
+++ b/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/RabbitConnectionFactoryBean.java
@@ -28,6 +28,7 @@ import java.security.SecureRandom;
 import java.security.UnrecoverableKeyException;
 import java.security.cert.CertificateException;
 import java.util.Arrays;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Properties;
 import java.util.concurrent.ExecutorService;
@@ -162,6 +163,9 @@ public class RabbitConnectionFactoryBean extends AbstractFactoryBean<ConnectionF
 
 	private String trustStoreAlgorithm = SUN_X509;
 
+	@Nullable
+	private URI uri;
+
 	public RabbitConnectionFactoryBean() {
 		this.connectionFactory.setAutomaticRecoveryEnabled(false);
 	}
@@ -485,12 +489,7 @@ public class RabbitConnectionFactoryBean extends AbstractFactoryBean<ConnectionF
 	 * @see com.rabbitmq.client.ConnectionFactory#setUri(java.net.URI)
 	 */
 	public void setUri(URI uri) {
-		try {
-			this.connectionFactory.setUri(uri);
-		}
-		catch (URISyntaxException | NoSuchAlgorithmException | KeyManagementException e) {
-			throw new IllegalArgumentException("Unable to set uri", e);
-		}
+		this.uri = uri;
 	}
 
 	/**
@@ -498,12 +497,7 @@ public class RabbitConnectionFactoryBean extends AbstractFactoryBean<ConnectionF
 	 * @see com.rabbitmq.client.ConnectionFactory#setUri(java.lang.String)
 	 */
 	public void setUri(String uriString) {
-		try {
-			this.connectionFactory.setUri(uriString);
-		}
-		catch (URISyntaxException | NoSuchAlgorithmException | KeyManagementException e) {
-			throw new IllegalArgumentException("Unable to set uri", e);
-		}
+		setUri(URI.create(uriString));
 	}
 
 	/**
@@ -755,9 +749,20 @@ public class RabbitConnectionFactoryBean extends AbstractFactoryBean<ConnectionF
 
 	@Override
 	protected ConnectionFactory createInstance() {
+		if (this.uri != null && !this.useSSL) {
+			this.useSSL = this.uri.getScheme().toLowerCase(Locale.ROOT).equals("amqps");
+		}
 		if (this.useSSL) {
 			setUpSSL();
 		}
+		if (this.uri != null) {
+			try {
+				this.connectionFactory.setUri(this.uri);
+			}
+			catch (URISyntaxException | NoSuchAlgorithmException | KeyManagementException e) {
+				throw new IllegalArgumentException("Unable to set uri", e);
+			}
+		}
 		return this.connectionFactory;
 	}
 
--- a/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/log4j2/AmqpAppender.java
+++ b/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/log4j2/AmqpAppender.java
@@ -22,6 +22,7 @@ import java.net.URI;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
 import java.util.Calendar;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Timer;
@@ -641,6 +642,10 @@ public class AmqpAppender extends AbstractAppender {
 					// overrides all preceding items when set
 					.acceptIfNotNull(this.uri, factoryBean::setUri);
 
+			if (this.uri != null && !this.useSsl) {
+				this.useSsl = this.uri.getScheme().toLowerCase(Locale.ROOT).equals("amqps");
+			}
+
 			if (this.useSsl) {
 				factoryBean.setUseSSL(true);
 				factoryBean.setEnableHostnameVerification(this.verifyHostname);
--- a/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/logback/AmqpAppender.java
+++ b/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/logback/AmqpAppender.java
@@ -19,6 +19,7 @@ package org.springframework.amqp.rabbit.logback;
 import java.io.UnsupportedEncodingException;
 import java.net.URI;
 import java.util.Calendar;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
@@ -739,6 +740,10 @@ public class AmqpAppender extends AppenderBase<ILoggingEvent> {
 				// overrides all preceding items when set
 				.acceptIfNotNull(this.uri, factoryBean::setUri);
 
+		if (this.uri != null && !this.useSsl) {
+			this.useSsl = this.uri.getScheme().toLowerCase(Locale.ROOT).equals("amqps");
+		}
+
 		if (this.useSsl) {
 			factoryBean.setUseSSL(true);
 			factoryBean.setEnableHostnameVerification(this.verifyHostname);
--- a/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/connection/SSLConnectionTests.java
+++ b/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/connection/SSLConnectionTests.java
@@ -24,6 +24,7 @@ import javax.net.ssl.SSLContext;
 import com.rabbitmq.client.Channel;
 import com.rabbitmq.client.Connection;
 import com.rabbitmq.client.ConnectionFactory;
+import com.rabbitmq.client.SslContextFactory;
 import com.rabbitmq.client.impl.CredentialsProvider;
 import com.rabbitmq.client.impl.CredentialsRefreshService;
 import org.apache.commons.logging.Log;
@@ -274,5 +275,17 @@ public class SSLConnectionTests {
 				.isSameAs(service);
 	}
 
+	@Test
+	public void sslUriEnablesSslAsWell() throws Exception {
+		RabbitConnectionFactoryBean fb = new RabbitConnectionFactoryBean();
+		fb.setUri("amqps://guest:guest@localhost:5671");
+		fb.afterPropertiesSet();
+		ConnectionFactory connectionFactory = fb.getObject();
+		assertThat(connectionFactory.isSSL()).isTrue();
+		SslContextFactory sslContextFactory =
+				(SslContextFactory) TestUtils.getPropertyValue(connectionFactory, "sslContextFactory");
+		SSLContext sslContext = sslContextFactory.create("test");
+		assertThat(sslContext.getProtocol()).isEqualTo("TLSv1.2");
+	}
 
 }
