--- a/java/org/apache/tomcat/util/net/SocketWrapperBase.java
+++ b/java/org/apache/tomcat/util/net/SocketWrapperBase.java
@@ -88,6 +88,8 @@
     protected int remotePort = -1;
     protected volatile ServletConnection servletConnection = null;
 
+    protected String sniHostName = null;
+
     /**
      * Used to record the first IOException that occurs during non-blocking
      * read/writes that can't be usefully propagated up the stack since there is
@@ -207,6 +209,20 @@
     public String getNegotiatedProtocol() { return negotiatedProtocol; }
     public void setNegotiatedProtocol(String negotiatedProtocol) {
         this.negotiatedProtocol = negotiatedProtocol;
+    }
+
+    /**
+     * @return the sniHostName
+     */
+    public String getSniHostName() {
+        return this.sniHostName;
+    }
+
+    /**
+     * @param sniHostName the SNI host name to set
+     */
+    public void setSniHostName(String sniHostName) {
+        this.sniHostName = sniHostName;
     }
 
     /**

--- a/java/org/apache/tomcat/util/net/AbstractEndpoint.java
+++ b/java/org/apache/tomcat/util/net/AbstractEndpoint.java
@@ -219,6 +219,17 @@
 
     // ----------------------------------------------------------------- Properties
 
+    private boolean strictSni = true;
+
+    public boolean getStrictSni() {
+        return strictSni;
+    }
+
+    public void setStrictSni(boolean strictSni) {
+        this.strictSni = strictSni;
+    }
+
+
     private String defaultSSLHostConfigName = SSLHostConfig.DEFAULT_SSL_HOST_NAME;
     /**
      * @return The host name for the default SSL configuration for this endpoint
@@ -505,6 +516,22 @@
             throw new IllegalStateException();
         }
         return result;
+    }
+
+
+    /**
+     * Check if two host names share the same SSLHostConfig.
+     *
+     * @param sniHostName the host name from SNI, null if SNI is not in use
+     * @param protocolHostName the host name from the protocol
+     * @return true if SNI is not checked, if the SNI host name matches the protocol host name,
+     *    if both host names use the same SSLHostConfig configuration, if there is no SNI and the
+     *    protocol host name uses the default SSLHostConfig configuration, and false otherwise
+     */
+    public boolean checkSni(String sniHostName, String protocolHostName) {
+        return (!strictSni || !isSSLEnabled()
+                || (sniHostName != null && sniHostName.equalsIgnoreCase(protocolHostName))
+                || getSSLHostConfig(sniHostName) == getSSLHostConfig(protocolHostName));
     }
 
 

--- a/java/org/apache/tomcat/util/net/SecureNioChannel.java
+++ b/java/org/apache/tomcat/util/net/SecureNioChannel.java
@@ -275,6 +275,7 @@
         switch (extractor.getResult()) {
         case COMPLETE:
             hostName = extractor.getSNIValue();
+            socketWrapper.setSniHostName(hostName);
             clientRequestedApplicationProtocols =
                     extractor.getClientRequestedApplicationProtocols();
             //$FALL-THROUGH$ to set the client requested ciphers

--- a/java/org/apache/tomcat/util/net/SecureNio2Channel.java
+++ b/java/org/apache/tomcat/util/net/SecureNio2Channel.java
@@ -398,6 +398,7 @@
         switch (extractor.getResult()) {
         case COMPLETE:
             hostName = extractor.getSNIValue();
+            socketWrapper.setSniHostName(hostName);
             clientRequestedApplicationProtocols =
                     extractor.getClientRequestedApplicationProtocols();
             //$FALL-THROUGH$ to set the client requested ciphers

--- a/java/org/apache/coyote/http11/AbstractHttp11Protocol.java
+++ b/java/org/apache/coyote/http11/AbstractHttp11Protocol.java
@@ -781,6 +781,10 @@
     }
 
 
+    public boolean checkSni(String sniHostName, String protocolHostName) {
+        return getEndpoint().checkSni(sniHostName, protocolHostName);
+    }
+
     // ------------------------------------------------------------- Common code
 
     @Override

--- a/java/org/apache/coyote/http11/Http11Processor.java
+++ b/java/org/apache/coyote/http11/Http11Processor.java
@@ -784,6 +784,11 @@
         // Validate host name and extract port if present
         parseHost(hostValueMB);
 
+        // Match host name with SNI if required
+        if (!protocol.checkSni(socketWrapper.getSniHostName(), request.serverName().toString())) {
+            badRequest("http11processor.request.sni");
+        }
+
         if (!getErrorState().isIoAllowed()) {
             getAdapter().log(request, response, 0);
         }

--- a/java/org/apache/coyote/http11/LocalStrings.properties
+++ b/java/org/apache/coyote/http11/LocalStrings.properties
@@ -34,6 +34,7 @@
 http11processor.request.nonNumericContentLength=The request contained a content-length header with a non-numeric value
 http11processor.request.prepare=Error preparing request
 http11processor.request.process=Error processing request
+http11processor.request.sni=The host header does not match the SNI host
 http11processor.response.finish=Error finishing response
 http11processor.response.invalidHeader=The HTTP response header [{0}] with value [{1}] has been removed from the response because it is invalid
 http11processor.sendfile.error=Error sending data using sendfile. May be caused by invalid request attributes for start/end points

--- a/java/org/apache/coyote/http2/Http2UpgradeHandler.java
+++ b/java/org/apache/coyote/http2/Http2UpgradeHandler.java
@@ -1905,6 +1905,11 @@
         }
     }
 
+
+    String getSniHostName() {
+        return socketWrapper.getSniHostName();
+    }
+
     protected class PingManager {
 
         protected boolean initiateDisabled = false;

--- a/java/org/apache/coyote/http2/Stream.java
+++ b/java/org/apache/coyote/http2/Stream.java
@@ -492,6 +492,11 @@
         } else {
             coyoteRequest.serverName().setString(value);
         }
+        // Match host name with SNI if required
+        if (!handler.getProtocol().getHttp11Protocol().checkSni(handler.getSniHostName(), coyoteRequest.serverName().getString())) {
+            throw new HpackException(sm.getString("stream.host.sni", getConnectionId(), getIdAsString(), value,
+                    handler.getSniHostName()));
+        }
     }
 
 

--- a/java/org/apache/coyote/http2/LocalStrings.properties
+++ b/java/org/apache/coyote/http2/LocalStrings.properties
@@ -100,6 +100,7 @@
 stream.header.unexpectedPseudoHeader=Connection [{0}], Stream [{1}], Pseudo header [{2}] received after a regular header
 stream.header.unknownPseudoHeader=Connection [{0}], Stream [{1}], Unknown pseudo header [{2}] received
 stream.host.inconsistent=Connection [{0}], Stream [{1}], The header host header [{2}] is inconsistent with previously provided values for host [{3}] and/or port [{4}]
+stream.host.sni=Connection [{0}], Stream [{1}], The host header [{2}] does not match the SNI host [{3}]
 stream.inputBuffer.copy=Copying [{0}] bytes from inBuffer to outBuffer
 stream.inputBuffer.dispatch=Data added to inBuffer when read interest is registered. Triggering a read dispatch
 stream.inputBuffer.empty=The Stream input buffer is empty. Waiting for more data

--- a/test/org/apache/catalina/startup/SimpleHttpClient.java
+++ b/test/org/apache/catalina/startup/SimpleHttpClient.java
@@ -193,20 +193,28 @@
         return redirectUri;
     }
 
-    public void connect(int connectTimeout, int soTimeout)
-           throws UnknownHostException, IOException {
+    public void connect(Socket socket, int connectTimeout, int soTimeout, boolean connect) throws UnknownHostException, IOException {
         SocketAddress addr = new InetSocketAddress("localhost", port);
-        socket = new Socket();
+        this.socket = socket;
         socket.setSoTimeout(soTimeout);
-        socket.connect(addr,connectTimeout);
+        if (connect) {
+            socket.connect(addr, connectTimeout);
+        }
         OutputStream os = createOutputStream(socket);
         writer = new OutputStreamWriter(os, requestBodyEncoding);
         InputStream is = socket.getInputStream();
         Reader r = new InputStreamReader(is, responseBodyEncoding);
         reader = new BufferedReader(r);
     }
-    public void connect() throws UnknownHostException, IOException {
-        connect(10000, 10000);
+
+    public void connect(int connectTimeout, int soTimeout) throws UnknownHostException, IOException {
+        connect(new Socket(), 10000, 10000, true);
+    }
+
+    public void connect(Socket socket) throws UnknownHostException, IOException {
+        connect(socket, 10000, 10000, false);
+    }
+
     }
 
     protected OutputStream createOutputStream(Socket socket) throws IOException {

--- a/test/org/apache/tomcat/util/net/TestSsl.java
+++ b/test/org/apache/tomcat/util/net/TestSsl.java
@@ -33,7 +33,10 @@
 import javax.net.SocketFactory;
 import javax.net.ssl.HandshakeCompletedEvent;
 import javax.net.ssl.HandshakeCompletedListener;
+import javax.net.ssl.SNIHostName;
+import javax.net.ssl.SNIServerName;
 import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLParameters;
 import javax.net.ssl.SSLSocket;
 import javax.net.ssl.SSLSocketFactory;
 
@@ -54,10 +57,13 @@
 import org.apache.catalina.connector.Connector;
 import org.apache.catalina.core.AprLifecycleListener;
 import org.apache.catalina.core.StandardServer;
+import org.apache.catalina.startup.SimpleHttpClient;
 import org.apache.catalina.startup.TesterServlet;
 import org.apache.catalina.startup.Tomcat;
 import org.apache.catalina.startup.TomcatBaseTest;
 import org.apache.tomcat.util.buf.ByteChunk;
+import org.apache.tomcat.util.net.SSLHostConfigCertificate.Type;
+import org.apache.tomcat.util.net.TesterSupport.ClientSSLSocketFactory;
 import org.apache.tomcat.websocket.server.WsContextListener;
 
 /**
@@ -263,6 +269,160 @@
         Assert.assertTrue("Checking no client issuer has been requested",
                 TesterSupport.getLastClientAuthRequestedIssuerCount() == 0);
     }
+
+    @Test
+    public void testSni() throws Exception {
+        System.setProperty("jsse.enableSNIExtension", "true");
+        ClientSSLSocketFactory clientSSLSocketFactory = TesterSupport.configureClientSsl();
+        Tomcat tomcat = getTomcatInstance();
+        tomcat.getConnector().setProperty("strictSni", "true");
+
+        File appDir = new File(getBuildDirectory(), "webapps/examples");
+        Context ctxt  = tomcat.addWebapp(null, "/examples", appDir.getAbsolutePath());
+        ctxt.addApplicationListener(WsContextListener.class.getName());
+
+        TesterSupport.initSsl(tomcat);
+
+        // Add another config for localhost
+        SSLHostConfig sslHostConfig = new SSLHostConfig();
+        SSLHostConfigCertificate certificate = new SSLHostConfigCertificate(sslHostConfig, Type.UNDEFINED);
+        sslHostConfig.addCertificate(certificate);
+        certificate.setCertificateKeystoreFile(new File(TesterSupport.LOCALHOST_RSA_JKS).getAbsolutePath());
+        certificate.setCertificateKeystorePassword(TesterSupport.JKS_PASS);
+        sslHostConfig.setHostName("localhost");
+        tomcat.getConnector().addSslHostConfig(sslHostConfig);
+
+        // Add another config for foobar
+        sslHostConfig = new SSLHostConfig();
+        certificate = new SSLHostConfigCertificate(sslHostConfig, Type.UNDEFINED);
+        sslHostConfig.addCertificate(certificate);
+        certificate.setCertificateKeystoreFile(new File(TesterSupport.LOCALHOST_RSA_JKS).getAbsolutePath());
+        certificate.setCertificateKeystorePassword(TesterSupport.JKS_PASS);
+        sslHostConfig.setHostName("foobar");
+        tomcat.getConnector().addSslHostConfig(sslHostConfig);
+
+        TesterSupport.configureSSLImplementation(tomcat, sslImplementationName);
+
+        tomcat.start();
+
+        // Send SNI and it matches
+        SSLSocket sslSocket = (SSLSocket) clientSSLSocketFactory.createSocket("localhost", getPort());
+        SNIHostName serverName = new SNIHostName("localhost");
+        List<SNIServerName> serverNames = new ArrayList<>(1);
+        serverNames.add(serverName);
+        SSLParameters params = sslSocket.getSSLParameters();
+        params.setServerNames(serverNames);
+        sslSocket.setSSLParameters(params);
+
+        Client client = new Client();
+        client.setPort(getPort());
+
+        // @formatter:off
+        client.setRequest(new String[] {
+                "GET /examples/servlets/servlet/HelloWorldExample HTTP/1.1" + SimpleHttpClient.CRLF +
+                    "Host: localhost" + SimpleHttpClient.CRLF +
+                    "Connection: Close" + SimpleHttpClient.CRLF +
+                    SimpleHttpClient.CRLF
+                });
+        // @formatter:on
+        client.connect(sslSocket);
+        client.processRequest(true);
+
+        Assert.assertEquals(HttpServletResponse.SC_OK, client.getStatusCode());
+        Assert.assertTrue(client.getResponseBody().contains("<a href=\"../helloworld.html\">"));
+        client.disconnect();
+        client.reset();
+
+        // Send SNI and it does not match
+        sslSocket = (SSLSocket) clientSSLSocketFactory.createSocket("localhost", getPort());
+        params = sslSocket.getSSLParameters();
+        params.setServerNames(serverNames);
+        sslSocket.setSSLParameters(params);
+
+        // @formatter:off
+        client.setRequest(new String[] {
+                "GET /examples/servlets/servlet/HelloWorldExample HTTP/1.1" + SimpleHttpClient.CRLF +
+                    "Host: foobar" + SimpleHttpClient.CRLF +
+                    "Connection: Close" + SimpleHttpClient.CRLF +
+                    SimpleHttpClient.CRLF
+                });
+        // @formatter:on
+        client.connect(sslSocket);
+        client.processRequest(true);
+
+        Assert.assertEquals(HttpServletResponse.SC_BAD_REQUEST, client.getStatusCode());
+        client.disconnect();
+        client.reset();
+
+        // Send SNI and it does not match, but this goes to the default host which is the same one
+        tomcat.getConnector().setProperty("defaultSSLHostConfigName", "localhost");
+        sslSocket = (SSLSocket) clientSSLSocketFactory.createSocket("localhost", getPort());
+        params = sslSocket.getSSLParameters();
+        params.setServerNames(serverNames);
+        sslSocket.setSSLParameters(params);
+
+        // @formatter:off
+        client.setRequest(new String[] {
+                "GET /examples/servlets/servlet/HelloWorldExample HTTP/1.1" + SimpleHttpClient.CRLF +
+                    "Host: something" + SimpleHttpClient.CRLF +
+                    "Connection: Close" + SimpleHttpClient.CRLF +
+                    SimpleHttpClient.CRLF
+                });
+        // @formatter:on
+        client.connect(sslSocket);
+        client.processRequest(true);
+
+        Assert.assertEquals(HttpServletResponse.SC_OK, client.getStatusCode());
+        client.disconnect();
+        client.reset();
+        tomcat.getConnector().setProperty("defaultSSLHostConfigName", "_default_");
+
+        tomcat.getConnector().setProperty("strictSni", "false");
+
+        // SNI is not verified
+        sslSocket = (SSLSocket) clientSSLSocketFactory.createSocket("localhost", getPort());
+        params = sslSocket.getSSLParameters();
+        params.setServerNames(serverNames);
+        sslSocket.setSSLParameters(params);
+
+        // @formatter:off
+        client.setRequest(new String[] {
+                "GET /examples/servlets/servlet/HelloWorldExample HTTP/1.1" + SimpleHttpClient.CRLF +
+                    "Host: foobar" + SimpleHttpClient.CRLF +
+                    "Connection: Close" + SimpleHttpClient.CRLF +
+                    SimpleHttpClient.CRLF
+                });
+        // @formatter:on
+        client.connect(sslSocket);
+        client.processRequest(true);
+
+        Assert.assertEquals(HttpServletResponse.SC_OK, client.getStatusCode());
+        client.disconnect();
+        client.reset();
+
+        tomcat.getConnector().setProperty("strictSni", "true");
+
+        // No SNI but this is the default config
+        tomcat.getConnector().setProperty("defaultSSLHostConfigName", "localhost");
+        Assert.assertEquals(HttpServletResponse.SC_OK,
+                getUrl("https://localhost:" + getPort() + "/examples/servlets/servlet/HelloWorldExample", new ByteChunk(), null));
+
+        // No SNI and this is not the default config
+        tomcat.getConnector().setProperty("defaultSSLHostConfigName", "_default_");
+        Assert.assertEquals(HttpServletResponse.SC_BAD_REQUEST,
+                getUrl("https://localhost:" + getPort() + "/examples/servlets/servlet/HelloWorldExample", new ByteChunk(), null));
+
+    }
+
+    private static final class Client extends SimpleHttpClient {
+
+        @Override
+        public boolean isResponseBodyOK() {
+            return true;
+        }
+
+    }
+
 
     @Test
     public void testClientInitiatedRenegotiation() throws Exception {

--- a/webapps/docs/config/http.xml
+++ b/webapps/docs/config/http.xml
@@ -267,6 +267,15 @@
       The default value is <code>false</code>.</p>
     </attribute>
 
+    <attribute name="strictSni" required="false">
+      <p>Set this attribute to <code>true</code> to verify that the
+      <code>SSLHostConfig</code> configuration associated with the SNI host name
+      is the same as the <code>SSLHostConfig</code> configuration associated
+      with the HTTP protocol virtual host name in use. Non matching requests
+      will be rejected.
+      The default value is <code>true</code>.</p>
+    </attribute>
+
     <attribute name="URIEncoding" required="false">
       <p>This specifies the character encoding used to decode the URI bytes,
       after %xx decoding the URL. The default value is <code>UTF-8</code>.</p>
