From b2536a5807052d0d6929b1f9067c7b12656619fb Mon Sep 17 00:00:00 2001
Date: Fri, 23 Jan 2026 14:27:35 +0530
Subject: [PATCH] patch

---
 dist/utils/getFilenameFromUrl.js | 31 +++++++++++++++++++++++++------
 1 file changed, 25 insertions(+), 6 deletions(-)

diff --git a/dist/utils/getFilenameFromUrl.js b/dist/utils/getFilenameFromUrl.js
index ae42d4f..70f8c21 100644
--- a/dist/utils/getFilenameFromUrl.js
+++ b/dist/utils/getFilenameFromUrl.js
 
 const memoizedParse = (0, _mem.default)(_url.parse);
 
+// Regex to detect path traversal attempts (e.g., "..", "..\\", "../")
+const UP_PATH_REGEXP = /(?:^|[\\/])\.\.(?:[\\/]|$)/;
+
 function getFilenameFromUrl(context, url) {
   const {
     options
@@ -48,16 +51,32 @@ function getFilenameFromUrl(context, url) {
       continue;
     }
 
-    if (urlObject.pathname && urlObject.pathname.startsWith(publicPathObject.pathname)) {
-      filename = outputPath; // Strip the `pathname` property from the `publicPath` option from the start of requested url
-      // `/complex/foo.js` => `foo.js`
+    if (urlObject.pathname) {
+      // Decode pathname for proper security validation
+      const pathname = _querystring.default.unescape(urlObject.pathname);
+      const publicPathPathname = _querystring.default.unescape(publicPathObject.pathname);
 
-      const pathname = urlObject.pathname.substr(publicPathObject.pathname.length);
+      if (!pathname.startsWith(publicPathPathname)) {
+        // eslint-disable-next-line no-continue
+        continue;
+      }
 
-      if (pathname) {
-        filename = _path.default.join(outputPath, _querystring.default.unescape(pathname));
+      // Null byte(s) - CVE-2024-29180 security fix
+      if (pathname.includes("\0")) {
+        return;
       }
 
+      // ".." is malicious - CVE-2024-29180 security fix
+      if (UP_PATH_REGEXP.test(_path.default.normalize(`./${pathname}`))) {
+        return;
+      }
+
+      // Strip the `pathname` property from the `publicPath` option from the start of requested url
+      // `/complex/foo.js` => `foo.js`
+      // and add outputPath
+      // `foo.js` => `/home/user/my-project/dist/foo.js`
+      filename = _path.default.join(outputPath, pathname.slice(publicPathPathname.length));
+
       let fsStats;
 
       try {
-- 
2.43.0

