--- a/package.json
+++ b/package.json
@@ -78,13 +78,13 @@
     ]
   },
   "dependencies": {
-    "@trysound/sax": "0.2.0",
     "commander": "^7.2.0",
     "css-select": "^5.1.0",
     "css-tree": "^2.3.1",
     "css-what": "^6.1.0",
     "csso": "^5.0.5",
-    "picocolors": "^1.0.0"
+    "picocolors": "^1.0.0",
+    "sax": "^1.5.0"
   },
   "devDependencies": {
     "@rollup/plugin-commonjs": "^22.0.2",
@@ -92,6 +92,7 @@
     "@types/css-tree": "^2.3.4",
     "@types/csso": "^5.0.4",
     "@types/jest": "^29.5.5",
+    "@types/sax": "^1.2.7",
     "del": "^6.0.0",
     "eslint": "^8.55.0",
     "jest": "^29.5.5",
--- a/lib/parser.js
+++ b/lib/parser.js
@@ -13,8 +13,7 @@
  * @typedef {import('./types').XastChild} XastChild
  */
 
-// @ts-ignore sax will be replaced with something else later
-const SAX = require('@trysound/sax');
+const SAX = require('sax');
 const { textElems } = require('../plugins/_collections');
 
 class SvgoParserError extends Error {
@@ -82,6 +81,7 @@
   lowercase: true,
   xmlns: true,
   position: true,
+  unparsedEntities: true,
 };
 
 /**
@@ -116,9 +116,6 @@
     current.children.push(node);
   };
 
-  /**
-   * @type {(doctype: string) => void}
-   */
   sax.ondoctype = (doctype) => {
     /**
      * @type {XastDoctype}
@@ -143,9 +140,6 @@
     }
   };
 
-  /**
-   * @type {(data: { name: string, body: string }) => void}
-   */
   sax.onprocessinginstruction = (data) => {
     /**
      * @type {XastInstruction}
@@ -158,9 +152,6 @@
     pushToContent(node);
   };
 
-  /**
-   * @type {(comment: string) => void}
-   */
   sax.oncomment = (comment) => {
     /**
      * @type {XastComment}
@@ -172,9 +163,6 @@
     pushToContent(node);
   };
 
-  /**
-   * @type {(cdata: string) => void}
-   */
   sax.oncdata = (cdata) => {
     /**
      * @type {XastCdata}
@@ -186,9 +174,6 @@
     pushToContent(node);
   };
 
-  /**
-   * @type {(data: { name: string, attributes: Record<string, { value: string }>}) => void}
-   */
   sax.onopentag = (data) => {
     /**
      * @type {XastElement}
@@ -207,9 +192,6 @@
     stack.push(element);
   };
 
-  /**
-   * @type {(text: string) => void}
-   */
   sax.ontext = (text) => {
     if (current.type === 'element') {
       // prevent trimming of meaningful whitespace inside textual tags
@@ -240,14 +222,12 @@
     current = stack[stack.length - 1];
   };
 
-  /**
-   * @type {(e: any) => void}
-   */
   sax.onerror = (e) => {
+    const reason = e.message.split('\n')[0];
     const error = new SvgoParserError(
-      e.reason,
-      e.line + 1,
-      e.column,
+      reason,
+      sax.line + 1,
+      sax.column,
       data,
       from,
     );
--- /dev/null
+++ b/test/svgo/billion-laughs.test.js
@@ -0,0 +1,27 @@
+const { readFile } = require('node:fs/promises');
+const path = require('path');
+const { optimize } = require('../../lib/svgo.js');
+
+describe('svgo', () => {
+  it('should throw on excessive expansion depth', async () => {
+    const original = await readFile(
+      path.join(__dirname, 'billion-laughs.svg'),
+      'utf-8',
+    );
+
+    expect(() => {
+      optimize(original);
+    }).toThrow('Parsed entity depth exceeds max entity depth');
+  });
+
+  it('should throw on excessive expansion count', async () => {
+    const original = await readFile(
+      path.join(__dirname, 'billion-laughs-flat.svg'),
+      'utf-8',
+    );
+
+    expect(() => {
+      optimize(original);
+    }).toThrow('Parsed entity count exceeds max entity count');
+  });
+});
--- /dev/null
+++ b/test/svgo/billion-laughs.svg
@@ -0,0 +1,15 @@
+<?xml version="1.0"?>
+<!DOCTYPE lolz [
+<!ENTITY lol "lol">
+<!ELEMENT lolz (#PCDATA)>
+<!ENTITY lol1 "&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;">
+<!ENTITY lol2 "&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;">
+<!ENTITY lol3 "&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;">
+<!ENTITY lol4 "&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;">
+<!ENTITY lol5 "&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;">
+<!ENTITY lol6 "&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;">
+<!ENTITY lol7 "&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;">
+<!ENTITY lol8 "&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;">
+<!ENTITY lol9 "&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;">
+]>
+<lolz>&lol9;</lolz>
--- /dev/null
+++ b/test/svgo/billion-laughs-flat.svg
@@ -0,0 +1,8 @@
+<?xml version="1.0"?>
+<!DOCTYPE lolz [
+<!ENTITY lol "lol">
+<!ELEMENT lolz (#PCDATA)>
+<!ENTITY lol1 "&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;">
+<!ENTITY lol2 "&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;">
+]>
+<lolz>&lol2;</lolz>
