--- a/dist/server/lib/router-server.js
+++ b/dist/server/lib/router-server.js
@@ -451,7 +451,7 @@
                     return developmentBundler.hotReloader.onHMR(req, socket, head);
                 }
             }
-            const { matchedOutput, parsedUrl } = await resolveRoutes({
+            const { finished, matchedOutput, parsedUrl, statusCode } = await resolveRoutes({
                 req,
                 res: socket,
                 isUpgradeReq: true,
@@ -462,8 +462,11 @@
             if (matchedOutput) {
                 return socket.end();
             }
-            if (parsedUrl.protocol) {
-                return await (0, _proxyrequest.proxyRequest)(req, socket, parsedUrl, head);
+            if (finished && parsedUrl.protocol) {
+                if (!statusCode) {
+                    return await (0, _proxyrequest.proxyRequest)(req, socket, parsedUrl, head);
+                }
+                return socket.end();
             }
         // If there's no matched output, we don't handle the request as user's
         // custom WS server may be listening on the same path.
--- a/dist/esm/server/lib/router-server.js
+++ b/dist/esm/server/lib/router-server.js
@@ -437,7 +437,7 @@
                     return developmentBundler.hotReloader.onHMR(req, socket, head);
                 }
             }
-            const { matchedOutput, parsedUrl } = await resolveRoutes({
+            const { finished, matchedOutput, parsedUrl, statusCode } = await resolveRoutes({
                 req,
                 res: socket,
                 isUpgradeReq: true,
@@ -448,8 +448,11 @@
             if (matchedOutput) {
                 return socket.end();
             }
-            if (parsedUrl.protocol) {
-                return await proxyRequest(req, socket, parsedUrl, head);
+            if (finished && parsedUrl.protocol) {
+                if (!statusCode) {
+                    return await proxyRequest(req, socket, parsedUrl, head);
+                }
+                return socket.end();
             }
         // If there's no matched output, we don't handle the request as user's
         // custom WS server may be listening on the same path.
