--- a/lib/tmp.js
+++ b/lib/tmp.js
@@ -512,6 +512,22 @@
 }
 
 /**
+ * Checks the given path value for relative path components.
+ *
+ * @param {string} value
+ * @returns {string} the unmodified value if it is safe
+ * @throws {Error} if the value contains a relative path component
+ * @private
+ */
+function _assertPath(value) {
+  if (value.includes('..')) {
+    throw new Error('Relative value not allowed');
+  }
+
+  return value;
+}
+
+/**
  * Asserts whether the specified options are valid, also sanitizes options and provides sane defaults for missing
  * options.
  *
@@ -555,8 +571,8 @@
 
   // for completeness' sake only, also keep (multiple) blanks if the user, purportedly sane, requests us to
   options.name = _isUndefined(options.name) ? undefined : options.name;
-  options.prefix = _isUndefined(options.prefix) ? '' : options.prefix;
-  options.postfix = _isUndefined(options.postfix) ? '' : options.postfix;
+  options.prefix = _isUndefined(options.prefix) ? '' : _assertPath(options.prefix);
+  options.postfix = _isUndefined(options.postfix) ? '' : _assertPath(options.postfix);
 }
 
 /**
--- /dev/null
+++ b/test/GHSA-ph9p-34f9-6g65-test.js
@@ -0,0 +1,24 @@
+const assert = require('assert');
+const tmp = require('../lib/tmp');
+
+describe('GHSA-ph9p-34f9-6g65', function () {
+  describe('#fileSync with attacker `prefix`', function () {
+    it('should not allow such prefixes', function (done) {
+      assert.throws(function () {
+        tmp.fileSync({ prefix: "../foo"});
+      }, new RegExp('^Error: Relative value not allowed'));
+
+      done();
+    });
+  });
+
+  describe('#fileSync with attacker `postfix`', function () {
+    it('should not allow such prefixes', function (done) {
+      assert.throws(function () {
+        tmp.fileSync({ postfix: "../foo"});
+      }, new RegExp('^Error: Relative value not allowed'));
+
+      done();
+    });
+  });
+});
