--- /dev/null
+++ b/package.json
@@ -0,0 +1,5 @@
+{
+  "name": "esbuild-source",
+  "private": true,
+  "version": "0.19.12"
+}
--- a/pkg/api/serve_other.go
+++ b/pkg/api/serve_other.go
@@ -48,6 +48,7 @@ type apiHandler struct {
 	keyfileToLower   string
 	certfileToLower  string
 	fallback         string
+	hosts            []string
 	serveWaitGroup   sync.WaitGroup
 	activeStreams    []chan serverSentEvent
 	currentHashes    map[string]string
@@ -103,12 +104,6 @@ func errorsToString(errors []Message) string {
 func (h *apiHandler) ServeHTTP(res http.ResponseWriter, req *http.Request) {
 	start := time.Now()
 
-	// Special-case the esbuild event stream
-	if req.Method == "GET" && req.URL.Path == "/esbuild" && req.Header.Get("Accept") == "text/event-stream" {
-		h.serveEventStream(start, req, res)
-		return
-	}
-
 	// HEAD requests omit the body
 	maybeWriteResponseBody := func(bytes []byte) { res.Write(bytes) }
 	isHEAD := req.Method == "HEAD"
@@ -116,9 +111,38 @@ func (h *apiHandler) ServeHTTP(res http.ResponseWriter, req *http.Request) {
 		maybeWriteResponseBody = func(bytes []byte) { res.Write(nil) }
 	}
 
+	// Check the "Host" header to prevent DNS rebinding attacks
+	// (security fix for GHSA-67mh-4wv8-2f99)
+	if strings.ContainsRune(req.Host, ':') {
+		// Try to strip off the port number
+		if host, _, err := net.SplitHostPort(req.Host); err == nil {
+			req.Host = host
+		}
+	}
+	if req.Host != "localhost" {
+		ok := false
+		for _, allowed := range h.hosts {
+			if req.Host == allowed {
+				ok = true
+				break
+			}
+		}
+		if !ok {
+			go h.notifyRequest(time.Since(start), req, http.StatusForbidden)
+			res.WriteHeader(http.StatusForbidden)
+			maybeWriteResponseBody([]byte(fmt.Sprintf("403 - Forbidden: The host %q is not allowed", req.Host)))
+			return
+		}
+	}
+
+	// Special-case the esbuild event stream
+	if req.Method == "GET" && req.URL.Path == "/esbuild" && req.Header.Get("Accept") == "text/event-stream" {
+		h.serveEventStream(start, req, res)
+		return
+	}
+
 	// Handle GET and HEAD requests
 	if (isHEAD || req.Method == "GET") && strings.HasPrefix(req.URL.Path, "/") {
-		res.Header().Set("Access-Control-Allow-Origin", "*")
 		queryPath := path.Clean(req.URL.Path)[1:]
 		result := h.rebuild()
 
@@ -360,7 +384,6 @@ func (h *apiHandler) serveEventStream(start time.Time, req *http.Request, res ht
 			res.Header().Set("Content-Type", "text/event-stream")
 			res.Header().Set("Connection", "keep-alive")
 			res.Header().Set("Cache-Control", "no-cache")
-			res.Header().Set("Access-Control-Allow-Origin", "*")
 			go h.notifyRequest(time.Since(start), req, http.StatusOK)
 			res.WriteHeader(http.StatusOK)
 			res.Write([]byte("retry: 500\n"))
@@ -785,11 +808,29 @@ func (ctx *internalContext) Serve(serveOptions ServeOptions) (ServeResult, error
 
 	// Extract the real port in case we passed a port of "0"
 	var result ServeResult
+	var boundHost string
 	if host, text, err := net.SplitHostPort(addr); err == nil {
 		if port, err := strconv.ParseInt(text, 10, 32); err == nil {
 			result.Port = uint16(port)
 			result.Host = host
+			boundHost = host
+		}
+	}
+
+	// Build up a list of allowed hosts for the "Host" header check
+	// (security fix for GHSA-67mh-4wv8-2f99)
+	var allowedHosts []string
+	if ip := net.ParseIP(boundHost); ip != nil && ip.IsUnspecified() {
+		// If this is "0.0.0.0" or "::", list all relevant IP addresses
+		if addrs, err := net.InterfaceAddrs(); err == nil {
+			for _, addr := range addrs {
+				if addr, ok := addr.(*net.IPNet); ok && (addr.IP.To4() != nil) == (ip.To4() != nil) && !addr.IP.IsLinkLocalUnicast() {
+					allowedHosts = append(allowedHosts, addr.IP.String())
+				}
+			}
 		}
+	} else if boundHost != "" {
+		allowedHosts = append(allowedHosts, boundHost)
 	}
 
 	// HTTPS-related files should be absolute paths
@@ -811,6 +852,7 @@ func (ctx *internalContext) Serve(serveOptions ServeOptions) (ServeResult, error
 		keyfileToLower:   strings.ToLower(serveOptions.Keyfile),
 		certfileToLower:  strings.ToLower(serveOptions.Certfile),
 		fallback:         serveOptions.Fallback,
+		hosts:            allowedHosts,
 		rebuild: func() BuildResult {
 			if atomic.LoadInt32(&shouldStop) != 0 {
 				// Don't start more rebuilds if we were told to stop
