diff --git a/lib/index.js b/lib/index.js
index 0a492de..1031511 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -58,8 +58,9 @@ function constructExportCall(path, exportIdent, exportNames, exportValues, expor
       const objectProperties = [];
       for (let i = 0; i < exportNames.length; i++) {
         const exportName = exportNames[i];
+        const exportNameNode = stringSpecifiers.has(exportName) ? _core.types.stringLiteral(exportName) : _core.types.identifier(exportName);
         const exportValue = exportValues[i];
-        objectProperties.push(_core.types.objectProperty(stringSpecifiers.has(exportName) ? _core.types.stringLiteral(exportName) : _core.types.identifier(exportName), exportValue));
+        objectProperties.push(_core.types.objectProperty(exportNameNode, exportValue));
       }
       statements.push(_core.types.expressionStatement(_core.types.callExpression(exportIdent, [_core.types.objectExpression(objectProperties)])));
     }
@@ -73,8 +74,10 @@ function constructExportCall(path, exportIdent, exportNames, exportValues, expor
     }));
     for (let i = 0; i < exportNames.length; i++) {
       const exportName = exportNames[i];
+      const computed = stringSpecifiers.has(exportName);
+      const exportNameNode = computed ? _core.types.stringLiteral(exportName) : _core.types.identifier(exportName);
       const exportValue = exportValues[i];
-      statements.push(_core.types.expressionStatement(_core.types.assignmentExpression("=", _core.types.memberExpression(_core.types.identifier(exportObj), _core.types.identifier(exportName)), exportValue)));
+      statements.push(_core.types.expressionStatement(_core.types.assignmentExpression("=", _core.types.memberExpression(_core.types.identifier(exportObj), exportNameNode, computed), exportValue)));
     }
     statements.push(_core.types.expressionStatement(_core.types.callExpression(exportIdent, [_core.types.identifier(exportObj)])));
   }
