--- a/lib/rules_inline/linkify.mjs
+++ b/lib/rules_inline/linkify.mjs
@@ -30,7 +30,14 @@ export default function linkify (state, silent) {
   if (url.length <= proto.length) return false

   // disallow '*' at the end of the link (conflicts with emphasis)
-  url = url.replace(/\*+$/, '')
+  // do manual backsearch to avoid perf issues with regex /\*+$/ on "****...****a".
+  let urlEnd = url.length
+  while (urlEnd > 0 && url.charCodeAt(urlEnd - 1) === 0x2A/* * */) {
+    urlEnd--
+  }
+  if (urlEnd !== url.length) {
+    url = url.slice(0, urlEnd)
+  }

   const fullUrl = state.md.normalizeLink(url)
   if (!state.md.validateLink(fullUrl)) return false
