diff --git a/plugins/sudoers/auth/aix_auth.c b/plugins/sudoers/auth/aix_auth.c
index 627281160..d0afc0428 100644
--- a/plugins/sudoers/auth/aix_auth.c
+++ b/plugins/sudoers/auth/aix_auth.c
@@ -275,17 +275,17 @@ sudo_aix_verify(struct passwd *pw, char *prompt, sudo_auth *auth, struct sudo_co
 	    /* password expired, user must change it */
 	    if (!sudo_aix_change_password(pw->pw_name)) {
 		sudo_warnx(U_("unable to change password for %s"), pw->pw_name);
-		ret = AUTH_FATAL;
+		ret = AUTH_ERROR;
 	    }
 	    break;
 	case 2:
 	    /* password expired, only admin can change it */
-	    ret = AUTH_FATAL;
+	    ret = AUTH_ERROR;
 	    break;
 	default:
 	    /* error (-1) */
 	    sudo_warn("passwdexpired");
-	    ret = AUTH_FATAL;
+	    ret = AUTH_ERROR;
 	    break;
 	}
     }
diff --git a/plugins/sudoers/auth/bsdauth.c b/plugins/sudoers/auth/bsdauth.c
index fec7402f2..b51049f0d 100644
--- a/plugins/sudoers/auth/bsdauth.c
+++ b/plugins/sudoers/auth/bsdauth.c
@@ -68,13 +68,13 @@ bsdauth_init(struct passwd *pw, sudo_auth *auth)
     if (state.lc == NULL) {
 	log_warning(0,
 	    N_("unable to get login class for user %s"), pw->pw_name);
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
     if ((state.as = auth_open()) == NULL) {
 	log_warning(0, N_("unable to begin bsd authentication"));
 	login_close(state.lc);
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
     /* XXX - maybe check the auth style earlier? */
@@ -83,7 +83,7 @@ bsdauth_init(struct passwd *pw, sudo_auth *auth)
 	log_warningx(0, N_("invalid authentication type"));
 	auth_close(state.as);
 	login_close(state.lc);
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
      if (auth_setitem(state.as, AUTHV_STYLE, login_style) < 0 ||
@@ -92,7 +92,7 @@ bsdauth_init(struct passwd *pw, sudo_auth *auth)
 	log_warningx(0, N_("unable to initialize BSD authentication"));
 	auth_close(state.as);
 	login_close(state.lc);
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
     auth->data = (void *) &state;
@@ -141,7 +141,7 @@ bsdauth_verify(struct passwd *pw, char *prompt, sudo_auth *auth, struct sudo_con
 		prompt[len--] = '\0';
 	    if (asprintf(&s, "%s [echo on]: ", prompt) == -1) {
 		log_warningx(0, N_("unable to allocate memory"));
-		debug_return_int(AUTH_FATAL);
+		debug_return_int(AUTH_ERROR);
 	    }
 	    free(pass);
 	    pass = auth_getpass(s, SUDO_CONV_PROMPT_ECHO_ON, callback);
diff --git a/plugins/sudoers/auth/fwtk.c b/plugins/sudoers/auth/fwtk.c
index 7f3385d0f..f3cdbcf42 100644
--- a/plugins/sudoers/auth/fwtk.c
+++ b/plugins/sudoers/auth/fwtk.c
@@ -52,22 +52,22 @@ sudo_fwtk_init(struct passwd *pw, sudo_auth *auth)
 
     if ((confp = cfg_read("sudo")) == (Cfg *)-1) {
 	sudo_warnx("%s", U_("unable to read fwtk config"));
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
     if (auth_open(confp)) {
 	sudo_warnx("%s", U_("unable to connect to authentication server"));
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
     /* Get welcome message from auth server */
     if (auth_recv(resp, sizeof(resp))) {
 	sudo_warnx("%s", U_("lost connection to authentication server"));
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
     if (strncmp(resp, "Authsrv ready", 13) != 0) {
 	sudo_warnx(U_("authentication server error:\n%s"), resp);
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
     debug_return_int(AUTH_SUCCESS);
@@ -87,7 +87,7 @@ sudo_fwtk_verify(struct passwd *pw, char *prompt, sudo_auth *auth, struct sudo_c
 restart:
     if (auth_send(buf) || auth_recv(resp, sizeof(resp))) {
 	sudo_warnx("%s", U_("lost connection to authentication server"));
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
     /* Get the password/response from the user. */
@@ -109,7 +109,7 @@ restart:
 	goto restart;
     } else {
 	sudo_warnx("%s", resp);
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
     if (!pass) {			/* ^C or error */
 	debug_return_int(AUTH_INTR);
@@ -119,7 +119,7 @@ restart:
     (void) snprintf(buf, sizeof(buf), "response '%s'", pass);
     if (auth_send(buf) || auth_recv(resp, sizeof(resp))) {
 	sudo_warnx("%s", U_("lost connection to authentication server"));
-	error = AUTH_FATAL;
+	error = AUTH_ERROR;
 	goto done;
     }
 
diff --git a/plugins/sudoers/auth/kerb5.c b/plugins/sudoers/auth/kerb5.c
index a6072269f..06bbc8577 100644
--- a/plugins/sudoers/auth/kerb5.c
+++ b/plugins/sudoers/auth/kerb5.c
@@ -117,7 +117,7 @@ sudo_krb5_setup(struct passwd *pw, char **promptp, sudo_auth *auth)
 	    if (asprintf(&krb5_prompt, "Password for %s: ", pname) == -1) {
 		log_warningx(0, N_("unable to allocate memory"));
 		free(pname);
-		debug_return_int(AUTH_FATAL);
+		debug_return_int(AUTH_ERROR);
 	    }
 	/*}*/
 	free(pname);
@@ -142,7 +142,7 @@ sudo_krb5_init(struct passwd *pw, sudo_auth *auth)
 	    sudo_krb5_instance[0] != '/' ? "/" : "", sudo_krb5_instance);
 	if (len == -1) {
 	    log_warningx(0, N_("unable to allocate memory"));
-	    debug_return_int(AUTH_FATAL);
+	    debug_return_int(AUTH_ERROR);
 	}
     }
 
diff --git a/plugins/sudoers/auth/pam.c b/plugins/sudoers/auth/pam.c
index 6310364b7..396f2426f 100644
--- a/plugins/sudoers/auth/pam.c
+++ b/plugins/sudoers/auth/pam.c
@@ -216,7 +216,7 @@ sudo_pam_init2(struct passwd *pw, sudo_auth *auth, bool quiet)
 	    &pam_conv, &pamh, errstr);
 	if (!quiet)
 	    log_warningx(0, N_("unable to initialize PAM: %s"), errstr);
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
     /* Initialize conversation function message filter. */
@@ -335,7 +335,7 @@ sudo_pam_verify(struct passwd *pw, char *prompt, sudo_auth *auth, struct sudo_co
 	default:
 	    s = sudo_pam_strerror(pamh, *pam_status);
 	    log_warningx(0, N_("PAM authentication error: %s"), s);
-	    debug_return_int(AUTH_FATAL);
+	    debug_return_int(AUTH_ERROR);
     }
 }
 
@@ -355,7 +355,7 @@ sudo_pam_approval(struct passwd *pw, sudo_auth *auth, bool exempt)
 	    case PAM_AUTH_ERR:
 		log_warningx(0, N_("account validation failure, "
 		    "is your account locked?"));
-		status = AUTH_FATAL;
+		status = AUTH_ERROR;
 		break;
 	    case PAM_NEW_AUTHTOK_REQD:
 		/* Ignore if user is exempt from password restrictions. */
@@ -382,13 +382,13 @@ sudo_pam_approval(struct passwd *pw, sudo_auth *auth, bool exempt)
 		/* Password expired, cannot be updated by user. */
 		log_warningx(0,
 		    N_("Password expired, contact your system administrator"));
-		status = AUTH_FATAL;
+		status = AUTH_ERROR;
 		break;
 	    case PAM_ACCT_EXPIRED:
 		log_warningx(0,
 		    N_("Account expired or PAM config lacks an \"account\" "
 		    "section for sudo, contact your system administrator"));
-		status = AUTH_FATAL;
+		status = AUTH_ERROR;
 		break;
 	    case PAM_AUTHINFO_UNAVAIL:
 	    case PAM_MAXTRIES:
@@ -400,7 +400,7 @@ sudo_pam_approval(struct passwd *pw, sudo_auth *auth, bool exempt)
 	    default:
 		s = sudo_pam_strerror(pamh, rc);
 		log_warningx(0, N_("PAM account management error: %s"), s);
-		status = AUTH_FATAL;
+		status = AUTH_ERROR;
 		break;
 	}
 	*pam_status = rc;
@@ -507,7 +507,7 @@ sudo_pam_begin_session(struct passwd *pw, char **user_envp[], sudo_auth *auth)
 		    "pam_end: %s", errstr);
 	    }
 	    pamh = NULL;
-	    status = AUTH_FATAL;
+	    status = AUTH_ERROR;
 	    goto done;
 	}
     }
@@ -523,7 +523,7 @@ sudo_pam_begin_session(struct passwd *pw, char **user_envp[], sudo_auth *auth)
 	if (pam_envp != NULL) {
 	    /* Merge pam env with user env. */
 	    if (!env_init(*user_envp) || !env_merge(pam_envp))
-		status = AUTH_FATAL;
+		status = AUTH_ERROR;
 	    *user_envp = env_get();
 	    (void)env_init(NULL);
 	    free(pam_envp);
@@ -576,7 +576,7 @@ sudo_pam_end_session(struct passwd *pw, sudo_auth *auth)
 	    errstr = sudo_pam_strerror(pamh, rc);
 	    sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
 		"pam_end: %s", errstr);
-	    status = AUTH_FATAL;
+	    status = AUTH_ERROR;
 	}
 	pamh = NULL;
     }
diff --git a/plugins/sudoers/auth/passwd.c b/plugins/sudoers/auth/passwd.c
index 566881f68..a9bbbbc84 100644
--- a/plugins/sudoers/auth/passwd.c
+++ b/plugins/sudoers/auth/passwd.c
@@ -52,7 +52,7 @@ sudo_passwd_init(struct passwd *pw, sudo_auth *auth)
     sudo_setspent();
     auth->data = sudo_getepw(pw);
     sudo_endspent();
-    debug_return_int(auth->data ? AUTH_SUCCESS : AUTH_FATAL);
+    debug_return_int(auth->data ? AUTH_SUCCESS : AUTH_ERROR);
 }
 
 #ifdef HAVE_CRYPT
@@ -62,7 +62,7 @@ sudo_passwd_verify(struct passwd *pw, char *pass, sudo_auth *auth, struct sudo_c
     char sav, *epass;
     char *pw_epasswd = auth->data;
     size_t pw_len;
-    int matched = 0;
+    int ret;
     debug_decl(sudo_passwd_verify, SUDOERS_DEBUG_AUTH);
 
     /* An empty plain-text password must match an empty encrypted password. */
@@ -75,8 +75,10 @@ sudo_passwd_verify(struct passwd *pw, char *pass, sudo_auth *auth, struct sudo_c
      */
     sav = pass[8];
     pw_len = strlen(pw_epasswd);
-    if (pw_len == DESLEN || HAS_AGEINFO(pw_epasswd, pw_len))
-	pass[8] = '\0';
+    if (pw_len == DESLEN || HAS_AGEINFO(pw_epasswd, pw_len)) {
+	(void)strlcpy(des_pass, pass, sizeof(des_pass));
+	pass = des_pass;
+    }
 
     /*
      * Normal UN*X password check.
@@ -84,28 +86,36 @@ sudo_passwd_verify(struct passwd *pw, char *pass, sudo_auth *auth, struct sudo_c
      * only compare the first DESLEN characters in that case.
      */
     epass = (char *) crypt(pass, pw_epasswd);
-    pass[8] = sav;
-    if (epass != NULL) {
-	if (HAS_AGEINFO(pw_epasswd, pw_len) && strlen(epass) == DESLEN)
-	    matched = !strncmp(pw_epasswd, epass, DESLEN);
-	else
-	    matched = !strcmp(pw_epasswd, epass);
+    ret = AUTH_FAILURE;
+     if (epass != NULL) {
+	if (HAS_AGEINFO(pw_epasswd, pw_len) && strlen(epass) == DESLEN) {
+	    if (strncmp(pw_epasswd, epass, DESLEN) == 0)
+		ret = AUTH_SUCCESS;
+	} else {
+	    if (strcmp(pw_epasswd, epass) == 0)
+		ret = AUTH_SUCCESS;
+	}
     }
 
-    debug_return_int(matched ? AUTH_SUCCESS : AUTH_FAILURE);
+    explicit_bzero(des_pass, sizeof(des_pass));
+
+    debug_return_int(ret);
 }
 #else
 int
 sudo_passwd_verify(struct passwd *pw, char *pass, sudo_auth *auth, struct sudo_conv_callback *callback)
 {
     char *pw_passwd = auth->data;
-    int matched;
+    int ret;
     debug_decl(sudo_passwd_verify, SUDOERS_DEBUG_AUTH);
 
     /* Simple string compare for systems without crypt(). */
-    matched = !strcmp(pass, pw_passwd);
+    if (strcmp(pass, pw_passwd) == 0)
+	ret = AUTH_SUCCESS;
+    else
+	ret = AUTH_FAILURE;
 
-    debug_return_int(matched ? AUTH_SUCCESS : AUTH_FAILURE);
+    debug_return_int(ret);
 }
 #endif
 
diff --git a/plugins/sudoers/auth/rfc1938.c b/plugins/sudoers/auth/rfc1938.c
index 2e4824c60..f2db2b1d3 100644
--- a/plugins/sudoers/auth/rfc1938.c
+++ b/plugins/sudoers/auth/rfc1938.c
@@ -97,7 +97,7 @@ sudo_rfc1938_setup(struct passwd *pw, char **promptp, sudo_auth *auth)
     if (rfc1938challenge(&rfc1938, pw->pw_name, challenge, sizeof(challenge))) {
 	if (IS_ONEANDONLY(auth)) {
 	    sudo_warnx(U_("you do not exist in the %s database"), auth->name);
-	    debug_return_int(AUTH_FATAL);
+	    debug_return_int(AUTH_ERROR);
 	} else {
 	    debug_return_int(AUTH_FAILURE);
 	}
@@ -109,7 +109,7 @@ sudo_rfc1938_setup(struct passwd *pw, char **promptp, sudo_auth *auth)
 	char *p = realloc(new_prompt, op_len + challenge_len + 7);
 	if (p == NULL) {
 	    sudo_warnx(U_("%s: %s"), __func__, U_("unable to allocate memory"));
-	    debug_return_int(AUTH_FATAL);
+	    debug_return_int(AUTH_ERROR);
 	}
 	np_size = op_len + challenge_len + 7;
 	new_prompt = p;
diff --git a/plugins/sudoers/auth/secureware.c b/plugins/sudoers/auth/secureware.c
index dd7aa42eb..9df6a2e05 100644
--- a/plugins/sudoers/auth/secureware.c
+++ b/plugins/sudoers/auth/secureware.c
@@ -63,7 +63,7 @@ sudo_secureware_init(struct passwd *pw, sudo_auth *auth)
     sudo_setspent();
     auth->data = sudo_getepw(pw);
     sudo_endspent();
-    debug_return_int(auth->data ? AUTH_SUCCESS : AUTH_FATAL);
+    debug_return_int(auth->data ? AUTH_SUCCESS : AUTH_ERROR);
 }
 
 int
diff --git a/plugins/sudoers/auth/securid5.c b/plugins/sudoers/auth/securid5.c
index 698953fc2..368bff4c7 100644
--- a/plugins/sudoers/auth/securid5.c
+++ b/plugins/sudoers/auth/securid5.c
@@ -70,7 +70,7 @@ sudo_securid_init(struct passwd *pw, sudo_auth *auth)
 	debug_return_int(AUTH_SUCCESS);
 
     sudo_warnx("%s", U_("failed to initialise the ACE API library"));
-    debug_return_int(AUTH_FATAL);
+    debug_return_int(AUTH_ERROR);
 }
 
 /*
@@ -96,7 +96,7 @@ sudo_securid_setup(struct passwd *pw, char **promptp, sudo_auth *auth)
     /* Re-initialize SecurID every time. */
     if (SD_Init(sd) != ACM_OK) {
 	sudo_warnx("%s", U_("unable to contact the SecurID server"));
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
     /* Lock new PIN code */
@@ -109,19 +109,19 @@ sudo_securid_setup(struct passwd *pw, char **promptp, sudo_auth *auth)
 
         case ACE_UNDEFINED_USERNAME:
 		sudo_warnx("%s", U_("invalid username length for SecurID"));
-		debug_return_int(AUTH_FATAL);
+		debug_return_int(AUTH_ERROR);
 
 	case ACE_ERR_INVALID_HANDLE:
 		sudo_warnx("%s", U_("invalid Authentication Handle for SecurID"));
-		debug_return_int(AUTH_FATAL);
+		debug_return_int(AUTH_ERROR);
 
 	case ACM_ACCESS_DENIED:
 		sudo_warnx("%s", U_("SecurID communication failed"));
-		debug_return_int(AUTH_FATAL);
+		debug_return_int(AUTH_ERROR);
 
 	default:
 		sudo_warnx("%s", U_("unknown SecurID error"));
-		debug_return_int(AUTH_FATAL);
+		debug_return_int(AUTH_ERROR);
 	}
 }
 
@@ -155,17 +155,17 @@ sudo_securid_verify(struct passwd *pw, char *pass, sudo_auth *auth, struct sudo_
 
 	case ACE_UNDEFINED_PASSCODE:
 		sudo_warnx("%s", U_("invalid passcode length for SecurID"));
-		ret = AUTH_FATAL;
+		ret = AUTH_ERROR;
 		break;
 
 	case ACE_UNDEFINED_USERNAME:
 		sudo_warnx("%s", U_("invalid username length for SecurID"));
-		ret = AUTH_FATAL;
+		ret = AUTH_ERROR;
 		break;
 
 	case ACE_ERR_INVALID_HANDLE:
 		sudo_warnx("%s", U_("invalid Authentication Handle for SecurID"));
-		ret = AUTH_FATAL;
+		ret = AUTH_ERROR;
 		break;
 
 	case ACM_ACCESS_DENIED:
@@ -203,12 +203,12 @@ then enter the new token code.\n", \
 		    "Your SecurID access has not yet been set up.\n");
 		sudo_printf(SUDO_CONV_ERROR_MSG|SUDO_CONV_PREFER_TTY, 
 		    "Please set up a PIN before you try to authenticate.\n");
-		ret = AUTH_FATAL;
+		ret = AUTH_ERROR;
 		break;
 
 	default:
 		sudo_warnx("%s", U_("unknown SecurID error"));
-		ret = AUTH_FATAL;
+		ret = AUTH_ERROR;
 		break;
     }
 
diff --git a/plugins/sudoers/auth/sia.c b/plugins/sudoers/auth/sia.c
index 7feb19e03..8f2c58a5f 100644
--- a/plugins/sudoers/auth/sia.c
+++ b/plugins/sudoers/auth/sia.c
@@ -57,7 +57,7 @@ sudo_sia_setup(struct passwd *pw, char **promptp, sudo_auth *auth)
     sudo_argv = reallocarray(NULL, sudo_argc + 1, sizeof(char *));
     if (sudo_argv == NULL) {
 	log_warningx(0, N_("unable to allocate memory"));
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
     sudo_argv[0] = "sudo";
     for (i = 0; i < NewArgc; i++)
@@ -67,7 +67,7 @@ sudo_sia_setup(struct passwd *pw, char **promptp, sudo_auth *auth)
     /* We don't let SIA prompt the user for input. */
     if (sia_ses_init(&siah, sudo_argc, sudo_argv, NULL, pw->pw_name, user_ttypath, 0, NULL) != SIASUCCESS) {
 	log_warning(0, N_("unable to initialize SIA session"));
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     }
 
     auth->data = siah;
@@ -95,7 +95,7 @@ sudo_sia_verify(struct passwd *pw, char *prompt, sudo_auth *auth,
     if (rc == SIASUCCESS)
 	debug_return_int(AUTH_SUCCESS);
     if (ISSET(rc, SIASTOP))
-	debug_return_int(AUTH_FATAL);
+	debug_return_int(AUTH_ERROR);
     debug_return_int(AUTH_FAILURE);
 }
 
@@ -115,7 +115,7 @@ int
 sudo_sia_begin_session(struct passwd *pw, char **user_envp[], sudo_auth *auth)
 {
     SIAENTITY *siah;
-    int status = AUTH_FATAL;
+    int status = AUTH_ERROR;
     debug_decl(sudo_sia_begin_session, SUDOERS_DEBUG_AUTH);
 
     /* Re-init sia for the target user's session. */
diff --git a/plugins/sudoers/auth/sudo_auth.c b/plugins/sudoers/auth/sudo_auth.c
index 188b65fde..c8b5a9d99 100644
--- a/plugins/sudoers/auth/sudo_auth.c
+++ b/plugins/sudoers/auth/sudo_auth.c
@@ -112,10 +112,16 @@ sudo_auth_init(struct passwd *pw)
 	if (auth->init && !IS_DISABLED(auth)) {
 	    /* Disable if it failed to init unless there was a fatal error. */
 	    status = (auth->init)(pw, auth);
-	    if (status == AUTH_FAILURE)
+	    switch (status) {
+	    case AUTH_SUCCESS:
+		break;
+	    case AUTH_FAILURE:
 		SET(auth->flags, FLAG_DISABLED);
-	    else if (status == AUTH_FATAL)
-		break;		/* assume error msg already printed */
+	    break;
+	    default:
+		/* Assume error msg already printed. */
+		debug_return_int(-1);
+	    }		/* assume error msg already printed */
 	}
     }
 
@@ -161,7 +167,7 @@ sudo_auth_init(struct passwd *pw)
 	}
     }
 
-    debug_return_int(status == AUTH_FATAL ? -1 : 0);
+    debug_return_int(0);
 }
 
 /*
@@ -202,7 +208,7 @@ sudo_auth_cleanup(struct passwd *pw, bool force)
     for (auth = auth_switch; auth->name; auth++) {
 	if (auth->cleanup && !IS_DISABLED(auth)) {
 	    int status = (auth->cleanup)(pw, auth, force);
-	    if (status == AUTH_FATAL) {
+	    if (status != AUTH_SUCCESS) {
 		/* Assume error msg already printed. */
 		debug_return_int(-1);
 	    }
@@ -297,7 +303,9 @@ verify_user(struct passwd *pw, char *prompt, int validated,
 		status = (auth->setup)(pw, &prompt, auth);
 		if (status == AUTH_FAILURE)
 		    SET(auth->flags, FLAG_DISABLED);
-		else if (status == AUTH_FATAL || user_interrupted())
+		else if (status == AUTH_NONINTERACTIVE)
+		    goto done;
+		else if (status != AUTH_SUCCESS || user_interrupted())
 		    goto done;		/* assume error msg already printed */
 	    }
 	}
@@ -348,7 +356,6 @@ done:
 	    log_auth_failure(validated, ntries);
 	    ret = false;
 	    break;
-	case AUTH_FATAL:
 	default:
 	    log_auth_failure(validated, 0);
 	    ret = -1;
@@ -360,24 +367,31 @@ done:
 
 /*
  * Call authentication method begin session hooks.
- * Returns 1 on success and -1 on error.
+ *  Returns true on success, false on failure and -1 on error.
  */
 int
 sudo_auth_begin_session(struct passwd *pw, char **user_env[])
 {
     sudo_auth *auth;
+    int ret = true;	
     debug_decl(sudo_auth_begin_session, SUDOERS_DEBUG_AUTH);
 
     for (auth = auth_switch; auth->name; auth++) {
 	if (auth->begin_session && !IS_DISABLED(auth)) {
 	    int status = (auth->begin_session)(pw, user_env, auth);
 	    if (status != AUTH_SUCCESS) {
-		/* Assume error msg already printed. */
-		debug_return_int(-1);
+		switch (status) {
+		case AUTH_FAILURE:
+		    ret = false;
+		    break;
+		default:
+		    /* Assume error msg already printed. */
+		    debug_return_int(-1);
+		}
 	    }
 	}
     }
-    debug_return_int(1);
+    debug_return_int(ret);
 }
 
 bool
@@ -398,25 +412,32 @@ sudo_auth_needs_end_session(void)
 
 /*
  * Call authentication method end session hooks.
- * Returns 1 on success and -1 on error.
+ *  Returns true on success, false on failure and -1 on error.
  */
 int
 sudo_auth_end_session(struct passwd *pw)
 {
     sudo_auth *auth;
+    int ret = true;
     int status;
     debug_decl(sudo_auth_end_session, SUDOERS_DEBUG_AUTH);
 
     for (auth = auth_switch; auth->name; auth++) {
 	if (auth->end_session && !IS_DISABLED(auth)) {
 	    status = (auth->end_session)(pw, auth);
-	    if (status == AUTH_FATAL) {
+	    switch (status) {
+	    case AUTH_SUCCESS:
+		break;
+	    case AUTH_FAILURE:
+		ret = false;
+		break;
+	    default:
 		/* Assume error msg already printed. */
 		debug_return_int(-1);
 	    }
 	}
     }
-    debug_return_int(1);
+    debug_return_int(ret);
 }
 
 /*
diff --git a/plugins/sudoers/auth/sudo_auth.h b/plugins/sudoers/auth/sudo_auth.h
index 9ee408d4f..17bb78ed3 100644
--- a/plugins/sudoers/auth/sudo_auth.h
+++ b/plugins/sudoers/auth/sudo_auth.h
@@ -20,10 +20,11 @@
 #define SUDO_AUTH_H
 
 /* Auth function return values.  */
-#define AUTH_SUCCESS	0
-#define AUTH_FAILURE	1
-#define AUTH_INTR	2
-#define AUTH_FATAL	3
+#define AUTH_SUCCESS		0x52a2925	/* 0101001010100010100100100101 */
+#define AUTH_FAILURE		0xad5d6da	/* 1010110101011101011011011010 */
+#define AUTH_INTR		0x69d61fc8	/* 1101001110101100001111111001000 */
+#define AUTH_ERROR		0x1629e037	/* 0010110001010011110000000110111 */
+#define AUTH_NONINTERACTIVE	0x1fc8d3ac	/* 11111110010001101001110101100 */
 
 typedef struct sudo_auth {
     int flags;			/* various flags, see below */
diff --git a/plugins/sudoers/match.c b/plugins/sudoers/match.c
index eb77cf2c3..88140854f 100644
--- a/plugins/sudoers/match.c
+++ b/plugins/sudoers/match.c
@@ -92,8 +92,13 @@ user_matches(struct sudoers_parse_tree *parse_tree, const struct passwd *pw,
 	    if ((a = alias_get(parse_tree, m->name, USERALIAS)) != NULL) {
 		/* XXX */
 		int rc = userlist_matches(parse_tree, pw, &a->members);
-		if (rc != UNSPEC)
-		    matched = m->negated ? !rc : rc;
+		if (SPECIFIED(rc)) {
+		    if (m->negated) {
+			matched = rc == ALLOW ? DENY : ALLOW;
+		    } else {
+			matched = rc;
+		    }
+		}
 		alias_put(a);
 		break;
 	    }
@@ -119,7 +124,8 @@ userlist_matches(struct sudoers_parse_tree *parse_tree, const struct passwd *pw,
     debug_decl(userlist_matches, SUDOERS_DEBUG_MATCH);
 
     TAILQ_FOREACH_REVERSE(m, list, member_list, entries) {
-	if ((matched = user_matches(parse_tree, pw, m)) != UNSPEC)
+	matched = user_matches(parse_tree, pw, m);
+	if (SPECIFIED(matched))
 	    break;
     }
     debug_return_int(matched);
@@ -189,8 +195,13 @@ runaslist_matches(struct sudoers_parse_tree *parse_tree,
 			if (a != NULL) {
 			    rc = runaslist_matches(parse_tree, &a->members,
 				&empty, matching_user, NULL);
-			    if (rc != UNSPEC)
-				user_matched = m->negated ? !rc : rc;
+			    if (SPECIFIED(rc)) {
+				if (m->negated) {
+				    user_matched = rc == ALLOW ? DENY : ALLOW;
+				} else {
+				    user_matched = rc;
+				}
+			    }
 			    alias_put(a);
 			    break;
 			}
@@ -205,7 +216,7 @@ runaslist_matches(struct sudoers_parse_tree *parse_tree,
 			    user_matched = !m->negated;
 			break;
 		}
-		if (user_matched != UNSPEC) {
+		if (SPECIFIED(user_matched)) {
 		    if (matching_user != NULL && m->type != ALIAS)
 			*matching_user = m;
 		    break;
@@ -233,8 +244,13 @@ runaslist_matches(struct sudoers_parse_tree *parse_tree,
 			if (a != NULL) {
 			    rc = runaslist_matches(parse_tree, &empty,
 				&a->members, NULL, matching_group);
-			    if (rc != UNSPEC)
-				group_matched = m->negated ? !rc : rc;
+			    if (SPECIFIED(rc)) {
+				if (m->negated) {
+				    group_matched = rc == ALLOW ? DENY : ALLOW;
+				} else {
+				    group_matched = rc;
+				}
+			    }
 			    alias_put(a);
 			    break;
 			}
@@ -244,14 +260,14 @@ runaslist_matches(struct sudoers_parse_tree *parse_tree,
 			    group_matched = !m->negated;
 			break;
 		}
-		if (group_matched != UNSPEC) {
+		if (SPECIFIED(group_matched)) {
 		    if (matching_group != NULL && m->type != ALIAS)
 			*matching_group = m;
 		    break;
 		}
 	    }
 	}
-	if (group_matched == UNSPEC) {
+	if (!SPECIFIED(group_matched)) {
 	    struct gid_list *runas_groups;
 	    /*
 	     * The runas group was not explicitly allowed by sudoers.
@@ -291,7 +307,7 @@ hostlist_matches_int(struct sudoers_parse_tree *parse_tree,
 {
     struct member *m;
     int matched = UNSPEC;
-    debug_decl(hostlist_matches, SUDOERS_DEBUG_MATCH);
+    debug_decl(hostlist_matches_int, SUDOERS_DEBUG_MATCH);
 
     TAILQ_FOREACH_REVERSE(m, list, member_list, entries) {
 	matched = host_matches(parse_tree, pw, lhost, shost, m);
@@ -346,8 +362,13 @@ host_matches(struct sudoers_parse_tree *parse_tree, const struct passwd *pw,
 		/* XXX */
 		int rc = hostlist_matches_int(parse_tree, pw, lhost, shost,
 		    &a->members);
-		if (rc != UNSPEC)
-		    matched = m->negated ? !rc : rc;
+		if (SPECIFIED(rc)) {
+		    if (m->negated) {
+			matched = rc == ALLOW ? DENY : ALLOW;
+		    } else {
+			matched = rc;
+		    }
+		}
 		alias_put(a);
 		break;
 	    }
@@ -375,7 +396,7 @@ cmndlist_matches(struct sudoers_parse_tree *parse_tree,
 
     TAILQ_FOREACH_REVERSE(m, list, member_list, entries) {
 	matched = cmnd_matches(parse_tree, m, runchroot, info);
-	if (matched != UNSPEC)
+	if (SPECIFIED(matched))
 	    break;
     }
     debug_return_int(matched);
@@ -410,8 +431,47 @@ cmnd_matches(struct sudoers_parse_tree *parse_tree, const struct member *m,
 	    a = alias_get(parse_tree, m->name, CMNDALIAS);
 	    if (a != NULL) {
 		rc = cmndlist_matches(parse_tree, &a->members, runchroot, info);
-		if (rc != UNSPEC)
+		if (SPECIFIED(rc)) {
 		    matched = m->negated ? !rc : rc;
+		}
+		alias_put(a);
+	    }
+	    break;
+    }
+    debug_return_int(matched);
+}
+
+/*
+ * Like cmnd_matches() but only matches against the ALL command.
+ * Returns ALLOW, DENY or UNSPEC.
+ */
+int
+cmnd_matches_all(const struct sudoers_parse_tree *parse_tree,
+    const struct member *m, const char *runchroot, struct cmnd_info *info)
+{
+    const bool negated = m->negated;
+    struct sudo_command *c;
+    int matched = UNSPEC;
+    struct alias *a;
+    debug_decl(cmnd_matches_all, SUDOERS_DEBUG_MATCH);
+
+    switch (m->type) {
+	case ALL:
+	    c = (struct sudo_command *)m->name;
+	    if (command_matches(c->cmnd, c->args, runchroot, info, &c->digests))
+		matched = negated ? DENY : ALLOW;
+	    break;
+	case ALIAS:
+	    a = alias_get((struct sudoers_parse_tree *)parse_tree, m->name, CMNDALIAS);
+	    if (a != NULL) {
+		TAILQ_FOREACH_REVERSE(m, &a->members, member_list, entries) {
+		    matched = cmnd_matches_all(parse_tree, m, runchroot, info);
+		    if (SPECIFIED(matched)) {
+			if (negated)
+			    matched = matched == ALLOW ? DENY : ALLOW;
+			break;
+		    }
+		}
 		alias_put(a);
 	    }
 	    break;
diff --git a/plugins/sudoers/parse.h b/plugins/sudoers/parse.h
index 711bf87e1..9d9f32b56 100644
--- a/plugins/sudoers/parse.h
+++ b/plugins/sudoers/parse.h
@@ -24,20 +24,40 @@
 #include "sudo_queue.h"
 
 /* Characters that must be quoted in sudoers. */
-#define SUDOERS_QUOTED	":\\,=#\""
+#define SUDOERS_QUOTED		":,=#\""
+#define SUDOERS_QUOTED_CMD	":,= \t#"
+#define SUDOERS_QUOTED_ARG	":,=#"
 
 /* Returns true if string 's' contains meta characters. */
 #define has_meta(s)	(strpbrk(s, "\\?*[]") != NULL)
 
+/* Match by name, not inode, when fuzzing. */
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
+# define SUDOERS_NAME_MATCH
+#endif
+
+/* Allowed by policy (rowhammer resistent). */
+#undef ALLOW
+#define ALLOW	 0x52a2925	/* 0101001010100010100100100101 */
+
+/* Denied by policy (rowhammer resistent). */
+#undef DENY
+#define DENY	 0xad5d6da	/* 1010110101011101011011011010 */
+
+/* Neither allowed, nor denied. */
 #undef UNSPEC
 #define UNSPEC	-1
-#undef DENY
-#define DENY	 0
-#undef ALLOW
-#define ALLOW	 1
+
+/* Tag implied by root access (SETENV only). */
 #undef IMPLIED
 #define IMPLIED	 2
 
+/*
+ * We must explicitly check against ALLOW and DENY instead testing
+ * that the value is not UNSPEC to avoid potential ROWHAMMER issues.
+ */
+#define SPECIFIED(_v)	((_v) == ALLOW || (_v) == DENY)
+
 /*
  * Initialize all tags to UNSPEC.
  */
@@ -84,7 +104,7 @@
  * Returns true if the specified tag is not UNSPEC or IMPLIED, else false.
  */
 #define TAG_SET(tt) \
-    ((tt) != UNSPEC && (tt) != IMPLIED)
+    ((tt) == true || (tt) == false)
 
 /*
  * Returns true if any tags set in nt differ between ot and nt, else false.
