--- a/x/mongo/driver/auth/internal/gssapi/gss_wrapper.c
+++ b/x/mongo/driver/auth/internal/gssapi/gss_wrapper.c
@@ -72,8 +72,8 @@ int gssapi_error_desc(
             free(*desc);
         }
 
-        *desc = malloc(desc_buffer.length+1);
-        memcpy(*desc, desc_buffer.value, desc_buffer.length+1);
+        *desc = calloc(1, desc_buffer.length + 1);
+        memcpy(*desc, desc_buffer.value, desc_buffer.length);
 
         gss_release_buffer(&local_min_stat, &desc_buffer);
     }
@@ -144,8 +144,8 @@ int gssapi_client_username(
         return GSSAPI_ERROR;
     }
 
-	*username = malloc(name_buffer.length+1);
-	memcpy(*username, name_buffer.value, name_buffer.length+1);
+    *username = calloc(1, name_buffer.length + 1);
+    memcpy(*username, name_buffer.value, name_buffer.length);
 
     gss_release_buffer(&ignored, &name_buffer);
     gss_release_name(&ignored, &name);
