diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/lang/java/avro/src/main/java/org/apache/avro/Schema.java target_draft/lang/java/avro/src/main/java/org/apache/avro/Schema.java
--- a/lang/java/avro/src/main/java/org/apache/avro/Schema.java	2026-03-09 08:53:44.650103114 +0000
+++ b/lang/java/avro/src/main/java/org/apache/avro/Schema.java	2026-03-09 09:36:49.729493434 +0000
@@ -1210,8 +1210,7 @@
 
     public FixedSchema(Name name, String doc, int size) {
       super(Type.FIXED, name, doc);
-      if (size < 0)
-        throw new IllegalArgumentException("Invalid fixed size: " + size);
+      SystemLimitException.checkMaxBytesLength(size);
       this.size = size;
     }
 
diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/lang/java/avro/src/main/java/org/apache/avro/SystemLimitException.java target_draft/lang/java/avro/src/main/java/org/apache/avro/SystemLimitException.java
--- a/lang/java/avro/src/main/java/org/apache/avro/SystemLimitException.java	1970-01-01 00:00:00.000000000 +0000
+++ b/lang/java/avro/src/main/java/org/apache/avro/SystemLimitException.java	2026-03-09 09:41:34.751386121 +0000
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.avro;
+
+import org.slf4j.LoggerFactory;
+
+/**
+ * Thrown to prevent making large allocations when reading potentially
+ * pathological input data from an untrusted source.
+ * <p/>
+ * The following system properties can be set to limit the size of bytes,
+ * strings and collection types to be allocated:
+ * <ul>
+ * <li><tt>org.apache.avro.limits.byte.maxLength</tt></li> limits the maximum
+ * size of <tt>byte</tt> types.</li>
+ * <li><tt>org.apache.avro.limits.collectionItems.maxLength</tt></li> limits the
+ * maximum number of <tt>map</tt> and <tt>list</tt> items that can be read at
+ * once single sequence.</li>
+ * <li><tt>org.apache.avro.limits.string.maxLength</tt></li> limits the maximum
+ * size of <tt>string</tt> types.</li>
+ * </ul>
+ *
+ * The default is to permit sizes up to {@link #MAX_ARRAY_VM_LIMIT}.
+ */
+public class SystemLimitException extends AvroRuntimeException {
+
+  /**
+   * The maximum length of array to allocate (unless necessary). Some VMs reserve
+   * some header words in an array. Attempts to allocate larger arrays may result
+   * in {@code OutOfMemoryError: Requested array size exceeds VM limit}
+   *
+   * @see <a href="https://bugs.openjdk.org/browse/JDK-8246725">JDK-8246725</a>
+   */
+  // VisibleForTesting
+  static final int MAX_ARRAY_VM_LIMIT = Integer.MAX_VALUE - 8;
+
+  public static final String MAX_BYTES_LENGTH_PROPERTY = "org.apache.avro.limits.bytes.maxLength";
+  public static final String MAX_COLLECTION_LENGTH_PROPERTY = "org.apache.avro.limits.collectionItems.maxLength";
+  public static final String MAX_STRING_LENGTH_PROPERTY = "org.apache.avro.limits.string.maxLength";
+
+  private static int maxBytesLength = MAX_ARRAY_VM_LIMIT;
+  private static int maxCollectionLength = MAX_ARRAY_VM_LIMIT;
+  private static int maxStringLength = MAX_ARRAY_VM_LIMIT;
+
+  static {
+    resetLimits();
+  }
+
+  public SystemLimitException(String message) {
+    super(message);
+  }
+
+  /**
+   * Get an integer value stored in a system property, used to configure the
+   * system behaviour of decoders
+   *
+   * @param property     The system property to fetch
+   * @param defaultValue The value to use if the system property is not present or
+   *                     parsable as an int
+   * @return The value from the system property
+   */
+  private static int getLimitFromProperty(String property, int defaultValue) {
+    String o = System.getProperty(property);
+    int i = defaultValue;
+    if (o != null) {
+      try {
+        i = Integer.parseUnsignedInt(o);
+      } catch (NumberFormatException nfe) {
+        LoggerFactory.getLogger(SystemLimitException.class).warn("Could not parse property " + property + ": " + o,
+            nfe);
+      }
+    }
+    return i;
+  }
+
+  /**
+   * Check to ensure that reading the bytes is within the specified limits.
+   *
+   * @param length The proposed size of the bytes to read
+   * @return The size of the bytes if and only if it is within the limit and
+   *         non-negative.
+   * @throws UnsupportedOperationException if reading the datum would allocate a
+   *                                       collection that the Java VM would be
+   *                                       unable to handle
+   * @throws SystemLimitException          if the decoding should fail because it
+   *                                       would otherwise result in an allocation
+   *                                       exceeding the set limit
+   * @throws AvroRuntimeException          if the length is negative
+   */
+  public static int checkMaxBytesLength(long length) {
+    if (length < 0) {
+      throw new AvroRuntimeException("Malformed data. Length is negative: " + length);
+    }
+    if (length > MAX_ARRAY_VM_LIMIT) {
+      throw new UnsupportedOperationException(
+          "Cannot read arrays longer than " + MAX_ARRAY_VM_LIMIT + " bytes in Java library");
+    }
+    if (length > maxBytesLength) {
+      throw new SystemLimitException("Bytes length " + length + " exceeds maximum allowed");
+    }
+    return (int) length;
+  }
+
+  /**
+   * Check to ensure that reading the specified number of items remains within the
+   * specified limits.
+   *
+   * @param existing The number of elements items read in the collection
+   * @param items    The next number of items to read. In normal usage, this is
+   *                 always a positive, permitted value. Negative and zero values
+   *                 have a special meaning in Avro decoding.
+   * @return The total number of items in the collection if and only if it is
+   *         within the limit and non-negative.
+   * @throws UnsupportedOperationException if reading the items would allocate a
+   *                                       collection that the Java VM would be
+   *                                       unable to handle
+   * @throws SystemLimitException          if the decoding should fail because it
+   *                                       would otherwise result in an allocation
+   *                                       exceeding the set limit
+   * @throws AvroRuntimeException          if the length is negative
+   */
+  public static int checkMaxCollectionLength(long existing, long items) {
+    long length = existing + items;
+    if (existing < 0) {
+      throw new AvroRuntimeException("Malformed data. Length is negative: " + existing);
+    }
+    if (items < 0) {
+      throw new AvroRuntimeException("Malformed data. Length is negative: " + items);
+    }
+    if (length > MAX_ARRAY_VM_LIMIT || length < existing) {
+      throw new UnsupportedOperationException(
+          "Cannot read collections larger than " + MAX_ARRAY_VM_LIMIT + " items in Java library");
+    }
+    if (length > maxCollectionLength) {
+      throw new SystemLimitException("Collection length " + length + " exceeds maximum allowed");
+    }
+    return (int) length;
+  }
+
+  /**
+   * Check to ensure that reading the string size is within the specified limits.
+   *
+   * @param length The proposed size of the string to read
+   * @return The size of the string if and only if it is within the limit and
+   *         non-negative.
+   * @throws UnsupportedOperationException if reading the items would allocate a
+   *                                       collection that the Java VM would be
+   *                                       unable to handle
+   * @throws SystemLimitException          if the decoding should fail because it
+   *                                       would otherwise result in an allocation
+   *                                       exceeding the set limit
+   * @throws AvroRuntimeException          if the length is negative
+   */
+  public static int checkMaxStringLength(long length) {
+    if (length < 0) {
+      throw new AvroRuntimeException("Malformed data. Length is negative: " + length);
+    }
+    if (length > MAX_ARRAY_VM_LIMIT) {
+      throw new UnsupportedOperationException("Cannot read strings longer than " + MAX_ARRAY_VM_LIMIT + " bytes");
+    }
+    if (length > maxStringLength) {
+      throw new SystemLimitException("String length " + length + " exceeds maximum allowed");
+    }
+    return (int) length;
+  }
+
+  /** Reread the limits from the system properties. */
+  // VisibleForTesting
+  static void resetLimits() {
+    maxBytesLength = getLimitFromProperty(MAX_BYTES_LENGTH_PROPERTY, MAX_ARRAY_VM_LIMIT);
+    maxCollectionLength = getLimitFromProperty(MAX_COLLECTION_LENGTH_PROPERTY, MAX_ARRAY_VM_LIMIT);
+    maxStringLength = getLimitFromProperty(MAX_STRING_LENGTH_PROPERTY, MAX_ARRAY_VM_LIMIT);
+  }
+}
diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/lang/java/avro/src/main/java/org/apache/avro/io/BinaryDecoder.java target_draft/lang/java/avro/src/main/java/org/apache/avro/io/BinaryDecoder.java
--- a/lang/java/avro/src/main/java/org/apache/avro/io/BinaryDecoder.java	2026-03-09 08:53:44.653103067 +0000
+++ b/lang/java/avro/src/main/java/org/apache/avro/io/BinaryDecoder.java	2026-03-09 09:56:23.829616938 +0000
@@ -25,6 +25,7 @@
 
 import org.apache.avro.AvroRuntimeException;
 import org.apache.avro.InvalidNumberEncodingException;
+import org.apache.avro.SystemLimitException;
 import org.apache.avro.util.Utf8;
 
 /**
@@ -38,16 +39,18 @@
  * 'direct'.
  *
  * @see Encoder
+ * @see SystemLimitException
  */
 
 public class BinaryDecoder extends Decoder {
 
   /**
-   * The maximum size of array to allocate. Some VMs reserve some header words in
-   * an array. Attempts to allocate larger arrays may result in OutOfMemoryError:
-   * Requested array size exceeds VM limit
+   * When reading a collection (MAP or ARRAY), this keeps track of the number of
+   * elements to ensure that the
+   * {@link SystemLimitException#checkMaxCollectionLength} constraint is
+   * respected.
    */
-  private static final long MAX_ARRAY_SIZE = (long) Integer.MAX_VALUE - 8L;
+  private long collectionCount = 0L;
 
   private ByteSource source = null;
   // we keep the buffer and its state variables in this class and not in a
@@ -276,17 +279,11 @@
 
   @Override
   public Utf8 readString(Utf8 old) throws IOException {
-    long length = readLong();
-    if (length > MAX_ARRAY_SIZE) {
-      throw new UnsupportedOperationException("Cannot read strings longer than " + MAX_ARRAY_SIZE + " bytes");
-    }
-    if (length < 0L) {
-      throw new AvroRuntimeException("Malformed data. Length is negative: " + length);
-    }
+    int length = SystemLimitException.checkMaxStringLength(readLong());
     Utf8 result = (old != null ? old : new Utf8());
-    result.setByteLength((int) length);
-    if (0L != length) {
-      doReadBytes(result.getBytes(), 0, (int) length);
+    result.setByteLength(length);
+    if (0 != length) {
+      doReadBytes(result.getBytes(), 0, length);
     }
     return result;
   }
@@ -305,7 +302,7 @@
 
   @Override
   public ByteBuffer readBytes(ByteBuffer old) throws IOException {
-    int length = readInt();
+    int length = SystemLimitException.checkMaxBytesLength(readLong());
     final ByteBuffer result;
     if (old != null && length <= old.capacity()) {
       result = old;
@@ -427,12 +424,15 @@
 
   @Override
   public long readArrayStart() throws IOException {
-    return doReadItemCount();
+    collectionCount = SystemLimitException.checkMaxCollectionLength(0L, doReadItemCount());
+    return collectionCount;
   }
 
   @Override
   public long arrayNext() throws IOException {
-    return doReadItemCount();
+    long length = doReadItemCount();
+    collectionCount = SystemLimitException.checkMaxCollectionLength(collectionCount, length);
+    return length;
   }
 
   @Override
@@ -442,12 +442,15 @@
 
   @Override
   public long readMapStart() throws IOException {
-    return doReadItemCount();
+    collectionCount = SystemLimitException.checkMaxCollectionLength(0L, doReadItemCount());
+    return collectionCount;
   }
 
   @Override
   public long mapNext() throws IOException {
-    return doReadItemCount();
+    long length = doReadItemCount();
+    collectionCount = SystemLimitException.checkMaxCollectionLength(collectionCount, length);
+    return length;
   }
 
   @Override
diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/lang/java/avro/src/main/java/org/apache/avro/io/DirectBinaryDecoder.java target_draft/lang/java/avro/src/main/java/org/apache/avro/io/DirectBinaryDecoder.java
--- a/lang/java/avro/src/main/java/org/apache/avro/io/DirectBinaryDecoder.java	2026-03-09 08:53:44.654103052 +0000
+++ b/lang/java/avro/src/main/java/org/apache/avro/io/DirectBinaryDecoder.java	2026-03-09 10:31:49.883656836 +0000
@@ -23,6 +23,7 @@
 import java.nio.ByteBuffer;
 
 import org.apache.avro.InvalidNumberEncodingException;
+import org.apache.avro.SystemLimitException;
 import org.apache.avro.util.ByteBufferInputStream;
 
 /**
@@ -67,7 +68,6 @@
         return bbi.readBuffer(length);
       }
     }
-
   }
 
   private ByteReader byteReader;
@@ -156,7 +156,7 @@
   @Override
   public ByteBuffer readBytes(ByteBuffer old) throws IOException {
     int length = readInt();
-    return byteReader.read(old, length);
+    return byteReader.read(old, SystemLimitException.checkMaxBytesLength(length));
   }
 
   @Override
diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/lang/java/avro/src/main/java/org/apache/avro/util/Utf8.java target_draft/lang/java/avro/src/main/java/org/apache/avro/util/Utf8.java
--- a/lang/java/avro/src/main/java/org/apache/avro/util/Utf8.java	2026-03-09 08:53:44.657103005 +0000
+++ b/lang/java/avro/src/main/java/org/apache/avro/util/Utf8.java	2026-03-09 10:15:19.063586923 +0000
@@ -19,9 +19,8 @@
 
 import java.nio.charset.StandardCharsets;
 
-import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.SystemLimitException;
 import org.apache.avro.io.BinaryData;
-import org.slf4j.LoggerFactory;
 
 /**
  * A Utf8 string. Unlike {@link String}, instances are mutable. This is more
@@ -54,8 +53,11 @@
   }
 
   public Utf8(String string) {
-    this.bytes = getBytesFor(string);
-    this.length = bytes.length;
+    byte[] bytes = getBytesFor(string);
+    int length = bytes.length;
+    SystemLimitException.checkMaxStringLength(length);
+    this.bytes = bytes;
+    this.length = length;
     this.string = string;
   }
 
@@ -67,8 +69,10 @@
   }
 
   public Utf8(byte[] bytes) {
+    int length = bytes.length;
+    SystemLimitException.checkMaxStringLength(length);
     this.bytes = bytes;
-    this.length = bytes.length;
+    this.length = length;
   }
 
   /**
@@ -109,9 +113,7 @@
    * length does not change, as this also clears the cached String.
    */
   public Utf8 setByteLength(int newLength) {
-    if (newLength > MAX_LENGTH) {
-      throw new AvroRuntimeException("String length " + newLength + " exceeds maximum allowed");
-    }
+    SystemLimitException.checkMaxStringLength(newLength);
     if (this.bytes.length < newLength) {
       byte[] newBytes = new byte[newLength];
       System.arraycopy(bytes, 0, newBytes, 0, this.length);
@@ -124,8 +126,11 @@
 
   /** Set to the contents of a String. */
   public Utf8 set(String string) {
-    this.bytes = getBytesFor(string);
-    this.length = bytes.length;
+    byte[] bytes = getBytesFor(string);
+    int length = bytes.length;
+    SystemLimitException.checkMaxStringLength(length);
+    this.bytes = bytes;
+    this.length = length;
     this.string = string;
     return this;
   }
diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/lang/java/avro/src/test/java/org/apache/avro/TestFixed.java target_draft/lang/java/avro/src/test/java/org/apache/avro/TestFixed.java
--- a/lang/java/avro/src/test/java/org/apache/avro/TestFixed.java	2026-03-09 08:53:44.658102989 +0000
+++ b/lang/java/avro/src/test/java/org/apache/avro/TestFixed.java	2026-03-09 10:17:56.361433949 +0000
@@ -33,4 +33,16 @@
     Assert.assertArrayEquals(new byte[16], (byte[]) field.defaultVal());
   }
 
+  @Test
+  void fixedLengthOutOfLimit() {
+    Exception ex = assertThrows(UnsupportedOperationException.class,
+        () -> Schema.createFixed("oversize", "doc", "space", Integer.MAX_VALUE));
+    assertEquals(TestSystemLimitException.ERROR_VM_LIMIT_BYTES, ex.getMessage());
+  }
+
+  @Test
+  void fixedNegativeLength() {
+    Exception ex = assertThrows(AvroRuntimeException.class, () -> Schema.createFixed("negative", "doc", "space", -1));
+    assertEquals(TestSystemLimitException.ERROR_NEGATIVE, ex.getMessage());
+  }
 }
diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/lang/java/avro/src/test/java/org/apache/avro/TestSystemLimitException.java target_draft/lang/java/avro/src/test/java/org/apache/avro/TestSystemLimitException.java
--- a/lang/java/avro/src/test/java/org/apache/avro/TestSystemLimitException.java	1970-01-01 00:00:00.000000000 +0000
+++ b/lang/java/avro/src/test/java/org/apache/avro/TestSystemLimitException.java	2026-03-09 10:21:07.857305178 +0000
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2017 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.avro;
+
+import static org.apache.avro.SystemLimitException.*;
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.function.Function;
+
+public class TestSystemLimitException {
+
+  /** Delegated here for package visibility. */
+  public static final int MAX_ARRAY_VM_LIMIT = SystemLimitException.MAX_ARRAY_VM_LIMIT;
+
+  public static final String ERROR_NEGATIVE = "Malformed data. Length is negative: -1";
+  public static final String ERROR_VM_LIMIT_BYTES = "Cannot read arrays longer than " + MAX_ARRAY_VM_LIMIT
+      + " bytes in Java library";
+  public static final String ERROR_VM_LIMIT_COLLECTION = "Cannot read collections larger than " + MAX_ARRAY_VM_LIMIT
+      + " items in Java library";
+  public static final String ERROR_VM_LIMIT_STRING = "Cannot read strings longer than " + MAX_ARRAY_VM_LIMIT + " bytes";
+
+  /** Delegated here for package visibility. */
+  public static void resetLimits() {
+    SystemLimitException.resetLimits();
+  }
+
+  @AfterEach
+  void reset() {
+    System.clearProperty(MAX_BYTES_LENGTH_PROPERTY);
+    System.clearProperty(MAX_COLLECTION_LENGTH_PROPERTY);
+    System.clearProperty(MAX_STRING_LENGTH_PROPERTY);
+    resetLimits();
+  }
+
+  /**
+   * A helper method that tests the consistent limit handling from system
+   * properties.
+   *
+   * @param f                The function to be tested.
+   * @param sysProperty      The system property used to control the custom limit.
+   * @param errorVmLimit     The error message used when the property would be
+   *                         over the VM limit.
+   * @param errorCustomLimit The error message used when the property would be
+   *                         over the custom limit of 1000.
+   */
+  void helpCheckSystemLimits(Function<Long, Integer> f, String sysProperty, String errorVmLimit,
+      String errorCustomLimit) {
+    // Correct values pass through
+    assertEquals(0, f.apply(0L));
+    assertEquals(1024, f.apply(1024L));
+    assertEquals(MAX_ARRAY_VM_LIMIT, f.apply((long) MAX_ARRAY_VM_LIMIT));
+
+    // Values that exceed the default system limits throw exceptions
+    Exception ex = assertThrows(UnsupportedOperationException.class, () -> f.apply(Long.MAX_VALUE));
+    assertEquals(errorVmLimit, ex.getMessage());
+    ex = assertThrows(UnsupportedOperationException.class, () -> f.apply((long) MAX_ARRAY_VM_LIMIT + 1));
+    assertEquals(errorVmLimit, ex.getMessage());
+    ex = assertThrows(AvroRuntimeException.class, () -> f.apply(-1L));
+    assertEquals(ERROR_NEGATIVE, ex.getMessage());
+
+    // Setting the system property to provide a custom limit.
+    System.setProperty(sysProperty, Long.toString(1000L));
+    resetLimits();
+
+    // Correct values pass through
+    assertEquals(0, f.apply(0L));
+    assertEquals(102, f.apply(102L));
+
+    // Values that exceed the custom system limits throw exceptions
+    ex = assertThrows(UnsupportedOperationException.class, () -> f.apply((long) MAX_ARRAY_VM_LIMIT + 1));
+    assertEquals(errorVmLimit, ex.getMessage());
+    ex = assertThrows(SystemLimitException.class, () -> f.apply(1024L));
+    assertEquals(errorCustomLimit, ex.getMessage());
+    ex = assertThrows(AvroRuntimeException.class, () -> f.apply(-1L));
+    assertEquals(ERROR_NEGATIVE, ex.getMessage());
+  }
+
+  @Test
+  void testCheckMaxBytesLength() {
+    helpCheckSystemLimits(SystemLimitException::checkMaxBytesLength, MAX_BYTES_LENGTH_PROPERTY, ERROR_VM_LIMIT_BYTES,
+        "Bytes length 1024 exceeds maximum allowed");
+  }
+
+  @Test
+  void testCheckMaxCollectionLengthFromZero() {
+    helpCheckSystemLimits(l -> checkMaxCollectionLength(0L, l), MAX_COLLECTION_LENGTH_PROPERTY,
+        ERROR_VM_LIMIT_COLLECTION, "Collection length 1024 exceeds maximum allowed");
+  }
+
+  @Test
+  void testCheckMaxStringLength() {
+    helpCheckSystemLimits(SystemLimitException::checkMaxStringLength, MAX_STRING_LENGTH_PROPERTY, ERROR_VM_LIMIT_STRING,
+        "String length 1024 exceeds maximum allowed");
+  }
+
+  @Test
+  void testCheckMaxCollectionLengthFromNonZero() {
+    // Correct values pass through
+    assertEquals(10, checkMaxCollectionLength(10L, 0L));
+    assertEquals(MAX_ARRAY_VM_LIMIT, checkMaxCollectionLength(10L, MAX_ARRAY_VM_LIMIT - 10L));
+    assertEquals(MAX_ARRAY_VM_LIMIT, checkMaxCollectionLength(MAX_ARRAY_VM_LIMIT - 10L, 10L));
+
+    // Values that exceed the default system limits throw exceptions
+    Exception ex = assertThrows(UnsupportedOperationException.class,
+        () -> checkMaxCollectionLength(10L, MAX_ARRAY_VM_LIMIT - 9L));
+    assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+    ex = assertThrows(UnsupportedOperationException.class,
+        () -> checkMaxCollectionLength(SystemLimitException.MAX_ARRAY_VM_LIMIT - 9L, 10L));
+    assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    ex = assertThrows(UnsupportedOperationException.class, () -> checkMaxCollectionLength(10L, Long.MAX_VALUE - 10L));
+    assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+    ex = assertThrows(UnsupportedOperationException.class, () -> checkMaxCollectionLength(Long.MAX_VALUE - 10L, 10L));
+    assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    // Overflow that adds to negative
+    ex = assertThrows(UnsupportedOperationException.class, () -> checkMaxCollectionLength(10L, Long.MAX_VALUE));
+    assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+    ex = assertThrows(UnsupportedOperationException.class, () -> checkMaxCollectionLength(Long.MAX_VALUE, 10L));
+    assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    ex = assertThrows(AvroRuntimeException.class, () -> checkMaxCollectionLength(10L, -1L));
+    assertEquals(ERROR_NEGATIVE, ex.getMessage());
+    ex = assertThrows(AvroRuntimeException.class, () -> checkMaxCollectionLength(-1L, 10L));
+    assertEquals(ERROR_NEGATIVE, ex.getMessage());
+
+    // Setting the system property to provide a custom limit.
+    System.setProperty(MAX_COLLECTION_LENGTH_PROPERTY, Long.toString(1000L));
+    resetLimits();
+
+    // Correct values pass through
+    assertEquals(10, checkMaxCollectionLength(10L, 0L));
+    assertEquals(102, checkMaxCollectionLength(10L, 92L));
+    assertEquals(102, checkMaxCollectionLength(92L, 10L));
+
+    // Values that exceed the custom system limits throw exceptions
+    ex = assertThrows(UnsupportedOperationException.class, () -> checkMaxCollectionLength(MAX_ARRAY_VM_LIMIT, 1));
+    assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+    ex = assertThrows(UnsupportedOperationException.class, () -> checkMaxCollectionLength(1, MAX_ARRAY_VM_LIMIT));
+    assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    ex = assertThrows(SystemLimitException.class, () -> checkMaxCollectionLength(999, 25));
+    assertEquals("Collection length 1024 exceeds maximum allowed", ex.getMessage());
+    ex = assertThrows(SystemLimitException.class, () -> checkMaxCollectionLength(25, 999));
+    assertEquals("Collection length 1024 exceeds maximum allowed", ex.getMessage());
+  }
+}
\ No newline at end of file
diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/lang/java/avro/src/test/java/org/apache/avro/io/TestBinaryDecoder.java target_draft/lang/java/avro/src/test/java/org/apache/avro/io/TestBinaryDecoder.java
--- a/lang/java/avro/src/test/java/org/apache/avro/io/TestBinaryDecoder.java	2026-03-09 08:53:44.660102958 +0000
+++ b/lang/java/avro/src/test/java/org/apache/avro/io/TestBinaryDecoder.java	2026-03-09 10:24:46.499151005 +0000
@@ -25,6 +25,7 @@
 
 import org.apache.avro.AvroRuntimeException;
 import org.apache.avro.Schema;
+import org.apache.avro.SystemLimitException;
 import org.apache.avro.generic.GenericDatumReader;
 import org.apache.avro.generic.GenericDatumWriter;
 import org.apache.avro.util.ByteBufferInputStream;
@@ -37,6 +38,11 @@
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 import org.junit.runners.Parameterized.Parameters;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import static org.apache.avro.TestSystemLimitException.*;
 
 @RunWith(Parameterized.class)
 public class TestBinaryDecoder {
@@ -76,6 +82,39 @@
     return factory.binaryDecoder(bytes, null);
   }
 
+  private BinaryDecoder newDecoder(byte[] bytes, boolean useDirect) {
+    return this.newDecoder(bytes, null, useDirect);
+  }
+
+  private BinaryDecoder newDecoder(byte[] bytes, BinaryDecoder reuse, boolean useDirect) {
+    if (useDirect) {
+      return this.factory.directBinaryDecoder(new ByteArrayInputStream(bytes), reuse);
+    } else {
+      return factory.binaryDecoder(bytes, reuse);
+    }
+  }
+
+  /**
+   * Create a decoder for simulating reading corrupt, unexpected or out-of-bounds
+   * data.
+   *
+   * @return a {@link org.apache.avro.io.BinaryDecoder that has been initialized
+   *         on a byte array containing the sequence of encoded longs in order.
+   */
+  private BinaryDecoder newDecoder(boolean useDirect, long... values) throws IOException {
+    try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+      BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(baos, null);
+      for (long v : values)
+        encoder.writeLong(v);
+      encoder.flush();
+      return newDecoder(baos.toByteArray(), useDirect);
+    }
+  }
+
+  private static void resetLimits() {
+    SystemLimitException.resetLimits();
+  }
+
   /** Verify EOFException throw at EOF */
 
   @Test(expected = EOFException.class)
@@ -359,6 +398,196 @@
     bd.readString();
   }
 
+  @ParameterizedTest
+  @ValueSource(booleans = { true, false })
+  public void testStringNegativeLength(boolean useDirect) throws IOException {
+    Exception ex = Assertions.assertThrows(AvroRuntimeException.class, this.newDecoder(useDirect, -1L)::readString);
+    Assertions.assertEquals(ERROR_NEGATIVE, ex.getMessage());
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = { true, false })
+  public void testStringVmMaxSize(boolean useDirect) throws IOException {
+    Exception ex = Assertions.assertThrows(UnsupportedOperationException.class,
+        newDecoder(useDirect, MAX_ARRAY_VM_LIMIT + 1L)::readString);
+    Assertions.assertEquals(ERROR_VM_LIMIT_STRING, ex.getMessage());
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = { true, false })
+  public void testStringMaxCustom(boolean useDirect) throws IOException {
+    try {
+      System.setProperty(SystemLimitException.MAX_STRING_LENGTH_PROPERTY, Long.toString(128));
+      resetLimits();
+      Exception ex = Assertions.assertThrows(SystemLimitException.class, newDecoder(useDirect, 129)::readString);
+      Assertions.assertEquals("String length 129 exceeds maximum allowed", ex.getMessage());
+    } finally {
+      System.clearProperty(SystemLimitException.MAX_STRING_LENGTH_PROPERTY);
+      resetLimits();
+    }
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = { true, false })
+  public void testBytesNegativeLength(boolean useDirect) throws IOException {
+    Exception ex = Assertions.assertThrows(AvroRuntimeException.class,
+        () -> this.newDecoder(useDirect, -1).readBytes(null));
+    Assertions.assertEquals(ERROR_NEGATIVE, ex.getMessage());
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = { true, false })
+  public void testBytesVmMaxSize(boolean useDirect) throws IOException {
+    Exception ex = Assertions.assertThrows(UnsupportedOperationException.class,
+        () -> this.newDecoder(useDirect, MAX_ARRAY_VM_LIMIT + 1).readBytes(null));
+    Assertions.assertEquals(ERROR_VM_LIMIT_BYTES, ex.getMessage());
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = { true, false })
+  public void testBytesMaxCustom(boolean useDirect) throws IOException {
+    try {
+      System.setProperty(SystemLimitException.MAX_BYTES_LENGTH_PROPERTY, Long.toString(128));
+      resetLimits();
+      Exception ex = Assertions.assertThrows(SystemLimitException.class,
+          () -> newDecoder(useDirect, 129).readBytes(null));
+      Assertions.assertEquals("Bytes length 129 exceeds maximum allowed", ex.getMessage());
+    } finally {
+      System.clearProperty(SystemLimitException.MAX_BYTES_LENGTH_PROPERTY);
+      resetLimits();
+    }
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = { true, false })
+  public void testArrayVmMaxSize(boolean useDirect) throws IOException {
+    // At start
+    Exception ex = Assertions.assertThrows(UnsupportedOperationException.class,
+        () -> this.newDecoder(useDirect, MAX_ARRAY_VM_LIMIT + 1).readArrayStart());
+    Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    // Next
+    ex = Assertions.assertThrows(UnsupportedOperationException.class,
+        () -> this.newDecoder(useDirect, MAX_ARRAY_VM_LIMIT + 1).arrayNext());
+    Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    // An OK reads followed by an overflow
+    Decoder bd = newDecoder(useDirect, MAX_ARRAY_VM_LIMIT - 100, Long.MAX_VALUE);
+    Assertions.assertEquals(MAX_ARRAY_VM_LIMIT - 100, bd.readArrayStart());
+    ex = Assertions.assertThrows(UnsupportedOperationException.class, bd::arrayNext);
+    Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    // Two OK reads followed by going over the VM limit.
+    bd = newDecoder(useDirect, MAX_ARRAY_VM_LIMIT - 100, 100, 1);
+    Assertions.assertEquals(MAX_ARRAY_VM_LIMIT - 100, bd.readArrayStart());
+    Assertions.assertEquals(100, bd.arrayNext());
+    ex = Assertions.assertThrows(UnsupportedOperationException.class, bd::arrayNext);
+    Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    // Two OK reads followed by going over the VM limit, where negative numbers are
+    // followed by the byte length of the items. For testing, the 999 values are
+    // read but ignored.
+    bd = newDecoder(useDirect, 100 - MAX_ARRAY_VM_LIMIT, 999, -100, 999, 1);
+    Assertions.assertEquals(MAX_ARRAY_VM_LIMIT - 100, bd.readArrayStart());
+    Assertions.assertEquals(100, bd.arrayNext());
+    ex = Assertions.assertThrows(UnsupportedOperationException.class, bd::arrayNext);
+    Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = { true, false })
+  public void testArrayMaxCustom(boolean useDirect) throws IOException {
+    try {
+      System.setProperty(SystemLimitException.MAX_COLLECTION_LENGTH_PROPERTY, Long.toString(128));
+      resetLimits();
+      Exception ex = Assertions.assertThrows(UnsupportedOperationException.class,
+          () -> newDecoder(useDirect, MAX_ARRAY_VM_LIMIT + 1).readArrayStart());
+      Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+      // Two OK reads followed by going over the custom limit.
+      Decoder bd = newDecoder(useDirect, 118, 10, 1);
+      Assertions.assertEquals(118, bd.readArrayStart());
+      Assertions.assertEquals(10, bd.arrayNext());
+      ex = Assertions.assertThrows(SystemLimitException.class, bd::arrayNext);
+      Assertions.assertEquals("Collection length 129 exceeds maximum allowed", ex.getMessage());
+
+      // Two OK reads followed by going over the VM limit, where negative numbers are
+      // followed by the byte length of the items. For testing, the 999 values are
+      // read but ignored.
+      bd = newDecoder(useDirect, -118, 999, -10, 999, 1);
+      Assertions.assertEquals(118, bd.readArrayStart());
+      Assertions.assertEquals(10, bd.arrayNext());
+      ex = Assertions.assertThrows(SystemLimitException.class, bd::arrayNext);
+      Assertions.assertEquals("Collection length 129 exceeds maximum allowed", ex.getMessage());
+
+    } finally {
+      System.clearProperty(SystemLimitException.MAX_COLLECTION_LENGTH_PROPERTY);
+      resetLimits();
+    }
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = { true, false })
+  public void testMapVmMaxSize(boolean useDirect) throws IOException {
+    // At start
+    Exception ex = Assertions.assertThrows(UnsupportedOperationException.class,
+        () -> this.newDecoder(useDirect, MAX_ARRAY_VM_LIMIT + 1).readMapStart());
+    Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    // Next
+    ex = Assertions.assertThrows(UnsupportedOperationException.class,
+        () -> this.newDecoder(useDirect, MAX_ARRAY_VM_LIMIT + 1).mapNext());
+    Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    // Two OK reads followed by going over the VM limit.
+    Decoder bd = newDecoder(useDirect, MAX_ARRAY_VM_LIMIT - 100, 100, 1);
+    Assertions.assertEquals(MAX_ARRAY_VM_LIMIT - 100, bd.readMapStart());
+    Assertions.assertEquals(100, bd.mapNext());
+    ex = Assertions.assertThrows(UnsupportedOperationException.class, bd::mapNext);
+    Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+    // Two OK reads followed by going over the VM limit, where negative numbers are
+    // followed by the byte length of the items. For testing, the 999 values are
+    // read but ignored.
+    bd = newDecoder(useDirect, 100 - MAX_ARRAY_VM_LIMIT, 999, -100, 999, 1);
+    Assertions.assertEquals(MAX_ARRAY_VM_LIMIT - 100, bd.readMapStart());
+    Assertions.assertEquals(100, bd.mapNext());
+    ex = Assertions.assertThrows(UnsupportedOperationException.class, bd::mapNext);
+    Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = { true, false })
+  public void testMapMaxCustom(boolean useDirect) throws IOException {
+    try {
+      System.setProperty(SystemLimitException.MAX_COLLECTION_LENGTH_PROPERTY, Long.toString(128));
+      resetLimits();
+      Exception ex = Assertions.assertThrows(UnsupportedOperationException.class,
+          () -> newDecoder(useDirect, MAX_ARRAY_VM_LIMIT + 1).readMapStart());
+      Assertions.assertEquals(ERROR_VM_LIMIT_COLLECTION, ex.getMessage());
+
+      // Two OK reads followed by going over the custom limit.
+      Decoder bd = newDecoder(useDirect, 118, 10, 1);
+      Assertions.assertEquals(118, bd.readMapStart());
+      Assertions.assertEquals(10, bd.mapNext());
+      ex = Assertions.assertThrows(SystemLimitException.class, bd::mapNext);
+      Assertions.assertEquals("Collection length 129 exceeds maximum allowed", ex.getMessage());
+
+      // Two OK reads followed by going over the VM limit, where negative numbers are
+      // followed by the byte length of the items. For testing, the 999 values are
+      // read but ignored.
+      bd = newDecoder(useDirect, -118, 999, -10, 999, 1);
+      Assertions.assertEquals(118, bd.readMapStart());
+      Assertions.assertEquals(10, bd.mapNext());
+      ex = Assertions.assertThrows(SystemLimitException.class, bd::mapNext);
+      Assertions.assertEquals("Collection length 129 exceeds maximum allowed", ex.getMessage());
+
+    } finally {
+      System.clearProperty(SystemLimitException.MAX_COLLECTION_LENGTH_PROPERTY);
+      resetLimits();
+    }
+  }
+
   @Test(expected = EOFException.class)
   public void testIntTooShort() throws IOException {
     byte[] badint = new byte[4];
diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/lang/java/avro/src/test/java/org/apache/avro/util/TestUtf8.java target_draft/lang/java/avro/src/test/java/org/apache/avro/util/TestUtf8.java
--- a/lang/java/avro/src/test/java/org/apache/avro/util/TestUtf8.java	2026-03-09 08:53:44.662102927 +0000
+++ b/lang/java/avro/src/test/java/org/apache/avro/util/TestUtf8.java	2026-03-09 10:28:37.685794645 +0000
@@ -19,9 +19,12 @@
 
 import static org.junit.Assert.assertSame;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThrows;
 
 import java.nio.charset.StandardCharsets;
 
+import org.apache.avro.SystemLimitException;
+import org.apache.avro.TestSystemLimitException;
 import org.junit.Test;
 
 public class TestUtf8 {
@@ -48,4 +51,24 @@
     assertEquals(4, u.getByteLength());
     assertSame(content, u.getBytes());
   }
+
+  @Test
+  void oversizeUtf8() {
+    Utf8 u = new Utf8();
+    u.setByteLength(1024);
+    assertEquals(1024, u.getByteLength());
+    assertThrows(UnsupportedOperationException.class,
+        () -> u.setByteLength(TestSystemLimitException.MAX_ARRAY_VM_LIMIT + 1));
+
+    try {
+      System.setProperty(SystemLimitException.MAX_STRING_LENGTH_PROPERTY, Long.toString(1000L));
+      TestSystemLimitException.resetLimits();
+
+      Exception ex = assertThrows(SystemLimitException.class, () -> u.setByteLength(1024));
+      assertEquals("String length 1024 exceeds maximum allowed", ex.getMessage());
+    } finally {
+      System.clearProperty(SystemLimitException.MAX_STRING_LENGTH_PROPERTY);
+      TestSystemLimitException.resetLimits();
+    }
+  }
 }

--- a/lang/java/avro/src/main/java/org/apache/avro/util/Utf8.java	2026-03-09 16:50:25.121070982 +0530
+++ b/lang/java/avro/src/main/java/org/apache/avro/util/Utf8.java	2026-03-09 16:51:39.239019313 +0530
@@ -29,23 +29,8 @@
  * as a single instance may be reused.
  */
 public class Utf8 implements Comparable<Utf8>, CharSequence {
-  private static final String MAX_LENGTH_PROPERTY = "org.apache.avro.limits.string.maxLength";
-  private static final int MAX_LENGTH;
   private static final byte[] EMPTY = new byte[0];
 
-  static {
-    String o = System.getProperty(MAX_LENGTH_PROPERTY);
-    int i = Integer.MAX_VALUE;
-    if (o != null) {
-      try {
-        i = Integer.parseUnsignedInt(o);
-      } catch (NumberFormatException nfe) {
-        LoggerFactory.getLogger(Utf8.class).warn("Could not parse property " + MAX_LENGTH_PROPERTY + ": " + o, nfe);
-      }
-    }
-    MAX_LENGTH = i;
-  }
-
   private byte[] bytes = EMPTY;
   private int length;
   private String string;

--- a/lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java
+++ b/lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java
@@ -24,6 +24,9 @@
 import org.apache.avro.io.ResolvingDecoder;
 import org.apache.avro.util.ClassUtils;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 
 /**
  * {@link org.apache.avro.io.DatumReader DatumReader} for generated Java
@@ -30,6 +33,16 @@
  * classes.
  */
 public class SpecificDatumReader<T> extends GenericDatumReader<T> {
+
+  public static final String[] SERIALIZABLE_PACKAGES;
+
+  static {
+    SERIALIZABLE_PACKAGES = System.getProperty("org.apache.avro.SERIALIZABLE_PACKAGES",
+        "java.lang,java.math,java.io,java.net,org.apache.avro.reflect").split(",");
+  }
+
+  private final List<String> trustedPackages = new ArrayList<>();
+
   public SpecificDatumReader() {
     this(null, null, SpecificData.get());
   }
@@ -54,6 +67,7 @@
    */
   public SpecificDatumReader(Schema writer, Schema reader, SpecificData data) {
     super(writer, reader, data);
+    trustedPackages.addAll(Arrays.asList(SERIALIZABLE_PACKAGES));
   }
 
   /** Construct given a {@link SpecificData}. */
@@ -101,12 +115,43 @@
     if (name == null)
       return null;
     try {
-      return ClassUtils.forName(getData().getClassLoader(), name);
+      Class clazz = ClassUtils.forName(getData().getClassLoader(), name);
+      checkSecurity(clazz);
+      return clazz;
     } catch (ClassNotFoundException e) {
       throw new AvroRuntimeException(e);
     }
   }
 
+  private boolean trustAllPackages() {
+    return (trustedPackages.size() == 1 && "*".equals(trustedPackages.get(0)));
+  }
+
+  private void checkSecurity(Class clazz) throws ClassNotFoundException {
+    if (trustAllPackages() || clazz.isPrimitive()) {
+      return;
+    }
+
+    boolean found = false;
+    Package thePackage = clazz.getPackage();
+    if (thePackage != null) {
+      for (String trustedPackage : getTrustedPackages()) {
+        if (thePackage.getName().equals(trustedPackage) || thePackage.getName().startsWith(trustedPackage + ".")) {
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        throw new SecurityException("Forbidden " + clazz
+            + "! This class is not trusted to be included in Avro schema using java-class. Please set org.apache.avro.SERIALIZABLE_PACKAGES system property with the packages you trust.");
+      }
+    }
+  }
+
+  public final List<String> getTrustedPackages() {
+    return trustedPackages;
+  }
+
   @Override
   protected Object readRecord(Object old, Schema expected, ResolvingDecoder in) throws IOException {
     SpecificData data = getSpecificData();
