--- a/xmlschemas.c
+++ b/xmlschemas.c
@@ -23647,6 +23647,7 @@
 			}
 			if (xmlSchemaItemListAdd(bind->dupls, bind->nodeTable[j]) == -1)
 			    goto internal_error;
+			dupls = (xmlSchemaPSVIIDCNodePtr *) bind->dupls->items;
 			/*
 			* Remove the duplicate entry from the IDC node-table.
 			*/
@@ -23863,6 +23864,8 @@
 				goto internal_error;
 			}
 			xmlSchemaItemListAdd(parBind->dupls, parNode);
+			dupls = (xmlSchemaPSVIIDCNodePtr *)
+			    parBind->dupls->items;
 		    } else {
 			/*
 			* Add the node-table entry (node and key-sequence) of
--- a/pattern.c
+++ b/pattern.c
@@ -1178,10 +1178,10 @@ xmlCompileStepPattern(xmlPatParserContextPtr ctxt) {
 			    goto error;
 			}
 		    } else {
-			PUSH(XML_OP_CHILD, token, URL);
+			PUSH(XML_OP_ELEM, token, URL);
 		    }
 		} else
-		    PUSH(XML_OP_CHILD, name, NULL);
+		    PUSH(XML_OP_ELEM, name, NULL);
 		return;
 	    } else if (xmlStrEqual(name, (const xmlChar *) "attribute")) {
 		XML_PAT_FREE_STRING(ctxt, name)
--- /dev/null
+++ b/result/pattern/child
@@ -0,0 +1,5 @@
+Node /a matches pattern child::a
+Node /a/b matches pattern child::b
+Node /a/b/c matches pattern child::c
+Node /a/b matches pattern a/child::b
+Node /a/b/c matches pattern a/b/child::c
--- /dev/null
+++ b/test/pattern/child.pat
@@ -0,0 +1,5 @@
+child::a
+child::b
+child::c
+a/child::b
+a/b/child::c
--- /dev/null
+++ b/test/pattern/child.xml
@@ -0,0 +1,5 @@
+<a>
+  <b>
+    <c d="attr">text</c>
+  </b>
+</a>
--- a/valid.c
+++ b/valid.c
@@ -5259,25 +5259,26 @@ xmlSnprintfElements(char *buf, int size, xmlNodePtr node, int glob) {
 	    return;
 	}
         switch (cur->type) {
-            case XML_ELEMENT_NODE:
+            case XML_ELEMENT_NODE: {
+                int qnameLen = xmlStrlen(cur->name);
+
+                if ((cur->ns != NULL) && (cur->ns->prefix != NULL))
+                    qnameLen += xmlStrlen(cur->ns->prefix) + 1;
+                if (size - len < qnameLen + 10) {
+                    if ((size - len > 4) && (buf[len - 1] != '.'))
+                        strcat(buf, " ...");
+                    return;
+                }
 		if ((cur->ns != NULL) && (cur->ns->prefix != NULL)) {
-		    if (size - len < xmlStrlen(cur->ns->prefix) + 10) {
-			if ((size - len > 4) && (buf[len - 1] != '.'))
-			    strcat(buf, " ...");
-			return;
-		    }
 		    strcat(buf, (char *) cur->ns->prefix);
 		    strcat(buf, ":");
 		}
-                if (size - len < xmlStrlen(cur->name) + 10) {
-		    if ((size - len > 4) && (buf[len - 1] != '.'))
-			strcat(buf, " ...");
-		    return;
-		}
-	        strcat(buf, (char *) cur->name);
+                if (cur->name != NULL)
+	            strcat(buf, (char *) cur->name);
 		if (cur->next != NULL)
 		    strcat(buf, " ");
 		break;
+            }
             case XML_TEXT_NODE:
 		if (xmlIsBlankNode(cur))
 		    break;
