diff --git a/lib/patch/parse.js b/lib/patch/parse.js
index f150104..07b292a 100644
--- a/lib/patch/parse.js
+++ b/lib/patch/parse.js
@@ -64,14 +64,13 @@ function parsePatch(uniDiff) {
 
 
   function parseFileHeader(index) {
-    var fileHeader = /^(---|\+\+\+)\s+(.*)$/.exec(diffstr[i]);
-
-    if (fileHeader) {
-      var keyPrefix = fileHeader[1] === '---' ? 'old' : 'new';
-      var data = fileHeader[2].split('\t', 2);
+    var fileHeaderMatch = (/^(---|\+\+\+)\s+/).exec(diffstr[i]);
+    if (fileHeaderMatch) {
+      var keyPrefix = fileHeaderMatch[1] === '---' ? 'old' : 'new';
+      var data = diffstr[i].substring(3).trim().split('\t', 2);
       var fileName = data[0].replace(/\\\\/g, '\\');
 
-      if (/^".*"$/.test(fileName)) {
+      if (fileName.startsWith('"') && fileName.endsWith('"')) {
         fileName = fileName.substr(1, fileName.length - 2);
       }
 
