--- a/src/clj/clojure/core.clj
+++ b/src/clj/clojure/core.clj
@@ -3003,7 +3003,8 @@
   [n x] (take n (repeat x)))

 (defn iterate
-  "Returns a lazy sequence of x, (f x), (f (f x)) etc. f must be free of side-effects"
+  "Returns a lazy (infinite!) sequence of x, (f x), (f (f x)) etc.
+  f must be free of side-effects"
   {:added "1.0"
    :static true}
   [f x] (clojure.lang.Iterate/create f x) )
--- a/src/jvm/clojure/lang/Cycle.java
+++ b/src/jvm/clojure/lang/Cycle.java
@@ -91,4 +91,13 @@
             s = all;
     }
 }
+
+public int hashCode(){
+    throw new UnsupportedOperationException();
+}
+
+public int hasheq(){
+    throw new UnsupportedOperationException();
+}
+
 }
--- a/src/jvm/clojure/lang/Iterate.java
+++ b/src/jvm/clojure/lang/Iterate.java
@@ -82,4 +82,12 @@
         v = f.invoke(v);
     }
 }
+
+public int hashCode(){
+    throw new UnsupportedOperationException();
+}
+
+public int hasheq(){
+    throw new UnsupportedOperationException();
+}
 }
--- a/src/jvm/clojure/lang/Repeat.java
+++ b/src/jvm/clojure/lang/Repeat.java
@@ -95,4 +95,17 @@
     }
 }

+public int hashCode(){
+    if(count <= 0)
+        throw new UnsupportedOperationException();
+    else
+        return super.hashCode();
+}
+
+public int hasheq(){
+    if(count <= 0)
+        throw new UnsupportedOperationException();
+    else
+        return super.hasheq();
+}
 }
--- a/test/clojure/test_clojure/sequences.clj
+++ b/test/clojure/test_clojure/sequences.clj
@@ -1361,3 +1361,13 @@
     (doseq [i (range 100)]
       (is (= s1 (concat (subseq s2 < i) (subseq s2 >= i))))
       (is (= (reverse s1) (concat (rsubseq s2 >= i) (rsubseq s2 < i)))))))
+
+(deftest infinite-seq-hash
+  (are [e] (thrown? Exception (.hashCode ^Object e))
+    (iterate identity nil)
+    (cycle [1])
+    (repeat 1))
+  (are [e] (thrown? Exception (.hasheq ^clojure.lang.IHashEq e))
+    (iterate identity nil)
+    (cycle [1])
+    (repeat 1)))
--- a/test/clojure/test_clojure/serialization.clj
+++ b/test/clojure/test_clojure/serialization.clj
@@ -173,4 +173,11 @@

     ;; stateful seqs
     (enumeration-seq (java.util.Collections/enumeration (range 50)))
-    (iterator-seq (.iterator (range 50)))))
\ No newline at end of file
+    (iterator-seq (.iterator (range 50)))))
+
+;; necessary for CVE-2024-22871
+(deftest CLJ-2839
+  (are [e] (thrown? Exception (.hashCode ^Object (-> e serialize deserialize)))
+    (repeat 1)
+    (iterate identity nil)
+    (cycle [1])))

--- a/src/clj/clojure/core.clj
+++ b/src/clj/clojure/core.clj
@@ -6981,7 +6981,7 @@
       version-string (.getProperty properties "version")
       [_ major minor incremental qualifier snapshot]
       (re-matches
-       #"(\d+)\.(\d+)\.(\d+)(?:-([a-zA-Z0-9_]+))?(?:-(SNAPSHOT))?"
+       #"(\d+)\.(\d+)\.(\d+)(?:-([a-zA-Z0-9_.]+))?(?:-(SNAPSHOT))?"
        version-string)
       clojure-version {:major       (Integer/valueOf ^String major)
                        :minor       (Integer/valueOf ^String minor)
