diff --git a/lib/validate.js b/lib/validate.js
index e4dc151..035aa4f 100644
--- a/lib/validate.js
+++ b/lib/validate.js
@@ -209,7 +209,7 @@ var validate = exports._validate = function(/*Any*/instance,/*Object*/schema,/*O
 			}
 			
 			for(var i in objTypeDef){ 
-				if(objTypeDef.hasOwnProperty(i)){
+				if(objTypeDef.hasOwnProperty(i) && i != '__proto__'){
 					var value = instance[i];
 					// skip _not_ specified properties
 					if (value === undefined && options.existingOnly) continue;
-- 
2.43.0

