From bed4bc674ce652e61e2a7aaa3b267482ed565113 Mon Sep 17 00:00:00 2001
Date: Thu, 5 Mar 2026 12:49:28 +0530
Subject: [PATCH] patch

---
 setuptools/_vendor/jaraco/context.py | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/setuptools/_vendor/jaraco/context.py b/setuptools/_vendor/jaraco/context.py
index 61b2713..6d6092b 100644
--- a/setuptools/_vendor/jaraco/context.py
+++ b/setuptools/_vendor/jaraco/context.py
@@ -62,12 +62,19 @@ def tarball(
     try:
         req = urllib.request.urlopen(url)
         with tarfile.open(fileobj=req, mode='r|*') as tf:
-            tf.extractall(path=target_dir, filter=strip_first_component)
+            tf.extractall(path=target_dir, filter=_default_filter)
         yield target_dir
     finally:
         shutil.rmtree(target_dir)
 
 
+def _compose_tarfile_filters(*filters):
+    def compose_two(f1, f2):
+        return lambda member, path: f1(f2(member, path), path)
+
+    return functools.reduce(compose_two, filters, lambda member, path: member)
+
+
 def strip_first_component(
     member: tarfile.TarInfo,
     path,
@@ -76,6 +83,9 @@ def strip_first_component(
     return member
 
 
+_default_filter = _compose_tarfile_filters(tarfile.data_filter, strip_first_component)
+
+
 def _compose(*cmgrs):
     """
     Compose any number of dependent context managers into a single one.
-- 
2.43.0

