--- a/common/iobuf.c
+++ b/common/iobuf.c
@@ -2043,6 +2043,8 @@
 	rc = 0;
       else
       {
+	size_t tmplen;
+
 	/* If no buffered data and drain buffer has been setup, and drain
 	 * buffer is largish, read data directly to drain buffer. */
 	if (a->d.len == 0
@@ -2055,8 +2057,10 @@
 	      log_debug ("iobuf-%d.%d: underflow: A->FILTER (%lu bytes, to external drain)\n",
 			 a->no, a->subno, (ulong)len);
 
+	    tmplen = len;  /* Used to check for bugs in the filter.  */
 	    rc = a->filter (a->filter_ov, IOBUFCTRL_UNDERFLOW, a->chain,
 			    a->e_d.buf, &len);
+	    log_assert (len <= tmplen);
 	    a->e_d.used = len;
 	    len = 0;
 	  }
@@ -2066,8 +2070,10 @@
 	      log_debug ("iobuf-%d.%d: underflow: A->FILTER (%lu bytes)\n",
 			 a->no, a->subno, (ulong)len);
 
+	    tmplen = len;
 	    rc = a->filter (a->filter_ov, IOBUFCTRL_UNDERFLOW, a->chain,
 			    &a->d.buf[a->d.len], &len);
+	    log_assert (len <= tmplen);
 	  }
       }
       a->d.len += len;
--- a/g10/armor.c
+++ b/g10/armor.c
@@ -1302,8 +1302,8 @@
 	n = 0;
 	if( afx->buffer_len ) {
             /* Copy the data from AFX->BUFFER to BUF.  */
-	    for(; n < size && afx->buffer_pos < afx->buffer_len; n++ )
-		buf[n++] = afx->buffer[afx->buffer_pos++];
+            for(; n < size && afx->buffer_pos < afx->buffer_len;)
+                buf[n++] = afx->buffer[afx->buffer_pos++];
 	    if( afx->buffer_pos >= afx->buffer_len )
 		afx->buffer_len = 0;
 	}
