--- a/dist/types/jxl.js
+++ b/dist/types/jxl.js
@@ -24,7 +24,7 @@ function extractPartialStreams(input) {
         if (!jxlpBox)
             break;
         partialStreams.push(input.slice(jxlpBox.offset + 12, jxlpBox.offset + jxlpBox.size));
-        offset = jxlpBox.offset + jxlpBox.size;
+        offset = jxlpBox.offset + (jxlpBox.size > 0 ? jxlpBox.size : 8);
     }
     return partialStreams;
 }
