--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/DefaultGrpcServerObservationConvention.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/DefaultGrpcServerObservationConvention.java
@@ -48,14 +48,18 @@ public class DefaultGrpcServerObservationConvention implements GrpcServerObserva
     public KeyValues getLowCardinalityKeyValues(GrpcServerObservationContext context) {
         KeyValue statusCodeKeyValue = context.getStatusCode() != null
                 ? LowCardinalityKeyNames.STATUS_CODE.withValue(context.getStatusCode().name()) : STATUS_CODE_UNKNOWN;
+        return KeyValues.of(statusCodeKeyValue, LowCardinalityKeyNames.METHOD.withValue(context.getMethodName()),
+                LowCardinalityKeyNames.SERVICE.withValue(context.getServiceName()),
+                LowCardinalityKeyNames.METHOD_TYPE.withValue(context.getMethodType().name()));
+    }
+
+    @Override
+    public KeyValues getHighCardinalityKeyValues(GrpcServerObservationContext context) {
         KeyValue peerNameKeyValue = context.getPeerName() != null
                 ? LowCardinalityKeyNames.PEER_NAME.withValue(context.getPeerName()) : PEER_NAME_UNKNOWN;
         KeyValue peerPortKeyValue = context.getPeerPort() != null
                 ? LowCardinalityKeyNames.PEER_PORT.withValue(context.getPeerPort().toString()) : PEER_PORT_UNKNOWN;
-        return KeyValues.of(statusCodeKeyValue, peerNameKeyValue, peerPortKeyValue,
-                LowCardinalityKeyNames.METHOD.withValue(context.getMethodName()),
-                LowCardinalityKeyNames.SERVICE.withValue(context.getServiceName()),
-                LowCardinalityKeyNames.METHOD_TYPE.withValue(context.getMethodType().name()));
+        return KeyValues.of(peerNameKeyValue, peerPortKeyValue);
     }
 
 }
--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/GrpcObservationDocumentation.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/GrpcObservationDocumentation.java
@@ -84,12 +84,22 @@ public enum GrpcObservationDocumentation implements ObservationDocumentation {
                 return "grpc.status_code";
             }
         },
+
+        /**
+         * This is low cardinality only for client observations, for server observations,
+         * it should be considered as high cardinality.
+         */
         PEER_NAME {
             @Override
             public String asString() {
                 return "net.peer.name";
             }
         },
+
+        /**
+         * This is low cardinality only for client observations, for server observations,
+         * it should be considered as high cardinality.
+         */
         PEER_PORT {
             @Override
             public String asString() {
--- a/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/grpc/GrpcObservationTest.java
+++ b/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/grpc/GrpcObservationTest.java
@@ -173,9 +173,11 @@ class GrpcObservationTest {
             assertThat(observationRegistry).hasAnObservation(observationContextAssert -> {
                 observationContextAssert.hasNameEqualTo("grpc.client");
                 assertCommonKeyValueNames(observationContextAssert);
+                assertClientKeyValues(observationContextAssert);
             }).hasAnObservation(observationContextAssert -> {
                 observationContextAssert.hasNameEqualTo("grpc.server");
                 assertCommonKeyValueNames(observationContextAssert);
+                assertServerKeyValues(observationContextAssert);
             });
             // end::assertion[]
             verifyHeaders();
@@ -357,6 +359,7 @@ class GrpcObservationTest {
                 observationContextAssert.doesNotHaveError();
                 observationContextAssert.hasNameEqualTo("grpc.client");
                 assertCommonKeyValueNames(observationContextAssert);
+                assertClientKeyValues(observationContextAssert);
             });
         }
 
@@ -397,9 +400,11 @@ class GrpcObservationTest {
         assertThat(observationRegistry).hasAnObservation(observationContextAssert -> {
             observationContextAssert.hasNameEqualTo("grpc.client");
             assertCommonKeyValueNames(observationContextAssert);
+            assertClientKeyValues(observationContextAssert);
         }).hasAnObservation(observationContextAssert -> {
             observationContextAssert.hasNameEqualTo("grpc.server");
             assertCommonKeyValueNames(observationContextAssert);
+            assertServerKeyValues(observationContextAssert);
         });
     }
 
@@ -512,6 +517,7 @@ class GrpcObservationTest {
             assertThat(observationRegistry).hasAnObservation(observationContextAssert -> {
                 observationContextAssert.hasNameEqualTo("grpc.server").hasError();
                 assertCommonKeyValueNames(observationContextAssert);
+                assertServerKeyValues(observationContextAssert);
             });
         }
 
@@ -635,9 +641,9 @@ class GrpcObservationTest {
             assertThat(serverContext.getMethodName()).isEqualTo(methodName);
             assertThat(serverContext.getFullMethodName()).isEqualTo(contextualName);
             assertThat(serverContext.getMethodType()).isEqualTo(methodType);
-            assertThat(serverContext.getAuthority()).isEqualTo("localhost");
-            assertThat(serverContext.getPeerName()).isEqualTo("localhost");
-            assertThat(serverContext.getPeerPort()).isEqualTo(-1);
+            assertThat(serverContext.getAuthority()).isEqualTo("test-host:1234");
+            assertThat(serverContext.getPeerName()).isEqualTo("test-host");
+            assertThat(serverContext.getPeerPort()).isEqualTo(1234);
         });
     }
 
@@ -649,6 +655,7 @@ class GrpcObservationTest {
             assertThat(clientContext.getMethodName()).isEqualTo(methodName);
             assertThat(clientContext.getFullMethodName()).isEqualTo(contextualName);
             assertThat(clientContext.getMethodType()).isEqualTo(methodType);
+            // TODO: this should be test-host:1234, test interceptor needs a fix
             assertThat(clientContext.getAuthority()).isEqualTo("localhost");
             assertThat(clientContext.getPeerName()).isEqualTo("localhost");
             assertThat(clientContext.getPeerPort()).isEqualTo(-1);
@@ -665,6 +672,21 @@ class GrpcObservationTest {
                     GrpcObservationDocumentation.LowCardinalityKeyNames.STATUS_CODE.asString());
     }
 
+    void assertClientKeyValues(ObservationContextAssert<?> observationContextAssert) {
+        // TODO: value should be asserted, test interceptor needs a fix
+        observationContextAssert
+            .hasLowCardinalityKeyValueWithKey(GrpcObservationDocumentation.LowCardinalityKeyNames.PEER_NAME.asString())
+            .hasLowCardinalityKeyValueWithKey(GrpcObservationDocumentation.LowCardinalityKeyNames.PEER_PORT.asString());
+    }
+
+    void assertServerKeyValues(ObservationContextAssert<?> observationContextAssert) {
+        observationContextAssert
+            .hasHighCardinalityKeyValue(GrpcObservationDocumentation.LowCardinalityKeyNames.PEER_NAME.asString(),
+                    "test-host")
+            .hasHighCardinalityKeyValue(GrpcObservationDocumentation.LowCardinalityKeyNames.PEER_PORT.asString(),
+                    "1234");
+    }
+
     // GRPC service extending SimpleService and provides echo implementation.
     static class EchoService extends SimpleServiceImplBase {
 
@@ -836,13 +858,15 @@ class GrpcObservationTest {
 
     static class ClientHeaderInterceptor implements ClientInterceptor {
 
+        private static final String AUTHORITY = "test-host:1234";
+
         private static final Metadata.Key<String> CLIENT_KEY = Metadata.Key.of("client",
                 Metadata.ASCII_STRING_MARSHALLER);
 
         @Override
         public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
                 CallOptions callOptions, Channel next) {
-            ClientCall<ReqT, RespT> call = next.newCall(method, callOptions);
+            ClientCall<ReqT, RespT> call = next.newCall(method, callOptions.withAuthority(AUTHORITY));
             return new SimpleForwardingClientCall<>(call) {
                 @Override
                 public void start(ClientCall.Listener<RespT> responseListener, Metadata headers) {
