--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/DefaultGrpcServerObservationConvention.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/DefaultGrpcServerObservationConvention.java
@@ -48,14 +48,18 @@ public class DefaultGrpcServerObservationConvention implements GrpcServerObserva
     public KeyValues getLowCardinalityKeyValues(GrpcServerObservationContext context) {
         KeyValue statusCodeKeyValue = context.getStatusCode() != null
                 ? LowCardinalityKeyNames.STATUS_CODE.withValue(context.getStatusCode().name()) : STATUS_CODE_UNKNOWN;
+        return KeyValues.of(statusCodeKeyValue, LowCardinalityKeyNames.METHOD.withValue(context.getMethodName()),
+                LowCardinalityKeyNames.SERVICE.withValue(context.getServiceName()),
+                LowCardinalityKeyNames.METHOD_TYPE.withValue(context.getMethodType().name()));
+    }
+
+    @Override
+    public KeyValues getHighCardinalityKeyValues(GrpcServerObservationContext context) {
         KeyValue peerNameKeyValue = context.getPeerName() != null
                 ? LowCardinalityKeyNames.PEER_NAME.withValue(context.getPeerName()) : PEER_NAME_UNKNOWN;
         KeyValue peerPortKeyValue = context.getPeerPort() != null
                 ? LowCardinalityKeyNames.PEER_PORT.withValue(context.getPeerPort().toString()) : PEER_PORT_UNKNOWN;
-        return KeyValues.of(statusCodeKeyValue, peerNameKeyValue, peerPortKeyValue,
-                LowCardinalityKeyNames.METHOD.withValue(context.getMethodName()),
-                LowCardinalityKeyNames.SERVICE.withValue(context.getServiceName()),
-                LowCardinalityKeyNames.METHOD_TYPE.withValue(context.getMethodType().name()));
+        return KeyValues.of(peerNameKeyValue, peerPortKeyValue);
     }
 
 }
--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/GrpcObservationDocumentation.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/GrpcObservationDocumentation.java
@@ -84,12 +84,22 @@ public enum GrpcObservationDocumentation implements ObservationDocumentation {
                 return "grpc.status_code";
             }
         },
+
+        /**
+         * This is low cardinality only for client observations, for server observations,
+         * it should be considered as high cardinality.
+         */
         PEER_NAME {
             @Override
             public String asString() {
                 return "net.peer.name";
             }
         },
+
+        /**
+         * This is low cardinality only for client observations, for server observations,
+         * it should be considered as high cardinality.
+         */
         PEER_PORT {
             @Override
             public String asString() {
--- a/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/grpc/GrpcObservationTest.java
+++ b/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/grpc/GrpcObservationTest.java
@@ -173,9 +173,11 @@ class GrpcObservationTest {
             assertThat(observationRegistry).hasAnObservation(observationContextAssert -> {
                 observationContextAssert.hasNameEqualTo("grpc.client");
                 assertCommonKeyValueNames(observationContextAssert);
+                assertClientKeyValues(observationContextAssert);
             }).hasAnObservation(observationContextAssert -> {
                 observationContextAssert.hasNameEqualTo("grpc.server");
                 assertCommonKeyValueNames(observationContextAssert);
+                assertServerKeyValues(observationContextAssert);
             });
             // end::assertion[]
             verifyHeaders();
@@ -357,6 +359,7 @@ class GrpcObservationTest {
                 observationContextAssert.doesNotHaveError();
                 observationContextAssert.hasNameEqualTo("grpc.client");
                 assertCommonKeyValueNames(observationContextAssert);
+                assertClientKeyValues(observationContextAssert);
             });
         }
 
@@ -397,9 +400,11 @@ class GrpcObservationTest {
         assertThat(observationRegistry).hasAnObservation(observationContextAssert -> {
             observationContextAssert.hasNameEqualTo("grpc.client");
             assertCommonKeyValueNames(observationContextAssert);
+            assertClientKeyValues(observationContextAssert);
         }).hasAnObservation(observationContextAssert -> {
             observationContextAssert.hasNameEqualTo("grpc.server");
             assertCommonKeyValueNames(observationContextAssert);
+            assertServerKeyValues(observationContextAssert);
         });
     }
 
@@ -512,6 +517,7 @@ class GrpcObservationTest {
             assertThat(observationRegistry).hasAnObservation(observationContextAssert -> {
                 observationContextAssert.hasNameEqualTo("grpc.server").hasError();
                 assertCommonKeyValueNames(observationContextAssert);
+                assertServerKeyValues(observationContextAssert);
             });
         }
 
@@ -635,9 +641,9 @@ class GrpcObservationTest {
             assertThat(serverContext.getMethodName()).isEqualTo(methodName);
             assertThat(serverContext.getFullMethodName()).isEqualTo(contextualName);
             assertThat(serverContext.getMethodType()).isEqualTo(methodType);
-            assertThat(serverContext.getAuthority()).isEqualTo("localhost");
-            assertThat(serverContext.getPeerName()).isEqualTo("localhost");
-            assertThat(serverContext.getPeerPort()).isEqualTo(-1);
+            assertThat(serverContext.getAuthority()).isEqualTo("test-host:1234");
+            assertThat(serverContext.getPeerName()).isEqualTo("test-host");
+            assertThat(serverContext.getPeerPort()).isEqualTo(1234);
         });
     }
 
@@ -649,6 +655,7 @@ class GrpcObservationTest {
             assertThat(clientContext.getMethodName()).isEqualTo(methodName);
             assertThat(clientContext.getFullMethodName()).isEqualTo(contextualName);
             assertThat(clientContext.getMethodType()).isEqualTo(methodType);
+            // TODO: this should be test-host:1234, test interceptor needs a fix
             assertThat(clientContext.getAuthority()).isEqualTo("localhost");
             assertThat(clientContext.getPeerName()).isEqualTo("localhost");
             assertThat(clientContext.getPeerPort()).isEqualTo(-1);
@@ -665,6 +672,21 @@ class GrpcObservationTest {
                     GrpcObservationDocumentation.LowCardinalityKeyNames.STATUS_CODE.asString());
     }
 
+    void assertClientKeyValues(ObservationContextAssert<?> observationContextAssert) {
+        // TODO: value should be asserted, test interceptor needs a fix
+        observationContextAssert
+            .hasLowCardinalityKeyValueWithKey(GrpcObservationDocumentation.LowCardinalityKeyNames.PEER_NAME.asString())
+            .hasLowCardinalityKeyValueWithKey(GrpcObservationDocumentation.LowCardinalityKeyNames.PEER_PORT.asString());
+    }
+
+    void assertServerKeyValues(ObservationContextAssert<?> observationContextAssert) {
+        observationContextAssert
+            .hasHighCardinalityKeyValue(GrpcObservationDocumentation.LowCardinalityKeyNames.PEER_NAME.asString(),
+                    "test-host")
+            .hasHighCardinalityKeyValue(GrpcObservationDocumentation.LowCardinalityKeyNames.PEER_PORT.asString(),
+                    "1234");
+    }
+
     // GRPC service extending SimpleService and provides echo implementation.
     static class EchoService extends SimpleServiceImplBase {
 
@@ -836,13 +858,15 @@ class GrpcObservationTest {
 
     static class ClientHeaderInterceptor implements ClientInterceptor {
 
+        private static final String AUTHORITY = "test-host:1234";
+
         private static final Metadata.Key<String> CLIENT_KEY = Metadata.Key.of("client",
                 Metadata.ASCII_STRING_MARSHALLER);
 
         @Override
         public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
                 CallOptions callOptions, Channel next) {
-            ClientCall<ReqT, RespT> call = next.newCall(method, callOptions);
+            ClientCall<ReqT, RespT> call = next.newCall(method, callOptions.withAuthority(AUTHORITY));
             return new SimpleForwardingClientCall<>(call) {
                 @Override
                 public void start(ClientCall.Listener<RespT> responseListener, Metadata headers) {

--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpJakartaServletRequestTags.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpJakartaServletRequestTags.java
@@ -49,7 +49,12 @@ public class HttpJakartaServletRequestTags {
      * @return the method tag whose value is a capitalized method (e.g. GET).
      */
     public static Tag method(@Nullable HttpServletRequest request) {
-        return (request != null) ? Tag.of("method", request.getMethod()) : METHOD_UNKNOWN;
+        if (request == null) {
+            return METHOD_UNKNOWN;
+        }
+
+        String method = request.getMethod();
+        return HttpMethods.isStandard(method) ? Tag.of("method", method) : METHOD_UNKNOWN;
     }

     /**
--- /dev/null
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpMethods.java
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2025 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.http;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Utility functions related to HTTP methods.
+ */
+final class HttpMethods {
+
+    private static final Set<String> STANDARD_HTTP_METHODS = new HashSet<>(9);
+
+    static {
+        STANDARD_HTTP_METHODS.add("GET");
+        STANDARD_HTTP_METHODS.add("HEAD");
+        STANDARD_HTTP_METHODS.add("POST");
+        STANDARD_HTTP_METHODS.add("PUT");
+        STANDARD_HTTP_METHODS.add("DELETE");
+        STANDARD_HTTP_METHODS.add("CONNECT");
+        STANDARD_HTTP_METHODS.add("OPTIONS");
+        STANDARD_HTTP_METHODS.add("TRACE");
+        STANDARD_HTTP_METHODS.add("PATCH");
+
+    }
+
+    private HttpMethods() {
+    }
+
+    /**
+     * Checks if the given method is a well-known HTTP method defined in
+     * <a href="https://www.rfc-editor.org/rfc/rfc9110.html#name-methods">RFC 9110</a> and
+     * <a href="https://www.rfc-editor.org/rfc/rfc5789.html">RFC 5789</a>. This is
+     * case-sensitive and known methods are all uppercase.
+     * @param method HTTP method to check
+     * @return {@code true} if the given method is a standard HTTP method
+     */
+    static boolean isStandard(String method) {
+        return STANDARD_HTTP_METHODS.contains(method);
+    }
+
+}
--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpRequestTags.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/http/HttpRequestTags.java
@@ -15,6 +15,7 @@
  */
 package io.micrometer.core.instrument.binder.http;

+import io.micrometer.common.lang.Nullable;
 import io.micrometer.common.util.StringUtils;
 import io.micrometer.core.annotation.Incubating;
 import io.micrometer.core.instrument.Tag;
@@ -46,8 +47,13 @@ public class HttpRequestTags {
      * @param request the request
      * @return the method tag whose value is a capitalized method (e.g. GET).
      */
-    public static Tag method(HttpServletRequest request) {
-        return (request != null) ? Tag.of("method", request.getMethod()) : METHOD_UNKNOWN;
+    public static Tag method(@Nullable HttpServletRequest request) {
+        if (request == null) {
+            return METHOD_UNKNOWN;
+        }
+
+        String method = request.getMethod();
+        return HttpMethods.isStandard(method) ? Tag.of("method", method) : METHOD_UNKNOWN;
     }

     /**
--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/DefaultJerseyObservationConvention.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/DefaultJerseyObservationConvention.java
@@ -57,7 +57,8 @@ public class DefaultJerseyObservationConvention implements JerseyObservationConv
         if (context.getCarrier() == null) {
             return null;
         }
-        return "HTTP " + context.getCarrier().getMethod();
+        String method = context.getCarrier().getMethod();
+        return "HTTP " + (HttpMethods.isStandard(method) ? method : "UNKNOWN");
     }

 }
--- /dev/null
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/HttpMethods.java
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2025 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.jersey.server;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Utility functions related to HTTP methods.
+ */
+final class HttpMethods {
+
+    private static final Set<String> STANDARD_HTTP_METHODS = new HashSet<>(9);
+
+    static {
+        STANDARD_HTTP_METHODS.add("GET");
+        STANDARD_HTTP_METHODS.add("HEAD");
+        STANDARD_HTTP_METHODS.add("POST");
+        STANDARD_HTTP_METHODS.add("PUT");
+        STANDARD_HTTP_METHODS.add("DELETE");
+        STANDARD_HTTP_METHODS.add("CONNECT");
+        STANDARD_HTTP_METHODS.add("OPTIONS");
+        STANDARD_HTTP_METHODS.add("TRACE");
+        STANDARD_HTTP_METHODS.add("PATCH");
+
+    }
+
+    private HttpMethods() {
+    }
+
+    /**
+     * Checks if the given method is a well-known HTTP method defined in
+     * <a href="https://www.rfc-editor.org/rfc/rfc9110.html#name-methods">RFC 9110</a> and
+     * <a href="https://www.rfc-editor.org/rfc/rfc5789.html">RFC 5789</a>. This is
+     * case-sensitive and known methods are all uppercase.
+     * @param method HTTP method to check
+     * @return {@code true} if the given method is a standard HTTP method
+     */
+    static boolean isStandard(String method) {
+        return STANDARD_HTTP_METHODS.contains(method);
+    }
+
+}
--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/JerseyKeyValues.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/JerseyKeyValues.java
@@ -16,6 +16,7 @@
 package io.micrometer.core.instrument.binder.jersey.server;

 import io.micrometer.common.KeyValue;
+import io.micrometer.common.lang.Nullable;
 import io.micrometer.common.util.StringUtils;
 import io.micrometer.core.instrument.binder.http.Outcome;
 import org.glassfish.jersey.server.ContainerRequest;
@@ -60,9 +61,14 @@ class JerseyKeyValues {
      * @param request the container request
      * @return the method KeyValue whose value is a capitalized method (e.g. GET).
      */
-    static KeyValue method(ContainerRequest request) {
-        return (request != null)
-                ? JerseyObservationDocumentation.JerseyLegacyLowCardinalityTags.METHOD.withValue(request.getMethod())
+    static KeyValue method(@Nullable ContainerRequest request) {
+        if (request == null) {
+            return METHOD_UNKNOWN;
+        }
+
+        String method = request.getMethod();
+        return HttpMethods.isStandard(method)
+                ? JerseyObservationDocumentation.JerseyLegacyLowCardinalityTags.METHOD.withValue(method)
                 : METHOD_UNKNOWN;
     }

--- a/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/JerseyTags.java
+++ b/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jersey/server/JerseyTags.java
@@ -67,8 +67,13 @@ public final class JerseyTags {
      * @param request the container request
      * @return the method tag whose value is a capitalized method (e.g. GET).
      */
-    public static Tag method(ContainerRequest request) {
-        return (request != null) ? Tag.of("method", request.getMethod()) : METHOD_UNKNOWN;
+    public static Tag method(@Nullable ContainerRequest request) {
+        if (request == null) {
+            return METHOD_UNKNOWN;
+        }
+
+        String method = request.getMethod();
+        return HttpMethods.isStandard(method) ? Tag.of("method", method) : METHOD_UNKNOWN;
     }

     /**
--- a/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/http/HttpJakartaServletRequestTagsTest.java
+++ b/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/http/HttpJakartaServletRequestTagsTest.java
@@ -43,6 +43,18 @@ class HttpJakartaServletRequestTagsTest {
         assertThat(result).isEqualTo(unknownMethod);
     }

+    @Test
+    void unknownMethodShouldContributeUnknownMethodTagValue() {
+        Tag unknownMethod = Tag.of("method", "UNKNOWN");
+        HttpServletRequest jakartaRequest = mockJakartaRequest("WELL");
+        Tag result = HttpJakartaServletRequestTags.method(jakartaRequest);
+        assertThat(result).isEqualTo(unknownMethod);
+
+        javax.servlet.http.HttpServletRequest javaxRequest = mockJavaxRequest("WELL");
+        result = HttpRequestTags.method(javaxRequest);
+        assertThat(result).isEqualTo(unknownMethod);
+    }
+
     @Test
     void requestShouldContributeMethodTag() {
         Tag httpGet = Tag.of("method", "GET");
--- /dev/null
+++ b/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/jersey/server/JerseyKeyValuesTests.java
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2026 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.jersey.server;
+
+import io.micrometer.common.KeyValue;
+import org.glassfish.jersey.server.ContainerRequest;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for {@link JerseyKeyValues}.
+ */
+class JerseyKeyValuesTests {
+
+    @Test
+    void nullRequestShouldResultInUnknownMethod() {
+        assertThat(JerseyKeyValues.method(null)).isEqualTo(KeyValue.of("method", "UNKNOWN"));
+    }
+
+    @Test
+    void unknownMethodShouldResultInUnknownTagValue() {
+        ContainerRequest request = mock(ContainerRequest.class);
+        when(request.getMethod()).thenReturn("WELL");
+        assertThat(JerseyKeyValues.method(request)).isEqualTo(KeyValue.of("method", "UNKNOWN"));
+    }
+
+    @Test
+    void knownMethodShouldResultInTagValue() {
+        ContainerRequest request = mock(ContainerRequest.class);
+        when(request.getMethod()).thenReturn("GET");
+        assertThat(JerseyKeyValues.method(request)).isEqualTo(KeyValue.of("method", "GET"));
+    }
+
+}
--- /dev/null
+++ b/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/jersey/server/JerseyTagsTests.java
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2026 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.jersey.server;
+
+import io.micrometer.core.instrument.Tag;
+import org.glassfish.jersey.server.ContainerRequest;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for {@link JerseyTags}.
+ */
+class JerseyTagsTests {
+
+    @Test
+    void nullRequestShouldResultInUnknownMethod() {
+        assertThat(JerseyTags.method(null)).isEqualTo(Tag.of("method", "UNKNOWN"));
+    }
+
+    @Test
+    void unknownMethodShouldResultInUnknownTagValue() {
+        ContainerRequest request = mock(ContainerRequest.class);
+        when(request.getMethod()).thenReturn("WELL");
+        assertThat(JerseyTags.method(request)).isEqualTo(Tag.of("method", "UNKNOWN"));
+    }
+
+    @Test
+    void knownMethodShouldResultInTagValue() {
+        ContainerRequest request = mock(ContainerRequest.class);
+        when(request.getMethod()).thenReturn("GET");
+        assertThat(JerseyTags.method(request)).isEqualTo(Tag.of("method", "GET"));
+    }
+
+}
