diff --git a/.mvn/extensions.xml b/.mvn/extensions.xml
deleted file mode 100644
index 91609f7618..0000000000
--- a/.mvn/extensions.xml
+++ /dev/null
@@ -1,9 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<extensions>
-    <!-- Workstation wide caching: just nuke your local repository and rebuild -->
-    <extension>
-        <groupId>eu.maveniverse.maven.mimir</groupId>
-        <artifactId>extension3</artifactId>
-        <version>0.11.2</version>
-    </extension>
-</extensions>
diff --git a/.mvn/maven.config b/.mvn/maven.config
index 1cbd53a74a..539dc8173f 100644
--- a/.mvn/maven.config
+++ b/.mvn/maven.config
@@ -1,5 +1 @@
 -Daether.dependencyCollector.impl=bf
--Daether.remoteRepositoryFilter.prefixes=true
--Daether.remoteRepositoryFilter.prefixes.basedir=${session.rootDirectory}/.mvn/rrf/
--Daether.remoteRepositoryFilter.groupId=true
--Daether.remoteRepositoryFilter.groupId.basedir=${session.rootDirectory}/.mvn/rrf/
\ No newline at end of file
diff --git a/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/DigestAuthentication.java b/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/DigestAuthentication.java
index 8cbf2ebe54..fe94d2ebb7 100644
--- a/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/DigestAuthentication.java
+++ b/jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/DigestAuthentication.java
@@ -14,6 +14,11 @@
 package org.eclipse.jetty.client;
 
 import java.net.URI;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.Charset;
+import java.nio.charset.CodingErrorAction;
 import java.nio.charset.StandardCharsets;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
@@ -28,21 +33,22 @@ import java.util.concurrent.atomic.AtomicInteger;
 import org.eclipse.jetty.http.HttpHeader;
 import org.eclipse.jetty.util.Attributes;
 import org.eclipse.jetty.util.StringUtil;
+import org.eclipse.jetty.util.TypeUtil;
 
 /**
- * Implementation of the HTTP "Digest" authentication defined in RFC 2617.
- * <p>
- * Applications should create objects of this class and add them to the
+ * <p>Implementation of the HTTP "Digest" authentication defined in RFC 7616.</p>
+ * <p>Applications should create objects of this class and add them to the
  * {@link AuthenticationStore} retrieved from the {@link HttpClient}
- * via {@link HttpClient#getAuthenticationStore()}.
+ * via {@link HttpClient#getAuthenticationStore()}.</p>
  */
 public class DigestAuthentication extends AbstractAuthentication
 {
     private final Random random;
     private final String user;
     private final String password;
+    private boolean userHashing;
 
-    /** Construct a DigestAuthentication with a {@link SecureRandom} nonce.
+    /**
      * @param uri the URI to match for the authentication
      * @param realm the realm to match for the authentication
      * @param user the user that wants to authenticate
@@ -75,10 +81,26 @@ public class DigestAuthentication extends AbstractAuthentication
         return "Digest";
     }
 
+    /**
+     * @return whether hashing of the username is supported
+     */
+    public boolean isUserHashing()
+    {
+        return userHashing;
+    }
+
+    /**
+     * @param userHashing whether hashing of the username is supported
+     */
+    public void setUserHashing(boolean userHashing)
+    {
+        this.userHashing = userHashing;
+    }
+
     @Override
     public boolean matches(String type, URI uri, String realm)
     {
-        // digest authenication requires a realm
+        // Digest authentication requires a realm.
         if (realm == null)
             return false;
 
@@ -90,9 +112,9 @@ public class DigestAuthentication extends AbstractAuthentication
     {
         Map<String, String> params = headerInfo.getParameters();
         String nonce = params.get("nonce");
-        if (nonce == null || nonce.length() == 0)
+        if (nonce == null || nonce.isEmpty())
             return null;
-        final String opaque = params.get("opaque");
+        String opaque = params.get("opaque");
         String algorithm = params.get("algorithm");
         if (algorithm == null)
             algorithm = "MD5";
@@ -109,11 +131,17 @@ public class DigestAuthentication extends AbstractAuthentication
             else if (serverQOPValues.contains("auth-int"))
                 clientQOP = "auth-int";
         }
+        String charsetName = params.get("charset");
+        Charset charset = StringUtil.isBlank(charsetName) ? null : Charset.forName(charsetName);
+        if (!"UTF-8".equals(charsetName))
+            charset = null;
+        boolean userHash = Boolean.parseBoolean(params.get("userhash"));
 
         String realm = getRealm();
         if (ANY_REALM.equals(realm))
             realm = headerInfo.getRealm();
-        return new DigestResult(headerInfo.getHeader(), response.getContent(), realm, user, password, algorithm, nonce, clientQOP, opaque);
+
+        return new DigestResult(headerInfo.getHeader(), realm, user, password, algorithm, nonce, clientQOP, opaque, charset, userHash);
     }
 
     private MessageDigest getMessageDigest(String algorithm)
@@ -132,7 +160,6 @@ public class DigestAuthentication extends AbstractAuthentication
     {
         private final AtomicInteger nonceCount = new AtomicInteger();
         private final HttpHeader header;
-        private final byte[] content;
         private final String realm;
         private final String user;
         private final String password;
@@ -140,11 +167,12 @@ public class DigestAuthentication extends AbstractAuthentication
         private final String nonce;
         private final String qop;
         private final String opaque;
+        private final Charset charset;
+        private final boolean userHash;
 
-        public DigestResult(HttpHeader header, byte[] content, String realm, String user, String password, String algorithm, String nonce, String qop, String opaque)
+        private DigestResult(HttpHeader header, String realm, String user, String password, String algorithm, String nonce, String qop, String opaque, Charset charset, boolean userHash)
         {
             this.header = header;
-            this.content = content;
             this.realm = realm;
             this.user = user;
             this.password = password;
@@ -152,6 +180,8 @@ public class DigestAuthentication extends AbstractAuthentication
             this.nonce = nonce;
             this.qop = qop;
             this.opaque = opaque;
+            this.charset = charset;
+            this.userHash = userHash;
         }
 
         @Override
@@ -165,72 +195,130 @@ public class DigestAuthentication extends AbstractAuthentication
         {
             MessageDigest digester = getMessageDigest(algorithm);
             if (digester == null)
-                return;
+                throw new IllegalArgumentException("Unsupported digest algorithm: " + algorithm);
+            if (!"auth".equals(qop))
+                throw new IllegalArgumentException("Unsupported quality of protection: " + qop);
+
+            // Retain ISO-8859-1 for old servers that don't send the charset parameter.
+            Charset cs = Objects.requireNonNullElse(charset, StandardCharsets.ISO_8859_1);
 
             String a1 = user + ":" + realm + ":" + password;
-            String hashA1 = toHexString(digester.digest(a1.getBytes(StandardCharsets.ISO_8859_1)));
+            String hashA1 = toHexString(digester.digest(strictEncode(cs, a1)));
 
             String query = request.getQuery();
             String path = request.getPath();
             String uri = (query == null) ? path : path + "?" + query;
             String a2 = request.getMethod() + ":" + uri;
-            if ("auth-int".equals(qop))
-                a2 += ":" + toHexString(digester.digest(content));
-            String hashA2 = toHexString(digester.digest(a2.getBytes(StandardCharsets.ISO_8859_1)));
+            String hashA2 = toHexString(digester.digest(strictEncode(cs, a2)));
 
             String nonceCount;
             String clientNonce;
-            String a3;
-            if (qop != null)
+            nonceCount = nextNonceCount();
+            clientNonce = newClientNonce();
+            String a3 = hashA1 + ":" + nonce + ":" + nonceCount + ":" + clientNonce + ":" + qop + ":" + hashA2;
+            String hashA3 = toHexString(digester.digest(strictEncode(cs, a3)));
+
+            // RFC-7616[3.4]: only username, realm, nonce, uri, response, cnonce, and opaque must be quoted.
+            // Parameters algorithm, username*, qop, and nc must not be quoted.
+            String digest = "Digest";
+            if (isUserHashing() && userHash)
+            {
+                digest += " username=\"%s\"".formatted(toHexString(digester.digest(strictEncode(cs, user + ":" + realm))));
+                digest += ", userhash=true";
+            }
+            else if (userNameNeedsEncoding(user))
             {
-                nonceCount = nextNonceCount();
-                clientNonce = newClientNonce();
-                a3 = hashA1 + ":" + nonce + ":" + nonceCount + ":" + clientNonce + ":" + qop + ":" + hashA2;
+                if (charset != null)
+                    digest += " username*=%s".formatted(encodeUserName(user, charset));
+                else
+                    throw new IllegalArgumentException("Unsupported username: " + user);
             }
             else
             {
-                nonceCount = null;
-                clientNonce = null;
-                a3 = hashA1 + ":" + nonce + ":" + hashA2;
+                digest += " username=\"%s\"".formatted(user);
             }
-            final String hashA3 = toHexString(digester.digest(a3.getBytes(StandardCharsets.ISO_8859_1)));
-
-            StringBuilder value = new StringBuilder("Digest");
-            value.append(" username=\"").append(user).append("\"");
-            value.append(", realm=\"").append(realm).append("\"");
-            value.append(", nonce=\"").append(nonce).append("\"");
+            digest += ", realm=\"%s\"".formatted(realm);
+            digest += ", nonce=\"%s\"".formatted(nonce);
             if (opaque != null)
-                value.append(", opaque=\"").append(opaque).append("\"");
-            value.append(", algorithm=\"").append(algorithm).append("\"");
-            value.append(", uri=\"").append(uri).append("\"");
-            if (qop != null)
+                digest += ", opaque=\"%s\"".formatted(opaque);
+            digest += ", algorithm=%s".formatted(algorithm);
+            digest += ", uri=\"%s\"".formatted(uri);
+            digest += ", qop=%s".formatted(qop);
+            digest += ", nc=%s".formatted(nonceCount);
+            digest += ", cnonce=\"%s\"".formatted(clientNonce);
+            digest += ", response=\"%s\"".formatted(hashA3);
+            String value = digest;
+
+            request.headers(headers -> headers.add(header, value));
+        }
+
+        private static byte[] strictEncode(Charset charset, String value)
+        {
+            try
+            {
+                ByteBuffer byteBuffer = charset.newEncoder()
+                    .onMalformedInput(CodingErrorAction.REPORT)
+                    .onUnmappableCharacter(CodingErrorAction.REPORT)
+                    .encode(CharBuffer.wrap(value));
+                byte[] bytes = new byte[byteBuffer.remaining()];
+                byteBuffer.get(bytes);
+                return bytes;
+            }
+            catch (CharacterCodingException x)
+            {
+                throw new RuntimeException(x);
+            }
+        }
+
+        private static boolean userNameNeedsEncoding(String user)
+        {
+            // Should be RFC 9110 quoted-string,
+            // but use here a simplified version.
+            for (int i = 0; i < user.length(); ++i)
             {
-                value.append(", qop=\"").append(qop).append("\"");
-                value.append(", nc=\"").append(nonceCount).append("\"");
-                value.append(", cnonce=\"").append(clientNonce).append("\"");
+                char c = user.charAt(i);
+                if (c < 0x20 || c > 0x7E || c == '"' || c == '\\')
+                    return true;
             }
-            value.append(", response=\"").append(hashA3).append("\"");
+            return false;
+        }
 
-            request.headers(headers -> headers.add(header, value.toString()));
+        private static String encodeUserName(String user, Charset charset)
+        {
+            byte[] bytes = strictEncode(charset, user);
+            StringBuilder builder = new StringBuilder(charset.name()).append("''");
+            for (byte b : bytes)
+            {
+                int c = b & 0xFF;
+                boolean unreserved = (c >= 'A' && c <= 'Z') ||
+                        (c >= 'a' && c <= 'z') ||
+                        (c >= '0' && c <= '9') ||
+                        c == '-' || c == '.' || c == '_' || c == '~';
+                if (unreserved)
+                    builder.append((char)c);
+                else
+                    builder.append(String.format("%%%02X", c));
+            }
+            return builder.toString();
         }
 
         private String nextNonceCount()
         {
             String padding = "00000000";
-            String next = Integer.toHexString(nonceCount.incrementAndGet()).toLowerCase(Locale.ENGLISH);
+            String next = Integer.toHexString(nonceCount.incrementAndGet()).toLowerCase(Locale.ROOT);
             return padding.substring(0, padding.length() - next.length()) + next;
         }
 
         private String newClientNonce()
         {
-            byte[] bytes = new byte[8];
+            byte[] bytes = new byte[16];
             random.nextBytes(bytes);
             return toHexString(bytes);
         }
 
         private String toHexString(byte[] bytes)
         {
-            return StringUtil.toHexString(bytes).toLowerCase(Locale.ENGLISH);
+            return TypeUtil.toString(bytes, 16);
         }
     }
 }
diff --git a/jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/DigestAuthenticator.java b/jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/DigestAuthenticator.java
index 7d3430b22f..614c564847 100644
--- a/jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/DigestAuthenticator.java
+++ b/jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/DigestAuthenticator.java
@@ -14,17 +14,21 @@
 package org.eclipse.jetty.security.authentication;
 
 import java.io.Serial;
-import java.nio.charset.StandardCharsets;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
 import java.security.MessageDigest;
 import java.security.SecureRandom;
 import java.util.Base64;
 import java.util.BitSet;
 import java.util.Iterator;
+import java.util.Locale;
+import java.util.Map;
 import java.util.Objects;
-import java.util.Queue;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentLinkedQueue;
-import java.util.concurrent.ConcurrentMap;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
 
 import org.eclipse.jetty.http.HttpHeader;
 import org.eclipse.jetty.http.HttpStatus;
@@ -43,6 +47,8 @@ import org.eclipse.jetty.util.thread.AutoLock;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import static java.nio.charset.StandardCharsets.UTF_8;
+
 /**
  * The nonce max age in ms can be set with the {@link SecurityHandler#setParameter(String, String)}
  * using the name "maxNonceAge".  The nonce max count can be set with {@link SecurityHandler#setParameter(String, String)}
@@ -54,10 +60,19 @@ public class DigestAuthenticator extends LoginAuthenticator
     private static final QuotedStringTokenizer TOKENIZER = QuotedStringTokenizer.builder().delimiters("=, ").returnDelimiters().allowEmbeddedQuotes().build();
 
     private final SecureRandom _random = new SecureRandom();
-    private final Queue<Nonce> _nonceQueue = new ConcurrentLinkedQueue<>();
-    private final ConcurrentMap<String, Nonce> _nonceMap = new ConcurrentHashMap<>();
+    private final Map<String, Nonce> _nonces = new ConcurrentHashMap<>();
+    private final SecretKey _secretKey;
     private long _maxNonceAgeMs = 60 * 1000;
-    private int _maxNC = 1024;
+    private int _maxNonceCount = 1024;
+    private String _algorithm = "SHA-256";
+    private boolean _userHashing;
+
+    public DigestAuthenticator()
+    {
+        byte[] bytes = new byte[32];
+        _random.nextBytes(bytes);
+        _secretKey = new SecretKeySpec(bytes, "HmacSHA256");
+    }
 
     @Override
     public void setConfiguration(Configuration configuration)
@@ -72,26 +87,70 @@ public class DigestAuthenticator extends LoginAuthenticator
             setMaxNonceCount(Integer.parseInt(mnc));
     }
 
+    /**
+     * @return the max number of times a nonce is used
+     */
     public int getMaxNonceCount()
     {
-        return _maxNC;
+        return _maxNonceCount;
     }
 
-    public void setMaxNonceCount(int maxNC)
+    /**
+     * @param maxNonceCount the max number of times a nonce is used
+     */
+    public void setMaxNonceCount(int maxNonceCount)
     {
-        _maxNC = maxNC;
+        _maxNonceCount = maxNonceCount;
     }
 
+    /**
+     * @return the max age of a nonce in milliseconds
+     */
     public long getMaxNonceAge()
     {
         return _maxNonceAgeMs;
     }
 
+    /**
+     * @param maxNonceAgeInMillis the max age of a nonce in milliseconds
+     */
     public void setMaxNonceAge(long maxNonceAgeInMillis)
     {
         _maxNonceAgeMs = maxNonceAgeInMillis;
     }
 
+    /**
+     * @return the {@link MessageDigest} algorithm
+     */
+    public String getAlgorithm()
+    {
+        return _algorithm;
+    }
+
+    /**
+     * @param algorithm the {@link MessageDigest} algorithm
+     */
+    public void setAlgorithm(String algorithm)
+    {
+        _algorithm = Objects.requireNonNull(algorithm);
+    }
+
+    /**
+     * @return whether the username is hashed
+     */
+    public boolean isUserHashing()
+    {
+        return _userHashing;
+    }
+
+    /**
+     * @param userHashing whether the username is hashed
+     */
+    public void setUserHashing(boolean userHashing)
+    {
+        _userHashing = userHashing;
+    }
+
     @Override
     public String getAuthenticationType()
     {
@@ -106,86 +165,186 @@ public class DigestAuthenticator extends LoginAuthenticator
         boolean stale = false;
         if (credentials != null)
         {
-            if (LOG.isDebugEnabled())
-                LOG.debug("Credentials: {}", credentials);
-            final Digest digest = new Digest(req.getMethod());
-            String last = null;
-            String name = null;
+            Digest digest = parseDigest(req.getMethod(), credentials);
+            if (digest != null)
+            {
+                if (verifyOpaque(digest))
+                {
+                    int n = checkNonce(digest);
+                    if (n > 0)
+                    {
+                        // Nonce correctness is a prerequisite for trusting digest.uri.
+                        // Check that the request URI matches the credential's URI.
+                        if (Objects.equals(digest.uri, req.getHttpURI().getPathQuery()))
+                        {
+                            UserIdentity user = login(digest.resolvedUserName, digest, req, res);
+                            if (user != null)
+                                return new UserAuthenticationSucceeded(getAuthenticationType(), user);
+                        }
+                    }
+                    else if (n == 0)
+                    {
+                        // Only good nonces can be stale.
+                        stale = true;
+                    }
+                }
+            }
+        }
+
+        if (AuthenticationState.Deferred.isDeferred(res))
+            return null;
 
-            for (Iterator<String> i = TOKENIZER.tokenize(credentials); i.hasNext();)
+        // Requests that don't have the Authorization header
+        // or that have it but it's invalid (e.g. missing/bad
+        // nonce, bad opaque, etc.) are replied with 401.
+
+        String domain = req.getContext().getContextPath();
+        if (domain == null)
+            domain = "/";
+
+        // RFC 7616[3.3]: only realm, domain, nonce, opaque, and qop must be quoted.
+        // Parameters stale and algorithm must not be quoted.
+        String value = "Digest realm=\"%s\"".formatted(_loginService.getName());
+        value += ", domain=\"%s\"".formatted(domain);
+        String nonce = newNonce(req);
+        value += ", nonce=\"%s\"".formatted(nonce);
+        value += ", opaque=\"%s\"".formatted(newOpaque(nonce));
+        value += ", stale=%s".formatted(stale);
+        value += ", algorithm=%s".formatted(getAlgorithm());
+        value += ", qop=\"auth\"";
+        value += ", charset=UTF-8";
+        value += ", userhash=%s".formatted(isUserHashing());
+
+        res.getHeaders().put(HttpHeader.WWW_AUTHENTICATE, value);
+
+        // Don't use AuthenticationState.writeError(), to avoid possibility of doing a Servlet error dispatch.
+        Response.writeError(req, res, callback, HttpStatus.UNAUTHORIZED_401);
+        return AuthenticationState.CHALLENGE;
+    }
+
+    private Digest parseDigest(String method, String credentials)
+    {
+        try
+        {
+            String name = null;
+            String value = null;
+            Digest digest = new Digest(method, getAlgorithm());
+            Iterator<String> i = TOKENIZER.tokenize(credentials);
+            while (i.hasNext())
             {
                 String tok = i.next();
                 char c = (tok.length() == 1) ? tok.charAt(0) : '\0';
-
                 switch (c)
                 {
-                    case '=':
-                        name = last;
-                        last = tok;
-                        break;
-                    case ',':
-                        name = null;
-                        break;
-                    case ' ':
-                        break;
-
-                    default:
-                        last = tok;
+                    case '=' -> name = value;
+                    case ',' -> name = null;
+                    case ' ' ->
+                    {
+                    }
+                    default ->
+                    {
+                        value = tok;
                         if (name != null)
                         {
-                            if ("username".equalsIgnoreCase(name))
-                                digest.username = tok;
-                            else if ("realm".equalsIgnoreCase(name))
-                                digest.realm = tok;
-                            else if ("nonce".equalsIgnoreCase(name))
-                                digest.nonce = tok;
-                            else if ("nc".equalsIgnoreCase(name))
-                                digest.nc = tok;
-                            else if ("cnonce".equalsIgnoreCase(name))
-                                digest.cnonce = tok;
-                            else if ("qop".equalsIgnoreCase(name))
-                                digest.qop = tok;
-                            else if ("uri".equalsIgnoreCase(name))
-                                digest.uri = tok;
-                            else if ("response".equalsIgnoreCase(name))
-                                digest.response = tok;
+                            switch (name.toLowerCase(Locale.ROOT))
+                            {
+                                case "response" -> digest.response = tok;
+                                case "username" -> digest.username = tok;
+                                case "userhash" -> digest.userhash = Boolean.parseBoolean(tok);
+                                case "username*" -> digest.usernameStar = tok;
+                                case "realm" -> digest.realm = tok;
+                                case "uri" -> digest.uri = tok;
+                                case "algorithm" -> digest.algorithm = tok;
+                                case "qop" -> digest.qop = tok;
+                                case "nonce" -> digest.nonce = tok;
+                                case "cnonce" -> digest.cnonce = tok;
+                                case "nc" -> digest.nc = tok;
+                                case "opaque" -> digest.opaque = tok;
+                            }
                             name = null;
                         }
+                    }
                 }
             }
 
-            int n = checkNonce(digest, req);
+            if (digest.username != null && digest.usernameStar != null)
+                return null;
+
+            if (digest.nc == null || digest.nc.length() != 8)
+                return null;
 
-            if (n > 0)
+            if (!getAlgorithm().equalsIgnoreCase(digest.algorithm))
+                return null;
+
+            if (!"auth".equalsIgnoreCase(digest.qop))
+                return null;
+
+            String resolvedUserName;
+            if (digest.userhash && isUserHashing())
             {
-                //UserIdentity user = _loginService.login(digest.username,digest);
-                UserIdentity user = login(digest.username, digest, req, res);
-                if (user != null)
-                {
-                    return new UserAuthenticationSucceeded(getAuthenticationType(), user);
-                }
+                if (digest.username == null)
+                    return null;
+                resolvedUserName = resolveHashedUserName(digest.username);
             }
-            else if (n == 0)
-                stale = true;
-        }
+            else if (digest.usernameStar != null)
+            {
+                resolvedUserName = resolveEncodedUserName(digest.usernameStar);
+            }
+            else
+            {
+                resolvedUserName = digest.username;
+            }
+            digest.resolvedUserName = resolvedUserName;
 
-        if (!AuthenticationState.Deferred.isDeferred(res))
+            return digest;
+        }
+        catch (Throwable x)
         {
-            String domain = req.getContext().getContextPath();
-            if (domain == null)
-                domain = "/";
-            res.getHeaders().put(HttpHeader.WWW_AUTHENTICATE.asString(), "Digest realm=\"" + _loginService.getName() +
-                    "\", domain=\"" + domain +
-                    "\", nonce=\"" + newNonce(req) +
-                    "\", algorithm=MD5" +
-                    ", qop=\"auth\"" +
-                    ", stale=" + stale);
-            Response.writeError(req, res, callback, HttpStatus.UNAUTHORIZED_401);
-
-            return AuthenticationState.CHALLENGE;
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to parse digest", x);
+            return null;
         }
+    }
 
-        return null;
+    /**
+     * <p>Resolves the hashed username received in the {@code Authorization} header.</p>
+     * <p>This method should look up the plain username from the username hash.</p>
+     * <p>By default, this method throws {@link UnsupportedOperationException}.</p>
+     *
+     * @param hashedUserName the hashed username
+     * @return the plain username
+     */
+    protected String resolveHashedUserName(String hashedUserName)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * <p>Resolves the encoded username received in the {@code Authorization} header.</p>
+     * <p>The username is encoded with RFC-5987 and this method should decode it.</p>
+     * <p>By default, this method only decodes RFC-5987 usernames encoded with UTF-8
+     * and no language; for example username: {@code UTF-8''caf%E2%82%AC} is decoded
+     * as {@code caf€}.</p>
+     *
+     * @param encodedUserName the encoded username
+     * @return the decoded username
+     */
+    protected String resolveEncodedUserName(String encodedUserName)
+    {
+        try
+        {
+            if (encodedUserName == null)
+                return null;
+            String encodedPrefix = "UTF-8''";
+            if (encodedUserName.startsWith(encodedPrefix))
+                return URI.create("scheme:" + encodedUserName.substring(encodedPrefix.length())).getSchemeSpecificPart();
+            return encodedUserName;
+        }
+        catch (Throwable x)
+        {
+            // Likely badly encoded username.
+            return null;
+        }
     }
 
     @Override
@@ -199,105 +358,202 @@ public class DigestAuthenticator extends LoginAuthenticator
 
     public String newNonce(Request request)
     {
-        Nonce nonce;
-
-        do
+        try
         {
-            byte[] nounce = new byte[24];
-            _random.nextBytes(nounce);
-
-            nonce = new Nonce(Base64.getEncoder().encodeToString(nounce), Request.getTimeStamp(request), getMaxNonceCount());
+            // The format of the nonce is such that it does
+            // not need server-side storage to be checked.
+            // nonce = base64(timestamp | random | hmac(timestamp | random))
+            Mac mac = newMac();
+            int dataLength = 2 * Long.BYTES;
+            byte[] bytes = new byte[dataLength + mac.getMacLength()];
+            ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
+            byteBuffer.putLong(Request.getTimeStamp(request)).putLong(_random.nextLong());
+            mac.update(bytes, 0, dataLength);
+            mac.doFinal(bytes, dataLength);
+            return Base64.getEncoder().encodeToString(bytes);
+        }
+        catch (GeneralSecurityException x)
+        {
+            throw new RuntimeException(x);
         }
-        while (_nonceMap.putIfAbsent(nonce._nonce, nonce) != null);
-        _nonceQueue.add(nonce);
-
-        return nonce._nonce;
     }
 
     /**
      * @param digest the digest data to check
-     * @param request the request object
-     * @return -1 for a bad nonce, 0 for a stale none, 1 for a good nonce
+     * @return -1 for a bad nonce, 0 for a stale nonce, 1 for a good nonce
      */
-    private int checkNonce(Digest digest, Request request)
+    private int checkNonce(Digest digest)
     {
-        // firstly let's expire old nonces
-        long expired = Request.getTimeStamp(request) - getMaxNonceAge();
-        Nonce nonce = _nonceQueue.peek();
-        while (nonce != null && nonce._ts < expired)
-        {
-            _nonceQueue.remove(nonce);
-            _nonceMap.remove(nonce._nonce);
-            nonce = _nonceQueue.peek();
-        }
-
-        // Now check the requested nonce
         try
         {
-            nonce = _nonceMap.get(digest.nonce);
+            String nonce = digest.nonce;
             if (nonce == null)
+                return -1;
+
+            String nc = digest.nc;
+            if (nc == null)
+                return -1;
+            long nonceNumber = Long.parseLong(digest.nc, 16);
+            if (nonceNumber >= getMaxNonceCount())
                 return 0;
 
-            long count = Long.parseLong(digest.nc, 16);
-            if (count >= _maxNC)
+            byte[] nonceBytes = Base64.getDecoder().decode(nonce);
+            ByteBuffer byteBuffer = ByteBuffer.wrap(nonceBytes);
+
+            long timestamp = byteBuffer.getLong();
+            long random = byteBuffer.getLong();
+
+            Mac mac = newMac();
+            int dataLength = 2 * Long.BYTES;
+            byte[] expected = new byte[dataLength + mac.getMacLength()];
+            ByteBuffer.wrap(expected).putLong(timestamp).putLong(random);
+            mac.update(expected, 0, dataLength);
+            mac.doFinal(expected, dataLength);
+
+            if (!MessageDigest.isEqual(expected, nonceBytes))
+                return -1;
+
+            long elapsed = System.currentTimeMillis() - timestamp;
+            if (elapsed < 0)
+                return -1;
+
+            if (elapsed > getMaxNonceAge())
                 return 0;
 
-            if (nonce.seen((int)count))
+            // Store the nonce to keep track of digest.nc, the nonce count.
+
+            // First, expire old nonces.
+            Iterator<Nonce> iterator = _nonces.values().iterator();
+            while (iterator.hasNext())
+            {
+                Nonce n = iterator.next();
+                elapsed = System.currentTimeMillis() - n._timestamp;
+                if (elapsed > getMaxNonceAge())
+                    iterator.remove();
+            }
+            // Store the used nonce.
+            Nonce n = _nonces.computeIfAbsent(nonce, k -> new Nonce(timestamp, getMaxNonceCount()));
+            // Check the nonce number to counter replay attacks.
+            if (n.seen((int)nonceNumber))
                 return -1;
 
             return 1;
         }
-        catch (Exception e)
+        catch (Throwable x)
         {
-            LOG.trace("IGNORED", e);
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to verify nonce", x);
+            return -1;
+        }
+    }
+
+    private String newOpaque(String nonce)
+    {
+        // The format of the opaque parameter is tied to the nonce:
+        // opaque = base64(hmac(nonce))
+        byte[] bytes = Base64.getDecoder().decode(nonce);
+        Mac mac = newMac();
+        byte[] macBytes = mac.doFinal(bytes);
+        return Base64.getEncoder().encodeToString(macBytes);
+    }
+
+    private boolean verifyOpaque(Digest digest)
+    {
+        try
+        {
+            // RFC-7616[3.3]: the opaque parameter SHOULD
+            // be sent by the client, but it may not.
+            if (digest.opaque == null)
+                return true;
+
+            String nonce = digest.nonce;
+            if (nonce == null)
+                return false;
+
+            Mac mac = newMac();
+            byte[] expected = mac.doFinal(Base64.getDecoder().decode(nonce));
+
+            byte[] actual = Base64.getDecoder().decode(digest.opaque);
+            return MessageDigest.isEqual(expected, actual);
+        }
+        catch (Throwable x)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to verify opaque", x);
+            return false;
+        }
+    }
+
+    private Mac newMac()
+    {
+        try
+        {
+            Mac mac = Mac.getInstance(_secretKey.getAlgorithm());
+            mac.init(_secretKey);
+            return mac;
+        }
+        catch (Throwable x)
+        {
+            throw new RuntimeException(x);
         }
-        return -1;
     }
 
     private static class Nonce
     {
         private final AutoLock _lock = new AutoLock();
-        final String _nonce;
-        final long _ts;
-        final BitSet _seen;
+        private final long _timestamp;
+        private final BitSet _seen;
 
-        public Nonce(String nonce, long ts, int size)
+        private Nonce(long timestamp, int size)
         {
-            _nonce = nonce;
-            _ts = ts;
+            _timestamp = timestamp;
             _seen = new BitSet(size);
         }
 
-        public boolean seen(int count)
+        private boolean seen(int number)
         {
             try (AutoLock ignored = _lock.lock())
             {
-                if (count >= _seen.size())
+                if (number >= _seen.size())
                     return true;
-                boolean s = _seen.get(count);
-                _seen.set(count);
+                boolean s = _seen.get(number);
+                _seen.set(number);
                 return s;
             }
         }
     }
 
+    // This class must remain {@code static} to avoid the hidden
+    // reference to the outer class, that would make it non-serializable.
     private static class Digest extends Credential
     {
         @Serial
         private static final long serialVersionUID = -2484639019549527724L;
-        final String method;
-        String username = "";
-        String realm = "";
-        String nonce = "";
-        String nc = "";
-        String cnonce = "";
-        String qop = "";
-        String uri = "";
-        String response = "";
-
-        Digest(String m)
+
+        private final String method;
+        private String algorithm;
+        private String username;
+        private String usernameStar;
+        private String resolvedUserName;
+        private String realm;
+        private String nonce;
+        private String nc;
+        private String cnonce;
+        private String qop;
+        private String uri;
+        private String response;
+        private boolean userhash;
+        private String opaque;
+
+        private Digest(String method, String algorithm)
         {
-            method = m;
+            this.method = method;
+            this.algorithm = algorithm;
+        }
+
+        private String getAlgorithm()
+        {
+            return algorithm;
         }
 
         @Override
@@ -309,68 +565,47 @@ public class DigestAuthenticator extends LoginAuthenticator
 
             try
             {
-                // MD5 required by the specification
-                MessageDigest md = MessageDigest.getInstance("MD5");
+                MessageDigest md = MessageDigest.getInstance(getAlgorithm());
                 byte[] ha1;
-                if (credentials instanceof MD5)
+                if (credentials instanceof MD5 md5)
                 {
                     // Credentials are already a MD5 digest - assume it's in
                     // form user:realm:password (we have no way to know since
                     // it's a digest, alright?)
-                    ha1 = ((MD5)credentials).getDigest();
+                    ha1 = md5.getDigest();
                 }
                 else
                 {
-                    // calc A1 digest
-                    md.update(username.getBytes(StandardCharsets.ISO_8859_1));
-                    md.update((byte)':');
-                    md.update(realm.getBytes(StandardCharsets.ISO_8859_1));
-                    md.update((byte)':');
-                    md.update(password.getBytes(StandardCharsets.ISO_8859_1));
-                    ha1 = md.digest();
+                    // Calculate H(A1).
+                    String a1 = resolvedUserName + ":" + realm + ":" + password;
+                    ha1 = md.digest(a1.getBytes(UTF_8));
                 }
-                // calc A2 digest
-                md.reset();
-                md.update(method.getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(uri.getBytes(StandardCharsets.ISO_8859_1));
-                byte[] ha2 = md.digest();
-
-                // calc digest
-                // request-digest = <"> < KD ( H(A1), unq(nonce-value) ":"
-                // nc-value ":" unq(cnonce-value) ":" unq(qop-value) ":" H(A2) )
-                // <">
-                // request-digest = <"> < KD ( H(A1), unq(nonce-value) ":" H(A2)
-                // ) > <">
-
-                md.update(TypeUtil.toString(ha1, 16).getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(nonce.getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(nc.getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(cnonce.getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(qop.getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(TypeUtil.toString(ha2, 16).getBytes(StandardCharsets.ISO_8859_1));
-                byte[] digest = md.digest();
-
-                // check digest
-                return stringEquals(TypeUtil.toString(digest, 16).toLowerCase(), response == null ? null : response.toLowerCase());
+
+                // Calculate H(A2).
+                String a2 = method + ":" + uri;
+                byte[] ha2 = md.digest(a2.getBytes(UTF_8));
+
+                // Calculate response, must match what the client sent.
+                String expected = TypeUtil.toString(ha1, 16) + ":" +
+                    nonce + ":" + nc + ":" + cnonce + ":" + qop + ":" +
+                    TypeUtil.toString(ha2, 16);
+                expected = TypeUtil.toString(md.digest(expected.getBytes(UTF_8)), 16);
+
+                // Check digest.
+                return stringEquals(expected, response == null ? null : response.toLowerCase(Locale.ROOT));
             }
-            catch (Exception e)
+            catch (Throwable x)
             {
-                LOG.warn("Unable to process digest", e);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Unable to process digest", x);
+                return false;
             }
-
-            return false;
         }
 
         @Override
         public String toString()
         {
-            return username + "," + response;
+            return "%s@%x[u=%s]".formatted(TypeUtil.toShortName(getClass()), hashCode(), resolvedUserName);
         }
     }
 }
diff --git a/jetty-ee9/jetty-ee9-security/src/main/java/org/eclipse/jetty/ee9/security/authentication/DigestAuthenticator.java b/jetty-ee9/jetty-ee9-security/src/main/java/org/eclipse/jetty/ee9/security/authentication/DigestAuthenticator.java
index 0b66bf0786..2ee2e12f72 100644
--- a/jetty-ee9/jetty-ee9-security/src/main/java/org/eclipse/jetty/ee9/security/authentication/DigestAuthenticator.java
+++ b/jetty-ee9/jetty-ee9-security/src/main/java/org/eclipse/jetty/ee9/security/authentication/DigestAuthenticator.java
@@ -15,17 +15,21 @@ package org.eclipse.jetty.ee9.security.authentication;
 
 import java.io.IOException;
 import java.io.Serial;
-import java.nio.charset.StandardCharsets;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
 import java.security.MessageDigest;
 import java.security.SecureRandom;
 import java.util.Base64;
 import java.util.BitSet;
 import java.util.Iterator;
+import java.util.Locale;
+import java.util.Map;
 import java.util.Objects;
-import java.util.Queue;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentLinkedQueue;
-import java.util.concurrent.ConcurrentMap;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
 
 import jakarta.servlet.ServletRequest;
 import jakarta.servlet.ServletResponse;
@@ -38,6 +42,7 @@ import org.eclipse.jetty.ee9.security.SecurityHandler;
 import org.eclipse.jetty.ee9.security.ServerAuthException;
 import org.eclipse.jetty.ee9.security.UserAuthentication;
 import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpStatus;
 import org.eclipse.jetty.security.UserIdentity;
 import org.eclipse.jetty.util.QuotedStringTokenizer;
 import org.eclipse.jetty.util.TypeUtil;
@@ -46,6 +51,8 @@ import org.eclipse.jetty.util.thread.AutoLock;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import static java.nio.charset.StandardCharsets.UTF_8;
+
 /**
  * The nonce max age in ms can be set with the {@link SecurityHandler#setInitParameter(String, String)}
  * using the name "maxNonceAge".  The nonce max count can be set with {@link SecurityHandler#setInitParameter(String, String)}
@@ -57,10 +64,19 @@ public class DigestAuthenticator extends LoginAuthenticator
     private static final QuotedStringTokenizer TOKENIZER = QuotedStringTokenizer.builder().delimiters("=, ").returnDelimiters().allowEmbeddedQuotes().build();
 
     private final SecureRandom _random = new SecureRandom();
-    private final ConcurrentMap<String, Nonce> _nonceMap = new ConcurrentHashMap<>();
-    private final Queue<Nonce> _nonceQueue = new ConcurrentLinkedQueue<>();
+    private final Map<String, Nonce> _nonces = new ConcurrentHashMap<>();
+    private final SecretKey _secretKey;
     private long _maxNonceAgeMs = 60 * 1000;
-    private int _maxNC = 1024;
+    private int _maxNonceCount = 1024;
+    private String _algorithm = "SHA-256";
+    private boolean _userHashing;
+
+    public DigestAuthenticator()
+    {
+        byte[] bytes = new byte[32];
+        _random.nextBytes(bytes);
+        _secretKey = new SecretKeySpec(bytes, "HmacSHA256");
+    }
 
     @Override
     public void setConfiguration(AuthConfiguration configuration)
@@ -75,26 +91,70 @@ public class DigestAuthenticator extends LoginAuthenticator
             setMaxNonceCount(Integer.parseInt(mnc));
     }
 
+    /**
+     * @return the max number of times a nonce is used
+     */
     public int getMaxNonceCount()
     {
-        return _maxNC;
+        return _maxNonceCount;
     }
 
-    public void setMaxNonceCount(int maxNC)
+    /**
+     * @param maxNonceCount the max number of times a nonce is used
+     */
+    public void setMaxNonceCount(int maxNonceCount)
     {
-        _maxNC = maxNC;
+        _maxNonceCount = maxNonceCount;
     }
 
+    /**
+     * @return the max age of a nonce in milliseconds
+     */
     public long getMaxNonceAge()
     {
         return _maxNonceAgeMs;
     }
 
+    /**
+     * @param maxNonceAgeInMillis the max age of a nonce in milliseconds
+     */
     public void setMaxNonceAge(long maxNonceAgeInMillis)
     {
         _maxNonceAgeMs = maxNonceAgeInMillis;
     }
 
+    /**
+     * @return the {@link MessageDigest} algorithm
+     */
+    public String getAlgorithm()
+    {
+        return _algorithm;
+    }
+
+    /**
+     * @param algorithm the {@link MessageDigest} algorithm
+     */
+    public void setAlgorithm(String algorithm)
+    {
+        _algorithm = Objects.requireNonNull(algorithm);
+    }
+
+    /**
+     * @return whether the username is hashed
+     */
+    public boolean isUserHashing()
+    {
+        return _userHashing;
+    }
+
+    /**
+     * @param userHashing whether the username is hashed
+     */
+    public void setUserHashing(boolean userHashing)
+    {
+        _userHashing = userHashing;
+    }
+
     @Override
     public String getAuthMethod()
     {
@@ -126,90 +186,188 @@ public class DigestAuthenticator extends LoginAuthenticator
             boolean stale = false;
             if (credentials != null)
             {
-                if (LOG.isDebugEnabled())
-                    LOG.debug("Credentials: {}", credentials);
-                final Digest digest = new Digest(request.getMethod());
-                String last = null;
-                String name = null;
-
-                for (Iterator<String> i = TOKENIZER.tokenize(credentials); i.hasNext();)
+                Digest digest = parseDigest(request.getMethod(), credentials);
+                if (digest != null)
                 {
-                    String tok = i.next();
-                    char c = (tok.length() == 1) ? tok.charAt(0) : '\0';
-
-                    switch (c)
+                    if (verifyOpaque(digest))
                     {
-                        case '=':
-                            name = last;
-                            last = tok;
-                            break;
-                        case ',':
-                            name = null;
-                            break;
-                        case ' ':
-                            break;
-
-                        default:
-                            last = tok;
-                            if (name != null)
+                        int n = checkNonce(digest);
+                        if (n > 0)
+                        {
+                            // Nonce correctness is a prerequisite for trusting digest.uri.
+                            // Check that the request URI matches the credential's URI.
+                            if (Objects.equals(digest.uri, baseRequest.getHttpURI().getPathQuery()))
                             {
-                                if ("username".equalsIgnoreCase(name))
-                                    digest.username = tok;
-                                else if ("realm".equalsIgnoreCase(name))
-                                    digest.realm = tok;
-                                else if ("nonce".equalsIgnoreCase(name))
-                                    digest.nonce = tok;
-                                else if ("nc".equalsIgnoreCase(name))
-                                    digest.nc = tok;
-                                else if ("cnonce".equalsIgnoreCase(name))
-                                    digest.cnonce = tok;
-                                else if ("qop".equalsIgnoreCase(name))
-                                    digest.qop = tok;
-                                else if ("uri".equalsIgnoreCase(name))
-                                    digest.uri = tok;
-                                else if ("response".equalsIgnoreCase(name))
-                                    digest.response = tok;
-                                name = null;
+                                UserIdentity user = login(digest.resolvedUserName, digest, request);
+                                if (user != null)
+                                    return new UserAuthentication(getAuthMethod(), user);
                             }
+                        }
+                        else if (n == 0)
+                        {
+                            // Only good nonces can be stale.
+                            stale = true;
+                        }
                     }
                 }
+            }
 
-                int n = checkNonce(digest, baseRequest);
+            if (DeferredAuthentication.isDeferred(response))
+                return Authentication.UNAUTHENTICATED;
 
-                if (n > 0)
+            // Requests that don't have the Authorization header
+            // or that have it but it's invalid (e.g. missing/bad
+            // nonce, bad opaque, etc.) are replied with 401.
+
+            String domain = request.getContextPath();
+            if (domain == null)
+                domain = "/";
+
+            // RFC 7616[3.3]: only realm, domain, nonce, opaque, and qop must be quoted.
+            // Parameters stale and algorithm must not be quoted.
+            String value = "Digest realm=\"%s\"".formatted(_loginService.getName());
+            value += ", domain=\"%s\"".formatted(domain);
+            String nonce = newNonce(baseRequest);
+            value += ", nonce=\"%s\"".formatted(nonce);
+            value += ", opaque=\"%s\"".formatted(newOpaque(nonce));
+            value += ", stale=%s".formatted(stale);
+            value += ", algorithm=%s".formatted(getAlgorithm());
+            value += ", qop=\"auth\"";
+            value += ", charset=UTF-8";
+            value += ", userhash=%s".formatted(isUserHashing());
+            response.setHeader(HttpHeader.WWW_AUTHENTICATE.asString(), value);
+
+            response.sendError(HttpServletResponse.SC_UNAUTHORIZED);
+            return Authentication.SEND_CONTINUE;
+        }
+        catch (IOException e)
+        {
+            throw new ServerAuthException(e);
+        }
+    }
+
+    private Digest parseDigest(String method, String credentials)
+    {
+        try
+        {
+            String name = null;
+            String value = null;
+            Digest digest = new Digest(method, getAlgorithm());
+            Iterator<String> i = TOKENIZER.tokenize(credentials);
+            while (i.hasNext())
+            {
+                String tok = i.next();
+                char c = (tok.length() == 1) ? tok.charAt(0) : '\0';
+                switch (c)
                 {
-                    //UserIdentity user = _loginService.login(digest.username,digest);
-                    UserIdentity user = login(digest.username, digest, req);
-                    if (user != null)
+                    case '=' -> name = value;
+                    case ',' -> name = null;
+                    case ' ' ->
+                    {
+                    }
+                    default ->
                     {
-                        return new UserAuthentication(getAuthMethod(), user);
+                        value = tok;
+                        if (name != null)
+                        {
+                            switch (name.toLowerCase(Locale.ROOT))
+                            {
+                                case "response" -> digest.response = tok;
+                                case "username" -> digest.username = tok;
+                                case "userhash" -> digest.userhash = Boolean.parseBoolean(tok);
+                                case "username*" -> digest.usernameStar = tok;
+                                case "realm" -> digest.realm = tok;
+                                case "uri" -> digest.uri = tok;
+                                case "algorithm" -> digest.algorithm = tok;
+                                case "qop" -> digest.qop = tok;
+                                case "nonce" -> digest.nonce = tok;
+                                case "cnonce" -> digest.cnonce = tok;
+                                case "nc" -> digest.nc = tok;
+                                case "opaque" -> digest.opaque = tok;
+                            }
+                            name = null;
+                        }
                     }
                 }
-                else if (n == 0)
-                    stale = true;
             }
 
-            if (!DeferredAuthentication.isDeferred(response))
+            if (digest.username != null && digest.usernameStar != null)
+                return null;
+
+            if (digest.nc == null || digest.nc.length() != 8)
+                return null;
+
+            if (!getAlgorithm().equalsIgnoreCase(digest.algorithm))
+                return null;
+
+            if (!"auth".equalsIgnoreCase(digest.qop))
+                return null;
+
+            String resolvedUserName;
+            if (digest.userhash && isUserHashing())
+            {
+                if (digest.username == null)
+                    return null;
+                resolvedUserName = resolveHashedUserName(digest.username);
+            }
+            else if (digest.usernameStar != null)
             {
-                String domain = request.getContextPath();
-                if (domain == null)
-                    domain = "/";
-                response.setHeader(HttpHeader.WWW_AUTHENTICATE.asString(), "Digest realm=\"" + _loginService.getName() +
-                    "\", domain=\"" + domain +
-                    "\", nonce=\"" + newNonce(baseRequest) +
-                    "\", algorithm=MD5" +
-                    ", qop=\"auth\"" +
-                    ", stale=" + stale);
-                response.sendError(HttpServletResponse.SC_UNAUTHORIZED);
-
-                return Authentication.SEND_CONTINUE;
+                resolvedUserName = resolveEncodedUserName(digest.usernameStar);
             }
+            else
+            {
+                resolvedUserName = digest.username;
+            }
+            digest.resolvedUserName = resolvedUserName;
 
-            return Authentication.UNAUTHENTICATED;
+            return digest;
         }
-        catch (IOException e)
+        catch (Throwable x)
         {
-            throw new ServerAuthException(e);
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to parse digest", x);
+            return null;
+        }
+    }
+
+    /**
+     * <p>Resolves the hashed username received in the {@code Authorization} header.</p>
+     * <p>This method should look up the plain username from the username hash.</p>
+     * <p>By default, this method throws {@link UnsupportedOperationException}.</p>
+     *
+     * @param hashedUserName the hashed username
+     * @return the plain username
+     */
+    protected String resolveHashedUserName(String hashedUserName)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * <p>Resolves the encoded username received in the {@code Authorization} header.</p>
+     * <p>The username is encoded with RFC-5987 and this method should decode it.</p>
+     * <p>By default, this method only decodes RFC-5987 usernames encoded with UTF-8
+     * and no language; for example username: {@code UTF-8''caf%E2%82%AC} is decoded
+     * as {@code caf€}.</p>
+     *
+     * @param encodedUserName the encoded username
+     * @return the decoded username
+     */
+    protected String resolveEncodedUserName(String encodedUserName)
+    {
+        try
+        {
+            if (encodedUserName == null)
+                return null;
+            String encodedPrefix = "UTF-8''";
+            if (encodedUserName.startsWith(encodedPrefix))
+                return URI.create("scheme:" + encodedUserName.substring(encodedPrefix.length())).getSchemeSpecificPart();
+            return encodedUserName;
+        }
+        catch (Throwable x)
+        {
+            // Likely badly encoded username.
+            return null;
         }
     }
 
@@ -224,105 +382,202 @@ public class DigestAuthenticator extends LoginAuthenticator
 
     public String newNonce(Request request)
     {
-        Nonce nonce;
-
-        do
+        try
         {
-            byte[] nounce = new byte[24];
-            _random.nextBytes(nounce);
-
-            nonce = new Nonce(Base64.getEncoder().encodeToString(nounce), request.getTimeStamp(), getMaxNonceCount());
+            // The format of the nonce is such that it does
+            // not need server-side storage to be checked.
+            // nonce = base64(timestamp | random | hmac(timestamp | random))
+            Mac mac = newMac();
+            int dataLength = 2 * Long.BYTES;
+            byte[] bytes = new byte[dataLength + mac.getMacLength()];
+            ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
+            byteBuffer.putLong(request.getTimeStamp()).putLong(_random.nextLong());
+            mac.update(bytes, 0, dataLength);
+            mac.doFinal(bytes, dataLength);
+            return Base64.getEncoder().encodeToString(bytes);
+        }
+        catch (GeneralSecurityException x)
+        {
+            throw new RuntimeException(x);
         }
-        while (_nonceMap.putIfAbsent(nonce._nonce, nonce) != null);
-        _nonceQueue.add(nonce);
-
-        return nonce._nonce;
     }
 
     /**
      * @param digest the digest data to check
-     * @param request the request object
-     * @return -1 for a bad nonce, 0 for a stale none, 1 for a good nonce
+     * @return -1 for a bad nonce, 0 for a stale nonce, 1 for a good nonce
      */
-    private int checkNonce(Digest digest, Request request)
+    private int checkNonce(Digest digest)
     {
-        // firstly let's expire old nonces
-        long expired = request.getTimeStamp() - getMaxNonceAge();
-        Nonce nonce = _nonceQueue.peek();
-        while (nonce != null && nonce._ts < expired)
-        {
-            _nonceQueue.remove(nonce);
-            _nonceMap.remove(nonce._nonce);
-            nonce = _nonceQueue.peek();
-        }
-
-        // Now check the requested nonce
         try
         {
-            nonce = _nonceMap.get(digest.nonce);
+            String nonce = digest.nonce;
             if (nonce == null)
+                return -1;
+
+            String nc = digest.nc;
+            if (nc == null)
+                return -1;
+            long nonceNumber = Long.parseLong(digest.nc, 16);
+            if (nonceNumber >= getMaxNonceCount())
                 return 0;
 
-            long count = Long.parseLong(digest.nc, 16);
-            if (count >= _maxNC)
+            byte[] nonceBytes = Base64.getDecoder().decode(nonce);
+            ByteBuffer byteBuffer = ByteBuffer.wrap(nonceBytes);
+
+            long timestamp = byteBuffer.getLong();
+            long random = byteBuffer.getLong();
+
+            Mac mac = newMac();
+            int dataLength = 2 * Long.BYTES;
+            byte[] expected = new byte[dataLength + mac.getMacLength()];
+            ByteBuffer.wrap(expected).putLong(timestamp).putLong(random);
+            mac.update(expected, 0, dataLength);
+            mac.doFinal(expected, dataLength);
+
+            if (!MessageDigest.isEqual(expected, nonceBytes))
+                return -1;
+
+            long elapsed = System.currentTimeMillis() - timestamp;
+            if (elapsed < 0)
+                return -1;
+
+            if (elapsed > getMaxNonceAge())
                 return 0;
 
-            if (nonce.seen((int)count))
+            // Store the nonce to keep track of digest.nc, the nonce count.
+
+            // First, expire old nonces.
+            Iterator<Nonce> iterator = _nonces.values().iterator();
+            while (iterator.hasNext())
+            {
+                Nonce n = iterator.next();
+                elapsed = System.currentTimeMillis() - n._timestamp;
+                if (elapsed > getMaxNonceAge())
+                    iterator.remove();
+            }
+            // Store the used nonce.
+            Nonce n = _nonces.computeIfAbsent(nonce, k -> new Nonce(timestamp, getMaxNonceCount()));
+            // Check the nonce number to counter replay attacks.
+            if (n.seen((int)nonceNumber))
                 return -1;
 
             return 1;
         }
-        catch (Exception e)
+        catch (Throwable x)
         {
-            LOG.trace("IGNORED", e);
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to verify nonce", x);
+            return -1;
+        }
+    }
+
+    private String newOpaque(String nonce)
+    {
+        // The format of the opaque parameter is tied to the nonce:
+        // opaque = base64(hmac(nonce))
+        byte[] bytes = Base64.getDecoder().decode(nonce);
+        Mac mac = newMac();
+        byte[] macBytes = mac.doFinal(bytes);
+        return Base64.getEncoder().encodeToString(macBytes);
+    }
+
+    private boolean verifyOpaque(Digest digest)
+    {
+        try
+        {
+            // RFC-7616[3.3]: the opaque parameter SHOULD
+            // be sent by the client, but it may not.
+            if (digest.opaque == null)
+                return true;
+
+            String nonce = digest.nonce;
+            if (nonce == null)
+                return false;
+
+            Mac mac = newMac();
+            byte[] expected = mac.doFinal(Base64.getDecoder().decode(nonce));
+
+            byte[] actual = Base64.getDecoder().decode(digest.opaque);
+            return MessageDigest.isEqual(expected, actual);
+        }
+        catch (Throwable x)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Unable to verify opaque", x);
+            return false;
+        }
+    }
+
+    private Mac newMac()
+    {
+        try
+        {
+            Mac mac = Mac.getInstance(_secretKey.getAlgorithm());
+            mac.init(_secretKey);
+            return mac;
+        }
+        catch (Throwable x)
+        {
+            throw new RuntimeException(x);
         }
-        return -1;
     }
 
     private static class Nonce
     {
         private final AutoLock _lock = new AutoLock();
-        final String _nonce;
-        final long _ts;
-        final BitSet _seen;
+        private final long _timestamp;
+        private final BitSet _seen;
 
-        public Nonce(String nonce, long ts, int size)
+        private Nonce(long timestamp, int size)
         {
-            _nonce = nonce;
-            _ts = ts;
+            _timestamp = timestamp;
             _seen = new BitSet(size);
         }
 
-        public boolean seen(int count)
+        private boolean seen(int number)
         {
             try (AutoLock ignored = _lock.lock())
             {
-                if (count >= _seen.size())
+                if (number >= _seen.size())
                     return true;
-                boolean s = _seen.get(count);
-                _seen.set(count);
+                boolean s = _seen.get(number);
+                _seen.set(number);
                 return s;
             }
         }
     }
 
+    // This class must remain {@code static} to avoid the hidden
+    // reference to the outer class, that would make it non-serializable.
     private static class Digest extends Credential
     {
         @Serial
         private static final long serialVersionUID = -2484639019549527724L;
-        final String method;
-        String username = "";
-        String realm = "";
-        String nonce = "";
-        String nc = "";
-        String cnonce = "";
-        String qop = "";
-        String uri = "";
-        String response = "";
-
-        Digest(String m)
+
+        private final String method;
+        private String algorithm;
+        private String username;
+        private String usernameStar;
+        private String resolvedUserName;
+        private String realm;
+        private String nonce;
+        private String nc;
+        private String cnonce;
+        private String qop;
+        private String uri;
+        private String response;
+        private boolean userhash;
+        private String opaque;
+
+        private Digest(String method, String algorithm)
         {
-            method = m;
+            this.method = method;
+            this.algorithm = algorithm;
+        }
+
+        private String getAlgorithm()
+        {
+            return algorithm;
         }
 
         @Override
@@ -334,68 +589,47 @@ public class DigestAuthenticator extends LoginAuthenticator
 
             try
             {
-                // MD5 required by the specification
-                MessageDigest md = MessageDigest.getInstance("MD5");
+                MessageDigest md = MessageDigest.getInstance(getAlgorithm());
                 byte[] ha1;
-                if (credentials instanceof Credential.MD5)
+                if (credentials instanceof MD5 md5)
                 {
                     // Credentials are already a MD5 digest - assume it's in
                     // form user:realm:password (we have no way to know since
                     // it's a digest, alright?)
-                    ha1 = ((Credential.MD5)credentials).getDigest();
+                    ha1 = md5.getDigest();
                 }
                 else
                 {
-                    // calc A1 digest
-                    md.update(username.getBytes(StandardCharsets.ISO_8859_1));
-                    md.update((byte)':');
-                    md.update(realm.getBytes(StandardCharsets.ISO_8859_1));
-                    md.update((byte)':');
-                    md.update(password.getBytes(StandardCharsets.ISO_8859_1));
-                    ha1 = md.digest();
+                    // Calculate H(A1).
+                    String a1 = resolvedUserName + ":" + realm + ":" + password;
+                    ha1 = md.digest(a1.getBytes(UTF_8));
                 }
-                // calc A2 digest
-                md.reset();
-                md.update(method.getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(uri.getBytes(StandardCharsets.ISO_8859_1));
-                byte[] ha2 = md.digest();
-
-                // calc digest
-                // request-digest = <"> < KD ( H(A1), unq(nonce-value) ":"
-                // nc-value ":" unq(cnonce-value) ":" unq(qop-value) ":" H(A2) )
-                // <">
-                // request-digest = <"> < KD ( H(A1), unq(nonce-value) ":" H(A2)
-                // ) > <">
-
-                md.update(TypeUtil.toString(ha1, 16).getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(nonce.getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(nc.getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(cnonce.getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(qop.getBytes(StandardCharsets.ISO_8859_1));
-                md.update((byte)':');
-                md.update(TypeUtil.toString(ha2, 16).getBytes(StandardCharsets.ISO_8859_1));
-                byte[] digest = md.digest();
-
-                // check digest
-                return stringEquals(TypeUtil.toString(digest, 16).toLowerCase(), response == null ? null : response.toLowerCase());
+
+                // Calculate H(A2).
+                String a2 = method + ":" + uri;
+                byte[] ha2 = md.digest(a2.getBytes(UTF_8));
+
+                // Calculate response, must match what the client sent.
+                String expected = TypeUtil.toString(ha1, 16) + ":" +
+                    nonce + ":" + nc + ":" + cnonce + ":" + qop + ":" +
+                    TypeUtil.toString(ha2, 16);
+                expected = TypeUtil.toString(md.digest(expected.getBytes(UTF_8)), 16);
+
+                // Check digest.
+                return stringEquals(expected, response == null ? null : response.toLowerCase(Locale.ROOT));
             }
-            catch (Exception e)
+            catch (Throwable x)
             {
-                LOG.warn("Unable to process digest", e);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Unable to process digest", x);
+                return false;
             }
-
-            return false;
         }
 
         @Override
         public String toString()
         {
-            return username + "," + response;
+            return "%s@%x[u=%s]".formatted(TypeUtil.toShortName(getClass()), hashCode(), resolvedUserName);
         }
     }
 }
diff --git a/pom.xml b/pom.xml
index e702bba681..76e349b5d7 100644
--- a/pom.xml
+++ b/pom.xml
@@ -2061,16 +2061,6 @@
       </plugin>
     </plugins>
     <extensions>
-      <extension>
-        <groupId>eu.maveniverse.maven.njord</groupId>
-        <artifactId>extension3</artifactId>
-        <version>${njord.version}</version>
-      </extension>
-      <extension>
-        <groupId>org.apache.maven.extensions</groupId>
-        <artifactId>maven-build-cache-extension</artifactId>
-        <version>${maven-build-cache.version}</version>
-      </extension>
     </extensions>
   </build>
 

--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java
@@ -969,6 +969,8 @@
             if (!_stream.compareAndSet(null, stream))
                 throw new IllegalStateException("Stream pending");
             _headerBuilder.clear();
+            if (_trailers != null)
+                _trailers.clear();
             _httpChannel.setHttpStream(stream);
         }
 
@@ -1029,9 +1031,15 @@
             if (stream._chunk != null)
                 throw new IllegalStateException();
             if (_trailers != null)
-                stream._chunk = new Trailers(_trailers.asImmutable());
+            {
+                if (_trailers.size() > 0)
+                    stream._chunk = new Trailers(_trailers.asImmutable());
+                _trailers = null;
+            }
             else
+            {
                 stream._chunk = Content.Chunk.EOF;
+            }
 
             getHttpChannel().getComplianceViolationListener().onRequestBegin(getHttpChannel().getRequest());
             return false;
@@ -1252,7 +1260,7 @@
             if (_uri.getPath() == null)
                 _uri.path("/");
 
-            _request = new MetaData.Request(_parser.getBeginNanoTime(), _method, _uri.asImmutable(), _version, _headerBuilder, _contentLength)
+            _request = new MetaData.Request(_parser.getBeginNanoTime(), _method, _uri.asImmutable(), _version, _headerBuilder.asImmutable(), _contentLength)
             {
                 @Override
                 public boolean is100ContinueExpected()
@@ -1337,6 +1345,7 @@
                 }
             }
 
+            _headerBuilder.clear();
             if (!persistent)
                 _generator.setPersistent(false);
 
