From 93b53d5fdd599d3f84b7b770f6efed536bf5e400 Mon Sep 17 00:00:00 2001
Date: Wed, 3 Dec 2025 14:55:33 +0530
Subject: [PATCH] adf

---
 index.js | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/index.js b/index.js
index 640c24c..a9d124f 100644
--- a/index.js
+++ b/index.js
@@ -2,13 +2,17 @@
 exports = module.exports = trim;
 
 function trim(str){
-  return str.replace(/^\s*|\s*$/g, '');
+  return exports.right(exports.left(str));
 }
 
 exports.left = function(str){
-  return str.replace(/^\s*/, '');
+  return str.replace(/^\s\s*/, '');
 };
 
 exports.right = function(str){
-  return str.replace(/\s*$/, '');
+  var whitespace_pattern = /\s/,
+      i = str.length;
+  while (whitespace_pattern.test(str.charAt(--i)));
+
+  return str.slice(0, i + 1);
 };
-- 
2.43.0

