--- package/lib/isLength.js	1985-10-26 13:45:00.000000000 +0530
+++ package/lib/isLength.js	2025-12-05 12:22:45.753150279 +0530
@@ -20,7 +20,7 @@
     min = arguments[1] || 0;
     max = arguments[2];
   }
-  var presentationSequences = str.match(/(\uFE0F|\uFE0E)/g) || [];
+  var presentationSequences = str.match(/[^\uFE0F\uFE0E][\uFE0F\uFE0E]/g) || [];
   var surrogatePairs = str.match(/[\uD800-\uDBFF][\uDC00-\uDFFF]/g) || [];
   var len = str.length - presentationSequences.length - surrogatePairs.length;
   return len >= min && (typeof max === 'undefined' || len <= max);
--- package/es/lib/isLength.js	1985-10-26 13:45:00.000000000 +0530
+++ package/es/lib/isLength.js	2025-12-05 12:22:47.063166804 +0530
@@ -14,7 +14,7 @@
     min = arguments[1] || 0;
     max = arguments[2];
   }
-  var presentationSequences = str.match(/(\uFE0F|\uFE0E)/g) || [];
+  var presentationSequences = str.match(/[^\uFE0F\uFE0E][\uFE0F\uFE0E]/g) || [];
   var surrogatePairs = str.match(/[\uD800-\uDBFF][\uDC00-\uDFFF]/g) || [];
   var len = str.length - presentationSequences.length - surrogatePairs.length;
   return len >= min && (typeof max === 'undefined' || len <= max);
--- package/validator.js	1985-10-26 13:45:00.000000000 +0530
+++ package/validator.js	2025-12-05 12:22:41.050119330 +0530
@@ -1967,7 +1967,7 @@
     min = arguments[1] || 0;
     max = arguments[2];
   }
-  var presentationSequences = str.match(/(\uFE0F|\uFE0E)/g) || [];
+  var presentationSequences = str.match(/[^\uFE0F\uFE0E][\uFE0F\uFE0E]/g) || [];
   var surrogatePairs = str.match(/[\uD800-\uDBFF][\uDC00-\uDFFF]/g) || [];
   var len = str.length - presentationSequences.length - surrogatePairs.length;
   return len >= min && (typeof max === 'undefined' || len <= max);
