--- a/driver/mysql/MysqlDriver.js
+++ b/driver/mysql/MysqlDriver.js
@@ -989,6 +989,7 @@
             trace: options.trace,
             multipleStatements: options.multipleStatements,
             flags: options.flags,
+            stringifyObjects: true,
         }, {
             host: credentials.host,
             user: credentials.username,
@@ -997,9 +998,10 @@
             port: credentials.port,
             ssl: options.ssl,
             socketPath: credentials.socketPath,
+            connectionLimit: options.poolSize,
         }, options.acquireTimeout === undefined
             ? {}
-            : { acquireTimeout: options.acquireTimeout }, { connectionLimit: options.poolSize }, options.extra || {});
+            : { acquireTimeout: options.acquireTimeout }, options.extra || {});
     }
     /**
      * Creates a new connection pool for a given database credentials.
