--- a/src/main/java/org/yaml/snakeyaml/LoaderOptions.java
+++ b/src/main/java/org/yaml/snakeyaml/LoaderOptions.java
@@ -13,6 +13,10 @@
  */
 package org.yaml.snakeyaml;
 
+import java.util.ArrayList;
+import org.yaml.snakeyaml.constructor.TagInspector;
+import org.yaml.snakeyaml.constructor.TrustedPrefixesTagInspector;
+
 /**
  * Configuration for loading
  */
@@ -29,6 +33,11 @@
   private int codePointLimit = 3 * 1024 * 1024; // 3 MB
 
   /**
+   * Secure by default - no custom classes are allowed
+   */
+  private TagInspector tagInspector = new TrustedPrefixesTagInspector(new ArrayList<String>());
+
+  /**
    * getter
    *
    * @return true when duplicate keys in mapping allowed (the latter overrides the former)
@@ -189,5 +198,13 @@
    */
   public void setCodePointLimit(int codePointLimit) {
     this.codePointLimit = codePointLimit;
+  }
+
+  public TagInspector getTagInspector() {
+    return tagInspector;
+  }
+
+  public void setTagInspector(TagInspector tagInspector) {
+    this.tagInspector = tagInspector;
   }
 }
--- a/src/main/java/org/yaml/snakeyaml/Yaml.java
+++ b/src/main/java/org/yaml/snakeyaml/Yaml.java
@@ -134,12 +134,13 @@
   /**
    * Create Yaml instance. It is safe to create a few instances and use them in different Threads.
    *
-   * @param constructor BaseConstructor to construct incoming documents
+   * @param constructor BaseConstructor to construct incoming documents. Its LoaderOptions will be
+   *        used everywhere
    * @param representer Representer to emit outgoing objects
    * @param dumperOptions DumperOptions to configure outgoing objects
    */
   public Yaml(BaseConstructor constructor, Representer representer, DumperOptions dumperOptions) {
-    this(constructor, representer, dumperOptions, new LoaderOptions(), new Resolver());
+    this(constructor, representer, dumperOptions, constructor.getLoadingConfig(), new Resolver());
   }
 
   /**
--- a/src/main/java/org/yaml/snakeyaml/composer/Composer.java
+++ b/src/main/java/org/yaml/snakeyaml/composer/Composer.java
@@ -228,6 +228,10 @@
       resolved = true;
     } else {
       nodeTag = new Tag(tag);
+      if (nodeTag.isCustomGlobal() && !loadingConfig.getTagInspector().allowGlobalTag(nodeTag)) {
+        throw new ComposerException(null, null, "Global tag is not allowed: " + tag,
+            ev.getStartMark());
+      }
     }
     Node node = new ScalarNode(nodeTag, resolved, ev.getValue(), ev.getStartMark(), ev.getEndMark(),
         ev.getScalarStyle());
@@ -251,6 +255,10 @@
       resolved = true;
     } else {
       nodeTag = new Tag(tag);
+      if (nodeTag.isCustomGlobal() && !loadingConfig.getTagInspector().allowGlobalTag(nodeTag)) {
+        throw new ComposerException(null, null, "Global tag is not allowed: " + tag,
+            startEvent.getStartMark());
+      }
     }
     final ArrayList<Node> children = new ArrayList<Node>();
     SequenceNode node = new SequenceNode(nodeTag, resolved, children, startEvent.getStartMark(),
@@ -291,6 +299,10 @@
       resolved = true;
     } else {
       nodeTag = new Tag(tag);
+      if (nodeTag.isCustomGlobal() && !loadingConfig.getTagInspector().allowGlobalTag(nodeTag)) {
+        throw new ComposerException(null, null, "Global tag is not allowed: " + tag,
+            startEvent.getStartMark());
+      }
     }
 
     final List<NodeTuple> children = new ArrayList<NodeTuple>();
--- a/src/main/java/org/yaml/snakeyaml/constructor/BaseConstructor.java
+++ b/src/main/java/org/yaml/snakeyaml/constructor/BaseConstructor.java
@@ -711,4 +711,8 @@
   public void setEnumCaseSensitive(boolean enumCaseSensitive) {
     this.enumCaseSensitive = enumCaseSensitive;
   }
+
+  public LoaderOptions getLoadingConfig() {
+    return loadingConfig;
+  }
 }
--- /dev/null
+++ b/src/main/java/org/yaml/snakeyaml/constructor/TagInspector.java
@@ -0,0 +1,33 @@
+/**
+ * Copyright (c) 2008, SnakeYAML
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.yaml.snakeyaml.constructor;
+
+import org.yaml.snakeyaml.nodes.Tag;
+
+/**
+ * Check if the global tags are allowed (the local tags are always allowed) It should control the
+ * classes to create to prevent possible remote code invocation when the data comes from untrusted
+ * source.
+ */
+public interface TagInspector {
+
+  /**
+   * Check
+   *
+   * @param tag - the global tag to allow
+   * @return true when the custom global tag is allowed to create a custom Java instance
+   */
+  boolean allowGlobalTag(Tag tag);
+
+}
--- /dev/null
+++ b/src/main/java/org/yaml/snakeyaml/constructor/TrustedPrefixesTagInspector.java
@@ -0,0 +1,51 @@
+/**
+ * Copyright (c) 2008, SnakeYAML
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.yaml.snakeyaml.constructor;
+
+import java.util.List;
+import org.yaml.snakeyaml.nodes.Tag;
+
+/**
+ * Allow to create classes with custom global tag if the class name matches any of the provided
+ * prefixes.
+ */
+public class TrustedPrefixesTagInspector implements TagInspector {
+
+  private final List<String> trustedList;
+
+  /**
+   * Create
+   *
+   * @param trustedList - list of prefixes to allow. It may be the package names
+   */
+  public TrustedPrefixesTagInspector(List<String> trustedList) {
+    this.trustedList = trustedList;
+  }
+
+  /**
+   * Check
+   *
+   * @param tag - the global tag to allow
+   * @return true when the custom global tag is allowed to create a custom Java instance
+   */
+  @Override
+  public boolean allowGlobalTag(Tag tag) {
+    for (String trusted : trustedList) {
+      if (tag.getClassName().startsWith(trusted)) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
--- /dev/null
+++ b/src/main/java/org/yaml/snakeyaml/constructor/TrustedTagInspector.java
@@ -0,0 +1,34 @@
+/**
+ * Copyright (c) 2008, SnakeYAML
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.yaml.snakeyaml.constructor;
+
+import org.yaml.snakeyaml.nodes.Tag;
+
+/**
+ * TagInspector which allows to create any custom instance. Should not be used when the data comes
+ * from untrusted source to prevent possible remote code invocation.
+ */
+public class TrustedTagInspector implements TagInspector {
+
+  /**
+   * Allow any
+   *
+   * @param tag - the global tag to allow
+   * @return always return true
+   */
+  @Override
+  public boolean allowGlobalTag(Tag tag) {
+    return true;
+  }
+}
--- a/src/main/java/org/yaml/snakeyaml/nodes/Tag.java
+++ b/src/main/java/org/yaml/snakeyaml/nodes/Tag.java
@@ -40,6 +40,27 @@
   public static final Tag STR = new Tag(PREFIX + "str");
   public static final Tag SEQ = new Tag(PREFIX + "seq");
   public static final Tag MAP = new Tag(PREFIX + "map");
+
+  // https://yaml.org/type/index.html
+  public static final Set<Tag> standardTags = new HashSet<Tag>(15);
+
+  static {
+    standardTags.add(YAML);
+    standardTags.add(MERGE);
+    standardTags.add(SET);
+    standardTags.add(PAIRS);
+    standardTags.add(OMAP);
+    standardTags.add(BINARY);
+    standardTags.add(INT);
+    standardTags.add(FLOAT);
+    standardTags.add(TIMESTAMP);
+    standardTags.add(BOOL);
+    standardTags.add(NULL);
+    standardTags.add(STR);
+    standardTags.add(SEQ);
+    standardTags.add(MAP);
+  }
+
   // For use to indicate a DUMMY node that contains comments, when there is no other (empty
   // document)
   public static final Tag COMMENT = new Tag(PREFIX + "comment");
@@ -159,4 +180,12 @@
     return value.equals(Tag.PREFIX + clazz.getName());
   }
 
+  /**
+   * Check if the that is global and not standard to provide it to TagInspector for verification.
+   *
+   * @return true when the tag must be verified to avoid remote code invocation
+   */
+  public boolean isCustomGlobal() {
+    return !secondary && !standardTags.contains(this);
+  }
 }
