--- a/java/org/apache/catalina/tribes/group/interceptors/EncryptInterceptor.java
+++ b/java/org/apache/catalina/tribes/group/interceptors/EncryptInterceptor.java
@@ -139,10 +139,10 @@
             xbb.clear();
             xbb.append(data, 0, data.length);

+            super.messageReceived(msg);
         } catch (GeneralSecurityException gse) {
             log.error(sm.getString("encryptInterceptor.decrypt.failed"), gse);
         }
-        super.messageReceived(msg);
     }

     /**

--- a/java/org/apache/catalina/tribes/membership/cloud/AbstractStreamProvider.java
+++ b/java/org/apache/catalina/tribes/membership/cloud/AbstractStreamProvider.java
@@ -81,7 +81,7 @@
     public URLConnection openConnection(String url, Map<String,String> headers, int connectTimeout, int readTimeout)
             throws IOException {
         if (log.isDebugEnabled()) {
-            log.debug(sm.getString("abstractStream.connection", getClass().getSimpleName(), url, headers,
+            log.debug(sm.getString("abstractStream.connection", getClass().getSimpleName(), url,
                     Integer.toString(connectTimeout), Integer.toString(readTimeout)));
         }
         URLConnection connection;
--- a/java/org/apache/catalina/tribes/membership/cloud/LocalStrings.properties
+++ b/java/org/apache/catalina/tribes/membership/cloud/LocalStrings.properties
@@ -17,7 +17,7 @@
 # To edit translations see: https://tomcat.apache.org/getinvolved.html#Translations

 abstractStream.CACertUndefined=CA cert file undefined
-abstractStream.connection=[{0}] opening connection: url [{1}], headers [{2}], connectTimeout [{3}], readTimeout [{4}]
+abstractStream.connection=[{0}] opening connection: url [{1}], connectTimeout [{2}], readTimeout [{3}]
 abstractStream.fileNotFound=CA cert file [{0}] not found
 abstractStream.invalidTimeout=Neither connectTimeout [{0}] nor readTimeout [{1}] can be less than 0 for URLConnection
 abstractStream.trustManagerError=Could not create trust manager for [{0}]
@@ -41,4 +41,4 @@
 kubernetesMembershipProvider.serviceAccountTokenMissing=Service account token not found at [{0}]
 kubernetesMembershipProvider.streamError=Failed to open stream

-tokenStream.failedConnection=Failed connection to [{0}] with token [{1}]
+tokenStream.failedConnection=Failed connection to [{0}]
--- a/java/org/apache/catalina/tribes/membership/cloud/TokenStreamProvider.java
+++ b/java/org/apache/catalina/tribes/membership/cloud/TokenStreamProvider.java
@@ -61,7 +61,7 @@
             return super.openStream(url, headers, connectTimeout, readTimeout);
         } catch (IOException ioe) {
             // Add debug information
-            throw new IOException(sm.getString("tokenStream.failedConnection", url, token), ioe);
+            throw new IOException(sm.getString("tokenStream.failedConnection", url), ioe);
         }
     }
 }
\ No newline at end of file

--- a/java/org/apache/catalina/valves/AbstractAccessLogValve.java
+++ b/java/org/apache/catalina/valves/AbstractAccessLogValve.java
@@ -1169,10 +1169,10 @@
                 } else {
                     buf.append(request.getMethod());
                     buf.append(' ');
-                    buf.append(request.getRequestURI());
+                    escapeAndAppend(request.getRequestURI(), buf);
                     if (request.getQueryString() != null) {
                         buf.append('?');
-                        buf.append(request.getQueryString());
+                        escapeAndAppend(request.getQueryString(), buf);
                     }
                     buf.append(' ');
                     buf.append(request.getProtocol());
@@ -1417,7 +1417,7 @@
             }
             if (query != null) {
                 buf.append('?');
-                buf.append(query);
+                escapeAndAppend(query, buf);
             }
         }
     }
@@ -1448,7 +1448,7 @@
         @Override
         public void addElement(CharArrayWriter buf, Request request, Response response, long time) {
             if (request != null) {
-                buf.append(request.getRequestURI());
+                escapeAndAppend(request.getRequestURI(), buf);
             } else {
                 buf.append('-');
             }

--- a/java/org/apache/tomcat/util/net/openssl/panama/OpenSSLEngine.java
+++ b/java/org/apache/tomcat/util/net/openssl/panama/OpenSSLEngine.java
@@ -1303,6 +1303,7 @@
     private static int processOCSPRequest(EngineState state, URL url, MemorySegment issuer, MemorySegment x509,
             MemorySegment /* X509_STORE_CTX */ x509ctx, Arena localArena) {
         if (openssl_h_Compatibility.BORINGSSL || openssl_h_Compatibility.isLibreSSLPre35()) {
+            X509_STORE_CTX_set_error(x509ctx, X509_V_ERR_UNABLE_TO_GET_CRL());
             return V_OCSP_CERTSTATUS_UNKNOWN();
         }
         MemorySegment ocspRequest = MemorySegment.NULL;
@@ -1315,20 +1316,24 @@
         try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
             ocspRequest = OCSP_REQUEST_new();
             if (MemorySegment.NULL.equals(ocspRequest)) {
+                X509_STORE_CTX_set_error(x509ctx, X509_V_ERR_UNABLE_TO_GET_CRL());
                 return V_OCSP_CERTSTATUS_UNKNOWN();
             }
             id = OCSP_cert_to_id(MemorySegment.NULL, x509, issuer);
             if (MemorySegment.NULL.equals(id)) {
+                X509_STORE_CTX_set_error(x509ctx, X509_V_ERR_UNABLE_TO_GET_CRL());
                 return V_OCSP_CERTSTATUS_UNKNOWN();
             }
             ocspOneReq = OCSP_request_add0_id(ocspRequest, id);
             if (MemorySegment.NULL.equals(ocspOneReq)) {
+                X509_STORE_CTX_set_error(x509ctx, X509_V_ERR_UNABLE_TO_GET_CRL());
                 return V_OCSP_CERTSTATUS_UNKNOWN();
             }
             OCSP_request_add1_nonce(ocspRequest, (char) 0, -1);
             MemorySegment bufPointer = localArena.allocateFrom(ValueLayout.ADDRESS, MemorySegment.NULL);
             int requestLength = i2d_OCSP_REQUEST(ocspRequest, bufPointer);
             if (requestLength <= 0) {
+                X509_STORE_CTX_set_error(x509ctx, X509_V_ERR_UNABLE_TO_GET_CRL());
                 return V_OCSP_CERTSTATUS_UNKNOWN();
             }
             MemorySegment buf = bufPointer.get(ValueLayout.ADDRESS, 0);
@@ -1350,6 +1355,7 @@
             connection.getOutputStream().write(ocspRequestData);
             int responseCode = connection.getResponseCode();
             if (responseCode != HttpURLConnection.HTTP_OK) {
+                X509_STORE_CTX_set_error(x509ctx, X509_V_ERR_UNABLE_TO_GET_CRL());
                 return V_OCSP_CERTSTATUS_UNKNOWN();
             }
             InputStream is = connection.getInputStream();
@@ -1357,6 +1363,7 @@
             byte[] responseBuf = new byte[1024];
             while ((read = is.read(responseBuf)) > 0) {
                 if (baos.size() > OCSP_MAX_RESPONSE_SIZE) {
+                    X509_STORE_CTX_set_error(x509ctx, X509_V_ERR_UNABLE_TO_GET_CRL());
                     return V_OCSP_CERTSTATUS_UNKNOWN();
                 }
                 baos.write(responseBuf, 0, read);

--- a/java/org/apache/tomcat/util/http/parser/HttpParser.java
+++ b/java/org/apache/tomcat/util/http/parser/HttpParser.java
@@ -33,9 +33,9 @@
  */
 public class HttpParser {

     private static final StringManager sm = StringManager.getManager(HttpParser.class);

-    private static final int ARRAY_SIZE = 128;
+    private static final int ARRAY_SIZE = 256;

     private static final boolean[] IS_CONTROL = new boolean[ARRAY_SIZE];
     private static final boolean[] IS_SEPARATOR = new boolean[ARRAY_SIZE];
@@ -47,6 +47,8 @@
     private static final boolean[] IS_SUBDELIM = new boolean[ARRAY_SIZE];
     private static final boolean[] IS_USERINFO = new boolean[ARRAY_SIZE];
     private static final boolean[] IS_RELAXABLE = new boolean[ARRAY_SIZE];
+    private static final boolean[] IS_FIELD_VCHAR = new boolean[ARRAY_SIZE];
+    private static final boolean[] IS_FIELD_CONTENT = new boolean[ARRAY_SIZE];

     private static final HttpParser DEFAULT;

@@ -66,7 +68,7 @@
             }

             // Token: Anything 0-127 that is not a control and not a separator
-            if (!IS_CONTROL[i] && !IS_SEPARATOR[i]) {
+            if (!IS_CONTROL[i] && !IS_SEPARATOR[i] && i < 128) {
                 IS_TOKEN[i] = true;
             }

@@ -75,7 +77,6 @@
                 IS_HEX[i] = true;
             }

-            // Not valid for HTTP protocol
             // "HTTP/" DIGIT "." DIGIT
             if (i == 'H' || i == 'T' || i == 'P' || i == '/' || i == '.' || (i >= '0' && i <= '9')) {
                 IS_HTTP_PROTOCOL[i] = true;
@@ -114,6 +115,16 @@
                     i == '{' || i == '|' || i == '}') {
                 IS_RELAXABLE[i] = true;
             }
+
+            // field-vchar is VCHAR / obs-text
+            if (i > 32 && i < 127 || i > 127) {
+                IS_FIELD_VCHAR[i] = true;
+            }
+
+            // field-content  = field-vchar [ 1*( SP / HTAB / field-vchar ) field-vchar ]
+            if (IS_FIELD_VCHAR[i] || i == '\t' || i == ' ') {
+                IS_FIELD_CONTENT[i] = true;
+            }
         }

         DEFAULT = new HttpParser(null, null);
@@ -393,8 +404,7 @@


     public static boolean isControl(int c) {
-        // Fast for valid control characters, slower for some incorrect
-        // ones
+        // Fast for valid control characters, slower for some incorrect ones
         try {
             return IS_CONTROL[c];
         } catch (ArrayIndexOutOfBoundsException ex) {
@@ -403,6 +413,26 @@
     }


+    public static boolean isFieldVChar(int c) {
+        // Fast for valid field-vchar characters, slower for some incorrect ones
+        try {
+            return IS_FIELD_VCHAR[c];
+        } catch (ArrayIndexOutOfBoundsException ex) {
+            return false;
+        }
+    }
+
+
+    public static boolean isFieldContent(int c) {
+        // Fast for valid field-content characters, slower for some incorrect ones
+        try {
+            return IS_FIELD_CONTENT[c];
+        } catch (ArrayIndexOutOfBoundsException ex) {
+            return false;
+        }
+    }
+
+
     /*
      *  Skip any whitespace and position to read the next character. The next character is returned as being able to
      *  'peek()' it allows a small optimisation in some cases.
--- a/java/org/apache/coyote/http2/HPackHuffman.java
+++ b/java/org/apache/coyote/http2/HPackHuffman.java
@@ -21,6 +21,7 @@
 import java.util.HashSet;
 import java.util.Set;

+import org.apache.tomcat.util.http.parser.HttpParser;
 import org.apache.tomcat.util.res.StringManager;

 public class HPackHuffman {
@@ -365,12 +366,35 @@
      * @param target The target for the decompressed data
      *
      * @throws HpackException If the Huffman encoded value in HPACK headers did not end with EOS padding
+     *
+     * @deprecated Will be removed in Tomcat 12. Use {@link #decode(ByteBuffer, int, StringBuilder, boolean)}
      */
+    @Deprecated
     public static void decode(ByteBuffer data, int length, StringBuilder target) throws HpackException {
+        decode(data, length, target, false);
+    }
+
+    /**
+     * Decodes a huffman encoded string into the target StringBuilder. There must be enough space left in the buffer for
+     * this method to succeed.
+     *
+     * @param data        The byte buffer
+     * @param length      The length of data from the buffer to decode
+     * @param target      The target for the decompressed data
+     * @param isFieldName {@code true} if a field name is being decoded (names have a more restrictive set of allowed
+     *                        characters than field values)
+     *
+     * @throws HpackException If the Huffman encoded value in HPACK headers did not end with EOS padding
+     */
+    public static void decode(ByteBuffer data, int length, StringBuilder target, boolean isFieldName)
+            throws HpackException {
+
         assert data.remaining() >= length;
         int treePos = 0;
         boolean eosBits = true;
         int eosBitCount = 0;
+        boolean firstChar = true;
+        char c = 'a';
         for (int i = 0; i < length; ++i) {
             byte b = data.get();
             int bitPos = 7;
@@ -384,9 +408,30 @@
                     if ((val & LOW_TERMINAL_BIT) == 0) {
                         treePos = val & LOW_MASK;
                         eosBits = false;
+                        // Found a zero, can't be counting EOS bits
                         eosBitCount = 0;
                     } else {
-                        target.append((char) (val & LOW_MASK));
+                        c = (char) (val & LOW_MASK);
+                        if (isFieldName) {
+                            if (!HttpParser.isToken(c) || Character.isUpperCase(c)) {
+                                throw new IllegalArgumentException(sm
+                                        .getString("hpackhuffman.decode.illegalCharacterName", Character.toString(c)));
+                            }
+                        } else {
+                            if (firstChar) {
+                                if (!HttpParser.isFieldVChar(c)) {
+                                    throw new IllegalArgumentException(sm.getString(
+                                            "hpackhuffman.decode.illegalCharacterValue.start", Character.toString(c)));
+                                }
+                                firstChar = false;
+                            } else {
+                                if (!HttpParser.isFieldContent(c)) {
+                                    throw new IllegalArgumentException(sm.getString(
+                                            "hpackhuffman.decode.illegalCharacterValue", Character.toString(c)));
+                                }
+                            }
+                        }
+                        target.append(c);
                         treePos = 0;
                         eosBits = true;
                     }
@@ -404,7 +448,27 @@
                             // as an error
                             throw new HpackException(sm.getString("hpackhuffman.stringLiteralEOS"));
                         }
-                        target.append((char) ((val >> 16) & LOW_MASK));
+                        c = (char) ((val >> 16) & LOW_MASK);
+                        if (isFieldName) {
+                            if (!HttpParser.isToken(c) || Character.isUpperCase(c)) {
+                                throw new IllegalArgumentException(sm
+                                        .getString("hpackhuffman.decode.illegalCharacterName", Character.toString(c)));
+                            }
+                        } else {
+                            if (firstChar) {
+                                if (!HttpParser.isFieldVChar(c)) {
+                                    throw new IllegalArgumentException(sm.getString(
+                                            "hpackhuffman.decode.illegalCharacterValue.start", Character.toString(c)));
+                                }
+                                firstChar = false;
+                            } else {
+                                if (!HttpParser.isFieldContent(c)) {
+                                    throw new IllegalArgumentException(sm.getString(
+                                            "hpackhuffman.decode.illegalCharacterValue", Character.toString(c)));
+                                }
+                            }
+                        }
+                        target.append(c);
                         treePos = 0;
                         eosBits = true;
                     }
@@ -420,6 +484,10 @@
         if (!eosBits) {
             throw new HpackException(sm.getString("hpackhuffman.huffmanEncodedHpackValueDidNotEndWithEOS"));
         }
+        if (!isFieldName && !HttpParser.isFieldVChar(c)) {
+            throw new IllegalArgumentException(
+                    sm.getString("hpackhuffman.decode.illegalCharacterValue.end", Character.toString(c)));
+        }
     }


--- a/java/org/apache/coyote/http2/HpackDecoder.java
+++ b/java/org/apache/coyote/http2/HpackDecoder.java
@@ -20,6 +20,7 @@

 import org.apache.juli.logging.Log;
 import org.apache.juli.logging.LogFactory;
+import org.apache.tomcat.util.http.parser.HttpParser;
 import org.apache.tomcat.util.res.StringManager;

 /**
@@ -114,7 +115,7 @@
                     buffer.position(originalPos);
                     return;
                 }
-                String headerValue = readHpackString(buffer);
+                String headerValue = readHpackString(buffer, false);
                 if (headerValue == null) {
                     buffer.position(originalPos);
                     return;
@@ -128,7 +129,7 @@
                     buffer.position(originalPos);
                     return;
                 }
-                String headerValue = readHpackString(buffer);
+                String headerValue = readHpackString(buffer, false);
                 if (headerValue == null) {
                     buffer.position(originalPos);
                     return;
@@ -141,7 +142,7 @@
                     buffer.position(originalPos);
                     return;
                 }
-                String headerValue = readHpackString(buffer);
+                String headerValue = readHpackString(buffer, false);
                 if (headerValue == null) {
                     buffer.position(originalPos);
                     return;
@@ -202,11 +203,11 @@
         } else if (index != 0) {
             return handleIndexedHeaderName(index);
         } else {
-            return readHpackString(buffer);
+            return readHpackString(buffer, true);
         }
     }

-    private String readHpackString(ByteBuffer buffer) throws HpackException {
+    private String readHpackString(ByteBuffer buffer, boolean isFieldName) throws HpackException {
         if (!buffer.hasRemaining()) {
             return null;
         }
@@ -218,18 +219,34 @@
         }
         boolean huffman = (data & 0b10000000) != 0;
         if (huffman) {
-            return readHuffmanString(length, buffer);
+            return readHuffmanString(length, buffer, isFieldName);
         }
         StringBuilder stringBuilder = new StringBuilder(length);
         for (int i = 0; i < length; ++i) {
-            stringBuilder.append((char) buffer.get());
+            char c = (char) (buffer.get() & 0xFF);
+            if (isFieldName) {
+                if (HttpParser.isToken(c) && !Character.isUpperCase(c)) {
+                    stringBuilder.append(c);
+                } else {
+                    throw new IllegalArgumentException(
+                            sm.getString("hpackdecoder.illegalCharacterName", Character.toString(c)));
+                }
+            } else {
+                if ((i == 0 || i == length - 1) && HttpParser.isFieldVChar(c) ||
+                        i > 0 && i < length - 1 && HttpParser.isFieldContent(c)) {
+                    stringBuilder.append(c);
+                } else {
+                    throw new IllegalArgumentException(
+                            sm.getString("hpackdecoder.illegalCharacterValue", Character.toString(c)));
+                }
+            }
         }
         return stringBuilder.toString();
     }

-    private String readHuffmanString(int length, ByteBuffer buffer) throws HpackException {
+    private String readHuffmanString(int length, ByteBuffer buffer, boolean isFieldName) throws HpackException {
         StringBuilder stringBuilder = new StringBuilder(length);
-        HPackHuffman.decode(buffer, length, stringBuilder);
+        HPackHuffman.decode(buffer, length, stringBuilder, isFieldName);
         return stringBuilder.toString();
     }

--- a/java/org/apache/coyote/http2/Http2Parser.java
+++ b/java/org/apache/coyote/http2/Http2Parser.java
@@ -523,6 +523,8 @@
             } catch (HpackException hpe) {
                 throw new ConnectionException(sm.getString("http2Parser.processFrameHeaders.decodingFailed"),
                         Http2Error.COMPRESSION_ERROR, hpe);
+            } catch (IllegalArgumentException iae) {
+                throw new StreamException("Invalid headers", Http2Error.PROTOCOL_ERROR, streamId, iae);
             }

             // switches to write mode
--- a/java/org/apache/coyote/http2/Stream.java
+++ b/java/org/apache/coyote/http2/Stream.java
@@ -22,8 +22,7 @@
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
 import java.util.Collections;
 import java.util.HashSet;
-import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -321,14 +320,7 @@
             log.trace(sm.getString("stream.header.debug", getConnectionId(), getIdAsString(), name, value));
         }

-        // Header names must be lowercase
-        if (!name.toLowerCase(Locale.US).equals(name)) {
-            headerException =
-                    new StreamException(sm.getString("stream.headercase", getConnectionId(), getIdAsString(), name),
-                            Http2Error.PROTOCOL_ERROR, getIdAsInt());
-            // No need for further processing. The stream will be reset.
-            return;
-        }
+        // Field header names being all lower case is enforced in HpackDecoder.

         if (HTTP_CONNECTION_SPECIFIC_HEADERS.contains(name)) {
             headerException = new StreamException(
--- a/java/org/apache/coyote/http2/StreamException.java
+++ b/java/org/apache/coyote/http2/StreamException.java
@@ -34,6 +34,12 @@
     }


+    StreamException(String msg, Http2Error error, int streamId, Throwable cause) {
+        super(msg, error, cause);
+        this.streamId = streamId;
+    }
+
+
     int getStreamId() {
         return streamId;
     }
--- a/java/org/apache/coyote/http2/LocalStrings.properties
+++ b/java/org/apache/coyote/http2/LocalStrings.properties
@@ -48,4 +48,6 @@
 hpackdecoder.headerTableIndexInvalid=The header table index [{0}] is not valid as there are [{1}] static entries and [{2}] dynamic entries
+hpackdecoder.illegalCharacterName=The illegal [{0}] character was found when decoding an HTTP/2 header field name
+hpackdecoder.illegalCharacterValue=The illegal [{0}] character was found when decoding an HTTP/2 header field value
 hpackdecoder.maxMemorySizeExceeded=The header table size [{0}] exceeds the maximum size [{1}]
 hpackdecoder.notImplemented=Not yet implemented
 hpackdecoder.nullHeader=Null header at index [{0}]
@@ -55,6 +57,10 @@
 hpackdecoder.useStatic=Using header from index [{0}] of static table
 hpackdecoder.zeroNotValidHeaderTableIndex=Zero is not a valid header table index

+hpackhuffman.decode.illegalCharacterName=The illegal [{0}] character was found when decoding an HTTP/2 header field name
+hpackhuffman.decode.illegalCharacterValue=The illegal [{0}] character was found when decoding an HTTP/2 header field value
+hpackhuffman.decode.illegalCharacterValue.end=The illegal [{0}] character was found when decoding the final character in an HTTP/2 header field value
+hpackhuffman.decode.illegalCharacterValue.start=The illegal [{0}] character was found when decoding the first character in an HTTP/2 header field value
 hpackhuffman.huffmanEncodedHpackValueDidNotEndWithEOS=Huffman encoded value in HPACK headers did not end with EOS padding
 hpackhuffman.stringLiteralEOS=Huffman encoded value in HPACK headers contained the EOS symbol
 hpackhuffman.stringLiteralTooMuchPadding=More than 7 bits of EOS padding were provided at the end of an Huffman encoded string literal
@@ -92,7 +98,6 @@

 stream.clientResetRequest=Client reset the stream before the request was fully read
 stream.closed=Connection [{0}], Stream [{1}], Unable to write to stream once it has been closed
-stream.header.case=Connection [{0}], Stream [{1}], HTTP header name [{2}] must be in lower case
 stream.header.connection=Connection [{0}], Stream [{1}], HTTP header [{2}] is not permitted in an HTTP/2 request
 stream.header.contentLength=Connection [{0}], Stream [{1}], The content length header value [{2}] does not agree with the size of the data received [{3}]
 stream.header.debug=Connection [{0}], Stream [{1}], HTTP header [{2}], Value [{3}]

--- a/java/org/apache/catalina/servlets/WebdavServlet.java
+++ b/java/org/apache/catalina/servlets/WebdavServlet.java
@@ -175,6 +175,8 @@
  *  &lt;/init-param&gt;
  * </pre>
  * <p>
+ * By default, WebDAV request bodies for LOCK and PROPFIND are limited to 4096 bytes. To change this limit, set the
+ * <code>maxRequestBodySize</code> <code>init-param</code> for the WebDAV servlet.
  *
  * @see <a href="https://tools.ietf.org/html/rfc4918">RFC 4918</a>
  */
@@ -202,6 +204,12 @@
     private static final int MAX_DEPTH = 3;


+    /*
+     * Default max request body size.
+     */
+    private static final int DEFAULT_MAX_REQUEST_BODY_SIZE = 4096;
+
+
     /**
      * Default namespace.
      */
@@ -215,6 +223,7 @@
             "\n  <D:lockentry><D:lockscope><D:exclusive/></D:lockscope><D:locktype><D:write/></D:locktype></D:lockentry>\n" +
                     "  <D:lockentry><D:lockscope><D:shared/></D:lockscope><D:locktype><D:write/></D:locktype></D:lockentry>\n";

+
     /**
      * Simple date format for the creation date ISO representation (partial).
      */
@@ -227,6 +236,7 @@
      */
     protected static final String LOCK_SCHEME = "urn:uuid:";

+
     // ----------------------------------------------------- Instance Variables

     /**
@@ -247,6 +257,9 @@
     private int maxDepth = MAX_DEPTH;


+    private int maxRequestBodySize = DEFAULT_MAX_REQUEST_BODY_SIZE;
+
+
     /**
      * Is access allowed via WebDAV to the special paths (/WEB-INF and /META-INF)?
      */
@@ -293,6 +306,10 @@
             maxDepth = Integer.parseInt(getServletConfig().getInitParameter("maxDepth"));
         }

+        if (getServletConfig().getInitParameter("maxRequestBodySize") != null) {
+            maxRequestBodySize = Integer.parseInt(getServletConfig().getInitParameter("maxRequestBodySize"));
+        }
+
         if (getServletConfig().getInitParameter("allowSpecialPaths") != null) {
             allowSpecialPaths = Boolean.parseBoolean(getServletConfig().getInitParameter("allowSpecialPaths"));
         }
@@ -829,13 +846,23 @@
             }
         }

+        // Short-cut if client provided a content length
+        if (req.getContentLengthLong() > maxRequestBodySize) {
+            resp.sendError(WebdavStatus.SC_REQUEST_TOO_LONG);
+            return;
+        }
+
         byte[] body;
-        try (InputStream is = req.getInputStream(); ByteArrayOutputStream os = new ByteArrayOutputStream()) {
+        try (InputStream is = req.getInputStream();
+                BoundedByteArrayOutputStream os = new BoundedByteArrayOutputStream(maxRequestBodySize)) {
             IOTools.flow(is, os);
             body = os.toByteArray();
         } catch (IOException ioe) {
             resp.sendError(WebdavStatus.SC_BAD_REQUEST);
             return;
+        } catch (ArrayIndexOutOfBoundsException e) {
+            resp.sendError(WebdavStatus.SC_REQUEST_TOO_LONG);
+            return;
         }
         if (body.length > 0) {
             DocumentBuilder documentBuilder = getDocumentBuilder();
@@ -1388,13 +1416,23 @@

         Node lockInfoNode = null;

+        // Short-cut if client provided a content length
+        if (req.getContentLengthLong() > maxRequestBodySize) {
+            resp.sendError(WebdavStatus.SC_REQUEST_TOO_LONG);
+            return;
+        }
+
         byte[] body;
-        try (InputStream is = req.getInputStream(); ByteArrayOutputStream os = new ByteArrayOutputStream()) {
+        try (InputStream is = req.getInputStream();
+                BoundedByteArrayOutputStream os = new BoundedByteArrayOutputStream(maxRequestBodySize)) {
             IOTools.flow(is, os);
             body = os.toByteArray();
         } catch (IOException ioe) {
             resp.sendError(WebdavStatus.SC_BAD_REQUEST);
             return;
+        } catch (ArrayIndexOutOfBoundsException e) {
+            resp.sendError(WebdavStatus.SC_REQUEST_TOO_LONG);
+            return;
         }
         if (body.length > 0) {
             DocumentBuilder documentBuilder = getDocumentBuilder();
@@ -3012,6 +3057,34 @@
     }


+    static class BoundedByteArrayOutputStream extends ByteArrayOutputStream {
+
+        private final int sizeLimit;
+        private int size;
+
+        BoundedByteArrayOutputStream(int sizeLimit) {
+            super();
+            this.sizeLimit = sizeLimit;
+        }
+
+        @Override
+        public synchronized void write(int b) {
+            size++;
+            if (size > sizeLimit) {
+                throw new ArrayIndexOutOfBoundsException();
+            }
+            super.write(b);
+        }
+
+        @Override
+        public synchronized void write(byte[] b, int off, int len) {
+            size += len;
+            if (size > sizeLimit) {
+                throw new ArrayIndexOutOfBoundsException();
+            }
+            super.write(b, off, len);
+        }
+    }
 }


--- /dev/null
+++ b/test/org/apache/catalina/servlets/TestWebdavBoundedByteArrayOutputStream.java
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.catalina.servlets;
+
+import java.io.IOException;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import org.apache.catalina.servlets.WebdavServlet.BoundedByteArrayOutputStream;
+
+public class TestWebdavBoundedByteArrayOutputStream {
+
+    private static final int TEST_LIMIT = 10;
+    private static final byte[] ONE_BYTE_ARRAY = new byte[] { 0 };
+
+
+    @Test
+    public void testWriteByte() {
+        BoundedByteArrayOutputStream bbaos = new BoundedByteArrayOutputStream(TEST_LIMIT);
+
+        for (int i = 0; i < TEST_LIMIT; i++) {
+            bbaos.write(0);
+        }
+
+        try {
+            bbaos.write(0);
+            Assert.fail("Writing 11th byte failed to trigger error");
+        } catch (ArrayIndexOutOfBoundsException e) {
+            // Pass
+        }
+    }
+
+
+    @Test
+    public void testWriteByteArray() throws IOException {
+        BoundedByteArrayOutputStream bbaos = new BoundedByteArrayOutputStream(TEST_LIMIT);
+
+        for (int i = 0; i < TEST_LIMIT; i++) {
+            bbaos.write(ONE_BYTE_ARRAY);
+        }
+
+        try {
+            bbaos.write(ONE_BYTE_ARRAY);
+            Assert.fail("Writing 11th byte failed to trigger error");
+        } catch (ArrayIndexOutOfBoundsException e) {
+            // Pass
+        }
+    }
+
+
+    @Test
+    public void testWriteByteSubArray() {
+        BoundedByteArrayOutputStream bbaos = new BoundedByteArrayOutputStream(TEST_LIMIT);
+
+        for (int i = 0; i < TEST_LIMIT; i++) {
+            bbaos.write(ONE_BYTE_ARRAY, 0, 1);
+        }
+
+        try {
+            bbaos.write(ONE_BYTE_ARRAY, 0, 1);
+            Assert.fail("Writing 11th byte failed to trigger error");
+        } catch (ArrayIndexOutOfBoundsException e) {
+            // Pass
+        }
+    }
+
+
+    @Test
+    public void testWriteBytes() {
+        BoundedByteArrayOutputStream bbaos = new BoundedByteArrayOutputStream(TEST_LIMIT);
+
+        for (int i = 0; i < TEST_LIMIT; i++) {
+            bbaos.writeBytes(ONE_BYTE_ARRAY);
+        }
+
+        try {
+            bbaos.writeBytes(ONE_BYTE_ARRAY);
+            Assert.fail("Writing 11th byte failed to trigger error");
+        } catch (ArrayIndexOutOfBoundsException e) {
+            // Pass
+        }
+    }
+}
--- a/test/org/apache/catalina/servlets/TestWebdavServlet.java
+++ b/test/org/apache/catalina/servlets/TestWebdavServlet.java
@@ -75,6 +75,7 @@

         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-specialpath"+UUID.randomUUID());
+        tempWebapp.deleteOnExit();
         Assert.assertTrue("Failed to mkdirs on "+tempWebapp.getCanonicalPath(),tempWebapp.mkdirs());
         Assert.assertTrue(new File(tempWebapp,"WEB-INF").mkdir());
         Assert.assertTrue(new File(tempWebapp,"META-INF").mkdir());
@@ -297,6 +298,7 @@

         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-properties");
+        tempWebapp.deleteOnExit();
         Assert.assertTrue(tempWebapp.mkdirs());
         Context ctxt = tomcat.addContext("", tempWebapp.getAbsolutePath());
         Wrapper webdavServlet = Tomcat.addServlet(ctxt, "webdav", new WebdavServlet());
@@ -441,6 +443,7 @@

         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-webapp");
+        tempWebapp.deleteOnExit();
         Assert.assertTrue(tempWebapp.mkdirs());
         Context ctxt = tomcat.addContext("", tempWebapp.getAbsolutePath());
         Wrapper webdavServlet = Tomcat.addServlet(ctxt, "webdav", new WebdavServlet());
@@ -922,6 +925,7 @@

         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-subpath");
+        tempWebapp.deleteOnExit();
         File subPath = new File(tempWebapp, "aaa");
         Assert.assertTrue(subPath.mkdirs());

@@ -1018,6 +1022,7 @@

         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-lock");
+        tempWebapp.deleteOnExit();
         Assert.assertTrue(tempWebapp.mkdirs());
         Context ctxt = tomcat.addContext("", tempWebapp.getAbsolutePath());
         Wrapper webdavServlet = Tomcat.addServlet(ctxt, "webdav", new WebdavServlet());
@@ -1405,6 +1410,7 @@

         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-if");
+        tempWebapp.deleteOnExit();
         File folder = new File(tempWebapp, "/myfolder/myfolder2/myfolder4/myfolder5");
         Assert.assertTrue(folder.mkdirs());
         File file = new File(folder, "myfile.txt");
@@ -1535,6 +1541,7 @@

         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-store");
+        tempWebapp.deleteOnExit();
         Assert.assertTrue(tempWebapp.mkdirs());
         Context ctxt = tomcat.addContext("", tempWebapp.getAbsolutePath());
         Wrapper webdavServlet = Tomcat.addServlet(ctxt, "webdav", new WebdavServlet());
@@ -1565,6 +1572,82 @@
         Assert.assertTrue(client.getResponseBody().contains(">testvalue</mytestproperty>"));
         validateXml(client.getResponseBody());
     }
+
+
+    /*
+     * Only tests LOCK bodies exceeding limit. Other tests cover valid LOCK bodies.
+     */
+    @Test
+    public void testLockBodyLimit() throws Exception {
+        doTestLimit("LOCK", LOCK_BODY);
+    }
+
+
+    /*
+     * Only tests PROPFIND bodies exceeding limit. Other tests cover valid PROPFIND bodies.
+     */
+    @Test
+    public void testPropFindBodyLimit() throws Exception {
+        doTestLimit("PROPFIND", PROPFIND_PROP);
+    }
+
+
+    private void doTestLimit(String method, String requestBody) throws Exception {
+
+        Tomcat tomcat = getTomcatInstance();
+
+        File appDir = new File("test/webapp");
+        Context ctxt = tomcat.addContext("", appDir.getAbsolutePath());
+
+        Wrapper webdavServlet = Tomcat.addServlet(ctxt, "webdav", new WebdavServlet());
+        webdavServlet.addInitParameter("listings", "true");
+        webdavServlet.addInitParameter("secret", "foo");
+        webdavServlet.addInitParameter("readonly", "false");
+        webdavServlet.addInitParameter("useStrongETags", "true");
+        webdavServlet.addInitParameter("maxRequestBodySize", "10");
+
+        ctxt.addServletMappingDecoded("/*", "webdav");
+        tomcat.start();
+
+        // With content length
+        Client client = new Client();
+        client.setPort(getPort());
+
+        // @formatter:off
+        client.setRequest(new String[] {
+                method + " / HTTP/1.1" + CRLF +
+                    "Host: localhost:" + getPort() + CRLF +
+                    "Content-Length: " + requestBody.length() + CRLF +
+                    "Connection: Close" + CRLF +
+                    CRLF +
+                    requestBody
+                });
+        // @formatter:on
+        client.connect();
+        client.processRequest(true);
+        Assert.assertEquals(WebdavStatus.SC_REQUEST_TOO_LONG, client.getStatusCode());
+
+        // Without content length
+        client.reset();
+
+        // @formatter:off
+        client.setRequest(new String[] {
+                method + " / HTTP/1.1" + CRLF +
+                    "Host: localhost:" + getPort() + CRLF +
+                    "Transfer-Encoding: chunked" + CRLF +
+                    "Connection: Close" + CRLF +
+                    CRLF +
+                    Integer.toHexString(requestBody.length()) + CRLF +
+                    requestBody + CRLF +
+                    "0" + CRLF +
+                    CRLF
+                });
+        // @formatter:on
+        client.connect();
+        client.processRequest(true);
+        Assert.assertEquals(WebdavStatus.SC_REQUEST_TOO_LONG, client.getStatusCode());
+    }
+

     public static class CustomPropertyStore implements PropertyStore {


--- a/java/org/apache/catalina/realm/LockOutRealm.java
+++ b/java/org/apache/catalina/realm/LockOutRealm.java
@@ -20,6 +20,7 @@
 import java.security.Principal;
 import java.security.cert.X509Certificate;
 import java.util.LinkedHashMap;
+import java.util.Locale;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicInteger;
 
@@ -72,7 +73,12 @@
      */
     protected Map<String,LockRecord> failedUsers = null;
 
+    /*
+     * Should user names (the keys) in the failedUsers Map be treated in a case sensitive manner. The default is false.
+     */
+    private boolean caseSensitive = false;
 
+
     @Override
     protected void startInternal() throws LifecycleException {
         /*
@@ -207,6 +213,9 @@
      * time will be recorded and any attempt to authenticate a locked user will log a warning.
      */
     public boolean isLocked(String username) {
+        if (!getCaseSensitive()) {
+            username = username.toLowerCase(Locale.ROOT);
+        }
         LockRecord lockRecord;
         synchronized (this) {
             lockRecord = failedUsers.get(username);
@@ -229,6 +238,9 @@
      * After successful authentication, any record of previous authentication failure is removed.
      */
     private synchronized void registerAuthSuccess(String username) {
+        if (!getCaseSensitive()) {
+            username = username.toLowerCase(Locale.ROOT);
+        }
         // Successful authentication means removal from the list of failed users
         failedUsers.remove(username);
     }
@@ -238,6 +250,9 @@
      * After a failed authentication, add the record of the failed authentication.
      */
     private void registerAuthFailure(String username) {
+        if (!getCaseSensitive()) {
+            username = username.toLowerCase(Locale.ROOT);
+        }
         LockRecord lockRecord;
         synchronized (this) {
             if (!failedUsers.containsKey(username)) {
@@ -336,6 +351,16 @@
      */
     public void setCacheRemovalWarningTime(int cacheRemovalWarningTime) {
         this.cacheRemovalWarningTime = cacheRemovalWarningTime;
+    }
+
+
+    public boolean getCaseSensitive() {
+        return caseSensitive;
+    }
+
+
+    public void setCaseSensitive(boolean caseSensitive) {
+        this.caseSensitive = caseSensitive;
     }
 
 

--- a/java/org/apache/catalina/realm/RealmBase.java
+++ b/java/org/apache/catalina/realm/RealmBase.java
@@ -1120,12 +1120,19 @@
      * @return the digest for the specified user
      */
     protected String getDigest(String username, String realmName, String algorithm) {
+        String password = getPassword(username);
+
+        // Short-cut null password case
+        if (password == null) {
+            return null;
+        }
+
         if (hasMessageDigest(algorithm)) {
             // Use pre-generated digest
-            return getPassword(username);
+            return password;
         }

-        String digestValue = username + ":" + realmName + ":" + getPassword(username);
+        String digestValue = username + ":" + realmName + ":" + password;

         byte[] valueBytes;
         try {

diff --git a/java/org/apache/tomcat/websocket/Authenticator.java b/java/org/apache/tomcat/websocket/Authenticator.java
index cdf20cfd06..9745c0286e 100644
--- a/java/org/apache/tomcat/websocket/Authenticator.java
+++ b/java/org/apache/tomcat/websocket/Authenticator.java
@@ -21,6 +21,7 @@ import java.util.Map;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import org.apache.tomcat.util.http.Method;
 import org.apache.tomcat.util.res.StringManager;
 
 /**
@@ -46,9 +47,34 @@ public abstract class Authenticator {
      * @return The generated authorization header value
      *
      * @throws AuthenticationException When an error occurs
+     *
+     * @deprecated Unused. Will be remove in Tomcat 12. Use
+     *                 {@link #getAuthorization(String, String, String, String, String, String)}
+     */
+    @Deprecated
+    public String getAuthorization(String requestUri, String authenticateHeader, String userName, String userPassword,
+            String userRealm) throws AuthenticationException {
+        return getAuthorization(Method.GET, requestUri, authenticateHeader, userName, userPassword, userRealm);
+    }
+
+
+    /**
+     * Generate the authorization header value that will be sent to the server.
+     *
+     * @param method             The request method
+     * @param requestUri         The request URI
+     * @param authenticateHeader The server authentication header received
+     * @param userName           The username
+     * @param userPassword       The user password
+     * @param userRealm          The realm for which the provided username and password are valid. {@code null} to
+     *                               indicate all realms.
+     *
+     * @return The generated authorization header value
+     *
+     * @throws AuthenticationException When an error occurs
      */
-    public abstract String getAuthorization(String requestUri, String authenticateHeader, String userName,
-            String userPassword, String userRealm) throws AuthenticationException;
+    public abstract String getAuthorization(String method, String requestUri, String authenticateHeader,
+            String userName, String userPassword, String userRealm) throws AuthenticationException;
 
 
     /**
diff --git a/java/org/apache/tomcat/websocket/BasicAuthenticator.java b/java/org/apache/tomcat/websocket/BasicAuthenticator.java
index b669d2eb94..cf11d72031 100644
--- a/java/org/apache/tomcat/websocket/BasicAuthenticator.java
+++ b/java/org/apache/tomcat/websocket/BasicAuthenticator.java
@@ -30,8 +30,8 @@ public class BasicAuthenticator extends Authenticator {
     public static final String charsetparam = "charset";
 
     @Override
-    public String getAuthorization(String requestUri, String authenticateHeader, String userName, String userPassword,
-            String userRealm) throws AuthenticationException {
+    public String getAuthorization(String method, String requestUri, String authenticateHeader, String userName,
+            String userPassword, String userRealm) throws AuthenticationException {
 
         validateUsername(userName);
         validatePassword(userPassword);
diff --git a/java/org/apache/tomcat/websocket/DigestAuthenticator.java b/java/org/apache/tomcat/websocket/DigestAuthenticator.java
index 9f018a7a5e..e68b33fab6 100644
--- a/java/org/apache/tomcat/websocket/DigestAuthenticator.java
+++ b/java/org/apache/tomcat/websocket/DigestAuthenticator.java
@@ -40,8 +40,8 @@ public class DigestAuthenticator extends Authenticator {
     private long cNonce;
 
     @Override
-    public String getAuthorization(String requestUri, String authenticateHeader, String userName, String userPassword,
-            String userRealm) throws AuthenticationException {
+    public String getAuthorization(String method, String requestUri, String authenticateHeader, String userName,
+            String userPassword, String userRealm) throws AuthenticationException {
 
         validateUsername(userName);
         validatePassword(userPassword);
@@ -79,8 +79,8 @@ public class DigestAuthenticator extends Authenticator {
 
         try {
             challenge.append("response=\"");
-            challenge.append(
-                    calculateRequestDigest(requestUri, userName, userPassword, realm, nonce, messageQop, algorithm));
+            challenge.append(calculateRequestDigest(method, requestUri, userName, userPassword, realm, nonce,
+                    messageQop, algorithm));
             challenge.append("\",");
         }
 
@@ -89,7 +89,9 @@ public class DigestAuthenticator extends Authenticator {
         }
 
         challenge.append("algorithm=").append(algorithm).append(",");
-        challenge.append("opaque=\"").append(opaque).append("\",");
+        if (opaque != null) {
+            challenge.append("opaque=\"").append(opaque).append("\",");
+        }
 
         if (!messageQop.isEmpty()) {
             challenge.append("qop=\"").append(messageQop).append("\"");
@@ -101,8 +103,8 @@ public class DigestAuthenticator extends Authenticator {
 
     }
 
-    private String calculateRequestDigest(String requestUri, String userName, String password, String realm,
-            String nonce, String qop, String algorithm) throws NoSuchAlgorithmException {
+    private String calculateRequestDigest(String method, String requestUri, String userName, String password,
+            String realm, String nonce, String qop, String algorithm) throws NoSuchAlgorithmException {
 
         boolean session = false;
         if (algorithm.endsWith("-sess")) {
@@ -123,7 +125,7 @@ public class DigestAuthenticator extends Authenticator {
          * If the "qop" value is "auth-int", then A2 is: A2 = Method ":" digest-uri-value ":" H(entity-body) since we do
          * not have an entity-body, A2 = Method ":" digest-uri-value for auth and auth_int
          */
-        String A2 = "GET:" + requestUri;
+        String A2 = method + ":" + requestUri;
 
         preDigest.append(encode(algorithm, A1));
         preDigest.append(':');
diff --git a/java/org/apache/tomcat/websocket/WsWebSocketContainer.java b/java/org/apache/tomcat/websocket/WsWebSocketContainer.java
index 4c542dc828..b039f4ab51 100644
--- a/java/org/apache/tomcat/websocket/WsWebSocketContainer.java
+++ b/java/org/apache/tomcat/websocket/WsWebSocketContainer.java
@@ -69,6 +69,7 @@ import org.apache.tomcat.InstanceManager;
 import org.apache.tomcat.InstanceManagerBindings;
 import org.apache.tomcat.util.buf.StringUtils;
 import org.apache.tomcat.util.collections.CaseInsensitiveKeyMap;
+import org.apache.tomcat.util.http.Method;
 import org.apache.tomcat.util.res.StringManager;
 
 public class WsWebSocketContainer implements WebSocketContainer, BackgroundProcess {
@@ -173,40 +174,41 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
 
 
     private Session connectToServerRecursive(ClientEndpointHolder clientEndpointHolder,
-            ClientEndpointConfig clientEndpointConfiguration, URI path, Set<URI> redirectSet)
+            ClientEndpointConfig clientEndpointConfiguration, URI serverEndpointUri, Set<URI> redirectSet)
             throws DeploymentException {
 
         if (log.isTraceEnabled()) {
-            log.trace(sm.getString("wsWebSocketContainer.connect.entry", clientEndpointHolder.getClassName(), path));
+            log.trace(sm.getString("wsWebSocketContainer.connect.entry", clientEndpointHolder.getClassName(),
+                    serverEndpointUri));
         }
 
         boolean secure = false;
         ByteBuffer proxyConnect = null;
-        URI proxyPath;
+        URI proxyUri;
 
         // Validate scheme (and build proxyPath)
-        String scheme = path.getScheme();
+        String scheme = serverEndpointUri.getScheme();
         if ("ws".equalsIgnoreCase(scheme)) {
-            proxyPath = URI.create("http" + path.toString().substring(2));
+            proxyUri = URI.create("http" + serverEndpointUri.toString().substring(2));
         } else if ("wss".equalsIgnoreCase(scheme)) {
-            proxyPath = URI.create("https" + path.toString().substring(3));
+            proxyUri = URI.create("https" + serverEndpointUri.toString().substring(3));
             secure = true;
         } else {
             throw new DeploymentException(sm.getString("wsWebSocketContainer.pathWrongScheme", scheme));
         }
 
-        // Validate host
-        String host = path.getHost();
-        if (host == null) {
+        // Validate server endpoint host
+        String serverEndpointHost = serverEndpointUri.getHost();
+        if (serverEndpointHost == null) {
             throw new DeploymentException(sm.getString("wsWebSocketContainer.pathNoHost"));
         }
-        int port = path.getPort();
+        int serverEndpointPort = serverEndpointUri.getPort();
 
         SocketAddress sa = null;
 
         // Check to see if a proxy is configured. Javadoc indicates return value
         // will never be null
-        List<Proxy> proxies = ProxySelector.getDefault().select(proxyPath);
+        List<Proxy> proxies = ProxySelector.getDefault().select(proxyUri);
         Proxy selectedProxy = null;
         for (Proxy proxy : proxies) {
             if (proxy.type().equals(Proxy.Type.HTTP)) {
@@ -223,12 +225,12 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
 
         // If the port is not explicitly specified, compute it based on the
         // scheme
-        if (port == -1) {
+        if (serverEndpointPort == -1) {
             if ("ws".equalsIgnoreCase(scheme)) {
-                port = 80;
+                serverEndpointPort = 80;
             } else {
                 // Must be wss due to scheme validation above
-                port = 443;
+                serverEndpointPort = 443;
             }
         }
 
@@ -236,21 +238,23 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
 
         // If sa is null, no proxy is configured so need to create sa
         if (sa == null) {
-            sa = new InetSocketAddress(host, port);
+            sa = new InetSocketAddress(serverEndpointHost, serverEndpointPort);
         } else {
-            proxyConnect = createProxyRequest(host, port,
+            proxyConnect = createProxyRequest(serverEndpointHost, serverEndpointPort,
                     (String) userProperties.get(Constants.PROXY_AUTHORIZATION_HEADER_NAME));
         }
 
         // Create the initial HTTP request to open the WebSocket connection
-        Map<String,List<String>> reqHeaders = createRequestHeaders(host, port, secure, clientEndpointConfiguration);
-        clientEndpointConfiguration.getConfigurator().beforeRequest(reqHeaders);
-        if (Constants.DEFAULT_ORIGIN_HEADER_VALUE != null && !reqHeaders.containsKey(Constants.ORIGIN_HEADER_NAME)) {
+        Map<String,List<String>> upgradeRequestHeaders =
+                createRequestHeaders(serverEndpointHost, serverEndpointPort, secure, clientEndpointConfiguration);
+        clientEndpointConfiguration.getConfigurator().beforeRequest(upgradeRequestHeaders);
+        if (Constants.DEFAULT_ORIGIN_HEADER_VALUE != null &&
+                !upgradeRequestHeaders.containsKey(Constants.ORIGIN_HEADER_NAME)) {
             List<String> originValues = new ArrayList<>(1);
             originValues.add(Constants.DEFAULT_ORIGIN_HEADER_VALUE);
-            reqHeaders.put(Constants.ORIGIN_HEADER_NAME, originValues);
+            upgradeRequestHeaders.put(Constants.ORIGIN_HEADER_NAME, originValues);
         }
-        ByteBuffer request = createRequest(path, reqHeaders);
+        ByteBuffer upgradeRequest = createRequest(serverEndpointUri, upgradeRequestHeaders);
 
         // Get the connection timeout
         long timeout = Constants.IO_TIMEOUT_MS_DEFAULT;
@@ -286,19 +290,23 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
                 writeRequest(channel, proxyConnect, timeout);
                 HttpResponse httpResponse = processResponse(response, channel, timeout);
                 if (httpResponse.status == Constants.PROXY_AUTHENTICATION_REQUIRED) {
-                    return processAuthenticationChallenge(clientEndpointHolder, clientEndpointConfiguration, path,
-                            redirectSet, userProperties, request, httpResponse, AuthenticationType.PROXY);
+                    return processAuthenticationChallenge(clientEndpointHolder, clientEndpointConfiguration,
+                            serverEndpointUri, redirectSet, userProperties, Method.CONNECT,
+                            serverEndpointHost + ":" + serverEndpointPort, httpResponse, AuthenticationType.PROXY);
                 } else if (httpResponse.status() != 200) {
                     throw new DeploymentException(sm.getString("wsWebSocketContainer.proxyConnectFail", selectedProxy,
                             Integer.toString(httpResponse.status())));
                 }
+                // Proxy authentication either successful or not required.
+                userProperties.remove(Constants.PROXY_AUTHORIZATION_HEADER_NAME);
             }
 
             if (secure) {
                 // Regardless of whether a non-secure wrapper was created for a
                 // proxy CONNECT, need to use TLS from this point on so wrap the
                 // original AsynchronousSocketChannel
-                SSLEngine sslEngine = createSSLEngine(clientEndpointConfiguration, host, port);
+                SSLEngine sslEngine =
+                        createSSLEngine(clientEndpointConfiguration, serverEndpointHost, serverEndpointPort);
                 channel = new AsyncChannelWrapperSecure(socketChannel, sslEngine);
             } else if (channel == null) {
                 // Only need to wrap as this point if it wasn't wrapped to process a
@@ -318,10 +326,10 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
                 } catch (IOException ioe) {
                     // Ignore
                 }
-                log.trace(sm.getString("wsWebSocketContainer.connect.write", Integer.valueOf(request.position()),
-                        Integer.valueOf(request.limit()), localAddress));
+                log.trace(sm.getString("wsWebSocketContainer.connect.write", Integer.valueOf(upgradeRequest.position()),
+                        Integer.valueOf(upgradeRequest.limit()), localAddress));
             }
-            writeRequest(channel, request, timeout);
+            writeRequest(channel, upgradeRequest, timeout);
 
             HttpResponse httpResponse = processResponse(response, channel, timeout);
 
@@ -334,6 +342,9 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
 
             if (httpResponse.status != 101) {
                 if (isRedirectStatus(httpResponse.status)) {
+                    // HTTP redirect. Authentication either successful or not required.
+                    userProperties.remove(Constants.AUTHORIZATION_HEADER_NAME);
+
                     List<String> locationHeader =
                             httpResponse.handshakeResponse().getHeaders().get(Constants.LOCATION_HEADER_NAME);
 
@@ -346,7 +357,7 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
                     URI redirectLocation = URI.create(locationHeader.get(0)).normalize();
 
                     if (!redirectLocation.isAbsolute()) {
-                        redirectLocation = path.resolve(redirectLocation);
+                        redirectLocation = serverEndpointUri.resolve(redirectLocation);
                     }
 
                     String redirectScheme = redirectLocation.getScheme().toLowerCase(Locale.ENGLISH);
@@ -367,14 +378,20 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
                             redirectSet);
 
                 } else if (httpResponse.status == Constants.UNAUTHORIZED) {
-                    return processAuthenticationChallenge(clientEndpointHolder, clientEndpointConfiguration, path,
-                            redirectSet, userProperties, request, httpResponse, AuthenticationType.WWW);
+                    String authenticationUri =
+                            new String(upgradeRequest.array(), StandardCharsets.ISO_8859_1).split("\\s", 3)[1];
+                    return processAuthenticationChallenge(clientEndpointHolder, clientEndpointConfiguration,
+                            serverEndpointUri, redirectSet, userProperties, Method.GET, authenticationUri, httpResponse,
+                            AuthenticationType.WWW);
 
                 } else {
                     throw new DeploymentException(
                             sm.getString("wsWebSocketContainer.invalidStatus", Integer.toString(httpResponse.status)));
                 }
             }
+            // HTTP upgrade successful. Authentication either successful or not required.
+            userProperties.remove(Constants.AUTHORIZATION_HEADER_NAME);
+
             HandshakeResponse handshakeResponse = httpResponse.handshakeResponse();
             clientEndpointConfiguration.getConfigurator().afterResponse(handshakeResponse);
 
@@ -417,7 +434,7 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
             success = true;
         } catch (ExecutionException | InterruptedException | SSLException | EOFException | TimeoutException |
                 URISyntaxException | AuthenticationException e) {
-            throw new DeploymentException(sm.getString("wsWebSocketContainer.httpRequestFailed", path), e);
+            throw new DeploymentException(sm.getString("wsWebSocketContainer.httpRequestFailed", serverEndpointUri), e);
         } finally {
             if (!success) {
                 if (channel != null) {
@@ -461,9 +478,10 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
 
 
     private Session processAuthenticationChallenge(ClientEndpointHolder clientEndpointHolder,
-            ClientEndpointConfig clientEndpointConfiguration, URI path, Set<URI> redirectSet,
-            Map<String,Object> userProperties, ByteBuffer request, HttpResponse httpResponse,
-            AuthenticationType authenticationType) throws DeploymentException, AuthenticationException {
+            ClientEndpointConfig clientEndpointConfiguration, URI serverEndpointUri, Set<URI> redirectSet,
+            Map<String,Object> userProperties, String authenticationMethod, String authenticationUri,
+            HttpResponse httpResponse, AuthenticationType authenticationType)
+            throws DeploymentException, AuthenticationException {
 
         if (userProperties.get(authenticationType.getAuthorizationHeaderName()) != null) {
             throw new DeploymentException(sm.getString("wsWebSocketContainer.failedAuthentication",
@@ -488,15 +506,14 @@ public class WsWebSocketContainer implements WebSocketContainer, BackgroundProce
                     Integer.valueOf(httpResponse.status), authScheme));
         }
 
-        String requestUri = new String(request.array(), StandardCharsets.ISO_8859_1).split("\\s", 3)[1];
-
         userProperties.put(authenticationType.getAuthorizationHeaderName(),
-                auth.getAuthorization(requestUri, authenticateHeaders.get(0),
+                auth.getAuthorization(authenticationMethod, authenticationUri, authenticateHeaders.get(0),
                         (String) userProperties.get(authenticationType.getUserNameProperty()),
                         (String) userProperties.get(authenticationType.getUserPasswordProperty()),
                         (String) userProperties.get(authenticationType.getUserRealmProperty())));
 
-        return connectToServerRecursive(clientEndpointHolder, clientEndpointConfiguration, path, redirectSet);
+        return connectToServerRecursive(clientEndpointHolder, clientEndpointConfiguration, serverEndpointUri,
+                redirectSet);
     }
 
 
diff --git a/test/org/apache/tomcat/websocket/TesterWebSocketClientProxy.java b/test/org/apache/tomcat/websocket/TesterWebSocketClientProxy.java
index 4182b5bb2b..021d729059 100644
--- a/test/org/apache/tomcat/websocket/TesterWebSocketClientProxy.java
+++ b/test/org/apache/tomcat/websocket/TesterWebSocketClientProxy.java
@@ -58,11 +58,22 @@ import org.apache.tomcat.websocket.TesterMessageCountClient.TesterProgrammaticEn
  *     ProxyVia On
  *     AllowCONNECT 0-65535
  *     <Proxy *>
- *         Order deny,allow
- *         Allow from all
  *         AuthType Basic
  *         AuthName "Proxy Password Required"
- *         AuthUserFile password.file
+ *         AuthUserFile "/etc/apache2/password.file"
+ *         Require valid-user
+ *     </Proxy>
+ * </VirtualHost>
+ *
+ * Listen 8890
+ * <VirtualHost *:8890>
+ *     ProxyRequests On
+ *     ProxyVia On
+ *     AllowCONNECT 0-65535
+ *     <Proxy *>
+ *         AuthType Digest
+ *         AuthName "Proxy Password Required"
+ *         AuthUserFile "/etc/apache2/password-digest.file"
  *         Require valid-user
  *     </Proxy>
  * </VirtualHost>
@@ -71,16 +82,20 @@ import org.apache.tomcat.websocket.TesterMessageCountClient.TesterProgrammaticEn
  * # htpasswd -c password.file proxy
  * New Password: proxy-pass
  *
+ * # htdigest -c password-digest.file proxy
+ * New Password: proxy-pass
+ *
  */
 public class TesterWebSocketClientProxy extends WebSocketBaseTest {
 
     private static final String MESSAGE_STRING = "proxy-test-message";
 
-    private static final String PROXY_ADDRESS = "192.168.0.200";
+    private static final String PROXY_ADDRESS = "192.168.23.32";
     private static final String PROXY_PORT_NO_AUTH = "8888";
-    private static final String PROXY_PORT_AUTH = "8889";
+    private static final String PROXY_PORT_BASIC_AUTH = "8889";
+    private static final String PROXY_PORT_DIGEST_AUTH = "8890";
     // The IP address of the test instance that is reachable from the proxy
-    private static final String TOMCAT_ADDRESS = "192.168.0.100";
+    private static final String TOMCAT_ADDRESS = "192.168.23.12";
 
     private static final String TOMCAT_USER = "tomcat";
     private static final String TOMCAT_PASSWORD = "tomcat-pass";
@@ -91,35 +106,49 @@ public class TesterWebSocketClientProxy extends WebSocketBaseTest {
 
     @Test
     public void testConnectToServerViaProxyWithNoAuthentication() throws Exception {
-        doTestConnectToServerViaProxy(false, false);
+        doTestConnectToServerViaProxy(false, null);
     }
 
 
     @Test
     public void testConnectToServerViaProxyWithServerAuthentication() throws Exception {
-        doTestConnectToServerViaProxy(true, false);
+        doTestConnectToServerViaProxy(true, null);
+    }
+
+
+    @Test
+    public void testConnectToServerViaProxyWithProxyBasicAuthentication() throws Exception {
+        doTestConnectToServerViaProxy(false, "basic");
+    }
+
+
+    @Test
+    public void testConnectToServerViaProxyWithServerAndProxyBasicAuthentication() throws Exception {
+        doTestConnectToServerViaProxy(true, "basic");
     }
 
 
     @Test
-    public void testConnectToServerViaProxyWithProxyAuthentication() throws Exception {
-        doTestConnectToServerViaProxy(false, true);
+    public void testConnectToServerViaProxyWithProxyDigestAuthentication() throws Exception {
+        doTestConnectToServerViaProxy(false, "digest");
     }
 
 
     @Test
-    public void testConnectToServerViaProxyWithServerAndProxyAuthentication() throws Exception {
-        doTestConnectToServerViaProxy(true, true);
+    public void testConnectToServerViaProxyWithServerAndProxyDigestAuthentication() throws Exception {
+        doTestConnectToServerViaProxy(true, "digest");
     }
 
 
-    private void doTestConnectToServerViaProxy(boolean serverAuthentication, boolean proxyAuthentication)
+    private void doTestConnectToServerViaProxy(boolean serverAuthentication, String proxyAuthentication)
             throws Exception {
 
         // Configure the proxy
         System.setProperty("http.proxyHost", PROXY_ADDRESS);
-        if (proxyAuthentication) {
-            System.setProperty("http.proxyPort", PROXY_PORT_AUTH);
+        if ("basic".equalsIgnoreCase(proxyAuthentication)) {
+            System.setProperty("http.proxyPort", PROXY_PORT_BASIC_AUTH);
+        } else if ("digest".equalsIgnoreCase(proxyAuthentication)) {
+            System.setProperty("http.proxyPort", PROXY_PORT_DIGEST_AUTH);
         } else {
             System.setProperty("http.proxyPort", PROXY_PORT_NO_AUTH);
         }
@@ -166,7 +195,7 @@ public class TesterWebSocketClientProxy extends WebSocketBaseTest {
             clientEndpointConfig.getUserProperties().put(Constants.WS_AUTHENTICATION_USER_NAME, TOMCAT_USER);
             clientEndpointConfig.getUserProperties().put(Constants.WS_AUTHENTICATION_PASSWORD, TOMCAT_PASSWORD);
         }
-        if (proxyAuthentication) {
+        if (proxyAuthentication != null) {
             clientEndpointConfig.getUserProperties().put(Constants.WS_AUTHENTICATION_PROXY_USER_NAME, PROXY_USER);
             clientEndpointConfig.getUserProperties().put(Constants.WS_AUTHENTICATION_PROXY_PASSWORD, PROXY_PASSWORD);
         }
* Unmerged path webapps/docs/changelog.xml

--- a/java/org/apache/catalina/realm/RealmBase.java
+++ b/java/org/apache/catalina/realm/RealmBase.java
@@ -658,8 +658,6 @@ public abstract class RealmBase extends LifecycleMBeanBase implements Realm {
                         constraints[i].included(uri, method));
             }
 
-            boolean matched = false;
-            int pos = -1;
             for (int j = 0; j < collection.length; j++) {
                 String[] patterns = collection[j].findPatterns();
 
@@ -669,6 +667,7 @@ public abstract class RealmBase extends LifecycleMBeanBase implements Realm {
                     continue;
                 }
 
+                boolean matched = false;
                 for (int k = 0; k < patterns.length && !matched; k++) {
                     String pattern = patterns[k];
                     if (pattern.startsWith("*.")) {
@@ -678,19 +677,18 @@ public abstract class RealmBase extends LifecycleMBeanBase implements Realm {
                                 uri.length() - dot == pattern.length() - 1) {
                             if (pattern.regionMatches(1, uri, dot, uri.length() - dot)) {
                                 matched = true;
-                                pos = j;
                             }
                         }
                     }
                 }
-            }
-            if (matched) {
-                found = true;
-                if (collection[pos].findMethod(method)) {
-                    if (results == null) {
-                        results = new ArrayList<>();
+                if (matched) {
+                    found = true;
+                    if (collection[j].findMethod(method)) {
+                        if (results == null) {
+                            results = new ArrayList<>();
+                        }
+                        results.add(constraints[i]);
                     }
-                    results.add(constraints[i]);
                 }
             }
         }
--- a/test/org/apache/catalina/realm/TestRealmBase.java
+++ b/test/org/apache/catalina/realm/TestRealmBase.java
@@ -790,4 +790,86 @@ public class TestRealmBase {
         Assert.assertFalse(mapRealm.hasResourcePermission(
                 request, response, constraintsDelete, null));
     }
+
+
+    @Test
+    public void testUncoveredMethods() throws IOException {
+        // Create a constraint for ROLE1
+        SecurityConstraint constraint = new SecurityConstraint();
+        constraint.addAuthRole(ROLE1);
+        // Add a collection for GET
+        SecurityCollection getCollection = new SecurityCollection();
+        getCollection.addMethod(Method.GET);
+        getCollection.addPatternDecoded("*.html");
+        constraint.addCollection(getCollection);
+        // Add a collection for POST
+        SecurityCollection postCollection = new SecurityCollection();
+        postCollection.addMethod(Method.POST);
+        postCollection.addPatternDecoded("*.html");
+        constraint.addCollection(postCollection);
+
+        TesterMapRealm mapRealm = new TesterMapRealm();
+
+        // Set up the mock request and response
+        TesterRequest request = new TesterRequest();
+        Response response = new TesterResponse();
+        Context context = request.getContext();
+        context.addSecurityRole(ROLE1);
+        context.addSecurityRole(ROLE2);
+        request.getMappingData().context = context;
+
+        // Create the principals
+        List<String> userRoles1 = new ArrayList<>();
+        userRoles1.add(ROLE1);
+        GenericPrincipal gp1 = new GenericPrincipal(USER1, userRoles1);
+
+        List<String> userRoles2 = new ArrayList<>();
+        userRoles2.add(ROLE2);
+        GenericPrincipal gp2 = new GenericPrincipal(USER2, userRoles2);
+
+        List<String> userRoles99 = new ArrayList<>();
+        GenericPrincipal gp99 = new GenericPrincipal(USER99, userRoles99);
+
+        // Add the constraint to the context
+        context.addConstraint(constraint);
+
+
+        // Only user1 should be able to perform a GET
+        request.setMethod(Method.GET);
+
+        SecurityConstraint[] constraintsGet =
+                mapRealm.findSecurityConstraints(request, context);
+
+        request.setUserPrincipal(null);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsGet, null));
+        request.setUserPrincipal(gp1);
+        Assert.assertTrue(mapRealm.hasResourcePermission(
+                request, response, constraintsGet, null));
+        request.setUserPrincipal(gp2);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsGet, null));
+        request.setUserPrincipal(gp99);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsGet, null));
+
+        // Only user1 should be able to perform a POST
+        request.setMethod(Method.POST);
+
+        SecurityConstraint[] constraintsPost =
+                mapRealm.findSecurityConstraints(request, context);
+
+        request.setUserPrincipal(null);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsPost, null));
+        request.setUserPrincipal(gp1);
+        Assert.assertTrue(mapRealm.hasResourcePermission(
+                request, response, constraintsPost, null));
+        request.setUserPrincipal(gp2);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsPost, null));
+        request.setUserPrincipal(gp99);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsPost, null));
+    }
 }
--- a/test/org/apache/tomcat/unittest/TesterRequest.java
+++ b/test/org/apache/tomcat/unittest/TesterRequest.java
@@ -31,6 +31,7 @@ import jakarta.servlet.SessionTrackingMode;
 import org.apache.catalina.Context;
 import org.apache.catalina.connector.Request;
 import org.apache.catalina.session.StandardSession;
+import org.apache.tomcat.util.buf.MessageBytes;
 
 public class TesterRequest extends Request {
 
@@ -81,6 +82,15 @@ public class TesterRequest extends Request {
         return "/level1/level2/foo.html";
     }
 
+
+    @Override
+    public MessageBytes getRequestPathMB() {
+        MessageBytes result = MessageBytes.newInstance();
+        result.setString(getRequestURI());
+        return result;
+    }
+
+
     @Override
     public String getDecodedRequestURI() {
         // Decoding not required
