--- a/src/standalone/index.ts
+++ b/src/standalone/index.ts
@@ -1,5 +1,6 @@
 import type { WithRequired } from '@apollo/utils.withrequired';
 import cors from 'cors';
+import { parse as parseContentType } from 'content-type';
 import express from 'express';
 import http, { type IncomingMessage, type ServerResponse } from 'http';
 import type { ListenOptions } from 'net';
@@ -29,6 +30,12 @@
   >;
 }
 
+// according to RFC8259, only UTF-8 is allowed in JSON text
+// (see https://datatracker.ietf.org/doc/html/rfc8259#section-8.1)
+// RFC 7159 also specifies that JSON could be UTF-16 or UTF-32,
+// so we allow for that, too
+const validCharset = /^utf-(8|((16|32)(le|be)?))$/i;
+
 export async function startStandaloneServer(
   server: ApolloServer<BaseContext>,
   options?: StartStandaloneServerOptions<BaseContext> & {
@@ -57,7 +64,24 @@
   const context = options?.context ?? (async () => ({}) as TContext);
   app.use(
     cors(),
-    express.json({ limit: '50mb' }),
+    express.json({
+      verify(req) {
+        const charset =
+          parseContentType(req).parameters.charset || 'utf-8';
+        if (!charset.match(validCharset)) {
+          throw Object.assign(
+            new Error(`unsupported charset "${charset.toUpperCase()}"`),
+            {
+              status: 415,
+              name: 'UnsupportedMediaTypeError',
+              charset,
+              type: 'charset.unsupported',
+            },
+          );
+        }
+      },
+      limit: '50mb',
+    }),
     expressMiddleware(server, { context }),
   );
 
--- a/dist/cjs/standalone/index.js
+++ b/dist/cjs/standalone/index.js
@@ -5,18 +5,37 @@
 Object.defineProperty(exports, "__esModule", { value: true });
 exports.startStandaloneServer = void 0;
 const cors_1 = __importDefault(require("cors"));
+const content_type_1 = require("content-type");
 const express_1 = __importDefault(require("express"));
 const http_1 = __importDefault(require("http"));
 const index_js_1 = require("../express4/index.js");
 const index_js_2 = require("../plugin/drainHttpServer/index.js");
 const urlForHttpServer_js_1 = require("../utils/urlForHttpServer.js");
+// according to RFC8259, only UTF-8 is allowed in JSON text
+// (see https://datatracker.ietf.org/doc/html/rfc8259#section-8.1)
+// RFC 7159 also specifies that JSON could be UTF-16 or UTF-32,
+// so we allow for that, too
+const validCharset = /^utf-(8|((16|32)(le|be)?))$/i;
 async function startStandaloneServer(server, options) {
     const app = (0, express_1.default)();
     const httpServer = http_1.default.createServer(app);
     server.addPlugin((0, index_js_2.ApolloServerPluginDrainHttpServer)({ httpServer: httpServer }));
     await server.start();
     const context = options?.context ?? (async () => ({}));
-    app.use((0, cors_1.default)(), express_1.default.json({ limit: '50mb' }), (0, index_js_1.expressMiddleware)(server, { context }));
+    app.use((0, cors_1.default)(), express_1.default.json({
+        verify(req) {
+            const charset = (0, content_type_1.parse)(req).parameters.charset || 'utf-8';
+            if (!charset.match(validCharset)) {
+                throw Object.assign(new Error(`unsupported charset "${charset.toUpperCase()}"`), {
+                    status: 415,
+                    name: 'UnsupportedMediaTypeError',
+                    charset,
+                    type: 'charset.unsupported',
+                });
+            }
+        },
+        limit: '50mb',
+    }), (0, index_js_1.expressMiddleware)(server, { context }));
     const listenOptions = options?.listen ?? { port: 4000 };
     await new Promise((resolve) => {
         httpServer.listen(listenOptions, resolve);
--- a/dist/esm/standalone/index.js
+++ b/dist/esm/standalone/index.js
@@ -1,16 +1,35 @@
 import cors from 'cors';
+import { parse as parseContentType } from 'content-type';
 import express from 'express';
 import http from 'http';
 import { expressMiddleware } from '../express4/index.js';
 import { ApolloServerPluginDrainHttpServer } from '../plugin/drainHttpServer/index.js';
 import { urlForHttpServer } from '../utils/urlForHttpServer.js';
+// according to RFC8259, only UTF-8 is allowed in JSON text
+// (see https://datatracker.ietf.org/doc/html/rfc8259#section-8.1)
+// RFC 7159 also specifies that JSON could be UTF-16 or UTF-32,
+// so we allow for that, too
+const validCharset = /^utf-(8|((16|32)(le|be)?))$/i;
 export async function startStandaloneServer(server, options) {
     const app = express();
     const httpServer = http.createServer(app);
     server.addPlugin(ApolloServerPluginDrainHttpServer({ httpServer: httpServer }));
     await server.start();
     const context = options?.context ?? (async () => ({}));
-    app.use(cors(), express.json({ limit: '50mb' }), expressMiddleware(server, { context }));
+    app.use(cors(), express.json({
+        verify(req) {
+            const charset = parseContentType(req).parameters.charset || 'utf-8';
+            if (!charset.match(validCharset)) {
+                throw Object.assign(new Error(`unsupported charset "${charset.toUpperCase()}"`), {
+                    status: 415,
+                    name: 'UnsupportedMediaTypeError',
+                    charset,
+                    type: 'charset.unsupported',
+                });
+            }
+        },
+        limit: '50mb',
+    }), expressMiddleware(server, { context }));
     const listenOptions = options?.listen ?? { port: 4000 };
     await new Promise((resolve) => {
         httpServer.listen(listenOptions, resolve);
--- a/package.json
+++ b/package.json
@@ -141,6 +141,7 @@
     "@types/express-serve-static-core": "^4.17.30",
     "@types/node-fetch": "^2.6.1",
     "async-retry": "^1.2.1",
+    "content-type": "^1.0.5",
     "cors": "^2.8.5",
     "express": "^4.17.1",
     "loglevel": "^1.6.8",
