--- a/src/main/java/org/apache/commons/compress/archivers/tar/TarArchiveEntry.java
+++ b/src/main/java/org/apache/commons/compress/archivers/tar/TarArchiveEntry.java
@@ -212,6 +212,17 @@
     /** Convert millis to seconds */
     public static final int MILLIS_PER_SECOND = 1000;

+    /**
+     * Regular expression pattern for validating values in pax extended header file time fields.
+     * These fields contain two numeric values (seconds and sub-second values) as per this definition:
+     * https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html#tag_20_92_13_05
+     * <p>
+     * Since they are parsed into long values, maximum length of each is the same as Long.MAX_VALUE
+     * which is 19 digits.
+     * </p>
+     */
+    private static final String PAX_EXTENDED_HEADER_FILE_TIMES_PATTERN = "-?\\d{1,19}(?:\\.\\d{1,19})?";
+
     private static FileTime fileTimeFromOptionalSeconds(long seconds) {
         if (seconds <= 0) {
             return null;
@@ -260,7 +271,12 @@
         return fileName;
     }

-    private static Instant parseInstantFromDecimalSeconds(final String value) {
+    private static Instant parseInstantFromDecimalSeconds(final String value) throws IOException {
+        // Validate field values to prevent denial of service attacks with BigDecimal values (see JDK-6560193)
+        if (!value.matches(TarArchiveEntry.PAX_EXTENDED_HEADER_FILE_TIMES_PATTERN)) {
+            throw new IOException("Corrupted PAX header. Time field value is invalid '" + value + "'");
+        }
+
         final BigDecimal epochSeconds = new BigDecimal(value);
         final long seconds = epochSeconds.longValue();
         final long nanos = epochSeconds.remainder(BigDecimal.ONE).movePointRight(9).longValue();
--- a/src/test/java/org/apache/commons/compress/archivers/tar/TarArchiveEntryTest.java
+++ b/src/test/java/org/apache/commons/compress/archivers/tar/TarArchiveEntryTest.java
@@ -40,6 +40,7 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
+import java.util.Random;

 import org.apache.commons.compress.AbstractTestCase;
 import org.apache.commons.compress.archivers.zip.ZipEncodingHelper;
@@ -134,6 +135,29 @@
         t.setSize(0100000000000L);
     }

+    @Test
+    public void testPaxTimeFieldsForInvalidValues() {
+        final String[] headerNames = { "LIBARCHIVE.creationtime", "atime", "mtime", "ctime" };
+        // @formatter:off
+        final String[] testValues = {
+                // Generate a number with a very large integer or fractional component
+                new Random().nextLong() + "." + String.join("",
+                        Collections.nCopies(15000, String.valueOf(Long.MAX_VALUE))),
+                // These two examples use the exponent notation
+                "9e9999999",
+                "9E9999999"
+        };
+        // @formatter:on
+
+        final TarArchiveEntry entry = new TarArchiveEntry("test.txt");
+        for (String name : headerNames) {
+            for (String value : testValues) {
+                Exception exp = assertThrows(IllegalArgumentException.class, () -> entry.addPaxHeader(name, value));
+                assert(exp.getCause().getMessage().startsWith("Corrupted PAX header. Time field value is invalid"));
+            }
+        }
+    }
+
     @Test
     public void testExtraPaxHeaders() throws IOException {
         final ByteArrayOutputStream bos = new ByteArrayOutputStream();

--- a/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveConstants.java
+++ b/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveConstants.java
@@ -69,5 +69,6 @@
      */
     public enum COMPRESSION_TYPE {
-        ZLIB(0),
+        UNKNOWN(-1),
+        ZLIB(0),
         BZLIB(1),
         LZO(2);
@@ -86,5 +86,5 @@
             }
 
-            return null;
+            return UNKNOWN;
         }
     }
--- a/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveUtil.java
+++ b/src/main/java/org/apache/commons/compress/archivers/dump/DumpArchiveUtil.java
@@ -58,4 +58,7 @@
      */
     public static final boolean verify(final byte[] buffer) {
+        if (buffer == null) {
+            return false;
+        }
         // verify magic. for now only accept NFS_MAGIC.
         final int magic = convert32(buffer, 24);
@@ -118,4 +118,7 @@
     static String decode(final ZipEncoding encoding, final byte[] b, final int offset, final int len)
         throws IOException {
+        if (offset > offset + len) {
+            throw new IOException("Invalid offset/length combination");
+        }
         return encoding.decode(Arrays.copyOfRange(b, offset, offset + len));
     }
--- a/src/main/java/org/apache/commons/compress/archivers/dump/TapeInputStream.java
+++ b/src/main/java/org/apache/commons/compress/archivers/dump/TapeInputStream.java
@@ -76,4 +76,7 @@
         }
         blockSize = RECORD_SIZE * recsPerBlock;
+        if (blockSize < 1) {
+            throw new IOException("Block size cannot be less than or equal to 0: " + blockSize);
+        }
 
         // save first block in case we need it again
