diff --git a/dist/node/chunks/dep-D4NMHUTW.js b/dist/node/chunks/dep-D4NMHUTW.js
index 5b079da..d146a99 100644
--- a/dist/node/chunks/dep-D4NMHUTW.js
+++ b/dist/node/chunks/dep-D4NMHUTW.js
@@ -35667,6 +35667,11 @@ async function loadAndTransform(environment, id, url, options, timestamp, mod, r
     err.code = ERR_DENIED_ID;
     throw err;
   }
+  if (!options.allowId && !options.skipFsCheck && environment.config.consumer !== "server" && !id.startsWith("\0") && !isFileLoadingAllowed(environment.getTopLevelConfig(), cleanUrl(id))) {
+    const err = new Error(`Denied ID ${id}`);
+    err.code = ERR_DENIED_ID;
+    throw err;
+  }
   let code = null;
   let map = null;
   const loadStart = debugLoad ? performance$1.now() : 0;
@@ -47884,6 +47889,11 @@ class DevEnvironment extends BaseEnvironment {
     this.hot = context.transport ? isWebSocketServer in context.transport ? context.transport : normalizeHotChannel(context.transport, context.hot) : normalizeHotChannel({}, context.hot);
     this.hot.setInvokeHandler({
       fetchModule: (id, importer, options2) => {
+        if (context.disableFetchModule) {
+          throw new Error(
+            `fetchModule is not supported in this environment`
+          );
+        }
         return this.fetchModule(id, importer, options2);
       }
     });
@@ -48445,7 +48455,8 @@ function defineConfig(config) {
 function defaultCreateClientDevEnvironment(name, config, context) {
   return new DevEnvironment(name, config, {
     hot: true,
-    transport: context.ws
+    transport: context.ws,
+    disableFetchModule: true
   });
 }
 function defaultCreateDevEnvironment(name, config) {
diff --git a/dist/node/index.d.ts b/dist/node/index.d.ts
index f725d4a..4afabc3 100644
--- a/dist/node/index.d.ts
+++ b/dist/node/index.d.ts
@@ -1242,6 +1242,7 @@ interface HotChannel<Api = any> {
      */
     close?(): Promise<unknown> | void;
     api?: Api;
+    skipFsCheck?: boolean;
 }
 /** @deprecated use `HotChannel` instead */
 type HMRChannel = HotChannel;
@@ -1889,6 +1890,7 @@ interface DevEnvironmentContext {
         inlineSourceMap?: boolean;
     };
     depsOptimizer?: DepsOptimizer;
+    disableFetchModule?: boolean;
 }
 declare class DevEnvironment extends BaseEnvironment {
     mode: "dev";
