--- a/core/src/main/java/org/bouncycastle/crypto/modes/G3413CTRBlockCipher.java
+++ b/core/src/main/java/org/bouncycastle/crypto/modes/G3413CTRBlockCipher.java
@@ -13,8 +13,6 @@
 public class G3413CTRBlockCipher
     extends StreamBlockCipher
 {
-
-
     private final int s;
     private byte[] CTR;
     private byte[] IV;
@@ -24,7 +22,6 @@ public class G3413CTRBlockCipher
     private int byteCount = 0;
     private boolean initialized;
 
-
     /**
      * Basic constructor.
      *
@@ -76,7 +73,6 @@ public void init(
         CipherParameters params)
         throws IllegalArgumentException
     {
-
         if (params instanceof ParametersWithIV)
         {
             ParametersWithIV ivParam = (ParametersWithIV)params;
@@ -95,7 +91,7 @@ public void init(
             {
                 CTR[i] = 0;
             }
-
+    
             // if null it's an IV changed only.
             if (ivParam.getParameters() != null)
             {
@@ -184,27 +180,33 @@ protected byte calculateByte(byte in)
         if (byteCount == s)
         {
             byteCount = 0;
-            generateCRT();
+            generateCTR();
         }
 
         return rv;
 
     }
 
-    private void generateCRT()
+    private void generateCTR()
     {
-        CTR[CTR.length - 1]++;
+        int start = CTR.length - 1;
+        while (++CTR[start] == 0)
+        {
+            start--;
+            if (start == IV.length - 1)
+            {
+                throw new IllegalStateException("attempt to process too many blocks");
+            }
+        }
     }
 
 
     private byte[] generateBuf()
     {
-
         byte[] encryptedCTR = new byte[CTR.length];
         cipher.processBlock(CTR, 0, encryptedCTR, 0);
 
         return GOST3413CipherUtil.MSB(encryptedCTR, s);
-
     }
 
 
--- a/core/src/test/java/org/bouncycastle/crypto/test/GOST3412Test.java
+++ b/core/src/test/java/org/bouncycastle/crypto/test/GOST3412Test.java
@@ -1,5 +1,6 @@
 package org.bouncycastle.crypto.test;
 
+import org.bouncycastle.crypto.StreamBlockCipher;
 import org.bouncycastle.crypto.engines.GOST3412_2015Engine;
 import org.bouncycastle.crypto.modes.G3413CBCBlockCipher;
 import org.bouncycastle.crypto.modes.G3413CFBBlockCipher;
@@ -7,6 +8,7 @@
 import org.bouncycastle.crypto.modes.G3413OFBBlockCipher;
 import org.bouncycastle.crypto.params.KeyParameter;
 import org.bouncycastle.crypto.params.ParametersWithIV;
+import org.bouncycastle.util.Arrays;
 import org.bouncycastle.util.encoders.Hex;
 import org.bouncycastle.util.test.SimpleTest;
 
@@ -87,10 +89,66 @@ public void performTest()
     {
         super.performTest();
 
+        ctrTest();
 //        cfbTest();
 //        ofbTest();
     }
 
+    private void ctrTest()
+        throws Exception
+    {
+        StreamBlockCipher sb = new G3413CTRBlockCipher(new GOST3412_2015Engine(), 8);
+
+        sb.init(true, new ParametersWithIV(new KeyParameter(Hex.decode("8899aabbccddeeff0011223344556677fedcba98765432100123456789abcdef")),
+            Hex.decode("0001020304050607")));
+
+        byte[] block = Hex.decode("000102030405060708090a0b0c0d0e0f");
+        byte[] output = new byte[16];
+        byte[] last = new byte[16];
+
+        sb.processBytes(block, 0, block.length, last, 0);
+
+        for (int i = 1; i < 255; i++)
+        {
+            sb.processBytes(block, 0, block.length, output, 0);
+            if (Arrays.areEqual(last, output))
+            {
+                fail("cipher text repeats 1");
+            }
+        }
+
+        sb.processBytes(block, 0, block.length, output, 0);
+        if (Arrays.areEqual(last, output))
+        {
+            fail("cipher text repeats 2");
+        }
+
+        sb = new G3413CTRBlockCipher(new GOST3412_2015Engine(), 8);
+
+        sb.init(true, new ParametersWithIV(new KeyParameter(Hex.decode("8899aabbccddeeff0011223344556677fedcba98765432100123456789abcdef")),
+            Hex.decode("0001020304050607")));
+
+        sb.processBytes(block, 0, block.length, last, 0);
+
+        for (int i = 1; i != ((1 << 15) - 1); i++)
+        {
+            sb.processBytes(block, 0, block.length, output, 0);
+            if (Arrays.areEqual(last, output))
+            {
+                fail("cipher text repeats 3");
+            }
+            byte[] tmp = last;
+            last = output;
+            output = tmp;
+        }
+
+        sb.processBytes(block, 0, block.length, output, 0);
+        if (Arrays.areEqual(last, output))
+        {
+            fail("cipher text repeats");
+        }
+    }
+
     public static void main(
         String[] args)
     {

diff --git a/prov/src/main/java/org/bouncycastle/jce/provider/X509LDAPCertStoreSpi.java b/prov/src/main/java/org/bouncycastle/jce/provider/X509LDAPCertStoreSpi.java
index 5789fd171..91e4150af 100644
--- a/prov/src/main/java/org/bouncycastle/jce/provider/X509LDAPCertStoreSpi.java
+++ b/prov/src/main/java/org/bouncycastle/jce/provider/X509LDAPCertStoreSpi.java
@@ -34,6 +34,7 @@
 import org.bouncycastle.asn1.ASN1InputStream;
 import org.bouncycastle.asn1.x509.CertificatePair;
 import org.bouncycastle.jce.X509LDAPCertStoreParameters;
+import org.bouncycastle.ldap.LDAPUtils;
 import org.bouncycastle.util.Strings;
 
 /**
@@ -50,26 +51,6 @@
 public class X509LDAPCertStoreSpi
     extends CertStoreSpi
 {
-    private static String[] FILTER_ESCAPE_TABLE = new String['\\' + 1];
-
-    static
-    {
-        // Filter encoding table -------------------------------------
-
-        // fill with char itself
-        for (char c = 0; c < FILTER_ESCAPE_TABLE.length; c++)
-        {
-            FILTER_ESCAPE_TABLE[c] = String.valueOf(c);
-        }
-
-        // escapes (RFC2254)
-        FILTER_ESCAPE_TABLE['*'] = "\\2a";
-        FILTER_ESCAPE_TABLE['('] = "\\28";
-        FILTER_ESCAPE_TABLE[')'] = "\\29";
-        FILTER_ESCAPE_TABLE['\\'] = "\\5c";
-        FILTER_ESCAPE_TABLE[0] = "\\00";
-    }
-
     /**
      * Initial Context Factory.
      */
@@ -124,42 +105,6 @@ private DirContext connectLDAP()
         return ctx;
     }
 
-    private String parseDN(String subject, String subjectAttributeName)
-    {
-        String temp = subject;
-        int begin = Strings.toLowerCase(temp).indexOf(Strings.toLowerCase(subjectAttributeName));
-        temp = temp.substring(begin + subjectAttributeName.length());
-        int end = temp.indexOf(',');
-        if (end == -1)
-        {
-            end = temp.length();
-        }
-        while (temp.charAt(end - 1) == '\\')
-        {
-            end = temp.indexOf(',', end + 1);
-            if (end == -1)
-            {
-                end = temp.length();
-            }
-        }
-        temp = temp.substring(0, end);
-        begin = temp.indexOf('=');
-        temp = temp.substring(begin + 1);
-        if (temp.charAt(0) == ' ')
-        {
-            temp = temp.substring(1);
-        }
-        if (temp.startsWith("\""))
-        {
-            temp = temp.substring(1);
-        }
-        if (temp.endsWith("\""))
-        {
-            temp = temp.substring(0, temp.length() - 1);
-        }
-        return filterEncode(temp);
-    }
-
     public Collection engineGetCertificates(CertSelector selector)
         throws CertStoreException
     {
@@ -277,7 +222,7 @@ private Set certSubjectSerialSearch(X509CertSelector xselector,
                         subject = xselector.getSubjectAsString();
                     }
                 }
-                String attrValue = parseDN(subject, subjectAttributeName);
+                String attrValue = LDAPUtils.parseDN(subject, subjectAttributeName);
                 set.addAll(search(attrName, "*" + attrValue + "*", attrs));
                 if (serial != null
                     && params.getSearchForSerialNumberIn() != null)
@@ -374,13 +319,13 @@ public Collection engineGetCRLs(CRLSelector selector)
                 {
                     String issuerAttributeName = params
                         .getCertificateRevocationListIssuerAttributeName();
-                    attrValue = parseDN((String)o, issuerAttributeName);
+                    attrValue = LDAPUtils.parseDN((String)o, issuerAttributeName);
                 }
                 else
                 {
                     String issuerAttributeName = params
                         .getCertificateRevocationListIssuerAttributeName();
-                    attrValue = parseDN(new X500Principal((byte[])o)
+                    attrValue = LDAPUtils.parseDN(new X500Principal((byte[])o)
                         .getName("RFC1779"), issuerAttributeName);
                 }
                 set.addAll(search(attrName, "*" + attrValue + "*", attrs));
@@ -415,43 +360,7 @@ public Collection engineGetCRLs(CRLSelector selector)
 
         return crlSet;
     }
-
-    /**
-     * Escape a value for use in a filter.
-     *
-     * @param value the value to escape.
-     * @return a properly escaped representation of the supplied value.
-     */
-    private String filterEncode(String value)
-    {
-        if (value == null)
-        {
-            return null;
-        }
-
-        // make buffer roomy
-        StringBuilder encodedValue = new StringBuilder(value.length() * 2);
-
-        int length = value.length();
-
-        for (int i = 0; i < length; i++)
-        {
-            char c = value.charAt(i);
-
-            if (c < FILTER_ESCAPE_TABLE.length)
-            {
-                encodedValue.append(FILTER_ESCAPE_TABLE[c]);
-            }
-            else
-            {
-                // default: add the char
-                encodedValue.append(c);
-            }
-        }
-
-        return encodedValue.toString();
-    }
-
+    
     /**
      * Returns a Set of byte arrays with the certificate or CRL encodings.
      *
diff --git a/prov/src/main/java/org/bouncycastle/ldap/LDAPUtils.java b/prov/src/main/java/org/bouncycastle/ldap/LDAPUtils.java
new file mode 100644
index 000000000..2d6bb3bc7
--- /dev/null
+++ b/prov/src/main/java/org/bouncycastle/ldap/LDAPUtils.java
@@ -0,0 +1,112 @@
+package org.bouncycastle.ldap;
+
+import org.bouncycastle.util.Strings;
+
+/**
+ * General utility methods for assisting with preparation of LDAP queries.
+ */
+public class LDAPUtils
+{
+    private static String[] FILTER_ESCAPE_TABLE = new String['\\' + 1];
+
+    static
+    {
+        // Filter encoding table -------------------------------------
+
+        // fill with char itself
+        for (char c = 0; c < FILTER_ESCAPE_TABLE.length; c++)
+        {
+            FILTER_ESCAPE_TABLE[c] = String.valueOf(c);
+        }
+
+        // escapes (RFC2254)
+        FILTER_ESCAPE_TABLE['*'] = "\\2a";
+        FILTER_ESCAPE_TABLE['('] = "\\28";
+        FILTER_ESCAPE_TABLE[')'] = "\\29";
+        FILTER_ESCAPE_TABLE['\\'] = "\\5c";
+        FILTER_ESCAPE_TABLE[0] = "\\00";
+    }
+
+    /**
+     * Parse out the contents of a particular subject attribute name from the string form of an X.500 DN.
+     *
+     * @param subject string form of an X.500 DN.
+     * @param subjectAttributeName the RDN attribute name of interest.
+     * @return an escaped string suitable for use in an LDAP query.
+     */
+    public static String parseDN(String subject, String subjectAttributeName)
+    {
+        String temp = subject;
+        int begin = Strings.toLowerCase(temp).indexOf(Strings.toLowerCase(subjectAttributeName));
+        if (begin == -1)
+        {
+            return "";
+        }
+        temp = temp.substring(begin + subjectAttributeName.length());
+        int end = temp.indexOf(',');
+        if (end == -1)
+        {
+            end = temp.length();
+        }
+        while (temp.charAt(end - 1) == '\\')
+        {
+            end = temp.indexOf(',', end + 1);
+            if (end == -1)
+            {
+                end = temp.length();
+            }
+        }
+        temp = temp.substring(0, end);
+        begin = temp.indexOf('=');
+        temp = temp.substring(begin + 1);
+        if (temp.charAt(0) == ' ')
+        {
+            temp = temp.substring(1);
+        }
+        if (temp.startsWith("\""))
+        {
+            temp = temp.substring(1);
+        }
+        if (temp.endsWith("\""))
+        {
+            temp = temp.substring(0, temp.length() - 1);
+        }
+        return filterEncode(temp);
+    }
+
+    /**
+     * Escape a value for use in a filter.
+     *
+     * @param value the value to escape.
+     * @return a properly escaped representation of the supplied value.
+     */
+    private static String filterEncode(String value)
+    {
+        if (value == null)
+        {
+            return null;
+        }
+
+        // make buffer roomy
+        StringBuilder encodedValue = new StringBuilder(value.length() * 2);
+
+        int length = value.length();
+
+        for (int i = 0; i < length; i++)
+        {
+            char c = value.charAt(i);
+
+            if (c < FILTER_ESCAPE_TABLE.length)
+            {
+                encodedValue.append(FILTER_ESCAPE_TABLE[c]);
+            }
+            else
+            {
+                // default: add the char
+                encodedValue.append(c);
+            }
+        }
+
+        return encodedValue.toString();
+    }
+}
diff --git a/prov/src/main/java/org/bouncycastle/x509/util/LDAPStoreHelper.java b/prov/src/main/java/org/bouncycastle/x509/util/LDAPStoreHelper.java
index 8121a498e..e1af28594 100644
--- a/prov/src/main/java/org/bouncycastle/x509/util/LDAPStoreHelper.java
+++ b/prov/src/main/java/org/bouncycastle/x509/util/LDAPStoreHelper.java
@@ -35,6 +35,7 @@
 import org.bouncycastle.jce.provider.X509CRLParser;
 import org.bouncycastle.jce.provider.X509CertPairParser;
 import org.bouncycastle.jce.provider.X509CertParser;
+import org.bouncycastle.ldap.LDAPUtils;
 import org.bouncycastle.util.StoreException;
 import org.bouncycastle.util.Strings;
 import org.bouncycastle.x509.X509AttributeCertStoreSelector;
@@ -65,7 +66,6 @@
  */
 public class LDAPStoreHelper
 {
-
     // TODO: cache results
 
     private X509LDAPCertStoreParameters params;
@@ -95,7 +95,8 @@ public LDAPStoreHelper(X509LDAPCertStoreParameters params)
      */
     private static final String URL_CONTEXT_PREFIX = "com.sun.jndi.url";
 
-    private DirContext connectLDAP() throws NamingException
+    private DirContext connectLDAP()
+        throws NamingException
     {
         Properties props = new Properties();
         props.setProperty(Context.INITIAL_CONTEXT_FACTORY, LDAP_PROVIDER);
@@ -111,47 +112,6 @@ private DirContext connectLDAP() throws NamingException
         return ctx;
     }
 
-    private String parseDN(String subject, String dNAttributeName)
-    {
-        String temp = subject;
-        int begin = Strings.toLowerCase(temp).indexOf(
-            Strings.toLowerCase(dNAttributeName) + "=");
-        if (begin == -1)
-        {
-            return "";
-        }
-        temp = temp.substring(begin + dNAttributeName.length());
-        int end = temp.indexOf(',');
-        if (end == -1)
-        {
-            end = temp.length();
-        }
-        while (temp.charAt(end - 1) == '\\')
-        {
-            end = temp.indexOf(',', end + 1);
-            if (end == -1)
-            {
-                end = temp.length();
-            }
-        }
-        temp = temp.substring(0, end);
-        begin = temp.indexOf('=');
-        temp = temp.substring(begin + 1);
-        if (temp.charAt(0) == ' ')
-        {
-            temp = temp.substring(1);
-        }
-        if (temp.startsWith("\""))
-        {
-            temp = temp.substring(1);
-        }
-        if (temp.endsWith("\""))
-        {
-            temp = temp.substring(0, temp.length() - 1);
-        }
-        return temp;
-    }
-
     private Set createCerts(List list, X509CertStoreSelector xselector)
         throws StoreException
     {
@@ -224,7 +184,7 @@ private List certSubjectSerialSearch(X509CertStoreSelector xselector,
         {
             for (int i = 0; i < subjectAttributeNames.length; i++)
             {
-                attrValue = parseDN(subject, subjectAttributeNames[i]);
+                attrValue = LDAPUtils.parseDN(subject, subjectAttributeNames[i]);
                 list
                     .addAll(search(attrNames, "*" + attrValue + "*",
                         attrs));
@@ -235,7 +195,7 @@ private List certSubjectSerialSearch(X509CertStoreSelector xselector,
             attrValue = serial;
             list.addAll(search(
                 splitString(params.getSearchForSerialNumberIn()),
-                                                  attrValue, attrs));
+                attrValue, attrs));
         }
         if (serial == null && subject == null)
         {
@@ -246,7 +206,6 @@ private List certSubjectSerialSearch(X509CertStoreSelector xselector,
     }
 
 
-
     /**
      * Can use the subject of the forward certificate of the set certificate
      * pair or the subject of the forward
@@ -290,7 +249,7 @@ private List crossCertificatePairSubjectSearch(
         {
             for (int i = 0; i < subjectAttributeNames.length; i++)
             {
-                attrValue = parseDN(subject, subjectAttributeNames[i]);
+                attrValue = LDAPUtils.parseDN(subject, subjectAttributeNames[i]);
                 list
                     .addAll(search(attrNames, "*" + attrValue + "*",
                         attrs));
@@ -385,7 +344,7 @@ private List attrCertSubjectSerialSearch(
         {
             for (int i = 0; i < subjectAttributeNames.length; i++)
             {
-                attrValue = parseDN(subject, subjectAttributeNames[i]);
+                attrValue = LDAPUtils.parseDN(subject, subjectAttributeNames[i]);
                 list
                     .addAll(search(attrNames, "*" + attrValue + "*",
                         attrs));
@@ -441,11 +400,11 @@ private List cRLIssuerSearch(X509CRLStoreSelector xselector,
         if (xselector.getAttrCertificateChecking() != null)
         {
             Principal principals[] = xselector.getAttrCertificateChecking().getIssuer().getPrincipals();
-            for (int i=0; i<principals.length; i++)
+            for (int i = 0; i < principals.length; i++)
             {
                 if (principals[i] instanceof X500Principal)
                 {
-                    issuers.add(principals[i]);        
+                    issuers.add(principals[i]);
                 }
             }
         }
@@ -457,7 +416,7 @@ private List cRLIssuerSearch(X509CRLStoreSelector xselector,
 
             for (int i = 0; i < issuerAttributeNames.length; i++)
             {
-                attrValue = parseDN(issuer, issuerAttributeNames[i]);
+                attrValue = LDAPUtils.parseDN(issuer, issuerAttributeNames[i]);
                 list
                     .addAll(search(attrNames, "*" + attrValue + "*",
                         attrs));
@@ -485,7 +444,8 @@ private List cRLIssuerSearch(X509CRLStoreSelector xselector,
      *                        directory.
      */
     private List search(String attributeNames[], String attributeValue,
-                        String[] attrs) throws StoreException
+                        String[] attrs)
+        throws StoreException
     {
         String filter = null;
         if (attributeNames == null)
@@ -599,7 +559,8 @@ private Set createCRLs(List list, X509CRLStoreSelector xselector)
     }
 
     private Set createCrossCertificatePairs(List list,
-                                            X509CertPairStoreSelector xselector) throws StoreException
+                                            X509CertPairStoreSelector xselector)
+        throws StoreException
     {
         Set certPairSet = new HashSet();
 
@@ -626,7 +587,7 @@ private Set createCrossCertificatePairs(List list,
                     pair = new X509CertificatePair(new CertificatePair(
                         Certificate
                             .getInstance(new ASN1InputStream(
-                            forward).readObject()),
+                                forward).readObject()),
                         Certificate
                             .getInstance(new ASN1InputStream(
                                 reverse).readObject())));
@@ -652,7 +613,8 @@ private Set createCrossCertificatePairs(List list,
     }
 
     private Set createAttributeCertificates(List list,
-                                            X509AttributeCertStoreSelector xselector) throws StoreException
+                                            X509AttributeCertStoreSelector xselector)
+        throws StoreException
     {
         Set certSet = new HashSet();
 
@@ -719,12 +681,14 @@ public Collection getAuthorityRevocationLists(X509CRLStoreSelector selector)
      * The attributeCertificateRevocationList holds a list of attribute
      * certificates that have been revoked.
      * </p>
+     *
      * @param selector The CRL selector to use to find the CRLs.
      * @return A possible empty collection with CRLs.
      * @throws StoreException
      */
     public Collection getAttributeCertificateRevocationLists(
-        X509CRLStoreSelector selector) throws StoreException
+        X509CRLStoreSelector selector)
+        throws StoreException
     {
         String[] attrs = splitString(params
             .getAttributeCertificateRevocationListAttribute());
@@ -754,12 +718,14 @@ public Collection getAttributeCertificateRevocationLists(
      * The attributeAuthorityList holds a list of AA certificates that have been
      * revoked.
      * </p>
+     *
      * @param selector The CRL selector to use to find the CRLs.
      * @return A possible empty collection with CRLs
      * @throws StoreException
      */
     public Collection getAttributeAuthorityRevocationLists(
-        X509CRLStoreSelector selector) throws StoreException
+        X509CRLStoreSelector selector)
+        throws StoreException
     {
         String[] attrs = splitString(params.getAttributeAuthorityRevocationListAttribute());
         String attrNames[] = splitString(params
@@ -789,7 +755,8 @@ public Collection getAttributeAuthorityRevocationLists(
      * @throws StoreException
      */
     public Collection getCrossCertificatePairs(
-        X509CertPairStoreSelector selector) throws StoreException
+        X509CertPairStoreSelector selector)
+        throws StoreException
     {
         String[] attrs = splitString(params.getCrossCertificateAttribute());
         String attrNames[] = splitString(params.getLdapCrossCertificateAttributeName());
@@ -850,6 +817,7 @@ public Collection getUserCertificates(X509CertStoreSelector selector)
      * <p>
      * The aAcertificate holds the privileges of an attribute authority.
      * </p>
+     *
      * @param selector The selector to find the attribute certificates.
      * @return A possible empty collection with attribute certificates.
      * @throws StoreException
@@ -882,12 +850,14 @@ public Collection getAACertificates(X509AttributeCertStoreSelector selector)
      * authority and holds a description of the privilege and its delegation
      * rules.
      * </p>
+     *
      * @param selector The selector to find the attribute certificates.
      * @return A possible empty collection with attribute certificates.
      * @throws StoreException
      */
     public Collection getAttributeDescriptorCertificates(
-        X509AttributeCertStoreSelector selector) throws StoreException
+        X509AttributeCertStoreSelector selector)
+        throws StoreException
     {
         String[] attrs = splitString(params.getAttributeDescriptorCertificateAttribute());
         String attrNames[] = splitString(params
@@ -916,6 +886,7 @@ public Collection getAttributeDescriptorCertificates(
      * store self-issued certificates (if any) and certificates issued to this
      * CA by CAs in the same realm as this CA.
      * </p>
+     *
      * @param selector The selector to find the certificates.
      * @return A possible empty collection with certificates.
      * @throws StoreException
@@ -948,7 +919,8 @@ public Collection getCACertificates(X509CertStoreSelector selector)
      * @throws StoreException
      */
     public Collection getDeltaCertificateRevocationLists(
-        X509CRLStoreSelector selector) throws StoreException
+        X509CRLStoreSelector selector)
+        throws StoreException
     {
         String[] attrs = splitString(params.getDeltaRevocationListAttribute());
         String attrNames[] = splitString(params.getLdapDeltaRevocationListAttributeName());
@@ -973,12 +945,14 @@ public Collection getDeltaCertificateRevocationLists(
      * <p>
      * The attributeCertificateAttribute holds the privileges of a user
      * </p>
+     *
      * @param selector The selector to find the attribute certificates.
      * @return A possible empty collection with attribute certificates.
      * @throws StoreException
      */
     public Collection getAttributeCertificateAttributes(
-        X509AttributeCertStoreSelector selector) throws StoreException
+        X509AttributeCertStoreSelector selector)
+        throws StoreException
     {
         String[] attrs = splitString(params.getAttributeCertificateAttributeAttribute());
         String attrNames[] = splitString(params
@@ -1007,7 +981,8 @@ public Collection getAttributeCertificateAttributes(
      * @throws StoreException
      */
     public Collection getCertificateRevocationLists(
-        X509CRLStoreSelector selector) throws StoreException
+        X509CRLStoreSelector selector)
+        throws StoreException
     {
         String[] attrs = splitString(params.getCertificateRevocationListAttribute());
         String attrNames[] = splitString(params

--- a/core/src/main/java/org/bouncycastle/pqc/crypto/frodo/FrodoEngine.java
+++ b/core/src/main/java/org/bouncycastle/pqc/crypto/frodo/FrodoEngine.java
@@ -100,31 +100,10 @@
         this.gen = mGen;
     }

-    private short sample(short r)
-    {
-        short t, e;
-        // 1. t = sum_{i=1}^{len_x - 1} r_i * 2^{i-1}
-        t = (short) ((r & 0xffff) >>> 1);
-        e = 0; // 2. e = 0
-        for (int z = 0; z < T_chi.length; z++)
-        {
-            if (t > T_chi[z]) // 4. if t > T_chi(z)
-                e++; // 5. e = e + 1
-        }
-        // 6. e = (-1)^{r_0} * e
-
-        if (((r & 0xffff) % 2) == 1)
-            e = (short) ((e) * (-1) & 0xffff);
-
-        return e;
-    }
-
     private short[] sample_matrix(short[] r, int offset, int n1, int n2)
     {
         short[] E = new short[n1 * n2];
-        for (int i = 0; i < n1; i++)
-            for (int j = 0; j < n2; j++)
-                E[i*n2+j] = sample(r[i * n2 + j + offset]);
+        Noise.sample(T_chi, r, offset, E);
         return E;
     }

@@ -461,20 +440,18 @@
     }


-    private short ctverify(short[] a1, short[] a2, short[] b1, short[] b2)
+    private static short ctverify(short[] a1, short[] a2, short[] b1, short[] b2)
     {
         // Compare two arrays in constant time.
         // Returns 0 if the byte arrays are equal, -1 otherwise.
-        short r = 0;
+        int r = 0;

-        for (short i = 0; i < a1.length; i++)
+        for (int i = 0; i < a1.length; i++)
             r |= a1[i] ^ b1[i];

-        for (short i = 0; i < a2.length; i++)
+        for (int i = 0; i < a2.length; i++)
             r |= a2[i] ^ b2[i];

-//        r = (short) ((-(short)(r >> 1) | -(short)(r & 1)) >> (8*2-1));
-        if (r == 0)
-            return 0;
-        return -1;
+        // Constant-time: returns 0 if equal, -1 if not equal.
+        return (short)((r | (-r)) >> 31);
     }

--- /dev/null
+++ b/core/src/main/java/org/bouncycastle/pqc/crypto/frodo/Noise.java
@@ -0,0 +1,30 @@
+package org.bouncycastle.pqc.crypto.frodo;
+
+abstract class Noise
+{
+    static void sample(short[] cdf, short[] r, int rOff, short[] s)
+    {
+        // No need to compare with the last value.
+//        assert cdf[cdf.length - 1] == 0x7FFF;
+
+        // Fills 's' with samples from the noise distribution 'cdf' using pseudo-random values 'r[rOff..]'
+        for (int i = 0, n = s.length; i < n; ++i)
+        {
+            int sample = 0;
+            int r_i = r[rOff + i] & 0xFFFF;
+            int prnd = r_i >>> 1;   // Drop the least significant bit
+            int sign = r_i & 1;     // Pick the least significant bit
+
+            for (int j = 0; j < cdf.length - 1; ++j)
+            {
+                // Constant time comparison: 1 if cdf[j] < prnd, 0 otherwise.
+                sample += (cdf[j] - prnd) >>> 31;
+            }
+
+            // Assuming that sign is either 0 or 1, flips sample iff sign = 1
+            sample = ((-sign) ^ sample) + sign;
+
+            s[i] = (short)sample;
+        }
+    }
+}

diff --git a/prov/build.gradle b/prov/build.gradle
index ac0900f08..bf2157313 100644
--- a/prov/build.gradle
+++ b/prov/build.gradle
@@ -112,8 +112,7 @@ compileJava21Java {
 }
 
 compileJava25Java {
-    options.release = 25
-    options.sourcepath = files(['src/main/java', 'src/main/jdk25'])
+    enabled = false
 }
 
 
@@ -135,9 +134,6 @@ task sourcesJar(type: Jar) {
     into('META-INF/versions/21') {
         from sourceSets.java21.allSource
     }
-    into('META-INF/versions/25') {
-        from sourceSets.java25.allSource
-    }
 }
 
 jar {
@@ -157,9 +153,6 @@ jar {
     into('META-INF/versions/21') {
         from sourceSets.java21.output
     }
-    into('META-INF/versions/25') {
-        from sourceSets.java25.output
-    }
     String v = "${rootProject.extensions.ext.bundle_version}"
     manifest.attributes('Multi-Release': 'true')
     manifest.attributes('Bundle-Name': 'bcprov')
@@ -270,8 +263,7 @@ compileTest21Java {
 }
 
 compileTest25Java {
-    options.release = 25
-    options.sourcepath = files(['src/test/java', 'src/test/jdk25'])
+    enabled = false
 }
 
 publishing {
diff --git a/tls/build.gradle b/tls/build.gradle
index ecfd90f17..ddd98bc70 100644
--- a/tls/build.gradle
+++ b/tls/build.gradle
@@ -142,19 +142,7 @@ compileTest21Java {
 }
 
 compileJava25Java {
-
-    options.release = 25
-
-    def prov_jar="${project(":prov").jar.outputs.files.getFiles().getAt(0)}"
-    def util_jar="${project(":util").jar.outputs.files.getFiles().getAt(0)}"
-    def pkix_jar="${project(":pkix").jar.outputs.files.getFiles().getAt(0)}"
-
-
-    options.compilerArgs += [
-            '--module-path', "${prov_jar}${File.pathSeparator}${util_jar}${File.pathSeparator}${pkix_jar}"
-    ]
-
-    options.sourcepath = files(['src/main/java', 'src/main/jdk25'])
+    enabled = false
 }
 
 
@@ -166,9 +154,6 @@ task sourcesJar(type: Jar) {
     into('META-INF/versions/9') {
         from sourceSets.java9.allSource
     }
-    into('META-INF/versions/25') {
-        from sourceSets.java25.allSource
-    }
 }
 
 jar {
@@ -176,9 +161,6 @@ jar {
     into('META-INF/versions/9') {
         from sourceSets.java9.output
     }
-    into('META-INF/versions/25') {
-        from sourceSets.java25.output
-    }
     String v = "${rootProject.extensions.ext.bundle_version}"
     manifest.attributes('Multi-Release': 'true')
     manifest.attributes('Bundle-Name': 'bctls')
