diff --git a/lib/form_data.js b/lib/form_data.js
index e05c8f1..c16eb3c 100644
--- a/lib/form_data.js
+++ b/lib/form_data.js
@@ -6,6 +6,7 @@ var https = require('https');
 var parseUrl = require('url').parse;
 var fs = require('fs');
 var Stream = require('stream').Stream;
+var crypto = require('crypto');
 var mime = require('mime-types');
 var asynckit = require('asynckit');
 var populate = require('./populate.js');
@@ -347,12 +348,7 @@ FormData.prototype.getBuffer = function() {
 FormData.prototype._generateBoundary = function() {
   // This generates a 50 character boundary similar to those used by Firefox.
   // They are optimized for boyer-moore parsing.
-  var boundary = '--------------------------';
-  for (var i = 0; i < 24; i++) {
-    boundary += Math.floor(Math.random() * 10).toString(16);
-  }
-
-  this._boundary = boundary;
+  this._boundary = '--------------------------' + crypto.randomBytes(12).toString('hex');
 };
 
 // Note: getLengthSync DOESN'T calculate streams length
-- 
2.43.0

