--- a/dist/shared/lib/router/utils/escape-path-delimiters.js
+++ b/dist/shared/lib/router/utils/escape-path-delimiters.js
@@ -10,7 +10,7 @@
     }
 });
 function escapePathDelimiters(segment, escapeEncoded) {
-    return segment.replace(new RegExp(`([/#?]${escapeEncoded ? '|%(2f|23|3f|5c)' : ''})`, 'gi'), (char)=>encodeURIComponent(char));
+    return segment.replace(new RegExp(`([/#?\\\\]${escapeEncoded ? '|%(2f|23|3f|5c)' : ''})`, 'gi'), (char)=>encodeURIComponent(char));
 }
 
 //# sourceMappingURL=escape-path-delimiters.js.map
\ No newline at end of file
--- a/dist/esm/shared/lib/router/utils/escape-path-delimiters.js
+++ b/dist/esm/shared/lib/router/utils/escape-path-delimiters.js
@@ -1,6 +1,6 @@
 // escape delimiters used by path-to-regexp
 export default function escapePathDelimiters(segment, escapeEncoded) {
-    return segment.replace(new RegExp(`([/#?]${escapeEncoded ? '|%(2f|23|3f|5c)' : ''})`, 'gi'), (char)=>encodeURIComponent(char));
+    return segment.replace(new RegExp(`([/#?\\\\]${escapeEncoded ? '|%(2f|23|3f|5c)' : ''})`, 'gi'), (char)=>encodeURIComponent(char));
 }
 
 //# sourceMappingURL=escape-path-delimiters.js.map
\ No newline at end of file
--- a/dist/server/lib/incremental-cache/file-system-cache.js
+++ b/dist/server/lib/incremental-cache/file-system-cache.js
@@ -310,18 +310,22 @@
         // Wait for all FS operations to complete.
         await writer.wait();
     }
-    getFilePath(pathname, kind) {
+    getFilePath(key, kind) {
+        let rootDir;
         switch(kind){
             case _responsecache.IncrementalCacheKind.FETCH:
                 // we store in .next/cache/fetch-cache so it can be persisted
                 // across deploys
-                return _path.default.join(this.serverDistDir, '..', 'cache', 'fetch-cache', pathname);
+                rootDir = _path.default.join(this.serverDistDir, '..', 'cache', 'fetch-cache');
+                break;
             case _responsecache.IncrementalCacheKind.PAGES:
-                return _path.default.join(this.serverDistDir, 'pages', pathname);
+                rootDir = _path.default.join(this.serverDistDir, 'pages');
+                break;
             case _responsecache.IncrementalCacheKind.IMAGE:
             case _responsecache.IncrementalCacheKind.APP_PAGE:
             case _responsecache.IncrementalCacheKind.APP_ROUTE:
-                return _path.default.join(this.serverDistDir, 'app', pathname);
+                rootDir = _path.default.join(this.serverDistDir, 'app');
+                break;
             default:
                 throw Object.defineProperty(new Error(`Unexpected file path kind: ${kind}`), "__NEXT_ERROR_CODE", {
                     value: "E479",
@@ -329,6 +333,15 @@
                     configurable: true
                 });
         }
+        const filePath = _path.default.join(rootDir, key);
+        if (!(filePath.startsWith(rootDir + _path.default.sep) || filePath === rootDir)) {
+            throw Object.defineProperty(new Error(`Invalid file path: ${filePath}`), "__NEXT_ERROR_CODE", {
+                value: "E1144",
+                enumerable: false,
+                configurable: true
+            });
+        }
+        return filePath;
     }
 }
 
--- a/dist/esm/server/lib/incremental-cache/file-system-cache.js
+++ b/dist/esm/server/lib/incremental-cache/file-system-cache.js
@@ -295,18 +295,22 @@
         // Wait for all FS operations to complete.
         await writer.wait();
     }
-    getFilePath(pathname, kind) {
+    getFilePath(key, kind) {
+        let rootDir;
         switch(kind){
             case IncrementalCacheKind.FETCH:
                 // we store in .next/cache/fetch-cache so it can be persisted
                 // across deploys
-                return path.join(this.serverDistDir, '..', 'cache', 'fetch-cache', pathname);
+                rootDir = path.join(this.serverDistDir, '..', 'cache', 'fetch-cache');
+                break;
             case IncrementalCacheKind.PAGES:
-                return path.join(this.serverDistDir, 'pages', pathname);
+                rootDir = path.join(this.serverDistDir, 'pages');
+                break;
             case IncrementalCacheKind.IMAGE:
             case IncrementalCacheKind.APP_PAGE:
             case IncrementalCacheKind.APP_ROUTE:
-                return path.join(this.serverDistDir, 'app', pathname);
+                rootDir = path.join(this.serverDistDir, 'app');
+                break;
             default:
                 throw Object.defineProperty(new Error(`Unexpected file path kind: ${kind}`), "__NEXT_ERROR_CODE", {
                     value: "E479",
@@ -314,6 +318,15 @@
                     configurable: true
                 });
         }
+        const filePath = path.join(rootDir, key);
+        if (!(filePath.startsWith(rootDir + path.sep) || filePath === rootDir)) {
+            throw Object.defineProperty(new Error(`Invalid file path: ${filePath}`), "__NEXT_ERROR_CODE", {
+                value: "E1144",
+                enumerable: false,
+                configurable: true
+            });
+        }
+        return filePath;
     }
 }
 

--- a/dist/server/image-optimizer.js
+++ b/dist/server/image-optimizer.js
@@ -169,7 +169,8 @@
     ICNS,
     BMP,
     JXL,
-    HEIC
+    HEIC,
+    AVIF
 ];
 const BLUR_IMG_SIZE = 8 // should match `next-image-loader`
 ;
--- a/dist/esm/server/image-optimizer.js
+++ b/dist/esm/server/image-optimizer.js
@@ -49,7 +49,8 @@
     ICNS,
     BMP,
     JXL,
-    HEIC
+    HEIC,
+    AVIF
 ];
 const BLUR_IMG_SIZE = 8 // should match `next-image-loader`
 ;
