--- /dev/null
+++ b/codec-http/src/main/java/io/netty/handler/codec/http/InvalidChunkExtensionException.java
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2025 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.handler.codec.CorruptedFrameException;
+
+/**
+ * Thrown when HTTP chunk extensions could not be parsed, typically due to incorrect use of CR LF delimiters.
+ * <p>
+ * <a href="https://datatracker.ietf.org/doc/html/rfc9112#name-chunked-transfer-coding">RFC 9112</a>
+ * specifies that chunk header lines must be terminated in a CR LF pair,
+ * and that a lone LF octet is not allowed within the chunk header line.
+ */
+public final class InvalidChunkExtensionException extends CorruptedFrameException {
+    private static final long serialVersionUID = 536224937231200736L;
+
+    public InvalidChunkExtensionException() {
+        super("Line Feed must be preceded by Carriage Return when terminating HTTP chunk header lines");
+    }
+
+    public InvalidChunkExtensionException(String message, Throwable cause) {
+        super(message, cause);
+    }
+
+    public InvalidChunkExtensionException(String message) {
+        super(message);
+    }
+
+    public InvalidChunkExtensionException(Throwable cause) {
+        super(cause);
+    }
+}
--- /dev/null
+++ b/codec-http/src/main/java/io/netty/handler/codec/http/InvalidChunkTerminationException.java
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2025 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.handler.codec.CorruptedFrameException;
+
+/**
+ * Thrown when HTTP chunks could not be parsed, typically due to incorrect use of CR LF delimiters.
+ * <p>
+ * <a href="https://datatracker.ietf.org/doc/html/rfc9112#name-chunked-transfer-coding">RFC 9112</a>
+ * specifies that chunk bodies must be terminated in a CR LF pair,
+ * and that the delimiter must follow the given chunk-size number of octets in chunk-data.
+ */
+public final class InvalidChunkTerminationException extends CorruptedFrameException {
+    private static final long serialVersionUID = 536224937231200736L;
+
+    public InvalidChunkTerminationException() {
+        super("Chunk data sections must be terminated by a CR LF octet pair");
+    }
+
+    public InvalidChunkTerminationException(String message, Throwable cause) {
+        super(message, cause);
+    }
+
+    public InvalidChunkTerminationException(String message) {
+        super(message);
+    }
+
+    public InvalidChunkTerminationException(Throwable cause) {
+        super(cause);
+    }
+}
--- /dev/null
+++ b/codec-http/src/main/java/io/netty/handler/codec/http/HttpChunkLineValidatingByteProcessor.java
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2026 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.util.ByteProcessor;
+
+import java.util.BitSet;
+
+/**
+ * Validates the chunk start line. That is, the chunk size and chunk extensions, until the CR LF pair.
+ * See <a href="https://www.rfc-editor.org/rfc/rfc9112#name-chunked-transfer-coding">RFC 9112 section 7.1</a>.
+ *
+ * <pre>{@code
+ *   chunked-body   = *chunk
+ *                    last-chunk
+ *                    trailer-section
+ *                    CRLF
+ *
+ *   chunk          = chunk-size [ chunk-ext ] CRLF
+ *                    chunk-data CRLF
+ *   chunk-size     = 1*HEXDIG
+ *   last-chunk     = 1*("0") [ chunk-ext ] CRLF
+ *
+ *   chunk-data     = 1*OCTET ; a sequence of chunk-size octets
+ *   chunk-ext      = *( BWS ";" BWS chunk-ext-name
+ *                       [ BWS "=" BWS chunk-ext-val ] )
+ *
+ *   chunk-ext-name = token
+ *   chunk-ext-val  = token / quoted-string
+ *   quoted-string  = DQUOTE *( qdtext / quoted-pair ) DQUOTE
+ *   qdtext         = HTAB / SP / %x21 / %x23-5B / %x5D-7E / obs-text
+ *   quoted-pair    = "\" ( HTAB / SP / VCHAR / obs-text )
+ *   obs-text       = %x80-FF
+ *   OWS            = *( SP / HTAB )
+ *                  ; optional whitespace
+ *   BWS            = OWS
+ *                  ; "bad" whitespace
+ *   VCHAR          =  %x21-7E
+ *                  ; visible (printing) characters
+ * }</pre>
+ */
+final class HttpChunkLineValidatingByteProcessor implements ByteProcessor {
+    private static final int SIZE = 0;
+    private static final int CHUNK_EXT_NAME = 1;
+    private static final int CHUNK_EXT_VAL_START = 2;
+    private static final int CHUNK_EXT_VAL_QUOTED = 3;
+    private static final int CHUNK_EXT_VAL_QUOTED_ESCAPE = 4;
+    private static final int CHUNK_EXT_VAL_QUOTED_END = 5;
+    private static final int CHUNK_EXT_VAL_TOKEN = 6;
+
+    static final class Match extends BitSet {
+        private static final long serialVersionUID = 49522994383099834L;
+        private final int then;
+
+        Match(int then) {
+            super(256);
+            this.then = then;
+        }
+
+        Match chars(String chars) {
+            return chars(chars, true);
+        }
+
+        Match chars(String chars, boolean value) {
+            for (int i = 0, len = chars.length(); i < len; i++) {
+                set(chars.charAt(i), value);
+            }
+            return this;
+        }
+
+        Match range(int from, int to) {
+            return range(from, to, true);
+        }
+
+        Match range(int from, int to, boolean value) {
+            for (int i = from; i <= to; i++) {
+                set(i, value);
+            }
+            return this;
+        }
+    }
+
+    private enum State {
+        Size(
+                new Match(SIZE).chars("0123456789abcdefABCDEF \t"),
+                new Match(CHUNK_EXT_NAME).chars(";")),
+        ChunkExtName(
+                new Match(CHUNK_EXT_NAME)
+                        .range(0x21, 0x7E)
+                        .chars(" \t")
+                        .chars("(),/:<=>?@[\\]{}", false),
+                new Match(CHUNK_EXT_VAL_START).chars("=")),
+        ChunkExtValStart(
+                new Match(CHUNK_EXT_VAL_START).chars(" \t"),
+                new Match(CHUNK_EXT_VAL_QUOTED).chars("\""),
+                new Match(CHUNK_EXT_VAL_TOKEN)
+                        .range(0x21, 0x7E)
+                        .chars("(),/:<=>?@[\\]{}", false)),
+        ChunkExtValQuoted(
+                new Match(CHUNK_EXT_VAL_QUOTED_ESCAPE).chars("\\"),
+                new Match(CHUNK_EXT_VAL_QUOTED_END).chars("\""),
+                new Match(CHUNK_EXT_VAL_QUOTED)
+                        .chars("\t !")
+                        .range(0x23, 0x5B)
+                        .range(0x5D, 0x7E)
+                        .range(0x80, 0xFF)),
+        ChunkExtValQuotedEscape(
+                new Match(CHUNK_EXT_VAL_QUOTED)
+                        .chars("\t ")
+                        .range(0x21, 0x7E)
+                        .range(0x80, 0xFF)),
+        ChunkExtValQuotedEnd(
+                new Match(CHUNK_EXT_VAL_QUOTED_END).chars("\t "),
+                new Match(CHUNK_EXT_NAME).chars(";")),
+        ChunkExtValToken(
+                new Match(CHUNK_EXT_VAL_TOKEN)
+                        .range(0x21, 0x7E, true)
+                        .chars("(),/:<=>?@[\\]{}", false),
+                new Match(CHUNK_EXT_NAME).chars(";")),
+        ;
+
+        private final Match[] matches;
+
+        State(Match... matches) {
+            this.matches = matches;
+        }
+
+        State match(byte value) {
+            for (Match match : matches) {
+                if (match.get(value)) {
+                    return STATES_BY_ORDINAL[match.then];
+                }
+            }
+            if (this == Size) {
+                throw new NumberFormatException("Invalid chunk size");
+            } else {
+                throw new InvalidChunkExtensionException("Invalid chunk extension");
+            }
+        }
+    }
+
+    private static final State[] STATES_BY_ORDINAL = State.values();
+
+    private State state = State.Size;
+
+    @Override
+    public boolean process(byte value) {
+        state = state.match(value);
+        return true;
+    }
+
+    public void finish() {
+        if (state != State.Size && state != State.ChunkExtName && state != State.ChunkExtValQuotedEnd) {
+            throw new InvalidChunkExtensionException("Invalid chunk extension");
+        }
+    }
+}
--- a/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java
+++ b/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java
@@ -457,6 +457,7 @@
             if (line == null) {
                 return;
             }
+            checkChunkExtensions(line);
             int chunkSize = getChunkSize(line.array(), line.arrayOffset() + line.readerIndex(), line.readableBytes());
             this.chunkSize = chunkSize;
             if (chunkSize == 0) {
@@ -491,16 +492,16 @@
             // fall-through
         }
         case READ_CHUNK_DELIMITER: {
-            final int wIdx = buffer.writerIndex();
-            int rIdx = buffer.readerIndex();
-            while (wIdx > rIdx) {
-                byte next = buffer.getByte(rIdx++);
-                if (next == HttpConstants.LF) {
+            if (buffer.readableBytes() >= 2) {
+                int rIdx = buffer.readerIndex();
+                if (buffer.getByte(rIdx) == HttpConstants.CR &&
+                        buffer.getByte(rIdx + 1) == HttpConstants.LF) {
+                    buffer.skipBytes(2);
                     currentState = State.READ_CHUNK_SIZE;
-                    break;
+                } else {
+                    out.add(invalidChunk(buffer, new InvalidChunkTerminationException()));
                 }
             }
-            buffer.readerIndex(rIdx);
             return;
         }
         case READ_CHUNK_FOOTER: try {
@@ -715,6 +716,16 @@
         HttpContent chunk = new DefaultLastHttpContent(Unpooled.EMPTY_BUFFER);
         chunk.setDecoderResult(DecoderResult.failure(cause));
         return chunk;
+    }
+
+    private static void checkChunkExtensions(ByteBuf line) {
+        int extensionsStart = line.bytesBefore((byte) ';');
+        if (extensionsStart == -1) {
+            return;
+        }
+        HttpChunkLineValidatingByteProcessor processor = new HttpChunkLineValidatingByteProcessor();
+        line.forEachByte(processor);
+        processor.finish();
     }
 
     private State readHeaders(ByteBuf buffer) {

--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java
@@ -116,7 +116,7 @@
     public static final int SMALLEST_MAX_CONCURRENT_STREAMS = 100;
     static final int DEFAULT_MAX_RESERVED_STREAMS = SMALLEST_MAX_CONCURRENT_STREAMS;
     static final int DEFAULT_MIN_ALLOCATION_CHUNK = 1024;
-
+    static final int DEFAULT_MAX_SMALL_CONTINUATION_FRAME = 16;
     /**
      * Calculate the threshold in bytes which should trigger a {@code GO_AWAY} if a set of headers exceeds this amount.
      * @param maxHeaderListSize
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java
@@ -17,6 +17,7 @@
 import io.netty.buffer.ByteBufAllocator;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.handler.codec.http2.Http2FrameReader.Configuration;
+import io.netty.util.internal.ObjectUtil;
 import io.netty.util.internal.PlatformDependent;
 import io.netty.util.internal.UnstableApi;

@@ -24,6 +25,7 @@
 import static io.netty.handler.codec.http2.Http2CodecUtil.DEFAULT_MAX_FRAME_SIZE;
 import static io.netty.handler.codec.http2.Http2CodecUtil.FRAME_HEADER_LENGTH;
 import static io.netty.handler.codec.http2.Http2CodecUtil.INT_FIELD_LENGTH;
+import static io.netty.handler.codec.http2.Http2CodecUtil.MAX_FRAME_SIZE_LOWER_BOUND;
 import static io.netty.handler.codec.http2.Http2CodecUtil.PING_FRAME_PAYLOAD_LENGTH;
 import static io.netty.handler.codec.http2.Http2CodecUtil.PRIORITY_ENTRY_LENGTH;
 import static io.netty.handler.codec.http2.Http2CodecUtil.SETTINGS_INITIAL_WINDOW_SIZE;
@@ -31,6 +33,7 @@
 import static io.netty.handler.codec.http2.Http2CodecUtil.headerListSizeExceeded;
 import static io.netty.handler.codec.http2.Http2CodecUtil.isMaxFrameSizeValid;
 import static io.netty.handler.codec.http2.Http2CodecUtil.readUnsignedInt;
+import static io.netty.handler.codec.http2.Http2Error.ENHANCE_YOUR_CALM;
 import static io.netty.handler.codec.http2.Http2Error.FLOW_CONTROL_ERROR;
 import static io.netty.handler.codec.http2.Http2Error.FRAME_SIZE_ERROR;
 import static io.netty.handler.codec.http2.Http2Error.PROTOCOL_ERROR;
@@ -53,6 +56,7 @@
  */
 @UnstableApi
 public class DefaultHttp2FrameReader implements Http2FrameReader, Http2FrameSizePolicy, Configuration {
+    private static final int FRAGMENT_THRESHOLD = MAX_FRAME_SIZE_LOWER_BOUND / 2;
     private final Http2HeadersDecoder headersDecoder;

     /**
@@ -69,7 +73,8 @@
     private Http2Flags flags;
     private int payloadLength;
     private HeadersContinuation headersContinuation;
-    private int maxFrameSize;
+    private int maxFrameSize = DEFAULT_MAX_FRAME_SIZE;
+    private final int maxSmallContinuationFrames;

     /**
      * Create a new instance.
@@ -90,8 +95,13 @@
     }

     public DefaultHttp2FrameReader(Http2HeadersDecoder headersDecoder) {
-        this.headersDecoder = headersDecoder;
-        maxFrameSize = DEFAULT_MAX_FRAME_SIZE;
+        this(headersDecoder, Http2CodecUtil.DEFAULT_MAX_SMALL_CONTINUATION_FRAME);
+    }
+
+    public DefaultHttp2FrameReader(Http2HeadersDecoder headersDecoder, int maxSmallContinuationFrames) {
+        this.headersDecoder = ObjectUtil.checkNotNull(headersDecoder, "headersDecoder");
+        this.maxSmallContinuationFrames = ObjectUtil.checkPositiveOrZero(
+                maxSmallContinuationFrames, "maxSmallContinuationFrames");
     }

     @Override
@@ -390,6 +400,12 @@
         if (streamId != headersContinuation.getStreamId()) {
             throw connectionError(PROTOCOL_ERROR, "Continuation stream ID does not match pending headers. "
                     + "Expected %d, but received %d.", headersContinuation.getStreamId(), streamId);
+        }
+
+        if (headersContinuation.numSmallFragments() >=  maxSmallContinuationFrames) {
+            throw connectionError(ENHANCE_YOUR_CALM,
+                    "Number of small consecutive continuations frames %d exceeds maximum: %d",
+                    headersContinuation.numSmallFragments(), maxSmallContinuationFrames);
         }
     }

@@ -651,6 +667,15 @@
         abstract void processFragment(boolean endOfHeaders, ByteBuf fragment, int len,
                 Http2FrameListener listener) throws Http2Exception;

+        /**
+         * Return the number of fragments that were used so far.
+         *
+         * @return the number of fragments
+         */
+        final int numSmallFragments() {
+            return builder.numSmallFragments();
+        }
+
         final HeadersBlockBuilder headersBlockBuilder() {
             return builder;
         }
@@ -670,6 +695,7 @@
      */
     protected class HeadersBlockBuilder {
         private ByteBuf headerBlock;
+        private int numSmallFragments;

         /**
          * The local header size maximum has been exceeded while accumulating bytes.
@@ -680,6 +706,15 @@
             headerListSizeExceeded(headersDecoder.configuration().maxHeaderListSizeGoAway());
         }

+        /**
+         * Return the number of fragments that was used so far.
+         *
+         * @return number of fragments.
+         */
+        int numSmallFragments() {
+            return numSmallFragments;
+        }
+
         /**
          * Adds a fragment to the block.
          *
@@ -691,6 +726,11 @@
          */
         final void addFragment(ByteBuf fragment, int len, ByteBufAllocator alloc,
                 boolean endOfHeaders) throws Http2Exception {
+            if (maxSmallContinuationFrames > 0 && !endOfHeaders && len < FRAGMENT_THRESHOLD) {
+                // Only count of the fragment is not the end of header and if its < 8kb.
+                numSmallFragments++;
+            }
+
             if (headerBlock == null) {
                 if (len > headersDecoder.configuration().maxHeaderListSizeGoAway()) {
                     headerSizeExceeded();
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2ConnectionHandlerBuilder.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2ConnectionHandlerBuilder.java
@@ -113,6 +113,7 @@
     private int maxConsecutiveEmptyFrames = 2;
     private Integer maxRstFramesPerWindow;
     private int secondsPerWindow = 30;
+    private int maxSmallContinuationFrames = Http2CodecUtil.DEFAULT_MAX_SMALL_CONTINUATION_FRAME;

     /**
      * Sets the {@link Http2Settings} to use for the initial connection settings exchange.
@@ -450,6 +451,30 @@
         return self();
     }

+    /**
+     * Returns the maximum number of small CONTINUATION frames per HEADERS block that are allowed
+     * before the connection is closed. Small is defined as 8 KiB, half the minimum allowed HTTP2 frame size.
+     * This setting is to protect against the remote peer flooding us with such frames.
+     *
+     * {@code 0} means no protection is in place.
+     */
+    protected int decoderEnforceMaxSmallContinuationFrames() {
+        return maxSmallContinuationFrames;
+    }
+
+    /**
+     * Returns the maximum number of small CONTINUATION frames per HEADERS block that are allowed
+     * before the connection is closed. Small is defined as 8 KiB, half the minimum allowed HTTP2 frame size.
+     * This setting is to protect against the remote peer flooding us with such frames.
+     * {@code 0} means no protection should be applied.
+     */
+    protected B decoderEnforceMaxSmallContinuationFrames(int maxSmallContinuationFrames) {
+        enforceNonCodecConstraints("maxSmallContinuationFrames");
+        this.maxSmallContinuationFrames = checkPositiveOrZero(
+                maxSmallContinuationFrames, "maxSmallContinuationFrames");
+        return self();
+    }
+
     /**
      * Determine if settings frame should automatically be acknowledged and applied.
      * @return this.
@@ -556,7 +581,7 @@
         Long maxHeaderListSize = initialSettings.maxHeaderListSize();
         Http2FrameReader reader = new DefaultHttp2FrameReader(new DefaultHttp2HeadersDecoder(isValidateHeaders(),
                 maxHeaderListSize == null ? DEFAULT_HEADER_LIST_SIZE : maxHeaderListSize,
-                /* initialHuffmanDecodeCapacity= */ -1));
+                /* initialHuffmanDecodeCapacity= */ -1), maxSmallContinuationFrames);
         Http2FrameWriter writer = encoderIgnoreMaxHeaderListSize == null ?
                 new DefaultHttp2FrameWriter(headerSensitivityDetector()) :
                 new DefaultHttp2FrameWriter(headerSensitivityDetector(), encoderIgnoreMaxHeaderListSize);
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodecBuilder.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodecBuilder.java
@@ -198,6 +198,17 @@
         return super.decoderEnforceMaxRstFramesPerWindow(maxRstFramesPerWindow, secondsPerWindow);
     }

+    @Override
+    public int decoderEnforceMaxSmallContinuationFrames() {
+        return super.decoderEnforceMaxSmallContinuationFrames();
+    }
+
+    @Override
+    public Http2FrameCodecBuilder decoderEnforceMaxSmallContinuationFrames(
+            int maxConsecutiveContinuationsFrames) {
+        return super.decoderEnforceMaxSmallContinuationFrames(maxConsecutiveContinuationsFrames);
+    }
+
     /**
      * Build a {@link Http2FrameCodec} object.
      */
@@ -211,7 +222,8 @@
             Long maxHeaderListSize = initialSettings().maxHeaderListSize();
             Http2FrameReader frameReader = new DefaultHttp2FrameReader(maxHeaderListSize == null ?
                     new DefaultHttp2HeadersDecoder(isValidateHeaders()) :
-                    new DefaultHttp2HeadersDecoder(isValidateHeaders(), maxHeaderListSize));
+                    new DefaultHttp2HeadersDecoder(isValidateHeaders(), maxHeaderListSize),
+                    decoderEnforceMaxSmallContinuationFrames());

             if (frameLogger() != null) {
                 frameWriter = new Http2OutboundFrameLogger(frameWriter, frameLogger());
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodecBuilder.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodecBuilder.java
@@ -214,6 +214,17 @@
         return super.decoderEnforceMaxRstFramesPerWindow(maxRstFramesPerWindow, secondsPerWindow);
     }

+    @Override
+    public int decoderEnforceMaxSmallContinuationFrames() {
+        return super.decoderEnforceMaxSmallContinuationFrames();
+    }
+
+    @Override
+    public Http2MultiplexCodecBuilder decoderEnforceMaxSmallContinuationFrames(
+            int maxConsecutiveContinuationsFrames) {
+        return super.decoderEnforceMaxSmallContinuationFrames(maxConsecutiveContinuationsFrames);
+    }
+
     @Override
     public Http2MultiplexCodec build() {
         Http2FrameWriter frameWriter = this.frameWriter;
@@ -225,7 +236,8 @@
             Long maxHeaderListSize = initialSettings().maxHeaderListSize();
             Http2FrameReader frameReader = new DefaultHttp2FrameReader(maxHeaderListSize == null ?
                     new DefaultHttp2HeadersDecoder(isValidateHeaders()) :
-                    new DefaultHttp2HeadersDecoder(isValidateHeaders(), maxHeaderListSize));
+                    new DefaultHttp2HeadersDecoder(isValidateHeaders(), maxHeaderListSize),
+                    decoderEnforceMaxSmallContinuationFrames());

             if (frameLogger() != null) {
                 frameWriter = new Http2OutboundFrameLogger(frameWriter, frameLogger());
--- a/codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2FrameReaderTest.java
+++ b/codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2FrameReaderTest.java
@@ -109,6 +109,59 @@
         }
     }

+    @Test
+    public void readHeaderFrameAndContinuationFrameExceedMax() throws Http2Exception {
+        frameReader = new DefaultHttp2FrameReader(new DefaultHttp2HeadersDecoder(true), 2);
+        final int streamId = 1;
+
+        final ByteBuf input = Unpooled.buffer();
+        try {
+            Http2Headers headers = new DefaultHttp2Headers()
+                    .authority("foo")
+                    .method("get")
+                    .path("/")
+                    .scheme("https");
+            writeHeaderFrame(input, streamId, headers,
+                    new Http2Flags().endOfHeaders(false).endOfStream(true));
+            writeContinuationFrame(input, streamId, new DefaultHttp2Headers().add("foo", "bar"),
+                    new Http2Flags().endOfHeaders(false));
+            writeContinuationFrame(input, streamId, new DefaultHttp2Headers().add("foo2", "bar2"),
+                    new Http2Flags().endOfHeaders(false));
+
+            Http2Exception ex = assertThrows(Http2Exception.class, new Executable() {
+                @Override
+                public void execute() throws Throwable {
+                    frameReader.readFrame(ctx, input, listener);
+                }
+            });
+            assertEquals(Http2Error.ENHANCE_YOUR_CALM, ex.error());
+        } finally {
+            input.release();
+        }
+    }
+
+    @Test
+    public void readHeaderFrameAndContinuationFrameDontExceedMax() throws Http2Exception {
+        frameReader = new DefaultHttp2FrameReader(new DefaultHttp2HeadersDecoder(true), 2);
+        final int streamId = 1;
+
+        final ByteBuf input = Unpooled.buffer();
+        try {
+            Http2Headers headers = new DefaultHttp2Headers()
+                    .authority("foo")
+                    .method("get")
+                    .path("/")
+                    .scheme("https");
+            writeHeaderFrame(input, streamId, headers,
+                    new Http2Flags().endOfHeaders(false).endOfStream(true));
+            writeContinuationFrame(input, streamId, new DefaultHttp2Headers().add("foo", "bar"),
+                    new Http2Flags().endOfHeaders(false));
+            frameReader.readFrame(ctx, input, listener);
+        } finally {
+            input.release();
+        }
+    }
+
     @Test
     public void readUnknownFrame() throws Http2Exception {
         ByteBuf input = Unpooled.buffer();

--- a/handler/src/main/java/io/netty/handler/ssl/SslUtils.java
+++ b/handler/src/main/java/io/netty/handler/ssl/SslUtils.java
@@ -314,8 +314,11 @@
      *                  the given {@link ByteBuf} is not encrypted at all.
      */
     static int getEncryptedPacketLength(ByteBuf buffer, int offset) {
+        int remaining = buffer.writerIndex() - offset;
+        if (remaining < SSL_RECORD_HEADER_LENGTH) {
+            return NOT_ENOUGH_DATA;
+        }
         int packetLength = 0;
-
         // SSLv3 or TLS - Check ContentType
         boolean tls;
         switch (buffer.getUnsignedByte(offset)) {
@@ -343,7 +346,7 @@
                     tls = false;
                 }
             } else if (version == DTLS_1_0 || version == DTLS_1_2 || version == DTLS_1_3) {
-                if (buffer.readableBytes() < offset + DTLS_RECORD_HEADER_LENGTH) {
+                if (remaining < DTLS_RECORD_HEADER_LENGTH) {
                     return NOT_ENOUGH_DATA;
                 }
                 // length is the last 2 bytes in the 13 byte header.
@@ -364,7 +367,8 @@
                 packetLength = headerLength == 2 ?
                         (shortBE(buffer, offset) & 0x7FFF) + 2 : (shortBE(buffer, offset) & 0x3FFF) + 3;
                 if (packetLength <= headerLength) {
-                    return NOT_ENOUGH_DATA;
+                    // If there's no data then consider this package as not encrypted.
+                    return NOT_ENCRYPTED;
                 }
             } else {
                 return NOT_ENCRYPTED;
@@ -417,7 +421,7 @@
         }

         // We need to copy 5 bytes into a temporary buffer so we can parse out the packet length easily.
-        ByteBuffer tmp = ByteBuffer.allocate(5);
+        ByteBuffer tmp = ByteBuffer.allocate(SSL_RECORD_HEADER_LENGTH);

         do {
             buffer = buffers[offset++].duplicate();
@@ -425,7 +429,7 @@
                 buffer.limit(buffer.position() + tmp.remaining());
             }
             tmp.put(buffer);
-        } while (tmp.hasRemaining());
+        } while (tmp.hasRemaining() && offset < buffers.length);

         // Done, flip the buffer so we can read from it.
         tmp.flip();
@@ -433,6 +437,10 @@
     }

     private static int getEncryptedPacketLength(ByteBuffer buffer) {
+        int remaining = buffer.remaining();
+        if (remaining < SSL_RECORD_HEADER_LENGTH) {
+            return NOT_ENOUGH_DATA;
+        }
         int packetLength = 0;
         int pos = buffer.position();
         // SSLv3 or TLS - Check ContentType
@@ -475,7 +483,8 @@
                 packetLength = headerLength == 2 ?
                         (shortBE(buffer, pos) & 0x7FFF) + 2 : (shortBE(buffer, pos) & 0x3FFF) + 3;
                 if (packetLength <= headerLength) {
-                    return NOT_ENOUGH_DATA;
+                    // If there's no data then consider this package as not encrypted.
+                    return NOT_ENCRYPTED;
                 }
             } else {
                 return NOT_ENCRYPTED;
--- a/handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java
+++ b/handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java
@@ -1200,7 +1200,9 @@
                 if (packetLength == SslUtils.NOT_ENCRYPTED) {
                     throw new NotSslRecordException("not an SSL/TLS record");
                 }

+                assert packetLength >= 0;
+
                 final int packetLengthDataOnly = packetLength - SSL_RECORD_HEADER_LENGTH;
                 if (packetLengthDataOnly > capacity) {
                     // Not enough space in the destination buffer so signal the caller that the buffer needs to be

diff --git a/transport-native-unix-common/pom.xml b/transport-native-unix-common/pom.xml
--- a/transport-native-unix-common/pom.xml
+++ b/transport-native-unix-common/pom.xml
@@ -45,6 +45,9 @@
     <nativeJarFile>${project.build.directory}/${project.build.finalName}-${jni.classifier}.jar</nativeJarFile>
     <!-- Always check JNI during test run so we catch bugs that could cause crashes -->
     <argLine.jni>-Xcheck:jni</argLine.jni>
+    <!-- Central-only coords: use properties so alias-updater does not rewrite bare io.netty in plugin config -->
+    <netty.jni.util.sources.groupId>io.netty</netty.jni.util.sources.groupId>
+    <netty.jni.util.sources.version>0.0.9.Final</netty.jni.util.sources.version>
   </properties>
 
   <build>
@@ -58,16 +61,19 @@
             <id>unpack</id>
             <phase>generate-sources</phase>
             <goals>
-              <goal>unpack-dependencies</goal>
+              <goal>unpack</goal>
             </goals>
             <configuration>
-              <includeGroupIds>io.netty</includeGroupIds>
-              <includeArtifactIds>netty-jni-util</includeArtifactIds>
-              <classifier>sources</classifier>
-              <outputDirectory>${jniUtilIncludeDir}</outputDirectory>
-              <includes>**.h,**.c</includes>
-              <overWriteReleases>false</overWriteReleases>
-              <overWriteSnapshots>true</overWriteSnapshots>
+              <artifactItems>
+                <artifactItem>
+                  <groupId>${netty.jni.util.sources.groupId}</groupId>
+                  <artifactId>netty-jni-util</artifactId>
+                  <version>${netty.jni.util.sources.version}</version>
+                  <classifier>sources</classifier>
+                  <outputDirectory>${jniUtilIncludeDir}</outputDirectory>
+                  <includes>**.h,**.c</includes>
+                </artifactItem>
+              </artifactItems>
             </configuration>
           </execution>
         </executions>
