--- a/Makefile
+++ b/Makefile
@@ -126,7 +126,9 @@
 
 .PHONY: evg-test-enterprise-auth
 evg-test-enterprise-auth:
-	go run -tags gssapi ./cmd/testentauth/main.go
+	# Sanitize to ensure no memory errors occur in C code used for GSSAPI
+	# authentication.
+	CGO_CFLAGS="-fsanitize=address" CGO_LDFLAGS="-fsanitize=address" go run -tags gssapi ./cmd/testentauth/main.go
 
 .PHONY: evg-test-oidc-auth
 evg-test-oidc-auth:
--- a/x/mongo/driver/auth/internal/gssapi/gss_wrapper.c
+++ b/x/mongo/driver/auth/internal/gssapi/gss_wrapper.c
@@ -72,8 +72,8 @@
             free(*desc);
         }
 
-        *desc = malloc(desc_buffer.length+1);
-        memcpy(*desc, desc_buffer.value, desc_buffer.length+1);
+        *desc = calloc(1, desc_buffer.length + 1);
+        memcpy(*desc, desc_buffer.value, desc_buffer.length);
 
         gss_release_buffer(&local_min_stat, &desc_buffer);
     }
@@ -144,8 +144,8 @@
         return GSSAPI_ERROR;
     }
 
-	*username = malloc(name_buffer.length+1);
-	memcpy(*username, name_buffer.value, name_buffer.length+1);
+    *username = calloc(1, name_buffer.length + 1);
+    memcpy(*username, name_buffer.value, name_buffer.length);
 
     gss_release_buffer(&ignored, &name_buffer);
     gss_release_name(&ignored, &name);
