diff --git a/dist/openpgp.mjs b/dist/openpgp.mjs
--- a/dist/openpgp.mjs
+++ b/dist/openpgp.mjs
@@ -30289,12 +30289,13 @@ class Message {
     if (literalDataList.length !== 1) {
       throw new Error('Can only verify message with one literal data packet.');
     }
-    if (isArrayStream(msg.packets.stream)) {
-      msg.packets.push(...await readToEnd(msg.packets.stream, _ => _ || []));
+    let packets = msg.packets;
+    if (isArrayStream(packets.stream)) {
+      packets = packets.concat(await readToEnd(packets.stream, _ => _ || []));
     }
-    const onePassSigList = msg.packets.filterByTag(enums.packet.onePassSignature).reverse();
-    const signatureList = msg.packets.filterByTag(enums.packet.signature);
-    if (onePassSigList.length && !signatureList.length && util.isStream(msg.packets.stream) && !isArrayStream(msg.packets.stream)) {
+    const onePassSigList = packets.filterByTag(enums.packet.onePassSignature).reverse();
+    const signatureList = packets.filterByTag(enums.packet.signature);
+    if (onePassSigList.length && !signatureList.length && util.isStream(packets.stream) && !isArrayStream(packets.stream)) {
       await Promise.all(onePassSigList.map(async onePassSig => {
         onePassSig.correspondingSig = new Promise((resolve, reject) => {
           onePassSig.correspondingSigResolve = resolve;
@@ -30304,7 +30305,7 @@ class Message {
         onePassSig.hashed = readToEnd(await onePassSig.hash(onePassSig.signatureType, literalDataList[0], undefined, false));
         onePassSig.hashed.catch(() => {});
       }));
-      msg.packets.stream = transformPair(msg.packets.stream, async (readable, writable) => {
+      packets.stream = transformPair(packets.stream, async (readable, writable) => {
         const reader = getReader(readable);
         const writer = getWriter(writable);
         try {
diff --git a/dist/openpgp.js b/dist/openpgp.js
--- a/dist/openpgp.js
+++ b/dist/openpgp.js
@@ -30292,12 +30292,13 @@ class Message {
       if (literalDataList.length !== 1) {
         throw new Error('Can only verify message with one literal data packet.');
       }
-      if (isArrayStream(msg.packets.stream)) {
-        msg.packets.push(...await readToEnd(msg.packets.stream, _ => _ || []));
+      let packets = msg.packets;
+      if (isArrayStream(packets.stream)) {
+        packets = packets.concat(await readToEnd(packets.stream, _ => _ || []));
       }
-      const onePassSigList = msg.packets.filterByTag(enums.packet.onePassSignature).reverse();
-      const signatureList = msg.packets.filterByTag(enums.packet.signature);
-      if (onePassSigList.length && !signatureList.length && util.isStream(msg.packets.stream) && !isArrayStream(msg.packets.stream)) {
+      const onePassSigList = packets.filterByTag(enums.packet.onePassSignature).reverse();
+      const signatureList = packets.filterByTag(enums.packet.signature);
+      if (onePassSigList.length && !signatureList.length && util.isStream(packets.stream) && !isArrayStream(packets.stream)) {
         await Promise.all(onePassSigList.map(async onePassSig => {
           onePassSig.correspondingSig = new Promise((resolve, reject) => {
             onePassSig.correspondingSigResolve = resolve;
@@ -30307,7 +30308,7 @@ class Message {
           onePassSig.hashed = readToEnd(await onePassSig.hash(onePassSig.signatureType, literalDataList[0], undefined, false));
           onePassSig.hashed.catch(() => {});
         }));
-        msg.packets.stream = transformPair(msg.packets.stream, async (readable, writable) => {
+        packets.stream = transformPair(packets.stream, async (readable, writable) => {
           const reader = getReader(readable);
           const writer = getWriter(writable);
           try {
diff --git a/dist/node/openpgp.mjs b/dist/node/openpgp.mjs
--- a/dist/node/openpgp.mjs
+++ b/dist/node/openpgp.mjs
@@ -30301,12 +30301,13 @@ class Message {
     if (literalDataList.length !== 1) {
       throw new Error('Can only verify message with one literal data packet.');
     }
-    if (isArrayStream(msg.packets.stream)) {
-      msg.packets.push(...await readToEnd(msg.packets.stream, _ => _ || []));
+    let packets = msg.packets;
+    if (isArrayStream(packets.stream)) {
+      packets = packets.concat(await readToEnd(packets.stream, _ => _ || []));
     }
-    const onePassSigList = msg.packets.filterByTag(enums.packet.onePassSignature).reverse();
-    const signatureList = msg.packets.filterByTag(enums.packet.signature);
-    if (onePassSigList.length && !signatureList.length && util.isStream(msg.packets.stream) && !isArrayStream(msg.packets.stream)) {
+    const onePassSigList = packets.filterByTag(enums.packet.onePassSignature).reverse();
+    const signatureList = packets.filterByTag(enums.packet.signature);
+    if (onePassSigList.length && !signatureList.length && util.isStream(packets.stream) && !isArrayStream(packets.stream)) {
       await Promise.all(onePassSigList.map(async onePassSig => {
         onePassSig.correspondingSig = new Promise((resolve, reject) => {
           onePassSig.correspondingSigResolve = resolve;
@@ -30316,7 +30317,7 @@ class Message {
         onePassSig.hashed = readToEnd(await onePassSig.hash(onePassSig.signatureType, literalDataList[0], undefined, false));
         onePassSig.hashed.catch(() => {});
       }));
-      msg.packets.stream = transformPair(msg.packets.stream, async (readable, writable) => {
+      packets.stream = transformPair(packets.stream, async (readable, writable) => {
         const reader = getReader(readable);
         const writer = getWriter(writable);
         try {
diff --git a/dist/node/openpgp.js b/dist/node/openpgp.js
--- a/dist/node/openpgp.js
+++ b/dist/node/openpgp.js
@@ -30315,12 +30315,13 @@ class Message {
     if (literalDataList.length !== 1) {
       throw new Error('Can only verify message with one literal data packet.');
     }
-    if (isArrayStream(msg.packets.stream)) {
-      msg.packets.push(...await readToEnd(msg.packets.stream, _ => _ || []));
+    let packets = msg.packets;
+    if (isArrayStream(packets.stream)) {
+      packets = packets.concat(await readToEnd(packets.stream, _ => _ || []));
     }
-    const onePassSigList = msg.packets.filterByTag(enums.packet.onePassSignature).reverse();
-    const signatureList = msg.packets.filterByTag(enums.packet.signature);
-    if (onePassSigList.length && !signatureList.length && util.isStream(msg.packets.stream) && !isArrayStream(msg.packets.stream)) {
+    const onePassSigList = packets.filterByTag(enums.packet.onePassSignature).reverse();
+    const signatureList = packets.filterByTag(enums.packet.signature);
+    if (onePassSigList.length && !signatureList.length && util.isStream(packets.stream) && !isArrayStream(packets.stream)) {
       await Promise.all(onePassSigList.map(async onePassSig => {
         onePassSig.correspondingSig = new Promise((resolve, reject) => {
           onePassSig.correspondingSigResolve = resolve;
@@ -30330,7 +30331,7 @@ class Message {
         onePassSig.hashed = readToEnd(await onePassSig.hash(onePassSig.signatureType, literalDataList[0], undefined, false));
         onePassSig.hashed.catch(() => {});
       }));
-      msg.packets.stream = transformPair(msg.packets.stream, async (readable, writable) => {
+      packets.stream = transformPair(packets.stream, async (readable, writable) => {
         const reader = getReader(readable);
         const writer = getWriter(writable);
         try {
