--- /dev/null
+++ b/package.json
@@ -0,0 +1,5 @@
+{
+  "name": "esbuild-source",
+  "private": true,
+  "version": "0.16.17"
+}
--- a/pkg/api/serve_other.go
+++ b/pkg/api/serve_other.go
@@ -37,6 +37,7 @@
 	onRequest        func(ServeOnRequestArgs)
 	rebuild          func() BuildResult
 	currentBuild     *runningBuild
+	hosts            []string
 	fs               fs.FS
 	serveError       error
 	absOutputDir     string
@@ -126,9 +127,32 @@
 func (h *apiHandler) ServeHTTP(res http.ResponseWriter, req *http.Request) {
 	start := time.Now()
 
+	// Check the "Host" header to prevent DNS rebinding attacks
+	reqHost := req.Host
+	if strings.ContainsRune(reqHost, ':') {
+		if host, _, err := net.SplitHostPort(reqHost); err == nil {
+			reqHost = host
+		}
+	}
+	if reqHost != "localhost" {
+		ok := false
+		for _, allowed := range h.hosts {
+			if reqHost == allowed {
+				ok = true
+				break
+			}
+		}
+		if !ok {
+			go h.notifyRequest(time.Since(start), req, http.StatusForbidden)
+			res.WriteHeader(http.StatusForbidden)
+			res.Write([]byte(fmt.Sprintf("403 - Forbidden: The host %q is not allowed", reqHost)))
+			return
+		}
+	}
+
+
 	// Handle get requests
 	if req.Method == "GET" && strings.HasPrefix(req.URL.Path, "/") {
-		res.Header().Set("Access-Control-Allow-Origin", "*")
 		queryPath := path.Clean(req.URL.Path)[1:]
 		result := h.build()
 
@@ -580,19 +604,36 @@
 
 	// Extract the real port in case we passed a port of "0"
 	var result ServeResult
+	var boundHost string
 	if host, text, err := net.SplitHostPort(addr); err == nil {
 		if port, err := strconv.ParseInt(text, 10, 32); err == nil {
 			result.Port = uint16(port)
-			result.Host = host
+			boundHost = host
 		}
 	}
 
+	// Build up a list of all hosts we use
+	if ip := net.ParseIP(boundHost); ip != nil && ip.IsUnspecified() {
+		// If this is "0.0.0.0" or "::", list all relevant IP addresses
+		if addrs, err := net.InterfaceAddrs(); err == nil {
+			for _, addr := range addrs {
+				if addr, ok := addr.(*net.IPNet); ok && (addr.IP.To4() != nil) == (ip.To4() != nil) && !addr.IP.IsLinkLocalUnicast() {
+					result.Hosts = append(result.Hosts, addr.IP.String())
+				}
+			}
+		}
+	} else {
+		result.Hosts = append(result.Hosts, boundHost)
+	}
+
+
 	var stoppingMutex sync.Mutex
 	isStopping := false
 
 	// The first build will just build normally
 	var handler *apiHandler
 	handler = &apiHandler{
+		hosts:            append([]string{}, result.Hosts...),
 		onRequest:        serveOptions.OnRequest,
 		outdirPathPrefix: outdirPathPrefix,
 		servedir:         serveOptions.Servedir,
--- a/pkg/api/api.go
+++ b/pkg/api/api.go
@@ -465,8 +465,8 @@
 
 // Documentation: https://esbuild.github.io/api/#serve-return-values
 type ServeResult struct {
-	Port uint16
-	Host string
+	Port  uint16
+	Hosts []string
 	Wait func() error
 	Stop func()
 }
--- a/cmd/esbuild/service.go
+++ b/cmd/esbuild/service.go
@@ -567,10 +567,15 @@
 	if err != nil {
 		return outgoingPacket{bytes: encodeErrorPacket(id, err)}
 	}
+	hosts := make([]interface{}, len(result.Hosts))
+	for i, host := range result.Hosts {
+		hosts[i] = host
+	}
 	response := map[string]interface{}{
-		"port": int(result.Port),
-		"host": result.Host,
+		"port":  int(result.Port),
+		"hosts": hosts,
 	}
+
 
 	activeBuild.refCount++ // Make sure the serve doesn't finish until "Wait" finishes
 	activeBuild.serveStop = result.Stop
--- a/lib/shared/stdio_protocol.ts
+++ b/lib/shared/stdio_protocol.ts
@@ -30,7 +30,7 @@
 
 export interface ServeResponse {
   port: number;
-  host: string;
+  hosts: string[];
 }
 
 export interface ServeStopRequest {
--- a/lib/shared/types.ts
+++ b/lib/shared/types.ts
@@ -248,7 +248,7 @@
 /** Documentation: https://esbuild.github.io/api/#serve-return-values */
 export interface ServeResult {
   port: number;
-  host: string;
+  hosts: string[];
   wait: Promise<void>;
   stop: () => void;
 }
--- a/lib/shared/common.ts
+++ b/lib/shared/common.ts
@@ -1070,7 +1070,7 @@ function buildOrServeImpl(
         refs.ref()
         let result: types.ServeResult = {
           port: serveResponse.port,
-          host: serveResponse.host,
+          hosts: serveResponse.hosts,
           wait: serve.wait,
           stop() {
             if (isStopped) return
--- a/pkg/cli/cli_impl.go
+++ b/pkg/cli/cli_impl.go
@@ -1376,26 +1376,10 @@ func serveImpl(osArgs []string) *cli_helpers.ErrorWithNote {
 
 	// Show what actually got bound if the port was 0
 	logger.PrintText(os.Stderr, logger.LevelInfo, filteredArgs, func(colors logger.Colors) string {
-		var hosts []string
+		hosts := result.Hosts
 		sb := strings.Builder{}
 		sb.WriteString(colors.Reset)
 
-		// If this is "0.0.0.0" or "::", list all relevant IP addresses
-		if ip := net.ParseIP(result.Host); ip != nil && ip.IsUnspecified() {
-			if addrs, err := net.InterfaceAddrs(); err == nil {
-				for _, addr := range addrs {
-					if addr, ok := addr.(*net.IPNet); ok && (addr.IP.To4() != nil) == (ip.To4() != nil) && !addr.IP.IsLinkLocalUnicast() {
-						hosts = append(hosts, addr.IP.String())
-					}
-				}
-			}
-		}
-
-		// Otherwise, just list the one IP address
-		if len(hosts) == 0 {
-			hosts = append(hosts, result.Host)
-		}
-
 		// Determine the host kinds
 		kinds := make([]string, len(hosts))
 		maxLen := 0
