From f9914ad9effc865ccc27d816255625890b31ca51 Mon Sep 17 00:00:00 2001
Subject: [PATCH] CVE-2026-73088: backport of f9914ad9effc (Object.create(null) + hasOwnProperty)

--- a/index.js
+++ b/index.js
@@ -185,8 +185,9 @@
             stats = stats.dataByBrowser;
         }
 
-        context.customUsage = { };
+        context.customUsage = Object.create(null);
         for ( var browser in stats ) {
+            if (!Object.prototype.hasOwnProperty.call(stats, browser)) continue;
             fillUsage(context.customUsage, browser, stats[browser]);
         }
     }
