--- a/lib/index.js
+++ b/lib/index.js
@@ -16,7 +16,7 @@
 */

 //                             1: type/subtype    2: params
-internals.contentTypeRegex = /^([^\/\s]+\/[^\s;]+)(.*)?$/;
+internals.contentTypeRegex = /^([^\/\s]+\/[^\s;]+)([ \t;][^\r\n]*)?$/;

 //                                             1: "b"   2: b
 internals.charsetParamRegex = /;\s*charset=(?:"([^"]+)"|([^;"\s]+))/i;
@@ -82,10 +82,10 @@
 */


-internals.contentDispositionRegex = /^\s*form-data\s*(?:;\s*(.+))?$/i;
+internals.contentDispositionRegex = /^\s*form-data\s*(?:;\s*(\S.*))?$/i;

 //                                        1: name     2: *            3: ext-value                      4: quoted  5: token
-internals.contentDispositionParamRegex = /([^\=\*\s]+)(\*)?\s*\=\s*(?:([^;'"\s]+\'[\w-]*\'[^;\s]+)|(?:\"([^"]*)\")|([^;\s]*))(?:\s*(?:;\s*)|$)/g;
+internals.contentDispositionParamRegex = /([^\=\*\s]+)(\*)?\s*\=\s*(?:([^;'"\s]+\'[\w-]*\'[^;\s]+)|(?:\"([^"]*)\")|([^;\s]*))/g;

 exports.disposition = function (header) {

