--- a/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/web/servlet/AbstractWebMvcEndpointHandlerMapping.java
+++ b/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/web/servlet/AbstractWebMvcEndpointHandlerMapping.java
@@ -103,6 +103,9 @@
 	private final Method handleMethod = ReflectionUtils.findMethod(OperationHandler.class, "handle",
 			HttpServletRequest.class, Map.class);
 
+	private final Method catchAllMethod = ReflectionUtils.findMethod(CatchAllHandler.class, "handle",
+			HttpServletResponse.class);
+
 	private RequestMappingInfo.BuilderConfiguration builderConfig = new RequestMappingInfo.BuilderConfiguration();
 
 	/**
@@ -193,6 +196,17 @@
 		return servletWebOperation;
 	}
 
+	/**
+	 * Register a "catch all" handler for the rest of actuator namespace, ensuring that
+	 * all requests are handled by this handler mapping.
+	 * @param responseStatus the response status to use for handled requests
+	 */
+	protected void registerCatchAllMapping(HttpStatus responseStatus) {
+		String subPath = this.endpointMapping.createSubPath("/**");
+		registerMapping(RequestMappingInfo.paths(subPath).options(this.builderConfig).build(),
+				new CatchAllHandler(responseStatus), this.catchAllMethod);
+	}
+
 	private RequestMappingInfo createRequestMappingInfo(WebOperationRequestPredicate predicate, String path) {
 		return RequestMappingInfo.paths(this.endpointMapping.createSubPath(path))
 			.options(this.builderConfig)
@@ -436,6 +449,23 @@
 
 	}
 
+	/**
+	 * Catch-all handler that always replies with a fixed HTTP status.
+	 */
+	private static final class CatchAllHandler {
+
+		private final HttpStatus responseStatus;
+
+		CatchAllHandler(HttpStatus responseStatus) {
+			this.responseStatus = responseStatus;
+		}
+
+		void handle(HttpServletResponse response) {
+			response.setStatus(this.responseStatus.value());
+		}
+
+	}
+
 	/**
 	 * {@link HandlerMethod} subclass for endpoint information logging.
 	 */
--- a/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/web/reactive/AbstractWebFluxEndpointHandlerMapping.java
+++ b/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/web/reactive/AbstractWebFluxEndpointHandlerMapping.java
@@ -55,6 +55,7 @@
 import org.springframework.http.HttpStatus;
 import org.springframework.http.MediaType;
 import org.springframework.http.ResponseEntity;
+import org.springframework.http.server.reactive.ServerHttpResponse;
 import org.springframework.security.authorization.AuthorityAuthorizationManager;
 import org.springframework.security.core.Authentication;
 import org.springframework.security.core.context.ReactiveSecurityContextHolder;
@@ -100,6 +101,9 @@
 	private final Method handleReadMethod = ReflectionUtils.findMethod(ReadOperationHandler.class, "handle",
 			ServerWebExchange.class);
 
+	private final Method handleCatchAllMethod = ReflectionUtils.findMethod(CatchAllHandler.class, "handle",
+			ServerWebExchange.class);
+
 	private final boolean shouldRegisterLinksMapping;
 
 	/**
@@ -171,6 +175,17 @@
 		return reactiveWebOperation;
 	}
 
+	/**
+	 * Register a "catch all" handler for the rest of actuator namespace, ensuring that
+	 * all requests are handled by this handler mapping.
+	 * @param responseStatus the response status to use for handled requests
+	 */
+	protected void registerCatchAllMapping(HttpStatus responseStatus) {
+		String subPath = this.endpointMapping.createSubPath("/**");
+		registerMapping(RequestMappingInfo.paths(subPath).build(), new CatchAllHandler(responseStatus),
+				this.handleCatchAllMethod);
+	}
+
 	private RequestMappingInfo createRequestMappingInfo(WebOperation operation) {
 		WebOperationRequestPredicate predicate = operation.getRequestPredicate();
 		String path = this.endpointMapping.createSubPath(predicate.getPath());
@@ -449,6 +463,30 @@
 
 	}
 
+	/**
+	 * Catch-all handler that always replies with a fixed HTTP status.
+	 */
+	private static final class CatchAllHandler {
+
+		private final HttpStatus responseStatus;
+
+		CatchAllHandler(HttpStatus responseStatus) {
+			this.responseStatus = responseStatus;
+		}
+
+		Mono<Void> handle(ServerWebExchange exchange) {
+			ServerHttpResponse response = exchange.getResponse();
+			response.setStatusCode(this.responseStatus);
+			return response.setComplete();
+		}
+
+		@Override
+		public String toString() {
+			return "Actuator catch-all endpoint";
+		}
+
+	}
+
 	private static class WebFluxEndpointHandlerMethod extends HandlerMethod {
 
 		WebFluxEndpointHandlerMethod(Object bean, Method method) {
--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/servlet/CloudFoundryWebEndpointServletHandlerMapping.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/servlet/CloudFoundryWebEndpointServletHandlerMapping.java
@@ -80,6 +80,12 @@
 		this.allEndpoints = allEndpoints;
 	}
 
+	@Override
+	protected void initHandlerMethods() {
+		super.initHandlerMethods();
+		registerCatchAllMapping(HttpStatus.FORBIDDEN);
+	}
+
 	@Override
 	protected ServletWebOperation wrapServletWebOperation(ExposableWebEndpoint endpoint, WebOperation operation,
 			ServletWebOperation servletWebOperation) {
--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/reactive/CloudFoundryWebFluxEndpointHandlerMapping.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/reactive/CloudFoundryWebFluxEndpointHandlerMapping.java
@@ -76,6 +76,12 @@
 		this.securityInterceptor = securityInterceptor;
 	}
 
+	@Override
+	protected void initHandlerMethods() {
+		super.initHandlerMethods();
+		registerCatchAllMapping(HttpStatus.FORBIDDEN);
+	}
+
 	@Override
 	protected ReactiveWebOperation wrapReactiveWebOperation(ExposableWebEndpoint endpoint, WebOperation operation,
 			ReactiveWebOperation reactiveWebOperation) {
--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/servlet/CloudFoundryActuatorAutoConfiguration.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/servlet/CloudFoundryActuatorAutoConfiguration.java
@@ -34,7 +34,6 @@
 import org.springframework.boot.actuate.endpoint.web.EndpointMapping;
 import org.springframework.boot.actuate.endpoint.web.EndpointMediaTypes;
 import org.springframework.boot.actuate.endpoint.web.ExposableWebEndpoint;
-import org.springframework.boot.actuate.endpoint.web.PathMappedEndpoints;
 import org.springframework.boot.actuate.endpoint.web.annotation.ControllerEndpointsSupplier;
 import org.springframework.boot.actuate.endpoint.web.annotation.ServletEndpointsSupplier;
 import org.springframework.boot.actuate.health.HealthEndpoint;
@@ -64,8 +63,6 @@
 import org.springframework.security.config.annotation.web.builders.WebSecurity;
 import org.springframework.security.config.annotation.web.configuration.WebSecurityCustomizer;
 import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
-import org.springframework.security.web.util.matcher.OrRequestMatcher;
-import org.springframework.security.web.util.matcher.RequestMatcher;
 import org.springframework.web.cors.CorsConfiguration;
 import org.springframework.web.servlet.DispatcherServlet;
 
@@ -165,29 +163,17 @@
 
 		@Bean
-		IgnoredCloudFoundryPathsWebSecurityCustomizer ignoreCloudFoundryPathsWebSecurityCustomizer(
-				CloudFoundryWebEndpointServletHandlerMapping handlerMapping) {
-			return new IgnoredCloudFoundryPathsWebSecurityCustomizer(handlerMapping);
+		IgnoredCloudFoundryPathsWebSecurityCustomizer ignoreCloudFoundryPathsWebSecurityCustomizer() {
+			return new IgnoredCloudFoundryPathsWebSecurityCustomizer();
 		}
 
 	}
 
 	@Order(SecurityProperties.IGNORED_ORDER)
 	static class IgnoredCloudFoundryPathsWebSecurityCustomizer implements WebSecurityCustomizer {
 
-		private final PathMappedEndpoints pathMappedEndpoints;
-
-		IgnoredCloudFoundryPathsWebSecurityCustomizer(CloudFoundryWebEndpointServletHandlerMapping handlerMapping) {
-			this.pathMappedEndpoints = new PathMappedEndpoints(BASE_PATH, handlerMapping::getAllEndpoints);
-		}
-
 		@Override
 		public void customize(WebSecurity web) {
-			List<RequestMatcher> requestMatchers = new ArrayList<>();
-			this.pathMappedEndpoints.getAllPaths()
-				.forEach((path) -> requestMatchers.add(new AntPathRequestMatcher(path + "/**")));
-			requestMatchers.add(new AntPathRequestMatcher(BASE_PATH));
-			requestMatchers.add(new AntPathRequestMatcher(BASE_PATH + "/"));
-			web.ignoring().requestMatchers(new OrRequestMatcher(requestMatchers));
+			web.ignoring().requestMatchers(new AntPathRequestMatcher(BASE_PATH + "/**"));
 		}
 
 	}
--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/reactive/ReactiveCloudFoundryActuatorAutoConfiguration.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/reactive/ReactiveCloudFoundryActuatorAutoConfiguration.java
@@ -21,7 +21,6 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
-import java.util.function.Supplier;
 
 import org.springframework.beans.BeansException;
 import org.springframework.beans.factory.ObjectProvider;
@@ -36,7 +35,6 @@
 import org.springframework.boot.actuate.endpoint.web.EndpointMapping;
 import org.springframework.boot.actuate.endpoint.web.EndpointMediaTypes;
 import org.springframework.boot.actuate.endpoint.web.ExposableWebEndpoint;
-import org.springframework.boot.actuate.endpoint.web.PathMappedEndpoints;
 import org.springframework.boot.actuate.endpoint.web.annotation.ControllerEndpointsSupplier;
 import org.springframework.boot.actuate.health.HealthEndpoint;
 import org.springframework.boot.actuate.health.ReactiveHealthEndpointWebExtension;
@@ -64,7 +62,6 @@
 import org.springframework.security.web.server.WebFilterChainProxy;
 import org.springframework.security.web.server.util.matcher.ServerWebExchangeMatcher;
 import org.springframework.security.web.server.util.matcher.ServerWebExchangeMatchers;
-import org.springframework.util.function.SingletonSupplier;
 import org.springframework.web.cors.CorsConfiguration;
 import org.springframework.web.reactive.function.client.WebClient;
 import org.springframework.web.server.WebFilter;
@@ -156,20 +153,15 @@
 	static class IgnoredPathsSecurityConfiguration {
 
 		@Bean
-		static WebFilterChainPostProcessor webFilterChainPostProcessor(
-				ObjectProvider<CloudFoundryWebFluxEndpointHandlerMapping> handlerMapping) {
-			return new WebFilterChainPostProcessor(handlerMapping);
+		static WebFilterChainPostProcessor webFilterChainPostProcessor() {
+			return new WebFilterChainPostProcessor();
 		}
 
 	}
 
 	static class WebFilterChainPostProcessor implements BeanPostProcessor {
 
-		private final Supplier<PathMappedEndpoints> pathMappedEndpoints;
-
-		WebFilterChainPostProcessor(ObjectProvider<CloudFoundryWebFluxEndpointHandlerMapping> handlerMapping) {
-			this.pathMappedEndpoints = SingletonSupplier
-				.of(() -> new PathMappedEndpoints(BASE_PATH, () -> handlerMapping.getObject().getAllEndpoints()));
+		WebFilterChainPostProcessor() {
 		}
 
 		@Override
@@ -183,8 +179,7 @@
 
 		private WebFilterChainProxy postProcess(WebFilterChainProxy existing) {
-			List<String> paths = getPaths(this.pathMappedEndpoints.get());
 			ServerWebExchangeMatcher cloudFoundryRequestMatcher = ServerWebExchangeMatchers
-				.pathMatchers(paths.toArray(new String[] {}));
+				.pathMatchers(BASE_PATH + "/**");
 			WebFilter noOpFilter = (exchange, chain) -> chain.filter(exchange);
 			MatcherSecurityWebFilterChain ignoredRequestFilterChain = new MatcherSecurityWebFilterChain(
 					cloudFoundryRequestMatcher, Collections.singletonList(noOpFilter));
@@ -193,14 +188,6 @@
 			return new WebFilterChainProxy(ignoredRequestFilterChain, allRequestsFilterChain);
 		}
 
-		private static List<String> getPaths(PathMappedEndpoints pathMappedEndpoints) {
-			List<String> paths = new ArrayList<>();
-			pathMappedEndpoints.getAllPaths().forEach((path) -> paths.add(path + "/**"));
-			paths.add(BASE_PATH);
-			paths.add(BASE_PATH + "/");
-			return paths;
-		}
-
 	}
 
 }
--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/servlet/CloudFoundryMvcWebEndpointIntegrationTests.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/servlet/CloudFoundryMvcWebEndpointIntegrationTests.java
@@ -199,6 +199,17 @@
 					.doesNotExist());
 	}
 
+	@Test
+	void unknownEndpointsAreForbidden() {
+		load(TestEndpointConfiguration.class,
+				(client) -> client.get()
+					.uri("/cfApplication/unknown")
+					.accept(MediaType.APPLICATION_JSON)
+					.exchange()
+					.expectStatus()
+					.isForbidden());
+	}
+
 	private void load(Class<?> configuration, Consumer<WebTestClient> clientConsumer) {
 		BiConsumer<ApplicationContext, WebTestClient> consumer = (context, client) -> clientConsumer.accept(client);
 		new WebApplicationContextRunner(AnnotationConfigServletWebServerApplicationContext::new)
--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/reactive/CloudFoundryWebFluxEndpointIntegrationTests.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/reactive/CloudFoundryWebFluxEndpointIntegrationTests.java
@@ -210,6 +210,12 @@
 			.doesNotExist()));
 	}
 
+	@Test
+	void unknownEndpointsAreForbidden() {
+		this.contextRunner.run(withWebTestClient(
+				(client) -> client.get().uri("/cfApplication/unknown").exchange().expectStatus().isForbidden()));
+	}
+
 	private ContextConsumer<AssertableReactiveWebApplicationContext> withWebTestClient(
 			Consumer<WebTestClient> clientConsumer) {
 		return (context) -> {
--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/servlet/CloudFoundryActuatorAutoConfigurationTests.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/servlet/CloudFoundryActuatorAutoConfigurationTests.java
@@ -183,7 +183,6 @@
 				testCloudFoundrySecurity(request, BASE_PATH + "/test", chain);
 				testCloudFoundrySecurity(request, BASE_PATH + "/test/a", chain);
-				request.setServletPath(BASE_PATH + "/other-path");
-				assertThat(chain.matches(request)).isFalse();
+				testCloudFoundrySecurity(request, BASE_PATH + "/other-path", chain);
 				request.setServletPath("/some-other-path");
 				assertThat(chain.matches(request)).isFalse();
 			});
--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/reactive/ReactiveCloudFoundryActuatorAutoConfigurationTests.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/reactive/ReactiveCloudFoundryActuatorAutoConfigurationTests.java
@@ -196,6 +196,6 @@
 				assertThat(cfRequestMatches).isTrue();
 				assertThat(cfRequestWithAdditionalPathMatches).isTrue();
-				assertThat(otherCfRequestMatches).isFalse();
+				assertThat(otherCfRequestMatches).isTrue();
 				assertThat(otherRequestMatches).isFalse();
 				otherRequestMatches = filters.get(1)
 					.matches(MockServerWebExchange.from(MockServerHttpRequest.get("/some-other-path").build()))

--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/security/reactive/EndpointRequest.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/security/reactive/EndpointRequest.java
@@ -219,11 +219,18 @@ public final class EndpointRequest {
 			if (this.includeLinks && StringUtils.hasText(pathMappedEndpoints.getBasePath())) {
 				delegateMatchers.add(new LinksServerWebExchangeMatcher());
 			}
+			if (delegateMatchers.isEmpty()) {
+				return EMPTY_MATCHER;
+			}
 			return new OrServerWebExchangeMatcher(delegateMatchers);
 		}
 
 		private Stream<String> streamPaths(List<Object> source, PathMappedEndpoints pathMappedEndpoints) {
-			return source.stream().filter(Objects::nonNull).map(this::getEndpointId).map(pathMappedEndpoints::getPath);
+			return source.stream()
+				.filter(Objects::nonNull)
+				.map(this::getEndpointId)
+				.map(pathMappedEndpoints::getPath)
+				.filter(Objects::nonNull);
 		}
 
 		@Override
@@ -236,6 +243,7 @@ public final class EndpointRequest {
 		}
 
 		private PathPatternParserServerWebExchangeMatcher getDelegateMatcher(String path) {
+			Assert.notNull(path, "'path' must not be null");
 			return new PathPatternParserServerWebExchangeMatcher(path + "/**");
 		}
 
--- a/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/security/servlet/EndpointRequest.java
+++ b/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/security/servlet/EndpointRequest.java
@@ -241,11 +241,18 @@ public final class EndpointRequest {
 			if (this.includeLinks && StringUtils.hasText(basePath)) {
 				delegateMatchers.addAll(getLinksMatchers(requestMatcherFactory, matcherProvider, basePath));
 			}
+			if (delegateMatchers.isEmpty()) {
+				return EMPTY_MATCHER;
+			}
 			return new OrRequestMatcher(delegateMatchers);
 		}
 
 		private Stream<String> streamPaths(List<Object> source, PathMappedEndpoints pathMappedEndpoints) {
-			return source.stream().filter(Objects::nonNull).map(this::getEndpointId).map(pathMappedEndpoints::getPath);
+			return source.stream()
+				.filter(Objects::nonNull)
+				.map(this::getEndpointId)
+				.map(pathMappedEndpoints::getPath)
+				.filter(Objects::nonNull);
 		}
 
 		private List<RequestMatcher> getDelegateMatchers(RequestMatcherFactory requestMatcherFactory,
@@ -316,6 +323,7 @@ public final class EndpointRequest {
 		RequestMatcher antPath(RequestMatcherProvider matcherProvider, String... parts) {
 			StringBuilder pattern = new StringBuilder();
 			for (String part : parts) {
+				Assert.notNull(part, "'part' must not be null");
 				pattern.append(part);
 			}
 			return matcherProvider.getRequestMatcher(pattern.toString());

diff --git a/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/JarEntriesStream.java b/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/JarEntriesStream.java
new file mode 100644
index 000000000000..35d9421874b4
--- /dev/null
+++ b/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/JarEntriesStream.java
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2012-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.loader.jar;
+
+import java.io.Closeable;
+import java.io.DataInputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.jar.JarEntry;
+import java.util.jar.JarInputStream;
+import java.util.zip.Inflater;
+import java.util.zip.ZipEntry;
+
+/**
+ * Helper class to iterate entries in a jar file and check that content matches a related
+ * entry.
+ *
+ * @author Phillip Webb
+ * @author Andy Wilkinson
+ */
+class JarEntriesStream implements Closeable {
+
+	private static final int BUFFER_SIZE = 4 * 1024;
+
+	private final JarInputStream in;
+
+	private final byte[] inBuffer = new byte[BUFFER_SIZE];
+
+	private final byte[] compareBuffer = new byte[BUFFER_SIZE];
+
+	private final Inflater inflater = new Inflater(true);
+
+	private JarEntry entry;
+
+	JarEntriesStream(InputStream in) throws IOException {
+		this.in = new JarInputStream(in);
+	}
+
+	JarEntry getNextEntry() throws IOException {
+		this.entry = this.in.getNextJarEntry();
+		if (this.entry != null) {
+			this.entry.getSize();
+		}
+		this.inflater.reset();
+		return this.entry;
+	}
+
+	boolean matches(boolean directory, int size, int compressionMethod, InputStreamSupplier streamSupplier)
+			throws IOException {
+		if (this.entry.isDirectory() != directory) {
+			fail("directory");
+		}
+		if (this.entry.getMethod() != compressionMethod) {
+			fail("compression method");
+		}
+		if (this.entry.isDirectory()) {
+			this.in.closeEntry();
+			return true;
+		}
+		try (DataInputStream expected = new DataInputStream(getInputStream(size, streamSupplier))) {
+			assertSameContent(expected);
+		}
+		return true;
+	}
+
+	private InputStream getInputStream(int size, InputStreamSupplier streamSupplier) throws IOException {
+		InputStream inputStream = streamSupplier.get();
+		return (this.entry.getMethod() != ZipEntry.DEFLATED) ? inputStream
+				: new ZipInflaterInputStream(inputStream, this.inflater, size);
+	}
+
+	private void assertSameContent(DataInputStream expected) throws IOException {
+		int len;
+		while ((len = this.in.read(this.inBuffer)) > 0) {
+			try {
+				expected.readFully(this.compareBuffer, 0, len);
+				if (Arrays.equals(this.inBuffer, 0, len, this.compareBuffer, 0, len)) {
+					continue;
+				}
+			}
+			catch (EOFException ex) {
+				// Continue and throw exception due to mismatched content length.
+			}
+			fail("content");
+		}
+		if (expected.read() != -1) {
+			fail("content");
+		}
+	}
+
+	private void fail(String check) {
+		throw new IllegalStateException("Content mismatch when reading security info for entry '%s' (%s check)"
+			.formatted(this.entry.getName(), check));
+	}
+
+	@Override
+	public void close() throws IOException {
+		this.inflater.end();
+		this.in.close();
+	}
+
+	@FunctionalInterface
+	interface InputStreamSupplier {
+
+		InputStream get() throws IOException;
+
+	}
+
+}
diff --git a/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/JarFileEntries.java b/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/JarFileEntries.java
index d151c8d80a85..bf4e3bcbbd0e 100644
--- a/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/JarFileEntries.java
+++ b/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/JarFileEntries.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012-2023 the original author or authors.
+ * Copyright 2012-2024 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -26,7 +26,6 @@
 import java.util.NoSuchElementException;
 import java.util.jar.Attributes;
 import java.util.jar.Attributes.Name;
-import java.util.jar.JarInputStream;
 import java.util.jar.Manifest;
 import java.util.zip.ZipEntry;
 
@@ -334,37 +333,30 @@ private AsciiBytes applyFilter(AsciiBytes name) {
 	JarEntryCertification getCertification(JarEntry entry) throws IOException {
 		JarEntryCertification[] certifications = this.certifications;
 		if (certifications == null) {
-			certifications = new JarEntryCertification[this.size];
-			// We fall back to use JarInputStream to obtain the certs. This isn't that
-			// fast, but hopefully doesn't happen too often.
-			try (JarInputStream certifiedJarStream = new JarInputStream(this.jarFile.getData().getInputStream())) {
-				java.util.jar.JarEntry certifiedEntry;
-				while ((certifiedEntry = certifiedJarStream.getNextJarEntry()) != null) {
-					// Entry must be closed to trigger a read and set entry certificates
-					certifiedJarStream.closeEntry();
-					int index = getEntryIndex(certifiedEntry.getName());
-					if (index != -1) {
-						certifications[index] = JarEntryCertification.from(certifiedEntry);
-					}
-				}
-			}
+			certifications = getCertifications();
 			this.certifications = certifications;
 		}
 		JarEntryCertification certification = certifications[entry.getIndex()];
 		return (certification != null) ? certification : JarEntryCertification.NONE;
 	}
 
-	private int getEntryIndex(CharSequence name) {
-		int hashCode = AsciiBytes.hashCode(name);
-		int index = getFirstIndex(hashCode);
-		while (index >= 0 && index < this.size && this.hashCodes[index] == hashCode) {
-			FileHeader candidate = getEntry(index, FileHeader.class, false, null);
-			if (candidate.hasName(name, NO_SUFFIX)) {
-				return index;
+	private JarEntryCertification[] getCertifications() throws IOException {
+		JarEntryCertification[] certifications = new JarEntryCertification[this.size];
+		try (JarEntriesStream entries = new JarEntriesStream(this.jarFile.getData().getInputStream())) {
+			java.util.jar.JarEntry entry = entries.getNextEntry();
+			while (entry != null) {
+				JarEntry relatedEntry = this.doGetEntry(entry.getName(), JarEntry.class, false, null);
+				if (relatedEntry != null && entries.matches(relatedEntry.isDirectory(), (int) relatedEntry.getSize(),
+						relatedEntry.getMethod(), () -> getEntryData(relatedEntry).getInputStream())) {
+					int index = relatedEntry.getIndex();
+					if (index != -1) {
+						certifications[index] = JarEntryCertification.from(entry);
+					}
+				}
+				entry = entries.getNextEntry();
 			}
-			index++;
 		}
-		return -1;
+		return certifications;
 	}
 
 	private static void swap(int[] array, int i, int j) {
diff --git a/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/ZipInflaterInputStream.java b/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/ZipInflaterInputStream.java
index 67624460ccd7..71750d1ab432 100644
--- a/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/ZipInflaterInputStream.java
+++ b/spring-boot-project/spring-boot-tools/spring-boot-loader-classic/src/main/java/org/springframework/boot/loader/jar/ZipInflaterInputStream.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012-2023 the original author or authors.
+ * Copyright 2012-2024 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -30,12 +30,23 @@
  */
 class ZipInflaterInputStream extends InflaterInputStream {
 
+	private final boolean ownsInflator;
+
 	private int available;
 
 	private boolean extraBytesWritten;
 
 	ZipInflaterInputStream(InputStream inputStream, int size) {
-		super(inputStream, new Inflater(true), getInflaterBufferSize(size));
+		this(inputStream, new Inflater(true), size, true);
+	}
+
+	ZipInflaterInputStream(InputStream inputStream, Inflater inflater, int size) {
+		this(inputStream, inflater, size, false);
+	}
+
+	private ZipInflaterInputStream(InputStream inputStream, Inflater inflater, int size, boolean ownsInflator) {
+		super(inputStream, inflater, getInflaterBufferSize(size));
+		this.ownsInflator = ownsInflator;
 		this.available = size;
 	}
 
@@ -59,7 +70,9 @@ public int read(byte[] b, int off, int len) throws IOException {
 	@Override
 	public void close() throws IOException {
 		super.close();
-		this.inf.end();
+		if (this.ownsInflator) {
+			this.inf.end();
+		}
 	}
 
 	@Override
diff --git a/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/JarEntriesStream.java b/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/JarEntriesStream.java
new file mode 100644
index 000000000000..35d9421874b4
--- /dev/null
+++ b/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/JarEntriesStream.java
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2012-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.loader.jar;
+
+import java.io.Closeable;
+import java.io.DataInputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.jar.JarEntry;
+import java.util.jar.JarInputStream;
+import java.util.zip.Inflater;
+import java.util.zip.ZipEntry;
+
+/**
+ * Helper class to iterate entries in a jar file and check that content matches a related
+ * entry.
+ *
+ * @author Phillip Webb
+ * @author Andy Wilkinson
+ */
+class JarEntriesStream implements Closeable {
+
+	private static final int BUFFER_SIZE = 4 * 1024;
+
+	private final JarInputStream in;
+
+	private final byte[] inBuffer = new byte[BUFFER_SIZE];
+
+	private final byte[] compareBuffer = new byte[BUFFER_SIZE];
+
+	private final Inflater inflater = new Inflater(true);
+
+	private JarEntry entry;
+
+	JarEntriesStream(InputStream in) throws IOException {
+		this.in = new JarInputStream(in);
+	}
+
+	JarEntry getNextEntry() throws IOException {
+		this.entry = this.in.getNextJarEntry();
+		if (this.entry != null) {
+			this.entry.getSize();
+		}
+		this.inflater.reset();
+		return this.entry;
+	}
+
+	boolean matches(boolean directory, int size, int compressionMethod, InputStreamSupplier streamSupplier)
+			throws IOException {
+		if (this.entry.isDirectory() != directory) {
+			fail("directory");
+		}
+		if (this.entry.getMethod() != compressionMethod) {
+			fail("compression method");
+		}
+		if (this.entry.isDirectory()) {
+			this.in.closeEntry();
+			return true;
+		}
+		try (DataInputStream expected = new DataInputStream(getInputStream(size, streamSupplier))) {
+			assertSameContent(expected);
+		}
+		return true;
+	}
+
+	private InputStream getInputStream(int size, InputStreamSupplier streamSupplier) throws IOException {
+		InputStream inputStream = streamSupplier.get();
+		return (this.entry.getMethod() != ZipEntry.DEFLATED) ? inputStream
+				: new ZipInflaterInputStream(inputStream, this.inflater, size);
+	}
+
+	private void assertSameContent(DataInputStream expected) throws IOException {
+		int len;
+		while ((len = this.in.read(this.inBuffer)) > 0) {
+			try {
+				expected.readFully(this.compareBuffer, 0, len);
+				if (Arrays.equals(this.inBuffer, 0, len, this.compareBuffer, 0, len)) {
+					continue;
+				}
+			}
+			catch (EOFException ex) {
+				// Continue and throw exception due to mismatched content length.
+			}
+			fail("content");
+		}
+		if (expected.read() != -1) {
+			fail("content");
+		}
+	}
+
+	private void fail(String check) {
+		throw new IllegalStateException("Content mismatch when reading security info for entry '%s' (%s check)"
+			.formatted(this.entry.getName(), check));
+	}
+
+	@Override
+	public void close() throws IOException {
+		this.inflater.end();
+		this.in.close();
+	}
+
+	@FunctionalInterface
+	interface InputStreamSupplier {
+
+		InputStream get() throws IOException;
+
+	}
+
+}
diff --git a/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/SecurityInfo.java b/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/SecurityInfo.java
index 3b20bebdbe4d..a6a0a08b1ec2 100644
--- a/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/SecurityInfo.java
+++ b/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/SecurityInfo.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012-2023 the original author or authors.
+ * Copyright 2012-2024 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -81,30 +81,31 @@ static SecurityInfo get(ZipContent content) {
 	 * @return the security info
 	 * @throws IOException on I/O error
 	 */
+	@SuppressWarnings("resource")
 	private static SecurityInfo load(ZipContent content) throws IOException {
 		int size = content.size();
 		boolean hasSecurityInfo = false;
 		Certificate[][] entryCertificates = new Certificate[size][];
 		CodeSigner[][] entryCodeSigners = new CodeSigner[size][];
-		try (JarInputStream in = new JarInputStream(content.openRawZipData().asInputStream())) {
-			JarEntry jarEntry = in.getNextJarEntry();
-			while (jarEntry != null) {
-				in.closeEntry(); // Close to trigger a read and set certs/signers
-				Certificate[] certificates = jarEntry.getCertificates();
-				CodeSigner[] codeSigners = jarEntry.getCodeSigners();
-				if (certificates != null || codeSigners != null) {
-					ZipContent.Entry contentEntry = content.getEntry(jarEntry.getName());
-					if (contentEntry != null) {
+		try (JarEntriesStream entries = new JarEntriesStream(content.openRawZipData().asInputStream())) {
+			JarEntry entry = entries.getNextEntry();
+			while (entry != null) {
+				ZipContent.Entry relatedEntry = content.getEntry(entry.getName());
+				if (relatedEntry != null && entries.matches(relatedEntry.isDirectory(),
+						relatedEntry.getUncompressedSize(), relatedEntry.getCompressionMethod(),
+						() -> relatedEntry.openContent().asInputStream())) {
+					Certificate[] certificates = entry.getCertificates();
+					CodeSigner[] codeSigners = entry.getCodeSigners();
+					if (certificates != null || codeSigners != null) {
 						hasSecurityInfo = true;
-						entryCertificates[contentEntry.getLookupIndex()] = certificates;
-						entryCodeSigners[contentEntry.getLookupIndex()] = codeSigners;
+						entryCertificates[relatedEntry.getLookupIndex()] = certificates;
+						entryCodeSigners[relatedEntry.getLookupIndex()] = codeSigners;
 					}
 				}
-				jarEntry = in.getNextJarEntry();
+				entry = entries.getNextEntry();
 			}
-			return (!hasSecurityInfo) ? NONE : new SecurityInfo(entryCertificates, entryCodeSigners);
 		}
-
+		return (!hasSecurityInfo) ? NONE : new SecurityInfo(entryCertificates, entryCodeSigners);
 	}
 
 }
diff --git a/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/ZipInflaterInputStream.java b/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/ZipInflaterInputStream.java
index 1528f0b9c507..095d24874916 100644
--- a/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/ZipInflaterInputStream.java
+++ b/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/ZipInflaterInputStream.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012-2023 the original author or authors.
+ * Copyright 2012-2024 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -29,7 +29,7 @@
  *
  * @author Phillip Webb
  */
-abstract class ZipInflaterInputStream extends InflaterInputStream {
+class ZipInflaterInputStream extends InflaterInputStream {
 
 	private int available;
 
