--- a/src/virtualenv/app_data/__init__.py
+++ b/src/virtualenv/app_data/__init__.py
@@ -34,12 +34,11 @@
     if is_read_only:
         return ReadOnlyAppData(folder)

-    if not os.path.isdir(folder):
-        try:
-            os.makedirs(folder)
-            logging.debug("created app data folder %s", folder)
-        except OSError as exception:
-            logging.info("could not create app data folder %s due to %r", folder, exception)
+    try:
+        os.makedirs(folder, exist_ok=True)
+        logging.debug("created app data folder %s", folder)
+    except OSError as exception:
+        logging.info("could not create app data folder %s due to %r", folder, exception)

     if os.access(folder, os.W_OK):
         return AppDataDiskFolder(folder)
--- a/src/virtualenv/util/lock.py
+++ b/src/virtualenv/util/lock.py
@@ -15,9 +15,8 @@
 class _CountedFileLock(FileLock):
     def __init__(self, lock_file) -> None:
         parent = os.path.dirname(lock_file)
-        if not os.path.isdir(parent):
-            with suppress(OSError):
-                os.makedirs(parent)
+        with suppress(OSError):
+            os.makedirs(parent, exist_ok=True)

         super().__init__(lock_file)
         self.count = 0
@@ -109,7 +108,7 @@
         # multiple processes might be trying to get a first lock... so we cannot check if this directory exist without
         # a lock, but that lock might then become expensive, and it's not clear where that lock should live.
         # Instead here we just ignore if we fail to create the directory.
         with suppress(OSError):
-            os.makedirs(str(self.path))
+            os.makedirs(str(self.path), exist_ok=True)

         try:
             lock.acquire(0.0001)
