--- a/validation/rules/OverlappingFieldsCanBeMergedRule.js
+++ b/validation/rules/OverlappingFieldsCanBeMergedRule.js
@@ -543,7 +543,7 @@
       ];
     } // Two field calls must have the same arguments.
 
-    if (stringifyArguments(node1) !== stringifyArguments(node2)) {
+    if (!sameArguments(node1, node2)) {
       return [
         [responseName, 'they have differing arguments'],
         [node1],
@@ -588,27 +588,37 @@
   }
 }
 
-function stringifyArguments(fieldNode) {
-  var _fieldNode$arguments;
+function sameArguments(node1, node2) {
+  const args1 = node1.arguments;
+  const args2 = node2.arguments;
 
-  // FIXME https://github.com/graphql/graphql-js/issues/2203
-  const args =
-    /* c8 ignore next */
-    (_fieldNode$arguments = fieldNode.arguments) !== null &&
-    _fieldNode$arguments !== void 0
-      ? _fieldNode$arguments
-      : [];
-  const inputObjectWithArgs = {
-    kind: _kinds.Kind.OBJECT,
-    fields: args.map((argNode) => ({
-      kind: _kinds.Kind.OBJECT_FIELD,
-      name: argNode.name,
-      value: argNode.value,
-    })),
-  };
-  return (0, _printer.print)(
-    (0, _sortValueNode.sortValueNode)(inputObjectWithArgs),
-  );
+  if (args1 === undefined || args1.length === 0) {
+    return args2 === undefined || args2.length === 0;
+  }
+
+  if (args2 === undefined || args2.length === 0) {
+    return false;
+  }
+
+  if (args1.length !== args2.length) {
+    return false;
+  }
+
+  const values2 = new Map(args2.map(({ name, value }) => [name.value, value]));
+  return args1.every((arg1) => {
+    const value1 = arg1.value;
+    const value2 = values2.get(arg1.name.value);
+
+    if (value2 === undefined) {
+      return false;
+    }
+
+    return stringifyValue(value1) === stringifyValue(value2);
+  });
+}
+
+function stringifyValue(value) {
+  return (0, _printer.print)((0, _sortValueNode.sortValueNode)(value));
 } // Two types conflict if both types could not apply to a value simultaneously.
 // Composite types are ignored as their individual field types will be compared
 // later recursively. However List and Non-Null types must match.
--- a/validation/rules/OverlappingFieldsCanBeMergedRule.mjs
+++ b/validation/rules/OverlappingFieldsCanBeMergedRule.mjs
@@ -537,7 +537,7 @@
       ];
     } // Two field calls must have the same arguments.
 
-    if (stringifyArguments(node1) !== stringifyArguments(node2)) {
+    if (!sameArguments(node1, node2)) {
       return [
         [responseName, 'they have differing arguments'],
         [node1],
@@ -582,25 +582,37 @@
   }
 }
 
-function stringifyArguments(fieldNode) {
-  var _fieldNode$arguments;
+function sameArguments(node1, node2) {
+  const args1 = node1.arguments;
+  const args2 = node2.arguments;
 
-  // FIXME https://github.com/graphql/graphql-js/issues/2203
-  const args =
-    /* c8 ignore next */
-    (_fieldNode$arguments = fieldNode.arguments) !== null &&
-    _fieldNode$arguments !== void 0
-      ? _fieldNode$arguments
-      : [];
-  const inputObjectWithArgs = {
-    kind: Kind.OBJECT,
-    fields: args.map((argNode) => ({
-      kind: Kind.OBJECT_FIELD,
-      name: argNode.name,
-      value: argNode.value,
-    })),
-  };
-  return print(sortValueNode(inputObjectWithArgs));
+  if (args1 === undefined || args1.length === 0) {
+    return args2 === undefined || args2.length === 0;
+  }
+
+  if (args2 === undefined || args2.length === 0) {
+    return false;
+  }
+
+  if (args1.length !== args2.length) {
+    return false;
+  }
+
+  const values2 = new Map(args2.map(({ name, value }) => [name.value, value]));
+  return args1.every((arg1) => {
+    const value1 = arg1.value;
+    const value2 = values2.get(arg1.name.value);
+
+    if (value2 === undefined) {
+      return false;
+    }
+
+    return stringifyValue(value1) === stringifyValue(value2);
+  });
+}
+
+function stringifyValue(value) {
+  return print(sortValueNode(value));
 } // Two types conflict if both types could not apply to a value simultaneously.
 // Composite types are ignored as their individual field types will be compared
 // later recursively. However List and Non-Null types must match.
