--- a/ecc/p384/point.go
+++ b/ecc/p384/point.go
@@ -87,6 +87,24 @@
 	return t
 }
 
+// oddMultiplesProjective calculates the points iP for i={1,3,5,7,..., 2^(n-1)-1}
+// Ensure that 1 < n < 31, otherwise it returns an empty slice.
+func (ap affinePoint) oddMultiplesProjective(n uint) []projectivePoint {
+	var t []projectivePoint
+	if n > 1 && n < 31 {
+		P := ap.toProjective()
+		s := int32(1) << (n - 1)
+		t = make([]projectivePoint, s)
+		t[0] = *P
+		var _2P projectivePoint
+		_2P.completeAdd(P, P)
+		for i := int32(1); i < s; i++ {
+			t[i].completeAdd(&t[i-1], &_2P)
+		}
+	}
+	return t
+}
+
 // p2Point is a point in P^2
 type p2Point struct{ x, y, z fp384 }
 
@@ -138,10 +156,8 @@
 	return &aP
 }
 
-func (P *jacobianPoint) cmov(Q *jacobianPoint, b int) { P.p2Point.cmov(&Q.p2Point, b) }
-
-// add calculates P=Q+R such that Q and R are different than the identity point,
-// and Q!==R. This function cannot be used for doublings.
+// add calculates P=Q+R such that Q and R are different than the identity point.
+// This function can be used for doublings.
 func (P *jacobianPoint) add(Q, R *jacobianPoint) {
 	if Q.isZero() {
 		*P = *R
@@ -171,6 +187,14 @@
 	fp384Sqr(HH, H)        // HH = H ^ 2
 	fp384Mul(HHH, H, HH)   // HHH = H * HH
 	fp384Sub(RR, S2, S1)   // r = S2 - S1
+
+	// Detect point doubling
+	if zero := (fp384{}); *H == zero && *RR == zero {
+		*P = *Q
+		P.double()
+		return
+	}
+
 	fp384Mul(V, U1, HH)    // V = U1 * HH
 	fp384Sqr(t2, RR)       // t2 = r ^ 2
 	fp384Add(t3, V, V)     // t3 = V + V
@@ -187,6 +211,7 @@
 // mixadd calculates P=Q+R such that P and Q different than the identity point,
 // and Q not in {P,-P, O}.
 func (P *jacobianPoint) mixadd(Q *jacobianPoint, R *affinePoint) {
+	// See https://www.hyperelliptic.org/EFD/g1p/auto-shortw-jacobian-3.html#addition-madd-2007-bl
 	if Q.isZero() {
 		*P = *R.toJacobian()
 		return
@@ -237,6 +262,7 @@
 }
 
 func (P *jacobianPoint) double() {
+	// See https://hyperelliptic.org/EFD/g1p/auto-shortw-jacobian-3.html#doubling-dbl-2001-b
 	delta, gamma, alpha, alpha2 := &fp384{}, &fp384{}, &fp384{}, &fp384{}
 	fp384Sqr(delta, &P.z)
 	fp384Sqr(gamma, &P.y)
@@ -292,6 +318,8 @@
 	zero := fp384{}
 	return P.x == zero && P.y != zero && P.z == zero
 }
+
+func (P *projectivePoint) cmov(Q *projectivePoint, b int) { P.p2Point.cmov(&Q.p2Point, b) }
 
 func (P *projectivePoint) toAffine() *affinePoint {
 	var aP affinePoint

--- a/ecc/p384/p384opt.go
+++ b/ecc/p384/p384opt.go
@@ -94,33 +94,32 @@
 	const bitsN = uint(384)
 	L := math.SignedDigit(&scalar, omega, bitsN)
 
-	var R jacobianPoint
-	Q := zeroPoint().toJacobian()
-	TabP := newAffinePoint(x1, y1).oddMultiples(omega)
+	var R projectivePoint
+	Q := zeroPoint().toProjective()
+	TabP := newAffinePoint(x1, y1).oddMultiplesProjective(omega)
 	for i := len(L) - 1; i > 0; i-- {
 		for j := uint(0); j < omega-1; j++ {
-			Q.double()
+			Q.completeAdd(Q, Q)
 		}
 		idx := absolute(L[i]) >> 1
 		for j := range TabP {
 			R.cmov(&TabP[j], subtle.ConstantTimeEq(int32(j), idx))
 		}
 		R.cneg(int(L[i]>>31) & 1)
-		Q.add(Q, &R)
+		Q.completeAdd(Q, &R)
 	}
 	// Calculate the last iteration using complete addition formula.
 	for j := uint(0); j < omega-1; j++ {
-		Q.double()
+		Q.completeAdd(Q, Q)
 	}
 	idx := absolute(L[0]) >> 1
 	for j := range TabP {
 		R.cmov(&TabP[j], subtle.ConstantTimeEq(int32(j), idx))
 	}
 	R.cneg(int(L[0]>>31) & 1)
-	QQ := Q.toProjective()
-	QQ.completeAdd(QQ, R.toProjective())
-	QQ.cneg(isEvenK)
-	return QQ.toAffine().toInt()
+	Q.completeAdd(Q, &R)
+	Q.cneg(isEvenK)
+	return Q.toAffine().toInt()
 }
 
 // ScalarBaseMult returns k*G, where G is the base point of the group

--- a/ecc/p384/point_test.go
+++ b/ecc/p384/point_test.go
@@ -130,13 +130,16 @@
 	})
 
 	t.Run("P+P=2P", func(t *testing.T) {
-		// This verifies that add function cannot be used for doublings.
+		// This verifies that add function can be used for doublings.
 		for i := 0; i < 128; i++ {
 			P = randomJacobian()
+			Q := *P
 
 			R.add(P, P)
 			gotX, gotY := R.toAffine().toInt()
-			wantX, wantY := zeroPoint().toInt()
+
+			Q.double()
+			wantX, wantY := Q.toAffine().toInt()
 
 			if gotX.Cmp(wantX) != 0 {
 				test.ReportError(t, gotX, wantX, P)
