--- a/src/main/java/org/apache/commons/text/StringSubstitutor.java
+++ b/src/main/java/org/apache/commons/text/StringSubstitutor.java
@@ -136,32 +136,34 @@
  *
  * <h2>Using Interpolation</h2>
  * <p>
- * The default interpolator let's you use string lookups like:
+ * The default interpolator lets you use string lookups like:
  * </p>
  *
  * <pre>
  * final StringSubstitutor interpolator = StringSubstitutor.createInterpolator();
- * interpolator.setEnableSubstitutionInVariables(true); // Allows for nested $'s.
- * final String text = interpolator.replace("Base64 Decoder:        ${base64Decoder:SGVsbG9Xb3JsZCE=}\n"
+ * final String text = interpolator.replace(
+ *       "Base64 Decoder:        ${base64Decoder:SGVsbG9Xb3JsZCE=}\n"
  *     + "Base64 Encoder:        ${base64Encoder:HelloWorld!}\n"
  *     + "Java Constant:         ${const:java.awt.event.KeyEvent.VK_ESCAPE}\n"
- *     + "Date:                  ${date:yyyy-MM-dd}\n" + "DNS:                   ${dns:address|apache.org}\n"
+ *     + "Date:                  ${date:yyyy-MM-dd}\n"
  *     + "Environment Variable:  ${env:USERNAME}\n"
  *     + "File Content:          ${file:UTF-8:src/test/resources/document.properties}\n"
- *     + "Java:                  ${java:version}\n" + "Localhost:             ${localhost:canonical-name}\n"
+ *     + "Java:                  ${java:version}\n"
+ *     + "Localhost:             ${localhost:canonical-name}\n"
  *     + "Properties File:       ${properties:src/test/resources/document.properties::mykey}\n"
- *     + "Resource Bundle:       ${resourceBundle:org.example.testResourceBundleLookup:mykey}\n"
- *     + "Script:                ${script:javascript:3 + 4}\n" + "System Property:       ${sys:user.dir}\n"
+ *     + "Resource Bundle:       ${resourceBundle:org.apache.commons.text.example.testResourceBundleLookup:mykey}\n"
+ *     + "System Property:       ${sys:user.dir}\n"
  *     + "URL Decoder:           ${urlDecoder:Hello%20World%21}\n"
  *     + "URL Encoder:           ${urlEncoder:Hello World!}\n"
- *     + "URL Content (HTTP):    ${url:UTF-8:http://www.apache.org}\n"
- *     + "URL Content (HTTPS):   ${url:UTF-8:https://www.apache.org}\n"
- *     + "URL Content (File):    ${url:UTF-8:file:///${sys:user.dir}/src/test/resources/document.properties}\n"
  *     + "XML XPath:             ${xml:src/test/resources/document.xml:/root/path/to/node}\n");
  * </pre>
  * <p>
- * For documentation of each lookup, see {@link StringLookupFactory}.
+ * For documentation and a full list of available lookups, see {@link StringLookupFactory}.
  * </p>
+ * <p><strong>NOTE:</strong> The list of lookups available by default in {@link #createInterpolator()} changed
+ * in version {@code 1.10.0}. See the {@link StringLookupFactory} documentation for details and an explanation
+ * on how to reproduce the previous functionality.
+ * </p>
  *
  * <h2>Using Recursive Variable Replacement</h2>
  * <p>
@@ -231,7 +233,6 @@
         public final int lengthChange;
 
         private Result(final boolean altered, final int lengthChange) {
-            super();
             this.altered = altered;
             this.lengthChange = lengthChange;
         }
@@ -288,14 +289,86 @@
      * Creates a new instance using the interpolator string lookup
      * {@link StringLookupFactory#interpolatorStringLookup()}.
      * <p>
-     * This StringSubstitutor lets you perform substituions like:
+     * This StringSubstitutor lets you perform substitutions like:
      * </p>
      *
      * <pre>
      * StringSubstitutor.createInterpolator().replace(
-     *   "OS name: ${sys:os.name}, " + "3 + 4 = ${script:javascript:3 + 4}");
+     *   "OS name: ${sys:os.name}, user: ${env:USER}");
      * </pre>
      *
+     * <p>The table below lists the lookups available by default in the returned instance. These
+     * may be modified through the use of the {@value StringLookupFactory#DEFAULT_STRING_LOOKUPS_PROPERTY}
+     * system property, as described in the {@link StringLookupFactory} documentation.</p>
+     *
+     * <p><strong>NOTE:</strong> The list of lookups available by default changed in version {@code 1.10.0}.
+     * Configuration via system property (as mentioned above) may be necessary to reproduce previous functionality.
+     * </p>
+     *
+     * <table>
+     * <caption>Default Lookups</caption>
+     * <tr>
+     * <th>Key</th>
+     * <th>Lookup</th>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_BASE64_DECODER}</td>
+     * <td>{@link StringLookupFactory#base64DecoderStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_BASE64_ENCODER}</td>
+     * <td>{@link StringLookupFactory#base64EncoderStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_CONST}</td>
+     * <td>{@link StringLookupFactory#constantStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_DATE}</td>
+     * <td>{@link StringLookupFactory#dateStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_ENV}</td>
+     * <td>{@link StringLookupFactory#environmentVariableStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_FILE}</td>
+     * <td>{@link StringLookupFactory#fileStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_JAVA}</td>
+     * <td>{@link StringLookupFactory#javaPlatformStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_LOCALHOST}</td>
+     * <td>{@link StringLookupFactory#localHostStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_PROPERTIES}</td>
+     * <td>{@link StringLookupFactory#propertiesStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_RESOURCE_BUNDLE}</td>
+     * <td>{@link StringLookupFactory#resourceBundleStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_SYS}</td>
+     * <td>{@link StringLookupFactory#systemPropertyStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_URL_DECODER}</td>
+     * <td>{@link StringLookupFactory#urlDecoderStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_URL_ENCODER}</td>
+     * <td>{@link StringLookupFactory#urlEncoderStringLookup()}</td>
+     * </tr>
+     * <tr>
+     * <td>{@value StringLookupFactory#KEY_XML}</td>
+     * <td>{@link StringLookupFactory#xmlStringLookup()}</td>
+     * </tr>
+     * </table>
+     *
      * @return a new instance using the interpolator string lookup.
      * @see StringLookupFactory#interpolatorStringLookup()
      * @since 1.8
@@ -351,7 +424,7 @@
         final Map<String, String> valueMap = new HashMap<>();
         final Enumeration<?> propNames = valueProperties.propertyNames();
         while (propNames.hasMoreElements()) {
-            final String propName = (String) propNames.nextElement();
+            final String propName = String.valueOf(propNames.nextElement());
             final String propValue = valueProperties.getProperty(propName);
             valueMap.put(propName, propValue);
         }
@@ -559,7 +632,7 @@
     /**
      * Creates a new instance based on the given StringSubstitutor.
      *
-     * @param other The StringSubstitutor is use as the source.
+     * @param other The StringSubstitutor used as the source.
      * @since 1.9
      */
     public StringSubstitutor(final StringSubstitutor other) {
@@ -735,6 +808,10 @@
      * @param length the length within the array to be processed, must be valid
      * @return The result of the replace operation
      * @throws IllegalArgumentException if variable is not found when its allowed to throw exception
+     * @throws StringIndexOutOfBoundsException if {@code offset} is not in the
+     *  range {@code 0 <= offset <= chars.length}
+     * @throws StringIndexOutOfBoundsException if {@code length < 0}
+     * @throws StringIndexOutOfBoundsException if {@code offset + length > chars.length}
      */
     public String replace(final char[] source, final int offset, final int length) {
         if (source == null) {
@@ -832,6 +909,10 @@
      * @param length the length within the source to be processed, must be valid
      * @return The result of the replace operation
      * @throws IllegalArgumentException if variable is not found when its allowed to throw exception
+     * @throws StringIndexOutOfBoundsException if {@code offset} is not in the
+     *  range {@code 0 <= offset <= source.length()}
+     * @throws StringIndexOutOfBoundsException if {@code length < 0}
+     * @throws StringIndexOutOfBoundsException if {@code offset + length > source.length()}
      */
     public String replace(final String source, final int offset, final int length) {
         if (source == null) {
@@ -1157,7 +1238,7 @@
      * @return this, to enable chaining
      */
     public StringSubstitutor setValueDelimiter(final String valueDelimiter) {
-        if (valueDelimiter == null || valueDelimiter.length() == 0) {
+        if (valueDelimiter == null || valueDelimiter.isEmpty()) {
             setValueDelimiterMatcher(null);
             return this;
         }
@@ -1317,7 +1398,7 @@
      * @param length the length within the builder to be processed, must be valid
      * @param priorVariables the stack keeping track of the replaced variables, may be null
      * @return The result.
-     * @throws IllegalArgumentException if variable is not found when its allowed to throw exception
+     * @throws IllegalArgumentException if variable is not found and <pre>isEnableUndefinedVariableException()==true</pre>
      * @since 1.9
      */
     private Result substitute(final TextStringBuilder builder, final int offset, final int length,
--- a/src/main/java/org/apache/commons/text/lookup/DefaultStringLookup.java
+++ b/src/main/java/org/apache/commons/text/lookup/DefaultStringLookup.java
@@ -21,6 +21,11 @@
  * <p>
  * This enum was adapted and expanded from Apache Commons Configuration 2.4.
  * </p>
+ * <p><strong>NOTE:</strong> Starting in version 1.10.0, not all lookups defined in this class are
+ * included by default in the
+ * {@link StringLookupFactory#addDefaultStringLookups(java.util.Map) StringLookupFactory.addDefaultStringLookups}
+ * method. See the {@link StringLookupFactory} class documentation for details.
+ * </p>
  *
  * @see StringLookupFactory
  * @see StringLookup
@@ -29,89 +34,122 @@
 public enum DefaultStringLookup {
 
     /**
-     * The lookup for Base64 decoding using the key {@value StringLookupFactory#KEY_BASE64_DECODER}.
+     * The lookup for Base64 decoding using the key {@code "base64Decoder"}.
+     * @see StringLookupFactory#KEY_BASE64_DECODER
+     * @see StringLookupFactory#base64DecoderStringLookup()
      */
     BASE64_DECODER(StringLookupFactory.KEY_BASE64_DECODER, StringLookupFactory.INSTANCE.base64DecoderStringLookup()),
 
     /**
-     * The lookup for Base64 decoding using the key {@value StringLookupFactory#KEY_BASE64_ENCODER}.
+     * The lookup for Base64 decoding using the key {@code "base64Encoder"}.
+     * @see StringLookupFactory#KEY_BASE64_ENCODER
+     * @see StringLookupFactory#base64EncoderStringLookup()
      */
     BASE64_ENCODER(StringLookupFactory.KEY_BASE64_ENCODER, StringLookupFactory.INSTANCE.base64EncoderStringLookup()),
 
     /**
-     * The lookup for constants using the key {@value StringLookupFactory#KEY_CONST}.
+     * The lookup for Java static class member constants using the key {@code "const"}.
+     * @see StringLookupFactory#KEY_CONST
+     * @see StringLookupFactory#constantStringLookup()
      */
     CONST(StringLookupFactory.KEY_CONST, StringLookupFactory.INSTANCE.constantStringLookup()),
 
     /**
-     * The lookup for dates using the key {@value StringLookupFactory#KEY_DATE}.
+     * The lookup for formatting the current date using the key {@code "date"}.
+     * @see StringLookupFactory#KEY_DATE
+     * @see StringLookupFactory#dateStringLookup()
      */
     DATE(StringLookupFactory.KEY_DATE, StringLookupFactory.INSTANCE.dateStringLookup()),
 
     /**
-     * The lookup for DNS using the key {@value StringLookupFactory#KEY_DNS}.
-     *
+     * The lookup for DNS using the key {@code "dns"}.
+     * @see StringLookupFactory#KEY_DNS
+     * @see StringLookupFactory#dnsStringLookup()
      * @since 1.8
      */
     DNS(StringLookupFactory.KEY_DNS, StringLookupFactory.INSTANCE.dnsStringLookup()),
 
     /**
-     * The lookup for environment properties using the key {@value StringLookupFactory#KEY_ENV}.
+     * The lookup for environment properties using the key {@code "env"}.
+     * @see StringLookupFactory#KEY_ENV
+     * @see StringLookupFactory#environmentVariableStringLookup()
      */
     ENVIRONMENT(StringLookupFactory.KEY_ENV, StringLookupFactory.INSTANCE.environmentVariableStringLookup()),
 
     /**
-     * The lookup for files using the key {@value StringLookupFactory#KEY_FILE}.
+     * The lookup for files using the key {@code "file"}.
+     * @see StringLookupFactory#KEY_FILE
+     * @see StringLookupFactory#fileStringLookup()
      */
     FILE(StringLookupFactory.KEY_FILE, StringLookupFactory.INSTANCE.fileStringLookup()),
 
     /**
-     * The lookup for Java platform information using the key {@value StringLookupFactory#KEY_JAVA}.
+     * The lookup for Java platform information using the key {@code "java"}.
+     * @see StringLookupFactory#KEY_JAVA
+     * @see StringLookupFactory#javaPlatformStringLookup()
      */
     JAVA(StringLookupFactory.KEY_JAVA, StringLookupFactory.INSTANCE.javaPlatformStringLookup()),
 
     /**
-     * The lookup for localhost information using the key {@value StringLookupFactory#KEY_LOCALHOST}.
+     * The lookup for localhost information using the key {@code "localhost"}.
+     * @see StringLookupFactory#KEY_LOCALHOST
+     * @see StringLookupFactory#localHostStringLookup()
      */
     LOCAL_HOST(StringLookupFactory.KEY_LOCALHOST, StringLookupFactory.INSTANCE.localHostStringLookup()),
 
     /**
-     * The lookup for properties using the key {@value StringLookupFactory#KEY_PROPERTIES}.
+     * The lookup for properties using the key {@code "properties"}.
+     * @see StringLookupFactory#KEY_PROPERTIES
+     * @see StringLookupFactory#propertiesStringLookup()
      */
     PROPERTIES(StringLookupFactory.KEY_PROPERTIES, StringLookupFactory.INSTANCE.propertiesStringLookup()),
 
     /**
-     * The lookup for resource bundles using the key {@value StringLookupFactory#KEY_RESOURCE_BUNDLE}.
+     * The lookup for resource bundles using the key {@code "resourceBundle"}.
+     * @see StringLookupFactory#KEY_RESOURCE_BUNDLE
+     * @see StringLookupFactory#resourceBundleStringLookup()
      */
     RESOURCE_BUNDLE(StringLookupFactory.KEY_RESOURCE_BUNDLE, StringLookupFactory.INSTANCE.resourceBundleStringLookup()),
 
     /**
-     * The lookup for scripts using the key {@value StringLookupFactory#KEY_SCRIPT}.
+     * The lookup for scripts using the key {@code "script"}.
+     * @see StringLookupFactory#KEY_SCRIPT
+     * @see StringLookupFactory#scriptStringLookup()
      */
     SCRIPT(StringLookupFactory.KEY_SCRIPT, StringLookupFactory.INSTANCE.scriptStringLookup()),
 
     /**
-     * The lookup for system properties using the key {@value StringLookupFactory#KEY_SYS}.
+     * The lookup for system properties using the key {@code "sys"}.
+     * @see StringLookupFactory#KEY_SYS
+     * @see StringLookupFactory#systemPropertyStringLookup()
      */
     SYSTEM_PROPERTIES(StringLookupFactory.KEY_SYS, StringLookupFactory.INSTANCE.systemPropertyStringLookup()),
 
     /**
-     * The lookup for URLs using the key {@value StringLookupFactory#KEY_URL}.
+     * The lookup for URLs using the key {@code "url"}.
+     * @see StringLookupFactory#KEY_URL
+     * @see StringLookupFactory#urlStringLookup()
      */
     URL(StringLookupFactory.KEY_URL, StringLookupFactory.INSTANCE.urlStringLookup()),
 
     /**
-     * The lookup for URL decoding using the key {@value StringLookupFactory#KEY_URL_DECODER}.
+     * The lookup for URL decoding using the key {@code "urlDecoder"}.
+     * @see StringLookupFactory#KEY_URL_DECODER
+     * @see StringLookupFactory#urlDecoderStringLookup()
      */
     URL_DECODER(StringLookupFactory.KEY_URL_DECODER, StringLookupFactory.INSTANCE.urlDecoderStringLookup()),
 
     /**
-     * The lookup for URL decoding using the key {@value StringLookupFactory#KEY_URL_ENCODER}.
+     * The lookup for URL decoding using the key {@code "urlEncoder"}.
+     * @see StringLookupFactory#KEY_URL_ENCODER
+     * @see StringLookupFactory#urlEncoderStringLookup()
      */
     URL_ENCODER(StringLookupFactory.KEY_URL_ENCODER, StringLookupFactory.INSTANCE.urlEncoderStringLookup()),
 
     /**
-     * The lookup for URL decoding using the key {@value StringLookupFactory#KEY_XML}.
+     * The lookup for URL decoding using the key {@code "xml"}.
+     * @see StringLookupFactory#KEY_XML
+     * @see StringLookupFactory#xmlStringLookup()
      */
     XML(StringLookupFactory.KEY_XML, StringLookupFactory.INSTANCE.xmlStringLookup());
 
--- a/src/main/java/org/apache/commons/text/lookup/InterpolatorStringLookup.java
+++ b/src/main/java/org/apache/commons/text/lookup/InterpolatorStringLookup.java
@@ -16,8 +16,8 @@
  */
 package org.apache.commons.text.lookup;
 
+import java.util.Collections;
 import java.util.HashMap;
-import java.util.Locale;
 import java.util.Map;
 import java.util.Map.Entry;
 
@@ -39,10 +39,6 @@
     /** Constant for the prefix separator. */
     private static final char PREFIX_SEPARATOR = ':';
 
-    static String toKey(final String key) {
-        return key.toLowerCase(Locale.ROOT);
-    }
-
     /** The default string lookup. */
     private final StringLookup defaultStringLookup;
 
@@ -50,7 +46,7 @@
     private final Map<String, StringLookup> stringLookupMap;
 
     /**
-     * Creates an instance using only lookups that work without initial properties and are stateless.
+     * Constructs an instance using only lookups that work without initial properties and are stateless.
      * <p>
      * Uses the {@link StringLookupFactory default lookups}.
      * </p>
@@ -60,7 +56,7 @@
     }
 
     /**
-     * Creates a fully customized instance.
+     * Constructs a fully customized instance.
      *
      * @param stringLookupMap the map of string lookups.
      * @param defaultStringLookup the default string lookup.
@@ -68,11 +64,10 @@
      */
     InterpolatorStringLookup(final Map<String, StringLookup> stringLookupMap, final StringLookup defaultStringLookup,
         final boolean addDefaultLookups) {
-        super();
         this.defaultStringLookup = defaultStringLookup;
         this.stringLookupMap = new HashMap<>(stringLookupMap.size());
         for (final Entry<String, StringLookup> entry : stringLookupMap.entrySet()) {
-            this.stringLookupMap.put(toKey(entry.getKey()), entry.getValue());
+            this.stringLookupMap.put(StringLookupFactory.toKey(entry.getKey()), entry.getValue());
         }
         if (addDefaultLookups) {
             StringLookupFactory.INSTANCE.addDefaultStringLookups(this.stringLookupMap);
@@ -80,7 +75,7 @@
     }
 
     /**
-     * Creates an instance using only lookups that work without initial properties and are stateless.
+     * Constructs an instance using only lookups that work without initial properties and are stateless.
      * <p>
      * Uses the {@link StringLookupFactory default lookups}.
      * </p>
@@ -89,16 +84,16 @@
      * @param defaultMap the default map for string lookups.
      */
     <V> InterpolatorStringLookup(final Map<String, V> defaultMap) {
-        this(StringLookupFactory.INSTANCE.mapStringLookup(defaultMap == null ? new HashMap<String, V>() : defaultMap));
+        this(StringLookupFactory.INSTANCE.mapStringLookup(defaultMap));
     }
 
     /**
-     * Creates an instance with the given lookup.
+     * Constructs an instance with the given lookup.
      *
      * @param defaultStringLookup the default lookup.
      */
     InterpolatorStringLookup(final StringLookup defaultStringLookup) {
-        this(new HashMap<>(), defaultStringLookup, true);
+        this(Collections.emptyMap(), defaultStringLookup, true);
     }
 
     /**
@@ -116,19 +111,19 @@
      * the prefix stripped to the lookup object registered for this prefix. If no prefix can be found or if the
      * associated lookup object cannot resolve this variable, the default lookup object will be used.
      *
-     * @param var the name of the variable whose value is to be looked up
+     * @param key the name of the variable whose value is to be looked up
      * @return The value of this variable or <b>null</b> if it cannot be resolved
      */
     @Override
-    public String lookup(String var) {
-        if (var == null) {
+    public String lookup(String key) {
+        if (key == null) {
             return null;
         }
 
-        final int prefixPos = var.indexOf(PREFIX_SEPARATOR);
+        final int prefixPos = key.indexOf(PREFIX_SEPARATOR);
         if (prefixPos >= 0) {
-            final String prefix = toKey(var.substring(0, prefixPos));
-            final String name = var.substring(prefixPos + 1);
+            final String prefix = StringLookupFactory.toKey(key.substring(0, prefixPos));
+            final String name = key.substring(prefixPos + 1);
             final StringLookup lookup = stringLookupMap.get(prefix);
             String value = null;
             if (lookup != null) {
@@ -138,10 +133,10 @@
             if (value != null) {
                 return value;
             }
-            var = var.substring(prefixPos + 1);
+            key = key.substring(prefixPos + 1);
         }
         if (defaultStringLookup != null) {
-            return defaultStringLookup.lookup(var);
+            return defaultStringLookup.lookup(key);
         }
         return null;
     }
--- a/src/main/java/org/apache/commons/text/lookup/StringLookupFactory.java
+++ b/src/main/java/org/apache/commons/text/lookup/StringLookupFactory.java
@@ -19,7 +19,11 @@
 
 import java.nio.charset.StandardCharsets;
 import java.util.Base64;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Locale;
 import java.util.Map;
+import java.util.Properties;
 import java.util.function.BiFunction;
 import java.util.function.Function;
 
@@ -40,7 +44,19 @@
  * <li>{@link #interpolatorStringLookup(Map, StringLookup, boolean)}.</li>
  * </ul>
  * <p>
- * The default lookups are:
+ * Unless explicitly requested otherwise, a set of default lookups are included for convenience with these
+ * variable interpolation methods. These defaults are listed in the table below. However, the exact lookups
+ * included can be configured through the use of the {@value #DEFAULT_STRING_LOOKUPS_PROPERTY} system property.
+ * If present, this system property will be parsed as a comma-separated list of lookup names, with the names
+ * being those defined by the {@link DefaultStringLookup} enum. For example, setting this system property to
+ * {@code "BASE64_ENCODER,ENVIRONMENT"} will only include the
+ * {@link DefaultStringLookup#BASE64_ENCODER BASE64_ENCODER} and {@link DefaultStringLookup#ENVIRONMENT ENVIRONMENT}
+ * lookups. Setting the property to the empty string will cause no defaults to be configured.
+ * Note that not all lookups defined here and in {@link DefaultStringLookup} are included by default.
+ * Specifically, lookups that can execute code (e.g., {@link DefaultStringLookup#SCRIPT SCRIPT}) and those
+ * that can result in contact with remote servers (e.g., {@link DefaultStringLookup#URL URL} and
+ * {@link DefaultStringLookup#DNS DNS}) are not included by default. The current set of default lookups can
+ * be accessed directly with {@link #addDefaultStringLookups(Map)}.
  * </p>
  * <table>
  * <caption>Default String Lookups</caption>
@@ -75,12 +91,6 @@
  * <td>1.5</td>
  * </tr>
  * <tr>
- * <td>{@value #KEY_DNS}</td>
- * <td>{@link StringLookup}</td>
- * <td>{@link #dnsStringLookup()}</td>
- * <td>1.8</td>
- * </tr>
- * <tr>
  * <td>{@value #KEY_ENV}</td>
  * <td>{@link StringLookup}</td>
  * <td>{@link #environmentVariableStringLookup()}</td>
@@ -117,24 +127,12 @@
  * <td>1.6</td>
  * </tr>
  * <tr>
- * <td>{@value #KEY_SCRIPT}</td>
- * <td>{@link StringLookup}</td>
- * <td>{@link #scriptStringLookup()}</td>
- * <td>1.5</td>
- * </tr>
- * <tr>
  * <td>{@value #KEY_SYS}</td>
  * <td>{@link StringLookup}</td>
  * <td>{@link #systemPropertyStringLookup()}</td>
  * <td>1.3</td>
  * </tr>
  * <tr>
- * <td>{@value #KEY_URL}</td>
- * <td>{@link StringLookup}</td>
- * <td>{@link #urlStringLookup()}</td>
- * <td>1.5</td>
- * </tr>
- * <tr>
  * <td>{@value #KEY_URL_DECODER}</td>
  * <td>{@link StringLookup}</td>
  * <td>{@link #urlDecoderStringLookup()}</td>
@@ -153,8 +151,37 @@
  * <td>1.5</td>
  * </tr>
  * </table>
+ *
+ * <table>
+ * <caption>Additional String Lookups (not included by default)</caption>
+ * <tr>
+ * <th>Key</th>
+ * <th>Interface</th>
+ * <th>Factory Method</th>
+ * <th>Since</th>
+ * </tr>
+ * <tr>
+ * <td>{@value #KEY_DNS}</td>
+ * <td>{@link StringLookup}</td>
+ * <td>{@link #dnsStringLookup()}</td>
+ * <td>1.8</td>
+ * </tr>
+ * <tr>
+ * <td>{@value #KEY_URL}</td>
+ * <td>{@link StringLookup}</td>
+ * <td>{@link #urlStringLookup()}</td>
+ * <td>1.5</td>
+ * </tr>
+ * <tr>
+ * <td>{@value #KEY_SCRIPT}</td>
+ * <td>{@link StringLookup}</td>
+ * <td>{@link #scriptStringLookup()}</td>
+ * <td>1.5</td>
+ * </tr>
+ * </table>
+ *
  * <p>
- * We also provide functional lookups used as building blocks for other lookups.
+ * This class also provides functional lookups used as building blocks for other lookups.
  * <table>
  * <caption>Functional String Lookups</caption>
  * <tr>
@@ -237,7 +264,7 @@
      * </p>
      *
      * <pre>
-     * StringLookupFactory.INSTANCE.dateStringLookup().lookup("USER");
+     * StringLookupFactory.INSTANCE.environmentVariableStringLookup().lookup("USER");
      * </pre>
      * <p>
      * Using a {@link StringSubstitutor}:
@@ -383,6 +410,17 @@
     public static final String KEY_XML = "xml";
 
     /**
+     * Name of the system property used to determine the string lookups added by the
+     * {@link #addDefaultStringLookups(Map)} method. Use of this property is only required
+     * in cases where the set of default lookups must be modified. (See the class documentation
+     * for details.)
+     *
+     * @since 1.10
+     */
+    public static final String DEFAULT_STRING_LOOKUPS_PROPERTY =
+            "org.apache.commons.text.lookup.StringLookupFactory.defaultStringLookups";
+
+    /**
      * Clears any static resources.
      *
      * @since 1.5
@@ -392,6 +430,18 @@
     }
 
     /**
+     * Returns the given map if the input is non-null or an empty immutable map if the input is null.
+     *
+     * @param <K> the class of the map keys
+     * @param <V> the class of the map values
+     * @param map The map to test
+     * @return the given map if the input is non-null or an empty immutable map if the input is null.
+     */
+    static <K, V> Map<K, V> toMap(final Map<K, V> map) {
+        return map == null ? Collections.emptyMap() : map;
+    }
+
+    /**
      * No need to build instances for now.
      */
     private StringLookupFactory() {
@@ -399,19 +449,17 @@
     }
 
     /**
-     * Adds the {@link StringLookupFactory default lookups}.
+     * Adds the default string lookups for this class to {@code stringLookupMap}. The default string
+     * lookups are a set of built-in lookups added for convenience during string interpolation. The
+     * defaults may be configured using the {@value #DEFAULT_STRING_LOOKUPS_PROPERTY} system property.
+     * See the class documentation for details and a list of lookups.
      *
-     * @param stringLookupMap the map of string lookups.
+     * @param stringLookupMap the map of string lookups to edit.
      * @since 1.5
      */
     public void addDefaultStringLookups(final Map<String, StringLookup> stringLookupMap) {
         if (stringLookupMap != null) {
-            // "base64" is deprecated in favor of KEY_BASE64_DECODER.
-            stringLookupMap.put("base64", StringLookupFactory.INSTANCE_BASE64_DECODER);
-            for (final DefaultStringLookup stringLookup : DefaultStringLookup.values()) {
-                stringLookupMap.put(InterpolatorStringLookup.toKey(stringLookup.getKey()),
-                    stringLookup.getStringLookup());
-            }
+            stringLookupMap.putAll(DefaultStringLookupsHolder.INSTANCE.getDefaultStringLookups());
         }
     }
 
@@ -435,7 +483,7 @@
      * The above examples convert {@code "SGVsbG9Xb3JsZCE="} to {@code "HelloWorld!"}.
      * </p>
      *
-     * @return The DateStringLookup singleton instance.
+     * @return The Base64DecoderStringLookup singleton instance.
      * @since 1.5
      */
     public StringLookup base64DecoderStringLookup() {
@@ -462,7 +510,7 @@
      * The above examples convert {@code } to {@code "SGVsbG9Xb3JsZCE="}.
      * </p>
      *
-     * @return The DateStringLookup singleton instance.
+     * @return The Base64EncoderStringLookup singleton instance.
      * @since 1.6
      */
     public StringLookup base64EncoderStringLookup() {
@@ -489,7 +537,7 @@
      * The above examples convert {@code "SGVsbG9Xb3JsZCE="} to {@code "HelloWorld!"}.
      * </p>
      *
-     * @return The DateStringLookup singleton instance.
+     * @return The Base64DecoderStringLookup singleton instance.
      * @since 1.5
      * @deprecated Use {@link #base64DecoderStringLookup()}.
      */
@@ -543,7 +591,7 @@
      * The above examples convert {@code java.awt.event.KeyEvent.VK_ESCAPE} to {@code "27"}.
      * </p>
      *
-     * @return The DateStringLookup singleton instance.
+     * @return The ConstantStringLookup singleton instance.
      * @since 1.5
      */
     public StringLookup constantStringLookup() {
@@ -593,17 +641,24 @@
      * StringLookupFactory.INSTANCE.dnsStringLookup().lookup("address|apache.org");
      * </pre>
      * <p>
-     * Using a {@link StringSubstitutor}:
+     * When used through a {@link StringSubstitutor}, this lookup must either be added programmatically
+     * (as below) or enabled as a default lookup using the {@value #DEFAULT_STRING_LOOKUPS_PROPERTY} system property
+     * (see class documentation).
      * </p>
      *
      * <pre>
-     * StringSubstitutor.createInterpolator().replace("... ${dns:address|apache.org} ..."));
+     * Map&lt;String, StringLookup&gt; lookupMap = new HashMap&lt;&gt;();
+     * lookupMap.put("dns", StringLookupFactory.INSTANCE.dnsStringLookup());
+     *
+     * StringLookup variableResolver = StringLookupFactory.INSTANCE.interpolatorStringLookup(lookupMap, null, false);
+     *
+     * new StringSubstitutor(variableResolver).replace("... ${dns:address|apache.org} ...");
      * </pre>
      * <p>
-     * The above examples convert {@code "address|apache.org"} to {@code "95.216.24.32} (or {@code "40.79.78.1"}).
+     * The above examples convert {@code "address|apache.org"} to the IP address of {@code apache.org}.
      * </p>
      *
-     * @return the DateStringLookup singleton instance.
+     * @return the DnsStringLookup singleton instance.
      * @since 1.8
      */
     public StringLookup dnsStringLookup() {
@@ -618,7 +673,7 @@
      * </p>
      *
      * <pre>
-     * StringLookupFactory.INSTANCE.dateStringLookup().lookup("USER");
+     * StringLookupFactory.INSTANCE.environmentVariableStringLookup().lookup("USER");
      * </pre>
      * <p>
      * Using a {@link StringSubstitutor}:
@@ -679,11 +734,10 @@
     }
 
     /**
-     * Returns a new InterpolatorStringLookup using the {@link StringLookupFactory default lookups}.
+     * Returns a {@link InterpolatorStringLookup} containing the configured
+     * {@link #addDefaultStringLookups(Map) default lookups}. See the class documentation for
+     * details on how these defaults are configured.
      * <p>
-     * The lookups available to an interpolator are defined in
-     * </p>
-     * <p>
      * Using a {@link StringLookup} from the {@link StringLookupFactory}:
      * </p>
      *
@@ -701,22 +755,22 @@
      * The above examples convert {@code "${sys:os.name}, ${env:USER}"} to the OS name and Linux user name.
      * </p>
      *
-     * @return a new InterpolatorStringLookup.
+     * @return the default {@link InterpolatorStringLookup}.
      */
     public StringLookup interpolatorStringLookup() {
         return InterpolatorStringLookup.INSTANCE;
     }
 
     /**
-     * Returns a new InterpolatorStringLookup using the {@link StringLookupFactory default lookups}.
-     * <p>
-     * If {@code addDefaultLookups} is true, the following lookups are used in addition to the ones provided in
-     * {@code stringLookupMap}:
-     * </p>
+     * Returns a new InterpolatorStringLookup. If {@code addDefaultLookups} is {@code true}, the configured
+     * {@link #addDefaultStringLookups(Map) default lookups} are included in addition to the ones
+     * provided in {@code stringLookupMap}. (See the class documentation for details on how default lookups
+     * are configured.)
      *
      * @param stringLookupMap the map of string lookups.
-     * @param defaultStringLookup the default string lookup.
-     * @param addDefaultLookups whether to use lookups as described above.
+     * @param defaultStringLookup the default string lookup; this lookup is used when a variable cannot be
+     *      resolved using the lookups in {@code stringLookupMap} or the configured default lookups (if enabled)
+     * @param addDefaultLookups whether to use default lookups as described above.
      * @return a new InterpolatorStringLookup.
      * @since 1.4
      */
@@ -726,7 +780,9 @@
     }
 
     /**
-     * Returns a new InterpolatorStringLookup using the {@link StringLookupFactory default lookups}.
+     * Returns a new InterpolatorStringLookup using the given key-value pairs and the configured
+     * {@link #addDefaultStringLookups(Map) default lookups} to resolve variables. (See the class
+     * documentation for details on how default lookups are configured.)
      *
      * @param <V> the value type the default string lookup's map.
      * @param map the default map for string lookups.
@@ -737,7 +793,9 @@
     }
 
     /**
-     * Returns a new InterpolatorStringLookup using the {@link StringLookupFactory default lookups}.
+     * Returns a new InterpolatorStringLookup using the given lookup and the configured
+     * {@link #addDefaultStringLookups(Map) default lookups} to resolve variables. (See the class
+     * documentation for details on how default lookups are configured.)
      *
      * @param defaultStringLookup the default string lookup.
      * @return a new InterpolatorStringLookup.
@@ -939,7 +997,9 @@
     }
 
     /**
-     * Returns the ScriptStringLookup singleton instance.
+     * Returns the ScriptStringLookup singleton instance. NOTE: This lookup is not included
+     * as a {@link #addDefaultStringLookups(Map) default lookup} unless explicitly enabled. See
+     * the class level documentation for details.
      * <p>
      * Looks up the value for the key in the format "ScriptEngineName:Script".
      * </p>
@@ -954,11 +1014,18 @@
      * StringLookupFactory.INSTANCE.scriptStringLookup().lookup("javascript:3 + 4");
      * </pre>
      * <p>
-     * Using a {@link StringSubstitutor}:
+     * When used through a {@link StringSubstitutor}, this lookup must either be added programmatically
+     * (as below) or enabled as a default lookup using the {@value #DEFAULT_STRING_LOOKUPS_PROPERTY} system property
+     * (see class documentation).
      * </p>
      *
      * <pre>
-     * StringSubstitutor.createInterpolator().replace("... ${javascript:3 + 4} ..."));
+     * Map&lt;String, StringLookup&gt; lookupMap = new HashMap&lt;&gt;();
+     * lookupMap.put("script", StringLookupFactory.INSTANCE.scriptStringLookup());
+     *
+     * StringLookup variableResolver = StringLookupFactory.INSTANCE.interpolatorStringLookup(lookupMap, null, false);
+     *
+     * String value = new StringSubstitutor(variableResolver).replace("${script:javascript:3 + 4}");
      * </pre>
      * <p>
      * The above examples convert {@code "javascript:3 + 4"} to {@code "7"}.
@@ -1065,7 +1132,9 @@
     }
 
     /**
-     * Returns the UrlStringLookup singleton instance.
+     * Returns the UrlStringLookup singleton instance. This lookup is not included
+     * as a {@link #addDefaultStringLookups(Map) default lookup} unless explicitly enabled. See
+     * the class level documentation for details.
      * <p>
      * Looks up the value for the key in the format "CharsetName:URL".
      * </p>
@@ -1084,11 +1153,18 @@
      * StringLookupFactory.INSTANCE.urlStringLookup().lookup("UTF-8:https://www.apache.org");
      * </pre>
      * <p>
-     * Using a {@link StringSubstitutor}:
+     * When used through a {@link StringSubstitutor}, this lookup must either be added programmatically
+     * (as below) or enabled as a default lookup using the {@value #DEFAULT_STRING_LOOKUPS_PROPERTY} system property
+     * (see class documentation).
      * </p>
      *
      * <pre>
-     * StringSubstitutor.createInterpolator().replace("... ${url:UTF-8:https://www.apache.org} ..."));
+     * Map&lt;String, StringLookup&gt; lookupMap = new HashMap&lt;&gt;();
+     * lookupMap.put("url", StringLookupFactory.INSTANCE.urlStringLookup());
+     *
+     * StringLookup variableResolver = StringLookupFactory.INSTANCE.interpolatorStringLookup(lookupMap, null, false);
+     *
+     * String value = new StringSubstitutor(variableResolver).replace("${url:UTF-8:https://www.apache.org}");
      * </pre>
      * <p>
      * The above examples convert {@code "UTF-8:https://www.apache.org"} to the contents of that page.
@@ -1135,4 +1211,109 @@
         return XmlStringLookup.INSTANCE;
     }
 
+    /**
+     * Get a string suitable for use as a key in the string lookup map.
+     * @param key string to convert to a string lookup map key
+     * @return string lookup map key
+     */
+    static String toKey(final String key) {
+        return key.toLowerCase(Locale.ROOT);
+    }
+
+    /**
+     * Internal class used to construct the default {@link StringLookup} map used by
+     * {@link StringLookupFactory#addDefaultStringLookups(Map)}.
+     */
+    static final class DefaultStringLookupsHolder {
+
+        /** Singleton instance, initialized with the system properties. */
+        static final DefaultStringLookupsHolder INSTANCE = new DefaultStringLookupsHolder(System.getProperties());
+
+        /** Default string lookup map. */
+        private final Map<String, StringLookup> defaultStringLookups;
+
+        /**
+         * Construct a new instance initialized with the given properties.
+         * @param props initialization properties
+         */
+        DefaultStringLookupsHolder(final Properties props) {
+            final Map<String, StringLookup> lookups =
+                    props.containsKey(StringLookupFactory.DEFAULT_STRING_LOOKUPS_PROPERTY)
+                        ? parseStringLookups(props.getProperty(StringLookupFactory.DEFAULT_STRING_LOOKUPS_PROPERTY))
+                        : createDefaultStringLookups();
+
+            defaultStringLookups = Collections.unmodifiableMap(lookups);
+        }
+
+        /**
+         * Get the default string lookups map.
+         * @return default string lookups map
+         */
+        Map<String, StringLookup> getDefaultStringLookups() {
+            return defaultStringLookups;
+        }
+
+        /**
+         * Create the lookup map used when the user has requested no customization.
+         * @return default lookup map
+         */
+        private static Map<String, StringLookup> createDefaultStringLookups() {
+            final Map<String, StringLookup> lookupMap = new HashMap<>();
+
+            addLookup(DefaultStringLookup.BASE64_DECODER, lookupMap);
+            addLookup(DefaultStringLookup.BASE64_ENCODER, lookupMap);
+            addLookup(DefaultStringLookup.CONST, lookupMap);
+            addLookup(DefaultStringLookup.DATE, lookupMap);
+            addLookup(DefaultStringLookup.ENVIRONMENT, lookupMap);
+            addLookup(DefaultStringLookup.FILE, lookupMap);
+            addLookup(DefaultStringLookup.JAVA, lookupMap);
+            addLookup(DefaultStringLookup.LOCAL_HOST, lookupMap);
+            addLookup(DefaultStringLookup.PROPERTIES, lookupMap);
+            addLookup(DefaultStringLookup.RESOURCE_BUNDLE, lookupMap);
+            addLookup(DefaultStringLookup.SYSTEM_PROPERTIES, lookupMap);
+            addLookup(DefaultStringLookup.URL_DECODER, lookupMap);
+            addLookup(DefaultStringLookup.URL_ENCODER, lookupMap);
+            addLookup(DefaultStringLookup.XML, lookupMap);
+
+            return lookupMap;
+        }
+
+        /**
+         * Construct a lookup map by parsing the given string. The string is expected to contain
+         * comma or space-separated names of values from the {@link DefaultStringLookup} enum. If
+         * the given string is null or empty, an empty map is returned.
+         * @param str string to parse; may be null or empty
+         * @return lookup map parsed from the given string
+         */
+        private static Map<String, StringLookup> parseStringLookups(final String str) {
+            final Map<String, StringLookup> lookupMap = new HashMap<>();
+
+            try {
+                for (final String lookupName : str.split("[\\s,]+")) {
+                    if (!lookupName.isEmpty()) {
+                        addLookup(DefaultStringLookup.valueOf(lookupName.toUpperCase()), lookupMap);
+                    }
+                }
+            } catch (IllegalArgumentException exc) {
+                throw new IllegalArgumentException("Invalid default string lookups definition: " + str, exc);
+            }
+
+            return lookupMap;
+        }
+
+        /**
+         * Add the key and string lookup from {@code lookup} to {@code map}, also adding any additional
+         * key aliases if needed. Keys are normalized using the {@link #toKey(String)} method.
+         * @param lookup lookup to add
+         * @param map map to add to
+         */
+        private static void addLookup(final DefaultStringLookup lookup, final Map<String, StringLookup> map) {
+            map.put(toKey(lookup.getKey()), lookup.getStringLookup());
+
+            if (DefaultStringLookup.BASE64_DECODER.equals(lookup)) {
+                // "base64" is deprecated in favor of KEY_BASE64_DECODER.
+                map.put(toKey("base64"), lookup.getStringLookup());
+            }
+        }
+    }
 }
--- a/src/test/java/org/apache/commons/text/StringSubstitutorWithInterpolatorStringLookupTest.java
+++ b/src/test/java/org/apache/commons/text/StringSubstitutorWithInterpolatorStringLookupTest.java
@@ -24,6 +24,7 @@
 import java.util.HashMap;
 import java.util.Map;
 
+import org.apache.commons.text.lookup.DefaultStringLookup;
 import org.apache.commons.text.lookup.StringLookup;
 import org.apache.commons.text.lookup.StringLookupFactory;
 import org.junit.jupiter.api.Assertions;
@@ -41,7 +42,7 @@
         final String value = "value";
         final Map<String, String> map = new HashMap<>();
         map.put(key, value);
-        final StringLookup mapStringLookup = StringLookupFactory.INSTANCE.functionStringLookup(k -> map.get(k));
+        final StringLookup mapStringLookup = StringLookupFactory.INSTANCE.functionStringLookup(map::get);
         final Map<String, StringLookup> stringLookupMap = new HashMap<>();
         stringLookupMap.put("customLookup", mapStringLookup);
         final StringSubstitutor strSubst = new StringSubstitutor(
@@ -91,27 +92,21 @@
         // Used to cut and paste into the docs.
         // @formatter:off
         final StringSubstitutor interpolator = StringSubstitutor.createInterpolator();
-        interpolator.setEnableSubstitutionInVariables(true); // Allows for nested $'s.
         final String text = interpolator.replace(
                 "Base64 Decoder:        ${base64Decoder:SGVsbG9Xb3JsZCE=}\n"
               + "Base64 Encoder:        ${base64Encoder:HelloWorld!}\n"
               + "Java Constant:         ${const:java.awt.event.KeyEvent.VK_ESCAPE}\n"
               + "Date:                  ${date:yyyy-MM-dd}\n"
-              + "DNS:                   ${dns:address|apache.org}\n"
               + "Environment Variable:  ${env:USERNAME}\n"
-              + "File Content:          ${file:UTF-8:src/test/resources/document.properties}\n"
+              + "File Content:          ${file:UTF-8:src/test/resources/org/apache/commons/text/document.properties}\n"
               + "Java:                  ${java:version}\n"
               + "Localhost:             ${localhost:canonical-name}\n"
-              + "Properties File:       ${properties:src/test/resources/document.properties::mykey}\n"
-              + "Resource Bundle:       ${resourceBundle:org.example.testResourceBundleLookup:mykey}\n"
-              + "Script:                ${script:javascript:3 + 4}\n"
+              + "Properties File:       ${properties:src/test/resources/org/apache/commons/text/document.properties::mykey}\n"
+              + "Resource Bundle:       ${resourceBundle:org.apache.commons.text.example.testResourceBundleLookup:mykey}\n"
               + "System Property:       ${sys:user.dir}\n"
               + "URL Decoder:           ${urlDecoder:Hello%20World%21}\n"
               + "URL Encoder:           ${urlEncoder:Hello World!}\n"
-              + "URL Content (HTTP):    ${url:UTF-8:http://www.apache.org}\n"
-              + "URL Content (HTTPS):   ${url:UTF-8:https://www.apache.org}\n"
-              + "URL Content (File):    ${url:UTF-8:file:///${sys:user.dir}/src/test/resources/document.properties}\n"
-              + "XML XPath:             ${xml:src/test/resources/document.xml:/root/path/to/node}\n"
+              + "XML XPath:             ${xml:src/test/resources/org/apache/commons/text/document.xml:/root/path/to/node}\n"
         );
         // @formatter:on
         Assertions.assertNotNull(text);
@@ -120,13 +115,12 @@
         Assertions.assertFalse(text.contains("${base64Encoder:HelloWorld!}"));
         Assertions.assertFalse(text.contains("${urlDecoder:Hello%20World%21}"));
         Assertions.assertFalse(text.contains("${urlEncoder:Hello World!}"));
-        Assertions.assertFalse(text.contains("${resourceBundle:org.example.testResourceBundleLookup:mykey}"));
-        // System.out.println(text);
+        Assertions.assertFalse(text.contains("${resourceBundle:org.apache.commons.text.example.testResourceBundleLookup:mykey}"));
     }
     @Test
     public void testDefaultValueForMissingKeyInResourceBundle() {
         final StringLookup interpolatorStringLookup = StringLookupFactory.INSTANCE.interpolatorStringLookup(
-            StringLookupFactory.INSTANCE.resourceBundleStringLookup("org.example.testResourceBundleLookup"));
+            StringLookupFactory.INSTANCE.resourceBundleStringLookup("org.apache.commons.text.example.testResourceBundleLookup"));
         assertEquals("${missingKey:-defaultValue}", interpolatorStringLookup.lookup("keyWithMissingKey"));
         final StringSubstitutor stringSubstitutor = new StringSubstitutor(interpolatorStringLookup);
         // The following would throw a MissingResourceException before TEXT-165.
@@ -135,7 +129,8 @@
 
     @Test
     public void testDnsLookup() throws UnknownHostException {
-        final StringSubstitutor strSubst = StringSubstitutor.createInterpolator();
+        final StringSubstitutor strSubst =
+                new StringSubstitutor(createInterpolatorWithLookups(DefaultStringLookup.DNS));
         final String hostName = InetAddress.getLocalHost().getHostName();
         Assertions.assertEquals(InetAddress.getByName(hostName).getHostAddress(),
             strSubst.replace("${dns:" + hostName + "}"));
@@ -143,14 +138,16 @@
 
     @Test
     public void testDnsLookupAddress() throws UnknownHostException {
-        final StringSubstitutor strSubst = StringSubstitutor.createInterpolator();
+        final StringSubstitutor strSubst =
+                new StringSubstitutor(createInterpolatorWithLookups(DefaultStringLookup.DNS));
         Assertions.assertEquals(InetAddress.getByName("apache.org").getHostAddress(),
             strSubst.replace("${dns:address|apache.org}"));
     }
 
     @Test
     public void testDnsLookupCanonicalName() throws UnknownHostException {
-        final StringSubstitutor strSubst = StringSubstitutor.createInterpolator();
+        final StringSubstitutor strSubst =
+                new StringSubstitutor(createInterpolatorWithLookups(DefaultStringLookup.DNS));
         final String address = InetAddress.getLocalHost().getHostAddress();
         final InetAddress inetAddress = InetAddress.getByName(address);
         Assertions.assertEquals(inetAddress.getCanonicalHostName(),
@@ -159,7 +156,8 @@
 
     @Test
     public void testDnsLookupName() throws UnknownHostException {
-        final StringSubstitutor strSubst = StringSubstitutor.createInterpolator();
+        final StringSubstitutor strSubst =
+                new StringSubstitutor(createInterpolatorWithLookups(DefaultStringLookup.DNS));
         final String address = InetAddress.getLocalHost().getHostAddress();
         final InetAddress inetAddress = InetAddress.getByName(address);
         Assertions.assertEquals(inetAddress.getHostName(), strSubst.replace("${dns:name|" + address + "}"));
@@ -167,7 +165,8 @@
 
     @Test
     public void testDnsLookupNameUntrimmed() throws UnknownHostException {
-        final StringSubstitutor strSubst = StringSubstitutor.createInterpolator();
+        final StringSubstitutor strSubst =
+                new StringSubstitutor(createInterpolatorWithLookups(DefaultStringLookup.DNS));
         final String address = InetAddress.getLocalHost().getHostAddress();
         final InetAddress inetAddress = InetAddress.getByName(address);
         Assertions.assertEquals(inetAddress.getHostName(), strSubst.replace("${dns:name| " + address + " }"));
@@ -175,20 +174,37 @@
 
     @Test
     public void testDnsLookupUnknown() {
-        final StringSubstitutor strSubst = StringSubstitutor.createInterpolator();
+        final StringSubstitutor strSubst =
+                new StringSubstitutor(createInterpolatorWithLookups(DefaultStringLookup.DNS));
         final String unknown = "${dns: u n k n o w n}";
         Assertions.assertEquals(unknown, strSubst.replace(unknown));
     }
 
     @Test
+    public void testDnsLookup_disabledByDefault() throws UnknownHostException {
+        final StringSubstitutor strSubst = StringSubstitutor.createInterpolator();
+        final String hostName = InetAddress.getLocalHost().getHostName();
+        final String input = "${dns:" + hostName + "}";
+        Assertions.assertEquals(input, strSubst.replace(input));
+    }
+
+    @Test
     public void testJavaScript() {
-        Assertions.assertEquals("Hello World!",
-                StringSubstitutor.createInterpolator().replace("${script:javascript:\"Hello World!\"}"));
-        Assertions.assertEquals("7",
-                StringSubstitutor.createInterpolator().replace("${script:javascript:3 + 4}"));
+        final StringSubstitutor strSubst =
+                new StringSubstitutor(createInterpolatorWithLookups(DefaultStringLookup.SCRIPT));
+
+        Assertions.assertEquals("Hello World!", strSubst.replace("${script:javascript:\"Hello World!\"}"));
+        Assertions.assertEquals("7", strSubst.replace("${script:javascript:3 + 4}"));
     }
 
     @Test
+    public void testJavaScript_disabledByDefault() {
+        final StringSubstitutor strSubst = StringSubstitutor.createInterpolator();
+
+        Assertions.assertEquals("${script:javascript:3 + 4}", strSubst.replace("${script:javascript:3 + 4}"));
+    }
+
+    @Test
     public void testLocalHostLookup_Address() throws UnknownHostException {
         final StringSubstitutor strSubst = StringSubstitutor.createInterpolator();
         Assertions.assertEquals(InetAddress.getLocalHost().getHostAddress(), strSubst.replace("${localhost:address}"));
@@ -227,4 +243,21 @@
         Assertions.assertEquals(System.getProperty(spKey), strSubst.replace("${sys:" + spKey + "}"));
     }
 
+    @Test
+    public void testSystemPropertyDefaultStringLookup() {
+        final StringSubstitutor strSubst = new StringSubstitutor(
+            StringLookupFactory.INSTANCE.interpolatorStringLookup(StringLookupFactory.INSTANCE.systemPropertyStringLookup()));
+        final String spKey = "user.name";
+        Assertions.assertEquals(System.getProperty(spKey), strSubst.replace("${" + spKey + "}"));
+        Assertions.assertEquals(System.getProperty(spKey), strSubst.replace("${sys:" + spKey + "}"));
+    }
+
+    private static StringLookup createInterpolatorWithLookups(final DefaultStringLookup... lookups) {
+        final Map<String, StringLookup> lookupMap = new HashMap<>();
+        for (final DefaultStringLookup lookup : lookups) {
+            lookupMap.put(lookup.getKey().toLowerCase(), lookup.getStringLookup());
+        }
+
+        return StringLookupFactory.INSTANCE.interpolatorStringLookup(lookupMap, null, false);
+    }
 }
--- a/src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java
+++ b/src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java
@@ -16,10 +16,11 @@
  */
 package org.apache.commons.text.lookup;
 
-import static org.junit.jupiter.api.Assertions.assertTrue;
-
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
 
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
@@ -30,24 +31,23 @@
 public class StringLookupFactoryTest {
 
     public static void assertDefaultKeys(final Map<String, StringLookup> stringLookupMap) {
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_BASE64_DECODER)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_BASE64_ENCODER)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_CONST)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_DATE)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_DNS)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_ENV)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_FILE)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_JAVA)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_LOCALHOST)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_PROPERTIES)));
-        assertTrue(
-            stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_RESOURCE_BUNDLE)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_SCRIPT)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_SYS)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_URL)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_URL_DECODER)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_URL_ENCODER)));
-        assertTrue(stringLookupMap.containsKey(InterpolatorStringLookup.toKey(StringLookupFactory.KEY_XML)));
+        // included
+        assertMappedLookups(stringLookupMap,
+                "base64",
+                StringLookupFactory.KEY_BASE64_DECODER,
+                StringLookupFactory.KEY_BASE64_ENCODER,
+                StringLookupFactory.KEY_CONST,
+                StringLookupFactory.KEY_DATE,
+                StringLookupFactory.KEY_ENV,
+                StringLookupFactory.KEY_FILE,
+                StringLookupFactory.KEY_JAVA,
+                StringLookupFactory.KEY_LOCALHOST,
+                StringLookupFactory.KEY_PROPERTIES,
+                StringLookupFactory.KEY_RESOURCE_BUNDLE,
+                StringLookupFactory.KEY_SYS,
+                StringLookupFactory.KEY_URL_DECODER,
+                StringLookupFactory.KEY_URL_ENCODER,
+                StringLookupFactory.KEY_XML);
     }
 
     @Test
@@ -89,5 +89,143 @@
         Assertions.assertSame(UrlEncoderStringLookup.INSTANCE, stringLookupFactory.urlEncoderStringLookup());
         Assertions.assertSame(UrlStringLookup.INSTANCE, stringLookupFactory.urlStringLookup());
         Assertions.assertSame(XmlStringLookup.INSTANCE, stringLookupFactory.xmlStringLookup());
+    }
+
+    @Test
+    public void testDefaultStringLookupsHolder_lookupsPropertyNotPresent() {
+        checkDefaultStringLookupsHolder(new Properties(),
+                "base64",
+                StringLookupFactory.KEY_BASE64_DECODER,
+                StringLookupFactory.KEY_BASE64_ENCODER,
+                StringLookupFactory.KEY_CONST,
+                StringLookupFactory.KEY_DATE,
+                StringLookupFactory.KEY_ENV,
+                StringLookupFactory.KEY_FILE,
+                StringLookupFactory.KEY_JAVA,
+                StringLookupFactory.KEY_LOCALHOST,
+                StringLookupFactory.KEY_PROPERTIES,
+                StringLookupFactory.KEY_RESOURCE_BUNDLE,
+                StringLookupFactory.KEY_SYS,
+                StringLookupFactory.KEY_URL_DECODER,
+                StringLookupFactory.KEY_URL_ENCODER,
+                StringLookupFactory.KEY_XML);
+    }
+
+    @Test
+    public void testDefaultStringLookupsHolder_lookupsPropertyEmptyAndBlank() {
+        final Properties propsWithNull = new Properties();
+        propsWithNull.setProperty(StringLookupFactory.DEFAULT_STRING_LOOKUPS_PROPERTY, "");
+
+        checkDefaultStringLookupsHolder(propsWithNull);
+
+        final Properties propsWithBlank = new Properties();
+        propsWithBlank.setProperty(StringLookupFactory.DEFAULT_STRING_LOOKUPS_PROPERTY, " ");
+
+        checkDefaultStringLookupsHolder(propsWithBlank);
+    }
+
+    @Test
+    public void testDefaultStringLookupsHolder_givenSingleLookup() {
+        final Properties props = new Properties();
+        props.setProperty(StringLookupFactory.DEFAULT_STRING_LOOKUPS_PROPERTY, "base64_encoder");
+
+        checkDefaultStringLookupsHolder(props,
+                "base64",
+                StringLookupFactory.KEY_BASE64_ENCODER);
+    }
+
+    @Test
+    public void testDefaultStringLookupsHolder_givenSingleLookup_weirdString() {
+        final Properties props = new Properties();
+        props.setProperty(StringLookupFactory.DEFAULT_STRING_LOOKUPS_PROPERTY, " \n \t  ,, DnS , , ");
+
+        checkDefaultStringLookupsHolder(props, StringLookupFactory.KEY_DNS);
+    }
+
+    @Test
+    public void testDefaultStringLookupsHolder_multipleLookups() {
+        final Properties props = new Properties();
+        props.setProperty(StringLookupFactory.DEFAULT_STRING_LOOKUPS_PROPERTY, "dns, url script ");
+
+        checkDefaultStringLookupsHolder(props,
+                StringLookupFactory.KEY_DNS,
+                StringLookupFactory.KEY_URL,
+                StringLookupFactory.KEY_SCRIPT);
+    }
+
+    @Test
+    public void testDefaultStringLookupsHolder_allLookups() {
+        final Properties props = new Properties();
+        props.setProperty(StringLookupFactory.DEFAULT_STRING_LOOKUPS_PROPERTY,
+                "BASE64_DECODER BASE64_ENCODER const, date, dns, environment "
+                + "file ,java, local_host properties, resource_bundle,script,system_properties "
+                + "url url_decoder  , url_encoder, xml");
+
+        checkDefaultStringLookupsHolder(props,
+                "base64",
+                StringLookupFactory.KEY_BASE64_DECODER,
+                StringLookupFactory.KEY_BASE64_ENCODER,
+                StringLookupFactory.KEY_CONST,
+                StringLookupFactory.KEY_DATE,
+                StringLookupFactory.KEY_ENV,
+                StringLookupFactory.KEY_FILE,
+                StringLookupFactory.KEY_JAVA,
+                StringLookupFactory.KEY_LOCALHOST,
+                StringLookupFactory.KEY_PROPERTIES,
+                StringLookupFactory.KEY_RESOURCE_BUNDLE,
+                StringLookupFactory.KEY_SYS,
+                StringLookupFactory.KEY_URL_DECODER,
+                StringLookupFactory.KEY_URL_ENCODER,
+                StringLookupFactory.KEY_XML,
+
+                StringLookupFactory.KEY_DNS,
+                StringLookupFactory.KEY_URL,
+                StringLookupFactory.KEY_SCRIPT);
+    }
+
+    @Test
+    public void testDefaultStringLookupsHolder_invalidLookupsDefinition() {
+        final Properties props = new Properties();
+        props.setProperty(StringLookupFactory.DEFAULT_STRING_LOOKUPS_PROPERTY, "base64_encoder nope");
+
+        final Exception exc = Assertions.assertThrows(IllegalArgumentException.class,
+                () -> new StringLookupFactory.DefaultStringLookupsHolder(props));
+        Assertions.assertEquals("Invalid default string lookups definition: base64_encoder nope", exc.getMessage());
+    }
+
+    private static void checkDefaultStringLookupsHolder(final Properties props, final String... keys) {
+        final StringLookupFactory.DefaultStringLookupsHolder holder =
+                new StringLookupFactory.DefaultStringLookupsHolder(props);
+
+        final Map<String, StringLookup> lookupMap = holder.getDefaultStringLookups();
+
+        assertMappedLookups(lookupMap, keys);
+    }
+
+    private static void assertMappedLookups(final Map<String, StringLookup> lookupMap, final String... keys) {
+        final Set<String> remainingKeys = new HashSet<>(lookupMap.keySet());
+
+        for (final String key : keys) {
+            final String normalizedKey = StringLookupFactory.toKey(key);
+            Assertions.assertNotNull(normalizedKey, () -> "Expected map to contain string lookup for key " + key);
+
+            remainingKeys.remove(normalizedKey);
+        }
+
+        Assertions.assertTrue(remainingKeys.isEmpty(), () -> "Unexpected keys in lookup map: " + remainingKeys);
+    }
+
+    /**
+     * Main method used to verify the default string lookups resolved during JVM execution.
+     * @param args
+     */
+    public static void main(final String[] args) {
+        final Map<String, StringLookup> lookupMap = new HashMap<>();
+        StringLookupFactory.INSTANCE.addDefaultStringLookups(lookupMap);
+
+        System.out.println("Default string lookups");
+        for (final String key : lookupMap.keySet()) {
+            System.out.println("- " + key);
+        }
     }
 }
