diff --git a/dist/defu.cjs b/dist/defu.cjs
index 72b30c4..3659320 100644
--- a/dist/defu.cjs
+++ b/dist/defu.cjs
@@ -23,7 +23,7 @@ function _defu(baseObject, defaults, namespace = ".", merger) {
   if (!isPlainObject(defaults)) {
     return _defu(baseObject, {}, namespace, merger);
   }
-  const object = Object.assign({}, defaults);
+  const object = { ...defaults };
   for (const key in baseObject) {
     if (key === "__proto__" || key === "constructor") {
       continue;
diff --git a/dist/defu.mjs b/dist/defu.mjs
index 889b83a..d1e5145 100644
--- a/dist/defu.mjs
+++ b/dist/defu.mjs
@@ -19,7 +19,7 @@ function _defu(baseObject, defaults, namespace = ".", merger) {
   if (!isPlainObject(defaults)) {
     return _defu(baseObject, {}, namespace, merger);
   }
-  const object = Object.assign({}, defaults);
+  const object = { ...defaults };
   for (const key in baseObject) {
     if (key === "__proto__" || key === "constructor") {
       continue;
