diff --git a/querydsl-core/src/main/java/com/querydsl/core/types/dsl/PathBuilderValidator.java b/querydsl-core/src/main/java/com/querydsl/core/types/dsl/PathBuilderValidator.java
index ab35ee0..26908c4 100644
--- a/querydsl-core/src/main/java/com/querydsl/core/types/dsl/PathBuilderValidator.java
+++ b/querydsl-core/src/main/java/com/querydsl/core/types/dsl/PathBuilderValidator.java
@@ -41,6 +41,9 @@ public interface PathBuilderValidator extends Serializable {
     PathBuilderValidator DEFAULT = new PathBuilderValidator() {
         @Override
         public Class<?> validate(Class<?> parent, String property, Class<?> propertyType) {
+            if (property.contains(" ")) {
+                throw new IllegalStateException("Unsafe due to CVE-2024-49203");
+            }
             return propertyType;
         }
     };
diff --git a/querydsl-core/src/test/java/com/querydsl/core/types/dsl/PathBuilderTest.java b/querydsl-core/src/test/java/com/querydsl/core/types/dsl/PathBuilderTest.java
index 5402088..089ac82 100644
--- a/querydsl-core/src/test/java/com/querydsl/core/types/dsl/PathBuilderTest.java
+++ b/querydsl-core/src/test/java/com/querydsl/core/types/dsl/PathBuilderTest.java
@@ -15,6 +15,8 @@ package com.querydsl.core.types.dsl;
 
 import static org.junit.Assert.*;
 
+import com.querydsl.core.types.Order;
+import com.querydsl.core.types.OrderSpecifier;
 import java.sql.Time;
 import java.util.Date;
 import java.util.Map;
@@ -130,4 +132,19 @@ public class PathBuilderTest {
             assertEquals(String.class, entity.get(pathName, Object.class).getType());
     }
 
+
+    @Test
+    public void order_HQL_injection() {
+        String orderBy = "breed";
+        PathBuilder<User> pathBuilder = new PathBuilder<User>(User.class, "entity");
+        new OrderSpecifier(Order.ASC, pathBuilder.get(orderBy));
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void unsafe_order_HQL_injection() {
+        // CVE-2024-49203
+        String orderBy = "test.name INTERSECT SELECT t FROM Test t WHERE 1=1 ORDER BY t.id";
+        PathBuilder<User> pathBuilder = new PathBuilder<User>(User.class, "entity");
+        new OrderSpecifier(Order.ASC, pathBuilder.get(orderBy));
+    }
 }
