--- a/index.js
+++ b/index.js
@@ -68,12 +68,15 @@ module.exports = function (args, opts) {
 
     function setKey (obj, keys, value) {
         var o = obj;
-        keys.slice(0,-1).forEach(function (key) {
+        for (var i = 0; i < keys.length - 1; i++) {
+            var key = keys[i];
+            if (key === '__proto__' || isConstructorOrProto(o, key)) return;
             if (o[key] === undefined) o[key] = {};
             o = o[key];
-        });
+        }
 
         var key = keys[keys.length - 1];
+        if (key === '__proto__' || isConstructorOrProto(o, key)) return;
         if (o[key] === undefined || flags.bools[key] || typeof o[key] === 'boolean') {
             o[key] = value;
         }
@@ -234,3 +237,6 @@ function isNumber (x) {
     return /^[-+]?(?:\d+(?:\.\d*)?|\.\d+)(e[-+]?\d+)?$/.test(x);
 }
 
+function isConstructorOrProto(obj, key) {
+    return key === 'constructor' && typeof obj[key] === 'function';
+}
