diff --git a/src/main/java/org/bouncycastle/asn1/ASN1ObjectIdentifier.java b/src/main/java/org/bouncycastle/asn1/ASN1ObjectIdentifier.java
--- a/src/main/java/org/bouncycastle/asn1/ASN1ObjectIdentifier.java	2026-06-03 14:22:17.024475731 +0530
+++ b/src/main/java/org/bouncycastle/asn1/ASN1ObjectIdentifier.java	2026-06-03 14:22:17.026260009 +0530
@@ -36,9 +36,14 @@
             return (ASN1ObjectIdentifier)obj;
         }
 
-        if (obj instanceof ASN1Encodable && ((ASN1Encodable)obj).toASN1Primitive() instanceof ASN1ObjectIdentifier)
+        if (obj instanceof ASN1Encodable)
         {
-            return (ASN1ObjectIdentifier)((ASN1Encodable)obj).toASN1Primitive();
+            ASN1Primitive primitive = ((ASN1Encodable)obj).toASN1Primitive();
+
+            if (primitive instanceof ASN1ObjectIdentifier)
+            {
+                return (ASN1ObjectIdentifier)primitive;
+            }
         }
 
         if (obj instanceof byte[])
@@ -83,11 +88,48 @@
         }
     }
 
+    /**
+     * Implementation limit on the length of the contents octets for an Object Identifier.
+     * <p/>
+     * We adopt the same value used by OpenJDK. In theory there is no limit on the length of the contents, or
+     * the number of subidentifiers, or the length of individual subidentifiers. In practice, supporting
+     * arbitrary lengths can lead to issues, e.g. denial-of-service attacks when attempting to convert a
+     * parsed value to its (decimal) string form.
+     */
+    private static final int MAX_CONTENTS_LENGTH = 4096;
+    private static final int MAX_IDENTIFIER_LENGTH = MAX_CONTENTS_LENGTH * 4 + 1;
+
     private static final long LONG_LIMIT = (Long.MAX_VALUE >> 7) - 0x7f;
 
+    static void checkContentsLength(int contentsLength)
+    {
+        if (contentsLength > MAX_CONTENTS_LENGTH)
+        {
+            throw new IllegalArgumentException("exceeded OID contents length limit");
+        }
+    }
+
+    static void checkIdentifier(String identifier)
+    {
+        if (identifier == null)
+        {
+            throw new NullPointerException("'identifier' cannot be null");
+        }
+        if (identifier.length() > MAX_IDENTIFIER_LENGTH)
+        {
+            throw new IllegalArgumentException("exceeded OID contents length limit");
+        }
+        if (!isValidIdentifier(identifier))
+        {
+            throw new IllegalArgumentException("string " + identifier + " not a valid OID");
+        }
+    }
+
     ASN1ObjectIdentifier(
         byte[] bytes)
     {
+        checkContentsLength(bytes.length);
+
         StringBuffer objId = new StringBuffer();
         long value = 0;
         BigInteger bigValue = null;
@@ -170,14 +212,7 @@
     public ASN1ObjectIdentifier(
         String identifier)
     {
-        if (identifier == null)
-        {
-            throw new IllegalArgumentException("'identifier' cannot be null");
-        }
-        if (!isValidIdentifier(identifier))
-        {
-            throw new IllegalArgumentException("string " + identifier + " not an OID");
-        }
+        checkIdentifier(identifier);
 
         this.identifier = identifier;
     }
@@ -196,6 +231,10 @@
         }
 
         this.identifier = oid.getId() + "." + branchID;
+        if (identifier.length() > MAX_IDENTIFIER_LENGTH)
+        {
+            throw new IllegalArgumentException("exceeded OID contents length limit");
+        }
     }
 
     /**
@@ -365,35 +404,40 @@
     private static boolean isValidBranchID(
         String branchID, int start)
     {
-        boolean periodAllowed = false;
+        int digitCount = 0;
 
         int pos = branchID.length();
         while (--pos >= start)
         {
             char ch = branchID.charAt(pos);
 
-            // TODO Leading zeroes?
-            if ('0' <= ch && ch <= '9')
-            {
-                periodAllowed = true;
-                continue;
-            }
-
             if (ch == '.')
             {
-                if (!periodAllowed)
+                if (0 == digitCount
+                    || (digitCount > 1 && branchID.charAt(pos + 1) == '0'))
                 {
                     return false;
                 }
 
-                periodAllowed = false;
-                continue;
+                digitCount = 0;
+            }
+            else if ('0' <= ch && ch <= '9')
+            {
+                ++digitCount;
             }
+            else
+            {
+                return false;
+            }
+        }
 
+        if (0 == digitCount
+            || (digitCount > 1 && branchID.charAt(pos + 1) == '0'))
+        {
             return false;
         }
 
-        return periodAllowed;
+        return true;
     }
 
     private static boolean isValidIdentifier(
@@ -477,4 +521,4 @@
         }
         return oid;
     }
-}
+}
\ No newline at end of file

diff --git a/src/main/java/META-INF/versions/9/module-info.java b/src/main/java/META-INF/versions/9/module-info.java
deleted file mode 100644
index 0fd13f8..0000000
--- a/src/main/java/META-INF/versions/9/module-info.java
+++ /dev/null
@@ -1,76 +0,0 @@
-module org.bouncycastle.fips.core
-{
-    provides java.security.Provider with org.bouncycastle.jcajce.provider.BouncyCastleFipsProvider;
-
-    exports org.bouncycastle;
-    exports org.bouncycastle.asn1;
-    exports org.bouncycastle.asn1.anssi;
-    exports org.bouncycastle.asn1.bc;
-    exports org.bouncycastle.asn1.bsi;
-    exports org.bouncycastle.asn1.cmp;
-    exports org.bouncycastle.asn1.cms;
-    exports org.bouncycastle.asn1.cms.ecc;
-    exports org.bouncycastle.asn1.crmf;
-    exports org.bouncycastle.asn1.cryptopro;
-    exports org.bouncycastle.asn1.dvcs;
-    exports org.bouncycastle.asn1.eac;
-    exports org.bouncycastle.asn1.edec;
-    exports org.bouncycastle.asn1.esf;
-    exports org.bouncycastle.asn1.ess;
-    exports org.bouncycastle.asn1.gm;
-    exports org.bouncycastle.asn1.gnu;
-    exports org.bouncycastle.asn1.iana;
-    exports org.bouncycastle.asn1.icao;
-    exports org.bouncycastle.asn1.isismtt;
-    exports org.bouncycastle.asn1.isismtt.ocsp;
-    exports org.bouncycastle.asn1.isismtt.x509;
-    exports org.bouncycastle.asn1.iso;
-    exports org.bouncycastle.asn1.kisa;
-    exports org.bouncycastle.asn1.microsoft;
-    exports org.bouncycastle.asn1.misc;
-    exports org.bouncycastle.asn1.mozilla;
-    exports org.bouncycastle.asn1.nist;
-    exports org.bouncycastle.asn1.nsri;
-    exports org.bouncycastle.asn1.ntt;
-    exports org.bouncycastle.asn1.ocsp;
-    exports org.bouncycastle.asn1.oiw;
-    exports org.bouncycastle.asn1.pkcs;
-    exports org.bouncycastle.asn1.rosstandart;
-    exports org.bouncycastle.asn1.sec;
-    exports org.bouncycastle.asn1.smime;
-    exports org.bouncycastle.asn1.teletrust;
-    exports org.bouncycastle.asn1.tsp;
-    exports org.bouncycastle.asn1.ua;
-    exports org.bouncycastle.asn1.util;
-    exports org.bouncycastle.asn1.x500;
-    exports org.bouncycastle.asn1.x500.style;
-    exports org.bouncycastle.asn1.x509;
-    exports org.bouncycastle.asn1.x509.qualified;
-    exports org.bouncycastle.asn1.x509.sigi;
-    exports org.bouncycastle.asn1.x9;
-    exports org.bouncycastle.crypto;
-    exports org.bouncycastle.crypto.asymmetric;
-    exports org.bouncycastle.crypto.fips;
-    exports org.bouncycastle.crypto.general;
-    exports org.bouncycastle.crypto.util;
-    exports org.bouncycastle.jcajce;
-    exports org.bouncycastle.jcajce.io;
-    exports org.bouncycastle.jcajce.interfaces;
-    exports org.bouncycastle.jcajce.provider;
-    exports org.bouncycastle.jcajce.spec;
-    exports org.bouncycastle.jcajce.util;
-    exports org.bouncycastle.math.ec;
-    exports org.bouncycastle.math.ec.custom.sec;
-    exports org.bouncycastle.math.ec.endo;
-    exports org.bouncycastle.math.ec.rfc7748;
-    exports org.bouncycastle.math.ec.rfc8032;
-    exports org.bouncycastle.math.field;
-    exports org.bouncycastle.util;
-    exports org.bouncycastle.util.encoders;
-    exports org.bouncycastle.util.io;
-    exports org.bouncycastle.util.io.pem;
-    exports org.bouncycastle.util.test;
-
-    requires java.logging;
-    requires java.naming;
-}
