--- a/pkix/src/main/java/org/bouncycastle/operator/jcajce/JcaContentVerifierProviderBuilder.java
+++ b/pkix/src/main/java/org/bouncycastle/operator/jcajce/JcaContentVerifierProviderBuilder.java
@@ -430,17 +430,20 @@
             {
                 ASN1Sequence sigSeq = ASN1Sequence.getInstance(expected);
                 boolean failed = false;
+                boolean atLeastOneChecked = false;
+
                 for (int i = 0; i != sigSeq.size(); i++)
                 {
                     if (sigs[i] != null)
                     {
+                        atLeastOneChecked = true;
                         if (!sigs[i].verify(ASN1BitString.getInstance(sigSeq.getObjectAt(i)).getOctets()))
                         {
                             failed = true;
                         }
                     }
                 }
-                return !failed;
+                return !failed & atLeastOneChecked;
             }
             catch (SignatureException e)
             {

--- a/prov/src/main/java/org/bouncycastle/x509/util/LDAPStoreHelper.java
+++ b/prov/src/main/java/org/bouncycastle/x509/util/LDAPStoreHelper.java
@@ -63,6 +63,26 @@
  */
 public class LDAPStoreHelper
 {
+    private static String[] FILTER_ESCAPE_TABLE = new String['\\' + 1];
+
+    static
+    {
+        // Filter encoding table -------------------------------------
+
+        // fill with char itself
+        for (char c = 0; c < FILTER_ESCAPE_TABLE.length; c++)
+        {
+            FILTER_ESCAPE_TABLE[c] = String.valueOf(c);
+        }
+
+        // escapes (RFC2254)
+        FILTER_ESCAPE_TABLE['*'] = "\\2a";
+        FILTER_ESCAPE_TABLE['('] = "\\28";
+        FILTER_ESCAPE_TABLE[')'] = "\\29";
+        FILTER_ESCAPE_TABLE['\\'] = "\\5c";
+        FILTER_ESCAPE_TABLE[0] = "\\00";
+    }
+

     // TODO: cache results

@@ -149,7 +169,7 @@
         {
             temp = temp.substring(0, temp.length() - 1);
         }
-        return temp;
+        return filterEncode(temp);
     }

     private Set createCerts(List list, X509CertStoreSelector xselector)
@@ -1114,4 +1134,40 @@
     {
         return cert.getIssuerX500Principal();
     }
+
+    /**
+     * Escape a value for use in an LDAP filter.
+     *
+     * @param value the value to escape.
+     * @return a properly escaped representation of the supplied value.
+     */
+    private String filterEncode(String value)
+    {
+        if (value == null)
+        {
+            return null;
+        }
+
+        // make buffer roomy
+        StringBuilder encodedValue = new StringBuilder(value.length() * 2);
+
+        int length = value.length();
+
+        for (int i = 0; i < length; i++)
+        {
+            char c = value.charAt(i);
+
+            if (c < FILTER_ESCAPE_TABLE.length)
+            {
+                encodedValue.append(FILTER_ESCAPE_TABLE[c]);
+            }
+            else
+            {
+                // default: add the char
+                encodedValue.append(c);
+            }
+        }
+
+        return encodedValue.toString();
+    }
 }

--- a/core/src/main/java/org/bouncycastle/pqc/crypto/frodo/FrodoEngine.java
+++ b/core/src/main/java/org/bouncycastle/pqc/crypto/frodo/FrodoEngine.java
@@ -100,31 +100,10 @@
         this.gen = mGen;
     }

-    private short sample(short r)
-    {
-        short t, e;
-        // 1. t = sum_{i=1}^{len_x - 1} r_i * 2^{i-1}
-        t = (short) ((r & 0xffff) >>> 1);
-        e = 0; // 2. e = 0
-        for (int z = 0; z < T_chi.length; z++)
-        {
-            if (t > T_chi[z]) // 4. if t > T_chi(z)
-                e++; // 5. e = e + 1
-        }
-        // 6. e = (-1)^{r_0} * e
-
-        if (((r & 0xffff) % 2) == 1)
-            e = (short) ((e) * (-1) & 0xffff);
-
-        return e;
-    }
-
     private short[] sample_matrix(short[] r, int offset, int n1, int n2)
     {
         short[] E = new short[n1 * n2];
-        for (int i = 0; i < n1; i++)
-            for (int j = 0; j < n2; j++)
-                E[i*n2+j] = sample(r[i * n2 + j + offset]);
+        Noise.sample(T_chi, r, offset, E);
         return E;
     }

@@ -461,22 +440,18 @@
     }


-    private short ctverify(short[] a1, short[] a2, short[] b1, short[] b2)
+    private static short ctverify(short[] a1, short[] a2, short[] b1, short[] b2)
     {
-        // Compare two arrays in constant time.
-        // Returns 0 if the byte arrays are equal, -1 otherwise.
-        short r = 0;
+        int r = 0;

-        for (short i = 0; i < a1.length; i++)
+        for (int i = 0; i < a1.length; i++)
             r |= a1[i] ^ b1[i];

-        for (short i = 0; i < a2.length; i++)
+        for (int i = 0; i < a2.length; i++)
             r |= a2[i] ^ b2[i];

-//        r = (short) ((-(short)(r >> 1) | -(short)(r & 1)) >> (8*2-1));
-        if (r == 0)
-            return 0;
-        return -1;
+        // Constant-time: returns 0 if equal, -1 if not equal.
+        return (short)((r | (-r)) >> 31);
     }

     private byte[] ctselect(byte[] a, byte[] b, short selector)
--- /dev/null
+++ b/core/src/main/java/org/bouncycastle/pqc/crypto/frodo/Noise.java
@@ -0,0 +1,30 @@
+package org.bouncycastle.pqc.crypto.frodo;
+
+abstract class Noise
+{
+    static void sample(short[] cdf, short[] r, int rOff, short[] s)
+    {
+        // No need to compare with the last value.
+//        assert cdf[cdf.length - 1] == 0x7FFF;
+
+        // Fills 's' with samples from the noise distribution 'cdf' using pseudo-random values 'r[rOff..]'
+        for (int i = 0, n = s.length; i < n; ++i)
+        {
+            int sample = 0;
+            int r_i = r[rOff + i] & 0xFFFF;
+            int prnd = r_i >>> 1;   // Drop the least significant bit
+            int sign = r_i & 1;     // Pick the least significant bit
+
+            for (int j = 0; j < cdf.length - 1; ++j)
+            {
+                // Constant time comparison: 1 if cdf[j] < prnd, 0 otherwise.
+                sample += (cdf[j] - prnd) >>> 31;
+            }
+
+            // Assuming that sign is either 0 or 1, flips sample iff sign = 1
+            sample = ((-sign) ^ sample) + sign;
+
+            s[i] = (short)sample;
+        }
+    }
+}
