--- a/lib/smtp-connection/index.js
+++ b/lib/smtp-connection/index.js
@@ -67,7 +67,7 @@
             this.secureConnection = true;
         }

-        this.name = this.options.name || this._getHostname();
+        this.name = (this.options.name || this._getHostname()).toString().replace(/[\r\n]+/g, '');

         this.logger = shared.getLogger(this.options, {
             component: this.options.component || 'smtp-connection',
