--- a/dist/node/chunks/dep-BK3b2jBa.js
+++ b/dist/node/chunks/dep-BK3b2jBa.js
@@ -62035,6 +62035,9 @@
         const depsOptimizer = getDepsOptimizer(server.config, false);
         if (depsOptimizer?.isOptimizedDepUrl(url)) {
           const sourcemapPath = url.startsWith(FS_PREFIX) ? fsPathFromId(url) : normalizePath$3(path$n.resolve(server.config.root, url.slice(1)));
+          if (!depsOptimizer.isOptimizedDepFile(sourcemapPath)) {
+            return next();
+          }
           try {
             const map = JSON.parse(
               await fsp.readFile(sourcemapPath, "utf-8")

--- a/dist/node/chunks/dep-BK3b2jBa.js
+++ b/dist/node/chunks/dep-BK3b2jBa.js
@@ -17229,11 +17229,11 @@
   } while (splitIndex !== -1);
   return parts;
 }
-const windowsDriveRE = /^[A-Z]:/;
+const windowsDriveRE$1 = /^[A-Z]:/;
 const replaceWindowsDriveRE = /^([A-Z]):\//;
 const linuxAbsolutePathRE = /^\/[^/]/;
 function escapeToLinuxLikePath(path2) {
-  if (windowsDriveRE.test(path2)) {
+  if (windowsDriveRE$1.test(path2)) {
     return path2.replace(replaceWindowsDriveRE, "/windows/$1/");
   }
   if (linuxAbsolutePathRE.test(path2)) {
@@ -51749,9 +51749,18 @@
 function isUriInFilePath(uri, filePath) {
   return isSameFileUri(uri, filePath) || isParentDirectory(uri, filePath);
 }
+const windowsDriveRE = /^[A-Z]:/i;
 function isFileLoadingAllowed(server, filePath) {
   const { fs } = server.config.server;
   if (!fs.strict) return true;
+  if (isWindows$3 && filePath.includes("~")) {
+    return false;
+  }
+  const hasDriveLetter = isWindows$3 && windowsDriveRE.test(filePath);
+  const hasColon = (hasDriveLetter ? filePath.slice(2) : filePath).includes(":");
+  if (hasColon) {
+    return false;
+  }
   const filePathWithoutTrailingSlash = filePath.endsWith("/") ? filePath.slice(0, -1) : filePath;
   if (server._fsDenyGlob(filePathWithoutTrailingSlash)) return false;
   if (server.moduleGraph.safeModulesPath.has(filePath)) return true;

--- a/dist/node/chunks/dep-BK3b2jBa.js
+++ b/dist/node/chunks/dep-BK3b2jBa.js
@@ -45399,6 +45399,10 @@
   let { fileName } = parsed;
   const { lineNumber, columnNumber } = parsed;
 
+  if (process.platform === "win32" && path$5.resolve(fileName).startsWith("\\\\")) {
+    return onErrorCallback(fileName, "UNC paths are not supported on Windows to avoid security issues. See https://github.com/vitejs/launch-editor/tree/main/packages/launch-editor#unc-paths-on-windows for details.");
+  }
+
   if (!fs$4.existsSync(fileName)) {
     return
   }
