--- a/dist/node/chunks/dep-BK3b2jBa.js
+++ b/dist/node/chunks/dep-BK3b2jBa.js
@@ -62035,6 +62035,9 @@
         const depsOptimizer = getDepsOptimizer(server.config, false);
         if (depsOptimizer?.isOptimizedDepUrl(url)) {
           const sourcemapPath = url.startsWith(FS_PREFIX) ? fsPathFromId(url) : normalizePath$3(path$n.resolve(server.config.root, url.slice(1)));
+          if (!depsOptimizer.isOptimizedDepFile(sourcemapPath)) {
+            return next();
+          }
           try {
             const map = JSON.parse(
               await fsp.readFile(sourcemapPath, "utf-8")
