--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImpl.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImpl.java
@@ -46,6 +46,7 @@
 import java.util.Iterator;
 import java.util.LinkedHashMap;
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Properties;
 import java.util.Set;
@@ -136,6 +137,8 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;

+import static org.apache.activemq.artemis.utils.PasswordMaskingUtil.isEncMasked;
+
 public class ConfigurationImpl implements Configuration, Serializable {

    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
@@ -144,6 +147,8 @@

    public static final String DOT_CLASS = ".class";

+   public static final String REDACTED = "**redacted**";
+
    private static final int DEFAULT_JMS_MESSAGE_SIZE = 1864;

    private static final int RANGE_SIZE_MIN = 0;
@@ -446,6 +451,7 @@
    private File artemisInstance;
    private transient JsonObject jsonStatus = JsonLoader.createObjectBuilder().build();
    private transient Checksum transientChecksum = null;
+   private final Set<String> keysToRedact = new HashSet<>();

    private JsonObject getJsonStatus() {
       if (jsonStatus == null) {
@@ -608,9 +614,14 @@
             checksum.update(value.getBytes(StandardCharsets.UTF_8));

             value = XMLUtil.replaceSystemPropsInString(value);
-            value = PasswordMaskingUtil.resolveMask(isMaskPassword(), value, getPasswordCodec());
             key = XMLUtil.replaceSystemPropsInString(key);
-            logger.debug("Property config, {}={}", key, value);
+            final boolean masked = isEncMasked(value);
+            if (masked) {
+               value = PasswordMaskingUtil.resolveMask(null, value, getPasswordCodec());
+               keysToRedact.add(key);
+            }
+            logger.debug("Property config, {}={}", key, (masked || shouldRedact(key)) ? REDACTED : value);
+            if (!masked) value = PasswordMaskingUtil.resolveMask(isMaskPassword(), value, getPasswordCodec());
             beanProperties.put(key, value);
          }
          alder32Hash = checksum.getValue();
@@ -623,6 +634,10 @@
       }
    }

+   private boolean shouldRedact(String name) {
+      return name.toUpperCase(Locale.ENGLISH).contains("PASSWORD") || keysToRedact.contains(name);
+   }
+
    public void populateWithProperties(final Object target, final String propsId, Map<String, Object> beanProperties) throws InvocationTargetException, IllegalAccessException {
       CollectionAutoFillPropertiesUtil autoFillCollections = new CollectionAutoFillPropertiesUtil(getBrokerPropertiesRemoveValue(beanProperties));
       BeanUtilsBean beanUtils = new BeanUtilsBean(new ConvertUtilsBean(), autoFillCollections) {
@@ -656,7 +671,7 @@
          @Override
          public void setProperty(final Object bean, String name, final Object value) throws InvocationTargetException, IllegalAccessException {
             {
                if (logger.isDebugEnabled()) {
-                  logger.debug("setProperty on {}, name: {}, value: {}", bean.getClass(), name, value);
+                  logger.debug("setProperty on {}, name: {}, value: {}", bean.getClass(), name, shouldRedact(name) ? REDACTED : value);
                }
                // Resolve any nested expression to get the actual target bean
                Object target = bean;

--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerSessionImpl.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerSessionImpl.java
@@ -756,7 +756,7 @@

       AddressSettings as = server.getAddressSettingsRepository().getMatch(queueConfiguration.getAddress().toString());

-      if (as.isAutoCreateAddresses() && server.getAddressInfo(queueConfiguration.getAddress()) == null) {
+      if (as.isAutoCreateAddresses() && (server.getAddressInfo(queueConfiguration.getAddress()) == null || !server.getAddressInfo(queueConfiguration.getAddress()).getRoutingTypes().contains(queueConfiguration.getRoutingType()))) {
          securityCheck(queueConfiguration.getAddress(), queueConfiguration.getName(), CheckType.CREATE_ADDRESS, this);
       }

--- a/artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java
+++ b/artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java
@@ -35,6 +35,7 @@
 import org.apache.activemq.artemis.api.core.ICoreMessage;
 import org.apache.activemq.artemis.api.core.Message;
 import org.apache.activemq.artemis.api.core.QueueConfiguration;
+import org.apache.activemq.artemis.api.core.RoutingType;
 import org.apache.activemq.artemis.api.core.SimpleString;
 import org.apache.activemq.artemis.core.client.impl.ClientConsumerImpl;
 import org.apache.activemq.artemis.core.protocol.openwire.OpenWireConstants;
@@ -215,6 +216,7 @@
          if (info.getDestination().isComposite()) {
             queueName =  queueName.concat(physicalName);
          }
+         QueueConfiguration queueConfiguration = new QueueConfiguration(queueName).setAddress(address).setRoutingType(RoutingType.MULTICAST).setFilterString(selector).setInternal(internalAddress);
          QueueQueryResult result = session.getCoreSession().executeQueueQuery(queueName);
          if (result.isExists()) {
             // Already exists
@@ -235,10 +237,10 @@
                session.getCoreSession().deleteQueue(queueName);

                // Create the new one
-               session.getCoreSession().createQueue(new QueueConfiguration(queueName).setAddress(address).setFilterString(selector).setInternal(internalAddress));
+               session.getCoreSession().createQueue(queueConfiguration);
             }
          } else {
-            session.getCoreSession().createQueue(new QueueConfiguration(queueName).setAddress(address).setFilterString(selector).setInternal(internalAddress));
+            session.getCoreSession().createQueue(queueConfiguration);
          }
       } else {
          /*
@@ -253,7 +255,7 @@
             queueName = new SimpleString(UUID.randomUUID().toString());
          }

-         session.getCoreSession().createQueue(new QueueConfiguration(queueName).setAddress(address).setFilterString(selector).setDurable(false).setTemporary(true).setInternal(internalAddress));
+         session.getCoreSession().createQueue(new QueueConfiguration(queueName).setAddress(address).setRoutingType(RoutingType.MULTICAST).setFilterString(selector).setDurable(false).setTemporary(true).setInternal(internalAddress));
       }

       return queueName;
--- a/tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/security/SecurityTest.java
+++ b/tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/security/SecurityTest.java
@@ -644,6 +644,123 @@
          assertTrue(e.getMessage().contains("does not have permission='BROWSE' for queue durableQueue on address address"));
       }
    }
+
+   @Test
+   public void testJAASSecurityManagerCreateQueueWithDifferentRoutingTypeAsAddressNegative() throws Exception {
+      final SimpleString ADDRESS = new SimpleString("address");
+      final SimpleString DURABLE_QUEUE = new SimpleString("durableQueue");
+      final SimpleString NON_DURABLE_QUEUE = new SimpleString("nonDurableQueue");
+      final SimpleString JMS = new SimpleString("jms");
+
+      ActiveMQJAASSecurityManager securityManager = new ActiveMQJAASSecurityManager("PropertiesLogin");
+      ActiveMQServer server = addServer(ActiveMQServers.newActiveMQServer(createDefaultInVMConfig().setSecurityEnabled(true), ManagementFactory.getPlatformMBeanServer(), securityManager, false));
+      Set<Role> roles = new HashSet<>();
+      roles.add(new Role("programmers", false, false, true, false, true, false, false, false, false, false, false, false));
+      server.getConfiguration().putSecurityRoles("#", roles);
+      server.start();
+      server.addAddressInfo(new AddressInfo(ADDRESS, RoutingType.ANYCAST));
+      server.addAddressInfo(new AddressInfo(JMS, RoutingType.ANYCAST));
+
+      ClientSessionFactory cf = createSessionFactory(locator);
+      ClientSession session = addClientSession(cf.createSession("first", "secret", false, true, true, false, 0));
+
+      ConnectionFactory connectionFactory = new ActiveMQConnectionFactory("vm://0");
+
+      // Explicit attempt to modify address routing type with a durable queue
+      try {
+         session.createQueue(new QueueConfiguration(DURABLE_QUEUE).setAddress(ADDRESS).setRoutingType(RoutingType.MULTICAST));
+         Assert.fail("should throw exception here");
+      } catch (ActiveMQException e) {
+         assertTrue(e.getMessage().contains("User: first"));
+         assertTrue(e.getMessage().contains("does not have permission='CREATE_ADDRESS' for queue durableQueue on address address"));
+      }
+
+      // Implicit attempt to modify address routing type with a durable queue using auto-create via JMS
+      try (Connection c = connectionFactory.createConnection("first", "secret")) {
+         c.setClientID("myClientID");
+         Session s = c.createSession();
+         s.createDurableSubscriber(s.createTopic(JMS.toString()), "foo");
+         Assert.fail("should throw exception here");
+      } catch (JMSException e) {
+         assertTrue(e.getMessage().contains("User: first"));
+         assertTrue(e.getMessage().contains("does not have permission='CREATE_ADDRESS' for queue myClientID.foo on address jms"));
+      }
+
+      // Explicit attempt to modify address routing type with a non-durable queue
+      try {
+         session.createQueue(new QueueConfiguration(NON_DURABLE_QUEUE).setAddress(ADDRESS).setDurable(false).setRoutingType(RoutingType.MULTICAST));
+         Assert.fail("should throw exception here");
+      } catch (ActiveMQException e) {
+         assertTrue(e.getMessage().contains("User: first"));
+         assertTrue(e.getMessage().contains("does not have permission='CREATE_ADDRESS' for queue nonDurableQueue on address address"));
+      }
+
+      // Implicit attempt to modify address routing type with a non-durable queue using auto-create via JMS
+      try (Connection c = connectionFactory.createConnection("first", "secret")) {
+         Session s = c.createSession();
+         s.createConsumer(s.createTopic(JMS.toString()));
+         Assert.fail("should throw exception here");
+      } catch (JMSException e) {
+         assertTrue(e.getMessage().contains("User: first"));
+         assertTrue(e.getMessage().contains("does not have permission='CREATE_ADDRESS'"));
+      }
+   }
+
+   @Test
+   public void testJAASSecurityManagerCreateQueueWithDifferentRoutingTypeAsAddress() throws Exception {
+      final SimpleString ADDRESS = new SimpleString("address");
+      final SimpleString DURABLE_QUEUE = new SimpleString("durableQueue");
+      final SimpleString NON_DURABLE_QUEUE = new SimpleString("nonDurableQueue");
+      final SimpleString JMS = new SimpleString("jms");
+
+      ActiveMQJAASSecurityManager securityManager = new ActiveMQJAASSecurityManager("PropertiesLogin");
+      ActiveMQServer server = addServer(ActiveMQServers.newActiveMQServer(createDefaultInVMConfig().setSecurityEnabled(true), ManagementFactory.getPlatformMBeanServer(), securityManager, false));
+      Set<Role> roles = new HashSet<>();
+      roles.add(new Role("programmers", false, true, true, false, true, false, false, false, true, false, false, false));
+      server.getConfiguration().putSecurityRoles("#", roles);
+      server.start();
+      server.addAddressInfo(new AddressInfo(ADDRESS, RoutingType.ANYCAST));
+
+      ClientSessionFactory cf = createSessionFactory(locator);
+      ClientSession session = addClientSession(cf.createSession("first", "secret", false, true, true, false, 0));
+
+      ConnectionFactory connectionFactory = new ActiveMQConnectionFactory("vm://0");
+
+      // Explicit attempt to modify address routing type with a durable queue
+      try {
+         session.createQueue(new QueueConfiguration(DURABLE_QUEUE).setAddress(ADDRESS).setRoutingType(RoutingType.MULTICAST));
+      } catch (ActiveMQException e) {
+         e.printStackTrace();
+         Assert.fail("should not throw exception here");
+      }
+
+      // Implicit attempt to modify address routing type with a durable queue using auto-create via JMS
+      try (Connection c = connectionFactory.createConnection("first", "secret")) {
+         c.setClientID("myClientID");
+         Session s = c.createSession();
+         s.createDurableSubscriber(s.createTopic(JMS.toString()), "foo");
+      } catch (JMSException e) {
+         e.printStackTrace();
+         Assert.fail("should not throw exception here");
+      }
+
+      // Explicit attempt to modify address routing type with a non-durable queue
+      try {
+         session.createQueue(new QueueConfiguration(NON_DURABLE_QUEUE).setAddress(ADDRESS).setDurable(false).setRoutingType(RoutingType.MULTICAST));
+      } catch (ActiveMQException e) {
+         e.printStackTrace();
+         Assert.fail("should not throw exception here");
+      }
+
+      // Implicit attempt to modify address routing type with a non-durable queue using auto-create via JMS
+      try (Connection c = connectionFactory.createConnection("first", "secret")) {
+         Session s = c.createSession();
+         s.createConsumer(s.createTopic(JMS.toString()));
+      } catch (JMSException e) {
+         e.printStackTrace();
+         Assert.fail("should not throw exception here");
+      }
+   }

    @Test
    // this is for backwards compatibility with the pre-FQQN syntax from ARTEMIS-592

--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/Configuration.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/Configuration.java
@@ -1508,5 +1508,10 @@
    String getManagementRbacPrefix();

    void setManagementRbacPrefix(String prefix);

+   List<String> getFederationDownstreamAuthorization();
+
+   void setFederationDownstreamAuthorization(List<String> roles);
+
+   Configuration addFederationDownstreamAuthorization(String role);
 }
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImpl.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImpl.java
@@ -220,6 +220,8 @@
    protected List<AMQPBrokerConnectConfiguration> amqpBrokerConnectConfigurations = new ArrayList<>();

    protected List<FederationConfiguration> federationConfigurations = new ArrayList<>();
+
+   protected List<String> downstreamAuthorization = new ArrayList<>();

    @Deprecated
    // this can eventually be replaced with List<QueueConfiguration>, but to keep existing semantics it must stay as is for now
@@ -3341,7 +3343,23 @@
    @Override
    public void setManagementRbacPrefix(String prefix) {
       this.managementRbacPrefix = prefix;
    }
+
+   @Override
+   public List<String> getFederationDownstreamAuthorization() {
+      return downstreamAuthorization;
+   }
+
+   @Override
+   public void setFederationDownstreamAuthorization(List<String> roles) {
+      this.downstreamAuthorization = roles;
+   }
+
+   @Override
+   public Configuration addFederationDownstreamAuthorization(String role) {
+      downstreamAuthorization.add(role);
+      return this;
+   }

    // extend property utils with ability to auto-fill and locate from collections
    // collection entries are identified by the name() property
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/deployers/impl/FileConfigurationParser.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/deployers/impl/FileConfigurationParser.java
@@ -29,6 +29,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.stream.Stream;
 import java.util.concurrent.TimeUnit;

 import org.apache.activemq.artemis.ArtemisConstants;
@@ -644,6 +645,13 @@
          parseBridgeConfiguration(mfNode, config);
       }

+      NodeList federations = e.getElementsByTagName("federations");
+
+      for (int i = 0; i < federations.getLength(); i++) {
+         Element fedNode = (Element) federations.item(i);
+         parseFederationsConfiguration(fedNode, config);
+      }
+
       NodeList fedNodes = e.getElementsByTagName("federation");

       for (int i = 0; i < fedNodes.getLength(); i++) {
@@ -2551,6 +2559,16 @@
       mainConfig.getBridgeConfigurations().add(config);
    }

+   private void parseFederationsConfiguration(final Element fedNode, final Configuration mainConfig) throws Exception {
+      String roles = fedNode.getAttribute("downstream-authorization");
+      if (!roles.isEmpty()) {
+         Stream.of(roles.split(","))
+            .map(String::trim)
+            .filter(s -> !s.isEmpty())
+            .forEach(mainConfig::addFederationDownstreamAuthorization);
+      }
+   }
+
    private void parseFederationConfiguration(final Element fedNode, final Configuration mainConfig) throws Exception {
       FederationConfiguration config = new FederationConfiguration();

--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/CoreProtocolManager.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/CoreProtocolManager.java
@@ -65,13 +65,16 @@
 import org.apache.activemq.artemis.core.remoting.impl.invm.InVMConnection;
 import org.apache.activemq.artemis.core.remoting.impl.netty.ActiveMQFrameDecoder2;
 import org.apache.activemq.artemis.core.remoting.impl.netty.NettyServerConnection;
 import org.apache.activemq.artemis.core.server.ActiveMQServer;
+import org.apache.activemq.artemis.core.server.ActiveMQServerLogger;
 import org.apache.activemq.artemis.spi.core.protocol.ConnectionEntry;
 import org.apache.activemq.artemis.spi.core.protocol.ProtocolManager;
 import org.apache.activemq.artemis.spi.core.protocol.ProtocolManagerFactory;
 import org.apache.activemq.artemis.spi.core.protocol.RemotingConnection;
 import org.apache.activemq.artemis.spi.core.remoting.Acceptor;
 import org.apache.activemq.artemis.spi.core.remoting.Connection;
+import org.apache.activemq.artemis.spi.core.security.jaas.UserPrincipal;
+import org.apache.activemq.artemis.utils.SecurityManagerUtil;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import java.lang.invoke.MethodHandles;
@@ -164,8 +167,7 @@
          .addClusterChannelHandler(rc.getChannel(CHANNEL_ID.CLUSTER.id, -1), acceptorUsed, rc,
                                    server.getActivation());

-      final Channel federationChannel =  rc.getChannel(CHANNEL_ID.FEDERATION.id, -1);
-      federationChannel.setHandler(new LocalChannelHandler(config, entry, channel0, acceptorUsed, rc));
+      rc.getChannel(CHANNEL_ID.FEDERATION.id, -1).setHandler(new FederationChannelHandler(acceptorUsed, rc));

       return entry;
    }
@@ -413,117 +415,152 @@
                   }
                });
             }
-         } else if (packet.getType() == PacketImpl.FEDERATION_DOWNSTREAM_CONNECT) {
-            //If we receive this packet then a remote broker is requesting us to create federated upstream connection
-            //back to it which simulates a downstream connection
-            final FederationDownstreamConnectMessage message = (FederationDownstreamConnectMessage) packet;
-            final FederationDownstreamConfiguration downstreamConfiguration = message.getStreamConfiguration();
-
-            //Create a new Upstream Federation configuration based on the received Downstream connection message
-            //from the remote broker
-            //The idea here is to set all the same configuration parameters that apply to the upstream connection
-            final FederationConfiguration config = new FederationConfiguration();
-            config.setName(message.getName() + FederationDownstreamConnectMessage.UPSTREAM_SUFFIX);
-            config.setCredentials(message.getCredentials());
-
-            //Add the policy map configuration
-            for (FederationPolicy policy : message.getFederationPolicyMap().values()) {
-               config.addFederationPolicy(policy);
-            }
-
-            //Add any transformer configurations
-            for (FederationTransformerConfiguration transformerConfiguration : message.getTransformerConfigurationMap().values()) {
-               config.addTransformerConfiguration(transformerConfiguration);
-            }
-
-            //Create an upstream configuration with the same name but apply the upstream suffix so it is unique
-            final FederationUpstreamConfiguration upstreamConfiguration = new FederationUpstreamConfiguration()
-               .setName(downstreamConfiguration.getName() + FederationDownstreamConnectMessage.UPSTREAM_SUFFIX)
-               .addPolicyRefs(downstreamConfiguration.getPolicyRefs());
-
-            //Use the provided Transport Configuration information to create an upstream connection back to the broker that
-            //created the downstream connection
-            final TransportConfiguration upstreamConfig = downstreamConfiguration.getUpstreamConfiguration();
-
-            //Initialize the upstream transport with the config from the acceptor as this will apply
-            //relevant settings such as SSL, then override with settings from the downstream config
-            final Map<String, Object> params = new HashMap<>(acceptorUsed.getConfiguration());
-            if (upstreamConfig.getParams() != null) {
-               params.putAll(upstreamConfig.getParams());
-            }
-            final Map<String, Object> extraParams = new HashMap<>();
-            if (upstreamConfig.getExtraParams() != null) {
-               extraParams.putAll(upstreamConfig.getExtraParams());
-            }
-
-            //Add the new upstream configuration that was created so we can connect back to the downstream server
-            final TransportConfiguration upstreamConf = new TransportConfiguration(
-               upstreamConfig.getFactoryClassName(), params, upstreamConfig.getName() + FederationDownstreamConnectMessage.UPSTREAM_SUFFIX,
-               extraParams);
-            server.getConfiguration()
-               .addConnectorConfiguration(upstreamConf.getName() + FederationDownstreamConnectMessage.UPSTREAM_SUFFIX, upstreamConf);
-
-            //Create a new upstream connection config based on the downstream configuration
-            FederationConnectionConfiguration downstreamConConf = downstreamConfiguration.getConnectionConfiguration();
-            FederationConnectionConfiguration upstreamConConf = upstreamConfiguration.getConnectionConfiguration();
-            List<String> connectorNames = new ArrayList<>();
-            connectorNames.add(upstreamConf.getName() + FederationDownstreamConnectMessage.UPSTREAM_SUFFIX);
-
-            //Configure all of the upstream connection parameters from the downstream connection that are relevant
-            //Note that HA and discoveryGroupName are skipped because the downstream connection will manage that
-            //In this case we just want to create a connection back to the broker that sent the downstream packet.
-            //If this broker goes down then the original broker (if configured with HA) will re-establish a new
-            //connection to another broker which will then create another upstream, etc
-            upstreamConConf.setStaticConnectors(connectorNames);
-            upstreamConConf.setUsername(downstreamConConf.getUsername());
-            upstreamConConf.setPassword(downstreamConConf.getPassword());
-            upstreamConConf.setShareConnection(downstreamConConf.isShareConnection());
-            upstreamConConf.setPriorityAdjustment(downstreamConConf.getPriorityAdjustment());
-            upstreamConConf.setClientFailureCheckPeriod(downstreamConConf.getClientFailureCheckPeriod());
-            upstreamConConf.setConnectionTTL(downstreamConConf.getConnectionTTL());
-            upstreamConConf.setRetryInterval(downstreamConConf.getRetryInterval());
-            upstreamConConf.setRetryIntervalMultiplier(downstreamConConf.getRetryIntervalMultiplier());
-            upstreamConConf.setMaxRetryInterval(downstreamConConf.getMaxRetryInterval());
-            upstreamConConf.setInitialConnectAttempts(downstreamConConf.getInitialConnectAttempts());
-            upstreamConConf.setReconnectAttempts(downstreamConConf.getReconnectAttempts());
-            upstreamConConf.setCallTimeout(downstreamConConf.getCallTimeout());
-            upstreamConConf.setCallFailoverTimeout(downstreamConConf.getCallFailoverTimeout());
-            config.addUpstreamConfiguration(upstreamConfiguration);
-
-            //Register close and failure listeners, if the initial downstream connection goes down then we
-            //want to terminate the upstream connection
-            rc.addCloseListener(() -> {
-               server.getFederationManager().undeploy(config.getName());
-            });
-
-            rc.addFailureListener(new FailureListener() {
-               @Override
-               public void connectionFailed(ActiveMQException exception, boolean failedOver) {
-                  server.getFederationManager().undeploy(config.getName());
-               }
-
-               @Override
-               public void connectionFailed(ActiveMQException exception, boolean failedOver,
-                                            String scaleDownTargetNodeID) {
-                  server.getFederationManager().undeploy(config.getName());
-               }
-            });
-
-            try {
-               server.getFederationManager().deploy(config);
-            } catch (Exception e) {
-               logger.error("Error deploying federation", e);
-            }
          }
       }

       private Pair<TransportConfiguration, TransportConfiguration> getPair(
          TransportConfiguration conn,
          boolean isBackup) {
          if (isBackup) {
             return new Pair<>(null, conn);
          }
          return new Pair<>(conn, null);
       }
    }
-}
+
+   private class FederationChannelHandler implements ChannelHandler {
+
+      private final Acceptor acceptorUsed;
+      private final CoreRemotingConnection rc;
+
+      private FederationChannelHandler(final Acceptor acceptorUsed, final CoreRemotingConnection rc) {
+         this.acceptorUsed = acceptorUsed;
+         this.rc = rc;
+      }
+
+      @Override
+      public void handlePacket(final Packet packet) {
+         if (packet.getType() == PacketImpl.FEDERATION_DOWNSTREAM_CONNECT) {
+            if (server.getSecurityStore().isSecurityEnabled()) {
+               if (rc.getSubject() == null) {
+                  ActiveMQServerLogger.LOGGER.federationDownstreamUnauthenticated(rc.getRemoteAddress());
+                  rc.close();
+                  return;
+               }
+               if (!server.getFederationManager().authorizeDownstreamDeployment(rc.getSubject())) {
+                  ActiveMQServerLogger.LOGGER.federationDownstreamUnauthorized(rc.getRemoteAddress(), SecurityManagerUtil.getUserFromSubject(rc.getSubject(), UserPrincipal.class));
+                  rc.close();
+                  return;
+               }
+            }
+            //If we receive this packet then a remote broker is requesting us to create federated upstream connection
+            //back to it which simulates a downstream connection
+            final FederationDownstreamConnectMessage message = (FederationDownstreamConnectMessage) packet;
+            final FederationDownstreamConfiguration downstreamConfiguration = message.getStreamConfiguration();
+
+            //Create a new Upstream Federation configuration based on the received Downstream connection message
+            //from the remote broker
+            //The idea here is to set all the same configuration parameters that apply to the upstream connection
+            final FederationConfiguration config = new FederationConfiguration();
+            config.setName(message.getName() + FederationDownstreamConnectMessage.UPSTREAM_SUFFIX);
+            config.setCredentials(message.getCredentials());
+
+            //Add the policy map configuration
+            for (FederationPolicy policy : message.getFederationPolicyMap().values()) {
+               config.addFederationPolicy(policy);
+            }
+
+            //Add any transformer configurations
+            for (FederationTransformerConfiguration transformerConfiguration : message.getTransformerConfigurationMap().values()) {
+               config.addTransformerConfiguration(transformerConfiguration);
+            }
+
+            //Create an upstream configuration with the same name but apply the upstream suffix so it is unique
+            final FederationUpstreamConfiguration upstreamConfiguration = new FederationUpstreamConfiguration()
+               .setName(downstreamConfiguration.getName() + FederationDownstreamConnectMessage.UPSTREAM_SUFFIX)
+               .addPolicyRefs(downstreamConfiguration.getPolicyRefs());
+
+            //Use the provided Transport Configuration information to create an upstream connection back to the broker that
+            //created the downstream connection
+            final TransportConfiguration upstreamConfig = downstreamConfiguration.getUpstreamConfiguration();
+
+            //Initialize the upstream transport with the config from the acceptor as this will apply
+            //relevant settings such as SSL, then override with settings from the downstream config
+            final Map<String, Object> params = new HashMap<>(acceptorUsed.getConfiguration());
+            if (upstreamConfig.getParams() != null) {
+               params.putAll(upstreamConfig.getParams());
+            }
+            final Map<String, Object> extraParams = new HashMap<>();
+            if (upstreamConfig.getExtraParams() != null) {
+               extraParams.putAll(upstreamConfig.getExtraParams());
+            }
+
+            //Add the new upstream configuration that was created so we can connect back to the downstream server
+            final TransportConfiguration upstreamConf = new TransportConfiguration(
+               upstreamConfig.getFactoryClassName(), params, upstreamConfig.getName() + FederationDownstreamConnectMessage.UPSTREAM_SUFFIX,
+               extraParams);
+            server.getConfiguration()
+               .addConnectorConfiguration(upstreamConf.getName() + FederationDownstreamConnectMessage.UPSTREAM_SUFFIX, upstreamConf);
+
+            //Create a new upstream connection config based on the downstream configuration
+            FederationConnectionConfiguration downstreamConConf = downstreamConfiguration.getConnectionConfiguration();
+            FederationConnectionConfiguration upstreamConConf = upstreamConfiguration.getConnectionConfiguration();
+            List<String> connectorNames = new ArrayList<>();
+            connectorNames.add(upstreamConf.getName() + FederationDownstreamConnectMessage.UPSTREAM_SUFFIX);
+
+            //Configure all of the upstream connection parameters from the downstream connection that are relevant
+            //Note that HA and discoveryGroupName are skipped because the downstream connection will manage that
+            //In this case we just want to create a connection back to the broker that sent the downstream packet.
+            //If this broker goes down then the original broker (if configured with HA) will re-establish a new
+            //connection to another broker which will then create another upstream, etc
+            upstreamConConf.setStaticConnectors(connectorNames);
+            upstreamConConf.setUsername(downstreamConConf.getUsername());
+            upstreamConConf.setPassword(downstreamConConf.getPassword());
+            upstreamConConf.setShareConnection(downstreamConConf.isShareConnection());
+            upstreamConConf.setPriorityAdjustment(downstreamConConf.getPriorityAdjustment());
+            upstreamConConf.setClientFailureCheckPeriod(downstreamConConf.getClientFailureCheckPeriod());
+            upstreamConConf.setConnectionTTL(downstreamConConf.getConnectionTTL());
+            upstreamConConf.setRetryInterval(downstreamConConf.getRetryInterval());
+            upstreamConConf.setRetryIntervalMultiplier(downstreamConConf.getRetryIntervalMultiplier());
+            upstreamConConf.setMaxRetryInterval(downstreamConConf.getMaxRetryInterval());
+            upstreamConConf.setInitialConnectAttempts(downstreamConConf.getInitialConnectAttempts());
+            upstreamConConf.setReconnectAttempts(downstreamConConf.getReconnectAttempts());
+            upstreamConConf.setCallTimeout(downstreamConConf.getCallTimeout());
+            upstreamConConf.setCallFailoverTimeout(downstreamConConf.getCallFailoverTimeout());
+            config.addUpstreamConfiguration(upstreamConfiguration);
+
+            //Register close and failure listeners, if the initial downstream connection goes down then we
+            //want to terminate the upstream connection
+            rc.addCloseListener(() -> {
+               handleClose(config.getName(), rc.getRemoteAddress());
+            });
+
+            rc.addFailureListener(new FailureListener() {
+               @Override
+               public void connectionFailed(ActiveMQException exception, boolean failedOver) {
+                  handleClose(config.getName(), rc.getRemoteAddress());
+               }
+
+               @Override
+               public void connectionFailed(ActiveMQException exception, boolean failedOver, String scaleDownTargetNodeID) {
+                  handleClose(config.getName(), rc.getRemoteAddress());
+               }
+            });
+
+            try {
+               server.getFederationManager().deploy(config);
+               String user = rc.getSubject() != null ? SecurityManagerUtil.getUserFromSubject(rc.getSubject(), UserPrincipal.class) : "anonymous";
+               ActiveMQServerLogger.LOGGER.federationDownstreamDeployedFromRemoteUser(config.getName(), user, rc.getRemoteAddress());
+            } catch (Exception e) {
+               logger.error("Error deploying federation", e);
+            }
+         }
+      }
+
+      private void handleClose(String configName, String remoteAddress) {
+         if (server.getFederationManager().undeploy(configName)) {
+            ActiveMQServerLogger.LOGGER.federationDownstreamConnectionClosed(remoteAddress, configName);
+         }
+      }
+   }
+}
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQServerLogger.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQServerLogger.java
@@ -1129,22 +1129,22 @@
    @LogMessage(id = 222278, value = "Unable to extract GroupSequence from message", level = LogMessage.Level.WARN)
    void unableToExtractGroupSequence(Throwable e);

-   @LogMessage(id = 222279, value = "Federation upstream {} policy ref {} could not be resolved in federation configuration", level = LogMessage.Level.WARN)
+   @LogMessage(id = 222279, value = "Core federation upstream {} policy ref {} could not be resolved in configuration", level = LogMessage.Level.WARN)
    void federationCantFindPolicyRef(String upstreamName, String policyRef);

-   @LogMessage(id = 222280, value = "Federation upstream {} policy ref {} is of unknown type in federation configuration", level = LogMessage.Level.WARN)
+   @LogMessage(id = 222280, value = "Core federation upstream {} policy ref {} is of unknown type in configuration", level = LogMessage.Level.WARN)
    void federationUnknownPolicyType(String upstreamName, String policyRef);

-   @LogMessage(id = 222281, value = "Federation upstream {} policy ref {} are too self referential, avoiding stack overflow , ", level = LogMessage.Level.WARN)
+   @LogMessage(id = 222281, value = "Core federation upstream {} policy ref {} are too self referential, avoiding stack overflow", level = LogMessage.Level.WARN)
    void federationAvoidStackOverflowPolicyRef(String upstreamName, String policyRef);

-   @LogMessage(id = 222282, value = "Federation downstream {} upstream transport configuration ref {} could not be resolved in federation configuration", level = LogMessage.Level.WARN)
+   @LogMessage(id = 222282, value = "Core federation downstream {} upstream transport configuration ref {} could not be resolved in configuration", level = LogMessage.Level.WARN)
    void federationCantFindUpstreamConnector(String downstreamName, String upstreamRef);

-   @LogMessage(id = 222283, value = "Federation downstream {} has been deployed", level = LogMessage.Level.INFO)
+   @LogMessage(id = 222283, value = "Core federation downstream {} has been deployed", level = LogMessage.Level.INFO)
    void federationDownstreamDeployed(String downstreamName);

-   @LogMessage(id = 222284, value = "Federation downstream {} has been undeployed", level = LogMessage.Level.INFO)
+   @LogMessage(id = 222284, value = "Core federation downstream {} has been undeployed", level = LogMessage.Level.INFO)
    void federationDownstreamUnDeployed(String downstreamName);

    @LogMessage(id = 222285, value = "File {} at {} is empty. Delete the empty file to stop this message.", level = LogMessage.Level.WARN)
@@ -1151,6 +1151,6 @@
    void emptyAddressFile(String addressFile, String directory);

-   @LogMessage(id = 222286, value = "Error executing {} federation plugin method.", level = LogMessage.Level.WARN)
+   @LogMessage(id = 222286, value = "Error executing {} Core federation plugin method.", level = LogMessage.Level.WARN)
    void federationPluginExecutionError(String pluginMethod, Throwable e);

    @LogMessage(id = 222287, value = "Error looking up bindings for address {}.", level = LogMessage.Level.WARN)
@@ -1207,6 +1207,6 @@
    void unableToLoadMessageFromJournal(Throwable t);

-   @LogMessage(id = 222305, value = "Error federating message {}.", level = LogMessage.Level.WARN)
+   @LogMessage(id = 222305, value = "Error during Core federation of message: {}.", level = LogMessage.Level.WARN)
    void federationDispatchError(String message, Throwable e);

    @LogMessage(id = 222306, value = "Failed to load prepared TX and it will be rolled back: {}", level = LogMessage.Level.WARN)
@@ -1611,4 +1611,16 @@

    @LogMessage(id = 224136, value = "Skipping correlation ID when converting message to OpenWire since byte[] value is not valid UTF-8: {}", level = LogMessage.Level.WARN)
    void unableToDecodeCorrelationId(String message);
+
+   @LogMessage(id = 224158, value = "Unable to process Core federation downstream request from {}. User is not authenticated. Closing connection.", level = LogMessage.Level.WARN)
+   void federationDownstreamUnauthenticated(String remoteAddress);
+
+   @LogMessage(id = 224159, value = "Unable to process Core federation downstream request from {}. User {} is not authorized. Closing connection.", level = LogMessage.Level.WARN)
+   void federationDownstreamUnauthorized(String remoteAddress, String user);
+
+   @LogMessage(id = 224160, value = "Core federation downstream {} has been deployed. Request sent by user {} from {}.", level = LogMessage.Level.INFO)
+   void federationDownstreamDeployedFromRemoteUser(String name, String userFromSubject, String remoteAddress);
+
+   @LogMessage(id = 224161, value = "Connection from {} closed. Undeployed Core federation {}.", level = LogMessage.Level.INFO)
+   void federationDownstreamConnectionClosed(String remoteAddress, String name);
 }
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/federation/FederationManager.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/federation/FederationManager.java
@@ -17,12 +17,15 @@
 package org.apache.activemq.artemis.core.server.federation;

+import javax.security.auth.Subject;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import org.apache.activemq.artemis.api.core.ActiveMQException;
 import org.apache.activemq.artemis.core.server.ActiveMQComponent;
 import org.apache.activemq.artemis.core.server.ActiveMQServer;
 import org.apache.activemq.artemis.core.config.FederationConfiguration;
+import org.apache.activemq.artemis.spi.core.security.jaas.RolePrincipal;

 public class FederationManager implements ActiveMQComponent {

@@ -29,6 +32,7 @@

    private Map<String, Federation> federations = new HashMap<>();
    private State state;
+   private List<String> downstreamAuthorization;

    enum State {
       STOPPED,
@@ -46,6 +50,7 @@

    public FederationManager(final ActiveMQServer server) {
       this.server = server;
+      this.downstreamAuthorization = server.getConfiguration().getFederationDownstreamAuthorization();
    }

    @Override
@@ -89,9 +94,10 @@
    public synchronized boolean undeploy(String name) {
       Federation federation = federations.remove(name);
       if (federation != null) {
          federation.stop();
+         return true;
       }
-      return true;
+      return false;
    }


@@ -129,5 +135,17 @@
    public void unregister(FederatedAbstract federatedAbstract) {
       server.unRegisterBrokerPlugin(federatedAbstract);
    }
+
+   public boolean authorizeDownstreamDeployment(Subject subject) {
+      if (!server.getSecurityStore().isSecurityEnabled()) {
+         return true;
+      }
+      for (RolePrincipal role : subject.getPrincipals(RolePrincipal.class)) {
+         if (downstreamAuthorization.contains(role.getName())) {
+            return true;
+         }
+      }
+      return false;
+   }

 }
--- a/artemis-server/src/main/resources/schema/artemis-configuration.xsd
+++ b/artemis-server/src/main/resources/schema/artemis-configuration.xsd
@@ -1544,7 +1544,8 @@
       <xsd:complexType>
          <xsd:sequence>
             <xsd:element name="federation" type="federationType" maxOccurs="unbounded" minOccurs="0"/>
          </xsd:sequence>
+         <xsd:attribute name="downstream-authorization" type="xsd:string" use="optional" />
          <xsd:attributeGroup ref="xml:specialAttrs"/>
       </xsd:complexType>
    </xsd:element>
--- a/artemis-server/src/test/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImplTest.java
+++ b/artemis-server/src/test/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImplTest.java
@@ -1413,6 +1413,20 @@
       Assert.assertEquals("u", configuration.getFederationConfigurations().get(0).getCredentials().getUser());
       Assert.assertEquals("secureexample", configuration.getFederationConfigurations().get(0).getCredentials().getPassword());
    }
+
+   @Test
+   public void testParseDownstreamAuthorization() throws Throwable {
+      Properties properties = new Properties();
+      properties.put("federationDownstreamAuthorization", "a,b,c,d,e");
+      ConfigurationImpl configuration = new ConfigurationImpl();
+      configuration.parsePrefixedProperties(properties, null);
+      Assert.assertEquals(5, configuration.getFederationDownstreamAuthorization().size());
+      Assert.assertEquals("a", configuration.getFederationDownstreamAuthorization().get(0));
+      Assert.assertEquals("b", configuration.getFederationDownstreamAuthorization().get(1));
+      Assert.assertEquals("c", configuration.getFederationDownstreamAuthorization().get(2));
+      Assert.assertEquals("d", configuration.getFederationDownstreamAuthorization().get(3));
+      Assert.assertEquals("e", configuration.getFederationDownstreamAuthorization().get(4));
+   }

    @Test
    public void testSetNestedPropertyOnCollections() throws Throwable {
--- a/artemis-server/src/test/java/org/apache/activemq/artemis/core/config/impl/FileConfigurationParserTest.java
+++ b/artemis-server/src/test/java/org/apache/activemq/artemis/core/config/impl/FileConfigurationParserTest.java
@@ -766,6 +766,20 @@
       Configuration configuration = parser.parseMainConfig(inputStream);
       Assert.assertFalse(configuration.isLargeMessageSync());
    }
+
+   @Test
+   public void testParseFederationDownstreamAuthorization() throws Exception {
+      String middlePart = "<federations downstream-authorization=\"a, b,c\"/>";
+      String configStr = firstPart + middlePart + lastPart;
+
+      final FileConfigurationParser parser = new FileConfigurationParser();
+      final ByteArrayInputStream input = new ByteArrayInputStream(configStr.getBytes(StandardCharsets.UTF_8));
+
+      final Configuration configuration = parser.parseMainConfig(input);
+      Assert.assertTrue(configuration.getFederationDownstreamAuthorization().contains("a"));
+      Assert.assertTrue(configuration.getFederationDownstreamAuthorization().contains("b"));
+      Assert.assertTrue(configuration.getFederationDownstreamAuthorization().contains("c"));
+   }

    private static String firstPart = "<core xmlns=\"urn:activemq:core\">" + "\n" +
       "<name>ActiveMQ.main.config</name>" + "\n" +

--- a/pom.xml
+++ b/pom.xml
@@ -121,7 +121,7 @@
       <pem-keystore.version>2.3.0</pem-keystore.version>
       <!-- This overrides the Apache parent POM due to https://issues.apache.org/jira/browse/MSHADE-462. Once that issue
            is resolved and a new Apache parent is released then we can get rid of this. -->
-      <version.maven-shade-plugin>3.4.1</version.maven-shade-plugin>
+      <version.maven-shade-plugin>3.6.0</version.maven-shade-plugin>
 
       <!-- This overrides the Apache parent POM. Once a new Apache parent is released with the latest version then we
            can get rid of this. -->
