--- a/src/java/org/apache/commons/collections/functors/FunctorUtils.java
+++ b/src/java/org/apache/commons/collections/functors/FunctorUtils.java
@@ -16,6 +16,8 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 import java.util.Collection;
 import java.util.Iterator;
 
@@ -34,6 +36,10 @@
  */
 class FunctorUtils {
     
+    /** System property key to enable unsafe serialization */
+    final static String UNSAFE_SERIALIZABLE_PROPERTY
+        = "org.apache.commons.collections.enableUnsafeSerialization";
+    
     /**
      * Restricted constructor.
      */
@@ -149,6 +155,35 @@
                 throw new IllegalArgumentException(
                     "The transformer array must not contain a null transformer, index " + i + " was null");
             }
+        }
+    }
+
+    /**
+     * Package-private helper method to check if serialization support is
+     * enabled for unsafe classes.
+     *
+     * @param clazz  the clazz to check for serialization support
+     * @throws UnsupportedOperationException if unsafe serialization is disabled
+     */
+    static void checkUnsafeSerialization(Class clazz) {
+        String unsafeSerializableProperty;
+        
+        try {
+            unsafeSerializableProperty = 
+                (String) AccessController.doPrivileged(new PrivilegedAction() {
+                    public Object run() {
+                        return System.getProperty(UNSAFE_SERIALIZABLE_PROPERTY);
+                    }
+                });
+        } catch (SecurityException ex) {
+            unsafeSerializableProperty = null;
+        }
+
+        if (!"true".equalsIgnoreCase(unsafeSerializableProperty)) {
+            throw new UnsupportedOperationException(
+                    "Serialization support for " + clazz.getName() + " is disabled for security reasons. " +
+                    "To enable it set system property '" + UNSAFE_SERIALIZABLE_PROPERTY + "' to 'true', " +
+                    "but you must ensure that your application does not de-serialize objects from untrusted sources.");
         }
     }
 
--- a/src/java/org/apache/commons/collections/functors/InvokerTransformer.java
+++ b/src/java/org/apache/commons/collections/functors/InvokerTransformer.java
@@ -16,6 +16,9 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
@@ -25,6 +28,16 @@
 
 /**
  * Transformer implementation that creates a new object instance by reflection.
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code InvokerTransformer}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
  * 
  * @since Commons Collections 3.0
  * @version $Revision$ $Date$
@@ -35,7 +48,7 @@
 
     /** The serial version */
     private static final long serialVersionUID = -8653385846894047688L;
-    
+
     /** The method name to call */
     private final String iMethodName;
     /** The array of reflection parameter types */
@@ -134,4 +147,21 @@
         }
     }
 
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(InvokerTransformer.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(InvokerTransformer.class);
+        is.defaultReadObject();
+    }
 }
--- a/src/java/org/apache/commons/collections/functors/CloneTransformer.java
+++ b/src/java/org/apache/commons/collections/functors/CloneTransformer.java
@@ -16,6 +16,9 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 
 import org.apache.commons.collections.Transformer;
@@ -24,7 +27,17 @@
  * Transformer implementation that returns a clone of the input object.
  * <p>
  * Clone is performed using <code>PrototypeFactory.getInstance(input).create()</code>.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code CloneTransformer}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision$ $Date$
  *
@@ -68,4 +81,21 @@
         return PrototypeFactory.getInstance(input).create();
     }
 
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(CloneTransformer.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(CloneTransformer.class);
+        is.defaultReadObject();
+    }
 }
--- a/src/java/org/apache/commons/collections/functors/ForClosure.java
+++ b/src/java/org/apache/commons/collections/functors/ForClosure.java
@@ -16,13 +16,26 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 
 import org.apache.commons.collections.Closure;
 
 /**
  * Closure implementation that calls another closure n times, like a for loop.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code ForClosure}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision$ $Date$
  *
@@ -100,6 +113,24 @@
      */
     public int getCount() {
         return iCount;
+    }
+
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(ForClosure.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(ForClosure.class);
+        is.defaultReadObject();
     }
 
 }
--- a/src/java/org/apache/commons/collections/functors/InstantiateFactory.java
+++ b/src/java/org/apache/commons/collections/functors/InstantiateFactory.java
@@ -16,6 +16,9 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
@@ -25,7 +28,17 @@
 
 /**
  * Factory implementation that creates a new object instance by reflection.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code InstantiateTransformer}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision$ $Date$
  *
@@ -136,5 +149,23 @@
             throw new FunctorException("InstantiateFactory: Constructor threw an exception", ex);
         }
     }
-    
+
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(InstantiateFactory.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(InstantiateFactory.class);
+        is.defaultReadObject();
+    }
+
 }
--- a/src/java/org/apache/commons/collections/functors/InstantiateTransformer.java
+++ b/src/java/org/apache/commons/collections/functors/InstantiateTransformer.java
@@ -16,6 +16,9 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
@@ -25,7 +28,17 @@
 
 /**
  * Transformer implementation that creates a new object instance by reflection.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code InstantiateTransformer}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision$ $Date$
  *
@@ -114,6 +127,24 @@
         } catch (InvocationTargetException ex) {
             throw new FunctorException("InstantiateTransformer: Constructor threw an exception", ex);
         }
+    }
+
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(InstantiateTransformer.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(InstantiateTransformer.class);
+        is.defaultReadObject();
     }
 
 }
--- a/src/java/org/apache/commons/collections/functors/WhileClosure.java
+++ b/src/java/org/apache/commons/collections/functors/WhileClosure.java
@@ -16,6 +16,9 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 
 import org.apache.commons.collections.Closure;
@@ -24,7 +27,17 @@
 /**
  * Closure implementation that executes a closure repeatedly until a condition is met,
  * like a do-while or while loop.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code WhileClosure}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision$ $Date$
  *
@@ -118,6 +131,24 @@
      */
     public boolean isDoLoop() {
         return iDoLoop;
+    }
+
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(WhileClosure.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(WhileClosure.class);
+        is.defaultReadObject();
     }
 
 }
--- a/src/java/org/apache/commons/collections/functors/package.html
+++ b/src/java/org/apache/commons/collections/functors/package.html
@@ -24,4 +24,32 @@
 {@link org.apache.commons.collections.Factory Factory} interfaces.
 These provide simple callbacks for processing with collections.
 </p>
+<p>
+<b>WARNING:</b> from v3.2.2 onwards this several classes in this package will
+throw an {@link UnsupportedOperationException} when trying to serialize or
+de-serialize an instance to prevent potential remote code execution exploits.
+</p>
+<p>
+Classes considered to be unsafe are:
+</p>
+<ul>
+  <li>CloneTransformer</li>
+  <li>ForClosure</li>
+  <li>InstantiateFactory</li>
+  <li>InstantiateTransformer</li>
+  <li>InvokerTransformer</li>
+  <li>WhileClosure</li>
+</ul>
+<p>
+In order to re-enable serialization support for these unsafe classes, the
+following system property can be used (via -Dproperty=true):
+<pre>
+org.apache.commons.collections.enableUnsafeSerialization
+</pre>
+<p>
+Be warned though that enabling serialization support for these unsafe classes
+makes your application vulnerable to known exploits and you must ensure that
+your application does not de-serialize objects from untrusted sources.
+</p>
+
 </BODY>

diff --git a/build.xml b/build.xml
index cb3c5d6..1813980 100644
--- a/build.xml
+++ b/build.xml
@@ -36,6 +36,7 @@
   <property name="component.title.full"    value="Apache Commons Collections"/>
 
   <!-- The current version number of this component -->
+  <property name="version"                 value="3.2.1"/>
   <property name="component.version"       value="3.2.1"/>
 
   <!-- The base directory for component configuration files -->
@@ -57,7 +58,7 @@
   <property name="build.src"               value="${build.home}/src-ide" />
   
   <!-- The name/location of the jar file to build -->
-  <property name="final.name"           value="${component.name}-${component.version}"/>
+  <property name="final.name"           value="${component.name}-${version}"/>
   <property name="jar.name"             value="${final.name}.jar"/>
   <property name="build.jar.name"       value="${build.home}/${jar.name}"/>
   
@@ -147,7 +148,7 @@
         <filter token="name"     value="${component.name}"/>
         <filter token="title"    value="${component.title}"/>
         <filter token="package"  value="${component.package}"/>
-        <filter token="version"  value="${component.version}"/>
+        <filter token="version"  value="${version}"/>
       </filterset>
       <fileset dir="${source.conf}" includes="*.MF"/>
     </copy>
@@ -365,7 +366,8 @@
                overview="${source.java}/org/apache/commons/collections/overview.html"
                doctitle="${component.title} ${component.version} API;"
             windowtitle="${component.title} ${component.version} API"
-                 bottom="Copyright &amp;copy; 2001-${year} Apache Software Foundation. All Rights Reserved.">
+                 bottom="Copyright &amp;copy; 2001-${year} Apache Software Foundation. All Rights Reserved."
+        additionalparam="-Xdoclint:none">
     </javadoc>
   </target>
 
diff --git a/src/java/org/apache/commons/collections/MultiHashMap.java b/src/java/org/apache/commons/collections/MultiHashMap.java
index 25158b4..f9f5525 100644
--- a/src/java/org/apache/commons/collections/MultiHashMap.java
+++ b/src/java/org/apache/commons/collections/MultiHashMap.java
@@ -331,7 +331,7 @@ public class MultiHashMap extends HashMap implements MultiMap {
      * @param item  the value to remove
      * @return the value removed (which was passed in), null if nothing removed
      */
-    public Object remove(Object key, Object item) {
+    public Object removeMapping(Object key, Object item) {
         Collection valuesForKey = getCollection(key);
         if (valuesForKey == null) {
             return null;
diff --git a/src/java/org/apache/commons/collections/MultiMap.java b/src/java/org/apache/commons/collections/MultiMap.java
index 8cd7212..0dae4cc 100644
--- a/src/java/org/apache/commons/collections/MultiMap.java
+++ b/src/java/org/apache/commons/collections/MultiMap.java
@@ -66,7 +66,7 @@ public interface MultiMap extends Map {
      * @throws ClassCastException if the key or value is of an invalid type
      * @throws NullPointerException if the key or value is null and null is invalid
      */
-    public Object remove(Object key, Object item);
+    public Object removeMapping(Object key, Object item);
 
     //-----------------------------------------------------------------------
     /**
diff --git a/src/java/org/apache/commons/collections/map/MultiKeyMap.java b/src/java/org/apache/commons/collections/map/MultiKeyMap.java
index 1ba1cea..15e779f 100644
--- a/src/java/org/apache/commons/collections/map/MultiKeyMap.java
+++ b/src/java/org/apache/commons/collections/map/MultiKeyMap.java
@@ -197,7 +197,7 @@ public class MultiKeyMap
      * @param key2  the second key
      * @return the value mapped to the removed key, null if key not in map
      */
-    public Object remove(Object key1, Object key2) {
+    public Object removeMultiKey(Object key1, Object key2) {
         int hashCode = hash(key1, key2);
         int index = map.hashIndex(hashCode, map.data.length);
         AbstractHashedMap.HashEntry entry = map.data[index];
diff --git a/src/java/org/apache/commons/collections/map/MultiValueMap.java b/src/java/org/apache/commons/collections/map/MultiValueMap.java
index cc74fd7..2b811a7 100644
--- a/src/java/org/apache/commons/collections/map/MultiValueMap.java
+++ b/src/java/org/apache/commons/collections/map/MultiValueMap.java
@@ -153,7 +153,7 @@ public class MultiValueMap extends AbstractMapDecorator implements MultiMap {
      * @param value the value to remove
      * @return the value removed (which was passed in), null if nothing removed
      */
-    public Object remove(Object key, Object value) {
+    public Object removeMapping(Object key, Object value) {
         Collection valuesForKey = getCollection(key);
         if (valuesForKey == null) {
             return null;
diff --git a/src/test/org/apache/commons/collections/TestMultiHashMap.java b/src/test/org/apache/commons/collections/TestMultiHashMap.java
index 722bec0..0b09b0a 100644
--- a/src/test/org/apache/commons/collections/TestMultiHashMap.java
+++ b/src/test/org/apache/commons/collections/TestMultiHashMap.java
@@ -217,7 +217,7 @@ public class TestMultiHashMap extends AbstractTestMap {
         MultiHashMap one = new MultiHashMap();
         Integer value = new Integer(1);
         one.put("One", value);
-        one.remove("One", value);
+        one.removeMapping("One", value);
         
         MultiHashMap two = new MultiHashMap();
         assertEquals(two, one);
@@ -269,7 +269,7 @@ public class TestMultiHashMap extends AbstractTestMap {
         assertEquals(4, map.totalSize());
         map.remove("A");
         assertEquals(3, map.totalSize());
-        map.remove("B", "BC");
+        map.removeMapping("B", "BC");
         assertEquals(2, map.totalSize());
     }
     
@@ -292,7 +292,7 @@ public class TestMultiHashMap extends AbstractTestMap {
         map.remove("A");
         assertEquals(0, map.size("A"));
         assertEquals(3, map.size("B"));
-        map.remove("B", "BC");
+        map.removeMapping("B", "BC");
         assertEquals(0, map.size("A"));
         assertEquals(2, map.size("B"));
     }
@@ -464,11 +464,11 @@ public class TestMultiHashMap extends AbstractTestMap {
         map.put("A", "AA");
         map.put("A", "AB");
         map.put("A", "AC");
-        assertEquals(null, map.remove("C", "CA"));
-        assertEquals(null, map.remove("A", "AD"));
-        assertEquals("AC", map.remove("A", "AC"));
-        assertEquals("AB", map.remove("A", "AB"));
-        assertEquals("AA", map.remove("A", "AA"));
+        assertEquals(null, map.removeMapping("C", "CA"));
+        assertEquals(null, map.removeMapping("A", "AD"));
+        assertEquals("AC", map.removeMapping("A", "AC"));
+        assertEquals("AB", map.removeMapping("A", "AB"));
+        assertEquals("AA", map.removeMapping("A", "AA"));
         assertEquals(new MultiHashMap(), map);
     }
 
diff --git a/src/test/org/apache/commons/collections/map/TestMultiKeyMap.java b/src/test/org/apache/commons/collections/map/TestMultiKeyMap.java
index 6e528fb..0854069 100644
--- a/src/test/org/apache/commons/collections/map/TestMultiKeyMap.java
+++ b/src/test/org/apache/commons/collections/map/TestMultiKeyMap.java
@@ -315,10 +315,10 @@ public class TestMultiKeyMap extends AbstractTestIterableMap {
             switch (key.size()) {
                 case 2:
                 assertEquals(true, multimap.containsKey(key.getKey(0), key.getKey(1)));
-                assertEquals(value, multimap.remove(key.getKey(0), key.getKey(1)));
+                assertEquals(value, multimap.removeMultiKey(key.getKey(0), key.getKey(1)));
                 assertEquals(false, multimap.containsKey(key.getKey(0), key.getKey(1)));
                 assertEquals(size - 1, multimap.size());
-                assertEquals(null, multimap.remove(key.getKey(0), key.getKey(1)));
+                assertEquals(null, multimap.removeMultiKey(key.getKey(0), key.getKey(1)));
                 assertEquals(false, multimap.containsKey(key.getKey(0), key.getKey(1)));
                 break;
                 case 3:
diff --git a/src/test/org/apache/commons/collections/map/TestMultiValueMap.java b/src/test/org/apache/commons/collections/map/TestMultiValueMap.java
index 7b322dc..439a14a 100644
--- a/src/test/org/apache/commons/collections/map/TestMultiValueMap.java
+++ b/src/test/org/apache/commons/collections/map/TestMultiValueMap.java
@@ -1,338 +1,338 @@
-/*
- * Licensed to the Apache Software Foundation (ASF) under one or more
- * contributor license agreements.  See the NOTICE file distributed with
- * this work for additional information regarding copyright ownership.
- * The ASF licenses this file to You under the Apache License, Version 2.0
- * (the "License"); you may not use this file except in compliance with
- * the License.  You may obtain a copy of the License at
- *
- *      http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-package org.apache.commons.collections.map;
-
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.LinkedList;
-import java.util.Map;
-
-import junit.framework.Test;
-import junit.framework.TestCase;
-import junit.framework.TestSuite;
-
-import org.apache.commons.collections.IteratorUtils;
-import org.apache.commons.collections.MultiMap;
-import org.apache.commons.collections.TestMultiHashMap;
-
-/**
- * TestMultiValueMap.
- *
- * @author <a href="mailto:jcarman@apache.org">James Carman</a>
- * @author Stephen Colebourne
- * @since Commons Collections 3.2
- */
-public class TestMultiValueMap extends TestCase {
-
-    public TestMultiValueMap(String testName) {
-        super(testName);
-    }
-
-    public static Test suite() {
-        return new TestSuite(TestMultiHashMap.class);
-    }
-
-    public static void main(String args[]) {
-        String[] testCaseName = { TestMultiHashMap.class.getName()};
-        junit.textui.TestRunner.main(testCaseName);
-    }
-
-    public void testNoMappingReturnsNull() {
-        final MultiValueMap map = createTestMap();
-        assertNull(map.get("whatever"));
-    }
-
-    public void testValueCollectionType() {
-        final MultiValueMap map = createTestMap(LinkedList.class);
-        assertTrue(map.get("one") instanceof LinkedList);
-    }
-
-    public void testMultipleValues() {
-        final MultiValueMap map = createTestMap(HashSet.class);
-        final HashSet expected = new HashSet();
-        expected.add("uno");
-        expected.add("un");
-        assertEquals(expected, map.get("one"));
-    }
-
-    public void testContainsValue() {
-        final MultiValueMap map = createTestMap(HashSet.class);
-        assertTrue(map.containsValue("uno"));
-        assertTrue(map.containsValue("un"));
-        assertTrue(map.containsValue("dos"));
-        assertTrue(map.containsValue("deux"));
-        assertTrue(map.containsValue("tres"));
-        assertTrue(map.containsValue("trois"));
-        assertFalse(map.containsValue("quatro"));
-    }
-
-    public void testKeyContainsValue() {
-        final MultiValueMap map = createTestMap(HashSet.class);
-        assertTrue(map.containsValue("one", "uno"));
-        assertTrue(map.containsValue("one", "un"));
-        assertTrue(map.containsValue("two", "dos"));
-        assertTrue(map.containsValue("two", "deux"));
-        assertTrue(map.containsValue("three", "tres"));
-        assertTrue(map.containsValue("three", "trois"));
-        assertFalse(map.containsValue("four", "quatro"));
-    }
-
-    public void testValues() {
-        final MultiValueMap map = createTestMap(HashSet.class);
-        final HashSet expected = new HashSet();
-        expected.add("uno");
-        expected.add("dos");
-        expected.add("tres");
-        expected.add("un");
-        expected.add("deux");
-        expected.add("trois");
-        final Collection c = map.values();
-        assertEquals(6, c.size());
-        assertEquals(expected, new HashSet(c));
-    }
-
-    private MultiValueMap createTestMap() {
-        return createTestMap(ArrayList.class);
-    }
-
-    private MultiValueMap createTestMap(Class collectionClass) {
-        final MultiValueMap map = MultiValueMap.decorate(new HashMap(), collectionClass);
-        map.put("one", "uno");
-        map.put("one", "un");
-        map.put("two", "dos");
-        map.put("two", "deux");
-        map.put("three", "tres");
-        map.put("three", "trois");
-        return map;
-    }
-
-    public void testKeyedIterator() {
-        final MultiValueMap map = createTestMap();
-        final ArrayList actual = new ArrayList(IteratorUtils.toList(map.iterator("one")));
-        final ArrayList expected = new ArrayList(Arrays.asList(new String[]{"uno", "un"}));
-        assertEquals(expected, actual);
-    }
-
-    public void testRemoveAllViaIterator() {
-        final MultiValueMap map = createTestMap();
-        for(Iterator i = map.values().iterator(); i.hasNext();) {
-            i.next();
-            i.remove();
-        }
-        assertNull(map.get("one"));
-        assertTrue(map.isEmpty());
-    }
-
-    public void testRemoveAllViaKeyedIterator() {
-        final MultiValueMap map = createTestMap();
-        for(Iterator i = map.iterator("one"); i.hasNext();) {
-            i.next();
-            i.remove();
-        }
-        assertNull(map.get("one"));
-        assertEquals(4, map.totalSize());
-    }
-
-    public void testTotalSizeA() {
-        assertEquals(6, createTestMap().totalSize());
-    }
-
-    //-----------------------------------------------------------------------
-    public void testMapEquals() {
-        MultiValueMap one = new MultiValueMap();
-        Integer value = new Integer(1);
-        one.put("One", value);
-        one.remove("One", value);
-        
-        MultiValueMap two = new MultiValueMap();
-        assertEquals(two, one);
-    }
-
-    //-----------------------------------------------------------------------
-    public void testGetCollection() {
-        MultiValueMap map = new MultiValueMap();
-        map.put("A", "AA");
-        assertSame(map.get("A"), map.getCollection("A"));
-    }
-    
-    public void testTotalSize() {
-        MultiValueMap map = new MultiValueMap();
-        assertEquals(0, map.totalSize());
-        map.put("A", "AA");
-        assertEquals(1, map.totalSize());
-        map.put("B", "BA");
-        assertEquals(2, map.totalSize());
-        map.put("B", "BB");
-        assertEquals(3, map.totalSize());
-        map.put("B", "BC");
-        assertEquals(4, map.totalSize());
-        map.remove("A");
-        assertEquals(3, map.totalSize());
-        map.remove("B", "BC");
-        assertEquals(2, map.totalSize());
-    }
-    
-    public void testSize() {
-        MultiValueMap map = new MultiValueMap();
-        assertEquals(0, map.size());
-        map.put("A", "AA");
-        assertEquals(1, map.size());
-        map.put("B", "BA");
-        assertEquals(2, map.size());
-        map.put("B", "BB");
-        assertEquals(2, map.size());
-        map.put("B", "BC");
-        assertEquals(2, map.size());
-        map.remove("A");
-        assertEquals(2, map.size());
-        map.remove("B", "BC");
-        assertEquals(2, map.size());
-    }
-    
-    public void testSize_Key() {
-        MultiValueMap map = new MultiValueMap();
-        assertEquals(0, map.size("A"));
-        assertEquals(0, map.size("B"));
-        map.put("A", "AA");
-        assertEquals(1, map.size("A"));
-        assertEquals(0, map.size("B"));
-        map.put("B", "BA");
-        assertEquals(1, map.size("A"));
-        assertEquals(1, map.size("B"));
-        map.put("B", "BB");
-        assertEquals(1, map.size("A"));
-        assertEquals(2, map.size("B"));
-        map.put("B", "BC");
-        assertEquals(1, map.size("A"));
-        assertEquals(3, map.size("B"));
-        map.remove("A");
-        assertEquals(0, map.size("A"));
-        assertEquals(3, map.size("B"));
-        map.remove("B", "BC");
-        assertEquals(0, map.size("A"));
-        assertEquals(2, map.size("B"));
-    }
-    
-    public void testIterator_Key() {
-        MultiValueMap map = new MultiValueMap();
-        assertEquals(false, map.iterator("A").hasNext());
-        map.put("A", "AA");
-        Iterator it = map.iterator("A");
-        assertEquals(true, it.hasNext());
-        it.next();
-        assertEquals(false, it.hasNext());
-    }
-    
-    public void testContainsValue_Key() {
-        MultiValueMap map = new MultiValueMap();
-        assertEquals(false, map.containsValue("A", "AA"));
-        assertEquals(false, map.containsValue("B", "BB"));
-        map.put("A", "AA");
-        assertEquals(true, map.containsValue("A", "AA"));
-        assertEquals(false, map.containsValue("A", "AB"));
-    }
-
-    public void testPutAll_Map1() {
-        MultiMap original = new MultiValueMap();
-        original.put("key", "object1");
-        original.put("key", "object2");
-
-        MultiValueMap test = new MultiValueMap();
-        test.put("keyA", "objectA");
-        test.put("key", "object0");
-        test.putAll(original);
-
-        assertEquals(2, test.size());
-        assertEquals(4, test.totalSize());
-        assertEquals(1, test.getCollection("keyA").size());
-        assertEquals(3, test.getCollection("key").size());
-        assertEquals(true, test.containsValue("objectA"));
-        assertEquals(true, test.containsValue("object0"));
-        assertEquals(true, test.containsValue("object1"));
-        assertEquals(true, test.containsValue("object2"));
-    }
-
-    public void testPutAll_Map2() {
-        Map original = new HashMap();
-        original.put("keyX", "object1");
-        original.put("keyY", "object2");
-
-        MultiValueMap test = new MultiValueMap();
-        test.put("keyA", "objectA");
-        test.put("keyX", "object0");
-        test.putAll(original);
-
-        assertEquals(3, test.size());
-        assertEquals(4, test.totalSize());
-        assertEquals(1, test.getCollection("keyA").size());
-        assertEquals(2, test.getCollection("keyX").size());
-        assertEquals(1, test.getCollection("keyY").size());
-        assertEquals(true, test.containsValue("objectA"));
-        assertEquals(true, test.containsValue("object0"));
-        assertEquals(true, test.containsValue("object1"));
-        assertEquals(true, test.containsValue("object2"));
-    }
-
-    public void testPutAll_KeyCollection() {
-        MultiValueMap map = new MultiValueMap();
-        Collection coll = Arrays.asList(new Object[] {"X", "Y", "Z"});
-        
-        assertEquals(true, map.putAll("A", coll));
-        assertEquals(3, map.size("A"));
-        assertEquals(true, map.containsValue("A", "X"));
-        assertEquals(true, map.containsValue("A", "Y"));
-        assertEquals(true, map.containsValue("A", "Z"));
-        
-        assertEquals(false, map.putAll("A", null));
-        assertEquals(3, map.size("A"));
-        assertEquals(true, map.containsValue("A", "X"));
-        assertEquals(true, map.containsValue("A", "Y"));
-        assertEquals(true, map.containsValue("A", "Z"));
-        
-        assertEquals(false, map.putAll("A", new ArrayList()));
-        assertEquals(3, map.size("A"));
-        assertEquals(true, map.containsValue("A", "X"));
-        assertEquals(true, map.containsValue("A", "Y"));
-        assertEquals(true, map.containsValue("A", "Z"));
-        
-        coll = Arrays.asList(new Object[] {"M"});
-        assertEquals(true, map.putAll("A", coll));
-        assertEquals(4, map.size("A"));
-        assertEquals(true, map.containsValue("A", "X"));
-        assertEquals(true, map.containsValue("A", "Y"));
-        assertEquals(true, map.containsValue("A", "Z"));
-        assertEquals(true, map.containsValue("A", "M"));
-    }
-
-    public void testRemove_KeyItem() {
-        MultiValueMap map = new MultiValueMap();
-        map.put("A", "AA");
-        map.put("A", "AB");
-        map.put("A", "AC");
-        assertEquals(null, map.remove("C", "CA"));
-        assertEquals(null, map.remove("A", "AD"));
-        assertEquals("AC", map.remove("A", "AC"));
-        assertEquals("AB", map.remove("A", "AB"));
-        assertEquals("AA", map.remove("A", "AA"));
-        assertEquals(new MultiValueMap(), map);
-    }
-
-}
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.collections.map;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Map;
+
+import junit.framework.Test;
+import junit.framework.TestCase;
+import junit.framework.TestSuite;
+
+import org.apache.commons.collections.IteratorUtils;
+import org.apache.commons.collections.MultiMap;
+import org.apache.commons.collections.TestMultiHashMap;
+
+/**
+ * TestMultiValueMap.
+ *
+ * @author <a href="mailto:jcarman@apache.org">James Carman</a>
+ * @author Stephen Colebourne
+ * @since Commons Collections 3.2
+ */
+public class TestMultiValueMap extends TestCase {
+
+    public TestMultiValueMap(String testName) {
+        super(testName);
+    }
+
+    public static Test suite() {
+        return new TestSuite(TestMultiHashMap.class);
+    }
+
+    public static void main(String args[]) {
+        String[] testCaseName = { TestMultiHashMap.class.getName()};
+        junit.textui.TestRunner.main(testCaseName);
+    }
+
+    public void testNoMappingReturnsNull() {
+        final MultiValueMap map = createTestMap();
+        assertNull(map.get("whatever"));
+    }
+
+    public void testValueCollectionType() {
+        final MultiValueMap map = createTestMap(LinkedList.class);
+        assertTrue(map.get("one") instanceof LinkedList);
+    }
+
+    public void testMultipleValues() {
+        final MultiValueMap map = createTestMap(HashSet.class);
+        final HashSet expected = new HashSet();
+        expected.add("uno");
+        expected.add("un");
+        assertEquals(expected, map.get("one"));
+    }
+
+    public void testContainsValue() {
+        final MultiValueMap map = createTestMap(HashSet.class);
+        assertTrue(map.containsValue("uno"));
+        assertTrue(map.containsValue("un"));
+        assertTrue(map.containsValue("dos"));
+        assertTrue(map.containsValue("deux"));
+        assertTrue(map.containsValue("tres"));
+        assertTrue(map.containsValue("trois"));
+        assertFalse(map.containsValue("quatro"));
+    }
+
+    public void testKeyContainsValue() {
+        final MultiValueMap map = createTestMap(HashSet.class);
+        assertTrue(map.containsValue("one", "uno"));
+        assertTrue(map.containsValue("one", "un"));
+        assertTrue(map.containsValue("two", "dos"));
+        assertTrue(map.containsValue("two", "deux"));
+        assertTrue(map.containsValue("three", "tres"));
+        assertTrue(map.containsValue("three", "trois"));
+        assertFalse(map.containsValue("four", "quatro"));
+    }
+
+    public void testValues() {
+        final MultiValueMap map = createTestMap(HashSet.class);
+        final HashSet expected = new HashSet();
+        expected.add("uno");
+        expected.add("dos");
+        expected.add("tres");
+        expected.add("un");
+        expected.add("deux");
+        expected.add("trois");
+        final Collection c = map.values();
+        assertEquals(6, c.size());
+        assertEquals(expected, new HashSet(c));
+    }
+
+    private MultiValueMap createTestMap() {
+        return createTestMap(ArrayList.class);
+    }
+
+    private MultiValueMap createTestMap(Class collectionClass) {
+        final MultiValueMap map = MultiValueMap.decorate(new HashMap(), collectionClass);
+        map.put("one", "uno");
+        map.put("one", "un");
+        map.put("two", "dos");
+        map.put("two", "deux");
+        map.put("three", "tres");
+        map.put("three", "trois");
+        return map;
+    }
+
+    public void testKeyedIterator() {
+        final MultiValueMap map = createTestMap();
+        final ArrayList actual = new ArrayList(IteratorUtils.toList(map.iterator("one")));
+        final ArrayList expected = new ArrayList(Arrays.asList(new String[]{"uno", "un"}));
+        assertEquals(expected, actual);
+    }
+
+    public void testRemoveAllViaIterator() {
+        final MultiValueMap map = createTestMap();
+        for(Iterator i = map.values().iterator(); i.hasNext();) {
+            i.next();
+            i.remove();
+        }
+        assertNull(map.get("one"));
+        assertTrue(map.isEmpty());
+    }
+
+    public void testRemoveAllViaKeyedIterator() {
+        final MultiValueMap map = createTestMap();
+        for(Iterator i = map.iterator("one"); i.hasNext();) {
+            i.next();
+            i.remove();
+        }
+        assertNull(map.get("one"));
+        assertEquals(4, map.totalSize());
+    }
+
+    public void testTotalSizeA() {
+        assertEquals(6, createTestMap().totalSize());
+    }
+
+    //-----------------------------------------------------------------------
+    public void testMapEquals() {
+        MultiValueMap one = new MultiValueMap();
+        Integer value = new Integer(1);
+        one.put("One", value);
+        one.removeMapping("One", value);
+        
+        MultiValueMap two = new MultiValueMap();
+        assertEquals(two, one);
+    }
+
+    //-----------------------------------------------------------------------
+    public void testGetCollection() {
+        MultiValueMap map = new MultiValueMap();
+        map.put("A", "AA");
+        assertSame(map.get("A"), map.getCollection("A"));
+    }
+    
+    public void testTotalSize() {
+        MultiValueMap map = new MultiValueMap();
+        assertEquals(0, map.totalSize());
+        map.put("A", "AA");
+        assertEquals(1, map.totalSize());
+        map.put("B", "BA");
+        assertEquals(2, map.totalSize());
+        map.put("B", "BB");
+        assertEquals(3, map.totalSize());
+        map.put("B", "BC");
+        assertEquals(4, map.totalSize());
+        map.remove("A");
+        assertEquals(3, map.totalSize());
+        map.removeMapping("B", "BC");
+        assertEquals(2, map.totalSize());
+    }
+    
+    public void testSize() {
+        MultiValueMap map = new MultiValueMap();
+        assertEquals(0, map.size());
+        map.put("A", "AA");
+        assertEquals(1, map.size());
+        map.put("B", "BA");
+        assertEquals(2, map.size());
+        map.put("B", "BB");
+        assertEquals(2, map.size());
+        map.put("B", "BC");
+        assertEquals(2, map.size());
+        map.remove("A");
+        assertEquals(2, map.size());
+        map.removeMapping("B", "BC");
+        assertEquals(2, map.size());
+    }
+    
+    public void testSize_Key() {
+        MultiValueMap map = new MultiValueMap();
+        assertEquals(0, map.size("A"));
+        assertEquals(0, map.size("B"));
+        map.put("A", "AA");
+        assertEquals(1, map.size("A"));
+        assertEquals(0, map.size("B"));
+        map.put("B", "BA");
+        assertEquals(1, map.size("A"));
+        assertEquals(1, map.size("B"));
+        map.put("B", "BB");
+        assertEquals(1, map.size("A"));
+        assertEquals(2, map.size("B"));
+        map.put("B", "BC");
+        assertEquals(1, map.size("A"));
+        assertEquals(3, map.size("B"));
+        map.remove("A");
+        assertEquals(0, map.size("A"));
+        assertEquals(3, map.size("B"));
+        map.removeMapping("B", "BC");
+        assertEquals(0, map.size("A"));
+        assertEquals(2, map.size("B"));
+    }
+    
+    public void testIterator_Key() {
+        MultiValueMap map = new MultiValueMap();
+        assertEquals(false, map.iterator("A").hasNext());
+        map.put("A", "AA");
+        Iterator it = map.iterator("A");
+        assertEquals(true, it.hasNext());
+        it.next();
+        assertEquals(false, it.hasNext());
+    }
+    
+    public void testContainsValue_Key() {
+        MultiValueMap map = new MultiValueMap();
+        assertEquals(false, map.containsValue("A", "AA"));
+        assertEquals(false, map.containsValue("B", "BB"));
+        map.put("A", "AA");
+        assertEquals(true, map.containsValue("A", "AA"));
+        assertEquals(false, map.containsValue("A", "AB"));
+    }
+
+    public void testPutAll_Map1() {
+        MultiMap original = new MultiValueMap();
+        original.put("key", "object1");
+        original.put("key", "object2");
+
+        MultiValueMap test = new MultiValueMap();
+        test.put("keyA", "objectA");
+        test.put("key", "object0");
+        test.putAll(original);
+
+        assertEquals(2, test.size());
+        assertEquals(4, test.totalSize());
+        assertEquals(1, test.getCollection("keyA").size());
+        assertEquals(3, test.getCollection("key").size());
+        assertEquals(true, test.containsValue("objectA"));
+        assertEquals(true, test.containsValue("object0"));
+        assertEquals(true, test.containsValue("object1"));
+        assertEquals(true, test.containsValue("object2"));
+    }
+
+    public void testPutAll_Map2() {
+        Map original = new HashMap();
+        original.put("keyX", "object1");
+        original.put("keyY", "object2");
+
+        MultiValueMap test = new MultiValueMap();
+        test.put("keyA", "objectA");
+        test.put("keyX", "object0");
+        test.putAll(original);
+
+        assertEquals(3, test.size());
+        assertEquals(4, test.totalSize());
+        assertEquals(1, test.getCollection("keyA").size());
+        assertEquals(2, test.getCollection("keyX").size());
+        assertEquals(1, test.getCollection("keyY").size());
+        assertEquals(true, test.containsValue("objectA"));
+        assertEquals(true, test.containsValue("object0"));
+        assertEquals(true, test.containsValue("object1"));
+        assertEquals(true, test.containsValue("object2"));
+    }
+
+    public void testPutAll_KeyCollection() {
+        MultiValueMap map = new MultiValueMap();
+        Collection coll = Arrays.asList(new Object[] {"X", "Y", "Z"});
+        
+        assertEquals(true, map.putAll("A", coll));
+        assertEquals(3, map.size("A"));
+        assertEquals(true, map.containsValue("A", "X"));
+        assertEquals(true, map.containsValue("A", "Y"));
+        assertEquals(true, map.containsValue("A", "Z"));
+        
+        assertEquals(false, map.putAll("A", null));
+        assertEquals(3, map.size("A"));
+        assertEquals(true, map.containsValue("A", "X"));
+        assertEquals(true, map.containsValue("A", "Y"));
+        assertEquals(true, map.containsValue("A", "Z"));
+        
+        assertEquals(false, map.putAll("A", new ArrayList()));
+        assertEquals(3, map.size("A"));
+        assertEquals(true, map.containsValue("A", "X"));
+        assertEquals(true, map.containsValue("A", "Y"));
+        assertEquals(true, map.containsValue("A", "Z"));
+        
+        coll = Arrays.asList(new Object[] {"M"});
+        assertEquals(true, map.putAll("A", coll));
+        assertEquals(4, map.size("A"));
+        assertEquals(true, map.containsValue("A", "X"));
+        assertEquals(true, map.containsValue("A", "Y"));
+        assertEquals(true, map.containsValue("A", "Z"));
+        assertEquals(true, map.containsValue("A", "M"));
+    }
+
+    public void testRemove_KeyItem() {
+        MultiValueMap map = new MultiValueMap();
+        map.put("A", "AA");
+        map.put("A", "AB");
+        map.put("A", "AC");
+        assertEquals(null, map.removeMapping("C", "CA"));
+        assertEquals(null, map.removeMapping("A", "AD"));
+        assertEquals("AC", map.removeMapping("A", "AC"));
+        assertEquals("AB", map.removeMapping("A", "AB"));
+        assertEquals("AA", map.removeMapping("A", "AA"));
+        assertEquals(new MultiValueMap(), map);
+    }
+
+}
