--- a/lib/cli.js
+++ b/lib/cli.js
@@ -63776,6 +63776,11 @@
       chown: false, // don't chown. just leave as it is
       map: function map(header) {
         header.mtime = now;
+        if (header.linkname) {
+          var basePath = path.posix.dirname(path.join('/', header.name));
+          var jailPath = path.posix.join(basePath, header.linkname);
+          header.linkname = path.posix.relative('/', jailPath);
+        }
         return header;
       },
       fs: patchedFs

--- a/lib/cli.js
+++ b/lib/cli.js
@@ -25770,7 +25770,27 @@
 
 var exec = exports.exec = (0, (_promise2 || _load_promise2()).promisify)(child.exec);
 
+function validate(program) {
+  var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
+
+  if (program.includes('/')) {
+    return true;
+  }
+
+  var cwd = opts.cwd || process.cwd();
+  var pathext = process.env.PATHEXT || '';
+  var exts = pathext.split(';');
+
+  for (var i = 0; i < exts.length; i++) {
+    var candidate = require('path').join(cwd, `${program}${exts[i]}`);
+    if (require('fs').existsSync(candidate)) {
+      throw new Error(`Potentially dangerous call to "${program}" in ${cwd}`);
+    }
+  }
+}
+
 function forkp(program, args, opts) {
+  validate(program, opts);
   return new (_promise || _load_promise()).default(function (resolve, reject) {
     var proc = child.fork(program, args, opts);
 
@@ -25785,6 +25805,7 @@
 }
 
 function spawnp(program, args, opts) {
+  validate(program, opts);
   return new (_promise || _load_promise()).default(function (resolve, reject) {
     var proc = child.spawn(program, args, opts);
 
@@ -25826,6 +25847,8 @@
   var key = opts.cwd || String(++uid);
   return queue.push(key, function () {
     return new (_promise || _load_promise()).default(function (resolve, reject) {
+      validate(program, opts);
+
       var proc = child.spawn(program, args, opts);
       spawnedProcesses[key] = proc;
 
