--- a/src/Document.jsx
+++ b/src/Document.jsx
@@ -115,7 +115,7 @@

         const { options, onLoadProgress, onPassword } = this.props;

-        this.loadingTask = pdfjs.getDocument({ ...source, ...options });
+        this.loadingTask = pdfjs.getDocument({ ...source, ...options, isEvalSupported: false });
         this.loadingTask.onPassword = onPassword;
         if (onLoadProgress) {
           this.loadingTask.onProgress = onLoadProgress;
