--- a/lib/path/evaluation.js
+++ b/lib/path/evaluation.js
@@ -285,7 +285,10 @@
       const property = callee.get("property");
       if (object.isIdentifier() && property.isIdentifier() && isValidCallee(object.node.name) && !isInvalidMethod(property.node.name)) {
         context = global[object.node.name];
-        func = context[property.node.name];
+        const key = property.node.name;
+        if (Object.hasOwnProperty.call(context, key)) {
+          func = context[key];
+        }
       }
       if (object.isLiteral() && property.isIdentifier()) {
         const type = typeof object.node.value;
