diff --git a/src/main/java/com/fasterxml/jackson/core/json/async/NonBlockingUtf8JsonParserBase.java b/src/main/java/com/fasterxml/jackson/core/json/async/NonBlockingUtf8JsonParserBase.java
index 0f9594dce..d750c0e2d 100644
--- a/src/main/java/com/fasterxml/jackson/core/json/async/NonBlockingUtf8JsonParserBase.java
+++ b/src/main/java/com/fasterxml/jackson/core/json/async/NonBlockingUtf8JsonParserBase.java
@@ -2,6 +2,7 @@ package com.fasterxml.jackson.core.json.async;
 
 import com.fasterxml.jackson.core.JsonParseException;
 import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.core.exc.StreamConstraintsException;
 import com.fasterxml.jackson.core.io.CharTypes;
 import com.fasterxml.jackson.core.io.IOContext;
 import com.fasterxml.jackson.core.json.JsonReadFeature;
@@ -362,7 +363,7 @@ public abstract class NonBlockingUtf8JsonParserBase
                 if (_numberNegative) {
                     --len;
                 }
-                _intLength = len;
+                _setIntLength(len);
             }
             return _valueComplete(JsonToken.VALUE_NUMBER_INT);
 
@@ -1299,7 +1300,7 @@ public abstract class NonBlockingUtf8JsonParserBase
         while (true) {
             if (ch < INT_0) {
                 if (ch == INT_PERIOD) {
-                    _intLength = outPtr;
+                    _setIntLength(outPtr);
                     ++_inputPtr;
                     return _startFloat(outBuf, outPtr, ch);
                 }
@@ -1307,7 +1308,7 @@ public abstract class NonBlockingUtf8JsonParserBase
             }
             if (ch > INT_9) {
                 if ((ch | 0x20) == INT_e) { // ~ 'eE'
-                    _intLength = outPtr;
+                    _setIntLength(outPtr);
                     ++_inputPtr;
                     return _startFloat(outBuf, outPtr, ch);
                 }
@@ -1326,7 +1327,7 @@ public abstract class NonBlockingUtf8JsonParserBase
             }
             ch = getByteFromBuffer(_inputPtr) & 0xFF;
         }
-        _intLength = outPtr;
+        _setIntLength(outPtr);
         _textBuffer.setCurrentLength(outPtr);
         return _valueComplete(JsonToken.VALUE_NUMBER_INT);
     }
@@ -1366,7 +1367,7 @@ public abstract class NonBlockingUtf8JsonParserBase
         while (true) {
             if (ch < INT_0) {
                 if (ch == INT_PERIOD) {
-                    _intLength = outPtr-1;
+                    _setIntLength(outPtr-1);
                     ++_inputPtr;
                     return _startFloat(outBuf, outPtr, ch);
                 }
@@ -1374,7 +1375,7 @@ public abstract class NonBlockingUtf8JsonParserBase
             }
             if (ch > INT_9) {
                 if ((ch | 0x20) == INT_e) { // ~ 'eE'
-                    _intLength = outPtr-1;
+                    _setIntLength(outPtr-1);
                     ++_inputPtr;
                     return _startFloat(outBuf, outPtr, ch);
                 }
@@ -1392,7 +1393,7 @@ public abstract class NonBlockingUtf8JsonParserBase
             }
             ch = getByteFromBuffer(_inputPtr) & 0xFF;
         }
-        _intLength = outPtr-1;
+        _setIntLength(outPtr-1);
         _textBuffer.setCurrentLength(outPtr);
         return _valueComplete(JsonToken.VALUE_NUMBER_INT);
     }
@@ -1438,7 +1439,7 @@ public abstract class NonBlockingUtf8JsonParserBase
         while (true) {
             if (ch < INT_0) {
                 if (ch == INT_PERIOD) {
-                    _intLength = outPtr-1;
+                    _setIntLength(outPtr-1);
                     ++_inputPtr;
                     return _startFloat(outBuf, outPtr, ch);
                 }
@@ -1446,7 +1447,7 @@ public abstract class NonBlockingUtf8JsonParserBase
             }
             if (ch > INT_9) {
                 if ((ch | 0x20) == INT_e) { // ~ 'eE'
-                    _intLength = outPtr-1;
+                    _setIntLength(outPtr-1);
                     ++_inputPtr;
                     return _startFloat(outBuf, outPtr, ch);
                 }
@@ -1464,7 +1465,7 @@ public abstract class NonBlockingUtf8JsonParserBase
             }
             ch = getByteFromBuffer(_inputPtr) & 0xFF;
         }
-        _intLength = outPtr-1;
+        _setIntLength(outPtr-1);
         _textBuffer.setCurrentLength(outPtr);
         return _valueComplete(JsonToken.VALUE_NUMBER_INT);
     }
@@ -1696,7 +1697,7 @@ public abstract class NonBlockingUtf8JsonParserBase
             int ch = getByteFromBuffer(_inputPtr) & 0xFF;
             if (ch < INT_0) {
                 if (ch == INT_PERIOD) {
-                    _intLength = outPtr+negMod;
+                    _setIntLength(outPtr+negMod);
                     ++_inputPtr;
                     return _startFloat(outBuf, outPtr, ch);
                 }
@@ -1704,7 +1705,7 @@ public abstract class NonBlockingUtf8JsonParserBase
             }
             if (ch > INT_9) {
                 if ((ch | 0x20) == INT_e) { // ~ 'eE'
-                    _intLength = outPtr+negMod;
+                    _setIntLength(outPtr+negMod);
                     ++_inputPtr;
                     return _startFloat(outBuf, outPtr, ch);
                 }
@@ -1718,7 +1719,7 @@ public abstract class NonBlockingUtf8JsonParserBase
             }
             outBuf[outPtr++] = (char) ch;
         }
-        _intLength = outPtr+negMod;
+        _setIntLength(outPtr+negMod);
         _textBuffer.setCurrentLength(outPtr);
         return _valueComplete(JsonToken.VALUE_NUMBER_INT);
     }
@@ -1735,7 +1736,7 @@ public abstract class NonBlockingUtf8JsonParserBase
                 if (_inputPtr >= _inputEnd) {
                     _textBuffer.setCurrentLength(outPtr);
                     _minorState = MINOR_NUMBER_FRACTION_DIGITS;
-                    _fractLength = fractLen;
+                    _setFractLength(fractLen);
                     return (_currToken = JsonToken.NOT_AVAILABLE);
                 }
                 ch = getNextSignedByteFromBuffer(); // ok to have sign extension for now
@@ -1756,7 +1757,7 @@ public abstract class NonBlockingUtf8JsonParserBase
                 ++fractLen;
             }
         }
-        _fractLength = fractLen;
+        _setFractLength(fractLen);
         int expLen = 0;
         if ((ch | 0x20) == INT_e) { // ~ 'eE' exponent?
             if (outPtr >= outBuf.length) {
@@ -1792,7 +1793,7 @@ public abstract class NonBlockingUtf8JsonParserBase
                 if (_inputPtr >= _inputEnd) {
                     _textBuffer.setCurrentLength(outPtr);
                     _minorState = MINOR_NUMBER_EXPONENT_DIGITS;
-                    _expLength = expLen;
+                    _setExpLength(expLen);
                     return (_currToken = JsonToken.NOT_AVAILABLE);
                 }
                 ch = getNextSignedByteFromBuffer();
@@ -1807,7 +1808,7 @@ public abstract class NonBlockingUtf8JsonParserBase
         --_inputPtr;
         _textBuffer.setCurrentLength(outPtr);
         // negative, int-length, fract-length already set, so...
-        _expLength = expLen;
+        _setExpLength(expLen);
         return _valueComplete(JsonToken.VALUE_NUMBER_FLOAT);
     }
 
@@ -1829,7 +1830,7 @@ public abstract class NonBlockingUtf8JsonParserBase
                 outBuf[outPtr++] = (char) ch;
                 if (_inputPtr >= _inputEnd) {
                     _textBuffer.setCurrentLength(outPtr);
-                    _fractLength = fractLen;
+                    _setFractLength(fractLen);
                     return JsonToken.NOT_AVAILABLE;
                 }
                 ch = getNextSignedByteFromBuffer();
@@ -1849,7 +1850,7 @@ public abstract class NonBlockingUtf8JsonParserBase
                 _reportUnexpectedNumberChar(ch, "Decimal point not followed by a digit");
             }
         }
-        _fractLength = fractLen;
+        _setFractLength(fractLen);
         _textBuffer.setCurrentLength(outPtr);
 
         // Ok: end of floating point number or exponent?
@@ -1899,7 +1900,7 @@ public abstract class NonBlockingUtf8JsonParserBase
             outBuf[outPtr++] = (char) ch;
             if (_inputPtr >= _inputEnd) {
                 _textBuffer.setCurrentLength(outPtr);
-                _expLength = expLen;
+                _setExpLength(expLen);
                 return JsonToken.NOT_AVAILABLE;
             }
             ch = getNextSignedByteFromBuffer();
@@ -1913,7 +1914,7 @@ public abstract class NonBlockingUtf8JsonParserBase
         --_inputPtr;
         _textBuffer.setCurrentLength(outPtr);
         // negative, int-length, fract-length already set, so...
-        _expLength = expLen;
+        _setExpLength(expLen);
         return _valueComplete(JsonToken.VALUE_NUMBER_FLOAT);
     }
 
@@ -2993,4 +2994,21 @@ public abstract class NonBlockingUtf8JsonParserBase
     /* Internal methods, other
     /**********************************************************************
      */
+
+    private void _setIntLength(final int len) throws StreamConstraintsException {
+        _streamReadConstraints.validateIntegerLength(len);
+        _intLength = len;
+    }
+
+    private void _setFractLength(final int len) throws StreamConstraintsException {
+        // assumes that the _intLength has been updated first
+        _streamReadConstraints.validateFPLength(_intLength + len);
+        _fractLength = len;
+    }
+
+    private void _setExpLength(final int len) throws StreamConstraintsException {
+        // assumes that the _intLength and _fractLength have been updated already
+        _streamReadConstraints.validateFPLength(_intLength + _fractLength + len);
+        _expLength = len;
+    }
 }
diff --git a/src/test/java/com/fasterxml/jackson/core/constraints/AsyncLargeNumberReadTest.java b/src/test/java/com/fasterxml/jackson/core/constraints/AsyncLargeNumberReadTest.java
new file mode 100644
index 000000000..63eebef2c
--- /dev/null
+++ b/src/test/java/com/fasterxml/jackson/core/constraints/AsyncLargeNumberReadTest.java
@@ -0,0 +1,107 @@
+package com.fasterxml.jackson.core.constraints;
+
+import com.fasterxml.jackson.core.async.ByteArrayFeeder;
+import org.junit.jupiter.api.Test;
+
+import com.fasterxml.jackson.core.*;
+import com.fasterxml.jackson.core.exc.StreamConstraintsException;
+import com.fasterxml.jackson.core.JsonFactory;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * Number Length Constraint Bypass in Non-Blocking (Async) JSON Parsers
+ */
+class AsyncLargeNumberReadTest
+    extends com.fasterxml.jackson.core.JUnit5TestBase
+{
+    private static final int TEST_NUMBER_LENGTH = StreamReadConstraints.DEFAULT_MAX_NUM_LEN * 2;
+
+    private final JsonFactory JSON_F = newStreamFactory();
+
+    @Test
+    void asyncParserFailsTooLongInt() throws Exception {
+        byte[] payload = buildPayloadWithLongInteger(TEST_NUMBER_LENGTH);
+
+        try (JsonParser p = JSON_F.createNonBlockingByteArrayParser()) {
+            ByteArrayFeeder byteArrayFeeder = (ByteArrayFeeder) p;
+            byteArrayFeeder.feedInput(payload, 0, payload.length);
+            byteArrayFeeder.endOfInput();
+    
+            _asyncParserFailsTooLongNumber(p, JsonToken.VALUE_NUMBER_INT);
+        }
+    }
+
+    @Test
+    void asyncParserFailsTooLongDecimal() throws Exception {
+        byte[] payload = buildPayloadWithLongDecimal(TEST_NUMBER_LENGTH);
+
+        try (JsonParser p = JSON_F.createNonBlockingByteArrayParser()) {
+            ByteArrayFeeder byteArrayFeeder = (ByteArrayFeeder) p;
+            byteArrayFeeder.feedInput(payload, 0, payload.length);
+            byteArrayFeeder.endOfInput();
+
+            _asyncParserFailsTooLongNumber(p, JsonToken.VALUE_NUMBER_FLOAT);
+        }
+    }
+
+    @Test
+    void asyncParserFailsTooLongDecimalWithExponent() throws Exception {
+        byte[] payload = buildPayloadWithLongExponent(TEST_NUMBER_LENGTH);
+
+        try (JsonParser p = JSON_F.createNonBlockingByteArrayParser()) {
+            ByteArrayFeeder byteArrayFeeder = (ByteArrayFeeder) p;
+            byteArrayFeeder.feedInput(payload, 0, payload.length);
+            byteArrayFeeder.endOfInput();
+
+            _asyncParserFailsTooLongNumber(p, JsonToken.VALUE_NUMBER_FLOAT);
+        }
+    }
+    
+    private void _asyncParserFailsTooLongNumber(JsonParser p, JsonToken tokenMatch) throws Exception {
+        boolean foundNumber = false;
+        try {
+            while (p.nextToken() != null) {
+                if (p.currentToken() == tokenMatch) {
+                    foundNumber = true;
+                    String numberText = p.getText();
+                    assertEquals(TEST_NUMBER_LENGTH, numberText.length(),
+                            "Async parser silently accepted all " + TEST_NUMBER_LENGTH + " digits");
+                }
+            }
+            fail("Async parser must reject a " + TEST_NUMBER_LENGTH + "-digit number (number found? "+foundNumber+")");
+        } catch (StreamConstraintsException e) {
+            verifyException(e, "Number value length (");
+            verifyException(e, " exceeds the maximum allowed");
+        }
+    }
+
+    private byte[] buildPayloadWithLongInteger(int numDigits) {
+        StringBuilder sb = new StringBuilder(numDigits + 10);
+        sb.append("{\"v\":");
+        for (int i = 0; i < numDigits; i++) {
+            sb.append((char) ('1' + (i % 9)));
+        }
+        sb.append('}');
+        return utf8Bytes(sb.toString());
+    }
+
+    private byte[] buildPayloadWithLongDecimal(int numDigits) {
+        StringBuilder sb = new StringBuilder(numDigits + 10);
+        sb.append("{\"v\":0.");
+        for (int i = 0; i < numDigits; i++) {
+            sb.append((char) ('1' + (i % 9)));
+        }
+        sb.append('}');
+        return utf8Bytes(sb.toString());
+    }
+
+    private byte[] buildPayloadWithLongExponent(int numDigits) {
+        StringBuilder sb = new StringBuilder(numDigits + 10);
+        sb.append("{\"v\":1.1E");
+        for (int i = 0; i < numDigits; i++) {
+            sb.append((char) ('1' + (i % 9)));
+        }
+        return utf8Bytes(sb.toString());
+    }
+}
diff --git a/src/test/java/com/fasterxml/jackson/core/constraints/LargeNumberReadTest.java b/src/test/java/com/fasterxml/jackson/core/constraints/LargeNumberReadTest.java
index 458f8ab3b..b86a3ba66 100644
--- a/src/test/java/com/fasterxml/jackson/core/constraints/LargeNumberReadTest.java
+++ b/src/test/java/com/fasterxml/jackson/core/constraints/LargeNumberReadTest.java
@@ -14,7 +14,6 @@ import static org.junit.jupiter.api.Assertions.fail;
  * Set of basic unit tests for verifying that "too big" number constraints
  * are caught by various JSON parser backends.
  */
-
 @SuppressWarnings("resource")
 class LargeNumberReadTest
         extends com.fasterxml.jackson.core.JUnit5TestBase
