--- a/dist/cjs/runHttpQuery.js
+++ b/dist/cjs/runHttpQuery.js
@@ -9,6 +9,7 @@
 const internalErrorClasses_js_1 = require("./internalErrorClasses.js");
 const negotiator_1 = __importDefault(require("negotiator"));
 const HeaderMap_js_1 = require("./utils/HeaderMap.js");
+const whatwg_mimetype_1 = __importDefault(require("whatwg-mimetype"));
 function fieldIfString(o, fieldName) {
     const value = o[fieldName];
     if (typeof value === 'string') {
@@ -112,6 +113,14 @@
             break;
         }
         case 'GET': {
+            const contentType = httpRequest.headers.get('content-type');
+            if (contentType !== undefined) {
+                const contentTypeParsed = whatwg_mimetype_1.default.parse(contentType);
+                if (contentTypeParsed === null ||
+                    contentTypeParsed.essence !== 'application/json') {
+                    throw new internalErrorClasses_js_1.BadRequestError('GET requests may not have a content-type header other than application/json.', { extensions: { http: newHTTPGraphQLHead(415) } });
+                }
+            }
             const searchParams = new URLSearchParams(httpRequest.search);
             graphQLRequest = {
                 query: searchParamIfSpecifiedOnce(searchParams, 'query'),
--- a/dist/esm/runHttpQuery.js
+++ b/dist/esm/runHttpQuery.js
@@ -3,6 +3,7 @@
 import { BadRequestError } from './internalErrorClasses.js';
 import Negotiator from 'negotiator';
 import { HeaderMap } from './utils/HeaderMap.js';
+import MIMEType from 'whatwg-mimetype';
 function fieldIfString(o, fieldName) {
     const value = o[fieldName];
     if (typeof value === 'string') {
@@ -106,6 +107,14 @@
             break;
         }
         case 'GET': {
+            const contentType = httpRequest.headers.get('content-type');
+            if (contentType !== undefined) {
+                const contentTypeParsed = MIMEType.parse(contentType);
+                if (contentTypeParsed === null ||
+                    contentTypeParsed.essence !== 'application/json') {
+                    throw new BadRequestError('GET requests may not have a content-type header other than application/json.', { extensions: { http: newHTTPGraphQLHead(415) } });
+                }
+            }
             const searchParams = new URLSearchParams(httpRequest.search);
             graphQLRequest = {
                 query: searchParamIfSpecifiedOnce(searchParams, 'query'),
--- a/src/runHttpQuery.ts
+++ b/src/runHttpQuery.ts
@@ -20,6 +20,7 @@
 import { BadRequestError } from './internalErrorClasses.js';
 import Negotiator from 'negotiator';
 import { HeaderMap } from './utils/HeaderMap.js';
+import MIMEType from 'whatwg-mimetype';
 
 function fieldIfString(
   o: Record<string, unknown>,
@@ -194,6 +195,20 @@
     }
 
     case 'GET': {
+      const contentType = httpRequest.headers.get('content-type');
+      if (contentType !== undefined) {
+        const contentTypeParsed = MIMEType.parse(contentType);
+        if (
+          contentTypeParsed === null ||
+          contentTypeParsed.essence !== 'application/json'
+        ) {
+          throw new BadRequestError(
+            'GET requests may not have a content-type header other than application/json.',
+            { extensions: { http: newHTTPGraphQLHead(415) } },
+          );
+        }
+      }
+
       const searchParams = new URLSearchParams(httpRequest.search);
 
       graphQLRequest = {
