--- a/activemq-broker/src/main/java/org/apache/activemq/broker/jmx/BrokerView.java
+++ b/activemq-broker/src/main/java/org/apache/activemq/broker/jmx/BrokerView.java
@@ -19,6 +19,7 @@
 import java.io.File;
 import java.io.IOException;
 import java.net.URI;
+import java.net.URISyntaxException;
 import java.util.*;
 import java.util.concurrent.atomic.AtomicInteger;
 
@@ -35,6 +36,7 @@
 import org.apache.activemq.command.*;
 import org.apache.activemq.network.NetworkConnector;
 import org.apache.activemq.util.BrokerSupport;
+import org.apache.activemq.util.URISupport;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -370,6 +372,8 @@
 
     @Override
     public String addConnector(String discoveryAddress) throws Exception {
+        // Verify VM transport is not used
+        validateAllowedUrl(discoveryAddress);
         TransportConnector connector = brokerService.addConnector(discoveryAddress);
         if (connector == null) {
             throw new NoSuchElementException("Not connector matched the given name: " + discoveryAddress);
@@ -380,6 +384,8 @@
 
     @Override
     public String addNetworkConnector(String discoveryAddress) throws Exception {
+        // Verify VM transport is not used
+        validateAllowedUrl(discoveryAddress);
         NetworkConnector connector = brokerService.addNetworkConnector(discoveryAddress);
         if (connector == null) {
             throw new NoSuchElementException("Not connector matched the given name: " + discoveryAddress);
@@ -556,4 +562,42 @@
     public long getTotalMaxUncommittedExceededCount() {
         return safeGetBroker().getDestinationStatistics().getMaxUncommittedExceededCount().getCount();
 	}
+
+    private static void validateAllowedUrl(String uriString) throws URISyntaxException {
+        validateAllowedUri(new URI(uriString), 0);
+    }
+
+    // Validate the URI does not contain VM transport
+    private static void validateAllowedUri(URI uri, int depth) throws URISyntaxException {
+        // Don't allow more than 5 nested URIs to prevent blowing the stack
+        if (depth > 5) {
+            throw new IllegalArgumentException("URI can't contain more than 5 nested composite URIs");
+        }
+
+        // First check the main URI scheme
+        validateAllowedScheme(uri.getScheme());
+
+        // If composite, iterate and check each of the composite URIs
+        if (URISupport.isCompositeURI(uri)) {
+            URISupport.CompositeData data = URISupport.parseComposite(uri);
+            depth++;
+            for (URI component : data.getComponents()) {
+                // Each URI could be a nested composite URI so call validateAllowedUri()
+                // to validate it. This check if composite first so we don't add to
+                // the recursive stack depth if there's a lot of URIs that are not composite
+                if (URISupport.isCompositeURI(uri)) {
+                    validateAllowedUri(component, depth);
+                } else {
+                    validateAllowedScheme(uri.getScheme());
+                }
+            }
+        }
+    }
+
+    // We don't allow VM transport scheme to be used
+    private static void validateAllowedScheme(String scheme) {
+        if (scheme.equals("vm")) {
+            throw new IllegalArgumentException("VM scheme is not allowed");
+        }
+    }
 }
--- a/activemq-unit-tests/src/test/java/org/apache/activemq/broker/jmx/MBeanTest.java
+++ b/activemq-unit-tests/src/test/java/org/apache/activemq/broker/jmx/MBeanTest.java
@@ -16,6 +16,9 @@
  */
 package org.apache.activemq.broker.jmx;
 
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
 import java.io.BufferedReader;
 import java.io.InputStreamReader;
 import java.net.URI;
@@ -64,6 +67,7 @@
 import org.apache.activemq.util.JMXSupport;
 import org.apache.activemq.util.URISupport;
 import org.apache.activemq.util.Wait;
+import org.junit.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -2055,4 +2059,45 @@
         assertFalse(subscription.isRetroactive());
         assertTrue(subscription.isExclusive());
     }
-}
\ No newline at end of file
+
+    // Test to verify VM transport is not allowed to be added as a connector
+    // through the Broker MBean
+    public void testAddVmConnectorBlockedBrokerView() throws Exception {
+        ObjectName brokerName = assertRegisteredObjectName(domain + ":type=Broker,brokerName=localhost");
+        BrokerViewMBean brokerView = MBeanServerInvocationHandler.newProxyInstance(mbeanServer, brokerName, BrokerViewMBean.class, true);
+
+        try {
+            brokerView.addConnector("vm://localhost");
+            fail("Should have failed trying to add vm connector");
+        } catch (IllegalArgumentException e) {
+            assertEquals("VM scheme is not allowed", e.getMessage());
+        }
+
+        try {
+            // verify any composite URI is blocked as well
+            brokerView.addConnector("failover:(tcp://0.0.0.0:0,vm://" + brokerName + ")");
+            fail("Should have failed trying to add vm connector");
+        } catch (IllegalArgumentException e) {
+            assertEquals("VM scheme is not allowed", e.getMessage());
+        }
+
+        try {
+            // verify nested composite URI is blocked
+            brokerView.addConnector("failover:(failover:(failover:(vm://localhost)))");
+            fail("Should have failed trying to add vm connector");
+        } catch (IllegalArgumentException e) {
+            assertEquals("VM scheme is not allowed", e.getMessage());
+        }
+
+        try {
+            // verify nested composite URI with more than 5 levels is blocked
+            brokerView.addConnector(
+                    "static:(failover:(failover:(failover:(failover:(failover:(tcp://localhost:0))))))");
+            fail("Should have failed trying to add vm connector bridge");
+        } catch (IllegalArgumentException e) {
+            assertEquals("URI can't contain more than 5 nested composite URIs", e.getMessage());
+        }
+
+    }
+
+}
--- a/activemq-unit-tests/src/test/java/org/apache/activemq/jmx/JmxCreateNCTest.java
+++ b/activemq-unit-tests/src/test/java/org/apache/activemq/jmx/JmxCreateNCTest.java
@@ -19,12 +19,16 @@
 import org.apache.activemq.broker.BrokerService;
 import org.apache.activemq.broker.jmx.BrokerViewMBean;
 import org.apache.activemq.broker.jmx.NetworkConnectorViewMBean;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
 import org.junit.Test;
 
 import javax.management.ObjectName;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.fail;
 
 /**
  * This test shows that when we create a network connector via JMX,
@@ -36,35 +40,89 @@
 
     private static final String BROKER_NAME = "jmx-broker";
 
-    @Test
-    public void testBridgeRegistration() throws Exception {
+    private BrokerService broker;
+    private BrokerViewMBean proxy;
 
-        System.setProperty("org.apache.activemq.audit", "all");
-
-        BrokerService broker = new BrokerService();
+    @Before
+    public void setUp() throws Exception {
+        broker = new BrokerService();
         broker.setBrokerName(BROKER_NAME);
         broker.setUseJmx(true); // explicitly set this so no funny issues
         broker.start();
         broker.waitUntilStarted();
 
-        // now create network connector over JMX
         ObjectName brokerObjectName = new ObjectName("org.apache.activemq:type=Broker,brokerName=" + BROKER_NAME);
-        BrokerViewMBean proxy = (BrokerViewMBean) broker.getManagementContext().newProxyInstance(brokerObjectName,
+        proxy = (BrokerViewMBean) broker.getManagementContext().newProxyInstance(brokerObjectName,
                 BrokerViewMBean.class, true);
+    }
 
+    @After
+    public void tearDown() throws Exception {
+        broker.stop();
+        broker.waitUntilStopped();
+    }
+
+    @Test
+    public void testBridgeRegistration() throws Exception {
         assertNotNull("We could not retrieve the broker from JMX", proxy);
 
         // let's add the NC
-        String connectoName = proxy.addNetworkConnector("static:(tcp://localhost:61617)");
-        assertEquals("NC", connectoName);
+        String connectorName = proxy.addNetworkConnector("static:(tcp://localhost:61617)");
+        assertEquals("NC", connectorName);
 
         // Make sure we can retrieve the NC through JMX
         ObjectName networkConnectorObjectName = new ObjectName("org.apache.activemq:type=Broker,brokerName=" + BROKER_NAME +
-                ",connector=networkConnectors,networkConnectorName=" + connectoName);
+                ",connector=networkConnectors,networkConnectorName=" + connectorName);
         NetworkConnectorViewMBean nc  = (NetworkConnectorViewMBean) broker.getManagementContext().newProxyInstance(networkConnectorObjectName,
                 NetworkConnectorViewMBean.class, true);
 
         assertNotNull(nc);
         assertEquals("NC", nc.getName());
+    }
+
+    @Test
+    public void testVmBridgeBlocked() throws Exception {
+        // Test composite network connector uri
+        try {
+            proxy.addNetworkConnector("static:(vm://localhost)");
+            fail("Should have failed trying to add vm connector bridge");
+        } catch (IllegalArgumentException e) {
+            assertEquals("VM scheme is not allowed", e.getMessage());
+        }
+
+        try {
+            proxy.addNetworkConnector("multicast:(vm://localhost)");
+            fail("Should have failed trying to add vm connector bridge");
+        } catch (IllegalArgumentException e) {
+            assertEquals("VM scheme is not allowed", e.getMessage());
+        }
+
+        // verify direct vm as well
+        try {
+            proxy.addNetworkConnector("vm://localhost");
+            fail("Should have failed trying to add vm connector bridge");
+        } catch (IllegalArgumentException e) {
+            assertEquals("VM scheme is not allowed", e.getMessage());
+        }
+
+        try {
+            // verify nested composite URI is blocked
+            proxy.addNetworkConnector("static:(failover:(failover:(tcp://localhost:0,vm://localhost)))");
+            fail("Should have failed trying to add vm connector bridge");
+        } catch (IllegalArgumentException e) {
+            assertEquals("VM scheme is not allowed", e.getMessage());
+        }
+    }
+
+    @Test
+    public void testAddNetworkConnectorMaxComposite() throws Exception {
+        try {
+            // verify nested composite URI with more than 5 levels is blocked
+            proxy.addNetworkConnector(
+                    "static:(failover:(failover:(failover:(failover:(failover:(tcp://localhost:0))))))");
+            fail("Should have failed trying to add vm connector bridge");
+        } catch (IllegalArgumentException e) {
+            assertEquals("URI can't contain more than 5 nested composite URIs", e.getMessage());
+        }
     }
 }

--- a/activemq-broker/src/main/java/org/apache/activemq/broker/jmx/BrokerView.java
+++ b/activemq-broker/src/main/java/org/apache/activemq/broker/jmx/BrokerView.java
@@ -44,6 +44,8 @@ public class BrokerView implements BrokerViewMBean {
 
     private static final Logger LOG = LoggerFactory.getLogger(BrokerView.class);
 
+    private static final Set<String> DENIED_TRANSPORT_SCHEMES = Set.of("vm", "http", "https", "ftp");
+
     ManagedRegionBroker broker;
 
     private final BrokerService brokerService;
@@ -372,7 +374,7 @@ public class BrokerView implements BrokerViewMBean {
 
     @Override
     public String addConnector(String discoveryAddress) throws Exception {
-        // Verify VM transport is not used
+        // Verify a denied transport scheme is not used
         validateAllowedUrl(discoveryAddress);
         TransportConnector connector = brokerService.addConnector(discoveryAddress);
         if (connector == null) {
@@ -384,7 +386,7 @@ public class BrokerView implements BrokerViewMBean {
 
     @Override
     public String addNetworkConnector(String discoveryAddress) throws Exception {
-        // Verify VM transport is not used
+        // Verify a denied transport scheme is not used
         validateAllowedUrl(discoveryAddress);
         NetworkConnector connector = brokerService.addNetworkConnector(discoveryAddress);
         if (connector == null) {
@@ -567,10 +569,11 @@ public class BrokerView implements BrokerViewMBean {
         validateAllowedUri(new URI(uriString), 0);
     }
 
-    // Validate the URI does not contain VM transport
+    // Validate the URI does not contain a denied transport scheme
     private static void validateAllowedUri(URI uri, int depth) throws URISyntaxException {
         // Don't allow more than 5 nested URIs to prevent blowing the stack
-        if (depth > 5) {
+        // If we are greater than 4 then this is the 5th level of composite
+        if (depth > 4) {
             throw new IllegalArgumentException("URI can't contain more than 5 nested composite URIs");
         }
 
@@ -585,19 +588,22 @@ public class BrokerView implements BrokerViewMBean {
                 // Each URI could be a nested composite URI so call validateAllowedUri()
                 // to validate it. This check if composite first so we don't add to
                 // the recursive stack depth if there's a lot of URIs that are not composite
-                if (URISupport.isCompositeURI(uri)) {
+                if (URISupport.isCompositeURI(component)) {
                     validateAllowedUri(component, depth);
                 } else {
-                    validateAllowedScheme(uri.getScheme());
+                    validateAllowedScheme(component.getScheme());
                 }
             }
         }
     }
 
-    // We don't allow VM transport scheme to be used
+    // Check all denied schemes
     private static void validateAllowedScheme(String scheme) {
-        if (scheme.equals("vm")) {
-            throw new IllegalArgumentException("VM scheme is not allowed");
+        for (String denied : DENIED_TRANSPORT_SCHEMES) {
+            // The schemes should be case-insensitive but ignore case as a precaution
+            if (scheme.equalsIgnoreCase(denied)) {
+                throw new IllegalArgumentException("Transport scheme '" + scheme + "' is not allowed");
+            }
         }
     }
 }

--- a/activemq-broker/src/main/java/org/apache/activemq/broker/BrokerService.java
+++ b/activemq-broker/src/main/java/org/apache/activemq/broker/BrokerService.java
@@ -1053,7 +1053,7 @@
     }
 
 
-    private static final String brokerNameReplacedCharsRegExp = "[^a-zA-Z0-9\\.\\_\\-\\:]";
+    private static final String INVALID_BROKER_NAME_CHAR_REG_EXP = "[^a-zA-Z0-9\\.\\_\\-\\:]";
 
     /**
      * Sets the name of this broker; which must be unique in the network.
@@ -1062,9 +1062,9 @@
         if (brokerName == null) {
             throw new NullPointerException("The broker name cannot be null");
         }
-        String str = brokerName.replaceAll(brokerNameReplacedCharsRegExp, "_");
+        String str = brokerName.replaceAll(INVALID_BROKER_NAME_CHAR_REG_EXP, "_");
         if (!str.equals(brokerName)) {
-            LOG.error("Broker Name: {} contained illegal characters matching regExp: {} - replaced with {}", brokerName, brokerNameReplacedCharsRegExp, str);
+            LOG.error("Broker Name: {} contained illegal characters matching regExp: {} - replaced with {}", brokerName, INVALID_BROKER_NAME_CHAR_REG_EXP, str);
         }
         this.brokerName = str.trim();
     }
@@ -2396,7 +2396,6 @@
         }
         destinationFactory.setRegionBroker(regionBroker);
         regionBroker.setKeepDurableSubsActive(keepDurableSubsActive);
-        regionBroker.setBrokerName(getBrokerName());
         regionBroker.getDestinationStatistics().setEnabled(enableStatistics);
         regionBroker.setAllowTempAutoCreationOnSend(isAllowTempAutoCreationOnSend());
         if (brokerId != null) {
--- a/activemq-broker/src/main/java/org/apache/activemq/broker/jmx/BrokerView.java
+++ b/activemq-broker/src/main/java/org/apache/activemq/broker/jmx/BrokerView.java
@@ -44,7 +44,9 @@
 
     private static final Logger LOG = LoggerFactory.getLogger(BrokerView.class);
 
-    private static final Set<String> DENIED_TRANSPORT_SCHEMES = Set.of("vm", "http", "https", "ftp");
+    public static final Set<String> DENIED_TRANSPORT_SCHEMES = Set.of("vm", "http",
+            "multicast", "zeroconf", "discovery", "fanout", "mock", "peer", "failover",
+            "proxy", "reliable", "simple", "udp");
 
     ManagedRegionBroker broker;
 
--- a/activemq-broker/src/main/java/org/apache/activemq/broker/jmx/DestinationView.java
+++ b/activemq-broker/src/main/java/org/apache/activemq/broker/jmx/DestinationView.java
@@ -17,6 +17,7 @@
 package org.apache.activemq.broker.jmx;
 
 import java.io.IOException;
+import java.net.URI;
 import java.net.URISyntaxException;
 import java.util.ArrayList;
 import java.util.Collections;
@@ -385,7 +386,7 @@
     @Override
     public String sendTextMessage(Map<String, String> headers, String body, String userName, @Sensitive String password) throws Exception {
 
-        String brokerUrl = "vm://" + broker.getBrokerName();
+        URI brokerUrl = broker.getVmConnectorURI();
         ActiveMQDestination dest = destination.getActiveMQDestination();
 
         ActiveMQConnectionFactory cf = new ActiveMQConnectionFactory(brokerUrl);
--- a/activemq-broker/src/main/java/org/apache/activemq/broker/region/RegionBroker.java
+++ b/activemq-broker/src/main/java/org/apache/activemq/broker/region/RegionBroker.java
@@ -24,6 +24,7 @@
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CopyOnWriteArrayList;
@@ -104,7 +105,7 @@
 
     private final LongSequenceGenerator sequenceGenerator = new LongSequenceGenerator();
     private BrokerId brokerId;
-    private String brokerName;
+    private final String brokerName;
     private final Map<String, ConnectionContext> clientIdSet = new HashMap<String, ConnectionContext>();
     private final DestinationInterceptor destinationInterceptor;
     private ConnectionContext adminConnectionContext;
@@ -138,7 +139,8 @@
 
     public RegionBroker(BrokerService brokerService, TaskRunnerFactory taskRunnerFactory, SystemUsage memoryManager, DestinationFactory destinationFactory,
         DestinationInterceptor destinationInterceptor, Scheduler scheduler, ThreadPoolExecutor executor) throws IOException {
-        this.brokerService = brokerService;
+        this.brokerService = Objects.requireNonNull(brokerService);
+        this.brokerName = Objects.requireNonNull(brokerService.getBrokerName(), "The broker name cannot be null");
         this.executor = executor;
         this.scheduler = scheduler;
         if (destinationFactory == null) {
@@ -564,20 +566,9 @@
 
     @Override
     public String getBrokerName() {
-        if (brokerName == null) {
-            try {
-                brokerName = InetAddressUtil.getLocalHostName().toLowerCase(Locale.ENGLISH);
-            } catch (Exception e) {
-                brokerName = "localhost";
-            }
-        }
         return brokerName;
     }
 
-    public void setBrokerName(String brokerName) {
-        this.brokerName = brokerName;
-    }
-
     public DestinationStatistics getDestinationStatistics() {
         return destinationStatistics;
     }
--- a/activemq-broker/src/test/java/org/apache/activemq/broker/BrokerServiceTest.java
+++ b/activemq-broker/src/test/java/org/apache/activemq/broker/BrokerServiceTest.java
@@ -17,6 +17,7 @@
 package org.apache.activemq.broker;
 
 import org.junit.Test;
+import static org.junit.Assert.assertEquals;
 
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
@@ -57,4 +58,43 @@
 
         assertTrue("Pre-shutdown hook should have been invoked", hook.invoked);
     }
+
+    @Test
+    public void testSetBrokerNameInvalidChars() {
+        final BrokerService brokerService = new BrokerService();
+
+        // All valid
+        brokerService.setBrokerName("valid");
+        assertEquals("valid", brokerService.getBrokerName());
+        brokerService.setBrokerName("valid123");
+        assertEquals("valid123", brokerService.getBrokerName());
+        brokerService.setBrokerName("this_is_valid");
+        assertEquals("this_is_valid", brokerService.getBrokerName());
+        brokerService.setBrokerName("this_123_valid");
+        assertEquals("this_123_valid", brokerService.getBrokerName());
+        brokerService.setBrokerName("valid-name123");
+        assertEquals("valid-name123", brokerService.getBrokerName());
+        brokerService.setBrokerName("1235.6789");
+        assertEquals("1235.6789", brokerService.getBrokerName());
+        brokerService.setBrokerName("valid:123");
+        assertEquals("valid:123", brokerService.getBrokerName());
+
+        // Test invalid names
+        brokerService.setBrokerName("abc?bad");
+        assertEquals("abc_bad", brokerService.getBrokerName());
+        brokerService.setBrokerName("#");
+        assertEquals("_", brokerService.getBrokerName());
+        brokerService.setBrokerName("?");
+        assertEquals("_", brokerService.getBrokerName());
+        brokerService.setBrokerName("invalid%");
+        assertEquals("invalid_", brokerService.getBrokerName());
+        brokerService.setBrokerName("\\");
+        assertEquals("_", brokerService.getBrokerName());
+        brokerService.setBrokerName("<>");
+        assertEquals("__", brokerService.getBrokerName());
+        brokerService.setBrokerName("abc=");
+        assertEquals("abc_", brokerService.getBrokerName());
+        brokerService.setBrokerName("name:abc=?bad");
+        assertEquals("name:abc__bad", brokerService.getBrokerName());
+    }
 }
--- a/activemq-spring/src/main/java/org/apache/activemq/xbean/XBeanBrokerFactory.java
+++ b/activemq-spring/src/main/java/org/apache/activemq/xbean/XBeanBrokerFactory.java
@@ -75,12 +75,10 @@
         if (broker == null) {
             // lets try find by type
             String[] names = context.getBeanNamesForType(BrokerService.class);
-            for (int i = 0; i < names.length; i++) {
-                String name = names[i];
-                broker = (BrokerService)context.getBean(name);
-                if (broker != null) {
-                    break;
-                }
+            for (String name : names) {
+                // No need to check for null, this will throw an exception if not found
+                broker = (BrokerService) context.getBean(name);
+                break;
             }
         }
         if (broker == null) {
--- a/activemq-unit-tests/src/test/java/org/apache/activemq/broker/jmx/MBeanTest.java
+++ b/activemq-unit-tests/src/test/java/org/apache/activemq/broker/jmx/MBeanTest.java
@@ -16,6 +16,7 @@
  */
 package org.apache.activemq.broker.jmx;
 
+import static org.apache.activemq.broker.jmx.BrokerView.DENIED_TRANSPORT_SCHEMES;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.fail;
 
@@ -2060,44 +2061,52 @@
         assertTrue(subscription.isExclusive());
     }
 
-    // Test to verify VM transport is not allowed to be added as a connector
+    // Test to verify blocked transport schemes are not allowed to be added as a connector
     // through the Broker MBean
-    public void testAddVmConnectorBlockedBrokerView() throws Exception {
+    public void testAddConnectorBlockedBrokerView() throws Exception {
+        for (String deniedScheme : DENIED_TRANSPORT_SCHEMES) {
+            LOG.info("verify testAddConnectorBlockedBrokerView scheme: {}", deniedScheme);
+            testAddTransportConnectorBlockedBrokerView(deniedScheme);
+        }
+    }
+
+    protected void testAddTransportConnectorBlockedBrokerView(String scheme) throws Exception {
         ObjectName brokerName = assertRegisteredObjectName(domain + ":type=Broker,brokerName=localhost");
         BrokerViewMBean brokerView = MBeanServerInvocationHandler.newProxyInstance(mbeanServer, brokerName, BrokerViewMBean.class, true);
 
         try {
-            brokerView.addConnector("vm://localhost");
-            fail("Should have failed trying to add vm connector");
+            brokerView.addConnector(scheme + "://localhost");
+            fail("Should have failed trying to add connector with scheme: " + scheme);
         } catch (IllegalArgumentException e) {
-            assertEquals("VM scheme is not allowed", e.getMessage());
+            assertEquals("Transport scheme '" + scheme + "' is not allowed", e.getMessage());
         }
 
         try {
-            // verify any composite URI is blocked as well
-            brokerView.addConnector("failover:(tcp://0.0.0.0:0,vm://" + brokerName + ")");
-            fail("Should have failed trying to add vm connector");
+            brokerView.addConnector("static:(tcp://0.0.0.0:0," + scheme + "://" + brokerName + ")");
+            fail("Should have failed trying to add connector with scheme: " + scheme);
         } catch (IllegalArgumentException e) {
-            assertEquals("VM scheme is not allowed", e.getMessage());
+            assertEquals("Transport scheme '" + scheme + "' is not allowed", e.getMessage());
         }
 
         try {
-            // verify nested composite URI is blocked
-            brokerView.addConnector("failover:(failover:(failover:(vm://localhost)))");
-            fail("Should have failed trying to add vm connector");
+            brokerView.addConnector("static:(static:(static:(" + scheme + "://localhost)))");
+            fail("Should have failed trying to add connector with scheme: " + scheme);
         } catch (IllegalArgumentException e) {
-            assertEquals("VM scheme is not allowed", e.getMessage());
+            assertEquals("Transport scheme '" + scheme + "' is not allowed", e.getMessage());
         }
+    }
+
+    public void testNestedAddTransportConnector() throws Exception {
+        ObjectName brokerName = assertRegisteredObjectName(domain + ":type=Broker,brokerName=localhost");
+        BrokerViewMBean brokerView = MBeanServerInvocationHandler.newProxyInstance(mbeanServer, brokerName, BrokerViewMBean.class, true);
 
         try {
-            // verify nested composite URI with more than 5 levels is blocked
             brokerView.addConnector(
-                    "static:(failover:(failover:(failover:(failover:(failover:(tcp://localhost:0))))))");
+                    "static:(static:(static:(static:(static:(static:(tcp://localhost:0))))))");
             fail("Should have failed trying to add vm connector bridge");
         } catch (IllegalArgumentException e) {
             assertEquals("URI can't contain more than 5 nested composite URIs", e.getMessage());
         }
-
     }
 
 }
--- a/activemq-unit-tests/src/test/java/org/apache/activemq/jmx/JmxCreateNCTest.java
+++ b/activemq-unit-tests/src/test/java/org/apache/activemq/jmx/JmxCreateNCTest.java
@@ -20,12 +20,14 @@
 import org.apache.activemq.broker.jmx.BrokerViewMBean;
 import org.apache.activemq.broker.jmx.NetworkConnectorViewMBean;
 import org.junit.After;
-import org.junit.AfterClass;
 import org.junit.Before;
 import org.junit.Test;
 
 import javax.management.ObjectName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import static org.apache.activemq.broker.jmx.BrokerView.DENIED_TRANSPORT_SCHEMES;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.fail;
@@ -34,10 +36,11 @@
  * This test shows that when we create a network connector via JMX,
  * the NC/bridge shows up in the MBean Server
  *
- * @author <a href="http://www.christianposta.com/blog">Christian Posta</a>
  */
 public class JmxCreateNCTest {
 
+    private static final Logger LOG = LoggerFactory.getLogger(JmxCreateNCTest.class);
+
     private static final String BROKER_NAME = "jmx-broker";
 
     private BrokerService broker;
@@ -47,7 +50,8 @@
     public void setUp() throws Exception {
         broker = new BrokerService();
         broker.setBrokerName(BROKER_NAME);
-        broker.setUseJmx(true); // explicitly set this so no funny issues
+        broker.setUseJmx(true);
+        broker.setPersistent(false);
         broker.start();
         broker.waitUntilStarted();
 
@@ -58,8 +62,10 @@
 
     @After
     public void tearDown() throws Exception {
-        broker.stop();
-        broker.waitUntilStopped();
+        if (broker != null) {
+            broker.stop();
+            broker.waitUntilStopped();
+        }
     }
 
     @Test
@@ -81,46 +87,45 @@
     }
 
     @Test
-    public void testVmBridgeBlocked() throws Exception {
-        // Test composite network connector uri
-        try {
-            proxy.addNetworkConnector("static:(vm://localhost)");
-            fail("Should have failed trying to add vm connector bridge");
-        } catch (IllegalArgumentException e) {
-            assertEquals("VM scheme is not allowed", e.getMessage());
+    public void testTransportSchemeBridgeBlocked() throws Exception {
+        for (String deniedScheme : DENIED_TRANSPORT_SCHEMES) {
+            LOG.info("verify testTransportSchemeBridgeBlocked scheme: {}", deniedScheme);
+            testTransportSchemeBridgeBlocked(deniedScheme);
         }
+    }
 
+    protected void testTransportSchemeBridgeBlocked(String scheme) throws Exception {
         try {
-            proxy.addNetworkConnector("multicast:(vm://localhost)");
-            fail("Should have failed trying to add vm connector bridge");
+            proxy.addNetworkConnector("static:(" + scheme + "://localhost)");
+            fail("Should have failed trying to add connector bridge with scheme: " + scheme);
         } catch (IllegalArgumentException e) {
-            assertEquals("VM scheme is not allowed", e.getMessage());
+            assertEquals("Transport scheme '" + scheme + "' is not allowed", e.getMessage());
         }
 
-        // verify direct vm as well
         try {
-            proxy.addNetworkConnector("vm://localhost");
-            fail("Should have failed trying to add vm connector bridge");
+            proxy.addNetworkConnector(scheme + "://localhost");
+            fail("Should have failed trying to add connector bridge with scheme: " + scheme);
         } catch (IllegalArgumentException e) {
-            assertEquals("VM scheme is not allowed", e.getMessage());
+            assertEquals("Transport scheme '" + scheme + "' is not allowed", e.getMessage());
         }
 
         try {
-            // verify nested composite URI is blocked
-            proxy.addNetworkConnector("static:(failover:(failover:(tcp://localhost:0,vm://localhost)))");
-            fail("Should have failed trying to add vm connector bridge");
+            proxy.addNetworkConnector("static:(static:(static:(tcp://localhost:0," + scheme + "://localhost)))");
+            fail("Should have failed trying to add connector bridge with scheme: " + scheme);
         } catch (IllegalArgumentException e) {
-            assertEquals("VM scheme is not allowed", e.getMessage());
+            assertEquals("Transport scheme '" + scheme + "' is not allowed", e.getMessage());
         }
     }
 
     @Test
     public void testAddNetworkConnectorMaxComposite() throws Exception {
+        assertNotNull(proxy.addNetworkConnector(
+                "static:(static:(static:(static:(static:(bad://localhost)))))"));
+
         try {
-            // verify nested composite URI with more than 5 levels is blocked
             proxy.addNetworkConnector(
-                    "static:(failover:(failover:(failover:(failover:(failover:(tcp://localhost:0))))))");
-            fail("Should have failed trying to add vm connector bridge");
+                    "static:(static:(static:(static:(static:(static:(tcp://localhost:0))))))");
+            fail("Should have failed trying to add more than 5 connector bridges");
         } catch (IllegalArgumentException e) {
             assertEquals("URI can't contain more than 5 nested composite URIs", e.getMessage());
         }

diff --git a/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientContext.java b/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientContext.java
--- a/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientContext.java
+++ b/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientContext.java
@@ -59,7 +59,12 @@ public class JMSClientContext {
     }
 
     public Connection createConnection(URI remoteURI, String username, String password, String clientId, boolean syncPublish) throws JMSException {
-        ConnectionFactory factory = createConnectionFactory(remoteURI, username, password, syncPublish);
+        return createConnection(remoteURI, username, password, clientId, syncPublish, null);
+    }
+
+    public Connection createConnection(URI remoteURI, String username, String password, String clientId, boolean syncPublish,
+            String sslProtocol) throws JMSException {
+        ConnectionFactory factory = createConnectionFactory(remoteURI, username, password, syncPublish, sslProtocol);
 
         Connection connection = factory.createConnection();
         connection.setExceptionListener(new ExceptionListener() {
@@ -166,7 +171,12 @@ public class JMSClientContext {
     }
 
     private ConnectionFactory createConnectionFactory(
-        URI remoteURI, String username, String password, boolean syncPublish) {
+            URI remoteURI, String username, String password, boolean syncPublish) {
+        return createConnectionFactory(remoteURI, username, password, syncPublish, null);
+    }
+
+    private ConnectionFactory createConnectionFactory(
+        URI remoteURI, String username, String password, boolean syncPublish, String sslProtocol) {
 
         String clientScheme;
         boolean useSSL = false;
@@ -204,6 +214,9 @@ public class JMSClientContext {
 
         if (useSSL) {
             amqpURI += "?transport.verifyHost=false";
+            if (sslProtocol != null) {
+                amqpURI += "&transport.enabledProtocols=" + sslProtocol;
+            }
         }
 
         LOG.debug("In createConnectionFactory using URI: {}", amqpURI);
diff --git a/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientNioPlusSslTest.java b/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientNioPlusSslTest.java
--- a/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientNioPlusSslTest.java
+++ b/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientNioPlusSslTest.java
@@ -16,8 +16,12 @@
  */
 package org.apache.activemq.transport.amqp;
 
+import jakarta.jms.Connection;
+import jakarta.jms.JMSException;
 import java.net.URI;
 
+import java.net.URISyntaxException;
+import org.junit.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -32,6 +36,13 @@ public class JMSClientNioPlusSslTest extends JMSClientSslTest {
         return amqpNioPlusSslURI;
     }
 
+    protected Connection createConnection(String clientId, boolean syncPublish) throws JMSException {
+        Connection connection = JMSClientContext.INSTANCE.createConnection(getBrokerURI(), "admin", "password", clientId, syncPublish,
+                enabledProtocols);
+        connection.start();
+        return connection;
+    }
+
     @Override
     protected boolean isUseTcpConnector() {
         return false;
@@ -46,4 +57,15 @@ public class JMSClientNioPlusSslTest extends JMSClientSslTest {
     protected String getTargetConnectorName() {
         return "amqp+nio+ssl";
     }
+
+    @Test(timeout=30000)
+    public void testSslHandshakeRenegotiationTlsv12() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.2");
+    }
+
+    @Test(timeout=30000)
+    public void testSslHandshakeRenegotiationTlsv13() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.3");
+    }
+
 }
diff --git a/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientSslTest.java b/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientSslTest.java
--- a/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientSslTest.java
+++ b/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/JMSClientSslTest.java
@@ -16,8 +16,29 @@
  */
 package org.apache.activemq.transport.amqp;
 
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import io.netty.channel.Channel;
+import io.netty.handler.ssl.SslHandler;
+import jakarta.jms.Message;
+import jakarta.jms.MessageConsumer;
+import jakarta.jms.MessageProducer;
+import jakarta.jms.Queue;
+import jakarta.jms.Session;
+import jakarta.jms.TextMessage;
+import java.lang.reflect.Field;
+import java.net.InetSocketAddress;
 import java.net.URI;
 
+import org.apache.activemq.broker.TransportConnector;
+import org.apache.activemq.transport.amqp.joram.ActiveMQAdmin;
+import org.apache.activemq.util.NioSslTestUtil;
+import org.apache.qpid.jms.JmsConnection;
+import org.apache.qpid.jms.provider.amqp.AmqpProvider;
+import org.apache.qpid.jms.transports.netty.NettyTcpTransport;
+import org.objectweb.jtests.jms.framework.TestConfig;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -27,6 +48,16 @@ import org.slf4j.LoggerFactory;
 public class JMSClientSslTest extends JMSClientTest {
     protected static final Logger LOG = LoggerFactory.getLogger(JMSClientSslTest.class);
 
+    protected String enabledProtocols = null;
+
+
+    @Override
+    public void setUp() throws Exception {
+        enabledProtocols = null;
+        super.setUp();
+    }
+
+
     @Override
     protected URI getBrokerURI() {
         return amqpSslURI;
@@ -46,4 +77,65 @@ public class JMSClientSslTest extends JMSClientTest {
     protected String getTargetConnectorName() {
         return "amqp+ssl";
     }
+
+    protected void testSslHandshakeRenegotiation(String protocol) throws Exception {
+        enabledProtocols = protocol;
+
+        ActiveMQAdmin.enableJMSFrameTracing();
+
+        connection = createConnection();
+
+        JmsConnection jmsCon = (JmsConnection) connection;
+        NettyTcpTransport transport = getNettyTransport(jmsCon);
+        Channel channel = getNettyChannel(transport);
+        SslHandler sslHandler = channel.pipeline().get(SslHandler.class);
+        assertEquals(protocol, sslHandler.engine().getSession().getProtocol());
+
+        // trigger handshakes
+        for (int i = 0; i < 10; i++) {
+            sslHandler.engine().beginHandshake();
+        }
+
+        // give some time for the handshake updates
+        Thread.sleep(100);
+
+        // check status advances if NIOSSL, then continue
+        // below to verify transports are not stuck
+        checkHandshakeStatusAdvances(((InetSocketAddress)channel.localAddress()).getPort());
+
+        // Make sure messages still work
+        Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        Queue queue = session.createQueue(getDestinationName());
+        MessageProducer p = session.createProducer(queue);
+
+        TextMessage message = session.createTextMessage();
+        message.setText("hello");
+        p.send(message);
+
+        MessageConsumer consumer = session.createConsumer(queue);
+        Message msg = consumer.receive(100);
+        assertNotNull(msg);
+        assertTrue(msg instanceof TextMessage);
+
+    }
+
+    // This only applies to NIO SSL
+    protected void checkHandshakeStatusAdvances(int localPort) throws Exception {
+        TransportConnector connector = brokerService.getTransportConnectorByScheme(getBrokerURI().getScheme());
+        NioSslTestUtil.checkHandshakeStatusAdvances(connector, localPort);
+    }
+
+    private NettyTcpTransport getNettyTransport(JmsConnection jmsCon) throws Exception {
+        Field providerField = JmsConnection.class.getDeclaredField("provider");
+        providerField.setAccessible(true);
+        AmqpProvider provider = (AmqpProvider) providerField.get(jmsCon);
+        return (NettyTcpTransport) provider.getTransport();
+    }
+
+    private Channel getNettyChannel(NettyTcpTransport transport) throws Exception {
+        Field channelField = NettyTcpTransport.class.getDeclaredField("channel");
+        channelField.setAccessible(true);
+        return (Channel) channelField.get(transport);
+    }
+
 }
diff --git a/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/auto/JMSClientAutoNioPlusSslTest.java b/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/auto/JMSClientAutoNioPlusSslTest.java
--- a/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/auto/JMSClientAutoNioPlusSslTest.java
+++ b/activemq-amqp/src/test/java/org/apache/activemq/transport/amqp/auto/JMSClientAutoNioPlusSslTest.java
@@ -16,9 +16,13 @@
  */
 package org.apache.activemq.transport.amqp.auto;
 
+import jakarta.jms.Connection;
+import jakarta.jms.JMSException;
 import java.net.URI;
 
+import org.apache.activemq.transport.amqp.JMSClientContext;
 import org.apache.activemq.transport.amqp.JMSClientSslTest;
+import org.junit.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -33,6 +37,13 @@ public class JMSClientAutoNioPlusSslTest extends JMSClientSslTest {
         return autoNioPlusSslURI;
     }
 
+    protected Connection createConnection(String clientId, boolean syncPublish) throws JMSException {
+        Connection connection = JMSClientContext.INSTANCE.createConnection(getBrokerURI(), "admin", "password", clientId, syncPublish,
+                enabledProtocols);
+        connection.start();
+        return connection;
+    }
+
     @Override
     protected boolean isUseTcpConnector() {
         return false;
@@ -47,4 +58,15 @@ public class JMSClientAutoNioPlusSslTest extends JMSClientSslTest {
     protected String getTargetConnectorName() {
         return "auto+nio+ssl";
     }
+
+    @Test(timeout=30000)
+    public void testSslHandshakeRenegotiationTlsv12() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.2");
+    }
+
+    @Test(timeout=30000)
+    public void testSslHandshakeRenegotiationTlsv13() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.3");
+    }
+
 }
diff --git a/activemq-broker/src/main/java/org/apache/activemq/transport/nio/AutoInitNioSSLTransport.java b/activemq-broker/src/main/java/org/apache/activemq/transport/nio/AutoInitNioSSLTransport.java
--- a/activemq-broker/src/main/java/org/apache/activemq/transport/nio/AutoInitNioSSLTransport.java
+++ b/activemq-broker/src/main/java/org/apache/activemq/transport/nio/AutoInitNioSSLTransport.java
@@ -171,7 +171,24 @@ public class AutoInitNioSSLTransport extends NIOSSLTransport {
                 if (!plain.hasRemaining()) {
                     int readCount = secureRead(plain);
 
-                    if (readCount == 0) {
+                    /*
+                     * 1) If data is read, continue below to the processCommand() call
+                     *    and handle processing the data in the buffer. This takes priority
+                     *    and some handshake status updates (like NEED_WRAP) can be handled
+                     *    concurrently with application data (like TLSv1.3 key updates)
+                     *    when the broker sends data to a client.
+                     *
+                     * 2) If no data is read, it's possible that the connection is waiting
+                     *    for us to process a handshake update (either KeyUpdate for
+                     *    TLS1.3 or renegotiation for TLSv1.2) so we need to check and process
+                     *    any handshake updates. If the handshake status was updated,
+                     *    we want to continue and loop again to recheck if we can now read new
+                     *    application data into the buffer after processing the updates.
+                     *
+                     * 3) If no data is read, and no handshake update is needed, then we
+                     *    are finished and can break.
+                     */
+                    if (readCount == 0 && !handleHandshakeUpdate()) {
                         break;
                     }
 
@@ -184,7 +201,11 @@ public class AutoInitNioSSLTransport extends NIOSSLTransport {
                     receiveCounter.addAndGet(readCount);
                 }
 
-                if (status == SSLEngineResult.Status.OK && handshakeStatus != SSLEngineResult.HandshakeStatus.NEED_UNWRAP) {
+                // Try and process commands if there is any data in plain if status is OK
+                // Handshake renegotiation can happen concurrently with application data reads
+                // so it's possible to have read data that needs processing even if the
+                // handshake status indicates NEED_UNWRAP
+                if (status == SSLEngineResult.Status.OK && plain.hasRemaining()) {
                     processCommand(plain);
                     //we have received enough bytes to detect the protocol
                     if (receiveCounter.get() >= 8) {
diff --git a/activemq-broker/src/test/java/org/apache/activemq/util/NioSslTestUtil.java b/activemq-broker/src/test/java/org/apache/activemq/util/NioSslTestUtil.java
new file mode 100644
--- /dev/null
+++ b/activemq-broker/src/test/java/org/apache/activemq/util/NioSslTestUtil.java
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.util;
+
+import static org.junit.Assert.assertTrue;
+
+import java.lang.reflect.Field;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+import javax.net.ssl.SSLSocket;
+import org.apache.activemq.broker.TransportConnection;
+import org.apache.activemq.broker.TransportConnector;
+import org.apache.activemq.transport.nio.NIOSSLTransport;
+
+public class NioSslTestUtil {
+
+    public static void checkHandshakeStatusAdvances(TransportConnector connector, SSLSocket socket) throws Exception {
+        checkHandshakeStatusAdvances(connector, socket.getLocalPort(), 10000, 10);
+    }
+
+    public static void checkHandshakeStatusAdvances(TransportConnector connector, int localPort) throws Exception {
+        checkHandshakeStatusAdvances(connector, localPort, 10000, 10);
+    }
+
+    public static void checkHandshakeStatusAdvances(TransportConnector connector, int localPort,
+            long duration, long sleepMillis) throws Exception {
+
+        TransportConnection con = connector.getConnections().stream()
+                .filter(tc -> tc.getRemoteAddress().contains(
+                Integer.toString(localPort))).findFirst().orElseThrow();
+
+        Field field = NIOSSLTransport.class.getDeclaredField("handshakeStatus");
+        field.setAccessible(true);
+        NIOSSLTransport t = con.getTransport().narrow(NIOSSLTransport.class);
+        // If this is the NIOSSLTransport then verify we exit NEED_WRAP and NEED_TASK
+        if (t != null) {
+            assertTrue(Wait.waitFor(() -> {
+                SSLEngineResult.HandshakeStatus status = (SSLEngineResult.HandshakeStatus) field.get(t);
+                return status != HandshakeStatus.NEED_WRAP && status != HandshakeStatus.NEED_TASK;
+            }, duration, sleepMillis));
+        }
+    }
+}
diff --git a/activemq-client/src/main/java/org/apache/activemq/transport/nio/NIOSSLTransport.java b/activemq-client/src/main/java/org/apache/activemq/transport/nio/NIOSSLTransport.java
--- a/activemq-client/src/main/java/org/apache/activemq/transport/nio/NIOSSLTransport.java
+++ b/activemq-client/src/main/java/org/apache/activemq/transport/nio/NIOSSLTransport.java
@@ -36,6 +36,7 @@ import javax.net.ssl.SSLContext;
 import javax.net.ssl.SSLEngine;
 import javax.net.ssl.SSLEngineResult;
 import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+import javax.net.ssl.SSLEngineResult.Status;
 import javax.net.ssl.SSLParameters;
 import javax.net.ssl.SSLPeerUnverifiedException;
 import javax.net.ssl.SSLSession;
@@ -271,10 +272,26 @@ public class NIOSSLTransport extends NIOTransport {
                 }
 
                 if (!plain.hasRemaining()) {
-
                     int readCount = secureRead(plain);
 
-                    if (readCount == 0) {
+                    /*
+                     * 1) If data is read, continue below to the processCommand() call
+                     *    and handle processing the data in the buffer. This takes priority
+                     *    and some handshake status updates (like NEED_WRAP) can be handled
+                     *    concurrently with application data (like TLSv1.3 key updates)
+                     *    when the broker sends data to a client.
+                     *
+                     * 2) If no data is read, it's possible that the connection is waiting
+                     *    for us to process a handshake update (either KeyUpdate for
+                     *    TLS1.3 or renegotiation for TLSv1.2) so we need to check and process
+                     *    any handshake updates. If the handshake status was updated,
+                     *    we want to continue and loop again to recheck if we can now read new
+                     *    application data into the buffer after processing the updates.
+                     *
+                     * 3) If no data is read, and no handshake update is needed, then we
+                     *    are finished and can break.
+                     */
+                    if (readCount == 0 && !handleHandshakeUpdate()) {
                         break;
                     }
 
@@ -288,7 +305,11 @@ public class NIOSSLTransport extends NIOTransport {
                     receiveCounter.addAndGet(readCount);
                 }
 
-                if (status == SSLEngineResult.Status.OK && handshakeStatus != SSLEngineResult.HandshakeStatus.NEED_UNWRAP) {
+                // Try and process commands if there is any data in plain if status is OK.
+                // Handshake renegotiation can happen concurrently with application data reads,
+                // so it's possible to have data in the buffer that needs processing even if the
+                // handshake status indicates NEED_UNWRAP
+                if (status == SSLEngineResult.Status.OK && plain.hasRemaining()) {
                     processCommand(plain);
                 }
             }
@@ -431,6 +452,43 @@ public class NIOSSLTransport extends NIOTransport {
         return plain.remaining();
     }
 
+    /**
+     * Checks if the connection is waiting for the broker to process an update
+     * to continue with handshaking and processes the necessary updates
+     *
+     * @return True if the handshake status was updated
+     * @throws IOException if an error occurs during processing
+     */
+    protected boolean handleHandshakeUpdate() throws IOException {
+        switch (handshakeStatus) {
+            // For NEED_WRAP, call write with a 0 size buffer which will flush any
+            // pending handshaking updates even if there is no application data to write
+            case NEED_WRAP:
+                ((NIOOutputStream) buffOut).write(ByteBuffer.allocate(0));
+                handshakeStatus = sslEngine.getHandshakeStatus();
+                return true;
+            // For NEED_TASK, process any outstanding tasks needed to continue the
+            // handshake negotiation
+            case NEED_TASK:
+                Runnable task;
+                while ((task = sslEngine.getDelegatedTask()) != null) {
+                    task.run();
+                }
+                handshakeStatus = sslEngine.getHandshakeStatus();
+                return true;
+            // We need to update the session when finished, and then
+            // we can update the handshake status which will advance to
+            // NOT_HANDSHAKING before we continue and try and read
+            // more data
+            case FINISHED:
+                sslSession = sslEngine.getSession();
+                handshakeStatus = sslEngine.getHandshakeStatus();
+                return true;
+            default:
+                return false;
+        }
+    }
+
     protected void doHandshake() throws Exception {
         handshakeInProgress = true;
         Selector selector = null;
diff --git a/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTNIOSSLTest.java b/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTNIOSSLTest.java
--- a/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTNIOSSLTest.java
+++ b/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTNIOSSLTest.java
@@ -16,6 +16,7 @@
  */
 package org.apache.activemq.transport.mqtt;
 
+import org.junit.Test;
 /**
  * Run the basic tests with the NIO Transport.
  */
@@ -30,4 +31,14 @@ public class MQTTNIOSSLTest extends MQTTTest {
     public boolean isUseSSL() {
         return true;
     }
+
+    @Test(timeout = 60000)
+    public void testHandshakeRenegotiationTlsv12() throws Exception {
+        testHandshakeRenegotiation("TLSv1.2");
+    }
+
+    @Test(timeout = 60000)
+    public void testHandshakeRenegotiationTlsv13() throws Exception {
+        testHandshakeRenegotiation("TLSv1.3");
+    }
 }
diff --git a/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTTest.java b/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTTest.java
--- a/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTTest.java
+++ b/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTTest.java
@@ -25,6 +25,9 @@ import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.net.InetSocketAddress;
 import java.net.ProtocolException;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -46,8 +49,11 @@ import jakarta.jms.Queue;
 import jakarta.jms.Session;
 import jakarta.jms.TextMessage;
 
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLSocket;
 import org.apache.activemq.ActiveMQConnection;
 import org.apache.activemq.ActiveMQConnectionFactory;
+import org.apache.activemq.broker.TransportConnector;
 import org.apache.activemq.broker.region.policy.LastImageSubscriptionRecoveryPolicy;
 import org.apache.activemq.broker.region.policy.PolicyEntry;
 import org.apache.activemq.broker.region.policy.PolicyMap;
@@ -55,8 +61,12 @@ import org.apache.activemq.broker.region.policy.RetainedMessageSubscriptionRecov
 import org.apache.activemq.command.ActiveMQMessage;
 import org.apache.activemq.command.ActiveMQTopic;
 import org.apache.activemq.util.ByteSequence;
+import org.apache.activemq.util.NioSslTestUtil;
 import org.apache.activemq.util.Wait;
+import org.fusesource.hawtdispatch.transport.SslTransport;
 import org.fusesource.mqtt.client.BlockingConnection;
+import org.fusesource.mqtt.client.CallbackConnection;
+import org.fusesource.mqtt.client.FutureConnection;
 import org.fusesource.mqtt.client.MQTT;
 import org.fusesource.mqtt.client.Message;
 import org.fusesource.mqtt.client.QoS;
@@ -64,6 +74,7 @@ import org.fusesource.mqtt.client.Topic;
 import org.fusesource.mqtt.client.Tracer;
 import org.fusesource.mqtt.codec.MQTTFrame;
 import org.fusesource.mqtt.codec.PUBLISH;
+import org.junit.Assume;
 import org.junit.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -1993,4 +2004,64 @@ public class MQTTTest extends MQTTTestSupport {
           }
        }
     }
+
+    protected void testHandshakeRenegotiation(String protocol) throws Exception {
+        MQTT mqtt = createMQTTSslConnection(null, true, protocol);
+        mqtt.setClientId("");
+        mqtt.setCleanSession(true);
+
+        CallbackConnection callbackConnection = mqtt.callbackConnection();
+        BlockingConnection connection = new BlockingConnection(new FutureConnection(callbackConnection));
+        connection.connect();
+
+        SslTransport transport = getSslTransport(callbackConnection);
+        SSLEngine engine = getSslTransport(transport);
+        assertEquals(protocol, engine.getSession().getProtocol());
+
+        // Run 100 key updates in a loop so that we can
+        // verify that the transport correctly processes them
+        // and that we are not stuck in NEED_WRAP state. This
+        // only applies to NIO, for regular SSL the state is not
+        // handled by the transport
+        for (int i = 0; i < 100; i++) {
+            try {
+                engine.beginHandshake();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        // give some time for the handshake updates
+        Thread.sleep(100);
+
+        // Wait to get past NEED_WRAP as that indicates we correctly handled
+        // the key updates issued for TLSv1.3 renegotiation
+        checkHandshakeStatusAdvances(((InetSocketAddress)transport.getLocalAddress()).getPort());
+
+        // Make sure we can still subscribe and receive
+        connection.subscribe(new Topic[] { new Topic("topic1", QoS.AT_LEAST_ONCE) });
+        connection.publish("topic1", "topic".getBytes(), QoS.AT_LEAST_ONCE, false);
+        assertNotNull(connection.receive(2000, TimeUnit.MILLISECONDS));
+
+        connection.disconnect();
+    }
+
+    private void checkHandshakeStatusAdvances(int port) throws Exception {
+        TransportConnector connector = brokerService.getTransportConnectorByScheme(
+                getProtocolScheme());
+        NioSslTestUtil.checkHandshakeStatusAdvances(connector, port);
+    }
+
+    private static SslTransport getSslTransport(CallbackConnection callbackConnection) throws Exception {
+        Field transportField = CallbackConnection.class.getDeclaredField("transport");
+        transportField.setAccessible(true);
+        return (SslTransport) transportField.get(callbackConnection);
+    }
+
+    private static SSLEngine getSslTransport(SslTransport transport) throws Exception {
+        Field engineField =  SslTransport.class.getDeclaredField("engine");
+        engineField.setAccessible(true);
+        return (SSLEngine) engineField.get(transport);
+    }
+
 }
diff --git a/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTTestSupport.java b/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTTestSupport.java
--- a/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTTestSupport.java
+++ b/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/MQTTTestSupport.java
@@ -103,6 +103,7 @@ public class MQTTTestSupport {
         System.setProperty("javax.net.ssl.keyStoreType", "jks");
 
         exceptions.clear();
+
         startBroker();
     }
 
@@ -413,7 +414,11 @@ public class MQTTTestSupport {
         return mqtt;
     }
 
-    private MQTT createMQTTSslConnection(String clientId, boolean clean) throws Exception {
+    protected MQTT createMQTTSslConnection(String clientId, boolean clean) throws Exception {
+        return createMQTTSslConnection(clientId, clean, null);
+    }
+
+    protected MQTT createMQTTSslConnection(String clientId, boolean clean, String sslProtocol) throws Exception {
         MQTT mqtt = new MQTT();
         mqtt.setConnectAttemptsMax(1);
         mqtt.setReconnectAttemptsMax(0);
@@ -433,6 +438,9 @@ public class MQTTTestSupport {
         sslContext.setKeyStorePassword("password");
         sslContext.setTrustStore(trustStore.getCanonicalPath());
         sslContext.setTrustStorePassword("password");
+        if (sslProtocol != null) {
+            sslContext.setProtocol(sslProtocol);
+        }
         sslContext.afterPropertiesSet();
 
         mqtt.setSslContext(sslContext.getSSLContext());
diff --git a/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/auto/MQTTAutoNioSslTest.java b/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/auto/MQTTAutoNioSslTest.java
--- a/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/auto/MQTTAutoNioSslTest.java
+++ b/activemq-mqtt/src/test/java/org/apache/activemq/transport/mqtt/auto/MQTTAutoNioSslTest.java
@@ -17,6 +17,7 @@
 package org.apache.activemq.transport.mqtt.auto;
 
 import org.apache.activemq.transport.mqtt.MQTTTest;
+import org.junit.Test;
 
 /**
  * Run the basic tests with the NIO Transport.
@@ -32,4 +33,14 @@ public class MQTTAutoNioSslTest extends MQTTTest {
     public boolean isUseSSL() {
         return true;
     }
+
+    @Test(timeout = 60000)
+    public void testHandshakeRenegotiationTlsv12() throws Exception {
+        testHandshakeRenegotiation("TLSv1.2");
+    }
+
+    @Test(timeout = 60000)
+    public void testHandshakeRenegotiationTlsv13() throws Exception {
+        testHandshakeRenegotiation("TLSv1.3");
+    }
 }
diff --git a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompNIOSSLTest.java b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompNIOSSLTest.java
--- a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompNIOSSLTest.java
+++ b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompNIOSSLTest.java
@@ -19,9 +19,6 @@ package org.apache.activemq.transport.stomp;
 import java.io.IOException;
 import java.net.Socket;
 
-import javax.net.SocketFactory;
-import javax.net.ssl.SSLSocketFactory;
-
 public class StompNIOSSLTest extends StompTest {
 
     @Override
@@ -36,7 +33,6 @@ public class StompNIOSSLTest extends StompTest {
 
     @Override
     protected Socket createSocket() throws IOException {
-        SocketFactory factory = SSLSocketFactory.getDefault();
-        return factory.createSocket("127.0.0.1", this.nioSslPort);
+        return createSslSocket("127.0.0.1", this.nioSslPort, this.clientSslProtocol);
     }
 }
diff --git a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompSslAuthTest.java b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompSslAuthTest.java
--- a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompSslAuthTest.java
+++ b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompSslAuthTest.java
@@ -19,9 +19,6 @@ package org.apache.activemq.transport.stomp;
 import java.io.IOException;
 import java.net.Socket;
 
-import javax.net.SocketFactory;
-import javax.net.ssl.SSLSocketFactory;
-
 import org.apache.activemq.ActiveMQConnectionFactory;
 import org.apache.activemq.broker.BrokerPlugin;
 import org.apache.activemq.broker.TransportConnector;
@@ -48,8 +45,7 @@ public class StompSslAuthTest extends StompTest {
 
     @Override
     protected Socket createSocket() throws IOException {
-        SocketFactory factory = SSLSocketFactory.getDefault();
-        return factory.createSocket("127.0.0.1", this.sslPort);
+        return createSslSocket("127.0.0.1", this.sslPort, this.clientSslProtocol);
     }
 
     @Override
diff --git a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompSslTest.java b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompSslTest.java
--- a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompSslTest.java
+++ b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompSslTest.java
@@ -19,9 +19,6 @@ package org.apache.activemq.transport.stomp;
 import java.io.IOException;
 import java.net.Socket;
 
-import javax.net.SocketFactory;
-import javax.net.ssl.SSLSocketFactory;
-
 public class StompSslTest extends StompTest {
 
     @Override
@@ -36,7 +33,6 @@ public class StompSslTest extends StompTest {
 
     @Override
     protected Socket createSocket() throws IOException {
-        SocketFactory factory = SSLSocketFactory.getDefault();
-        return factory.createSocket("127.0.0.1", this.sslPort);
+        return createSslSocket("127.0.0.1", this.sslPort, this.clientSslProtocol);
     }
 }
diff --git a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompTestSupport.java b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompTestSupport.java
--- a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompTestSupport.java
+++ b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompTestSupport.java
@@ -21,6 +21,7 @@ import java.io.File;
 import java.io.IOException;
 import java.net.Socket;
 import java.security.ProtectionDomain;
+import java.security.SecureRandom;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Vector;
@@ -29,6 +30,7 @@ import jakarta.jms.JMSException;
 import javax.management.MalformedObjectNameException;
 import javax.management.ObjectName;
 
+import javax.net.ssl.SSLContext;
 import org.apache.activemq.ActiveMQConnectionFactory;
 import org.apache.activemq.broker.BrokerPlugin;
 import org.apache.activemq.broker.BrokerService;
@@ -70,6 +72,7 @@ public class StompTestSupport {
     protected StompConnection stompConnection;
     protected ActiveMQConnectionFactory cf;
     protected Vector<Throwable> exceptions = new Vector<Throwable>();
+    protected String transportConnectorName = null;
 
     @Rule public TestName name = new TestName();
 
@@ -118,6 +121,30 @@ public class StompTestSupport {
         LOG.info("========== finished " + getName() + " ==========");
     }
 
+    protected Socket createSslSocket(String host, int port, String protocol) throws IOException {
+        try {
+            final SSLContext context;
+            if (protocol != null) {
+                File keyStore = new File(basedir(), "src/test/resources/server.keystore");
+                File trustStore = new File(basedir(), "src/test/resources/client.keystore");
+
+                final ResourceLoadingSslContext sslContext = new ResourceLoadingSslContext();
+                sslContext.setKeyStore(keyStore.getCanonicalPath());
+                sslContext.setKeyStorePassword("password");
+                sslContext.setTrustStore(trustStore.getCanonicalPath());
+                sslContext.setTrustStorePassword("password");
+                sslContext.setProtocol(protocol);
+                sslContext.afterPropertiesSet();
+                context = sslContext.getSSLContext();
+            } else {
+                context = SSLContext.getDefault();
+            }
+            return context.getSocketFactory().createSocket(host, port);
+        } catch (Exception e) {
+            throw new IOException(e);
+        }
+    }
+
     public void startBroker() throws Exception {
         if (brokerService != null) {
             stopBroker();
@@ -151,7 +178,7 @@ public class StompTestSupport {
 
         ArrayList<BrokerPlugin> plugins = new ArrayList<BrokerPlugin>();
 
-        addTranportConnectors();
+        transportConnectorName = addTranportConnectors();
         addOpenWireConnector();
 
         BrokerPlugin authenticationPlugin = configureAuthentication();
@@ -291,7 +318,7 @@ public class StompTestSupport {
         cf = new ActiveMQConnectionFactory(jmsUri);
     }
 
-    protected void addTranportConnectors() throws Exception {
+    protected String addTranportConnectors() throws Exception {
         TransportConnector connector = null;
 
         if (isUseTcpConnector()) {
@@ -342,6 +369,8 @@ public class StompTestSupport {
             autoNioSslPort = connector.getConnectUri().getPort();
             LOG.debug("Using auto+nio+ssl port " + autoNioSslPort);
         }
+
+        return connector.getName();
     }
 
     protected boolean isPersistent() {
diff --git a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/auto/StompAutoNioSslTest.java b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/auto/StompAutoNioSslTest.java
--- a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/auto/StompAutoNioSslTest.java
+++ b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/auto/StompAutoNioSslTest.java
@@ -18,10 +18,6 @@ package org.apache.activemq.transport.stomp.auto;
 
 import java.io.IOException;
 import java.net.Socket;
-
-import javax.net.SocketFactory;
-import javax.net.ssl.SSLSocketFactory;
-
 import org.apache.activemq.transport.stomp.StompTest;
 
 public class StompAutoNioSslTest extends StompTest {
@@ -38,7 +34,7 @@ public class StompAutoNioSslTest extends StompTest {
 
     @Override
     protected Socket createSocket() throws IOException {
-        SocketFactory factory = SSLSocketFactory.getDefault();
-        return factory.createSocket("127.0.0.1", this.autoNioSslPort);
+        return createSslSocket("127.0.0.1", this.autoNioSslPort, this.clientSslProtocol);
     }
+
 }
diff --git a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/auto/StompAutoSslTest.java b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/auto/StompAutoSslTest.java
--- a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/auto/StompAutoSslTest.java
+++ b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/auto/StompAutoSslTest.java
@@ -19,9 +19,6 @@ package org.apache.activemq.transport.stomp.auto;
 import java.io.IOException;
 import java.net.Socket;
 
-import javax.net.SocketFactory;
-import javax.net.ssl.SSLSocketFactory;
-
 import org.apache.activemq.transport.stomp.StompTest;
 
 public class StompAutoSslTest extends StompTest {
@@ -38,7 +35,6 @@ public class StompAutoSslTest extends StompTest {
 
     @Override
     protected Socket createSocket() throws IOException {
-        SocketFactory factory = SSLSocketFactory.getDefault();
-        return factory.createSocket("127.0.0.1", this.autoSslPort);
+        return createSslSocket("127.0.0.1", this.autoSslPort, this.clientSslProtocol);
     }
 }
diff --git a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/TransportBrokerTestSupport.java b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/TransportBrokerTestSupport.java
--- a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/TransportBrokerTestSupport.java
+++ b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/TransportBrokerTestSupport.java
@@ -16,23 +16,42 @@
  */
 package org.apache.activemq.transport;
 
+import java.io.IOException;
+import java.lang.reflect.Field;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.util.ArrayList;
 import java.util.Iterator;
 
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+import javax.net.ssl.SSLSocket;
 import org.apache.activemq.broker.BrokerService;
 import org.apache.activemq.broker.BrokerTest;
 import org.apache.activemq.broker.StubConnection;
+import org.apache.activemq.broker.TransportConnection;
 import org.apache.activemq.broker.TransportConnector;
+import org.apache.activemq.command.ActiveMQDestination;
+import org.apache.activemq.command.ConnectionInfo;
+import org.apache.activemq.command.ConsumerInfo;
+import org.apache.activemq.command.Message;
+import org.apache.activemq.command.ProducerInfo;
+import org.apache.activemq.command.SessionInfo;
+import org.apache.activemq.transport.nio.NIOSSLTransport;
+import org.apache.activemq.transport.tcp.TcpTransport;
+import org.apache.activemq.util.NioSslTestUtil;
+import org.apache.activemq.util.Wait;
 
 public abstract class TransportBrokerTestSupport extends BrokerTest {
 
     protected TransportConnector connector;
     private ArrayList<StubConnection> connections = new ArrayList<StubConnection>();
+    protected String enabledProtocols = null;
+
 
     @Override
     protected void setUp() throws Exception {
+        enabledProtocols = null;
         super.setUp();
     }
 
@@ -87,4 +106,59 @@ public abstract class TransportBrokerTestSupport extends BrokerTest {
         return connection;
     }
 
+
+    protected void testSslHandshakeRenegotiation(String protocol) throws Exception {
+        this.enabledProtocols = protocol;
+
+        // Start a producer
+        StubConnection connection = createConnection();
+        ConnectionInfo connectionInfo = createConnectionInfo();
+        SessionInfo sessionInfo = createSessionInfo(connectionInfo);
+        ProducerInfo producerInfo = createProducerInfo(sessionInfo);
+        connection.send(connectionInfo);
+        connection.send(sessionInfo);
+        connection.send(producerInfo);
+
+        destination = createDestinationInfo(connection, connectionInfo, ActiveMQDestination.QUEUE_TYPE);
+
+        TcpTransport tcpTransport = connection.getTransport().narrow(TcpTransport.class);
+        Field socketField = TcpTransport.class.getDeclaredField("socket");
+        socketField.setAccessible(true);
+        SSLSocket socket = (SSLSocket) socketField.get(tcpTransport);
+        assertEquals(protocol, socket.getSession().getProtocol());
+
+        // trigger a bunch of updates (either TLSv1.2 full updates or TLSv1.3 key updates)
+        for (int i = 0; i < 10; i++) {
+            try {
+                socket.startHandshake();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        // give some time for the handshake updates
+        Thread.sleep(100);
+
+        // check status advances if NIOSSL, then continue
+        // below to verify transports are not stuck
+        checkHandshakeStatusAdvances(socket);
+
+        // Send a message to the broker.
+        connection.send(createMessage(producerInfo, destination, deliveryMode));
+
+        // Start the consumer
+        ConsumerInfo consumerInfo = createConsumerInfo(sessionInfo, destination);
+        connection.send(consumerInfo);
+
+        // Make sure the message was delivered.
+        Message m = receiveMessage(connection);
+        assertNotNull(m);
+    }
+
+    // This only applies to NIO SSL
+    protected void checkHandshakeStatusAdvances(SSLSocket socket) throws Exception {
+        TransportConnector connector = broker.getTransportConnectorByScheme(getBindURI().getScheme());
+        NioSslTestUtil.checkHandshakeStatusAdvances(connector, socket);
+    }
+
 }
diff --git a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/auto/AutoNIOSslTransportBrokerTest.java b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/auto/AutoNIOSslTransportBrokerTest.java
--- a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/auto/AutoNIOSslTransportBrokerTest.java
+++ b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/auto/AutoNIOSslTransportBrokerTest.java
@@ -16,13 +16,37 @@
  */
 package org.apache.activemq.transport.auto;
 
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.net.Socket;
 import java.net.URI;
 import java.net.URISyntaxException;
 
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+import javax.net.ssl.SSLSocket;
 import junit.framework.Test;
 import junit.textui.TestRunner;
 
+import org.apache.activemq.broker.StubConnection;
+import org.apache.activemq.broker.TransportConnection;
+import org.apache.activemq.broker.TransportConnector;
+import org.apache.activemq.command.ActiveMQDestination;
+import org.apache.activemq.command.ConnectionInfo;
+import org.apache.activemq.command.ConsumerInfo;
+import org.apache.activemq.command.Message;
+import org.apache.activemq.command.ProducerInfo;
+import org.apache.activemq.command.SessionInfo;
+import org.apache.activemq.transport.Transport;
 import org.apache.activemq.transport.TransportBrokerTestSupport;
+import org.apache.activemq.transport.TransportFilter;
+import org.apache.activemq.transport.nio.NIOSSLTransport;
+import org.apache.activemq.transport.tcp.SslTransport;
+import org.apache.activemq.transport.tcp.TcpTransport;
+import org.apache.activemq.util.Wait;
 
 public class AutoNIOSslTransportBrokerTest extends TransportBrokerTestSupport {
 
@@ -38,7 +62,9 @@ public class AutoNIOSslTransportBrokerTest extends TransportBrokerTestSupport {
 
     @Override
     protected URI getBindURI() throws URISyntaxException {
-        return new URI("auto+nio+ssl://localhost:0?soWriteTimeout=20000");
+        String uri = super.getBindURI().toString() + "?soWriteTimeout=20000";
+        uri = enabledProtocols != null ? uri + "&socket.enabledProtocols=" + enabledProtocols : uri;
+        return new URI(uri);
     }
 
     @Override
@@ -54,6 +80,14 @@ public class AutoNIOSslTransportBrokerTest extends TransportBrokerTestSupport {
         super.setUp();
     }
 
+    public void testSslHandshakeRenegotiationTlsv12() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.2");
+    }
+
+    public void testSslHandshakeRenegotiationTlsv13() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.3");
+    }
+
     public static Test suite() {
         return suite(AutoNIOSslTransportBrokerTest.class);
     }
@@ -62,5 +96,4 @@ public class AutoNIOSslTransportBrokerTest extends TransportBrokerTestSupport {
         TestRunner.run(suite());
     }
 
-
 }
diff --git a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/auto/AutoSslTransportBrokerTest.java b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/auto/AutoSslTransportBrokerTest.java
--- a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/auto/AutoSslTransportBrokerTest.java
+++ b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/auto/AutoSslTransportBrokerTest.java
@@ -36,7 +36,9 @@ public class AutoSslTransportBrokerTest extends TransportBrokerTestSupport {
 
     @Override
     protected URI getBindURI() throws URISyntaxException {
-        return new URI("auto+ssl://localhost:0?soWriteTimeout=20000");
+        String uri = super.getBindURI().toString() + "?soWriteTimeout=20000";
+        uri = enabledProtocols != null ? uri + "&socket.enabledProtocols=" + enabledProtocols : uri;
+        return new URI(uri);
     }
 
     @Override
@@ -53,6 +55,14 @@ public class AutoSslTransportBrokerTest extends TransportBrokerTestSupport {
         super.setUp();
     }
 
+    public void testSslHandshakeRenegotiationTlsv12() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.2");
+    }
+
+    public void testSslHandshakeRenegotiationTlsv13() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.3");
+    }
+
     public static Test suite() {
         return suite(AutoSslTransportBrokerTest.class);
     }
diff --git a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/nio/NIOSSLBasicTest.java b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/nio/NIOSSLBasicTest.java
--- a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/nio/NIOSSLBasicTest.java
+++ b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/nio/NIOSSLBasicTest.java
@@ -24,6 +24,9 @@ import jakarta.jms.Queue;
 import jakarta.jms.Session;
 import jakarta.jms.TextMessage;
 
+import javax.net.ServerSocketFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLServerSocketFactory;
 import org.apache.logging.log4j.Level;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.core.LogEvent;
@@ -98,7 +101,10 @@ public class NIOSSLBasicTest {
 
     @Test
     public void enabledCipherSuites() throws Exception {
-        BrokerService broker = createBroker("nio+ssl", getTransportType() + "://localhost:0?transport.needClientAuth=true&transport.verifyHostName=false&transport.enabledCipherSuites=TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256");
+        SSLServerSocketFactory s = (SSLServerSocketFactory) SSLServerSocketFactory.getDefault();
+        String[] defaultCipherSuites = s.getSupportedCipherSuites();
+        assertTrue(defaultCipherSuites.length > 0);
+        BrokerService broker = createBroker("nio+ssl", getTransportType() + "://localhost:0?transport.needClientAuth=true&transport.verifyHostName=false&transport.enabledCipherSuites=" + defaultCipherSuites[0]);
         basicSendReceive("ssl://localhost:" + broker.getConnectorByName("nio+ssl").getConnectUri().getPort() + "?socket.verifyHostName=false");
         stopBroker(broker);
     }
diff --git a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/nio/NIOSSLTransportBrokerTest.java b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/nio/NIOSSLTransportBrokerTest.java
--- a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/nio/NIOSSLTransportBrokerTest.java
+++ b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/nio/NIOSSLTransportBrokerTest.java
@@ -36,7 +36,9 @@ public class NIOSSLTransportBrokerTest extends TransportBrokerTestSupport {
 
     @Override
     protected URI getBindURI() throws URISyntaxException {
-        return new URI("nio+ssl://localhost:0?soWriteTimeout=20000");
+        String uri = super.getBindURI().toString() + "?soWriteTimeout=20000";
+        uri = enabledProtocols != null ? uri + "&socket.enabledProtocols=" + enabledProtocols : uri;
+        return new URI(uri);
     }
 
     @Override
@@ -58,6 +60,14 @@ public class NIOSSLTransportBrokerTest extends TransportBrokerTestSupport {
         super.tearDown();
     }
 
+    public void testSslHandshakeRenegotiationTlsv12() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.2");
+    }
+
+    public void testSslHandshakeRenegotiationTlsv13() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.3");
+    }
+
     public static Test suite() {
         return suite(NIOSSLTransportBrokerTest.class);
     }
diff --git a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/tcp/SslTransportBrokerTest.java b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/tcp/SslTransportBrokerTest.java
--- a/activemq-unit-tests/src/test/java/org/apache/activemq/transport/tcp/SslTransportBrokerTest.java
+++ b/activemq-unit-tests/src/test/java/org/apache/activemq/transport/tcp/SslTransportBrokerTest.java
@@ -35,7 +35,9 @@ public class SslTransportBrokerTest extends TransportBrokerTestSupport {
 
     @Override
     protected URI getBindURI() throws URISyntaxException {
-        return new URI("ssl://localhost:0?soWriteTimeout=20000");
+        String uri = super.getBindURI().toString() + "?soWriteTimeout=20000";
+        uri = enabledProtocols != null ? uri + "&socket.enabledProtocols=" + enabledProtocols : uri;
+        return new URI(uri);
     }
 
     protected void setUp() throws Exception {
@@ -51,6 +53,14 @@ public class SslTransportBrokerTest extends TransportBrokerTestSupport {
         super.setUp();
     }
 
+    public void testSslHandshakeRenegotiationTlsv12() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.2");
+    }
+
+    public void testSslHandshakeRenegotiationTlsv13() throws Exception {
+        testSslHandshakeRenegotiation("TLSv1.3");
+    }
+
     public static Test suite() {
         return suite(SslTransportBrokerTest.class);
     }
--- a/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompTest.java
+++ b/activemq-stomp/src/test/java/org/apache/activemq/transport/stomp/StompTest.java
@@ -25,6 +25,7 @@
 
 import java.io.IOException;
 import java.io.StringReader;
+import java.lang.reflect.Field;
 import java.net.SocketTimeoutException;
 import java.util.Arrays;
 import java.util.HashMap;
@@ -50,7 +51,12 @@
 import jakarta.jms.TextMessage;
 import javax.management.ObjectName;
 
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+import javax.net.ssl.SSLSocket;
 import org.apache.activemq.broker.BrokerService;
+import org.apache.activemq.broker.TransportConnection;
+import org.apache.activemq.broker.TransportConnector;
 import org.apache.activemq.broker.jmx.BrokerViewMBean;
 import org.apache.activemq.broker.jmx.QueueViewMBean;
 import org.apache.activemq.broker.region.AbstractSubscription;
@@ -61,7 +67,10 @@
 import org.apache.activemq.command.ActiveMQDestination;
 import org.apache.activemq.command.ActiveMQQueue;
 import org.apache.activemq.command.ActiveMQTextMessage;
+import org.apache.activemq.transport.nio.NIOSSLTransport;
+import org.apache.activemq.util.NioSslTestUtil;
 import org.apache.activemq.util.Wait;
+import org.junit.Assume;
 import org.junit.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -79,6 +88,7 @@
     protected Session session;
     protected ActiveMQQueue queue;
     protected XStream xstream;
+    protected String clientSslProtocol;
 
     private final String xmlObject = "<pojo>\n"
             + "  <name>Dejan</name>\n"
@@ -134,6 +144,7 @@
         queue = new ActiveMQQueue(getQueueName());
         super.setUp();
 
+        clientSslProtocol = null;
         stompConnect();
 
         connection = cf.createConnection("system", "manager");
@@ -2435,6 +2446,97 @@
         assertFalse(frame.contains("backslash:\\\\\n"));
         assertFalse(frame.contains("carriagereturn:\\\\r\n"));
 
+    }
+
+    // Verify that full handshake renegotiation is handled by both SSL
+    // and NIO ssl transports and we don't get stuck. This is primarily
+    // to test NIO handles NEED_TASK but also double checks SSL works
+    @Test(timeout = 60000)
+    public void testHandshakeRenegotiationTlsv12() throws Exception {
+        Assume.assumeTrue(stompConnection.getStompSocket() instanceof SSLSocket);
+
+        // Reset
+        clientSslProtocol = "TLSv1.2";
+        stompDisconnect();
+        stompConnect();
+
+        connectForSslHandshakeTest();
+
+        SSLSocket socket = (SSLSocket) stompConnection.getStompSocket();
+        assertEquals("TLSv1.2", socket.getSession().getProtocol());
+
+        socket.startHandshake();
+        Thread.sleep(100);
+
+        // Wait to get past NEED_TASK as that indicates we correctly handled
+        // the tasks issued for TLSv1.2 renegotiation
+        checkHandshakeStatusAdvances(socket);
+
+        receiveForSslHandshakeTest();
+    }
+
+    // Verify that TLS1.3 KeyUpdates during handshake renegotiation will be processed
+    // by the server correctly
+    @Test(timeout = 60000)
+    public void testHandshakeRenegotiationTlsv13() throws Exception {
+        Assume.assumeTrue(stompConnection.getStompSocket() instanceof SSLSocket);
+
+        // Set to TLSv1.3
+        clientSslProtocol = "TLSv1.3";
+        stompDisconnect();
+        stompConnect();
+
+        connectForSslHandshakeTest();
+
+        SSLSocket socket = (SSLSocket) stompConnection.getStompSocket();
+        assertEquals("TLSv1.3", socket.getSession().getProtocol());
+
+        // Run 100 key updates in a loop so that we can
+        // verify that the transport correctly processes them
+        // and that we are not stuck in NEED_WRAP state. This
+        // only applies to NIO, for regular SSL the state is not
+        // handled by the transport
+        for (int i = 0; i < 100; i++) {
+            try {
+                socket.startHandshake();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        // give some time for the handshake updates
+        Thread.sleep(100);
+
+        // Wait to get past NEED_WRAP as that indicates we correctly handled
+        // the key updates issued for TLSv1.3 renegotiation
+        checkHandshakeStatusAdvances(socket);
+
+        // Make sure we can still subscribe and receive
+        receiveForSslHandshakeTest();
+    }
+
+    private void checkHandshakeStatusAdvances(SSLSocket socket) throws Exception {
+        TransportConnector connector = brokerService.getTransportConnectorByName(transportConnectorName);
+        NioSslTestUtil.checkHandshakeStatusAdvances(connector, socket);
+    }
+
+    private void connectForSslHandshakeTest() throws Exception {
+        String frame = "CONNECT\n" + "login:system\n" + "passcode:manager\n\n" + Stomp.NULL;
+        stompConnection.sendFrame(frame);
+        frame = stompConnection.receiveFrame();
+        assertTrue(frame.startsWith("CONNECTED"));
+        frame = "SEND\n" + "destination:/queue/" + getQueueName() + "\ncontent-length:5" + " \n\n" + "\u0001\u0002\u0000\u0004\u0005" + Stomp.NULL;
+        stompConnection.sendFrame(frame);
+    }
+
+    private void receiveForSslHandshakeTest() throws Exception {
+        String frame = "SUBSCRIBE\n" + "destination:/queue/" + getQueueName() + "\n" + "ack:auto\n\n" + Stomp.NULL;
+        stompConnection.sendFrame(frame);
+        StompFrame message = stompConnection.receive();
+        assertTrue(message.getAction().startsWith("MESSAGE"));
+        String length = message.getHeaders().get("content-length");
+        assertEquals("5", length);
+        assertEquals(5, message.getContent().length);
     }
 
     public void doTestAckInTransaction(boolean topic) throws Exception {
