--- a/dist/server/app-render/action-handler.js
+++ b/dist/server/app-render/action-handler.js
@@ -831,6 +831,7 @@
  * It pre-parses the FormData to ensure that any action IDs referred to are actual action IDs for
  * this Next.js application.
  */ function areAllActionIdsValid(mpaFormData, serverModuleMap) {
+    let seenActionRefs = 0;
     let hasAtLeastOneAction = false;
     // Before we attempt to decode the payload for a possible MPA action, assert that all
     // action IDs are valid IDs. If not we should disregard the payload
@@ -845,6 +846,12 @@
             }
             hasAtLeastOneAction = true;
         } else if (key.startsWith($ACTION_REF_)) {
+            if (++seenActionRefs > 2) {
+                // We only expect to see at most 2 $ACTION_REF_ fields in the form data:
+                // one from <form action="..." method="post">
+                // and one from <input action="..." type="submit">
+                return false;
+            }
             // Bound args case
             const actionDescriptorField = $ACTION_ + key.slice($ACTION_REF_.length) + ':0';
             const actionFields = mpaFormData.getAll(actionDescriptorField);
--- a/dist/esm/server/app-render/action-handler.js
+++ b/dist/esm/server/app-render/action-handler.js
@@ -804,6 +804,7 @@
  * It pre-parses the FormData to ensure that any action IDs referred to are actual action IDs for
  * this Next.js application.
  */ function areAllActionIdsValid(mpaFormData, serverModuleMap) {
+    let seenActionRefs = 0;
     let hasAtLeastOneAction = false;
     // Before we attempt to decode the payload for a possible MPA action, assert that all
     // action IDs are valid IDs. If not we should disregard the payload
@@ -818,6 +819,12 @@
             }
             hasAtLeastOneAction = true;
         } else if (key.startsWith($ACTION_REF_)) {
+            if (++seenActionRefs > 2) {
+                // We only expect to see at most 2 $ACTION_REF_ fields in the form data:
+                // one from <form action="..." method="post">
+                // and one from <input action="..." type="submit">
+                return false;
+            }
             // Bound args case
             const actionDescriptorField = $ACTION_ + key.slice($ACTION_REF_.length) + ':0';
             const actionFields = mpaFormData.getAll(actionDescriptorField);

--- a/dist/shared/lib/router/utils/prepare-destination.js
+++ b/dist/shared/lib/router/utils/prepare-destination.js
@@ -212,7 +212,8 @@
     let destHostnameCompiler;
     if (destHostname) {
         destHostnameCompiler = (0, _routematchutils.safeCompile)(destHostname, {
-            validate: false
+            validate: false,
+            encode: encodeURIComponent
         });
     }
     // update any params in query values
--- a/dist/esm/shared/lib/router/utils/prepare-destination.js
+++ b/dist/esm/shared/lib/router/utils/prepare-destination.js
@@ -182,7 +182,8 @@
     let destHostnameCompiler;
     if (destHostname) {
         destHostnameCompiler = safeCompile(destHostname, {
-            validate: false
+            validate: false,
+            encode: encodeURIComponent
         });
     }
     // update any params in query values
--- a/dist/server/lib/router-utils/proxy-request.js
+++ b/dist/server/lib/router-utils/proxy-request.js
@@ -22,6 +22,10 @@
     delete parsedUrl.query;
     parsedUrl.search = (0, _serverrouteutils.stringifyQuery)(req, query);
     const target = _url.default.format(parsedUrl);
+    // Keep in mind that a WHATWG URL's hostname and the parsedUrl's hostname
+    // are not strictly equal due to lowercasing, IDN translation, IPv4 and IPv6 normalization, etc.
+    // We just make sure this is a valid URL since http-proxy doesn't validate.
+    new URL(target);
     const HttpProxy = require('next/dist/compiled/http-proxy');
     const proxy = new HttpProxy({
         target,
--- a/dist/esm/server/lib/router-utils/proxy-request.js
+++ b/dist/esm/server/lib/router-utils/proxy-request.js
@@ -7,6 +7,10 @@
     delete parsedUrl.query;
     parsedUrl.search = stringifyQuery(req, query);
     const target = url.format(parsedUrl);
+    // Keep in mind that a WHATWG URL's hostname and the parsedUrl's hostname
+    // are not strictly equal due to lowercasing, IDN translation, IPv4 and IPv6 normalization, etc.
+    // We just make sure this is a valid URL since http-proxy doesn't validate.
+    new URL(target);
     const HttpProxy = require('next/dist/compiled/http-proxy');
     const proxy = new HttpProxy({
         target,

--- a/dist/server/app-render/action-handler.js
+++ b/dist/server/app-render/action-handler.js
@@ -130,7 +130,19 @@
     const proto = ((_getRequestMeta = (0, _requestmeta.getRequestMeta)(req, 'initProtocol')) == null ? void 0 : _getRequestMeta.replace(/:+$/, '')) || 'https';
     // For standalone or the serverful mode, use the internal origin directly
     // other than the host headers from the request.
-    const origin = process.env.__NEXT_PRIVATE_ORIGIN || `${proto}://${host.value}`;
+    let origin = process.env.__NEXT_PRIVATE_ORIGIN;
+    if (origin === undefined) {
+        const initUrl = (0, _requestmeta.getRequestMeta)(req, 'initURL');
+        if (initUrl !== undefined) {
+            try {
+                origin = new URL(initUrl).origin;
+            } catch (error) {
+                throw Object.defineProperty(new Error('Could not determine origin for forwarded Server Actions request. This can happen if port or hostname are not configured for this server.', { cause: error }), "__NEXT_ERROR_CODE", { value: "E794", enumerable: false, configurable: true });
+            }
+        } else {
+            throw Object.defineProperty(new _invarianterror.InvariantError('Missing initURL'), "__NEXT_ERROR_CODE", { value: "E793", enumerable: false, configurable: true });
+        }
+    }
     const fetchUrl = new URL(`${origin}${basePath}${workerPathname}`);
     try {
         var _response_headers_get;
@@ -229,7 +241,19 @@
         const proto = ((_getRequestMeta = (0, _requestmeta.getRequestMeta)(req, 'initProtocol')) == null ? void 0 : _getRequestMeta.replace(/:+$/, '')) || 'https';
         // For standalone or the serverful mode, use the internal origin directly
         // other than the host headers from the request.
-        const origin = process.env.__NEXT_PRIVATE_ORIGIN || `${proto}://${originalHost.value}`;
+        let origin = process.env.__NEXT_PRIVATE_ORIGIN;
+        if (origin === undefined) {
+            const initUrl = (0, _requestmeta.getRequestMeta)(req, 'initURL');
+            if (initUrl !== undefined) {
+                try {
+                    origin = new URL(initUrl).origin;
+                } catch (error) {
+                    throw Object.defineProperty(new Error('Could not determine origin for forwarded Server Actions request. This can happen if port or hostname are not configured for this server.', { cause: error }), "__NEXT_ERROR_CODE", { value: "E794", enumerable: false, configurable: true });
+                }
+            } else {
+                throw Object.defineProperty(new _invarianterror.InvariantError('Missing initURL'), "__NEXT_ERROR_CODE", { value: "E793", enumerable: false, configurable: true });
+            }
+        }
         const fetchUrl = new URL(`${origin}${appRelativeRedirectUrl.pathname}${appRelativeRedirectUrl.search}`);
         if (workStore.pendingRevalidatedTags) {
             var _workStore_incrementalCache_prerenderManifest_preview, _workStore_incrementalCache_prerenderManifest, _workStore_incrementalCache;
--- a/dist/esm/server/app-render/action-handler.js
+++ b/dist/esm/server/app-render/action-handler.js
@@ -103,7 +103,19 @@
     const proto = ((_getRequestMeta = getRequestMeta(req, 'initProtocol')) == null ? void 0 : _getRequestMeta.replace(/:+$/, '')) || 'https';
     // For standalone or the serverful mode, use the internal origin directly
     // other than the host headers from the request.
-    const origin = process.env.__NEXT_PRIVATE_ORIGIN || `${proto}://${host.value}`;
+    let origin = process.env.__NEXT_PRIVATE_ORIGIN;
+    if (origin === undefined) {
+        const initUrl = getRequestMeta(req, 'initURL');
+        if (initUrl !== undefined) {
+            try {
+                origin = new URL(initUrl).origin;
+            } catch (error) {
+                throw Object.defineProperty(new Error('Could not determine origin for forwarded Server Actions request. This can happen if port or hostname are not configured for this server.', { cause: error }), "__NEXT_ERROR_CODE", { value: "E794", enumerable: false, configurable: true });
+            }
+        } else {
+            throw Object.defineProperty(new InvariantError('Missing initURL'), "__NEXT_ERROR_CODE", { value: "E793", enumerable: false, configurable: true });
+        }
+    }
     const fetchUrl = new URL(`${origin}${basePath}${workerPathname}`);
     try {
         var _response_headers_get;
@@ -202,7 +214,19 @@
         const proto = ((_getRequestMeta = getRequestMeta(req, 'initProtocol')) == null ? void 0 : _getRequestMeta.replace(/:+$/, '')) || 'https';
         // For standalone or the serverful mode, use the internal origin directly
         // other than the host headers from the request.
-        const origin = process.env.__NEXT_PRIVATE_ORIGIN || `${proto}://${originalHost.value}`;
+        let origin = process.env.__NEXT_PRIVATE_ORIGIN;
+        if (origin === undefined) {
+            const initUrl = getRequestMeta(req, 'initURL');
+            if (initUrl !== undefined) {
+                try {
+                    origin = new URL(initUrl).origin;
+                } catch (error) {
+                    throw Object.defineProperty(new Error('Could not determine origin for forwarded Server Actions request. This can happen if port or hostname are not configured for this server.', { cause: error }), "__NEXT_ERROR_CODE", { value: "E794", enumerable: false, configurable: true });
+                }
+            } else {
+                throw Object.defineProperty(new InvariantError('Missing initURL'), "__NEXT_ERROR_CODE", { value: "E793", enumerable: false, configurable: true });
+            }
+        }
         const fetchUrl = new URL(`${origin}${appRelativeRedirectUrl.pathname}${appRelativeRedirectUrl.search}`);
         if (workStore.pendingRevalidatedTags) {
             var _workStore_incrementalCache_prerenderManifest_preview, _workStore_incrementalCache_prerenderManifest, _workStore_incrementalCache;
