diff --git a/helpers/esm/wrapRegExp.js b/helpers/esm/wrapRegExp.js
index f4cff77..369db24 100644
--- a/helpers/esm/wrapRegExp.js
+++ b/helpers/esm/wrapRegExp.js
@@ -33,9 +33,10 @@ export default function _wrapRegExp() {
   }, BabelRegExp.prototype[Symbol.replace] = function (t, p) {
     if ("string" == typeof p) {
       var o = r.get(this);
-      return e[Symbol.replace].call(this, t, p.replace(/\$<([^>]+)>/g, function (e, r) {
-        var t = o[r];
-        return "$" + (Array.isArray(t) ? t.join("$") : t);
+      return e[Symbol.replace].call(this, t, p.replace(/\$<([^>]+)(>|$)/g, function (e, r, n) {
+        if ("" === n) return e;
+        var t = Object.prototype.hasOwnProperty.call(o, r) ? o[r] : void 0;
+        return Array.isArray(t) ? "$" + t.join("$") : "number" == typeof t ? "$" + t : "";
       }));
     }
     if ("function" == typeof p) {
diff --git a/helpers/wrapRegExp.js b/helpers/wrapRegExp.js
index b104663..1dd4c1a 100644
--- a/helpers/wrapRegExp.js
+++ b/helpers/wrapRegExp.js
@@ -33,9 +33,10 @@ function _wrapRegExp() {
   }, BabelRegExp.prototype[Symbol.replace] = function (t, p) {
     if ("string" == typeof p) {
       var o = r.get(this);
-      return e[Symbol.replace].call(this, t, p.replace(/\$<([^>]+)>/g, function (e, r) {
-        var t = o[r];
-        return "$" + (Array.isArray(t) ? t.join("$") : t);
+      return e[Symbol.replace].call(this, t, p.replace(/\$<([^>]+)(>|$)/g, function (e, r, n) {
+        if ("" === n) return e;
+        var t = Object.prototype.hasOwnProperty.call(o, r) ? o[r] : void 0;
+        return Array.isArray(t) ? "$" + t.join("$") : "number" == typeof t ? "$" + t : "";
       }));
     }
     if ("function" == typeof p) {
-- 
2.43.0

