diff --git a/lib/index.js b/lib/index.js
--- a/lib/index.js
+++ b/lib/index.js
@@ -63,6 +63,10 @@
         const keys = Object.getOwnPropertyNames(obj);
         for (let i = 0; i < keys.length; ++i) {
             const key = keys[i];
+            if (key === '__proto__') {
+                continue;
+            }
+
             const descriptor = Object.getOwnPropertyDescriptor(obj, key);
             if (descriptor &&
                 (descriptor.get ||
