--- a/lib/previous-map.js
+++ b/lib/previous-map.js
@@ -88,7 +88,7 @@ var PreviousMap = function () {
     };
 
     PreviousMap.prototype.loadAnnotation = function loadAnnotation(css) {
-        var match = css.match(/\/\*\s*# sourceMappingURL=(.*)\s*\*\//);
+        var match = css.match(/\/\*\s*# sourceMappingURL=((?:(?!sourceMappingURL=).)*)\*\//);
         if (match) this.annotation = match[1].trim();
     };
