---
 dist-src/iterator.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dist-src/iterator.js b/dist-src/iterator.js
index 2bd96f8..2459555 100644
--- a/dist-src/iterator.js
+++ b/dist-src/iterator.js
@@ -14,7 +14,7 @@ function iterator(octokit, route, parameters) {
           const response = await requestMethod({ method, url, headers });
           const normalizedResponse = normalizePaginatedListResponse(response);
           url = ((normalizedResponse.headers.link || "").match(
-            /<([^>]+)>;\s*rel="next"/
+            /<([^<>]+)>;\s*rel="next"/
           ) || [])[1];
           return { value: normalizedResponse };
         } catch (error) {
-- 
2.43.0

