diff --git a/fesm2022/debug_node.mjs b/fesm2022/debug_node.mjs
index be43067..0f0075e 100755
--- a/fesm2022/debug_node.mjs
+++ b/fesm2022/debug_node.mjs
@@ -24017,7 +24017,7 @@ var localeEn = ["en", [["a", "p"], ["AM", "PM"]], [["AM", "PM"]], [["S", "M", "T
 /**
  * This const is used to store the locale data registered with `registerLocaleData`
  */
-let LOCALE_DATA = {};
+let LOCALE_DATA = /* @__PURE__ */ Object.create(null);
 /**
  * Register locale data to be used internally by Angular. See the
  * ["I18n guide"](guide/i18n/format-data-locale) to know how to import additional locale
@@ -24103,7 +24103,7 @@ function getLocaleData(normalizedLocale) {
  * Helper function to remove all the locale data from `LOCALE_DATA`.
  */
 function unregisterAllLocaleData() {
-    LOCALE_DATA = {};
+    LOCALE_DATA = Object.create(null);
 }
 /**
  * Index of each type of locale data from the locale data array
