--- a/index.js
+++ b/index.js
@@ -80,6 +80,21 @@
 /** Used to match unescaped characters in compiled string literals. */
 var reUnescapedString = /['\n\r\u2028\u2029\\]/g;
 
+/** Used as error message for invalid template variable option. */
+var INVALID_TEMPL_VAR_ERROR_TEXT = 'Invalid `variable` option passed into `_.template`';
+
+/**
+ * Used to validate the `variable` option in `_.template`.
+ *
+ * Forbids characters which could potentially change the meaning of the function argument definition:
+ * - "()," (modification of function parameters)
+ * - "=" (default value)
+ * - "[]{}" (destructuring of function parameters)
+ * - "/" (beginning of a comment)
+ * - whitespace
+ */
+var reForbiddenIdentifierChars = /[()=,{}\[\]\/\s]/;
+
 /** Used to identify `toStringTag` values of typed arrays. */
 var typedArrayTags = {};
 typedArrayTags[float32Tag] = typedArrayTags[float64Tag] =
@@ -1522,6 +1537,11 @@
   if (!variable) {
     source = 'with (obj) {\n' + source + '\n}\n';
   }
+  // Throw an error if a forbidden character was found in `variable`, to prevent
+  // potential command injection attacks.
+  else if (reForbiddenIdentifierChars.test(variable)) {
+    throw new Error(INVALID_TEMPL_VAR_ERROR_TEXT);
+  }
   // Cleanup code by stripping empty strings.
   source = (isEvaluating ? source.replace(reEmptyStringLeading, '') : source)
     .replace(reEmptyStringMiddle, '$1')

--- a/index.js
+++ b/index.js
@@ -81,6 +81,9 @@
 /** Used as error message for invalid template variable option. */
 var INVALID_TEMPL_VAR_ERROR_TEXT = 'Invalid `variable` option passed into `_.template`';
 
+/** Used as error message for invalid template imports option. */
+var INVALID_TEMPL_IMPORTS_ERROR_TEXT = 'Invalid `imports` option passed into `_.template`';
+
 /**
  * Used to validate the `variable` option in `_.template`.
  *
@@ -1478,6 +1481,12 @@
       importsKeys = keys(imports),
       importsValues = baseValues(imports, importsKeys);
 
+  for (var i = 0; i < importsKeys.length; i++) {
+    if (reForbiddenIdentifierChars.test(importsKeys[i])) {
+      throw new Error(INVALID_TEMPL_IMPORTS_ERROR_TEXT);
+    }
+  }
+
   var isEscaping,
       isEvaluating,
       index = 0,
