--- a/core/src/main/java/org/bouncycastle/pqc/crypto/frodo/FrodoEngine.java
+++ b/core/src/main/java/org/bouncycastle/pqc/crypto/frodo/FrodoEngine.java
@@ -440,20 +440,18 @@
     }


-    private short ctverify(short[] a1, short[] a2, short[] b1, short[] b2)
+    private static short ctverify(short[] a1, short[] a2, short[] b1, short[] b2)
     {
         // Compare two arrays in constant time.
         // Returns 0 if the byte arrays are equal, -1 otherwise.
-        short r = 0;
+        int r = 0;

-        for (short i = 0; i < a1.length; i++)
+        for (int i = 0; i < a1.length; i++)
             r |= a1[i] ^ b1[i];

-        for (short i = 0; i < a2.length; i++)
+        for (int i = 0; i < a2.length; i++)
             r |= a2[i] ^ b2[i];

-//        r = (short) ((-(short)(r >> 1) | -(short)(r & 1)) >> (8*2-1));
-        if (r == 0)
-            return 0;
-        return -1;
+        // Constant-time: returns 0 if equal, -1 if not equal.
+        return (short)((r | (-r)) >> 31);
     }
