--- a/index.js
+++ b/index.js
@@ -150,7 +150,7 @@
     var x = numeric(n[0]);
     var y = numeric(n[1]);
     var width = Math.max(n[0].length, n[1].length)
     var incr = n.length == 3
-      ? Math.abs(numeric(n[2]))
+      ? Math.max(Math.abs(numeric(n[2])), 1)
       : 1;
     var test = lte;
     var reverse = y < x;

--- a/index.js
+++ b/index.js
@@ -109,7 +109,7 @@
   var isOptions = m.body.indexOf(',') >= 0;
   if (!isSequence && !isOptions) {
     // {a},b}
-    if (m.post.match(/,.*\}/)) {
+    if (m.post.match(/,(?!,).*\}/)) {
       str = m.pre + '{' + m.body + escClose + m.post;
       return expand(str);
     }
--- /dev/null
+++ b/test/redos.js
@@ -0,0 +1,13 @@
+var test = require('tape');
+var expand = require('..');
+
+test('redos', function(t) {
+  var str = '{a}' + ','.repeat(100000) + ' ';
+  var startTime = Date.now();
+  expand(str);
+  var endTime = Date.now();
+  var timeTaken = endTime - startTime;
+  t.ok(timeTaken < 1000, 'Expected time (' + timeTaken + 'ms) to be less than 1000ms');
+  t.end();
+});
+
