diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/index.js target_draft/index.js
--- a/index.js	2026-01-23 09:10:24.122762649 +0000
+++ b/index.js	2026-01-23 09:20:04.007790761 +0000
@@ -1,13 +1,37 @@
 'use strict';
 
 var fn = module.exports = function (x) {
-	return fn.end(fn.start(x));
+	let start = 0;
+	let end = x.length;
+
+	while (start < end && (x[start] === '\r' || x[start] === '\n')) {
+		start++;
+	}
+
+	while (end > 0 && (x[end - 1] === '\r' || x[end - 1] === '\n')) {
+		end--;
+	}
+
+	return (start > 0 || end < x.length) ? x.slice(start, end) : x;
 };
 
 fn.start = function (x) {
-	return x.replace(/^[\r\n]+/, '');
+	const end = x.length;
+	let start = 0;
+
+	while (start < end && (x[start] === '\r' || x[start] === '\n')) {
+		start++;
+	}
+
+	return start > 0 ? x.slice(start, end) : x;
 };
 
 fn.end = function (x) {
-	return x.replace(/[\r\n]+$/, '');
+	let end = x.length;
+
+	while (end > 0 && (x[end - 1] === '\r' || x[end - 1] === '\n')) {
+		end--;
+	}
+
+	return end < x.length ? x.slice(0, end) : x;
 };

