--- a/aiohttp/http_parser.py
+++ b/aiohttp/http_parser.py
@@ -847,6 +847,13 @@
                         i = chunk.find(CHUNK_EXT, 0, pos)
                         if i >= 0:
                             size_b = chunk[:i]  # strip chunk-extensions
+                            # Verify no LF in the chunk-extension
+                            if b"\n" in (ext := chunk[i:pos]):
+                                exc = BadHttpMessage(
+                                    f"Unexpected LF in chunk-extension: {ext!r}"
+                                )
+                                set_exception(self.payload, exc)
+                                raise exc
                         else:
                             size_b = chunk[:pos]


--- a/aiohttp/multipart.py
+++ b/aiohttp/multipart.py
@@ -325,11 +325,8 @@
         self._read_bytes += len(chunk)
         if self._read_bytes == self._length:
             self._at_eof = True
-        if self._at_eof:
-            clrf = await self._content.readline()
-            assert (
-                b"\r\n" == clrf
-            ), "reader did not read all the data or it is malformed"
+        if self._at_eof and await self._content.readline() != b"\r\n":
+            raise ValueError("Reader did not read all the data or it is malformed")
         return chunk

     async def _read_chunk_from_length(self, size: int) -> bytes:
@@ -354,7 +351,8 @@

         chunk = await self._content.read(size)
         self._content_eof += int(self._content.at_eof())
-        assert self._content_eof < 3, "Reading after EOF"
+        if self._content_eof > 2:
+            raise ValueError("Reading after EOF")
         assert self._prev_chunk is not None
         window = self._prev_chunk + chunk
         sub = b"\r\n" + self._boundary
--- a/aiohttp/web_request.py
+++ b/aiohttp/web_request.py
@@ -712,13 +712,13 @@
             multipart = await self.multipart()
             max_size = self._client_max_size

-            field = await multipart.next()
-            while field is not None:
-                size = 0
+            while (field := await multipart.next()) is not None:
+                size = 0
                 field_ct = field.headers.get(hdrs.CONTENT_TYPE)

                 if isinstance(field, BodyPartReader):
-                    assert field.name is not None
+                    if field.name is None:
+                        raise ValueError("Multipart field missing name.")

                     # Note that according to RFC 7578, the Content-Type header
                     # is optional, even for files, so we can't assume it's
@@ -769,9 +769,7 @@
                 else:
                     raise ValueError(
                         "To decode nested multipart you need " "to use custom reader",
                     )
-
-                field = await multipart.next()
         else:
             data = await self.read()
             if data:
