--- a/cairosvg/defs.py
+++ b/cairosvg/defs.py
@@ -331,6 +331,9 @@

 def use(surface, node):
     """Draw the content of another SVG node."""
+    surface.reference_count += 1
+    if not node.unsafe and surface.reference_count > 100_000:
+        raise ValueError('Abort rendering: more than 100 000 referenced elements')
     surface.context.save()
     surface.context.translate(
         size(surface, node.get('x'), 'x'), size(surface, node.get('y'), 'y'))
--- a/cairosvg/surface.py
+++ b/cairosvg/surface.py
@@ -159,6 +159,7 @@
         self.cursor_d_position = [0, 0]
         self.text_path_width = 0
         self.tree_cache = {(tree.url, tree.get('id')): tree}
+        self.reference_count = 0
         if parent_surface:
             self.markers = parent_surface.markers
             self.gradients = parent_surface.gradients
