diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/lib/form_data.js target_draft/lib/form_data.js
--- a/lib/form_data.js	2025-12-02 06:56:39.091230700 +0000
+++ b/lib/form_data.js	2025-12-02 07:05:00.437096801 +0000
@@ -5,6 +5,7 @@
 var https = require('https');
 var parseUrl = require('url').parse;
 var fs = require('fs');
+var crypto = require('crypto');
 var mime = require('mime-types');
 var asynckit = require('asynckit');
 var populate = require('./populate.js');
@@ -346,12 +347,7 @@
 FormData.prototype._generateBoundary = function() {
   // This generates a 50 character boundary similar to those used by Firefox.
   // They are optimized for boyer-moore parsing.
-  var boundary = '--------------------------';
-  for (var i = 0; i < 24; i++) {
-    boundary += Math.floor(Math.random() * 10).toString(16);
-  }
-
-  this._boundary = boundary;
+  this._boundary = '--------------------------' + crypto.randomBytes(12).toString('hex');
 };
 
 // Note: getLengthSync DOESN'T calculate streams length
diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/package.json target_draft/package.json
--- a/package.json	2025-12-02 06:56:39.091230700 +0000
+++ b/package.json	2025-12-02 07:03:02.115679604 +0000
@@ -56,9 +56,10 @@
     "is-node-modern": "^1.0.0",
     "istanbul": "^0.4.5",
     "obake": "^0.1.2",
-    "puppeteer": "^1.19.0",
     "pkgfiles": "^2.3.0",
-    "pre-commit": "^1.1.3",
+    "pre-commit": "^1.2.2",
+    "predict-v8-randomness": "^1.0.35",
+    "puppeteer": "^1.20.0",
     "request": "^2.88.0",
     "rimraf": "^2.7.1",
     "tape": "^4.6.2",
diff -urN '--exclude=.git' '--exclude=node_modules' '--exclude=.cache' '--exclude=*.log' --no-dereference '--unified=3' target/test/integration/test-boundary-prediction.js target_draft/test/integration/test-boundary-prediction.js
--- a/test/integration/test-boundary-prediction.js	1970-01-01 00:00:00.000000000 +0000
+++ b/test/integration/test-boundary-prediction.js	2025-12-02 07:04:02.516299136 +0000
@@ -0,0 +1,57 @@
+var common = require('../common');
+var assert = common.assert;
+var FormData = require(common.dir.lib + '/form_data');
+var predictV8Randomness = require('predict-v8-randomness');
+
+var initialSequence = [
+  Math.random(),
+  Math.random(),
+  Math.random(),
+  Math.random(),
+];
+var predictor = new predictV8Randomness.Predictor(initialSequence);
+
+predictor.predictNext(24).then(function (next24RandomOutputs) {
+  var predictedBoundary = next24RandomOutputs
+    .map(function (v) {
+      return Math.floor(v * 10).toString(16);
+    })
+    .join('');
+
+  var boundaryIntro = '----------------------------';
+
+  var payload =
+    'zzz\r\n' +
+    boundaryIntro +
+    predictedBoundary +
+    '\r\nContent-Disposition: form-data; name="is_admin"\r\n\r\ntrue\r\n' +
+    boundaryIntro +
+    predictedBoundary +
+    '--\r\n';
+
+  var FIELDS = {
+    my_field: {
+      value: payload,
+    },
+  };
+
+  // count total
+  var fieldsPassed = Object.keys(FIELDS).length;
+
+  // prepare form-receiving http server
+  var server = common.testFields(FIELDS, function (fields) {
+    fieldsPassed = fields;
+  });
+
+  server.listen(common.port, function () {
+    var form = new FormData();
+
+    common.actions.populateFields(form, FIELDS);
+
+    common.actions.submit(form, server);
+  });
+
+  process.on('exit', function () {
+    assert.strictEqual(fieldsPassed, 0);
+  });
+});
\ No newline at end of file
