--- a/java/org/apache/catalina/realm/RealmBase.java
+++ b/java/org/apache/catalina/realm/RealmBase.java
@@ -1145,12 +1145,19 @@
      * @return the digest for the specified user
      */
     protected String getDigest(String username, String realmName, String algorithm) {
+        String password = getPassword(username);
+
+        // Short-cut null password case
+        if (password == null) {
+            return null;
+        }
+
         if (hasMessageDigest(algorithm)) {
             // Use pre-generated digest
-            return getPassword(username);
+            return password;
         }

-        String digestValue = username + ":" + realmName + ":" + getPassword(username);
+        String digestValue = username + ":" + realmName + ":" + password;

         byte[] valueBytes;
         try {

--- a/java/org/apache/tomcat/websocket/Authenticator.java
+++ b/java/org/apache/tomcat/websocket/Authenticator.java
@@ -21,6 +21,7 @@
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import org.apache.tomcat.util.http.Method;
 import org.apache.tomcat.util.res.StringManager;
 
 /**
@@ -36,21 +37,46 @@
 
 
     /**
-     * Generate the authorization header value that will be sent to the server.
+     * Generate the authorization header value that will be sent to the server.
      *
-     * @param requestUri         The request URI
-     * @param authenticateHeader The server authentication header received
-     * @param userName           The username
-     * @param userPassword       The user password
-     * @param userRealm          The realm for which the provided username and password are valid. {@code null} to
-     *                               indicate all realms.
+     * @param requestUri         The request URI
+     * @param authenticateHeader The server authentication header received
+     * @param userName           The username
+     * @param userPassword       The user password
+     * @param userRealm          The realm for which the provided username and password are valid. {@code null} to
+     *                               indicate all realms.
      *
-     * @return The generated authorization header value
+     * @return The generated authorization header value
      *
      * @throws AuthenticationException When an error occurs
+     *
+     * @deprecated Unused. Will be remove in Tomcat 12. Use
+     *                 {@link #getAuthorization(String, String, String, String, String, String)}
+     */
+    @Deprecated
+    public String getAuthorization(String requestUri, String authenticateHeader, String userName, String userPassword,
+            String userRealm) throws AuthenticationException {
+        return getAuthorization(Method.GET, requestUri, authenticateHeader, userName, userPassword, userRealm);
+    }
+
+
+    /**
+     * Generate the authorization header value that will be sent to the server.
+     *
+     * @param method             The request method
+     * @param requestUri         The request URI
+     * @param authenticateHeader The server authentication header received
+     * @param userName           The username
+     * @param userPassword       The user password
+     * @param userRealm          The realm for which the provided username and password are valid. {@code null} to
+     *                               indicate all realms.
+     *
+     * @return The generated authorization header value
+     *
+     * @throws AuthenticationException When an error occurs
      */
-    public abstract String getAuthorization(String requestUri, String authenticateHeader, String userName,
-            String userPassword, String userRealm) throws AuthenticationException;
+    public abstract String getAuthorization(String method, String requestUri, String authenticateHeader,
+            String userName, String userPassword, String userRealm) throws AuthenticationException;
 
 
     /**
--- a/java/org/apache/tomcat/websocket/BasicAuthenticator.java
+++ b/java/org/apache/tomcat/websocket/BasicAuthenticator.java
@@ -30,8 +30,8 @@
     public static final String charsetparam = "charset";
 
     @Override
-    public String getAuthorization(String requestUri, String authenticateHeader, String userName, String userPassword,
-            String userRealm) throws AuthenticationException {
+    public String getAuthorization(String method, String requestUri, String authenticateHeader, String userName,
+            String userPassword, String userRealm) throws AuthenticationException {
 
         validateUsername(userName);
         validatePassword(userPassword);
--- a/java/org/apache/tomcat/websocket/DigestAuthenticator.java
+++ b/java/org/apache/tomcat/websocket/DigestAuthenticator.java
@@ -40,8 +40,8 @@
     private long cNonce;
 
     @Override
-    public String getAuthorization(String requestUri, String authenticateHeader, String userName, String userPassword,
-            String userRealm) throws AuthenticationException {
+    public String getAuthorization(String method, String requestUri, String authenticateHeader, String userName,
+            String userPassword, String userRealm) throws AuthenticationException {
 
         validateUsername(userName);
         validatePassword(userPassword);
@@ -79,8 +79,8 @@
 
         try {
             challenge.append("response=\"");
-            challenge.append(
-                    calculateRequestDigest(requestUri, userName, userPassword, realm, nonce, messageQop, algorithm));
+            challenge.append(calculateRequestDigest(method, requestUri, userName, userPassword, realm, nonce,
+                    messageQop, algorithm));
             challenge.append("\",");
         }
 
@@ -89,7 +89,9 @@
         }
 
         challenge.append("algorithm=").append(algorithm).append(",");
-        challenge.append("opaque=\"").append(opaque).append("\",");
+        if (opaque != null) {
+            challenge.append("opaque=\"").append(opaque).append("\",");
+        }
 
         if (!messageQop.isEmpty()) {
             challenge.append("qop=\"").append(messageQop).append("\"");
@@ -101,8 +103,8 @@
 
     }
 
-    private String calculateRequestDigest(String requestUri, String userName, String password, String realm,
-            String nonce, String qop, String algorithm) throws NoSuchAlgorithmException {
+    private String calculateRequestDigest(String method, String requestUri, String userName, String password,
+            String realm, String nonce, String qop, String algorithm) throws NoSuchAlgorithmException {
 
         boolean session = false;
         if (algorithm.endsWith("-sess")) {
@@ -123,7 +125,7 @@
          * If the "qop" value is "auth-int", then A2 is: A2 = Method ":" digest-uri-value ":" H(entity-body) since we do
          * not have an entity-body, A2 = Method ":" digest-uri-value for auth and auth_int
          */
-        String A2 = "GET:" + requestUri;
+        String A2 = method + ":" + requestUri;
 
         preDigest.append(encode(algorithm, A1));
         preDigest.append(':');
--- a/java/org/apache/tomcat/websocket/WsWebSocketContainer.java
+++ b/java/org/apache/tomcat/websocket/WsWebSocketContainer.java
@@ -75,6 +75,7 @@
 import org.apache.tomcat.InstanceManagerBindings;
 import org.apache.tomcat.util.buf.StringUtils;
 import org.apache.tomcat.util.collections.CaseInsensitiveKeyMap;
+import org.apache.tomcat.util.http.Method;
 import org.apache.tomcat.util.res.StringManager;
 import org.apache.tomcat.util.security.KeyStoreUtil;
 
@@ -181,40 +182,41 @@
 
 
     private Session connectToServerRecursive(ClientEndpointHolder clientEndpointHolder,
-            ClientEndpointConfig clientEndpointConfiguration, URI path, Set<URI> redirectSet)
+            ClientEndpointConfig clientEndpointConfiguration, URI serverEndpointUri, Set<URI> redirectSet)
             throws DeploymentException {
 
         if (log.isTraceEnabled()) {
-            log.trace(sm.getString("wsWebSocketContainer.connect.entry", clientEndpointHolder.getClassName(), path));
+            log.trace(sm.getString("wsWebSocketContainer.connect.entry", clientEndpointHolder.getClassName(),
+                    serverEndpointUri));
         }
 
         boolean secure = false;
         ByteBuffer proxyConnect = null;
-        URI proxyPath;
+        URI proxyUri;
 
         // Validate scheme (and build proxyPath)
-        String scheme = path.getScheme();
+        String scheme = serverEndpointUri.getScheme();
         if ("ws".equalsIgnoreCase(scheme)) {
-            proxyPath = URI.create("http" + path.toString().substring(2));
+            proxyUri = URI.create("http" + serverEndpointUri.toString().substring(2));
         } else if ("wss".equalsIgnoreCase(scheme)) {
-            proxyPath = URI.create("https" + path.toString().substring(3));
+            proxyUri = URI.create("https" + serverEndpointUri.toString().substring(3));
             secure = true;
         } else {
             throw new DeploymentException(sm.getString("wsWebSocketContainer.pathWrongScheme", scheme));
         }
 
-        // Validate host
-        String host = path.getHost();
-        if (host == null) {
+        // Validate server endpoint host
+        String serverEndpointHost = serverEndpointUri.getHost();
+        if (serverEndpointHost == null) {
             throw new DeploymentException(sm.getString("wsWebSocketContainer.pathNoHost"));
         }
-        int port = path.getPort();
+        int serverEndpointPort = serverEndpointUri.getPort();
 
         SocketAddress sa = null;
 
         // Check to see if a proxy is configured. Javadoc indicates return value
         // will never be null
-        List<Proxy> proxies = ProxySelector.getDefault().select(proxyPath);
+        List<Proxy> proxies = ProxySelector.getDefault().select(proxyUri);
         Proxy selectedProxy = null;
         for (Proxy proxy : proxies) {
             if (proxy.type().equals(Proxy.Type.HTTP)) {
@@ -232,34 +234,36 @@
 
         // If the port is not explicitly specified, compute it based on the
         // scheme
-        if (port == -1) {
+        if (serverEndpointPort == -1) {
             if ("ws".equalsIgnoreCase(scheme)) {
-                port = 80;
+                serverEndpointPort = 80;
             } else {
                 // Must be wss due to scheme validation above
-                port = 443;
+                serverEndpointPort = 443;
             }
         }
 
         Map<String,Object> userProperties = clientEndpointConfiguration.getUserProperties();
 
         // If sa is null, no proxy is configured so need to create sa
         if (sa == null) {
-            sa = new InetSocketAddress(host, port);
+            sa = new InetSocketAddress(serverEndpointHost, serverEndpointPort);
         } else {
-            proxyConnect = createProxyRequest(host, port,
+            proxyConnect = createProxyRequest(serverEndpointHost, serverEndpointPort,
                     (String) userProperties.get(Constants.PROXY_AUTHORIZATION_HEADER_NAME));
         }
 
         // Create the initial HTTP request to open the WebSocket connection
-        Map<String,List<String>> reqHeaders = createRequestHeaders(host, port, secure, clientEndpointConfiguration);
-        clientEndpointConfiguration.getConfigurator().beforeRequest(reqHeaders);
-        if (Constants.DEFAULT_ORIGIN_HEADER_VALUE != null && !reqHeaders.containsKey(Constants.ORIGIN_HEADER_NAME)) {
+        Map<String,List<String>> upgradeRequestHeaders =
+                createRequestHeaders(serverEndpointHost, serverEndpointPort, secure, clientEndpointConfiguration);
+        clientEndpointConfiguration.getConfigurator().beforeRequest(upgradeRequestHeaders);
+        if (Constants.DEFAULT_ORIGIN_HEADER_VALUE != null &&
+                !upgradeRequestHeaders.containsKey(Constants.ORIGIN_HEADER_NAME)) {
             List<String> originValues = new ArrayList<>(1);
             originValues.add(Constants.DEFAULT_ORIGIN_HEADER_VALUE);
-            reqHeaders.put(Constants.ORIGIN_HEADER_NAME, originValues);
+            upgradeRequestHeaders.put(Constants.ORIGIN_HEADER_NAME, originValues);
         }
-        ByteBuffer request = createRequest(path, reqHeaders);
+        ByteBuffer upgradeRequest = createRequest(serverEndpointUri, upgradeRequestHeaders);
 
         // Get the connection timeout
         long timeout = Constants.IO_TIMEOUT_MS_DEFAULT;
@@ -289,12 +293,15 @@
                 writeRequest(channel, proxyConnect, timeout);
                 HttpResponse httpResponse = processResponse(response, channel, timeout);
                 if (httpResponse.status == Constants.PROXY_AUTHENTICATION_REQUIRED) {
-                    return processAuthenticationChallenge(clientEndpointHolder, clientEndpointConfiguration, path,
-                            redirectSet, userProperties, request, httpResponse, AuthenticationType.PROXY);
+                    return processAuthenticationChallenge(clientEndpointHolder, clientEndpointConfiguration,
+                            serverEndpointUri, redirectSet, userProperties, Method.CONNECT,
+                            serverEndpointHost + ":" + serverEndpointPort, httpResponse, AuthenticationType.PROXY);
                 } else if (httpResponse.getStatus() != 200) {
                     throw new DeploymentException(sm.getString("wsWebSocketContainer.proxyConnectFail", selectedProxy,
                             Integer.toString(httpResponse.getStatus())));
                 }
+                // Proxy authentication either successful or not required.
+                userProperties.remove(Constants.PROXY_AUTHORIZATION_HEADER_NAME);
             }
 
             if (secure) {
@@ -302,6 +309,7 @@
                 // proxy CONNECT, need to use TLS from this point on so wrap the
                 // original AsynchronousSocketChannel
-                SSLEngine sslEngine = createSSLEngine(clientEndpointConfiguration, host, port);
+                SSLEngine sslEngine =
+                        createSSLEngine(clientEndpointConfiguration, serverEndpointHost, serverEndpointPort);
                 channel = new AsyncChannelWrapperSecure(socketChannel, sslEngine);
             } else if (channel == null) {
                 // Only need to wrap as this point if it wasn't wrapped to process a
@@ -321,10 +329,10 @@
                 } catch (IOException ioe) {
                     // Ignore
                 }
-                log.trace(sm.getString("wsWebSocketContainer.connect.write", Integer.valueOf(request.position()),
-                        Integer.valueOf(request.limit()), localAddress));
+                log.trace(sm.getString("wsWebSocketContainer.connect.write", Integer.valueOf(upgradeRequest.position()),
+                        Integer.valueOf(upgradeRequest.limit()), localAddress));
             }
-            writeRequest(channel, request, timeout);
+            writeRequest(channel, upgradeRequest, timeout);
 
             HttpResponse httpResponse = processResponse(response, channel, timeout);
 
@@ -342,6 +350,9 @@
 
             if (httpResponse.status != 101) {
                 if (isRedirectStatus(httpResponse.status)) {
+                    // HTTP redirect. Authentication either successful or not required.
+                    userProperties.remove(Constants.AUTHORIZATION_HEADER_NAME);
+
                     List<String> locationHeader =
                             httpResponse.getHandshakeResponse().getHeaders().get(Constants.LOCATION_HEADER_NAME);
 
@@ -355,7 +366,7 @@
                     URI redirectLocation = URI.create(locationHeader.get(0)).normalize();
 
                     if (!redirectLocation.isAbsolute()) {
-                        redirectLocation = path.resolve(redirectLocation);
+                        redirectLocation = serverEndpointUri.resolve(redirectLocation);
                     }
 
                     String redirectScheme = redirectLocation.getScheme().toLowerCase(Locale.ENGLISH);
@@ -376,7 +387,10 @@
 
                 } else if (httpResponse.status == Constants.UNAUTHORIZED) {
-                    return processAuthenticationChallenge(clientEndpointHolder, clientEndpointConfiguration, path,
-                            redirectSet, userProperties, request, httpResponse, AuthenticationType.WWW);
+                    String authenticationUri =
+                            new String(upgradeRequest.array(), StandardCharsets.ISO_8859_1).split("\\s", 3)[1];
+                    return processAuthenticationChallenge(clientEndpointHolder, clientEndpointConfiguration,
+                            serverEndpointUri, redirectSet, userProperties, Method.GET, authenticationUri, httpResponse,
+                            AuthenticationType.WWW);
 
                 } else {
                     throw new DeploymentException(
@@ -384,6 +398,9 @@
                             sm.getString("wsWebSocketContainer.invalidStatus", Integer.toString(httpResponse.status)));
                 }
             }
+            // HTTP upgrade successful. Authentication either successful or not required.
+            userProperties.remove(Constants.AUTHORIZATION_HEADER_NAME);
+
             HandshakeResponse handshakeResponse = httpResponse.getHandshakeResponse();
             clientEndpointConfiguration.getConfigurator().afterResponse(handshakeResponse);
 
@@ -425,6 +441,6 @@
         } catch (ExecutionException | InterruptedException | SSLException | EOFException | TimeoutException |
                 URISyntaxException | AuthenticationException e) {
-            throw new DeploymentException(sm.getString("wsWebSocketContainer.httpRequestFailed", path), e);
+            throw new DeploymentException(sm.getString("wsWebSocketContainer.httpRequestFailed", serverEndpointUri), e);
         } finally {
             if (!success) {
                 if (channel != null) {
@@ -470,8 +486,9 @@
 
 
     private Session processAuthenticationChallenge(ClientEndpointHolder clientEndpointHolder,
-            ClientEndpointConfig clientEndpointConfiguration, URI path, Set<URI> redirectSet,
-            Map<String,Object> userProperties, ByteBuffer request, HttpResponse httpResponse,
+            ClientEndpointConfig clientEndpointConfiguration, URI serverEndpointUri, Set<URI> redirectSet,
+            Map<String,Object> userProperties, String authenticationMethod, String authenticationUri,
+            HttpResponse httpResponse,
             AuthenticationType authenticationType) throws DeploymentException, AuthenticationException {
 
         if (userProperties.get(authenticationType.getAuthorizationHeaderName()) != null) {
@@ -495,14 +512,13 @@
                     Integer.valueOf(httpResponse.status), authScheme));
         }
 
-        String requestUri = new String(request.array(), StandardCharsets.ISO_8859_1).split("\\s", 3)[1];
-
         userProperties.put(authenticationType.getAuthorizationHeaderName(),
-                auth.getAuthorization(requestUri, authenticateHeaders.get(0),
+                auth.getAuthorization(authenticationMethod, authenticationUri, authenticateHeaders.get(0),
                         (String) userProperties.get(authenticationType.getUserNameProperty()),
                         (String) userProperties.get(authenticationType.getUserPasswordProperty()),
                         (String) userProperties.get(authenticationType.getUserRealmProperty())));
 
-        return connectToServerRecursive(clientEndpointHolder, clientEndpointConfiguration, path, redirectSet);
+        return connectToServerRecursive(clientEndpointHolder, clientEndpointConfiguration, serverEndpointUri,
+                redirectSet);
     }
 

--- a/java/org/apache/catalina/realm/LockOutRealm.java
+++ b/java/org/apache/catalina/realm/LockOutRealm.java
@@ -19,6 +19,7 @@ package org.apache.catalina.realm;
 import java.security.Principal;
 import java.security.cert.X509Certificate;
 import java.util.LinkedHashMap;
+import java.util.Locale;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicInteger;

@@ -71,6 +72,11 @@ public class LockOutRealm extends CombinedRealm {
      */
     protected Map<String,LockRecord> failedUsers = null;

+    /*
+     * Should user names (the keys) in the failedUsers Map be treated in a case sensitive manner. The default is false.
+     */
+    private boolean caseSensitive = false;
+

     @Override
     protected void startInternal() throws LifecycleException {
@@ -205,6 +211,9 @@ public class LockOutRealm extends CombinedRealm {
      * time will be recorded and any attempt to authenticate a locked user will log a warning.
      */
     public boolean isLocked(String username) {
+        if (!getCaseSensitive()) {
+            username = username.toLowerCase(Locale.ROOT);
+        }
         LockRecord lockRecord;
         synchronized (this) {
             lockRecord = failedUsers.get(username);
@@ -227,6 +236,9 @@ public class LockOutRealm extends CombinedRealm {
      * After successful authentication, any record of previous authentication failure is removed.
      */
     private synchronized void registerAuthSuccess(String username) {
+        if (!getCaseSensitive()) {
+            username = username.toLowerCase(Locale.ROOT);
+        }
         // Successful authentication means removal from the list of failed users
         failedUsers.remove(username);
     }
@@ -236,6 +248,9 @@ public class LockOutRealm extends CombinedRealm {
      * After a failed authentication, add the record of the failed authentication.
      */
     private void registerAuthFailure(String username) {
+        if (!getCaseSensitive()) {
+            username = username.toLowerCase(Locale.ROOT);
+        }
         LockRecord lockRecord;
         synchronized (this) {
             if (!failedUsers.containsKey(username)) {
@@ -337,6 +352,16 @@ public class LockOutRealm extends CombinedRealm {
     }


+    public boolean getCaseSensitive() {
+        return caseSensitive;
+    }
+
+
+    public void setCaseSensitive(boolean caseSensitive) {
+        this.caseSensitive = caseSensitive;
+    }
+
+
     protected static class LockRecord {
         private final AtomicInteger failures = new AtomicInteger(0);
         private long lastFailureTime = 0;
--- /dev/null
+++ b/test/org/apache/catalina/realm/TestLockoutRealm.java
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.catalina.realm;
+
+import java.security.Principal;
+import java.util.Locale;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import org.apache.catalina.Context;
+import org.apache.catalina.startup.TesterMapRealm;
+import org.apache.tomcat.unittest.TesterContext;
+
+public class TestLockoutRealm {
+
+    private static final String USER_NAME = "user";
+    private static final String PASSWORD = "password";
+
+    private LockOutRealm realm;
+
+
+    @Before
+    public void init() throws Exception {
+        Context context = new TesterContext();
+        TesterMapRealm tmr = new TesterMapRealm();
+        tmr.setContainer(context);
+        MessageDigestCredentialHandler ch = new MessageDigestCredentialHandler();
+        tmr.setCredentialHandler(ch);
+        tmr.addUser(USER_NAME, PASSWORD);
+        tmr.start();
+
+        realm = new LockOutRealm();
+        realm.setContainer(context);
+        realm.addRealm(tmr);
+        realm.setFailureCount(2);
+        realm.start();
+    }
+
+
+    @Test
+    public void testLockoutAfterFailure() {
+        Principal p = realm.authenticate(USER_NAME, PASSWORD);
+        Assert.assertNotNull(p);
+
+        p = realm.authenticate(USER_NAME, "wrong");
+        p = realm.authenticate(USER_NAME, "wrong");
+        // Should be locked now
+        p = realm.authenticate(USER_NAME, PASSWORD);
+        Assert.assertNull(p);
+    }
+
+
+    @Test
+    public void testLockoutAfterFailureCaseSensitiveDefault() {
+        Principal p = realm.authenticate(USER_NAME, PASSWORD);
+        Assert.assertNotNull(p);
+
+        p = realm.authenticate(USER_NAME, "wrong");
+        p = realm.authenticate(USER_NAME.toUpperCase(Locale.ENGLISH), "wrong");
+        // Should be locked now
+        p = realm.authenticate(USER_NAME, PASSWORD);
+        Assert.assertNull(p);
+    }
+
+
+    @Test
+    public void testLockoutAfterFailureCaseSensitiveFalse() {
+        realm.setCaseSensitive(false);
+
+        Principal p = realm.authenticate(USER_NAME, PASSWORD);
+        Assert.assertNotNull(p);
+
+        p = realm.authenticate(USER_NAME, "wrong");
+        p = realm.authenticate(USER_NAME.toUpperCase(Locale.ENGLISH), "wrong");
+        // Should be locked now
+        p = realm.authenticate(USER_NAME, PASSWORD);
+        Assert.assertNull(p);
+    }
+
+
+    @Test
+    public void testLockoutAfterFailureCaseSensitiveTrue() {
+        realm.setCaseSensitive(true);
+
+        Principal p = realm.authenticate(USER_NAME, PASSWORD);
+        Assert.assertNotNull(p);
+
+        p = realm.authenticate(USER_NAME, "wrong");
+        p = realm.authenticate(USER_NAME.toUpperCase(Locale.ENGLISH), "wrong");
+        // Should not be locked yet
+        p = realm.authenticate(USER_NAME, PASSWORD);
+        Assert.assertNotNull(p);
+
+        p = realm.authenticate(USER_NAME, "wrong");
+        p = realm.authenticate(USER_NAME, "wrong");
+
+        // Both should be locked now
+        p = realm.authenticate(USER_NAME, PASSWORD);
+        Assert.assertNull(p);
+    }
+}
--- a/webapps/docs/config/realm.xml
+++ b/webapps/docs/config/realm.xml
@@ -1009,6 +1009,12 @@
        1000.</p>
       </attribute>

+      <attribute name="caseSensitive" required="false">
+       <p>The manner in which user names will be treated when making locking
+       decisions. Defaults to <code>false</code>, meaning user names will be
+       treated in a case insensitive manner.</p>
+      </attribute>
+
       <attribute name="failureCount" required="false">
        <p>The number of times in a row a user has to fail authentication to be
        locked out. Defaults to 5.</p>

--- a/java/org/apache/catalina/realm/RealmBase.java
+++ b/java/org/apache/catalina/realm/RealmBase.java
@@ -666,8 +666,6 @@
                         constraints[i].included(uri, method));
             }

-            boolean matched = false;
-            int pos = -1;
             for (int j = 0; j < collection.length; j++) {
                 String[] patterns = collection[j].findPatterns();

@@ -677,6 +675,7 @@
                     continue;
                 }

+                boolean matched = false;
                 for (int k = 0; k < patterns.length && !matched; k++) {
                     String pattern = patterns[k];
                     if (pattern.startsWith("*.")) {
@@ -686,19 +685,18 @@
                                 uri.length() - dot == pattern.length() - 1) {
                             if (pattern.regionMatches(1, uri, dot, uri.length() - dot)) {
                                 matched = true;
-                                pos = j;
                             }
                         }
                     }
                 }
-            }
-            if (matched) {
-                found = true;
-                if (collection[pos].findMethod(method)) {
-                    if (results == null) {
-                        results = new ArrayList<>();
+                if (matched) {
+                    found = true;
+                    if (collection[j].findMethod(method)) {
+                        if (results == null) {
+                            results = new ArrayList<>();
+                        }
+                        results.add(constraints[i]);
                     }
-                    results.add(constraints[i]);
                 }
             }
         }
--- a/test/org/apache/catalina/realm/TestRealmBase.java
+++ b/test/org/apache/catalina/realm/TestRealmBase.java
@@ -790,4 +790,86 @@
         Assert.assertFalse(mapRealm.hasResourcePermission(
                 request, response, constraintsDelete, null));
     }
+
+
+    @Test
+    public void testUncoveredMethods() throws IOException {
+        // Create a constraint for ROLE1
+        SecurityConstraint constraint = new SecurityConstraint();
+        constraint.addAuthRole(ROLE1);
+        // Add a collection for GET
+        SecurityCollection getCollection = new SecurityCollection();
+        getCollection.addMethod(Method.GET);
+        getCollection.addPatternDecoded("*.html");
+        constraint.addCollection(getCollection);
+        // Add a collection for POST
+        SecurityCollection postCollection = new SecurityCollection();
+        postCollection.addMethod(Method.POST);
+        postCollection.addPatternDecoded("*.html");
+        constraint.addCollection(postCollection);
+
+        TesterMapRealm mapRealm = new TesterMapRealm();
+
+        // Set up the mock request and response
+        TesterRequest request = new TesterRequest();
+        Response response = new TesterResponse();
+        Context context = request.getContext();
+        context.addSecurityRole(ROLE1);
+        context.addSecurityRole(ROLE2);
+        request.getMappingData().context = context;
+
+        // Create the principals
+        List<String> userRoles1 = new ArrayList<>();
+        userRoles1.add(ROLE1);
+        GenericPrincipal gp1 = new GenericPrincipal(USER1, userRoles1);
+
+        List<String> userRoles2 = new ArrayList<>();
+        userRoles2.add(ROLE2);
+        GenericPrincipal gp2 = new GenericPrincipal(USER2, userRoles2);
+
+        List<String> userRoles99 = new ArrayList<>();
+        GenericPrincipal gp99 = new GenericPrincipal(USER99, userRoles99);
+
+        // Add the constraint to the context
+        context.addConstraint(constraint);
+
+
+        // Only user1 should be able to perform a GET
+        request.setMethod(Method.GET);
+
+        SecurityConstraint[] constraintsGet =
+                mapRealm.findSecurityConstraints(request, context);
+
+        request.setUserPrincipal(null);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsGet, null));
+        request.setUserPrincipal(gp1);
+        Assert.assertTrue(mapRealm.hasResourcePermission(
+                request, response, constraintsGet, null));
+        request.setUserPrincipal(gp2);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsGet, null));
+        request.setUserPrincipal(gp99);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsGet, null));
+
+        // Only user1 should be able to perform a POST
+        request.setMethod(Method.POST);
+
+        SecurityConstraint[] constraintsPost =
+                mapRealm.findSecurityConstraints(request, context);
+
+        request.setUserPrincipal(null);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsPost, null));
+        request.setUserPrincipal(gp1);
+        Assert.assertTrue(mapRealm.hasResourcePermission(
+                request, response, constraintsPost, null));
+        request.setUserPrincipal(gp2);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsPost, null));
+        request.setUserPrincipal(gp99);
+        Assert.assertFalse(mapRealm.hasResourcePermission(
+                request, response, constraintsPost, null));
+    }
 }
--- a/test/org/apache/tomcat/unittest/TesterRequest.java
+++ b/test/org/apache/tomcat/unittest/TesterRequest.java
@@ -31,6 +31,7 @@
 import org.apache.catalina.Context;
 import org.apache.catalina.connector.Request;
 import org.apache.catalina.session.StandardSession;
+import org.apache.tomcat.util.buf.MessageBytes;

 public class TesterRequest extends Request {

@@ -81,6 +82,15 @@
         return "/level1/level2/foo.html";
     }

+
+    @Override
+    public MessageBytes getRequestPathMB() {
+        MessageBytes result = MessageBytes.newInstance();
+        result.setString(getRequestURI());
+        return result;
+    }
+
+
     @Override
     public String getDecodedRequestURI() {
         // Decoding not required

--- a/java/org/apache/tomcat/util/http/parser/HttpParser.java
+++ b/java/org/apache/tomcat/util/http/parser/HttpParser.java
@@ -33,7 +33,7 @@
 
     private static final StringManager sm = StringManager.getManager(HttpParser.class);
 
-    private static final int ARRAY_SIZE = 128;
+    private static final int ARRAY_SIZE = 256;
 
     private static final boolean[] IS_CONTROL = new boolean[ARRAY_SIZE];
     private static final boolean[] IS_SEPARATOR = new boolean[ARRAY_SIZE];
@@ -47,6 +47,8 @@
     private static final boolean[] IS_SUBDELIM = new boolean[ARRAY_SIZE];
     private static final boolean[] IS_USERINFO = new boolean[ARRAY_SIZE];
     private static final boolean[] IS_RELAXABLE = new boolean[ARRAY_SIZE];
+    private static final boolean[] IS_FIELD_VCHAR = new boolean[ARRAY_SIZE];
+    private static final boolean[] IS_FIELD_CONTENT = new boolean[ARRAY_SIZE];
 
     private static final HttpParser DEFAULT;
 
@@ -66,7 +68,7 @@
             }
 
             // Token: Anything 0-127 that is not a control and not a separator
-            if (!IS_CONTROL[i] && !IS_SEPARATOR[i]) {
+            if (!IS_CONTROL[i] && !IS_SEPARATOR[i] && i < 128) {
                 IS_TOKEN[i] = true;
             }
 
@@ -75,7 +77,6 @@
                 IS_HEX[i] = true;
             }
 
-            // Not valid for HTTP protocol
             // "HTTP/" DIGIT "." DIGIT
             if (i == 'H' || i == 'T' || i == 'P' || i == '/' || i == '.' || (i >= '0' && i <= '9')) {
                 IS_HTTP_PROTOCOL[i] = true;
@@ -114,6 +115,16 @@
                     i == '{' || i == '|' || i == '}') {
                 IS_RELAXABLE[i] = true;
             }
+
+            // field-vchar is VCHAR / obs-text
+            if (i > 20 && i < 127 || i > 127) {
+                IS_FIELD_VCHAR[i] = true;
+            }
+
+            // field-content  = field-vchar [ 1*( SP / HTAB / field-vchar ) field-vchar ]
+            if (IS_FIELD_VCHAR[i] || i == '\t' || i == ' ') {
+                IS_FIELD_CONTENT[i] = true;
+            }
         }
 
         DEFAULT = new HttpParser(null, null);
@@ -393,10 +404,29 @@
 
 
     public static boolean isControl(int c) {
-        // Fast for valid control characters, slower for some incorrect
-        // ones
+        // Fast for valid control characters, slower for some incorrect ones
         try {
             return IS_CONTROL[c];
+        } catch (ArrayIndexOutOfBoundsException ex) {
+            return false;
+        }
+    }
+
+
+    public static boolean isFieldVChar(int c) {
+        // Fast for valid field-vchar characters, slower for some incorrect ones
+        try {
+            return IS_FIELD_VCHAR[c];
+        } catch (ArrayIndexOutOfBoundsException ex) {
+            return false;
+        }
+    }
+
+
+    public static boolean isFieldContent(int c) {
+        // Fast for valid field-content characters, slower for some incorrect ones
+        try {
+            return IS_FIELD_CONTENT[c];
         } catch (ArrayIndexOutOfBoundsException ex) {
             return false;
         }

--- a/java/org/apache/coyote/http2/HpackDecoder.java
+++ b/java/org/apache/coyote/http2/HpackDecoder.java
@@ -20,6 +20,7 @@
 
 import org.apache.juli.logging.Log;
 import org.apache.juli.logging.LogFactory;
+import org.apache.tomcat.util.http.parser.HttpParser;
 import org.apache.tomcat.util.res.StringManager;
 
 /**
@@ -114,7 +115,7 @@
                     buffer.position(originalPos);
                     return;
                 }
-                String headerValue = readHpackString(buffer);
+                String headerValue = readHpackString(buffer, false);
                 if (headerValue == null) {
                     buffer.position(originalPos);
                     return;
@@ -128,7 +129,7 @@
                     buffer.position(originalPos);
                     return;
                 }
-                String headerValue = readHpackString(buffer);
+                String headerValue = readHpackString(buffer, false);
                 if (headerValue == null) {
                     buffer.position(originalPos);
                     return;
@@ -141,7 +142,7 @@
                     buffer.position(originalPos);
                     return;
                 }
-                String headerValue = readHpackString(buffer);
+                String headerValue = readHpackString(buffer, false);
                 if (headerValue == null) {
                     buffer.position(originalPos);
                     return;
@@ -202,11 +203,11 @@
         } else if (index != 0) {
             return handleIndexedHeaderName(index);
         } else {
-            return readHpackString(buffer);
+            return readHpackString(buffer, true);
         }
     }
 
-    private String readHpackString(ByteBuffer buffer) throws HpackException {
+    private String readHpackString(ByteBuffer buffer, boolean isFieldName) throws HpackException {
         if (!buffer.hasRemaining()) {
             return null;
         }
@@ -218,18 +219,34 @@
         }
         boolean huffman = (data & 0b10000000) != 0;
         if (huffman) {
-            return readHuffmanString(length, buffer);
+            return readHuffmanString(length, buffer, isFieldName);
         }
         StringBuilder stringBuilder = new StringBuilder(length);
         for (int i = 0; i < length; ++i) {
-            stringBuilder.append((char) buffer.get());
+            char c = (char) (buffer.get() & 0xFF);
+            if (isFieldName) {
+                if (HttpParser.isToken(c) && !Character.isUpperCase(c)) {
+                    stringBuilder.append(c);
+                } else {
+                    throw new IllegalArgumentException(
+                            sm.getString("hpackdecoder.illegalCharacterName", Character.toString(c)));
+                }
+            } else {
+                if ((i == 0 || i == length - 1) && HttpParser.isFieldVChar(c) ||
+                        i > 0 && i < length - 1 && HttpParser.isFieldContent(c)) {
+                    stringBuilder.append(c);
+                } else {
+                    throw new IllegalArgumentException(
+                            sm.getString("hpackdecoder.illegalCharacterValue", Character.toString(c)));
+                }
+            }
         }
         return stringBuilder.toString();
     }
 
-    private String readHuffmanString(int length, ByteBuffer buffer) throws HpackException {
+    private String readHuffmanString(int length, ByteBuffer buffer, boolean isFieldName) throws HpackException {
         StringBuilder stringBuilder = new StringBuilder(length);
-        HPackHuffman.decode(buffer, length, stringBuilder);
+        HPackHuffman.decode(buffer, length, stringBuilder, isFieldName);
         return stringBuilder.toString();
     }
 
@@ -238,11 +255,7 @@
             return Hpack.STATIC_TABLE[index].name;
         } else {
             // index is 1 based
-            if (index > Hpack.STATIC_TABLE_LENGTH + filledTableSlots) {
-                throw new HpackException(sm.getString("hpackdecoder.headerTableIndexInvalid", Integer.valueOf(index),
-                        Integer.valueOf(Hpack.STATIC_TABLE_LENGTH), Integer.valueOf(filledTableSlots)));
-            }
-            int adjustedIndex = getRealIndex(index - Hpack.STATIC_TABLE_LENGTH);
+            int adjustedIndex = getRealIndex(index);
             Hpack.HeaderField res = headerTable[adjustedIndex];
             if (res == null) {
                 throw new HpackException(sm.getString("hpackdecoder.nullHeader", Integer.valueOf(index)));
@@ -263,11 +276,7 @@
             addStaticTableEntry(index);
         } else {
             // index is 1 based
-            if (index > Hpack.STATIC_TABLE_LENGTH + filledTableSlots) {
-                throw new HpackException(sm.getString("hpackdecoder.headerTableIndexInvalid", Integer.valueOf(index),
-                        Integer.valueOf(Hpack.STATIC_TABLE_LENGTH), Integer.valueOf(filledTableSlots)));
-            }
-            int adjustedIndex = getRealIndex(index - Hpack.STATIC_TABLE_LENGTH);
+            int adjustedIndex = getRealIndex(index);
             if (log.isTraceEnabled()) {
                 log.trace(sm.getString("hpackdecoder.useDynamic", Integer.valueOf(adjustedIndex)));
             }
@@ -287,15 +296,15 @@
      * @return the real index into the array
      */
     int getRealIndex(int index) throws HpackException {
-        // the index is one based, but our table is zero based, hence -1
-        // also because of our ring buffer set up the indexes are reversed
+        int dynamicIndex = index - Hpack.STATIC_TABLE_LENGTH;
+        // The index is one based, but our table is zero based
+        // Also, because of our ring buffer set up, the indexes are reversed
         // index = 1 is at position firstSlotPosition + filledSlots
-        int realIndex = (firstSlotPosition + (filledTableSlots - index)) % headerTable.length;
-        if (realIndex < 0) {
+        if (dynamicIndex < 1 || dynamicIndex > filledTableSlots) {
             throw new HpackException(sm.getString("hpackdecoder.headerTableIndexInvalid", Integer.valueOf(index),
                     Integer.valueOf(Hpack.STATIC_TABLE_LENGTH), Integer.valueOf(filledTableSlots)));
         }
-        return realIndex;
+        return (firstSlotPosition + (filledTableSlots - dynamicIndex)) % headerTable.length;
     }
 
     private void addStaticTableEntry(int index) throws HpackException {

--- a/java/org/apache/coyote/http2/HPackHuffman.java
+++ b/java/org/apache/coyote/http2/HPackHuffman.java
@@ -19,8 +19,10 @@
 import java.nio.ByteBuffer;
 import java.util.Arrays;
 import java.util.HashSet;
+import java.util.Locale;
 import java.util.Set;
 
+import org.apache.tomcat.util.http.parser.HttpParser;
 import org.apache.tomcat.util.res.StringManager;
 
 public class HPackHuffman {
@@ -371,12 +373,35 @@
      * @param target The target for the decompressed data
      *
      * @throws HpackException If the Huffman encoded value in HPACK headers did not end with EOS padding
+     *
+     * @deprecated Will be removed in Tomcat 12. Use {@link #decode(ByteBuffer, int, StringBuilder, boolean)}
      */
+    @Deprecated
     public static void decode(ByteBuffer data, int length, StringBuilder target) throws HpackException {
+        decode(data, length, target, false);
+    }
+
+    /**
+     * Decodes a huffman encoded string into the target StringBuilder. There must be enough space left in the buffer for
+     * this method to succeed.
+     *
+     * @param data        The byte buffer
+     * @param length      The length of data from the buffer to decode
+     * @param target      The target for the decompressed data
+     * @param isFieldName {@code true} if a field name is being decoded (names have a more restrictive set of allowed
+     *                        characters than field values)
+     *
+     * @throws HpackException If the Huffman encoded value in HPACK headers did not end with EOS padding
+     */
+    public static void decode(ByteBuffer data, int length, StringBuilder target, boolean isFieldName)
+            throws HpackException {
+
         assert data.remaining() >= length;
         int treePos = 0;
         boolean eosBits = true;
         int eosBitCount = 0;
+        boolean firstChar = true;
+        char c = 'a';
         for (int i = 0; i < length; ++i) {
             byte b = data.get();
             int bitPos = 7;
@@ -387,11 +412,34 @@
                     if ((val & LOW_TERMINAL_BIT) == 0) {
                         treePos = val & LOW_MASK;
                         eosBits = false;
+                        // Found a zero, can't be counting EOS bits
                         eosBitCount = 0;
                     } else {
-                        target.append((char) (val & LOW_MASK));
+                        c = (char) (val & LOW_MASK);
+                        if (isFieldName) {
+                            if (!HttpParser.isToken(c) || Character.isUpperCase(c)) {
+                                throw new IllegalArgumentException(
+                                        sm.getString("hpackhuffman.decode.illegalCharacterName"));
+                            }
+                        } else {
+                            if (firstChar) {
+                                if (!HttpParser.isFieldVChar(c)) {
+                                    throw new IllegalArgumentException(
+                                            sm.getString("hpackhuffman.decode.illegalCharacterValue.start"));
+                                }
+                                firstChar = false;
+                            } else {
+                                if (!HttpParser.isFieldContent(c)) {
+                                    throw new IllegalArgumentException(
+                                            sm.getString("hpackhuffman.decode.illegalCharacterValue"));
+                                }
+                            }
+                        }
+                        target.append(c);
                         treePos = 0;
                         eosBits = true;
+                        // Output a character, reset eosBitCount
+                        eosBitCount = 0;
                     }
                 } else {
                     if (eosBits) {
@@ -406,9 +454,31 @@
                             // as an error
                             throw new HpackException(sm.getString("hpackhuffman.stringLiteralEOS"));
                         }
-                        target.append((char) ((val >> 16) & LOW_MASK));
+                        c = (char) ((val >> 16) & LOW_MASK);
+                        if (isFieldName) {
+                            if (!HttpParser.isToken(c) || Character.isUpperCase(c)) {
+                                throw new IllegalArgumentException(
+                                        sm.getString("hpackhuffman.decode.illegalCharacterName"));
+                            }
+                        } else {
+                            if (firstChar) {
+                                if (!HttpParser.isFieldVChar(c)) {
+                                    throw new IllegalArgumentException(
+                                            sm.getString("hpackhuffman.decode.illegalCharacterValue.start"));
+                                }
+                                firstChar = false;
+                            } else {
+                                if (!HttpParser.isFieldContent(c)) {
+                                    throw new IllegalArgumentException(
+                                            sm.getString("hpackhuffman.decode.illegalCharacterValue"));
+                                }
+                            }
+                        }
+                        target.append(c);
                         treePos = 0;
                         eosBits = true;
+                        // Output a character, reset eosBitCount
+                        eosBitCount = 0;
                     }
                 }
                 bitPos--;
@@ -420,6 +490,9 @@
         if (!eosBits) {
             throw new HpackException(sm.getString("hpackhuffman.huffmanEncodedHpackValueDidNotEndWithEOS"));
         }
+        if (!HttpParser.isFieldVChar(c)) {
+            throw new IllegalArgumentException(sm.getString("hpackhuffman.decode.illegalCharacterValue.end"));
+        }
     }
 
 
@@ -432,8 +505,29 @@
      * @param forceLowercase If the string should be encoded in lower case
      *
      * @return true if encoding succeeded
+     *
+     * @deprecated Unused. This method will be removed in Tomcat 12 onwards.
      */
+    @Deprecated
     public static boolean encode(ByteBuffer buffer, String toEncode, boolean forceLowercase) {
+        if (forceLowercase) {
+            return encode(buffer, toEncode.toLowerCase(Locale.ENGLISH));
+        } else {
+            return encode(buffer, toEncode);
+        }
+    }
+
+
+    /**
+     * Encodes the given string into the buffer. If there is not enough space in the buffer, or the encoded version is
+     * bigger than the original it will return false and not modify the buffers position.
+     *
+     * @param buffer   The buffer to encode into
+     * @param toEncode The string to encode
+     *
+     * @return true if encoding succeeded
+     */
+    public static boolean encode(ByteBuffer buffer, String toEncode) {
         if (buffer.remaining() <= toEncode.length()) {
             return false;
         }
@@ -448,9 +542,6 @@
                 throw new IllegalArgumentException(
                         sm.getString("hpack.invalidCharacter", Character.toString(c), Integer.valueOf(c)));
             }
-            if (forceLowercase) {
-                c = Hpack.toLower(c);
-            }
             HuffmanCode code = HUFFMAN_CODES[c];
             length += code.length;
         }
@@ -464,9 +555,6 @@
         byte currentBufferByte = 0;
         for (int i = 0; i < toEncode.length(); ++i) {
             char c = toEncode.charAt(i);
-            if (forceLowercase) {
-                c = Hpack.toLower(c);
-            }
             HuffmanCode code = HUFFMAN_CODES[c];
             if (code.length + bytePos <= 8) {
                 // it fits in the current byte

--- a/java/org/apache/coyote/http2/HpackEncoder.java
+++ b/java/org/apache/coyote/http2/HpackEncoder.java
@@ -124,6 +124,22 @@
      * @return The state of the encoding process
      */
     State encode(MimeHeaders headers, ByteBuffer target) {
+        return encode(headers, target, true);
+    }
+
+    /**
+     * Encodes the headers into a buffer.
+     *
+     * @param headers        The headers to encode
+     * @param target         The buffer to which to write the encoded headers
+     * @param forceLowerCase Normally {@code true} to ensure that header field names are lower case as required for
+     *                           HTTP/2 but some tests may deliberately allow upper case characters to test Tomcat's
+     *                           handling of such invalid field header names.
+     *
+     * @return The state of the encoding process
+     */
+    State encode(MimeHeaders headers, ByteBuffer target, boolean forceLowerCase) {
+
         int it = headersIterator;
         if (headersIterator == -1) {
             handleTableSizeChange(target);
@@ -136,8 +152,10 @@
             }
         }
         while (it < currentHeaders.size()) {
-            // FIXME: Review lowercase policy
-            String headerName = headers.getName(it).toString().toLowerCase(Locale.US);
+            String headerName = headers.getName(it).toString();
+            if (forceLowerCase) {
+                headerName = headerName.toLowerCase(Locale.US);
+            }
             boolean skip = false;
             if (firstPass) {
                 if (headerName.charAt(0) != ':') {
@@ -211,23 +229,29 @@
         return State.COMPLETE;
     }
 
+    /*
+     * headerName must be lower case by the time this method is called.
+     *
+     * The exception to the above rule is test cases which may deliberately use some upper case characters to test how
+     * Tomcat responds to such invalid input.
+     */
     private void writeHuffmanEncodableName(ByteBuffer target, String headerName) {
         if (hpackHeaderFunction.shouldUseHuffman(headerName)) {
-            if (HPackHuffman.encode(target, headerName, true)) {
+            if (HPackHuffman.encode(target, headerName)) {
                 return;
             }
         }
         target.put((byte) 0); // to use encodeInteger we need to place the first byte in the buffer.
         Hpack.encodeInteger(target, headerName.length(), 7);
         for (int j = 0; j < headerName.length(); ++j) {
-            target.put((byte) Hpack.toLower(headerName.charAt(j)));
+            target.put((byte) headerName.charAt(j));
         }
 
     }
 
     private void writeHuffmanEncodableValue(ByteBuffer target, String headerName, String val) {
         if (hpackHeaderFunction.shouldUseHuffman(headerName, val)) {
-            if (!HPackHuffman.encode(target, val, false)) {
+            if (!HPackHuffman.encode(target, val)) {
                 writeValueString(target, val);
             }
         } else {

--- a/java/org/apache/coyote/http2/Hpack.java
+++ b/java/org/apache/coyote/http2/Hpack.java
@@ -217,7 +217,10 @@
         }
     }
 
-
+    /*
+     * Unused. Will be removed in Tomcat 12 onwards.
+     */
+    @Deprecated
     static char toLower(char c) {
         if (c >= 'A' && c <= 'Z') {
             return (char) (c + LOWER_DIFF);
@@ -225,7 +228,7 @@
         return c;
     }
 
+
     private Hpack() {
     }
-
 }

--- a/java/org/apache/coyote/http2/Stream.java
+++ b/java/org/apache/coyote/http2/Stream.java
@@ -25,7 +25,6 @@
 import java.security.PrivilegedExceptionAction;
 import java.util.Collections;
 import java.util.HashSet;
-import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -49,6 +48,7 @@
 import org.apache.juli.logging.LogFactory;
 import org.apache.tomcat.util.buf.ByteChunk;
 import org.apache.tomcat.util.buf.MessageBytes;
+import org.apache.tomcat.util.http.HeaderUtil;
 import org.apache.tomcat.util.http.Method;
 import org.apache.tomcat.util.http.MimeHeaders;
 import org.apache.tomcat.util.http.parser.Host;
@@ -327,14 +327,7 @@
             log.trace(sm.getString("stream.header.debug", getConnectionId(), getIdAsString(), name, value));
         }
 
-        // Header names must be lowercase
-        if (!name.toLowerCase(Locale.US).equals(name)) {
-            headerException =
-                    new StreamException(sm.getString("stream.headercase", getConnectionId(), getIdAsString(), name),
-                            Http2Error.PROTOCOL_ERROR, getIdAsInt());
-            // No need for further processing. The stream will be reset.
-            return;
-        }
+        // Field header names being all lower case is enforced in HpackDecoder.
 
         if (HTTP_CONNECTION_SPECIFIC_HEADERS.contains(name)) {
             headerException = new StreamException(
@@ -347,7 +340,7 @@
         if ("te".equals(name)) {
             if (!"trailers".equals(value)) {
                 headerException =
-                        new StreamException(sm.getString("stream.header.te", getConnectionId(), getIdAsString(), name),
+                        new StreamException(sm.getString("stream.header.te", getConnectionId(), getIdAsString(), value),
                                 Http2Error.PROTOCOL_ERROR, getIdAsInt());
                 // No need for further processing. The stream will be reset.
                 return;
@@ -362,7 +355,7 @@
 
         if (name.isEmpty()) {
             headerException =
-                    new StreamException(sm.getString("stream.header.empty", getConnectionId(), getIdAsString(), name),
+                    new StreamException(sm.getString("stream.header.empty", getConnectionId(), getIdAsString()),
                             Http2Error.PROTOCOL_ERROR, getIdAsInt());
             // No need for further processing. The stream will be reset.
             return;
@@ -393,20 +386,23 @@
                     headerException = new StreamException(
                             sm.getString("stream.header.duplicate", getConnectionId(), getIdAsString(), ":method"),
                             Http2Error.PROTOCOL_ERROR, getIdAsInt());
-                    // No need for further processing. The stream will be reset.
-                    return;
                 }
                 break;
             }
             case ":scheme": {
                 if (coyoteRequest.scheme().isNull()) {
                     coyoteRequest.scheme().setString(value);
+                    // Check scheme is consistent with TLS usage
+                    if ("https".equals(value) != handler.getProtocol().getHttp11Protocol().isSSLEnabled()) {
+                        headerException = new StreamException(
+                                sm.getString("stream.header.inconsistentScheme", getConnectionId(), getIdAsString(),
+                                value, Boolean.toString(handler.getProtocol().getHttp11Protocol().isSSLEnabled())),
+                                Http2Error.PROTOCOL_ERROR, getIdAsInt());
+                    }
                 } else {
                     headerException = new StreamException(
                             sm.getString("stream.header.duplicate", getConnectionId(), getIdAsString(), ":scheme"),
                             Http2Error.PROTOCOL_ERROR, getIdAsInt());
-                    // No need for further processing. The stream will be reset.
-                    return;
                 }
                 break;
             }
@@ -415,31 +411,27 @@
                     headerException = new StreamException(
                             sm.getString("stream.header.duplicate", getConnectionId(), getIdAsString(), ":path"),
                             Http2Error.PROTOCOL_ERROR, getIdAsInt());
-                    // No need for further processing. The stream will be reset.
-                    return;
-                }
-                if (value.isEmpty()) {
+                } else if (value.isEmpty()) {
                     headerException = new StreamException(
                             sm.getString("stream.header.noPath", getConnectionId(), getIdAsString()),
                             Http2Error.PROTOCOL_ERROR, getIdAsInt());
-                    // No need for further processing. The stream will be reset.
-                    return;
-                }
-                int queryStart = value.indexOf('?');
-                String uri;
-                if (queryStart == -1) {
-                    uri = value;
                 } else {
-                    uri = value.substring(0, queryStart);
-                    String query = value.substring(queryStart + 1);
-                    coyoteRequest.queryString().setString(query);
+                    int queryStart = value.indexOf('?');
+                    String uri;
+                    if (queryStart == -1) {
+                        uri = value;
+                    } else {
+                        uri = value.substring(0, queryStart);
+                        String query = value.substring(queryStart + 1);
+                        coyoteRequest.queryString().setString(query);
+                    }
+                    // Bug 61120. Set the URI as bytes rather than String so:
+                    // - any path parameters are correctly processed
+                    // - the normalization security checks are performed that prevent
+                    // directory traversal attacks
+                    byte[] uriBytes = uri.getBytes(StandardCharsets.ISO_8859_1);
+                    coyoteRequest.requestURI().setBytes(uriBytes, 0, uriBytes.length);
                 }
-                // Bug 61120. Set the URI as bytes rather than String so:
-                // - any path parameters are correctly processed
-                // - the normalization security checks are performed that prevent
-                // directory traversal attacks
-                byte[] uriBytes = uri.getBytes(StandardCharsets.ISO_8859_1);
-                coyoteRequest.requestURI().setBytes(uriBytes, 0, uriBytes.length);
                 break;
             }
             case ":authority": {
@@ -449,8 +441,6 @@
                     headerException = new StreamException(
                             sm.getString("stream.header.duplicate", getConnectionId(), getIdAsString(), ":authority"),
                             Http2Error.PROTOCOL_ERROR, getIdAsInt());
-                    // No need for further processing. The stream will be reset.
-                    return;
                 }
                 break;
             }
@@ -479,8 +469,6 @@
                     headerException = new StreamException(
                             sm.getString("stream.header.duplicate", getConnectionId(), getIdAsString(), "host"),
                             Http2Error.PROTOCOL_ERROR, getIdAsInt());
-                    // No need for further processing. The stream will be reset.
-                    return;
                 }
                 break;
             }
@@ -511,9 +499,7 @@
                     headerException = new StreamException(
                             sm.getString("stream.header.unknownPseudoHeader", getConnectionId(), getIdAsString(), name),
                             Http2Error.PROTOCOL_ERROR, getIdAsInt());
-                }
-
-                if (headerState == HEADER_STATE_TRAILER) {
+                } else if (headerState == HEADER_STATE_TRAILER) {
                     // HTTP/2 headers are already always lower case
                     coyoteRequest.getMimeTrailerFields().addValue(name).setString(value);
                 } else {
@@ -547,11 +533,8 @@
         // Match host name with SNI if required
         if (!handler.getProtocol().getHttp11Protocol().checkSni(handler.getSniHostName(),
                 coyoteRequest.serverName().getString())) {
-            headerException = new StreamException(
-                    sm.getString("stream.host.sni", getConnectionId(), getIdAsString(), value, handler.getSniHostName()),
-                    Http2Error.PROTOCOL_ERROR, getIdAsInt());
-            // No need for further processing. The stream will be reset.
-            return;
+            headerException = new StreamException(sm.getString("stream.host.sni", getConnectionId(), getIdAsString(),
+                    value, handler.getSniHostName()), Http2Error.PROTOCOL_ERROR, getIdAsInt());
         }
     }
 
@@ -567,10 +550,9 @@
                 // Host value inconsistent
                 headerException = new StreamException(
                         sm.getString("stream.host.inconsistent", getConnectionId(), getIdAsString(), value,
-                                coyoteRequest.serverName().getString(), Integer.toString(coyoteRequest.getServerPort())),
+                                coyoteRequest.serverName().getString(),
+                                Integer.toString(coyoteRequest.getServerPort())),
                         Http2Error.PROTOCOL_ERROR, getIdAsInt());
-                // No need for further processing. The stream will be reset.
-                return;
             }
         } catch (IllegalArgumentException iae) {
             // Bad :authority / host header -> 400 response
@@ -598,12 +580,32 @@
     }
 
 
-    final boolean receivedEndOfHeaders() throws ConnectionException {
-        if (coyoteRequest.getMethod() == null || coyoteRequest.scheme().isNull() ||
-                !Method.CONNECT.equals(coyoteRequest.getMethod()) && coyoteRequest.requestURI().isNull()) {
-            throw new ConnectionException(sm.getString("stream.header.required", getConnectionId(), getIdAsString()),
-                    Http2Error.PROTOCOL_ERROR);
+    final boolean receivedEndOfHeaders() throws StreamException {
+        boolean missingHeader = false;
+
+        if (coyoteRequest.getMethod() == null) {
+            missingHeader = true;
+        } else if (Method.CONNECT.equals(coyoteRequest.getMethod())) {
+            // CONNECT only
+            if (!coyoteRequest.scheme().isNull() || !coyoteRequest.requestURI().isNull()) {
+                throw new StreamException(sm.getString("stream.header.invalidConnect", getConnectionId(),
+                        getIdAsString()), Http2Error.PROTOCOL_ERROR, getIdAsInt());
+            }
+            if (coyoteRequest.serverName().isNull()) {
+                missingHeader = true;
+            }
+        } else {
+            // All other methods
+            if (coyoteRequest.scheme().isNull() || coyoteRequest.requestURI().isNull()) {
+                missingHeader = true;
+            }
         }
+
+        if (missingHeader) {
+            throw new StreamException(sm.getString("stream.header.required", getConnectionId(), getIdAsString()),
+                    Http2Error.PROTOCOL_ERROR, getIdAsInt());
+        }
+
         // Cookie headers need to be concatenated into a single header
         // See RFC 7540 8.1.2.5
         // Can only do this once the headers are fully received
@@ -627,8 +629,8 @@
 
 
     final void writeTrailers() throws IOException {
-        Supplier<Map<String,String>> supplier = coyoteResponse.getTrailerFields();
-        if (supplier == null) {
+        Supplier<Map<String,String>> trailerFieldsSupplier = coyoteResponse.getTrailerFields();
+        if (trailerFieldsSupplier == null) {
             // No supplier was set, end of stream will already have been sent
             return;
         }
@@ -639,17 +641,22 @@
          */
         MimeHeaders mimeHeaders = new MimeHeaders();
 
-        Map<String,String> headerMap = supplier.get();
-        if (headerMap == null) {
-            headerMap = Collections.emptyMap();
+        Map<String,String> trailerFields = trailerFieldsSupplier.get();
+        if (trailerFields == null) {
+            trailerFields = Collections.emptyMap();
         }
 
         // Copy the contents of the Map to the MimeHeaders
         // TODO: Is there benefit in refactoring this? Is MimeHeaders too
         // heavyweight? Can we reduce the copy/conversions?
-        for (Map.Entry<String,String> headerEntry : headerMap.entrySet()) {
-            MessageBytes mb = mimeHeaders.addValue(headerEntry.getKey());
-            mb.setString(headerEntry.getValue());
+        for (Map.Entry<String,String> trailerField : trailerFields.entrySet()) {
+            // Ignore disallowed headers
+            if (HeaderUtil.isHeaderDisallowedInTrailers(trailerField.getKey())) {
+                continue;
+            }
+
+            MessageBytes mb = mimeHeaders.addValue(trailerField.getKey());
+            mb.setString(trailerField.getValue());
         }
 
         handler.writeHeaders(this, 0, mimeHeaders, true, Constants.DEFAULT_HEADERS_FRAME_SIZE);

--- a/java/org/apache/coyote/http2/Http2UpgradeHandler.java
+++ b/java/org/apache/coyote/http2/Http2UpgradeHandler.java
@@ -1189,7 +1189,8 @@
                             break;
                         }
 
-                        int share = (int) (s.getConnectionAllocationRequested() * remaining /
+                        // Cast to long during the intermediate calculation to avoid integer overflow in multiplication
+                        int share = (int) ((long) s.getConnectionAllocationRequested() * remaining /
                                 requestedAllocationForIncrementalStreams);
                         if (share == 0) {
                             share = 1;

--- a/java/org/apache/coyote/http2/LocalStrings.properties
+++ b/java/org/apache/coyote/http2/LocalStrings.properties
@@ -46,6 +46,8 @@
 hpackdecoder.clearDynamic=Emptying dynamic table
 hpackdecoder.emitHeader=Emitting header with name [{0}] and value [{1}]
 hpackdecoder.headerTableIndexInvalid=The header table index [{0}] is not valid as there are [{1}] static entries and [{2}] dynamic entries
+hpackdecoder.illegalCharacterName=The illegal [{0}] character was found when decoding an HTTP/2 header field name
+hpackdecoder.illegalCharacterValue=The illegal [{0}] character was found when decoding an HTTP/2 header field value
 hpackdecoder.maxMemorySizeExceeded=The header table size [{0}] exceeds the maximum size [{1}]
 hpackdecoder.notImplemented=Not yet implemented
 hpackdecoder.nullHeader=Null header at index [{0}]
@@ -54,6 +56,10 @@
 hpackdecoder.useStatic=Using header from index [{0}] of static table
 hpackdecoder.zeroNotValidHeaderTableIndex=Zero is not a valid header table index
 
+hpackhuffman.decode.illegalCharacterName=The illegal [{0}] character was found when decoding an HTTP/2 header field name
+hpackhuffman.decode.illegalCharacterValue=The illegal [{0}] character was found when decoding an HTTP/2 header field value
+hpackhuffman.decode.illegalCharacterValue.end=The illegal [{0}] character was found when decoding the final character in an HTTP/2 header field value
+hpackhuffman.decode.illegalCharacterValue.start=The illegal [{0}] character was found when decoding the first character in an HTTP/2 header field value
 hpackhuffman.huffmanEncodedHpackValueDidNotEndWithEOS=Huffman encoded value in HPACK headers did not end with EOS padding
 hpackhuffman.stringLiteralEOS=Huffman encoded value in HPACK headers contained the EOS symbol
 hpackhuffman.stringLiteralTooMuchPadding=More than 7 bits of EOS padding were provided at the end of an Huffman encoded string literal
@@ -92,19 +98,20 @@
 
 stream.clientResetRequest=Client reset the stream before the request was fully read
 stream.closed=Connection [{0}], Stream [{1}], Unable to write to stream once it has been closed
-stream.header.case=Connection [{0}], Stream [{1}], HTTP header name [{2}] must be in lower case
 stream.header.connection=Connection [{0}], Stream [{1}], HTTP header [{2}] is not permitted in an HTTP/2 request
 stream.header.contentLength=Connection [{0}], Stream [{1}], The content length header value [{2}] does not agree with the size of the data received [{3}]
 stream.header.debug=Connection [{0}], Stream [{1}], HTTP header [{2}], Value [{3}]
 stream.header.duplicate=Connection [{0}], Stream [{1}], received multiple [{2}] headers
 stream.header.empty=Connection [{0}], Stream [{1}], Invalid empty header name
+stream.header.inconsistentScheme=Connection [{0}], Stream [{1}], The scheme [{2}] is not consistent with the TLS enabled setting of [{3}]
 stream.header.invalid=Connection [{0}], Stream [{1}], The header [{2}] contained invalid value [{3}]
+stream.header.invalidConnect=Connection [{0}], Stream [{1}], The CONNECT request was invalid as neither :scheme nor :path should be present
 stream.header.noPath=Connection [{0}], Stream [{1}], The [:path] pseudo header was empty
 stream.header.required=Connection [{0}], Stream [{1}], One or more required headers was missing
 stream.header.te=Connection [{0}], Stream [{1}], HTTP header [te] is not permitted to have the value [{2}] in an HTTP/2 request
 stream.header.unexpectedPseudoHeader=Connection [{0}], Stream [{1}], Pseudo header [{2}] received after a regular header
 stream.header.unknownPseudoHeader=Connection [{0}], Stream [{1}], Unknown pseudo header [{2}] received
-stream.host.inconsistent=Connection [{0}], Stream [{1}], The header host header [{2}] is inconsistent with previously provided values for host [{3}] and/or port [{4}]
+stream.host.inconsistent=Connection [{0}], Stream [{1}], The host header [{2}] is inconsistent with previously provided values for host [{3}] and/or port [{4}]
 stream.host.sni=Connection [{0}], Stream [{1}], The host header [{2}] does not match the SNI host [{3}]
 stream.inputBuffer.copy=Copying [{0}] bytes from inBuffer to outBuffer
 stream.inputBuffer.dispatch=Data added to inBuffer when read interest is registered. Triggering a read dispatch

--- a/test/org/apache/coyote/http2/Http2TestBase.java
+++ b/test/org/apache/coyote/http2/Http2TestBase.java
@@ -226,7 +226,11 @@
             String url) {
         List<Header> headers = new ArrayList<>(4);
         headers.add(new Header(":method", Method.GET));
-        headers.add(new Header(":scheme", "http"));
+        if (getTomcatInstance().getConnector().getSecure()) {
+            headers.add(new Header(":scheme", "https"));
+        } else {
+            headers.add(new Header(":scheme", "http"));
+        }
         headers.add(new Header(":path", url));
         headers.add(new Header(":authority", "localhost:" + getPort()));
 
@@ -243,7 +247,8 @@
         for (Header header : headers) {
             mimeHeaders.addValue(header.getName()).setString(header.getValue());
         }
-        hpackEncoder.encode(mimeHeaders, headersPayload);
+        // Don't force lower case to allow testing with upper case field names
+        hpackEncoder.encode(mimeHeaders, headersPayload, false);
         if (padding != null) {
             headersPayload.put(padding);
         }
@@ -264,7 +269,11 @@
     protected void buildSimpleGetRequestPart1(byte[] frameHeader, ByteBuffer headersPayload, int streamId) {
         List<Header> headers = new ArrayList<>(3);
         headers.add(new Header(":method", Method.GET));
-        headers.add(new Header(":scheme", "http"));
+        if (getTomcatInstance().getConnector().getSecure()) {
+            headers.add(new Header(":scheme", "https"));
+        } else {
+            headers.add(new Header(":scheme", "http"));
+        }
         headers.add(new Header(":path", "/simple"));
 
         buildSimpleGetRequestPart1(frameHeader, headersPayload, headers, streamId);
@@ -377,7 +386,11 @@
 
         MimeHeaders headers = new MimeHeaders();
         headers.addValue(":method").setString(Method.POST);
-        headers.addValue(":scheme").setString("http");
+        if (getTomcatInstance().getConnector().getSecure()) {
+            headers.addValue(":scheme").setString("https");
+        } else {
+            headers.addValue(":scheme").setString("http");
+        }
         headers.addValue(":path").setString(path);
         headers.addValue(":authority").setString("localhost:" + getPort());
         if (useExpectation) {
@@ -453,7 +466,11 @@
     protected void buildHeadRequest(byte[] headersFrameHeader, ByteBuffer headersPayload, int streamId, String path) {
         MimeHeaders headers = new MimeHeaders();
         headers.addValue(":method").setString(Method.HEAD);
-        headers.addValue(":scheme").setString("http");
+        if (getTomcatInstance().getConnector().getSecure()) {
+            headers.addValue(":scheme").setString("https");
+        } else {
+            headers.addValue(":scheme").setString("http");
+        }
         headers.addValue(":path").setString(path);
         headers.addValue(":authority").setString("localhost:" + getPort());
         hpackEncoder.encode(headers, headersPayload);
@@ -1184,7 +1201,7 @@
         public void headersEnd(int streamId, boolean endOfStream) {
             trace.append(streamId + "-HeadersEnd\n");
             if (endOfStream) {
-                receivedEndOfStream(streamId) ;
+                receivedEndOfStream(streamId);
             }
         }
 
@@ -1456,7 +1473,7 @@
         @Override
         protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
 
-            Map<String, String[]> params = req.getParameterMap();
+            Map<String,String[]> params = req.getParameterMap();
 
             resp.setContentType("text/plain");
             resp.setCharacterEncoding("UTF-8");

--- a/test/org/apache/coyote/http2/TestHttp2Section_8_1.java
+++ b/test/org/apache/coyote/http2/TestHttp2Section_8_1.java
@@ -554,6 +554,6 @@
         headers.add(new Header(":path", "/simple"));
         headers.add(new Header("host", "localhost:" + getPort()));
 
-        doInvalidPseudoHeaderTest(headers, "0-Goaway-[3]-[1]-");
+        doInvalidPseudoHeaderTest(headers, "3-RST-[1]\n");
     }
 }

--- a/java/org/apache/tomcat/util/http/HeaderUtil.java
+++ b/java/org/apache/tomcat/util/http/HeaderUtil.java
@@ -16,7 +16,37 @@
  */
 package org.apache.tomcat.util.http;
 
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Set;
+
 public class HeaderUtil {
+
+    private static final Set<String> disallowedTrailerFieldNames = new HashSet<>();
+
+    static {
+        // Always add these in lower case
+        disallowedTrailerFieldNames.add("age");
+        disallowedTrailerFieldNames.add("cache-control");
+        disallowedTrailerFieldNames.add("content-length");
+        disallowedTrailerFieldNames.add("content-encoding");
+        disallowedTrailerFieldNames.add("content-range");
+        disallowedTrailerFieldNames.add("content-type");
+        disallowedTrailerFieldNames.add("date");
+        disallowedTrailerFieldNames.add("expires");
+        disallowedTrailerFieldNames.add("location");
+        disallowedTrailerFieldNames.add("retry-after");
+        disallowedTrailerFieldNames.add("trailer");
+        disallowedTrailerFieldNames.add("transfer-encoding");
+        disallowedTrailerFieldNames.add("vary");
+        disallowedTrailerFieldNames.add("warning");
+    }
+
+
+    public static boolean isHeaderDisallowedInTrailers(String headerName) {
+        return disallowedTrailerFieldNames.contains(headerName.toLowerCase(Locale.ENGLISH));
+    }
+
 
     /**
      * Converts an HTTP header line in byte form to a printable String. Bytes corresponding to visible ASCII characters

--- a/java/org/apache/coyote/http11/filters/ChunkedOutputFilter.java
+++ b/java/org/apache/coyote/http11/filters/ChunkedOutputFilter.java
@@ -21,16 +21,14 @@
 import java.io.OutputStreamWriter;
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
-import java.util.HashSet;
-import java.util.Locale;
 import java.util.Map;
-import java.util.Set;
 import java.util.function.Supplier;
 
 import org.apache.coyote.Response;
 import org.apache.coyote.http11.HttpOutputBuffer;
 import org.apache.coyote.http11.OutputFilter;
 import org.apache.tomcat.util.buf.HexUtils;
+import org.apache.tomcat.util.http.HeaderUtil;
 
 /**
  * Chunked output filter.
@@ -41,26 +39,6 @@
     private static final byte[] CRLF_BYTES = { (byte) '\r', (byte) '\n' };
     private static final byte[] END_CHUNK_BYTES = { (byte) '0', (byte) '\r', (byte) '\n', (byte) '\r', (byte) '\n' };
 
-    private static final Set<String> disallowedTrailerFieldNames = new HashSet<>();
-
-    static {
-        // Always add these in lower case
-        disallowedTrailerFieldNames.add("age");
-        disallowedTrailerFieldNames.add("cache-control");
-        disallowedTrailerFieldNames.add("content-length");
-        disallowedTrailerFieldNames.add("content-encoding");
-        disallowedTrailerFieldNames.add("content-range");
-        disallowedTrailerFieldNames.add("content-type");
-        disallowedTrailerFieldNames.add("date");
-        disallowedTrailerFieldNames.add("expires");
-        disallowedTrailerFieldNames.add("location");
-        disallowedTrailerFieldNames.add("retry-after");
-        disallowedTrailerFieldNames.add("trailer");
-        disallowedTrailerFieldNames.add("transfer-encoding");
-        disallowedTrailerFieldNames.add("vary");
-        disallowedTrailerFieldNames.add("warning");
-    }
-
     /**
      * Next buffer in the pipeline.
      */
@@ -178,13 +156,13 @@
             try (OutputStreamWriter osw = new OutputStreamWriter(baos, StandardCharsets.ISO_8859_1)) {
                 for (Map.Entry<String,String> trailerField : trailerFields.entrySet()) {
                     // Ignore disallowed headers
-                    if (disallowedTrailerFieldNames.contains(trailerField.getKey().toLowerCase(Locale.ENGLISH))) {
+                    if (HeaderUtil.isHeaderDisallowedInTrailers(trailerField.getKey())) {
                         continue;
                     }
-                    osw.write(trailerField.getKey());
+                    osw.write(filterForHeaders(trailerField.getKey()));
                     osw.write(':');
                     osw.write(' ');
-                    osw.write(trailerField.getValue());
+                    osw.write(filterForHeaders(trailerField.getValue()));
                     osw.write("\r\n");
                 }
             }
@@ -195,6 +173,33 @@
             crlfChunk.position(0).limit(crlfChunk.capacity());
         }
         buffer.end();
+    }
+
+
+    /*
+     * Filters out CTLs excluding TAB and any code points above 255 (since this is meant to be ISO-8859-1).
+     *
+     * This doesn't perform full HTTP validation. For example, it does not limit field names to tokens.
+     *
+     * Strictly, correct trailer fields is an application concern. The filtering here is a basic attempt to help
+     * mis-behaving applications prevent the worst of the potential side-effects of invalid trailer fields.
+     */
+    // package private so it is visible for testing
+    static String filterForHeaders(String input) {
+        char[] chars = input.toCharArray();
+        boolean updated = false;
+        for (int i = 0; i < chars.length; i++) {
+            if (chars[i] < 32 && chars [i] != 9 || chars[i] == 127 || chars[i] > 255) {
+                chars[i] = ' ';
+                updated = true;
+            }
+        }
+
+        if (updated) {
+            return new String(chars);
+        } else {
+            return input;
+        }
     }
 
 


diff --git a/java/org/apache/catalina/servlets/WebdavServlet.java b/java/org/apache/catalina/servlets/WebdavServlet.java
--- a/java/org/apache/catalina/servlets/WebdavServlet.java
+++ b/java/org/apache/catalina/servlets/WebdavServlet.java
@@ -175,6 +175,8 @@ import org.xml.sax.SAXException;
  *  &lt;/init-param&gt;
  * </pre>
  * <p>
+ * By default, WebDAV request bodies for LOCK and PROPFIND are limited to 4096 bytes. To change this limit, set the
+ * <code>maxRequestBodySize</code> <code>init-param</code> for the WebDAV servlet.
  *
  * @see <a href="https://tools.ietf.org/html/rfc4918">RFC 4918</a>
  */
@@ -202,6 +204,12 @@ public class WebdavServlet extends DefaultServlet implements PeriodicEventListen
     private static final int MAX_DEPTH = 3;
 
 
+    /*
+     * Default max request body size.
+     */
+    private static final int DEFAULT_MAX_REQUEST_BODY_SIZE = 4096;
+
+
     /**
      * Default namespace.
      */
@@ -215,6 +223,7 @@ public class WebdavServlet extends DefaultServlet implements PeriodicEventListen
             "\n  <D:lockentry><D:lockscope><D:exclusive/></D:lockscope><D:locktype><D:write/></D:locktype></D:lockentry>\n" +
                     "  <D:lockentry><D:lockscope><D:shared/></D:lockscope><D:locktype><D:write/></D:locktype></D:lockentry>\n";
 
+
     /**
      * Simple date format for the creation date ISO representation (partial).
      */
@@ -227,6 +236,7 @@ public class WebdavServlet extends DefaultServlet implements PeriodicEventListen
      */
     protected static final String LOCK_SCHEME = "urn:uuid:";
 
+
     // ----------------------------------------------------- Instance Variables
 
     /**
@@ -247,6 +257,9 @@ public class WebdavServlet extends DefaultServlet implements PeriodicEventListen
     private int maxDepth = MAX_DEPTH;
 
 
+    private int maxRequestBodySize = DEFAULT_MAX_REQUEST_BODY_SIZE;
+
+
     /**
      * Is access allowed via WebDAV to the special paths (/WEB-INF and /META-INF)?
      */
@@ -293,6 +306,10 @@ public class WebdavServlet extends DefaultServlet implements PeriodicEventListen
             maxDepth = Integer.parseInt(getServletConfig().getInitParameter("maxDepth"));
         }
 
+        if (getServletConfig().getInitParameter("maxRequestBodySize") != null) {
+            maxRequestBodySize = Integer.parseInt(getServletConfig().getInitParameter("maxRequestBodySize"));
+        }
+
         if (getServletConfig().getInitParameter("allowSpecialPaths") != null) {
             allowSpecialPaths = Boolean.parseBoolean(getServletConfig().getInitParameter("allowSpecialPaths"));
         }
@@ -829,13 +846,23 @@ public class WebdavServlet extends DefaultServlet implements PeriodicEventListen
             }
         }
 
+        // Short-cut if client provided a content length
+        if (req.getContentLengthLong() > maxRequestBodySize) {
+            resp.sendError(WebdavStatus.SC_REQUEST_TOO_LONG);
+            return;
+        }
+
         byte[] body;
-        try (InputStream is = req.getInputStream(); ByteArrayOutputStream os = new ByteArrayOutputStream()) {
+        try (InputStream is = req.getInputStream();
+                BoundedByteArrayOutputStream os = new BoundedByteArrayOutputStream(maxRequestBodySize)) {
             IOTools.flow(is, os);
             body = os.toByteArray();
         } catch (IOException ioe) {
             resp.sendError(WebdavStatus.SC_BAD_REQUEST);
             return;
+        } catch (ArrayIndexOutOfBoundsException e) {
+            resp.sendError(WebdavStatus.SC_REQUEST_TOO_LONG);
+            return;
         }
         if (body.length > 0) {
             DocumentBuilder documentBuilder = getDocumentBuilder();
@@ -1389,13 +1416,23 @@ public class WebdavServlet extends DefaultServlet implements PeriodicEventListen
 
         Node lockInfoNode = null;
 
+        // Short-cut if client provided a content length
+        if (req.getContentLengthLong() > maxRequestBodySize) {
+            resp.sendError(WebdavStatus.SC_REQUEST_TOO_LONG);
+            return;
+        }
+
         byte[] body;
-        try (InputStream is = req.getInputStream(); ByteArrayOutputStream os = new ByteArrayOutputStream()) {
+        try (InputStream is = req.getInputStream();
+                BoundedByteArrayOutputStream os = new BoundedByteArrayOutputStream(maxRequestBodySize)) {
             IOTools.flow(is, os);
             body = os.toByteArray();
         } catch (IOException ioe) {
             resp.sendError(WebdavStatus.SC_BAD_REQUEST);
             return;
+        } catch (ArrayIndexOutOfBoundsException e) {
+            resp.sendError(WebdavStatus.SC_REQUEST_TOO_LONG);
+            return;
         }
         if (body.length > 0) {
             DocumentBuilder documentBuilder = getDocumentBuilder();
@@ -3020,6 +3057,34 @@ public class WebdavServlet extends DefaultServlet implements PeriodicEventListen
     }
 
 
+    static class BoundedByteArrayOutputStream extends ByteArrayOutputStream {
+
+        private final int sizeLimit;
+        private int size;
+
+        BoundedByteArrayOutputStream(int sizeLimit) {
+            super();
+            this.sizeLimit = sizeLimit;
+        }
+
+        @Override
+        public synchronized void write(int b) {
+            size++;
+            if (size > sizeLimit) {
+                throw new ArrayIndexOutOfBoundsException();
+            }
+            super.write(b);
+        }
+
+        @Override
+        public synchronized void write(byte[] b, int off, int len) {
+            size += len;
+            if (size > sizeLimit) {
+                throw new ArrayIndexOutOfBoundsException();
+            }
+            super.write(b, off, len);
+        }
+    }
 }
 
 
diff --git a/test/org/apache/catalina/servlets/TestWebdavBoundedByteArrayOutputStream.java b/test/org/apache/catalina/servlets/TestWebdavBoundedByteArrayOutputStream.java
new file mode 100644
--- /dev/null
+++ b/test/org/apache/catalina/servlets/TestWebdavBoundedByteArrayOutputStream.java
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.catalina.servlets;
+
+import java.io.IOException;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import org.apache.catalina.servlets.WebdavServlet.BoundedByteArrayOutputStream;
+
+public class TestWebdavBoundedByteArrayOutputStream {
+
+    private static final int TEST_LIMIT = 10;
+    private static final byte[] ONE_BYTE_ARRAY = new byte[] { 0 };
+
+
+    @Test
+    public void testWriteByte() {
+        BoundedByteArrayOutputStream bbaos = new BoundedByteArrayOutputStream(TEST_LIMIT);
+
+        for (int i = 0; i < TEST_LIMIT; i++) {
+            bbaos.write(0);
+        }
+
+        try {
+            bbaos.write(0);
+            Assert.fail("Writing 11th byte failed to trigger error");
+        } catch (ArrayIndexOutOfBoundsException e) {
+            // Pass
+        }
+    }
+
+
+    @Test
+    public void testWriteByteArray() throws IOException {
+        BoundedByteArrayOutputStream bbaos = new BoundedByteArrayOutputStream(TEST_LIMIT);
+
+        for (int i = 0; i < TEST_LIMIT; i++) {
+            bbaos.write(ONE_BYTE_ARRAY);
+        }
+
+        try {
+            bbaos.write(ONE_BYTE_ARRAY);
+            Assert.fail("Writing 11th byte failed to trigger error");
+        } catch (ArrayIndexOutOfBoundsException e) {
+            // Pass
+        }
+    }
+
+
+    @Test
+    public void testWriteByteSubArray() {
+        BoundedByteArrayOutputStream bbaos = new BoundedByteArrayOutputStream(TEST_LIMIT);
+
+        for (int i = 0; i < TEST_LIMIT; i++) {
+            bbaos.write(ONE_BYTE_ARRAY, 0, 1);
+        }
+
+        try {
+            bbaos.write(ONE_BYTE_ARRAY, 0, 1);
+            Assert.fail("Writing 11th byte failed to trigger error");
+        } catch (ArrayIndexOutOfBoundsException e) {
+            // Pass
+        }
+    }
+
+
+    @Test
+    public void testWriteBytes() {
+        BoundedByteArrayOutputStream bbaos = new BoundedByteArrayOutputStream(TEST_LIMIT);
+
+        for (int i = 0; i < TEST_LIMIT; i++) {
+            bbaos.writeBytes(ONE_BYTE_ARRAY);
+        }
+
+        try {
+            bbaos.writeBytes(ONE_BYTE_ARRAY);
+            Assert.fail("Writing 11th byte failed to trigger error");
+        } catch (ArrayIndexOutOfBoundsException e) {
+            // Pass
+        }
+    }
+}
diff --git a/test/org/apache/catalina/servlets/TestWebdavServlet.java b/test/org/apache/catalina/servlets/TestWebdavServlet.java
--- a/test/org/apache/catalina/servlets/TestWebdavServlet.java
+++ b/test/org/apache/catalina/servlets/TestWebdavServlet.java
@@ -75,6 +75,7 @@ public class TestWebdavServlet extends TomcatBaseTest {
 
         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-specialpath"+UUID.randomUUID());
+        tempWebapp.deleteOnExit();
         Assert.assertTrue("Failed to mkdirs on "+tempWebapp.getCanonicalPath(),tempWebapp.mkdirs());
         Assert.assertTrue(new File(tempWebapp,"WEB-INF").mkdir());
         Assert.assertTrue(new File(tempWebapp,"META-INF").mkdir());
@@ -297,6 +298,7 @@ public class TestWebdavServlet extends TomcatBaseTest {
 
         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-properties");
+        tempWebapp.deleteOnExit();
         Assert.assertTrue(tempWebapp.mkdirs());
         Context ctxt = tomcat.addContext("", tempWebapp.getAbsolutePath());
         Wrapper webdavServlet = Tomcat.addServlet(ctxt, "webdav", new WebdavServlet());
@@ -441,6 +443,7 @@ public class TestWebdavServlet extends TomcatBaseTest {
 
         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-webapp");
+        tempWebapp.deleteOnExit();
         Assert.assertTrue(tempWebapp.mkdirs());
         Context ctxt = tomcat.addContext("", tempWebapp.getAbsolutePath());
         Wrapper webdavServlet = Tomcat.addServlet(ctxt, "webdav", new WebdavServlet());
@@ -922,6 +925,7 @@ public class TestWebdavServlet extends TomcatBaseTest {
 
         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-subpath");
+        tempWebapp.deleteOnExit();
         File subPath = new File(tempWebapp, "aaa");
         Assert.assertTrue(subPath.mkdirs());
 
@@ -1018,6 +1022,7 @@ public class TestWebdavServlet extends TomcatBaseTest {
 
         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-lock");
+        tempWebapp.deleteOnExit();
         Assert.assertTrue(tempWebapp.mkdirs());
         Context ctxt = tomcat.addContext("", tempWebapp.getAbsolutePath());
         Wrapper webdavServlet = Tomcat.addServlet(ctxt, "webdav", new WebdavServlet());
@@ -1405,6 +1410,7 @@ public class TestWebdavServlet extends TomcatBaseTest {
 
         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-if");
+        tempWebapp.deleteOnExit();
         File folder = new File(tempWebapp, "/myfolder/myfolder2/myfolder4/myfolder5");
         Assert.assertTrue(folder.mkdirs());
         File file = new File(folder, "myfile.txt");
@@ -1535,6 +1541,7 @@ public class TestWebdavServlet extends TomcatBaseTest {
 
         // Create a temp webapp that can be safely written to
         File tempWebapp = new File(getTemporaryDirectory(), "webdav-store");
+        tempWebapp.deleteOnExit();
         Assert.assertTrue(tempWebapp.mkdirs());
         Context ctxt = tomcat.addContext("", tempWebapp.getAbsolutePath());
         Wrapper webdavServlet = Tomcat.addServlet(ctxt, "webdav", new WebdavServlet());
@@ -1566,6 +1573,82 @@ public class TestWebdavServlet extends TomcatBaseTest {
         validateXml(client.getResponseBody());
     }
 
+
+    /*
+     * Only tests LOCK bodies exceeding limit. Other tests cover valid LOCK bodies.
+     */
+    @Test
+    public void testLockBodyLimit() throws Exception {
+        doTestLimit("LOCK", LOCK_BODY);
+    }
+
+
+    /*
+     * Only tests PROPFIND bodies exceeding limit. Other tests cover valid PROPFIND bodies.
+     */
+    @Test
+    public void testPropFindBodyLimit() throws Exception {
+        doTestLimit("PROPFIND", PROPFIND_PROP);
+    }
+
+
+    private void doTestLimit(String method, String requestBody) throws Exception {
+
+        Tomcat tomcat = getTomcatInstance();
+
+        File appDir = new File("test/webapp");
+        Context ctxt = tomcat.addContext("", appDir.getAbsolutePath());
+
+        Wrapper webdavServlet = Tomcat.addServlet(ctxt, "webdav", new WebdavServlet());
+        webdavServlet.addInitParameter("listings", "true");
+        webdavServlet.addInitParameter("secret", "foo");
+        webdavServlet.addInitParameter("readonly", "false");
+        webdavServlet.addInitParameter("useStrongETags", "true");
+        webdavServlet.addInitParameter("maxRequestBodySize", "10");
+
+        ctxt.addServletMappingDecoded("/*", "webdav");
+        tomcat.start();
+
+        // With content length
+        Client client = new Client();
+        client.setPort(getPort());
+
+        // @formatter:off
+        client.setRequest(new String[] {
+                method + " / HTTP/1.1" + CRLF +
+                    "Host: localhost:" + getPort() + CRLF +
+                    "Content-Length: " + requestBody.length() + CRLF +
+                    "Connection: Close" + CRLF +
+                    CRLF +
+                    requestBody
+                });
+        // @formatter:on
+        client.connect();
+        client.processRequest(true);
+        Assert.assertEquals(WebdavStatus.SC_REQUEST_TOO_LONG, client.getStatusCode());
+
+        // Without content length
+        client.reset();
+
+        // @formatter:off
+        client.setRequest(new String[] {
+                method + " / HTTP/1.1" + CRLF +
+                    "Host: localhost:" + getPort() + CRLF +
+                    "Transfer-Encoding: chunked" + CRLF +
+                    "Connection: Close" + CRLF +
+                    CRLF +
+                    Integer.toHexString(requestBody.length()) + CRLF +
+                    requestBody + CRLF +
+                    "0" + CRLF +
+                    CRLF
+                });
+        // @formatter:on
+        client.connect();
+        client.processRequest(true);
+        Assert.assertEquals(WebdavStatus.SC_REQUEST_TOO_LONG, client.getStatusCode());
+    }
+
+
     public static class CustomPropertyStore implements PropertyStore {
 
         private String propertyName = null;
