--- a/lib/path/evaluation.js
+++ b/lib/path/evaluation.js
@@ -373,7 +373,9 @@
 
         if (_object.isIdentifier() && _property.isIdentifier() && VALID_CALLEES.indexOf(_object.node.name) >= 0 && INVALID_METHODS.indexOf(_property.node.name) < 0) {
           context = global[_object.node.name];
-          func = context[_property.node.name];
+          if (Object.hasOwnProperty.call(context, _property.node.name)) {
+            func = context[_property.node.name];
+          }
         }
 
         if (_object.isLiteral() && _property.isIdentifier()) {
