diff --git a/lib/patch/parse.js b/lib/patch/parse.js
--- a/lib/patch/parse.js
+++ b/lib/patch/parse.js
@@ -24,8 +24,8 @@ function parsePatch(uniDiff) {
       }
 
       // Diff index
-      var header = /^(?:Index:|diff(?: -r \w+)+)\s+(.+?)\s*$/.exec(line);
+      var header = /^(?:Index:|diff(?: -r \w+)+)\s+/.exec(line);
       if (header) {
-        index.index = header[1];
+        index.index = line.substring(header[0].length).trim();
       }
       i++;
@@ -56,15 +56,16 @@ function parsePatch(uniDiff) {
   // are consumed.
   function parseFileHeader(index) {
-    var fileHeader = /^(---|\+\+\+)\s+(.*)\r?$/.exec(diffstr[i]);
-    if (fileHeader) {
-      var keyPrefix = fileHeader[1] === '---' ? 'old' : 'new';
-      var data = fileHeader[2].split('\t', 2);
+    var fileHeaderMatch = /^(---|\+\+\+)\s+/.exec(diffstr[i]);
+    if (fileHeaderMatch) {
+      var prefix = fileHeaderMatch[1];
+      var data = diffstr[i].substring(3).trim().split('\t', 2);
       var fileName = data[0].replace(/\\\\/g, '\\');
-      if (/^".*"$/.test(fileName)) {
+      if (fileName.startsWith('"') && fileName.endsWith('"')) {
         fileName = fileName.substr(1, fileName.length - 2);
       }
+      var keyPrefix = prefix === '---' ? 'old' : 'new';
       index[keyPrefix + 'FileName'] = fileName;
       index[keyPrefix + 'Header'] = (data[1] || '').trim();
       i++;
