diff --git a/logback-core/src/main/java/ch/qos/logback/core/model/processor/conditional/IfModelHandler.java b/logback-core/src/main/java/ch/qos/logback/core/model/processor/conditional/IfModelHandler.java
index 7dcaf4a..047746e 100644
--- a/logback-core/src/main/java/ch/qos/logback/core/model/processor/conditional/IfModelHandler.java
+++ b/logback-core/src/main/java/ch/qos/logback/core/model/processor/conditional/IfModelHandler.java
@@ -33,6 +33,9 @@ public class IfModelHandler extends ModelHandlerBase {
     public static final String MISSING_JANINO_MSG = "Could not find Janino library on the class path. Skipping conditional processing.";
     public static final String MISSING_JANINO_SEE = "See also " + CoreConstants.CODES_URL + "#ifJanino";
 
+    public static final String NEW_OPERATOR_DISALLOWED_MSG = "The 'condition' attribute may not contain the 'new' operator.";
+    public static final String NEW_OPERATOR_DISALLOWED_SEE = "See also " + CoreConstants.CODES_URL + "#conditionNew";
+
     enum Branch {IF_BRANCH, ELSE_BRANCH; }
     
     IfModel ifModel = null;
@@ -75,6 +78,13 @@ public class IfModelHandler extends ModelHandlerBase {
                return;
             }
 
+            // do not allow 'new' operator
+            if(hasNew(conditionStr)) {
+                addError(NEW_OPERATOR_DISALLOWED_MSG);
+                addError(NEW_OPERATOR_DISALLOWED_SEE);
+                return;
+            }
+
             try {
                 PropertyEvalScriptBuilder pesb = new PropertyEvalScriptBuilder(mic);
                 pesb.setContext(context);
@@ -96,8 +106,11 @@ public class IfModelHandler extends ModelHandlerBase {
             }
         }
     }
-    
-    
+
+    private boolean hasNew(String conditionStr) {
+        return conditionStr.contains("new ");
+    }
+
     @Override
     public void postHandle(ModelInterpretationContext mic, Model model) throws ModelHandlerException {
 

diff --git a/logback-classic/src/main/java/ch/qos/logback/classic/joran/ModelClassToModelHandlerLinker.java b/logback-classic/src/main/java/ch/qos/logback/classic/joran/ModelClassToModelHandlerLinker.java
index 3c99410..11ae8d1 100644
--- a/logback-classic/src/main/java/ch/qos/logback/classic/joran/ModelClassToModelHandlerLinker.java
+++ b/logback-classic/src/main/java/ch/qos/logback/classic/joran/ModelClassToModelHandlerLinker.java
@@ -33,6 +33,7 @@ import ch.qos.logback.core.model.AppenderRefModel;
 import ch.qos.logback.core.model.InsertFromJNDIModel;
 import ch.qos.logback.core.model.ModelHandlerFactoryMethod;
 import ch.qos.logback.core.model.processor.AppenderModelHandler;
+import ch.qos.logback.core.model.processor.AppenderDeclarationAnalyser;
 import ch.qos.logback.core.model.processor.AppenderRefDependencyAnalyser;
 import ch.qos.logback.core.model.processor.AppenderRefModelHandler;
 import ch.qos.logback.core.model.processor.DefaultProcessor;
@@ -80,6 +81,8 @@ public class ModelClassToModelHandlerLinker extends ModelClassToModelHandlerLink
         defaultProcessor.addAnalyser(AppenderModel.class,
                 () -> new RefContainerDependencyAnalyser(context, AppenderModel.class));
 
+        defaultProcessor.addAnalyser(AppenderModel.class, () -> new AppenderDeclarationAnalyser(context));
+
         defaultProcessor.addAnalyser(AppenderRefModel.class, () -> new AppenderRefDependencyAnalyser(context));
 
         sealModelFilters(defaultProcessor);
diff --git a/logback-core/src/main/java/ch/qos/logback/core/joran/action/ImplicitModelDataForComplexProperty.java b/logback-core/src/main/java/ch/qos/logback/core/joran/action/ImplicitModelDataForComplexProperty.java
index f53fb31..ac1dcbd 100644
--- a/logback-core/src/main/java/ch/qos/logback/core/joran/action/ImplicitModelDataForComplexProperty.java
+++ b/logback-core/src/main/java/ch/qos/logback/core/joran/action/ImplicitModelDataForComplexProperty.java
@@ -24,6 +24,7 @@ import ch.qos.logback.core.util.AggregationType;
 public class ImplicitModelDataForComplexProperty extends ImplicitModelData {
 
     private Object nestedComplexProperty;
+    private Class<?> expectedPropertyType;
 
     public ImplicitModelDataForComplexProperty(PropertySetter parentBean, AggregationType aggregationType,
             String propertyName) {
@@ -34,6 +35,14 @@ public class ImplicitModelDataForComplexProperty extends ImplicitModelData {
         return nestedComplexProperty;
     }
 
+    public Class<?> getExpectedPropertyType() {
+        return expectedPropertyType;
+    }
+
+    public void setExpectedPropertyType(Class<?> expectedPropertyType) {
+        this.expectedPropertyType = expectedPropertyType;
+    }
+
     public void setNestedComplexProperty(Object nestedComplexProperty) {
         this.nestedComplexProperty = nestedComplexProperty;
     }
diff --git a/logback-core/src/main/java/ch/qos/logback/core/joran/util/PropertySetter.java b/logback-core/src/main/java/ch/qos/logback/core/joran/util/PropertySetter.java
index 100ffd3..b394e3d 100644
--- a/logback-core/src/main/java/ch/qos/logback/core/joran/util/PropertySetter.java
+++ b/logback-core/src/main/java/ch/qos/logback/core/joran/util/PropertySetter.java
@@ -392,4 +392,38 @@ public class PropertySetter extends ContextAwareBase {
         return getByConcreteType(name, relevantMethod);
     }
 
+    public Class<?> getTypeForComplexProperty(String nestedElementTagName, AggregationType aggregationType) {
+
+        String cName = capitalizeFirstLetter(nestedElementTagName);
+
+        Method aMethod = null;
+        switch (aggregationType) {
+            case AS_COMPLEX_PROPERTY:
+                aMethod = findSetterMethod(nestedElementTagName);
+                break;
+            case AS_COMPLEX_PROPERTY_COLLECTION:
+                aMethod = findAdderMethod(cName);
+                break;
+        }
+
+        checkParameterCount(aMethod, nestedElementTagName);
+
+        Class<?>[] paramTypes = aMethod.getParameterTypes();
+        return paramTypes[0];
+    }
+
+    private void checkParameterCount(Method aMethod, String nestedElementTagName) {
+        if (aMethod == null) {
+            String msg = "Could not find method for property [" + nestedElementTagName + "].";
+            addError(msg);
+            throw new IllegalStateException(msg);
+        }
+        int parameterCount = aMethod.getParameterCount();
+        if (parameterCount != 1) {
+            String msg = "Expected [" + aMethod.getName() + "] for property [" + nestedElementTagName + "] to have exactly one parameter.";
+            addError(msg);
+            throw new IllegalStateException(msg);
+        }
+    }
+
 }
diff --git a/logback-core/src/main/java/ch/qos/logback/core/model/processor/AppenderDeclarationAnalyser.java b/logback-core/src/main/java/ch/qos/logback/core/model/processor/AppenderDeclarationAnalyser.java
new file mode 100644
index 0000000..9b6c64d
--- /dev/null
+++ b/logback-core/src/main/java/ch/qos/logback/core/model/processor/AppenderDeclarationAnalyser.java
@@ -0,0 +1,76 @@
+/*
+ * Logback: the reliable, generic, fast and flexible logging framework.
+ *  Copyright (C) 1999-2026, QOS.ch. All rights reserved.
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation
+ *
+ *     or (per the licensee's choosing)
+ *
+ * under the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation.
+ */
+
+package ch.qos.logback.core.model.processor;
+
+import ch.qos.logback.core.Context;
+import ch.qos.logback.core.model.AppenderModel;
+import ch.qos.logback.core.model.Model;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * The AppenderDeclarationAnalyser class is responsible for analyzing the availability
+ * of appenders. By available, we mean whether an appender with a given name is declared
+ * somewhere in the configuration. This availability information is later used by
+ * AppenderRefModelHandler to attempt to attach only those appenders that were previously
+ * declared.
+ */
+@PhaseIndicator(phase = ProcessingPhase.DEPENDENCY_ANALYSIS)
+public class AppenderDeclarationAnalyser extends ModelHandlerBase {
+
+    static final String DECLARED_APPENDER_NAME_SET_KEY = "DECLARED_APPENDER_NAME_SET";
+
+
+    public AppenderDeclarationAnalyser(Context context) {
+        super(context);
+    }
+
+    @Override
+    protected Class<AppenderModel> getSupportedModelClass() {
+        return AppenderModel.class;
+    }
+
+
+    @Override
+    public void handle(ModelInterpretationContext mic, Model model) throws ModelHandlerException {
+        AppenderModel appenderModel = (AppenderModel) model;
+
+        String appenderName = mic.subst(appenderModel.getName());
+        addAppenderDeclaration(mic, appenderName);
+    }
+
+
+    static public Set<String> getAppenderNameSet(ModelInterpretationContext mic) {
+        Set<String> set = (Set<String>) mic.getObjectMap().get(DECLARED_APPENDER_NAME_SET_KEY);
+        if(set == null) {
+            set = new HashSet<>();
+            mic.getObjectMap().put(DECLARED_APPENDER_NAME_SET_KEY, set);
+        }
+        return set;
+    }
+
+    static public void addAppenderDeclaration(ModelInterpretationContext mic, String appenderName) {
+        Set<String> set = getAppenderNameSet(mic);
+        set.add(appenderName);
+    }
+
+
+    static public boolean isAppenderDeclared(ModelInterpretationContext mic, String appenderName) {
+        Set<String> set = getAppenderNameSet(mic);
+        return set.contains(appenderName);
+    }
+
+}
diff --git a/logback-core/src/main/java/ch/qos/logback/core/model/processor/AppenderRefModelHandler.java b/logback-core/src/main/java/ch/qos/logback/core/model/processor/AppenderRefModelHandler.java
index 12e7318..ef6d439 100755
--- a/logback-core/src/main/java/ch/qos/logback/core/model/processor/AppenderRefModelHandler.java
+++ b/logback-core/src/main/java/ch/qos/logback/core/model/processor/AppenderRefModelHandler.java
@@ -1,5 +1,6 @@
 package ch.qos.logback.core.model.processor;
 
+import static ch.qos.logback.core.model.processor.AppenderDeclarationAnalyser.isAppenderDeclared;
 import java.util.Map;
 
 import ch.qos.logback.core.Appender;
@@ -50,6 +51,11 @@ public class AppenderRefModelHandler extends ModelHandlerBase {
             AppenderAttachable<?> appenderAttachable) {
         String appenderName = mic.subst(appenderRefModel.getRef());
 
+        if(!isAppenderDeclared(mic, appenderName)) {
+            addWarn("Appender named [" + appenderName + "] could not be found. Skipping attachment to "+appenderAttachable+".");
+            return;
+        }
+
         Map<String, Appender> appenderBag = (Map<String, Appender>) mic.getObjectMap().get(JoranConstants.APPENDER_BAG);
 
         Appender appender = appenderBag.get(appenderName);
diff --git a/logback-core/src/main/java/ch/qos/logback/core/model/processor/DefaultProcessor.java b/logback-core/src/main/java/ch/qos/logback/core/model/processor/DefaultProcessor.java
index 89549d2..7cb1711 100755
--- a/logback-core/src/main/java/ch/qos/logback/core/model/processor/DefaultProcessor.java
+++ b/logback-core/src/main/java/ch/qos/logback/core/model/processor/DefaultProcessor.java
@@ -15,6 +15,7 @@ package ch.qos.logback.core.model.processor;
 
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.function.Supplier;
@@ -42,7 +43,7 @@ public class DefaultProcessor extends ContextAwareBase {
 
     final protected ModelInterpretationContext mic;
     final HashMap<Class<? extends Model>, ModelHandlerFactoryMethod> modelClassToHandlerMap = new HashMap<>();
-    final HashMap<Class<? extends Model>, Supplier<ModelHandlerBase>> modelClassToDependencyAnalyserMap = new HashMap<>();
+    final HashMap<Class<? extends Model>, List<Supplier<ModelHandlerBase>>> modelClassToDependencyAnalyserMap = new HashMap<>();
 
     ChainedModelFilter phaseOneFilter = new ChainedModelFilter();
     ChainedModelFilter phaseTwoFilter = new ChainedModelFilter();
@@ -81,7 +82,7 @@ public class DefaultProcessor extends ContextAwareBase {
     }
 
     public void addAnalyser(Class<? extends Model> modelClass, Supplier<ModelHandlerBase> analyserSupplier) {
-        modelClassToDependencyAnalyserMap.put(modelClass, analyserSupplier);
+        modelClassToDependencyAnalyserMap.computeIfAbsent(modelClass, x -> new ArrayList<>()).add(analyserSupplier);
     }
 
     private void traversalLoop(TraverseMethod traverseMethod, Model model, ModelFilter modelfFilter, String phaseName) {
@@ -127,25 +128,31 @@ public class DefaultProcessor extends ContextAwareBase {
 
 
     protected void analyseDependencies(Model model) {
-        Supplier<ModelHandlerBase> analyserSupplier = modelClassToDependencyAnalyserMap.get(model.getClass());
 
-        ModelHandlerBase analyser = null;
+        List<Supplier<ModelHandlerBase>> analyserSupplierList = modelClassToDependencyAnalyserMap.get(model.getClass());
 
-        if (analyserSupplier != null) {
-            analyser = analyserSupplier.get();
-        }
+        if (analyserSupplierList != null) {
+            for (Supplier<ModelHandlerBase> analyserSupplier : analyserSupplierList) {
+                ModelHandlerBase analyser = null;
+
+                if (analyserSupplier != null) {
+                    analyser = analyserSupplier.get();
+                }
+
+                if (analyser != null && !model.isSkipped()) {
+                    callAnalyserHandleOnModel(model, analyser);
+                }
 
-        if (analyser != null && !model.isSkipped()) {
-            callAnalyserHandleOnModel(model, analyser);
+                if (analyser != null && !model.isSkipped()) {
+                    callAnalyserPostHandleOnModel(model, analyser);
+                }
+            }
         }
 
         for (Model m : model.getSubModels()) {
             analyseDependencies(m);
         }
 
-        if (analyser != null && !model.isSkipped()) {
-            callAnalyserPostHandleOnModel(model, analyser);
-        }
     }
 
     private void callAnalyserPostHandleOnModel(Model model, ModelHandlerBase analyser) {
@@ -281,14 +288,20 @@ public class DefaultProcessor extends ContextAwareBase {
     }
 
     private boolean allDependenciesStarted(Model model) {
+        // assumes that DependencyDefinitions have been registered
         List<String> dependencyNames = mic.getDependeeNamesForModel(model);
 
         if (dependencyNames == null || dependencyNames.isEmpty()) {
             return true;
         }
         for (String name : dependencyNames) {
+            boolean isRegistered = AppenderDeclarationAnalyser.isAppenderDeclared(mic, name);
+            if (!isRegistered) {
+                // non registered dependencies are not taken into account
+                continue;
+            }
             boolean isStarted = mic.isNamedDependeeStarted(name);
-            if (isStarted == false) {
+            if (!isStarted) {
                 return false;
             }
         }
diff --git a/logback-core/src/main/java/ch/qos/logback/core/model/processor/ImplicitModelHandler.java b/logback-core/src/main/java/ch/qos/logback/core/model/processor/ImplicitModelHandler.java
index aaa7b66..b386b0f 100755
--- a/logback-core/src/main/java/ch/qos/logback/core/model/processor/ImplicitModelHandler.java
+++ b/logback-core/src/main/java/ch/qos/logback/core/model/processor/ImplicitModelHandler.java
@@ -88,10 +88,12 @@ public class ImplicitModelHandler extends ModelHandlerBase {
         // we only push action data if NestComponentIA is applicable
         case AS_COMPLEX_PROPERTY_COLLECTION:
         case AS_COMPLEX_PROPERTY:
-            ImplicitModelDataForComplexProperty adComplex = new ImplicitModelDataForComplexProperty(parentBean,
+            Class<?> propertyType = parentBean.getTypeForComplexProperty(nestedElementTagName, aggregationType);
+            ImplicitModelDataForComplexProperty imdForComplexProperty = new ImplicitModelDataForComplexProperty(parentBean,
                     aggregationType, nestedElementTagName);
-            implicitModelData = adComplex;
-            doComplex(mic, implicitModel, adComplex);
+            imdForComplexProperty.setExpectedPropertyType(propertyType);
+            implicitModelData = imdForComplexProperty;
+            doComplex(mic, implicitModel, imdForComplexProperty);
             return;
         default:
             addError("PropertySetter.computeAggregationType returned " + aggregationType);
@@ -119,7 +121,7 @@ public class ImplicitModelHandler extends ModelHandlerBase {
     }
 
     public void doComplex(ModelInterpretationContext interpretationContext, ComponentModel componentModel,
-            ImplicitModelDataForComplexProperty actionData) {
+            ImplicitModelDataForComplexProperty imdForComplexProperty) {
 
         String className = componentModel.getClassName();
         // perform variable name substitution
@@ -134,13 +136,13 @@ public class ImplicitModelHandler extends ModelHandlerBase {
                 componentClass = Loader.loadClass(fqcn, context);
             } else {
                 // guess class name via implicit rules
-                PropertySetter parentBean = actionData.parentBean;
-                componentClass = parentBean.getClassNameViaImplicitRules(actionData.propertyName,
-                        actionData.getAggregationType(), interpretationContext.getDefaultNestedComponentRegistry());
+                PropertySetter parentBean = imdForComplexProperty.parentBean;
+                componentClass = parentBean.getClassNameViaImplicitRules(imdForComplexProperty.propertyName,
+                        imdForComplexProperty.getAggregationType(), interpretationContext.getDefaultNestedComponentRegistry());
             }
 
             if (componentClass == null) {
-                actionData.inError = true;
+                imdForComplexProperty.inError = true;
                 String errMsg = "Could not find an appropriate class for property [" + componentModel.getTag() + "]";
                 addError(errMsg);
                 return;
@@ -151,18 +153,22 @@ public class ImplicitModelHandler extends ModelHandlerBase {
                         + "] property");
             }
 
-            actionData.setNestedComplexProperty(componentClass.getConstructor().newInstance());
+            Class<?> expectedPropertyType = imdForComplexProperty.getExpectedPropertyType();
 
-            // pass along the repository
-            if (actionData.getNestedComplexProperty() instanceof ContextAware) {
-                ((ContextAware) actionData.getNestedComplexProperty()).setContext(this.context);
+            Object object = OptionHelper.instantiateClassWithSuperclassRestriction(componentClass, expectedPropertyType);
+
+            imdForComplexProperty.setNestedComplexProperty(object);
+
+            // pass along the context
+            if (imdForComplexProperty.getNestedComplexProperty() instanceof ContextAware) {
+                ((ContextAware) imdForComplexProperty.getNestedComplexProperty()).setContext(this.context);
             }
             // addInfo("Pushing component [" + localName
             // + "] on top of the object stack.");
-            interpretationContext.pushObject(actionData.getNestedComplexProperty());
+            interpretationContext.pushObject(imdForComplexProperty.getNestedComplexProperty());
 
         } catch (Exception oops) {
-            actionData.inError = true;
+            imdForComplexProperty.inError = true;
             String msg = "Could not create component [" + componentModel.getTag() + "] of type [" + fqcn + "]";
             addError(msg, oops);
         }
diff --git a/logback-core/src/main/java/ch/qos/logback/core/util/OptionHelper.java b/logback-core/src/main/java/ch/qos/logback/core/util/OptionHelper.java
index de4961e..58a2be0 100644
--- a/logback-core/src/main/java/ch/qos/logback/core/util/OptionHelper.java
+++ b/logback-core/src/main/java/ch/qos/logback/core/util/OptionHelper.java
@@ -14,6 +14,7 @@
 package ch.qos.logback.core.util;
 
 import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
 import java.util.Properties;
 
 import ch.qos.logback.core.Context;
@@ -44,6 +45,19 @@ public class OptionHelper {
         return instantiateByClassNameAndParameter(className, superClass, classLoader, null, null);
     }
 
+    public static Object instantiateClassWithSuperclassRestriction(Class<?> classObj, Class<?> superClass)
+            throws IncompatibleClassException, DynamicClassLoadingException {
+        if (!superClass.isAssignableFrom(classObj)) {
+            throw new IncompatibleClassException(superClass, classObj);
+        }
+
+        try {
+            return classObj.getConstructor().newInstance();
+        } catch (NoSuchMethodException|InstantiationException|IllegalAccessException|InvocationTargetException e) {
+            throw new DynamicClassLoadingException("Failed to instantiate type " + classObj.getName(), e);
+        }
+    }
+
     public static Object instantiateByClassNameAndParameter(String className, Class<?> superClass,
             ClassLoader classLoader, Class<?> type, Object parameter)
             throws IncompatibleClassException, DynamicClassLoadingException {
