--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/Configuration.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/Configuration.java
@@ -1457,4 +1457,10 @@ public interface Configuration {
        Eventually with some coordination we can update it from various server components. */
    // Inspired by https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#:~:text=The%20status%20describes%20the%20current,the%20desired%20state%20you%20supplied
    void setStatus(String status);
+
+   List<String> getFederationDownstreamAuthorization();
+
+   void setFederationDownstreamAuthorization(List<String> roles);
+
+   Configuration addFederationDownstreamAuthorization(String role);
 }
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImpl.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImpl.java
@@ -216,6 +216,8 @@ public class ConfigurationImpl implements Configuration, Serializable {
 
    protected List<FederationConfiguration> federationConfigurations = new ArrayList<>();
 
+   protected List<String> downstreamAuthorization = new ArrayList<>();
+
    @Deprecated
    // this can eventually be replaced with List<QueueConfiguration>, but to keep existing semantics it must stay as is for now
    private List<CoreQueueConfiguration> coreQueueConfigurations = new ArrayList<>();
@@ -3194,6 +3196,22 @@ public class ConfigurationImpl implements Configuration, Serializable {
       this.jsonStatus = JsonUtil.mergeAndUpdate(getJsonStatus(), update);
    }
 
+
+   @Override
+   public List<String> getFederationDownstreamAuthorization() {
+      return downstreamAuthorization;
+   }
+
+   @Override
+   public void setFederationDownstreamAuthorization(List<String> roles) {
+      this.downstreamAuthorization = roles;
+   }
+
+   @Override
+   public Configuration addFederationDownstreamAuthorization(String role) {
+      downstreamAuthorization.add(role);
+      return this;
+   }
    // extend property utils with ability to auto-fill and locate from collections
    // collection entries are identified by the name() property
    private static class CollectionAutoFillPropertiesUtil extends PropertyUtilsBean {
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/deployers/impl/FileConfigurationParser.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/deployers/impl/FileConfigurationParser.java
@@ -29,6 +29,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
 
 import org.apache.activemq.artemis.ArtemisConstants;
 import org.apache.activemq.artemis.api.config.ActiveMQDefaultConfiguration;
@@ -615,6 +616,13 @@ public final class FileConfigurationParser extends XMLConfigurationUtil {
          parseBridgeConfiguration(mfNode, config);
       }
 
+      NodeList federations = e.getElementsByTagName("federations");
+
+      for (int i = 0; i < federations.getLength(); i++) {
+         Element fedNode = (Element) federations.item(i);
+         parseFederationsConfiguration(fedNode, config);
+      }
+
       NodeList fedNodes = e.getElementsByTagName("federation");
 
       for (int i = 0; i < fedNodes.getLength(); i++) {
@@ -2551,6 +2559,16 @@ public final class FileConfigurationParser extends XMLConfigurationUtil {
       mainConfig.getBridgeConfigurations().add(config);
    }
 
+   private void parseFederationsConfiguration(final Element fedNode, final Configuration mainConfig) throws Exception {
+      String roles = fedNode.getAttribute("downstream-authorization");
+      if (!roles.isEmpty()) {
+         Stream.of(roles.split(","))
+            .map(String::trim)
+            .filter(s -> !s.isEmpty())
+            .forEach(mainConfig::addFederationDownstreamAuthorization);
+      }
+   }
+
    private void parseFederationConfiguration(final Element fedNode, final Configuration mainConfig) throws Exception {
       FederationConfiguration config = new FederationConfiguration();
 
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/CoreProtocolManager.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/CoreProtocolManager.java
@@ -65,6 +65,9 @@ import org.apache.activemq.artemis.core.remoting.FailureListener;
 import org.apache.activemq.artemis.core.remoting.impl.netty.ActiveMQFrameDecoder2;
 import org.apache.activemq.artemis.core.remoting.impl.netty.NettyServerConnection;
 import org.apache.activemq.artemis.core.server.ActiveMQServer;
+import org.apache.activemq.artemis.core.server.ActiveMQServerLogger;
+import org.apache.activemq.artemis.spi.core.security.jaas.UserPrincipal;
+import org.apache.activemq.artemis.utils.SecurityManagerUtil;
 import org.apache.activemq.artemis.spi.core.protocol.ConnectionEntry;
 import org.apache.activemq.artemis.spi.core.protocol.ProtocolManager;
 import org.apache.activemq.artemis.spi.core.protocol.ProtocolManagerFactory;
@@ -163,8 +166,7 @@ public class CoreProtocolManager implements ProtocolManager<Interceptor, ActiveM
          .addClusterChannelHandler(rc.getChannel(CHANNEL_ID.CLUSTER.id, -1), acceptorUsed, rc,
                                    server.getActivation());
 
-      final Channel federationChannel =  rc.getChannel(CHANNEL_ID.FEDERATION.id, -1);
-      federationChannel.setHandler(new LocalChannelHandler(config, entry, channel0, acceptorUsed, rc));
+      rc.getChannel(CHANNEL_ID.FEDERATION.id, -1).setHandler(new FederationChannelHandler(acceptorUsed, rc));
 
       return entry;
    }
@@ -412,7 +414,44 @@ public class CoreProtocolManager implements ProtocolManager<Interceptor, ActiveM
                   }
                });
             }
-         } else if (packet.getType() == PacketImpl.FEDERATION_DOWNSTREAM_CONNECT) {
+         }
+      }
+
+      private Pair<TransportConfiguration, TransportConfiguration> getPair(
+         TransportConfiguration conn,
+         boolean isBackup) {
+         if (isBackup) {
+            return new Pair<>(null, conn);
+         }
+         return new Pair<>(conn, null);
+      }
+   }
+
+   private class FederationChannelHandler implements ChannelHandler {
+
+      private final Acceptor acceptorUsed;
+      private final CoreRemotingConnection rc;
+
+      private FederationChannelHandler(final Acceptor acceptorUsed, final CoreRemotingConnection rc) {
+         this.acceptorUsed = acceptorUsed;
+         this.rc = rc;
+      }
+
+      @Override
+      public void handlePacket(final Packet packet) {
+         if (packet.getType() == PacketImpl.FEDERATION_DOWNSTREAM_CONNECT) {
+            if (server.getSecurityStore().isSecurityEnabled()) {
+               if (rc.getSubject() == null) {
+                  ActiveMQServerLogger.LOGGER.federationDownstreamUnauthenticated(rc.getRemoteAddress());
+                  rc.destroy();
+                  return;
+               }
+               if (!server.getFederationManager().authorizeDownstreamDeployment(rc.getSubject())) {
+                  ActiveMQServerLogger.LOGGER.federationDownstreamUnauthorized(rc.getRemoteAddress(), SecurityManagerUtil.getUserFromSubject(rc.getSubject(), UserPrincipal.class));
+                  rc.destroy();
+                  return;
+               }
+            }
             //If we receive this packet then a remote broker is requesting us to create federated upstream connection
             //back to it which simulates a downstream connection
             final FederationDownstreamConnectMessage message = (FederationDownstreamConnectMessage) packet;
@@ -492,37 +531,37 @@ public class CoreProtocolManager implements ProtocolManager<Interceptor, ActiveM
             //Register close and failure listeners, if the initial downstream connection goes down then we
             //want to terminate the upstream connection
             rc.addCloseListener(() -> {
-               server.getFederationManager().undeploy(config.getName());
+               handleClose(config.getName(), rc.getRemoteAddress());
             });
 
             rc.addFailureListener(new FailureListener() {
                @Override
                public void connectionFailed(ActiveMQException exception, boolean failedOver) {
-                  server.getFederationManager().undeploy(config.getName());
+                  handleClose(config.getName(), rc.getRemoteAddress());
                }
 
                @Override
                public void connectionFailed(ActiveMQException exception, boolean failedOver,
                                             String scaleDownTargetNodeID) {
-                  server.getFederationManager().undeploy(config.getName());
+                  handleClose(config.getName(), rc.getRemoteAddress());
                }
             });
 
             try {
                server.getFederationManager().deploy(config);
+               String user = rc.getSubject() != null ? SecurityManagerUtil.getUserFromSubject(rc.getSubject(), UserPrincipal.class) : "anonymous";
+               ActiveMQServerLogger.LOGGER.federationDownstreamDeployedFromRemoteUser(config.getName(), user, rc.getRemoteAddress());
             } catch (Exception e) {
                logger.error("Error deploying federation", e);
             }
          }
       }
 
-      private Pair<TransportConfiguration, TransportConfiguration> getPair(
-         TransportConfiguration conn,
-         boolean isBackup) {
-         if (isBackup) {
-            return new Pair<>(null, conn);
+      private void handleClose(String configName, String remoteAddress) {
+         if (server.getFederationManager().undeploy(configName)) {
+            ActiveMQServerLogger.LOGGER.federationDownstreamConnectionClosed(remoteAddress, configName);
          }
-         return new Pair<>(conn, null);
       }
    }
 }
+
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQServerLogger.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQServerLogger.java
@@ -1593,4 +1593,16 @@ public interface ActiveMQServerLogger {
 
    @LogMessage(id = 224130, value = "The {} value {} will override the {} value {} since both are set.", level = LogMessage.Level.INFO)
    void configParamOverride(String overridingParam, Object overridingParamValue, String overriddenParam, Object overriddenParamValue);
+
+   @LogMessage(id = 224131, value = "Unable to process Core federation downstream request from {}. User is not authenticated. Closing connection.", level = LogMessage.Level.WARN)
+   void federationDownstreamUnauthenticated(String remoteAddress);
+
+   @LogMessage(id = 224132, value = "Unable to process Core federation downstream request from {}. User {} is not authorized. Closing connection.", level = LogMessage.Level.WARN)
+   void federationDownstreamUnauthorized(String remoteAddress, String user);
+
+   @LogMessage(id = 224133, value = "Core federation downstream {} has been deployed. Request sent by user {} from {}.", level = LogMessage.Level.INFO)
+   void federationDownstreamDeployedFromRemoteUser(String name, String userFromSubject, String remoteAddress);
+
+   @LogMessage(id = 224134, value = "Connection from {} closed. Undeployed Core federation {}.", level = LogMessage.Level.INFO)
+   void federationDownstreamConnectionClosed(String remoteAddress, String name);
 }
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/federation/FederationManager.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/federation/FederationManager.java
@@ -16,13 +16,16 @@
  */
 package org.apache.activemq.artemis.core.server.federation;
 
+import javax.security.auth.Subject;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import org.apache.activemq.artemis.api.core.ActiveMQException;
 import org.apache.activemq.artemis.core.server.ActiveMQComponent;
 import org.apache.activemq.artemis.core.server.ActiveMQServer;
 import org.apache.activemq.artemis.core.config.FederationConfiguration;
+import org.apache.activemq.artemis.spi.core.security.jaas.RolePrincipal;
 
 public class FederationManager implements ActiveMQComponent {
 
@@ -30,6 +33,7 @@ public class FederationManager implements ActiveMQComponent {
 
    private Map<String, Federation> federations = new HashMap<>();
    private State state;
+   private List<String> downstreamAuthorization;
 
    enum State {
       STOPPED,
@@ -47,6 +51,7 @@ public class FederationManager implements ActiveMQComponent {
 
    public FederationManager(final ActiveMQServer server) {
       this.server = server;
+      this.downstreamAuthorization = server.getConfiguration().getFederationDownstreamAuthorization();
    }
 
    @Override
@@ -90,8 +95,9 @@ public class FederationManager implements ActiveMQComponent {
       Federation federation = federations.remove(name);
       if (federation != null) {
          federation.stop();
+         return true;
       }
-      return true;
+      return false;
    }
 
 
@@ -129,4 +135,15 @@ public class FederationManager implements ActiveMQComponent {
       server.unRegisterBrokerPlugin(federatedAbstract);
    }
 
+   public boolean authorizeDownstreamDeployment(Subject subject) {
+      if (!server.getSecurityStore().isSecurityEnabled()) {
+         return true;
+      }
+      for (RolePrincipal role : subject.getPrincipals(RolePrincipal.class)) {
+         if (downstreamAuthorization.contains(role.getName())) {
+            return true;
+         }
+      }
+      return false;
+   }
 }
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/utils/SecurityManagerUtil.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/utils/SecurityManagerUtil.java
@@ -32,6 +32,15 @@ import org.apache.activemq.artemis.spi.core.security.jaas.RolePrincipal;
 public class SecurityManagerUtil {
 
    private static final String WILDCARD = "*";
+   public static String getUserFromSubject(Subject subject, Class<? extends Principal> principalClass) {
+      if (subject != null) {
+         for (Principal candidate : subject.getPrincipals(principalClass)) {
+            return candidate.getName();
+         }
+      }
+      return null;
+   }
+
 
    public static Set<RolePrincipal> getPrincipalsInRole(final CheckType checkType, final Set<Role> roles, final String rolePrincipalClass) {
       Set principals = new HashSet<>();
--- a/artemis-server/src/main/resources/schema/artemis-configuration.xsd
+++ b/artemis-server/src/main/resources/schema/artemis-configuration.xsd
@@ -1489,6 +1489,7 @@
          <xsd:sequence>
             <xsd:element name="federation" type="federationType" maxOccurs="unbounded" minOccurs="0"/>
          </xsd:sequence>
+         <xsd:attribute name="downstream-authorization" type="xsd:string" use="optional" />
          <xsd:attributeGroup ref="xml:specialAttrs"/>
       </xsd:complexType>
    </xsd:element>
