--- a/src/wheel/cli/unpack.py
+++ b/src/wheel/cli/unpack.py
@@ -19,12 +19,12 @@
         destination = Path(dest) / namever
         print(f"Unpacking to: {destination}...", end="", flush=True)
         for zinfo in wf.filelist:
-            wf.extract(zinfo, destination)
+            target_path = Path(wf.extract(zinfo, destination))

             # Set permissions to the same values as they were set in the archive
             # We have to do this manually due to
             # https://github.com/python/cpython/issues/59999
             permissions = zinfo.external_attr >> 16 & 0o777
-            destination.joinpath(zinfo.filename).chmod(permissions)
+            target_path.chmod(permissions)

     print("OK")
