--- a/src/utils.js
+++ b/src/utils.js
@@ -1,28 +1,8 @@
 // Utility functions

-var PREFIXES = 'Webkit Moz O ms'.split(' ');
-var FLOAT_COMPARISON_EPSILON = 0.001;
-
-// Copy all attributes from source object to destination object.
-// destination object is mutated.
-function extend(destination, source, recursive) {
-    destination = destination || {};
-    source = source || {};
-    recursive = recursive || false;
-
-    for (var attrName in source) {
-        if (source.hasOwnProperty(attrName)) {
-            var destVal = destination[attrName];
-            var sourceVal = source[attrName];
-            if (recursive && isObject(destVal) && isObject(sourceVal)) {
-                destination[attrName] = extend(destVal, sourceVal, recursive);
-            } else {
-                destination[attrName] = sourceVal;
-            }
-        }
-    }
-
-    return destination;
-}
+var merge = require('lodash.merge');
+
+var PREFIXES = 'Webkit Moz O ms'.split(' ');
+var FLOAT_COMPARISON_EPSILON = 0.001;

 // Renders templates with given variables. Variables must be surrounded with
@@ -123,7 +103,7 @@
 }

 module.exports = {
-    extend: extend,
+    extend: merge,
     render: render,
     setStyle: setStyle,
     setStyles: setStyles,
--- a/package.json
+++ b/package.json
@@ -4,6 +4,7 @@
   "description": "Responsive and slick progress bars with animated SVG paths",
   "main": "src/main.js",
   "dependencies": {
+    "lodash.merge": "^4.6.2",
     "shifty": "^2.1.2"
   },
   "devDependencies": {
