--- a/dist/http-proxy-middleware.js
+++ b/dist/http-proxy-middleware.js
@@ -73,8 +73,14 @@ class HttpProxyMiddleware {
          * @return {Boolean}
          */
         this.shouldProxy = (context, req) => {
-            const path = req.originalUrl || req.url;
-            return contextMatcher.match(context, path, req);
+            try {
+                const path = req.originalUrl || req.url;
+                return contextMatcher.match(context, path, req);
+            }
+            catch (error) {
+                this.logger.error(error);
+                return false;
+            }
         };
         /**
          * Apply option.router and option.pathRewrite
--- a/package.json
+++ b/package.json
@@ -15,8 +15,8 @@
     "eslint:fix": "yarn eslint --fix",
     "prettier": "prettier --list-different \"**/*.{js,ts,md,yml,json,html}\"",
     "prettier:fix": "prettier --write \"**/*.{js,ts,md,yml,json,html}\"",
-    "prebuild": "yarn clean",
-    "build": "tsc",
+    "prebuild": "echo skip-clean-preserve-patched-dist",
+    "build": "echo skip-build-preserve-patched-dist",
     "pretest": "yarn build",
     "test": "jest",
     "precoverage": "yarn build",
