diff --git a/providers/smtp/src/main/java/org/eclipse/angus/mail/smtp/SMTPTransport.java b/providers/smtp/src/main/java/org/eclipse/angus/mail/smtp/SMTPTransport.java
--- a/providers/smtp/src/main/java/org/eclipse/angus/mail/smtp/SMTPTransport.java
+++ b/providers/smtp/src/main/java/org/eclipse/angus/mail/smtp/SMTPTransport.java
@@ -2455,10 +2455,21 @@
  //logger.fine("SENT: " + new String(cmdBytes, 0));
 
  try {
+ validateCommand(cmdBytes);
  serverOutput.write(cmdBytes);
  serverOutput.write(CRLF);
  serverOutput.flush();
- } catch (IOException ex) {
+ } catch (IOException | RuntimeException ex) {
  throw new MessagingException("Can't send command to SMTP host", ex);
+ }
+ }
+
+ private void validateCommand(byte[] cmdBytes) throws MessagingException {
+ final byte CR = '\r';
+ final byte LF = '\n';
+ for (byte b : cmdBytes) {
+ if (b == LF || b == CR) {
+ throw new IllegalArgumentException("Command contains illegal character: " + String.format("0x%02x",b));
+ }
  }
  }
