--- a/angular.js
+++ b/angular.js
@@ -22624,11 +22624,15 @@
     // If the currency symbol is empty, trim whitespace around the symbol
     var currencySymbolRe = !currencySymbol ? /\s*\u00A4\s*/g : /\u00A4/g;
 
     // if null or undefined pass it through
-    return (amount == null)
-        ? amount
-        : formatNumber(amount, formats.PATTERNS[1], formats.GROUP_SEP, formats.DECIMAL_SEP, fractionSize).
-            replace(currencySymbolRe, currencySymbol);
+    if (amount == null) {
+      return amount;
+    }
+
+    var formattedNumber = formatNumber(amount, formats.PATTERNS[1], formats.GROUP_SEP, formats.DECIMAL_SEP, fractionSize);
+    return formattedNumber.indexOf('\u00A4') !== -1
+        ? formattedNumber.replace(currencySymbolRe, currencySymbol)
+        : formattedNumber;
   };
 }
 

--- a/angular.js
+++ b/angular.js
@@ -9410,43 +9410,17 @@
-      // Such values are a bit too complex to handle automatically inside $sce.
-      // Instead, we sanitize each of the URIs individually, which works, even dynamically.
-
-      // It's not possible to work around this using `$sce.trustAsMediaUrl`.
-      // If you want to programmatically set explicitly trusted unsafe URLs, you should use
-      // `$sce.trustAsHtml` on the whole `img` tag and inject it into the DOM using the
-      // `ng-bind-html` directive.
-
-      var result = '';
-
-      // first check if there are spaces because it's not the same pattern
-      var trimmedSrcset = trim(value);
-      //                (   999x   ,|   999w   ,|   ,|,   )
-      var srcPattern = /(\s+\d+x\s*,|\s+\d+w\s*,|\s+,|,\s+)/;
-      var pattern = /\s/.test(trimmedSrcset) ? srcPattern : /(,)/;
-
-      // split srcset into tuple of uri and descriptor except for the last item
-      var rawUris = trimmedSrcset.split(pattern);
-
-      // for each tuples
-      var nbrUrisWith2parts = Math.floor(rawUris.length / 2);
-      for (var i = 0; i < nbrUrisWith2parts; i++) {
-        var innerIdx = i * 2;
-        // sanitize the uri
-        result += $sce.getTrustedMediaUrl(trim(rawUris[innerIdx]));
-        // add the descriptor
-        result += ' ' + trim(rawUris[innerIdx + 1]);
+      var srcSetValues = value.split(',');
+      var sanitizedSrcSet = [];
+      for (var i = 0; i < srcSetValues.length; i++) {
+        var srcSetValue = trim(srcSetValues[i]);
+        if (!srcSetValue) {
+          continue;
+        }
+
+        var srcSetParts = srcSetValue.split(/\s+/);
+        var uri = $sce.getTrustedMediaUrl(srcSetParts[0]);
+        sanitizedSrcSet.push(uri + (srcSetParts[1] ? ' ' + srcSetParts[1] : ''));
       }
 
-      // split the last item into uri and descriptor
-      var lastTuple = trim(rawUris[i * 2]).split(/\s/);
-
-      // sanitize the last uri
-      result += $sce.getTrustedMediaUrl(trim(lastTuple[0]));
-
-      // and add the last descriptor if any
-      if (lastTuple.length === 2) {
-        result += (' ' + trim(lastTuple[1]));
-      }
-      return result;
+      return sanitizedSrcSet.join(',');
     }
 
 
