diff --git a/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImpl.java b/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImpl.java
index 2a532d1..a25a379 100644
--- a/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImpl.java
+++ b/artemis-server/src/main/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImpl.java
@@ -47,6 +47,7 @@ import java.util.HashSet;
 import java.util.Iterator;
 import java.util.LinkedHashMap;
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Properties;
 import java.util.Set;
@@ -137,6 +138,8 @@ import org.apache.commons.beanutils.expression.Resolver;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import static org.apache.activemq.artemis.utils.PasswordMaskingUtil.isEncMasked;
+
 public class ConfigurationImpl implements Configuration, Serializable {
 
    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
@@ -145,6 +148,8 @@ public class ConfigurationImpl implements Configuration, Serializable {
 
    public static final String PROPERTY_CLASS_SUFFIX = ".class";
 
+   public static final String REDACTED = "**redacted**";
+
    private static final int DEFAULT_JMS_MESSAGE_SIZE = 1864;
 
    private static final int RANGE_SIZE_MIN = 0;
@@ -455,6 +460,7 @@ public class ConfigurationImpl implements Configuration, Serializable {
    private File artemisInstance;
    private transient JsonObject jsonStatus = JsonLoader.createObjectBuilder().build();
    private transient Checksum transientChecksum = null;
+   private final Set<String> keysToRedact = new HashSet<>();
 
    private JsonObject getJsonStatus() {
       if (jsonStatus == null) {
@@ -625,9 +631,13 @@ public class ConfigurationImpl implements Configuration, Serializable {
             checksum.update(value.getBytes(StandardCharsets.UTF_8));
 
             value = XMLUtil.replaceSystemPropsInString(value);
-            value = PasswordMaskingUtil.resolveMask(isMaskPassword(), value, getPasswordCodec());
+            final boolean masked = isEncMasked(value);
+            if (masked) {
+               value = PasswordMaskingUtil.resolveMask(null, value, getPasswordCodec());
+               keysToRedact.add(key);
+            }
             key = XMLUtil.replaceSystemPropsInString(key);
-            logger.debug("Property config, {}={}", key, value);
+            logger.debug("Property config, {}={}", key, (masked || shouldRedact(key)) ? REDACTED : value);
             beanProperties.put(key, value);
          }
          alder32Hash = checksum.getValue();
@@ -639,6 +649,10 @@ public class ConfigurationImpl implements Configuration, Serializable {
       }
    }
 
+   private boolean shouldRedact(String name) {
+      return name.toUpperCase(Locale.ENGLISH).contains("PASSWORD") || keysToRedact.contains(name);
+   }
+
    public void populateWithProperties(final Object target, final String propsId, Map<String, Object> beanProperties) throws InvocationTargetException, IllegalAccessException {
       CollectionAutoFillPropertiesUtil autoFillCollections = new CollectionAutoFillPropertiesUtil(getBrokerPropertiesRemoveValue(beanProperties));
       BeanUtilsBean beanUtils = new BeanUtilsBean(new ConvertUtilsBean(), autoFillCollections) {
