diff --git a/src/server-interceptors.ts b/src/server-interceptors.ts
--- a/src/server-interceptors.ts
+++ b/src/server-interceptors.ts
@@ -19,7 +19,7 @@ import { PartialStatusObject } from './call-interface';
 import { ServerMethodDefinition } from './make-client';
 import { Metadata } from './metadata';
 import { ChannelOptions } from './channel-options';
-import { Handler, ServerErrorResponse } from './server-call';
+import { Handler } from './server-call';
 import { Deadline } from './deadline';
 import {
   DEFAULT_MAX_RECEIVE_MESSAGE_LENGTH,
@@ -518,14 +518,6 @@ export class BaseServerInterceptingCall
     private readonly handler: Handler<any, any>,
     options: ChannelOptions
   ) {
-    this.stream.once('error', (err: ServerErrorResponse) => {
-      /* We need an error handler to avoid uncaught error event exceptions, but
-       * there is nothing we can reasonably do here. Any error event should
-       * have a corresponding close event, which handles emitting the cancelled
-       * event. And the stream is now in a bad state, so we can't reasonably
-       * expect to be able to send an error over it. */
-    });
-
     this.stream.once('close', () => {
       trace(
         'Request to method ' +
diff --git a/src/server.ts b/src/server.ts
--- a/src/server.ts
+++ b/src/server.ts
@@ -1332,6 +1332,13 @@ export class Server {
     stream: http2.ServerHttp2Stream,
     headers: http2.IncomingHttpHeaders
   ) {
+    stream.once('error', (err: ServerErrorResponse) => {
+      /* We need an error handler to avoid uncaught error event exceptions, but
+       * there is nothing we can reasonably do here. Any error event should
+       * have a corresponding close event, which handles emitting the cancelled
+       * event. And the stream is now in a bad state, so we can't reasonably
+       * expect to be able to send an error over it. */
+    });
     // for handling idle timeout
     this.onStreamOpened(stream);
 
@@ -1416,6 +1423,13 @@ export class Server {
     stream: http2.ServerHttp2Stream,
     headers: http2.IncomingHttpHeaders
   ) {
+    stream.once('error', (err: ServerErrorResponse) => {
+      /* We need an error handler to avoid uncaught error event exceptions, but
+       * there is nothing we can reasonably do here. Any error event should
+       * have a corresponding close event, which handles emitting the cancelled
+       * event. And the stream is now in a bad state, so we can't reasonably
+       * expect to be able to send an error over it. */
+    });
     // for handling idle timeout
     this.onStreamOpened(stream);
 

diff --git a/src/compression-filter.ts b/src/compression-filter.ts
index e4428a1f..5277d8c1 100644
--- a/src/compression-filter.ts
+++ b/src/compression-filter.ts
@@ -119,6 +119,12 @@ class DeflateHandler extends CompressionHandler {
       let totalLength = 0;
       const messageParts: Buffer[] = [];
       const decompresser = zlib.createInflate();
+      decompresser.on('error', (error: Error) => {
+        reject({
+          code: Status.INTERNAL,
+          details: 'Failed to decompress deflate-encoded message'
+        });
+      });
       decompresser.on('data', (chunk: Buffer) => {
         messageParts.push(chunk);
         totalLength += chunk.byteLength;
@@ -161,6 +167,12 @@ class GzipHandler extends CompressionHandler {
       let totalLength = 0;
       const messageParts: Buffer[] = [];
       const decompresser = zlib.createGunzip();
+      decompresser.on('error', (error: Error) => {
+        reject({
+          code: Status.INTERNAL,
+          details: 'Failed to decompress gzip-encoded message'
+        });
+      });
       decompresser.on('data', (chunk: Buffer) => {
         messageParts.push(chunk);
         totalLength += chunk.byteLength;
diff --git a/src/server-interceptors.ts b/src/server-interceptors.ts
index 518b30ea..d98a73e2 100644
--- a/src/server-interceptors.ts
+++ b/src/server-interceptors.ts
@@ -699,6 +699,12 @@ export class BaseServerInterceptingCall
       return new Promise((resolve, reject) => {
         let totalLength = 0
         const messageParts: Buffer[] = [];
+        decompresser.on('error', (error: Error) => {
+          reject({
+            code: Status.INTERNAL,
+            details: 'Failed to decompress message'
+          });
+        });
         decompresser.on('data', (chunk: Buffer) => {
           messageParts.push(chunk);
           totalLength += chunk.byteLength;
