--- a/lib/dot/pattern.jst
+++ b/lib/dot/pattern.jst
@@ -3,12 +3,22 @@
 {{# def.setupKeyword }}
 {{# def.$data }}
 
-{{
-  var $regexp = $isData
-                ? '(new RegExp(' + $schemaValue + '))'
-                : it.usePattern($schema);
-}}
-
-if ({{# def.$dataNotType:'string' }} !{{=$regexp}}.test({{=$data}}) ) {
+{{? $isData }}
+{{
+  var $patternValid = 'patternValid' + $lvl;
+}}
+var {{=$patternValid}};
+try {
+  {{=$patternValid}} = (new RegExp({{=$schemaValue}})).test({{=$data}});
+} catch(e) {
+  {{=$patternValid}} = false;
+}
+if ({{# def.$dataNotType:'string' }} !{{=$patternValid}} ) {
+{{??}}
+{{
+  var $regexp = it.usePattern($schema);
+}}
+if ({{# def.$dataNotType:'string' }} !{{=$regexp}}.test({{=$data}}) ) {
+{{?}}
   {{# def.error:'pattern' }}
 } {{? $breakOnError }} else { {{?}}
--- a/lib/dotjs/pattern.js
+++ b/lib/dotjs/pattern.js
@@ -16,12 +16,18 @@
   } else {
     $schemaValue = $schema;
   }
-  var $regexp = $isData ? '(new RegExp(' + $schemaValue + '))' : it.usePattern($schema);
-  out += 'if ( ';
   if ($isData) {
-    out += ' (' + ($schemaValue) + ' !== undefined && typeof ' + ($schemaValue) + ' != \'string\') || ';
+    var $patternValid = 'patternValid' + $lvl;
+    out += ' var ' + $patternValid + '; ';
+    out += ' try { ' + $patternValid + ' = (new RegExp(' + $schemaValue + ')).test(' + $data + '); } catch(e) { ' + $patternValid + ' = false; } ';
+    out += 'if ( ';
+    out += ' (' + ($schemaValue) + ' !== undefined && typeof ' + ($schemaValue) + ' != \'string\') || ';
+    out += ' !' + $patternValid + ' ) {   ';
+  } else {
+    var $regexp = it.usePattern($schema);
+    out += 'if ( ';
+    out += ' !' + ($regexp) + '.test(' + ($data) + ') ) {   ';
   }
-  out += ' !' + ($regexp) + '.test(' + ($data) + ') ) {   ';
   var $$outStack = $$outStack || [];
   $$outStack.push(out);
   out = ''; /* istanbul ignore else */
