--- a/coders/jxl.c
+++ b/coders/jxl.c
@@ -654,7 +654,7 @@
                       ThrowJXLReaderException(ResourceLimitError,MemoryAllocationFailed,image);
                   }
                 grayscale=MagickTrue;
-                pixel_format.num_channels=1;
+                pixel_format.num_channels=image->matte ? 2 : 1;
                 pixel_format.data_type=(basic_info.bits_per_sample <= 8 ? JXL_TYPE_UINT8 :
                                         (basic_info.bits_per_sample <= 16 ? JXL_TYPE_UINT16 :
                                          JXL_TYPE_FLOAT));
