--- a/coders/jxl.c
+++ b/coders/jxl.c
@@ -636,6 +636,9 @@
               image->matte=MagickTrue;

             image->orientation=convert_orientation(basic_info.orientation);
+
+            if (CheckImagePixelLimits(image, exception) != MagickPass)
+              ThrowJXLReaderException(ResourceLimitError,ImagePixelLimitExceeded,image);

             pixel_format.endianness=JXL_NATIVE_ENDIAN;
             pixel_format.align=0;
--- a/coders/jxl.c
+++ b/coders/jxl.c
@@ -654,7 +654,7 @@
                       ThrowJXLReaderException(ResourceLimitError,MemoryAllocationFailed,image);
                   }
                 grayscale=MagickTrue;
-                pixel_format.num_channels=1;
+                pixel_format.num_channels=image->matte ? 2 : 1;
                 pixel_format.data_type=(basic_info.bits_per_sample <= 8 ? JXL_TYPE_UINT8 :
                                         (basic_info.bits_per_sample <= 16 ? JXL_TYPE_UINT16 :
                                          JXL_TYPE_FLOAT));
