--- a/src/index.ts
+++ b/src/index.ts
@@ -890,63 +890,64 @@
   #pool : ThreadPool;
 
   constructor (options : Options = {}) {
-    super({ ...options, name: 'Piscina' });
+    super(Object.assign(Object.create(null), options, { name: 'Piscina' }) as Options);
+    const opts: Options = Object.assign(Object.create(null), options) as Options;
 
-    if (typeof options.filename !== 'string' && options.filename != null) {
+    if (typeof opts.filename !== 'string' && opts.filename != null) {
       throw new TypeError('options.filename must be a string or null');
     }
-    if (typeof options.name !== 'string' && options.name != null) {
+    if (typeof opts.name !== 'string' && opts.name != null) {
       throw new TypeError('options.name must be a string or null');
     }
-    if (options.minThreads !== undefined &&
-        (typeof options.minThreads !== 'number' || options.minThreads < 0)) {
+    if (opts.minThreads !== undefined &&
+        (typeof opts.minThreads !== 'number' || opts.minThreads < 0)) {
       throw new TypeError('options.minThreads must be a non-negative integer');
     }
-    if (options.maxThreads !== undefined &&
-        (typeof options.maxThreads !== 'number' || options.maxThreads < 1)) {
+    if (opts.maxThreads !== undefined &&
+        (typeof opts.maxThreads !== 'number' || opts.maxThreads < 1)) {
       throw new TypeError('options.maxThreads must be a positive integer');
     }
-    if (options.minThreads !== undefined && options.maxThreads !== undefined &&
-        options.minThreads > options.maxThreads) {
+    if (opts.minThreads !== undefined && opts.maxThreads !== undefined &&
+        opts.minThreads > opts.maxThreads) {
       throw new RangeError('options.minThreads and options.maxThreads must not conflict');
     }
-    if (options.idleTimeout !== undefined &&
-        (typeof options.idleTimeout !== 'number' || options.idleTimeout < 0)) {
+    if (opts.idleTimeout !== undefined &&
+        (typeof opts.idleTimeout !== 'number' || opts.idleTimeout < 0)) {
       throw new TypeError('options.idleTimeout must be a non-negative integer');
     }
-    if (options.maxQueue !== undefined &&
-        options.maxQueue !== 'auto' &&
-        (typeof options.maxQueue !== 'number' || options.maxQueue < 0)) {
+    if (opts.maxQueue !== undefined &&
+        opts.maxQueue !== 'auto' &&
+        (typeof opts.maxQueue !== 'number' || opts.maxQueue < 0)) {
       throw new TypeError('options.maxQueue must be a non-negative integer');
     }
-    if (options.concurrentTasksPerWorker !== undefined &&
-        (typeof options.concurrentTasksPerWorker !== 'number' ||
-         options.concurrentTasksPerWorker < 1)) {
+    if (opts.concurrentTasksPerWorker !== undefined &&
+        (typeof opts.concurrentTasksPerWorker !== 'number' ||
+         opts.concurrentTasksPerWorker < 1)) {
       throw new TypeError(
         'options.concurrentTasksPerWorker must be a positive integer');
     }
-    if (options.useAtomics !== undefined &&
-        typeof options.useAtomics !== 'boolean') {
+    if (opts.useAtomics !== undefined &&
+        typeof opts.useAtomics !== 'boolean') {
       throw new TypeError('options.useAtomics must be a boolean value');
     }
-    if (options.resourceLimits !== undefined &&
-        (typeof options.resourceLimits !== 'object' ||
-         options.resourceLimits === null)) {
+    if (opts.resourceLimits !== undefined &&
+        (typeof opts.resourceLimits !== 'object' ||
+         opts.resourceLimits === null)) {
       throw new TypeError('options.resourceLimits must be an object');
     }
-    if (options.taskQueue !== undefined && !isTaskQueue(options.taskQueue)) {
+    if (opts.taskQueue !== undefined && !isTaskQueue(opts.taskQueue)) {
       throw new TypeError('options.taskQueue must be a TaskQueue object');
     }
-    if (options.niceIncrement !== undefined &&
-        (typeof options.niceIncrement !== 'number' || options.niceIncrement < 0)) {
+    if (opts.niceIncrement !== undefined &&
+        (typeof opts.niceIncrement !== 'number' || opts.niceIncrement < 0)) {
       throw new TypeError('options.niceIncrement must be a non-negative integer');
     }
-    if (options.trackUnmanagedFds !== undefined &&
-        typeof options.trackUnmanagedFds !== 'boolean') {
+    if (opts.trackUnmanagedFds !== undefined &&
+        typeof opts.trackUnmanagedFds !== 'boolean') {
       throw new TypeError('options.trackUnmanagedFds must be a boolean value');
     }
 
-    this.#pool = new ThreadPool(this, options);
+    this.#pool = new ThreadPool(this, opts);
   }
 
   /** @deprecated Use run(task, options) instead **/
@@ -1002,12 +1003,16 @@
       return Promise.reject(
         new TypeError('options must be an object'));
     }
+
     const {
       transferList,
-      filename,
-      name,
       signal
     } = options;
+    const filename = Object.prototype.hasOwnProperty.call(options, 'filename')
+      ? (options as any).filename : null;
+    const name = Object.prototype.hasOwnProperty.call(options, 'name')
+      ? (options as any).name : null;
+
     if (transferList !== undefined && !Array.isArray(transferList)) {
       return Promise.reject(
         new TypeError('transferList argument must be an Array'));
--- a/test/option-validation.ts
+++ b/test/option-validation.ts
@@ -1,5 +1,6 @@
 import Piscina from '..';
 import { test } from 'tap';
+import { resolve } from 'path';
 
 test('filename cannot be non-null/non-string', async ({ throws }) => {
   throws(() => new Piscina(({
@@ -7,10 +8,32 @@
   }) as any), /options.filename must be a string or null/);
 });
 
+test('filename is not tampered', async ({ rejects }) => {
+  const polluted = { filename: 'polluted' };
+  const opts1 = {};
+  const opts2 = { '__proto__': polluted };
+
+  Object.setPrototypeOf(opts1, polluted);
+  rejects(new Piscina(opts2 as any).run({}), /filename must be provided/);
+  rejects(new Piscina(opts1 as any).run({}), /filename must be provided/);
+});
+
 test('name cannot be non-null/non-string', async ({ throws }) => {
   throws(() => new Piscina(({
     name: 12
   }) as any), /options.name must be a string or null/);
+});
+
+test('name is not tampered', async ({ equal }) => {
+  const polluted = { name: 'b' };
+  const opts1 = { filename: resolve(__dirname, 'fixtures/multiple.js') };
+  const opts2 = { '__proto__': polluted, filename: resolve(__dirname, 'fixtures/multiple.js') };
+
+  const res1 = await new Piscina(opts1 as any).run({});
+  const res2 = await new Piscina(opts2 as any).run({});
+
+  equal(res1, 'a');
+  equal(res2, 'a');
 });
 
 test('minThreads must be non-negative integer', async ({ throws }) => {
