--- a/index.js
+++ b/index.js
@@ -10,7 +10,7 @@
             return '"' + s.replace(/(["\\$`!])/g, '\\$1') + '"';
         }
         else {
-            s = String(s).replace(/([A-z]:)?([#!"$&'()*,:;<=>?@\[\\\]^`{|}])/g, '$1\\$2');
+            s = String(s).replace(/([A-Za-z]:)?([#!"$&'()*,:;<=>?@\[\\\]^`{|}])/g, '$1\\$2');

             // unescape paths in windows
             s = s.replace(/\\\\/g, '\\')
--- a/test/quote.js
+++ b/test/quote.js
@@ -39,3 +39,9 @@

     t.end()
 })
+
+test("chars for windows paths don't break out", function (t) {
+    var x = '`:\\a\\b'
+    t.equal(quote([x]), '\\`\\:\\\\a\\\\b')
+    t.end()
+})
