--- a/src/java/org/apache/commons/collections/functors/FunctorUtils.java
+++ b/src/java/org/apache/commons/collections/functors/FunctorUtils.java
@@ -15,6 +15,8 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 import java.util.Collection;
 import java.util.Iterator;
 
@@ -24,24 +26,28 @@
 
 /**
  * Internal utilities for functors.
- * 
+ *
  * @since Commons Collections 3.0
  * @version $Revision: 1.7 $ $Date: 2004/05/16 11:47:38 $
  *
  * @author Stephen Colebourne
  */
 class FunctorUtils {
-    
+
+    /** System property key to enable unsafe serialization */
+    final static String UNSAFE_SERIALIZABLE_PROPERTY
+        = "org.apache.commons.collections.enableUnsafeSerialization";
+
     /**
      * Restricted constructor.
      */
     private FunctorUtils() {
         super();
     }
-    
+
     /**
      * Clone the predicates to ensure that the internal reference can't be messed with.
-     * 
+     *
      * @param predicates  the predicates to copy
      * @return the cloned predicates
      */
@@ -51,10 +57,10 @@
         }
         return (Predicate[]) predicates.clone();
     }
-    
+
     /**
      * Validate the predicates to ensure that all is well.
-     * 
+     *
      * @param predicates  the predicates to validate
      */
     static void validate(Predicate[] predicates) {
@@ -67,10 +73,10 @@
             }
         }
     }
-    
+
     /**
      * Validate the predicates to ensure that all is well.
-     * 
+     *
      * @param predicates  the predicates to validate
      */
     static void validateMin2(Predicate[] predicates) {
@@ -90,7 +96,7 @@
 
     /**
      * Validate the predicates to ensure that all is well.
-     * 
+     *
      * @param predicates  the predicates to validate
      * @return predicate array
      */
@@ -114,10 +120,10 @@
         }
         return preds;
     }
-    
+
     /**
      * Clone the closures to ensure that the internal reference can't be messed with.
-     * 
+     *
      * @param closures  the closures to copy
      * @return the cloned closures
      */
@@ -127,10 +133,10 @@
         }
         return (Closure[]) closures.clone();
     }
-    
+
     /**
      * Validate the closures to ensure that all is well.
-     * 
+     *
      * @param closures  the closures to validate
      */
     static void validate(Closure[] closures) {
@@ -146,7 +152,7 @@
 
     /**
      * Copy method
-     * 
+     *
      * @param transformers  the transformers to copy
      * @return a clone of the transformers
      */
@@ -156,10 +162,10 @@
         }
         return (Transformer[]) transformers.clone();
     }
-    
+
     /**
      * Validate method
-     * 
+     *
      * @param transformers  the transformers to validate
      */
     static void validate(Transformer[] transformers) {
@@ -171,6 +177,35 @@
                 throw new IllegalArgumentException(
                     "The transformer array must not contain a null transformer, index " + i + " was null");
             }
+        }
+    }
+
+    /**
+     * Package-private helper method to check if serialization support is
+     * enabled for unsafe classes.
+     *
+     * @param clazz  the clazz to check for serialization support
+     * @throws UnsupportedOperationException if unsafe serialization is disabled
+     */
+    static void checkUnsafeSerialization(Class clazz) {
+        String unsafeSerializableProperty;
+
+        try {
+            unsafeSerializableProperty =
+                (String) AccessController.doPrivileged(new PrivilegedAction() {
+                    public Object run() {
+                        return System.getProperty(UNSAFE_SERIALIZABLE_PROPERTY);
+                    }
+                });
+        } catch (SecurityException ex) {
+            unsafeSerializableProperty = null;
+        }
+
+        if (!"true".equalsIgnoreCase(unsafeSerializableProperty)) {
+            throw new UnsupportedOperationException(
+                    "Serialization support for " + clazz.getName() + " is disabled for security reasons. " +
+                    "To enable it set system property '" + UNSAFE_SERIALIZABLE_PROPERTY + "' to 'true', " +
+                    "but you must ensure that your application does not de-serialize objects from untrusted sources.");
         }
     }
 
--- a/src/java/org/apache/commons/collections/functors/InvokerTransformer.java
+++ b/src/java/org/apache/commons/collections/functors/InvokerTransformer.java
@@ -15,6 +15,9 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
@@ -24,7 +27,17 @@
 
 /**
  * Transformer implementation that creates a new object instance by reflection.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code InvokerTransformer}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision: 1.7 $ $Date: 2004/05/26 21:44:05 $
  *
@@ -34,7 +47,7 @@
 
     /** The serial version */
     static final long serialVersionUID = -8653385846894047688L;
-    
+
     /** The method name to call */
     private final String iMethodName;
     /** The array of reflection parameter types */
@@ -44,7 +57,7 @@
 
     /**
      * Gets an instance of this transformer calling a specific method with no arguments.
-     * 
+     *
      * @param methodName  the method name to call
      * @return an invoker transformer
      * @since Commons Collections 3.1
@@ -58,7 +71,7 @@
 
     /**
      * Gets an instance of this transformer calling a specific method with specific values.
-     * 
+     *
      * @param methodName  the method name to call
      * @param paramTypes  the parameter types of the method
      * @param args  the arguments to pass to the method
@@ -84,7 +97,7 @@
 
     /**
      * Constructor for no arg instance.
-     * 
+     *
      * @param methodName  the method to call
      */
     private InvokerTransformer(String methodName) {
@@ -97,7 +110,7 @@
     /**
      * Constructor that performs no validation.
      * Use <code>getInstance</code> if you want that.
-     * 
+     *
      * @param methodName  the method to call
      * @param paramTypes  the constructor parameter types, not cloned
      * @param args  the constructor arguments, not cloned
@@ -111,7 +124,7 @@
 
     /**
      * Transforms the input to result by invoking a method on the input.
-     * 
+     *
      * @param input  the input object to transform
      * @return the transformed result, null if null input
      */
@@ -123,7 +136,7 @@
             Class cls = input.getClass();
             Method method = cls.getMethod(iMethodName, iParamTypes);
             return method.invoke(input, iArgs);
-                
+
         } catch (NoSuchMethodException ex) {
             throw new FunctorException("InvokerTransformer: The method '" + iMethodName + "' on '" + input.getClass() + "' does not exist");
         } catch (IllegalAccessException ex) {
@@ -131,6 +144,24 @@
         } catch (InvocationTargetException ex) {
             throw new FunctorException("InvokerTransformer: The method '" + iMethodName + "' on '" + input.getClass() + "' threw an exception", ex);
         }
+    }
+
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(InvokerTransformer.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(InvokerTransformer.class);
+        is.defaultReadObject();
     }
 
 }
--- a/src/java/org/apache/commons/collections/functors/CloneTransformer.java
+++ b/src/java/org/apache/commons/collections/functors/CloneTransformer.java
@@ -15,6 +15,9 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 
 import org.apache.commons.collections.Transformer;
@@ -23,7 +26,17 @@
  * Transformer implementation that returns a clone of the input object.
  * <p>
  * Clone is performed using <code>PrototypeFactory.getInstance(input).create()</code>.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code CloneTransformer}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision: 1.6 $ $Date: 2004/05/16 11:36:31 $
  *
@@ -39,7 +52,7 @@
 
     /**
      * Factory returning the singleton instance.
-     * 
+     *
      * @return the singleton instance
      * @since Commons Collections 3.1
      */
@@ -56,7 +69,7 @@
 
     /**
      * Transforms the input to result by cloning it.
-     * 
+     *
      * @param input  the input object to transform
      * @return the transformed result
      */
@@ -65,6 +78,24 @@
             return null;
         }
         return PrototypeFactory.getInstance(input).create();
+    }
+
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(CloneTransformer.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(CloneTransformer.class);
+        is.defaultReadObject();
     }
 
 }
--- a/src/java/org/apache/commons/collections/functors/ForClosure.java
+++ b/src/java/org/apache/commons/collections/functors/ForClosure.java
@@ -15,13 +15,26 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 
 import org.apache.commons.collections.Closure;
 
 /**
  * Closure implementation that calls another closure n times, like a for loop.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code ForClosure}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision: 1.5 $ $Date: 2004/05/16 11:47:38 $
  *
@@ -42,7 +55,7 @@
      * <p>
      * A null closure or zero count returns the <code>NOPClosure</code>.
      * A count of one returns the specified closure.
-     * 
+     *
      * @param count  the number of times to execute the closure
      * @param closure  the closure to execute, not null
      * @return the <code>for</code> closure
@@ -60,7 +73,7 @@
     /**
      * Constructor that performs no validation.
      * Use <code>getInstance</code> if you want that.
-     * 
+     *
      * @param count  the number of times to execute the closure
      * @param closure  the closure to execute, not null
      */
@@ -72,7 +85,7 @@
 
     /**
      * Executes the closure <code>count</code> times.
-     * 
+     *
      * @param input  the input object
      */
     public void execute(Object input) {
@@ -83,7 +96,7 @@
 
     /**
      * Gets the closure.
-     * 
+     *
      * @return the closure
      * @since Commons Collections 3.1
      */
@@ -93,12 +106,30 @@
 
     /**
      * Gets the count.
-     * 
+     *
      * @return the count
      * @since Commons Collections 3.1
      */
     public int getCount() {
         return iCount;
+    }
+
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(ForClosure.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(ForClosure.class);
+        is.defaultReadObject();
     }
 
 }
--- a/src/java/org/apache/commons/collections/functors/InstantiateFactory.java
+++ b/src/java/org/apache/commons/collections/functors/InstantiateFactory.java
@@ -15,6 +15,9 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
@@ -24,7 +27,17 @@
 
 /**
  * Factory implementation that creates a new object instance by reflection.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code InstantiateFactory}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision: 1.6 $ $Date: 2004/05/16 11:47:38 $
  *
@@ -46,7 +59,7 @@
 
     /**
      * Factory method that performs validation.
-     * 
+     *
      * @param classToInstantiate  the class to instantiate, not null
      * @param paramTypes  the constructor parameter types
      * @param args  the constructor arguments
@@ -74,7 +87,7 @@
     /**
      * Constructor that performs no validation.
      * Use <code>getInstance</code> if you want that.
-     * 
+     *
      * @param classToInstantiate  the class to instantiate
      */
     public InstantiateFactory(Class classToInstantiate) {
@@ -88,7 +101,7 @@
     /**
      * Constructor that performs no validation.
      * Use <code>getInstance</code> if you want that.
-     * 
+     *
      * @param classToInstantiate  the class to instantiate
      * @param paramTypes  the constructor parameter types, not cloned
      * @param args  the constructor arguments, not cloned
@@ -115,7 +128,7 @@
 
     /**
      * Creates an object using the stored constructor.
-     * 
+     *
      * @return the new object
      */
     public Object create() {
@@ -135,5 +148,23 @@
             throw new FunctorException("InstantiateFactory: Constructor threw an exception", ex);
         }
     }
-    
+
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(InstantiateFactory.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(InstantiateFactory.class);
+        is.defaultReadObject();
+    }
+
 }
--- a/src/java/org/apache/commons/collections/functors/InstantiateTransformer.java
+++ b/src/java/org/apache/commons/collections/functors/InstantiateTransformer.java
@@ -15,6 +15,9 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
@@ -24,7 +27,17 @@
 
 /**
  * Transformer implementation that creates a new object instance by reflection.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code InstantiateTransformer}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision: 1.6 $ $Date: 2004/05/16 11:36:31 $
  *
@@ -34,7 +47,7 @@
 
     /** The serial version */
     static final long serialVersionUID = 3786388740793356347L;
-    
+
     /** Singleton instance that uses the no arg constructor */
     public static final Transformer NO_ARG_INSTANCE = new InstantiateTransformer();
 
@@ -45,7 +58,7 @@
 
     /**
      * Transformer method that performs validation.
-     * 
+     *
      * @param paramTypes  the constructor parameter types
      * @param args  the constructor arguments
      * @return an instantiate transformer
@@ -78,7 +91,7 @@
     /**
      * Constructor that performs no validation.
      * Use <code>getInstance</code> if you want that.
-     * 
+     *
      * @param paramTypes  the constructor parameter types, not cloned
      * @param args  the constructor arguments, not cloned
      */
@@ -90,7 +103,7 @@
 
     /**
      * Transforms the input Class object to a result by instantiation.
-     * 
+     *
      * @param input  the input object to transform
      * @return the transformed result
      */
@@ -113,6 +126,24 @@
         } catch (InvocationTargetException ex) {
             throw new FunctorException("InstantiateTransformer: Constructor threw an exception", ex);
         }
+    }
+
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(InstantiateTransformer.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(InstantiateTransformer.class);
+        is.defaultReadObject();
     }
 
 }
--- a/src/java/org/apache/commons/collections/functors/WhileClosure.java
+++ b/src/java/org/apache/commons/collections/functors/WhileClosure.java
@@ -15,6 +15,9 @@
  */
 package org.apache.commons.collections.functors;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 
 import org.apache.commons.collections.Closure;
@@ -23,7 +26,17 @@
 /**
  * Closure implementation that executes a closure repeatedly until a condition is met,
  * like a do-while or while loop.
- * 
+ * <p>
+ * <b>WARNING:</b> from v3.2.2 onwards this class will throw an
+ * {@link UnsupportedOperationException} when trying to serialize or
+ * de-serialize an instance to prevent potential remote code execution exploits.
+ * <p>
+ * In order to re-enable serialization support for {@code WhileClosure}
+ * the following system property can be used (via -Dproperty=true):
+ * <pre>
+ * org.apache.commons.collections.enableUnsafeSerialization
+ * </pre>
+ *
  * @since Commons Collections 3.0
  * @version $Revision: 1.5 $ $Date: 2004/05/16 11:47:38 $
  *
@@ -43,7 +56,7 @@
 
     /**
      * Factory method that performs validation.
-     * 
+     *
      * @param predicate  the predicate used to evaluate when the loop terminates, not null
      * @param closure  the closure the execute, not null
      * @param doLoop  true to act as a do-while loop, always executing the closure once
@@ -63,7 +76,7 @@
     /**
      * Constructor that performs no validation.
      * Use <code>getInstance</code> if you want that.
-     * 
+     *
      * @param predicate  the predicate used to evaluate when the loop terminates, not null
      * @param closure  the closure the execute, not null
      * @param doLoop  true to act as a do-while loop, always executing the closure once
@@ -77,7 +90,7 @@
 
     /**
      * Executes the closure until the predicate is false.
-     * 
+     *
      * @param input  the input object
      */
     public void execute(Object input) {
@@ -91,7 +104,7 @@
 
     /**
      * Gets the predicate in use.
-     * 
+     *
      * @return the predicate
      * @since Commons Collections 3.1
      */
@@ -101,7 +114,7 @@
 
     /**
      * Gets the closure.
-     * 
+     *
      * @return the closure
      * @since Commons Collections 3.1
      */
@@ -111,12 +124,30 @@
 
     /**
      * Is the loop a do-while loop.
-     * 
+     *
      * @return true is do-while, false if while
      * @since Commons Collections 3.1
      */
     public boolean isDoLoop() {
         return iDoLoop;
+    }
+
+    /**
+     * Overrides the default writeObject implementation to prevent
+     * serialization (see COLLECTIONS-580).
+     */
+    private void writeObject(ObjectOutputStream os) throws IOException {
+        FunctorUtils.checkUnsafeSerialization(WhileClosure.class);
+        os.defaultWriteObject();
+    }
+
+    /**
+     * Overrides the default readObject implementation to prevent
+     * de-serialization (see COLLECTIONS-580).
+     */
+    private void readObject(ObjectInputStream is) throws ClassNotFoundException, IOException {
+        FunctorUtils.checkUnsafeSerialization(WhileClosure.class);
+        is.defaultReadObject();
     }
 
 }
--- a/src/java/org/apache/commons/collections/functors/package.html
+++ b/src/java/org/apache/commons/collections/functors/package.html
@@ -24,3 +24,31 @@
 These provide simple callbacks for processing with collections.
 </pre>
+<p>
+<b>WARNING:</b> from v3.2.2 onwards this several classes in this package will
+throw an {@link UnsupportedOperationException} when trying to serialize or
+de-serialize an instance to prevent potential remote code execution exploits.
+</p>
+<p>
+Classes considered to be unsafe are:
+</p>
+<ul>
+  <li>CloneTransformer</li>
+  <li>ForClosure</li>
+  <li>InstantiateFactory</li>
+  <li>InstantiateTransformer</li>
+  <li>InvokerTransformer</li>
+  <li>WhileClosure</li>
+</ul>
+<p>
+In order to re-enable serialization support for these unsafe classes, the
+following system property can be used (via -Dproperty=true):
+<pre>
+org.apache.commons.collections.enableUnsafeSerialization
+</pre>
+<p>
+Be warned though that enabling serialization support for these unsafe classes
+makes your application vulnerable to known exploits and you must ensure that
+your application does not de-serialize objects from untrusted sources.
+</p>
+
 </BODY>

diff --git a/build.xml b/build.xml
index 87aab6d..b5ad803 100644
--- a/build.xml
+++ b/build.xml
@@ -35,6 +35,7 @@
   <property name="component.title.full"    value="Apache Jakarta Commons Collections"/>
 
   <!-- The current version number of this component -->
+  <property name="version"                 value="3.1"/>
   <property name="component.version"       value="3.1"/>
 
   <!-- The base directory for component configuration files -->
@@ -54,7 +55,7 @@
   <property name="build.docs"              value="${build.home}/docs/apidocs"/>
   
   <!-- The name/location of the jar file to build -->
-  <property name="jar.name"             value="${component.name}-${component.version}.jar"/>
+  <property name="jar.name"             value="${component.name}-${version}.jar"/>
   <property name="build.jar.name"       value="${build.home}/${jar.name}"/>
   
   <!-- The name/location of the zip files to build -->
@@ -139,7 +140,7 @@
         <filter token="name"     value="${component.name}"/>
         <filter token="title"    value="${component.title}"/>
         <filter token="package"  value="${component.package}"/>
-        <filter token="version"  value="${component.version}"/>
+        <filter token="version"  value="${version}"/>
       </filterset>
       <fileset dir="${source.conf}" includes="*.MF"/>
     </copy>
@@ -366,7 +367,8 @@
                overview="${source.java}/org/apache/commons/collections/overview.html"
                doctitle="${component.title} ${component.version} API;"
             windowtitle="${component.title} ${component.version} API"
-                 bottom="Copyright &amp;copy; 2001-${year} Apache Software Foundation. All Rights Reserved.">
+                 bottom="Copyright &amp;copy; 2001-${year} Apache Software Foundation. All Rights Reserved."
+        additionalparam="-Xdoclint:none">
     </javadoc>
   </target>
 
diff --git a/src/java/org/apache/commons/collections/MultiHashMap.java b/src/java/org/apache/commons/collections/MultiHashMap.java
index 2b0c43f..24882cd 100644
--- a/src/java/org/apache/commons/collections/MultiHashMap.java
+++ b/src/java/org/apache/commons/collections/MultiHashMap.java
@@ -309,7 +309,7 @@ public class MultiHashMap extends HashMap implements MultiMap {
      * @param item  the value to remove
      * @return the value removed (which was passed in), null if nothing removed
      */
-    public Object remove(Object key, Object item) {
+    public Object removeMapping(Object key, Object item) {
         Collection valuesForKey = getCollection(key);
         if (valuesForKey == null) {
             return null;
diff --git a/src/java/org/apache/commons/collections/MultiMap.java b/src/java/org/apache/commons/collections/MultiMap.java
index 784dfa5..2293d19 100644
--- a/src/java/org/apache/commons/collections/MultiMap.java
+++ b/src/java/org/apache/commons/collections/MultiMap.java
@@ -65,7 +65,7 @@ public interface MultiMap extends Map {
      * @throws ClassCastException if the key or value is of an invalid type
      * @throws NullPointerException if the key or value is null and null is invalid
      */
-    public Object remove(Object key, Object item);
+    public Object removeMapping(Object key, Object item);
 
     //-----------------------------------------------------------------------
     /**
diff --git a/src/java/org/apache/commons/collections/map/MultiKeyMap.java b/src/java/org/apache/commons/collections/map/MultiKeyMap.java
index 37695bc..282d64a 100644
--- a/src/java/org/apache/commons/collections/map/MultiKeyMap.java
+++ b/src/java/org/apache/commons/collections/map/MultiKeyMap.java
@@ -191,7 +191,7 @@ public class MultiKeyMap
      * @param key2  the second key
      * @return the value mapped to the removed key, null if key not in map
      */
-    public Object remove(Object key1, Object key2) {
+    public Object removeMultiKey(Object key1, Object key2) {
         int hashCode = hash(key1, key2);
         int index = map.hashIndex(hashCode, map.data.length);
         AbstractHashedMap.HashEntry entry = map.data[index];
diff --git a/src/test/org/apache/commons/collections/TestMultiHashMap.java b/src/test/org/apache/commons/collections/TestMultiHashMap.java
index c3a467c..f76b8da 100644
--- a/src/test/org/apache/commons/collections/TestMultiHashMap.java
+++ b/src/test/org/apache/commons/collections/TestMultiHashMap.java
@@ -216,7 +216,7 @@ public class TestMultiHashMap extends AbstractTestMap {
         MultiHashMap one = new MultiHashMap();
         Integer value = new Integer(1);
         one.put("One", value);
-        one.remove("One", value);
+        one.removeMapping("One", value);
         
         MultiHashMap two = new MultiHashMap();
         assertEquals(two, one);
@@ -268,7 +268,7 @@ public class TestMultiHashMap extends AbstractTestMap {
         assertEquals(4, map.totalSize());
         map.remove("A");
         assertEquals(3, map.totalSize());
-        map.remove("B", "BC");
+        map.removeMapping("B", "BC");
         assertEquals(2, map.totalSize());
     }
     
@@ -291,7 +291,7 @@ public class TestMultiHashMap extends AbstractTestMap {
         map.remove("A");
         assertEquals(0, map.size("A"));
         assertEquals(3, map.size("B"));
-        map.remove("B", "BC");
+        map.removeMapping("B", "BC");
         assertEquals(0, map.size("A"));
         assertEquals(2, map.size("B"));
     }
diff --git a/src/test/org/apache/commons/collections/map/TestMultiKeyMap.java b/src/test/org/apache/commons/collections/map/TestMultiKeyMap.java
index 6a3f09c..4bb7083 100644
--- a/src/test/org/apache/commons/collections/map/TestMultiKeyMap.java
+++ b/src/test/org/apache/commons/collections/map/TestMultiKeyMap.java
@@ -314,10 +314,10 @@ public class TestMultiKeyMap extends AbstractTestIterableMap {
             switch (key.size()) {
                 case 2:
                 assertEquals(true, multimap.containsKey(key.getKey(0), key.getKey(1)));
-                assertEquals(value, multimap.remove(key.getKey(0), key.getKey(1)));
+                assertEquals(value, multimap.removeMultiKey(key.getKey(0), key.getKey(1)));
                 assertEquals(false, multimap.containsKey(key.getKey(0), key.getKey(1)));
                 assertEquals(size - 1, multimap.size());
-                assertEquals(null, multimap.remove(key.getKey(0), key.getKey(1)));
+                assertEquals(null, multimap.removeMultiKey(key.getKey(0), key.getKey(1)));
                 assertEquals(false, multimap.containsKey(key.getKey(0), key.getKey(1)));
                 break;
                 case 3:
