--- a/src/server-interceptors.ts
+++ b/src/server-interceptors.ts
@@ -1020,15 +1020,17 @@
     return this.host;
   }
   getAuthContext(): AuthContext {
-    if (this.stream.session?.socket instanceof TLSSocket) {
-      const peerCertificate = this.stream.session.socket.getPeerCertificate();
-      return {
-        transportSecurityType: 'ssl',
-        sslPeerCertificate: peerCertificate.raw ? peerCertificate : undefined
-      }
-    } else {
+    if (!(this.stream.session?.socket instanceof TLSSocket)) {
+      return {};
+    }
+    if (!this.stream.session.socket.authorized) {
       return {};
     }
+    const peerCertificate = this.stream.session.socket.getPeerCertificate();
+    return {
+      transportSecurityType: 'ssl',
+      sslPeerCertificate: peerCertificate.raw ? peerCertificate : undefined
+    }
   }
   getConnectionInfo(): ConnectionInfo {
     return this.connectionInfo;
--- a/src/transport.ts
+++ b/src/transport.ts
@@ -259,7 +259,7 @@
       this.maybeStartKeepalivePingTimer();
     }
 
-    if (session.socket instanceof TLSSocket) {
+    if (session.socket instanceof TLSSocket && session.socket.authorized) {
       this.authContext = {
         transportSecurityType: 'ssl',
         sslPeerCertificate: session.socket.getPeerCertificate()
--- a/build/src/server-interceptors.js
+++ b/build/src/server-interceptors.js
@@ -781,16 +781,17 @@
     }
     getAuthContext() {
         var _a;
-        if (((_a = this.stream.session) === null || _a === void 0 ? void 0 : _a.socket) instanceof tls_1.TLSSocket) {
-            const peerCertificate = this.stream.session.socket.getPeerCertificate();
-            return {
-                transportSecurityType: 'ssl',
-                sslPeerCertificate: peerCertificate.raw ? peerCertificate : undefined
-            };
+        if (!(((_a = this.stream.session) === null || _a === void 0 ? void 0 : _a.socket) instanceof tls_1.TLSSocket)) {
+            return {};
         }
-        else {
+        if (!this.stream.session.socket.authorized) {
             return {};
         }
+        const peerCertificate = this.stream.session.socket.getPeerCertificate();
+        return {
+            transportSecurityType: 'ssl',
+            sslPeerCertificate: peerCertificate.raw ? peerCertificate : undefined
+        };
     }
     getConnectionInfo() {
         return this.connectionInfo;
--- a/build/src/transport.js
+++ b/build/src/transport.js
@@ -147,7 +147,7 @@
         if (this.keepaliveWithoutCalls) {
             this.maybeStartKeepalivePingTimer();
         }
-        if (session.socket instanceof tls_1.TLSSocket) {
+        if (session.socket instanceof tls_1.TLSSocket && session.socket.authorized) {
             this.authContext = {
                 transportSecurityType: 'ssl',
                 sslPeerCertificate: session.socket.getPeerCertificate()
