--- a/index.js
+++ b/index.js
@@ -196,6 +196,15 @@ module.exports = function (zipPath, opts, cb) {
             function writeSymlink () {
               readStream.pipe(concat(function (data) {
                 var link = data.toString()
+
+                var linkTarget = path.resolve(path.dirname(dest), link)
+                var relativeLinkTarget = path.relative(opts.dir, linkTarget)
+
+                if (relativeLinkTarget.split(path.sep).indexOf('..') !== -1 || path.isAbsolute(relativeLinkTarget)) {
+                  cancelled = true
+                  return done(new Error('Out of bound symlink target "' + link + '" found while processing file ' + entry.fileName))
+                }
+
                 debug('creating symlink', link, dest)
                 fs.symlink(link, dest, function (err) {
                   if (err) cancelled = true
