--- a/lib/postcss.d.ts
+++ b/lib/postcss.d.ts
@@ -229,7 +229,7 @@ declare namespace postcss {
   export interface Parser<RootNode = Document | Root> {
     (
       css: { toString(): string } | string,
-      opts?: Pick<ProcessOptions, 'document' | 'from' | 'map'>
+      opts?: Pick<ProcessOptions, 'document' | 'from' | 'map' | 'unsafeMap'>
     ): RootNode
   }

@@ -354,6 +354,9 @@ declare namespace postcss {

     /**
      * Disable source map file protections.
+     *
+     * By default source map is limited only for `.map` files
+     * in the `from` folder.
      */
     unsafeMap?: boolean
   }
--- a/lib/previous-map.js
+++ b/lib/previous-map.js
@@ -1,7 +1,7 @@
 'use strict'

 let { existsSync, readFileSync } = require('fs')
-let { dirname, join } = require('path')
+let { dirname, isAbsolute, join, relative, sep } = require('path')
 let { SourceMapConsumer, SourceMapGenerator } = require('source-map-js')

 function fromBase64(str) {
@@ -85,11 +85,21 @@ class PreviousMap {
   }

   loadFile(path, cssFile, trusted) {
-    /* c8 ignore next 5 */
     if (!trusted && !this.unsafeMap) {
       if (!/\.map$/i.test(path)) {
         return undefined
       }
+      if (!cssFile) {
+        return undefined
+      }
+      let relativePath = relative(dirname(cssFile), path)
+      if (
+        relativePath === '..' ||
+        relativePath.startsWith('..' + sep) ||
+        isAbsolute(relativePath)
+      ) {
+        return undefined
+      }
     }
     this.root = dirname(path)
     if (existsSync(path)) {
