--- a/setuptools/package_index.py
+++ b/setuptools/package_index.py
@@ -827,6 +827,10 @@ class PackageIndex(Environment):
             name = name[:-4]  # strip the extra .zip before download

         filename = os.path.join(tmpdir, name)
+
+        # ensure path resolves within the tmpdir
+        if not filename.startswith(str(tmpdir)):
+            raise ValueError(f"Invalid filename {filename}")

         # Download the file
         #

--- a/setuptools/package_index.py
+++ b/setuptools/package_index.py
@@ -1,6 +1,7 @@
 """PyPI and direct package downloading."""
 
 import sys
+import subprocess
 import os
 import re
 import io
@@ -882,16 +883,12 @@
         url, rev = self._vcs_split_rev_from_url(url, pop_prefix=True)
 
         self.info("Doing git clone from %s to %s", url, filename)
-        os.system("git clone --quiet %s %s" % (url, filename))
+        subprocess.check_call(["git", "clone", "--quiet", url, filename])
 
         if rev is not None:
             self.info("Checking out %s", rev)
-            os.system(
-                "git -C %s checkout --quiet %s"
-                % (
-                    filename,
-                    rev,
-                )
+            subprocess.check_call(
+                ["git", "-C", filename, "checkout", "--quiet", rev],
             )
 
         return filename
@@ -901,16 +898,12 @@
         url, rev = self._vcs_split_rev_from_url(url, pop_prefix=True)
 
         self.info("Doing hg clone from %s to %s", url, filename)
-        os.system("hg clone --quiet %s %s" % (url, filename))
+        subprocess.check_call(["hg", "clone", "--quiet", url, filename])
 
         if rev is not None:
             self.info("Updating to %s", rev)
-            os.system(
-                "hg --cwd %s up -C -r %s -q"
-                % (
-                    filename,
-                    rev,
-                )
+            subprocess.check_call(
+                ["hg", "--cwd", filename, "up", "-C", "-r", rev, "-q"],
             )
 
         return filename
--- a/setuptools/tests/test_packageindex.py
+++ b/setuptools/tests/test_packageindex.py
@@ -190,37 +190,38 @@
         url = 'git+https://github.example/group/project@master#egg=foo'
         index = setuptools.package_index.PackageIndex()
 
-        with mock.patch("os.system") as os_system_mock:
+        with mock.patch("subprocess.check_call") as subprocess_mock:
             result = index.download(url, str(tmpdir))
 
-        os_system_mock.assert_called()
+        subprocess_mock.assert_called()
 
         expected_dir = str(tmpdir / 'project@master')
-        expected = (
-            'git clone --quiet ' 'https://github.example/group/project {expected_dir}'
-        ).format(**locals())
-        first_call_args = os_system_mock.call_args_list[0][0]
+        expected = [
+            'git', 'clone', '--quiet',
+            'https://github.example/group/project', expected_dir,
+        ]
+        first_call_args = subprocess_mock.call_args_list[0][0]
         assert first_call_args == (expected,)
 
-        tmpl = 'git -C {expected_dir} checkout --quiet master'
-        expected = tmpl.format(**locals())
-        assert os_system_mock.call_args_list[1][0] == (expected,)
+        expected = ['git', '-C', expected_dir, 'checkout', '--quiet', 'master']
+        assert subprocess_mock.call_args_list[1][0] == (expected,)
         assert result == expected_dir
 
     def test_download_git_no_rev(self, tmpdir):
         url = 'git+https://github.example/group/project#egg=foo'
         index = setuptools.package_index.PackageIndex()
 
-        with mock.patch("os.system") as os_system_mock:
+        with mock.patch("subprocess.check_call") as subprocess_mock:
             result = index.download(url, str(tmpdir))
 
-        os_system_mock.assert_called()
+        subprocess_mock.assert_called()
 
         expected_dir = str(tmpdir / 'project')
-        expected = (
-            'git clone --quiet ' 'https://github.example/group/project {expected_dir}'
-        ).format(**locals())
-        os_system_mock.assert_called_once_with(expected)
+        expected = [
+            'git', 'clone', '--quiet',
+            'https://github.example/group/project', expected_dir,
+        ]
+        subprocess_mock.assert_called_once_with(expected)
 
     def test_download_svn(self, tmpdir):
         url = 'svn+https://svn.example/project#egg=foo'
