--- a/src/standalone/index.ts
+++ b/src/standalone/index.ts
@@ -1,6 +1,7 @@
 import type { WithRequired } from '@apollo/utils.withrequired';
 import cors from 'cors';
 import bodyParser from 'body-parser';
+import { parse as parseContentType } from 'content-type';
 import http, { type IncomingMessage, type ServerResponse } from 'http';
 import type { ListenOptions } from 'net';
 import { parse as urlParse } from 'url';
@@ -27,6 +28,12 @@
   >;
 }
 
+// according to RFC8259, only UTF-8 is allowed in JSON text
+// (see https://datatracker.ietf.org/doc/html/rfc8259#section-8.1)
+// RFC 7159 also specifies that JSON could be UTF-16 or UTF-32,
+// so we allow for that, too
+const validCharset = /^utf-(8|((16|32)(le|be)?))$/i;
+
 export async function startStandaloneServer(
   server: ApolloServer<BaseContext>,
   options?: StartStandaloneServerOptions<BaseContext> & {
@@ -45,7 +52,23 @@
 ): Promise<{ url: string }> {
   const context = options?.context ?? (async () => ({}) as TContext);
   const corsHandler = cors();
-  const jsonHandler = bodyParser.json({ limit: '50mb' });
+  const jsonHandler = bodyParser.json({
+    verify(req) {
+      const charset = parseContentType(req).parameters.charset || 'utf-8';
+      if (!charset.match(validCharset)) {
+        throw Object.assign(
+          new Error(`unsupported charset "${charset.toUpperCase()}"`),
+          {
+            status: 415,
+            name: 'UnsupportedMediaTypeError',
+            charset,
+            type: 'charset.unsupported',
+          },
+        );
+      }
+    },
+    limit: '50mb',
+  });
   const httpServer = http.createServer((req, res) => {
     const errorHandler = finalhandler(req, res, {
       // Use the same onerror as Express.
--- a/dist/cjs/standalone/index.js
+++ b/dist/cjs/standalone/index.js
@@ -6,16 +6,31 @@
 exports.startStandaloneServer = startStandaloneServer;
 const cors_1 = __importDefault(require("cors"));
 const body_parser_1 = __importDefault(require("body-parser"));
+const content_type_1 = require("content-type");
 const http_1 = __importDefault(require("http"));
 const url_1 = require("url");
 const index_js_1 = require("../plugin/drainHttpServer/index.js");
 const urlForHttpServer_js_1 = require("../utils/urlForHttpServer.js");
 const HeaderMap_js_1 = require("../utils/HeaderMap.js");
 const finalhandler_1 = __importDefault(require("finalhandler"));
+const validCharset = /^utf-(8|((16|32)(le|be)?))$/i;
 async function startStandaloneServer(server, options) {
     const context = options?.context ?? (async () => ({}));
     const corsHandler = (0, cors_1.default)();
-    const jsonHandler = body_parser_1.default.json({ limit: '50mb' });
+    const jsonHandler = body_parser_1.default.json({
+        verify(req) {
+            const charset = (0, content_type_1.parse)(req).parameters.charset || 'utf-8';
+            if (!charset.match(validCharset)) {
+                throw Object.assign(new Error(`unsupported charset \"${charset.toUpperCase()}\"`), {
+                    status: 415,
+                    name: 'UnsupportedMediaTypeError',
+                    charset,
+                    type: 'charset.unsupported',
+                });
+            }
+        },
+        limit: '50mb',
+    });
     const httpServer = http_1.default.createServer((req, res) => {
         const errorHandler = (0, finalhandler_1.default)(req, res, {
             onerror(err) {
--- a/dist/esm/standalone/index.js
+++ b/dist/esm/standalone/index.js
@@ -1,15 +1,30 @@
 import cors from 'cors';
 import bodyParser from 'body-parser';
+import { parse as parseContentType } from 'content-type';
 import http from 'http';
 import { parse as urlParse } from 'url';
 import { ApolloServerPluginDrainHttpServer } from '../plugin/drainHttpServer/index.js';
 import { urlForHttpServer } from '../utils/urlForHttpServer.js';
 import { HeaderMap } from '../utils/HeaderMap.js';
 import finalhandler from 'finalhandler';
+const validCharset = /^utf-(8|((16|32)(le|be)?))$/i;
 export async function startStandaloneServer(server, options) {
     const context = options?.context ?? (async () => ({}));
     const corsHandler = cors();
-    const jsonHandler = bodyParser.json({ limit: '50mb' });
+    const jsonHandler = bodyParser.json({
+        verify(req) {
+            const charset = parseContentType(req).parameters.charset || 'utf-8';
+            if (!charset.match(validCharset)) {
+                throw Object.assign(new Error(`unsupported charset \"${charset.toUpperCase()}\"`), {
+                    status: 415,
+                    name: 'UnsupportedMediaTypeError',
+                    charset,
+                    type: 'charset.unsupported',
+                });
+            }
+        },
+        limit: '50mb',
+    });
     const httpServer = http.createServer((req, res) => {
         const errorHandler = finalhandler(req, res, {
             onerror(err) {
--- a/package.json
+++ b/package.json
@@ -138,6 +138,7 @@
     "@graphql-tools/schema": "^10.0.0",
     "async-retry": "^1.2.1",
     "body-parser": "^2.2.0",
+    "content-type": "^1.0.5",
     "cors": "^2.8.5",
     "finalhandler": "^2.1.0",
     "loglevel": "^1.6.8",
