diff --git a/lib/index.js b/lib/index.js
index 29ff69c..cefc75f 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -71,8 +71,10 @@ function constructExportCall(path, exportIdent, exportNames, exportValues, expor
     }));
     for (let i = 0; i < exportNames.length; i++) {
       const exportName = exportNames[i];
+      const computed = stringSpecifiers.has(exportName);
+      const exportNameNode = computed ? _core.types.stringLiteral(exportName) : _core.types.identifier(exportName);
       const exportValue = exportValues[i];
-      statements.push(_core.types.expressionStatement(_core.types.assignmentExpression("=", _core.types.memberExpression(_core.types.identifier(exportObj), _core.types.identifier(exportName)), exportValue)));
+      statements.push(_core.types.expressionStatement(_core.types.assignmentExpression("=", _core.types.memberExpression(_core.types.identifier(exportObj), exportNameNode, computed), exportValue)));
     }
     statements.push(_core.types.expressionStatement(_core.types.callExpression(exportIdent, [_core.types.identifier(exportObj)])));
   }
