--- a/build/transports-uws/polling.js
+++ b/build/transports-uws/polling.js
@@ -113,7 +113,8 @@
         }
         const isBinary = "application/octet-stream" === req.headers["content-type"];
         if (isBinary && this.protocol === 4) {
-            return this.onError("invalid content");
+            this.onError("invalid content");
+            return res.writeStatus("400 Bad Request").end();
         }
         this.dataReq = req;
         this.dataRes = res;
--- a/build/transports/polling.js
+++ b/build/transports/polling.js
@@ -101,7 +101,8 @@
         }
         const isBinary = "application/octet-stream" === req.headers["content-type"];
         if (isBinary && this.protocol === 4) {
-            return this.onError("invalid content");
+            this.onError("invalid content");
+            return res.writeHead(400).end();
         }
         this.dataReq = req;
         this.dataRes = res;
