--- a/src/ntlm.c
+++ b/src/ntlm.c
@@ -278,6 +278,7 @@ static int ntlm_decode_oem_str(struct wire_field_hdr *str_hdr,
     str_offs = le32toh(str_hdr->offset);
     if ((str_offs < payload_offs) ||
         (str_offs > buffer->length) ||
+        (UINT32_MAX - str_offs < str_len) ||
         (str_offs + str_len > buffer->length)) {
         return ERR_DECODE;
     }
@@ -330,6 +331,7 @@ static int ntlm_decode_ucs2_str_hdr(struct ntlm_ctx *ctx,
     str_offs = le32toh(str_hdr->offset);
     if ((str_offs < payload_offs) ||
         (str_offs > buffer->length) ||
+        (UINT32_MAX - str_offs < str_len) ||
         (str_offs + str_len > buffer->length)) {
         return ERR_DECODE;
     }
@@ -416,6 +418,7 @@ static int ntlm_decode_field(struct wire_field_hdr *hdr,
     offs = le32toh(hdr->offset);
     if ((offs < payload_offs) ||
         (offs > buffer->length) ||
+        (UINT32_MAX - offs < len) ||
         (offs + len > buffer->length)) {
         return ERR_DECODE;
     }
--- a/src/ntlm.c
+++ b/src/ntlm.c
@@ -751,8 +751,6 @@ int ntlm_decode_target_info(struct ntlm_ctx *ctx, struct ntlm_buffer *buffer,

 done:
     if (ret) {
-        ntlm_free_buffer_data(&sh);
-        ntlm_free_buffer_data(&cb);
         safefree(nb_computer);
         safefree(nb_domain);
         safefree(dns_computer);--- a/src/ntlm.c
+++ b/src/ntlm.c
@@ -321,7 +321,7 @@ static int ntlm_decode_ucs2_str_hdr(struct ntlm_ctx *ctx,
     char *in, *out = NULL;
     uint16_t str_len;
     uint32_t str_offs;
-    size_t outlen;
+    size_t outlen = 0;
     int ret = 0;

     str_len = le16toh(str_hdr->len);
@@ -342,13 +342,14 @@ static int ntlm_decode_ucs2_str_hdr(struct ntlm_ctx *ctx,

     ret = ntlm_str_convert(ctx->to_oem, in, out, str_len, &outlen);

-    /* make sure to terminate output string */
-    out[outlen] = '\0';
-
 done:
     if (ret) {
         safefree(out);
+    } else {
+        /* make sure to terminate output string */
+        out[outlen] = '\0';
     }
+
     *str = out;
     return ret;
 }
