--- a/build/src/server-call.js
+++ b/build/src/server-call.js
@@ -233,13 +233,6 @@
         this.messagesToPush = [];
         this.maxSendMessageSize = constants_1.DEFAULT_MAX_SEND_MESSAGE_LENGTH;
         this.maxReceiveMessageSize = constants_1.DEFAULT_MAX_RECEIVE_MESSAGE_LENGTH;
-        this.stream.once('error', (err) => {
-            /* We need an error handler to avoid uncaught error event exceptions, but
-             * there is nothing we can reasonably do here. Any error event should
-             * have a corresponding close event, which handles emitting the cancelled
-             * event. And the stream is now in a bad state, so we can't reasonably
-             * expect to be able to send an error over it. */
-        });
         this.stream.once('close', () => {
             var _a;
             trace('Request to method ' +
--- a/build/src/server.js
+++ b/build/src/server.js
@@ -638,6 +638,13 @@
         call.sendError(err);
     }
     _channelzHandler(stream, headers) {
+        stream.once('error', (err) => {
+            /* We need an error handler to avoid uncaught error event exceptions, but
+             * there is nothing we can reasonably do here. Any error event should
+             * have a corresponding close event, which handles emitting the cancelled
+             * event. And the stream is now in a bad state, so we can't reasonably
+             * expect to be able to send an error over it. */
+        });
         const channelzSessionInfo = this.sessions.get(stream.session);
         this.callTracker.addCallStarted();
         channelzSessionInfo === null || channelzSessionInfo === void 0 ? void 0 : channelzSessionInfo.streamTracker.addCallStarted();
@@ -689,6 +696,13 @@
         }
     }
     _streamHandler(stream, headers) {
+        stream.once('error', (err) => {
+            /* We need an error handler to avoid uncaught error event exceptions, but
+             * there is nothing we can reasonably do here. Any error event should
+             * have a corresponding close event, which handles emitting the cancelled
+             * event. And the stream is now in a bad state, so we can't reasonably
+             * expect to be able to send an error over it. */
+        });
         if (this._verifyContentType(stream, headers) !== true) {
             return;
         }
--- a/src/server-call.ts
+++ b/src/server-call.ts
@@ -424,14 +424,6 @@
   ) {
     super();
 
-    this.stream.once('error', (err: ServerErrorResponse) => {
-      /* We need an error handler to avoid uncaught error event exceptions, but
-       * there is nothing we can reasonably do here. Any error event should
-       * have a corresponding close event, which handles emitting the cancelled
-       * event. And the stream is now in a bad state, so we can't reasonably
-       * expect to be able to send an error over it. */
-    });
-
     this.stream.once('close', () => {
       trace(
         'Request to method ' +
--- a/src/server.ts
+++ b/src/server.ts
@@ -915,6 +915,14 @@
     stream: http2.ServerHttp2Stream,
     headers: http2.IncomingHttpHeaders
   ) {
+    stream.once('error', (err: ServerErrorResponse) => {
+      /* We need an error handler to avoid uncaught error event exceptions, but
+       * there is nothing we can reasonably do here. Any error event should
+       * have a corresponding close event, which handles emitting the cancelled
+       * event. And the stream is now in a bad state, so we can't reasonably
+       * expect to be able to send an error over it. */
+    });
+
     const channelzSessionInfo = this.sessions.get(
       stream.session as http2.ServerHttp2Session
     );
@@ -983,6 +991,14 @@
     stream: http2.ServerHttp2Stream,
     headers: http2.IncomingHttpHeaders
   ) {
+    stream.once('error', (err: ServerErrorResponse) => {
+      /* We need an error handler to avoid uncaught error event exceptions, but
+       * there is nothing we can reasonably do here. Any error event should
+       * have a corresponding close event, which handles emitting the cancelled
+       * event. And the stream is now in a bad state, so we can't reasonably
+       * expect to be able to send an error over it. */
+    });
+
     if (this._verifyContentType(stream, headers) !== true) {
       return;
     }

--- a/src/compression-filter.ts
+++ b/src/compression-filter.ts
@@ -119,6 +119,12 @@
       let totalLength = 0;
       const messageParts: Buffer[] = [];
       const decompresser = zlib.createInflate();
+      decompresser.on('error', (error: Error) => {
+        reject({
+          code: Status.INTERNAL,
+          details: 'Failed to decompress deflate-encoded message'
+        });
+      });
       decompresser.on('data', (chunk: Buffer) => {
         messageParts.push(chunk);
         totalLength += chunk.byteLength;
@@ -161,6 +167,12 @@
       let totalLength = 0;
       const messageParts: Buffer[] = [];
       const decompresser = zlib.createGunzip();
+      decompresser.on('error', (error: Error) => {
+        reject({
+          code: Status.INTERNAL,
+          details: 'Failed to decompress deflate-encoded message'
+        });
+      });
       decompresser.on('data', (chunk: Buffer) => {
         messageParts.push(chunk);
         totalLength += chunk.byteLength;
--- a/src/server-call.ts
+++ b/src/server-call.ts
@@ -490,6 +490,12 @@
       return new Promise((resolve, reject) => {
         let totalLength = 0
         const messageParts: Buffer[] = [];
+        decompresser.on('error', (error: Error) => {
+          reject({
+            code: Status.INTERNAL,
+            details: 'Failed to decompress message'
+          });
+        });
         decompresser.on('data', (chunk: Buffer) => {
           messageParts.push(chunk);
           totalLength += chunk.byteLength;
--- a/build/src/compression-filter.js
+++ b/build/src/compression-filter.js
@@ -94,6 +94,12 @@
             let totalLength = 0;
             const messageParts = [];
             const decompresser = zlib.createInflate();
+            decompresser.on('error', (error) => {
+                reject({
+                    code: constants_1.Status.INTERNAL,
+                    details: 'Failed to decompress deflate-encoded message'
+                });
+            });
             decompresser.on('data', (chunk) => {
                 messageParts.push(chunk);
                 totalLength += chunk.byteLength;
@@ -135,6 +141,12 @@
             let totalLength = 0;
             const messageParts = [];
             const decompresser = zlib.createGunzip();
+            decompresser.on('error', (error) => {
+                reject({
+                    code: constants_1.Status.INTERNAL,
+                    details: 'Failed to decompress deflate-encoded message'
+                });
+            });
             decompresser.on('data', (chunk) => {
                 messageParts.push(chunk);
                 totalLength += chunk.byteLength;
--- a/build/src/server-call.js
+++ b/build/src/server-call.js
@@ -293,6 +293,12 @@
             return new Promise((resolve, reject) => {
                 let totalLength = 0;
                 const messageParts = [];
+                decompresser.on('error', (error) => {
+                    reject({
+                        code: constants_1.Status.INTERNAL,
+                        details: 'Failed to decompress message'
+                    });
+                });
                 decompresser.on('data', (chunk) => {
                     messageParts.push(chunk);
                     totalLength += chunk.byteLength;
