--- a/xstream/src/java/com/thoughtworks/xstream/io/binary/BinaryStreamReader.java
+++ b/xstream/src/java/com/thoughtworks/xstream/io/binary/BinaryStreamReader.java
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2006 Joe Walnes.
- * Copyright (C) 2006, 2007, 2011, 2013 XStream Committers.
+ * Copyright (C) 2006, 2007, 2011, 2013, 2024 XStream Committers.
  * All rights reserved.
  *
  * The software in this package is published under the terms of the BSD
@@ -14,6 +14,7 @@ import com.thoughtworks.xstream.converters.ErrorWriter;
 import com.thoughtworks.xstream.io.ExtendedHierarchicalStreamReader;
 import com.thoughtworks.xstream.io.HierarchicalStreamReader;
 import com.thoughtworks.xstream.io.StreamException;
+import com.thoughtworks.xstream.security.InputManipulationException;

 import java.io.DataInputStream;
 import java.io.IOException;
@@ -148,14 +149,19 @@ public class BinaryStreamReader implements ExtendedHierarchicalStreamReader {
     private Token readToken() {
         if (pushback == null) {
             try {
-                Token token = tokenFormatter.read(in);
-                switch (token.getType()) {
-                    case Token.TYPE_MAP_ID_TO_VALUE:
-                        idRegistry.put(token.getId(), token.getValue());
-                        return readToken(); // Next one please.
-                    default:
-                        return token;
-                }
+                boolean mapping = false;
+                do {
+                    final Token token = tokenFormatter.read(in);
+                    switch (token.getType()) {
+                    case Token.TYPE_MAP_ID_TO_VALUE:
+                        idRegistry.put(token.getId(), token.getValue());
+                        mapping ^= true;
+                        continue; // Next one please.
+                    default:
+                        return token;
+                    }
+                } while (mapping);
+                throw new InputManipulationException("Binary stream will never have two mapping tokens in sequence");
             } catch (IOException e) {
                 throw new StreamException(e);
             }

--- a/pom.xml
+++ b/pom.xml
@@ -109,7 +109,7 @@
       <modules>
         <!--module>xstream-its</module--><!-- requires installed repo -->
         <module>xstream-jmh</module>
-        <module>xstream-distribution</module>
+        <!--module>xstream-distribution</module--><!-- requires javadoc artifacts skipped in CI -->
       </modules>
     </profile>
     <profile>
