--- a/data_row.go
+++ b/data_row.go
@@ -51,11 +51,9 @@ func (dst *DataRow) Decode(src []byte) error {
 		// null
 		if msgSize == -1 {
 			dst.Values[i] = nil
+		} else if msgSize < 0 || len(src[rp:]) < msgSize {
+			return &invalidMessageFormatErr{messageType: "DataRow"}
 		} else {
-			if len(src[rp:]) < msgSize {
-				return &invalidMessageFormatErr{messageType: "DataRow"}
-			}
-
 			dst.Values[i] = src[rp : rp+msgSize : rp+msgSize]
 			rp += msgSize
 		}
