--- a/lib/db.js
+++ b/lib/db.js
@@ -542,7 +542,9 @@ var createCollection = function(self, name, options, callback) {
       // Execute command
       self.command(cmd, finalOptions, function(err) {
         if(err) return handleCallback(callback, err);
-        handleCallback(callback, null, new Collection(self, self.s.topology, self.s.databaseName, name, self.s.pkFactory, options));
+
+        try { return handleCallback(callback, null, new Collection(self, self.s.topology, self.s.databaseName, name, self.s.pkFactory, options)); }
+        catch(err) { return handleCallback(callback, err); }
       });
   });
 }
