--- a/src/main/java/org/yaml/snakeyaml/LoaderOptions.java
+++ b/src/main/java/org/yaml/snakeyaml/LoaderOptions.java
@@ -15,9 +15,16 @@
  */
 package org.yaml.snakeyaml;
 
+import java.util.ArrayList;
+
+import org.yaml.snakeyaml.constructor.TagInspector;
+import org.yaml.snakeyaml.constructor.TrustedPrefixesTagInspector;
+
 public class LoaderOptions {
 
     private boolean allowDuplicateKeys = true;
+    // Secure by default - no custom classes are allowed
+    private TagInspector tagInspector = new TrustedPrefixesTagInspector(new ArrayList<String>());
 
     public boolean isAllowDuplicateKeys() {
         return allowDuplicateKeys;
@@ -42,4 +49,11 @@
         this.allowDuplicateKeys = allowDuplicateKeys;
     }
 
+    public TagInspector getTagInspector() {
+        return tagInspector;
+    }
+
+    public void setTagInspector(TagInspector tagInspector) {
+        this.tagInspector = tagInspector;
+    }
 }
--- a/src/main/java/org/yaml/snakeyaml/Yaml.java
+++ b/src/main/java/org/yaml/snakeyaml/Yaml.java
@@ -519,7 +519,7 @@
     }
 
     private Object loadFromReader(StreamReader sreader, Class<?> type) {
-        Composer composer = new Composer(new ParserImpl(sreader), resolver);
+        Composer composer = new Composer(new ParserImpl(sreader), resolver, loadingConfig);
         constructor.setComposer(composer);
         return constructor.getSingleData(type);
     }
@@ -534,7 +534,7 @@
      *         sequence
      */
     public Iterable<Object> loadAll(Reader yaml) {
-        Composer composer = new Composer(new ParserImpl(new StreamReader(yaml)), resolver);
+        Composer composer = new Composer(new ParserImpl(new StreamReader(yaml)), resolver, loadingConfig);
         constructor.setComposer(composer);
         Iterator<Object> result = new Iterator<Object>() {
             @Override
@@ -606,7 +606,7 @@
      * @return parsed root Node for the specified YAML document
      */
     public Node compose(Reader yaml) {
-        Composer composer = new Composer(new ParserImpl(new StreamReader(yaml)), resolver);
+        Composer composer = new Composer(new ParserImpl(new StreamReader(yaml)), resolver, loadingConfig);
         return composer.getSingleNode();
     }
 
@@ -620,7 +620,7 @@
      * @return parsed root Nodes for all the specified YAML documents
      */
     public Iterable<Node> composeAll(Reader yaml) {
-        final Composer composer = new Composer(new ParserImpl(new StreamReader(yaml)), resolver);
+        final Composer composer = new Composer(new ParserImpl(new StreamReader(yaml)), resolver, loadingConfig);
         Iterator<Node> result = new Iterator<Node>() {
             @Override
             public boolean hasNext() {
--- a/src/main/java/org/yaml/snakeyaml/composer/Composer.java
+++ b/src/main/java/org/yaml/snakeyaml/composer/Composer.java
@@ -22,6 +22,7 @@
 import java.util.Map;
 import java.util.Set;
 
+import org.yaml.snakeyaml.LoaderOptions;
 import org.yaml.snakeyaml.events.AliasEvent;
 import org.yaml.snakeyaml.events.Event;
 import org.yaml.snakeyaml.events.MappingStartEvent;
@@ -50,12 +51,18 @@
     private final Resolver resolver;
     private final Map<String, Node> anchors;
     private final Set<Node> recursiveNodes;
+    private final LoaderOptions loadingConfig;
 
     public Composer(Parser parser, Resolver resolver) {
+        this(parser, resolver, new LoaderOptions());
+    }
+
+    public Composer(Parser parser, Resolver resolver, LoaderOptions loadingConfig) {
         this.parser = parser;
         this.resolver = resolver;
         this.anchors = new HashMap<String, Node>();
         this.recursiveNodes = new HashSet<Node>();
+        this.loadingConfig = loadingConfig;
     }
 
     /**
@@ -159,6 +166,10 @@
             resolved = true;
         } else {
             nodeTag = new Tag(tag);
+            if (nodeTag.isCustomGlobal() && !loadingConfig.getTagInspector().allowGlobalTag(nodeTag)) {
+                throw new ComposerException(null, null, "Global tag is not allowed: " + tag,
+                        ev.getStartMark());
+            }
         }
         Node node = new ScalarNode(nodeTag, resolved, ev.getValue(), ev.getStartMark(),
                 ev.getEndMark(), ev.getScalarStyle());
@@ -178,6 +189,10 @@
             resolved = true;
         } else {
             nodeTag = new Tag(tag);
+            if (nodeTag.isCustomGlobal() && !loadingConfig.getTagInspector().allowGlobalTag(nodeTag)) {
+                throw new ComposerException(null, null, "Global tag is not allowed: " + tag,
+                        startEvent.getStartMark());
+            }
         }
         final ArrayList<Node> children = new ArrayList<Node>();
         SequenceNode node = new SequenceNode(nodeTag, resolved, children, startEvent.getStartMark(),
@@ -203,6 +218,10 @@
             resolved = true;
         } else {
             nodeTag = new Tag(tag);
+            if (nodeTag.isCustomGlobal() && !loadingConfig.getTagInspector().allowGlobalTag(nodeTag)) {
+                throw new ComposerException(null, null, "Global tag is not allowed: " + tag,
+                        startEvent.getStartMark());
+            }
         }
 
         final List<NodeTuple> children = new ArrayList<NodeTuple>();
new file mode 100644
--- /dev/null
+++ b/src/main/java/org/yaml/snakeyaml/constructor/TagInspector.java
@@ -0,0 +1,34 @@
+/**
+ * Copyright (c) 2008, SnakeYAML
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.yaml.snakeyaml.constructor;
+
+import org.yaml.snakeyaml.nodes.Tag;
+
+/**
+ * Check if the global tags are allowed (the local tags are always allowed).
+ * It should control the classes to create to prevent possible remote code
+ * invocation when the data comes from untrusted source.
+ */
+public interface TagInspector {
+
+    /**
+     * Check whether the global tag is allowed.
+     *
+     * @param tag - the global tag to allow
+     * @return true when the custom global tag is allowed to create a custom Java instance
+     */
+    boolean allowGlobalTag(Tag tag);
+}
new file mode 100644
--- /dev/null
+++ b/src/main/java/org/yaml/snakeyaml/constructor/TrustedPrefixesTagInspector.java
@@ -0,0 +1,48 @@
+/**
+ * Copyright (c) 2008, SnakeYAML
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.yaml.snakeyaml.constructor;
+
+import java.util.List;
+
+import org.yaml.snakeyaml.nodes.Tag;
+
+/**
+ * Allow to create classes with custom global tag if the class name matches
+ * any of the provided prefixes.
+ */
+public class TrustedPrefixesTagInspector implements TagInspector {
+
+    private final List<String> trustedList;
+
+    /**
+     * Create
+     *
+     * @param trustedList - list of prefixes to allow. It may be the package names
+     */
+    public TrustedPrefixesTagInspector(List<String> trustedList) {
+        this.trustedList = trustedList;
+    }
+
+    @Override
+    public boolean allowGlobalTag(Tag tag) {
+        for (String trusted : trustedList) {
+            if (tag.getClassName().startsWith(trusted)) {
+                return true;
+            }
+        }
+        return false;
+    }
+}
new file mode 100644
--- /dev/null
+++ b/src/main/java/org/yaml/snakeyaml/constructor/TrustedTagInspector.java
@@ -0,0 +1,31 @@
+/**
+ * Copyright (c) 2008, SnakeYAML
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.yaml.snakeyaml.constructor;
+
+import org.yaml.snakeyaml.nodes.Tag;
+
+/**
+ * TagInspector which allows to create any custom instance. Should not be used
+ * when the data comes from untrusted source to prevent possible remote code
+ * invocation.
+ */
+public class TrustedTagInspector implements TagInspector {
+
+    @Override
+    public boolean allowGlobalTag(Tag tag) {
+        return true;
+    }
+}
--- a/src/main/java/org/yaml/snakeyaml/nodes/Tag.java
+++ b/src/main/java/org/yaml/snakeyaml/nodes/Tag.java
@@ -44,6 +44,25 @@
     public static final Tag STR = new Tag(PREFIX + "str");
     public static final Tag SEQ = new Tag(PREFIX + "seq");
     public static final Tag MAP = new Tag(PREFIX + "map");
+
+    // https://yaml.org/type/index.html
+    public static final Set<Tag> standardTags = new HashSet<Tag>();
+    static {
+        standardTags.add(YAML);
+        standardTags.add(MERGE);
+        standardTags.add(SET);
+        standardTags.add(PAIRS);
+        standardTags.add(OMAP);
+        standardTags.add(BINARY);
+        standardTags.add(INT);
+        standardTags.add(FLOAT);
+        standardTags.add(TIMESTAMP);
+        standardTags.add(BOOL);
+        standardTags.add(NULL);
+        standardTags.add(STR);
+        standardTags.add(SEQ);
+        standardTags.add(MAP);
+    }
     public static final Map<Tag, Set<Class<?>>> COMPATIBILITY_MAP;
     static {
         COMPATIBILITY_MAP = new HashMap<Tag, Set<Class<?>>>();
@@ -165,4 +184,13 @@
         return value.equals(Tag.PREFIX + clazz.getName());
     }
 
+    /**
+     * Check if the tag is global and not standard to provide it to TagInspector
+     * for verification.
+     *
+     * @return true when the tag must be verified to avoid remote code invocation
+     */
+    public boolean isCustomGlobal() {
+        return !secondary && !standardTags.contains(this);
+    }
 }

--- a/src/main/java/org/yaml/snakeyaml/LoaderOptions.java
+++ b/src/main/java/org/yaml/snakeyaml/LoaderOptions.java
@@ -24,6 +24,7 @@
     private boolean allowDuplicateKeys = true;
     // Secure by default - no custom classes are allowed
     private TagInspector tagInspector = new TrustedPrefixesTagInspector(new ArrayList<String>());
+    private int nestingDepthLimit = 50;
 
     public boolean isAllowDuplicateKeys() {
         return allowDuplicateKeys;
@@ -56,5 +57,19 @@
 
     public void setTagInspector(TagInspector tagInspector) {
         this.tagInspector = tagInspector;
     }
+
+    public int getNestingDepthLimit() {
+        return nestingDepthLimit;
+    }
+
+    /**
+     * Set max depth of nested collections. When the limit is exceeded an exception is thrown.
+     * Aliases/Anchors are not counted.
+     * This is to prevent a DoS attack
+     * @param nestingDepthLimit - depth to be accepted (50 by default)
+     */
+    public void setNestingDepthLimit(int nestingDepthLimit) {
+        this.nestingDepthLimit = nestingDepthLimit;
+    }
 }
--- a/src/main/java/org/yaml/snakeyaml/composer/Composer.java
+++ b/src/main/java/org/yaml/snakeyaml/composer/Composer.java
@@ -25,6 +25,7 @@
 import org.yaml.snakeyaml.LoaderOptions;
 import org.yaml.snakeyaml.events.AliasEvent;
 import org.yaml.snakeyaml.events.Event;
+import org.yaml.snakeyaml.error.YAMLException;
 import org.yaml.snakeyaml.events.MappingStartEvent;
 import org.yaml.snakeyaml.events.NodeEvent;
 import org.yaml.snakeyaml.events.ScalarEvent;
@@ -53,6 +54,9 @@
     private final Map<String, Node> anchors;
     private final Set<Node> recursiveNodes;
     private final LoaderOptions loadingConfig;
+    // keep the nesting of collections inside other collections
+    private int nestingDepth = 0;
+    private final int nestingDepthLimit;
 
     public Composer(Parser parser, Resolver resolver) {
         this(parser, resolver, new LoaderOptions());
@@ -64,6 +68,7 @@
         this.anchors = new HashMap<String, Node>();
         this.recursiveNodes = new HashSet<Node>();
         this.loadingConfig = loadingConfig;
+        this.nestingDepthLimit = loadingConfig.getNestingDepthLimit();
     }
 
     /**
@@ -135,6 +140,7 @@
         } else {
             NodeEvent event = (NodeEvent) parser.peekEvent();
             String anchor = event.getAnchor();
+            increaseNestingDepth();
             // the check for duplicate anchors has been removed (issue 174)
             if (parser.checkEvent(Event.ID.Scalar)) {
                 node = composeScalarNode(anchor);
@@ -143,6 +149,7 @@
             } else {
                 node = composeMappingNode(anchor);
             }
+            decreaseNestingDepth();
         }
         recursiveNodes.remove(parent);
         return node;
@@ -252,5 +259,26 @@
 
     protected Node composeValueNode(MappingNode node) {
         return composeNode(node);
     }
+
+    /**
+     * Increase nesting depth and fail when it exceeds the denied limit
+     */
+    private void increaseNestingDepth() {
+        if (nestingDepth > nestingDepthLimit) {
+            throw new YAMLException("Nesting Depth exceeded max " + nestingDepthLimit);
+        }
+        nestingDepth++;
+    }
+
+    /**
+     * Indicate that the collection is finished and the nesting is decreased
+     */
+    private void decreaseNestingDepth() {
+        if (nestingDepth > 0) {
+            nestingDepth--;
+        } else {
+            throw new YAMLException("Nesting Depth cannot be negative");
+        }
+    }
 }

diff --git a/src/main/java/org/yaml/snakeyaml/LoaderOptions.java b/src/main/java/org/yaml/snakeyaml/LoaderOptions.java
index 3902547..aa8ba36 100644
--- a/src/main/java/org/yaml/snakeyaml/LoaderOptions.java
+++ b/src/main/java/org/yaml/snakeyaml/LoaderOptions.java
@@ -26,6 +26,8 @@ public class LoaderOptions {
     // Secure by default - no custom classes are allowed
     private TagInspector tagInspector = new TrustedPrefixesTagInspector(new ArrayList<String>());
     private int nestingDepthLimit = 50;
+    private int maxAliasesForCollections = 50; //to prevent YAML at https://en.wikipedia.org/wiki/Billion_laughs_attack
+    private boolean allowRecursiveKeys;
 
     public boolean isAllowDuplicateKeys() {
         return allowDuplicateKeys;
@@ -71,4 +73,20 @@ public class LoaderOptions {
     public void setNestingDepthLimit(int nestingDepthLimit) {
         this.nestingDepthLimit = nestingDepthLimit;
     }
+
+    public int getMaxAliasesForCollections() {
+        return maxAliasesForCollections;
+    }
+
+    public void setMaxAliasesForCollections(int maxAliasesForCollections) {
+        this.maxAliasesForCollections = maxAliasesForCollections;
+    }
+
+    public void setAllowRecursiveKeys(boolean allowRecursiveKeys) {
+        this.allowRecursiveKeys = allowRecursiveKeys;
+    }
+
+    public boolean getAllowRecursiveKeys() {
+        return allowRecursiveKeys;
+    }
 }
diff --git a/src/main/java/org/yaml/snakeyaml/Yaml.java b/src/main/java/org/yaml/snakeyaml/Yaml.java
index 41b8315..d0e160c 100644
--- a/src/main/java/org/yaml/snakeyaml/Yaml.java
+++ b/src/main/java/org/yaml/snakeyaml/Yaml.java
@@ -82,7 +82,7 @@ public class Yaml {
      *            LoadingConfig to control load behavior
      */
     public Yaml(LoaderOptions loadingConfig) {
-        this(new Constructor(), new Representer(), new DumperOptions(), loadingConfig);
+        this(new Constructor(loadingConfig), new Representer(), new DumperOptions(), loadingConfig);
     }
 
     /**
diff --git a/src/main/java/org/yaml/snakeyaml/composer/Composer.java b/src/main/java/org/yaml/snakeyaml/composer/Composer.java
index 34e5d3a..4eba2f1 100644
--- a/src/main/java/org/yaml/snakeyaml/composer/Composer.java
+++ b/src/main/java/org/yaml/snakeyaml/composer/Composer.java
@@ -52,6 +52,7 @@ public class Composer {
     private final Resolver resolver;
     private final Map<String, Node> anchors;
     private final Set<Node> recursiveNodes;
+    private int nonScalarAliasesCount = 0;
     private final LoaderOptions loadingConfig;
     // keep the nesting of collections inside other collections
     private int nestingDepth = 0;
@@ -141,6 +142,12 @@ public class Composer {
                         event.getStartMark());
             }
             node = anchors.get(anchor);
+            if (!(node instanceof ScalarNode)) {
+                this.nonScalarAliasesCount++;
+                if (this.nonScalarAliasesCount > loadingConfig.getMaxAliasesForCollections()) {
+                    throw new YAMLException("Number of aliases for non-scalar nodes exceeds the specified max=" + loadingConfig.getMaxAliasesForCollections());
+                }
+            }
             if (recursiveNodes.remove(node)) {
                 node.setTwoStepsConstruction(true);
             }
diff --git a/src/main/java/org/yaml/snakeyaml/constructor/BaseConstructor.java b/src/main/java/org/yaml/snakeyaml/constructor/BaseConstructor.java
index 8937c34..433392f 100644
--- a/src/main/java/org/yaml/snakeyaml/constructor/BaseConstructor.java
+++ b/src/main/java/org/yaml/snakeyaml/constructor/BaseConstructor.java
@@ -32,6 +32,7 @@ import java.util.SortedSet;
 import java.util.TreeMap;
 import java.util.TreeSet;
 
+import org.yaml.snakeyaml.LoaderOptions;
 import org.yaml.snakeyaml.TypeDescription;
 import org.yaml.snakeyaml.composer.Composer;
 import org.yaml.snakeyaml.composer.ComposerException;
@@ -81,7 +82,13 @@ public abstract class BaseConstructor {
     protected final Map<Class<? extends Object>, TypeDescription> typeDefinitions;
     protected final Map<Tag, Class<? extends Object>> typeTags;
 
+    protected LoaderOptions loadingConfig;
+
     public BaseConstructor() {
+        this(new LoaderOptions());
+    }
+
+    public BaseConstructor(LoaderOptions loadingConfig) {
         constructedObjects = new HashMap<Node, Object>();
         recursiveObjects = new HashSet<Node>();
         maps2fill = new ArrayList<RecursiveTuple<Map<Object, Object>, RecursiveTuple<Object, Object>>>();
@@ -96,6 +103,7 @@ public abstract class BaseConstructor {
                 TreeMap.class));
         typeDefinitions.put(SortedSet.class, new TypeDescription(SortedSet.class, Tag.SET,
                 TreeSet.class));
+        this.loadingConfig = loadingConfig;
     }
 
     public void setComposer(Composer composer) {
@@ -452,15 +460,19 @@ public abstract class BaseConstructor {
             }
             Object value = constructObject(valueNode);
             if (keyNode.isTwoStepsConstruction()) {
-                /*
-                 * if keyObject is created it 2 steps we should postpone putting
-                 * it in map because it may have different hash after
-                 * initialization compared to clean just created one. And map of
-                 * course does not observe key hashCode changes.
-                 */
-                maps2fill.add(0,
-                        new RecursiveTuple<Map<Object, Object>, RecursiveTuple<Object, Object>>(
-                                mapping, new RecursiveTuple<Object, Object>(key, value)));
+                if (loadingConfig.getAllowRecursiveKeys()) {
+                    /*
+                     * if keyObject is created it 2 steps we should postpone putting
+                     * it in map because it may have different hash after
+                     * initialization compared to clean just created one. And map of
+                     * course does not observe key hashCode changes.
+                     */
+                    maps2fill.add(0,
+                            new RecursiveTuple<Map<Object, Object>, RecursiveTuple<Object, Object>>(
+                                    mapping, new RecursiveTuple<Object, Object>(key, value)));
+                } else {
+                    throw new YAMLException("Recursive key for mapping is detected but it is not configured to be allowed.");
+                }
             } else {
                 mapping.put(key, value);
             }
diff --git a/src/main/java/org/yaml/snakeyaml/constructor/Constructor.java b/src/main/java/org/yaml/snakeyaml/constructor/Constructor.java
index 77582ea..dc59e5b 100644
--- a/src/main/java/org/yaml/snakeyaml/constructor/Constructor.java
+++ b/src/main/java/org/yaml/snakeyaml/constructor/Constructor.java
@@ -26,6 +26,7 @@ import java.util.Map;
 import java.util.Set;
 import java.util.UUID;
 
+import org.yaml.snakeyaml.LoaderOptions;
 import org.yaml.snakeyaml.TypeDescription;
 import org.yaml.snakeyaml.error.YAMLException;
 import org.yaml.snakeyaml.introspector.Property;
@@ -46,6 +47,10 @@ public class Constructor extends SafeConstructor {
         this(Object.class);
     }
 
+    public Constructor(LoaderOptions loadingConfig) {
+        this(Object.class, loadingConfig);
+    }
+
     /**
      * Create Constructor for the specified class as the root.
      *
@@ -56,6 +61,10 @@ public class Constructor extends SafeConstructor {
         this(new TypeDescription(checkRoot(theRoot)));
     }
 
+    public Constructor(Class<? extends Object> theRoot, LoaderOptions loadingConfig) {
+        this(new TypeDescription(checkRoot(theRoot)), loadingConfig);
+    }
+
     /**
      * Ugly Java way to check the argument in the constructor
      */
@@ -67,10 +76,19 @@ public class Constructor extends SafeConstructor {
     }
 
     public Constructor(TypeDescription theRoot) {
-        this(theRoot, null);
+        this(theRoot, null, new LoaderOptions());
+    }
+
+    public Constructor(TypeDescription theRoot, LoaderOptions loadingConfig) {
+        this(theRoot, null, loadingConfig);
     }
 
     public Constructor(TypeDescription theRoot, Collection<TypeDescription> moreTDs) {
+        this(theRoot, moreTDs, new LoaderOptions());
+    }
+
+    public Constructor(TypeDescription theRoot, Collection<TypeDescription> moreTDs, LoaderOptions loadingConfig) {
+        super(loadingConfig);
         if (theRoot == null) {
             throw new NullPointerException("Root type must be provided.");
         }
@@ -102,6 +120,10 @@ public class Constructor extends SafeConstructor {
         this(Class.forName(check(theRoot)));
     }
 
+    public Constructor(String theRoot, LoaderOptions loadingConfig) throws ClassNotFoundException {
+        this(Class.forName(check(theRoot)), loadingConfig);
+    }
+
     private static final String check(String s) {
         if (s == null) {
             throw new NullPointerException("Root type must be provided.");
diff --git a/src/main/java/org/yaml/snakeyaml/constructor/SafeConstructor.java b/src/main/java/org/yaml/snakeyaml/constructor/SafeConstructor.java
index a9c0f68..a522c4f 100644
--- a/src/main/java/org/yaml/snakeyaml/constructor/SafeConstructor.java
+++ b/src/main/java/org/yaml/snakeyaml/constructor/SafeConstructor.java
@@ -29,6 +29,7 @@ import java.util.TreeSet;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import org.yaml.snakeyaml.LoaderOptions;
 import org.yaml.snakeyaml.error.YAMLException;
 import org.yaml.snakeyaml.external.biz.base64Coder.Base64Coder;
 import org.yaml.snakeyaml.nodes.MappingNode;
@@ -47,6 +48,11 @@ public class SafeConstructor extends BaseConstructor {
     public static final ConstructUndefined undefinedConstructor = new ConstructUndefined();
 
     public SafeConstructor() {
+        this(new LoaderOptions());
+    }
+
+    public SafeConstructor(LoaderOptions loadingConfig) {
+        super(loadingConfig);
         this.yamlConstructors.put(Tag.NULL, new ConstructYamlNull());
         this.yamlConstructors.put(Tag.BOOL, new ConstructYamlBool());
         this.yamlConstructors.put(Tag.INT, new ConstructYamlInt());
diff --git a/src/test/java/org/yaml/snakeyaml/issues/issue377/BillionLaughsAttackTest.java b/src/test/java/org/yaml/snakeyaml/issues/issue377/BillionLaughsAttackTest.java
new file mode 100644
index 0000000..9a0f1da
--- /dev/null
+++ b/src/test/java/org/yaml/snakeyaml/issues/issue377/BillionLaughsAttackTest.java
@@ -0,0 +1,97 @@
+/**
+ * Copyright (c) 2008, http://www.snakeyaml.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.yaml.snakeyaml.issues.issue377;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import java.util.Map;
+
+import org.junit.Test;
+import org.yaml.snakeyaml.LoaderOptions;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.error.YAMLException;
+
+/**
+ * https://en.wikipedia.org/wiki/Billion_laughs_attack#Variations
+ */
+public class BillionLaughsAttackTest {
+    public static final String data = "a: &a [\"lol\",\"lol\",\"lol\",\"lol\",\"lol\",\"lol\",\"lol\",\"lol\",\"lol\"]\n" +
+            "b: &b [*a,*a,*a,*a,*a,*a,*a,*a,*a]\n" +
+            "c: &c [*b,*b,*b,*b,*b,*b,*b,*b,*b]\n" +
+            "d: &d [*c,*c,*c,*c,*c,*c,*c,*c,*c]\n" +
+            "e: &e [*d,*d,*d,*d,*d,*d,*d,*d,*d]\n" +
+            "f: &f [*e,*e,*e,*e,*e,*e,*e,*e,*e]\n" +
+            "g: &g [*f,*f,*f,*f,*f,*f,*f,*f,*f]\n" +
+            "h: &h [*g,*g,*g,*g,*g,*g,*g,*g,*g]\n" +
+            "i: &i [*h,*h,*h,*h,*h,*h,*h,*h,*h]";
+
+    public static final String scalarAliasesData = "a: &a foo\n" +
+            "b:  *a\n" +
+            "c:  *a\n" +
+            "d:  *a\n" +
+            "e:  *a\n" +
+            "f:  *a\n" +
+            "g:  *a\n";
+
+    @Test
+    public void billionLaughsAttackLoaded() {
+        LoaderOptions settings = new LoaderOptions();
+        settings.setMaxAliasesForCollections(72);
+        Yaml yaml = new Yaml(settings);
+        Map map = (Map) yaml.load(data);
+        assertNotNull(map);
+    }
+
+    @Test
+    public void billionLaughsAttackExpanded() {
+        LoaderOptions settings = new LoaderOptions();
+        settings.setMaxAliasesForCollections(100);
+        Yaml yaml = new Yaml(settings);
+        Map map = (Map) yaml.load(data);
+        assertNotNull(map);
+        try {
+            map.toString();
+            fail("Expected overflow");
+        } catch (Throwable e) {
+            assertTrue(e.getMessage().contains("heap"));
+        }
+    }
+
+    @Test
+    public void billionLaughsAttackWithRestrictedAliases() {
+        LoaderOptions settings = new LoaderOptions();
+        Yaml yaml = new Yaml(settings);
+        try {
+            yaml.load(data);
+            fail();
+        } catch (YAMLException e) {
+            assertEquals("Number of aliases for non-scalar nodes exceeds the specified max=50", e.getMessage());
+        }
+    }
+
+    @Test
+    public void doNotRestrictScalarAliases() {
+        LoaderOptions settings = new LoaderOptions();
+        settings.setMaxAliasesForCollections(5); //smaller than number of aliases for scalars
+        Yaml yaml = new Yaml(settings);
+        Map map = (Map) yaml.load(scalarAliasesData);
+        assertNotNull(map);
+    }
+
+}
diff --git a/src/test/java/org/yaml/snakeyaml/issues/issue377/ReferencesTest.java b/src/test/java/org/yaml/snakeyaml/issues/issue377/ReferencesTest.java
index b7428f1..e1c4b8c 100644
--- a/src/test/java/org/yaml/snakeyaml/issues/issue377/ReferencesTest.java
+++ b/src/test/java/org/yaml/snakeyaml/issues/issue377/ReferencesTest.java
@@ -16,27 +16,126 @@
 package org.yaml.snakeyaml.issues.issue377;
 
 import org.junit.Test;
+import org.yaml.snakeyaml.LoaderOptions;
 import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.constructor.SafeConstructor;
 
-import java.util.Map;
+import java.util.HashMap;
 
-import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
 
 public class ReferencesTest {
 
+    /**
+     * Create data which is difficult to parse.
+     *
+     * @param size - size of the map, defines the complexity
+     * @return YAML to parse
+     */
+    private String createDump(int size) {
+        HashMap root = new HashMap();
+        HashMap s1, s2, t1, t2;
+        s1 = root;
+        s2 = new HashMap();
+        /*
+        the time to parse grows very quickly
+        SIZE -> time to parse in seconds
+        25 -> 1
+        26 -> 2
+        27 -> 3
+        28 -> 8
+        29 -> 13
+        30 -> 28
+        31 -> 52
+        32 -> 113
+        33 -> 245
+        34 -> 500
+         */
+        for (int i = 0; i < size; i++) {
+
+            t1 = new HashMap();
+            t2 = new HashMap();
+            t1.put("foo", "1");
+            t2.put("bar", "2");
+
+            s1.put("a", t1);
+            s1.put("b", t2);
+            s2.put("a", t1);
+            s2.put("b", t2);
+
+            s1 = t1;
+            s2 = t2;
+        }
+
+        // this is VERY BAD code
+        // the map has itself as a key (no idea why it may be used except of a DoS attack)
+        HashMap f = new HashMap();
+        f.put(f, "a");
+        f.put("g", root);
+
+        Yaml yaml = new Yaml(new SafeConstructor());
+        String output = yaml.dump(f);
+        return output;
+    }
+
     @Test
-    public void billionLaughsAttack() {
-        String data = "a: &a [\"lol\",\"lol\",\"lol\",\"lol\",\"lol\",\"lol\",\"lol\",\"lol\",\"lol\"]\n" +
-                "b: &b [*a,*a,*a,*a,*a,*a,*a,*a,*a]\n" +
-                "c: &c [*b,*b,*b,*b,*b,*b,*b,*b,*b]\n" +
-                "d: &d [*c,*c,*c,*c,*c,*c,*c,*c,*c]\n" +
-                "e: &e [*d,*d,*d,*d,*d,*d,*d,*d,*d]\n" +
-                "f: &f [*e,*e,*e,*e,*e,*e,*e,*e,*e]\n" +
-                "g: &g [*f,*f,*f,*f,*f,*f,*f,*f,*f]\n" +
-                "h: &h [*g,*g,*g,*g,*g,*g,*g,*g,*g]\n" +
-                "i: &i [*h,*h,*h,*h,*h,*h,*h,*h,*h]";
-        Yaml yaml = new Yaml();
-        Map map = (Map) yaml.load(data);
-        assertNotNull(map);
+    public void referencesWithRecursiveKeysNotAllowedByDefault() {
+        String output = createDump(30);
+        //System.out.println(output);
+        long time1 = System.currentTimeMillis();
+        // Load
+        LoaderOptions settings = new LoaderOptions();
+        settings.setMaxAliasesForCollections(150);
+        Yaml yaml = new Yaml(settings);
+        try {
+            yaml.load(output);
+            fail();
+        } catch (Exception e) {
+            assertEquals("Recursive key for mapping is detected but it is not configured to be allowed.", e.getMessage());
+        }
+        long time2 = System.currentTimeMillis();
+        float duration = (time2 - time1) / 1000;
+        assertTrue("It should fail quickly. Time was " + duration + " seconds.", duration < 1.0);
     }
-}
\ No newline at end of file
+
+    @Test
+    public void parseManyAliasesForCollections() {
+        String output = createDump(25);
+        // Load
+        // long time1 = System.currentTimeMillis();
+        LoaderOptions settings = new LoaderOptions();
+        settings.setMaxAliasesForCollections(50);
+        settings.setAllowRecursiveKeys(true);
+        Yaml yaml = new Yaml(settings);
+        yaml.load(output);
+        // Disabling this as it runs slower than 0.9 on my machine
+        // long time2 = System.currentTimeMillis();
+        // double duration = (time2 - time1) / 1000.0;
+        // assertTrue("It should take time. Time was " + duration + " seconds.", duration > 0.9);
+        // assertTrue("Time was " + duration + " seconds.", duration < 5.0);
+    }
+
+    @Test
+    public void referencesWithRestrictedAliases() {
+        // without alias restriction this size should occupy tons of CPU, memory and time to parse
+        String bigYAML = createDump(35);
+        // Load
+        long time1 = System.currentTimeMillis();
+        LoaderOptions settings = new LoaderOptions();
+        settings.setMaxAliasesForCollections(40);
+        settings.setAllowRecursiveKeys(true);
+        Yaml yaml = new Yaml(settings);
+        try {
+            yaml.load(bigYAML);
+            fail();
+        } catch (Exception e) {
+            assertEquals("Number of aliases for non-scalar nodes exceeds the specified max=40", e.getMessage());
+        }
+        long time2 = System.currentTimeMillis();
+        float duration = (time2 - time1) / 1000;
+        assertTrue("It should fail quickly. Time was " + duration + " seconds.", duration < 1.0);
+    }
+
+}
diff --git a/src/test/java/org/yaml/snakeyaml/issues/issue38/BigNumberIdTest.java b/src/test/java/org/yaml/snakeyaml/issues/issue38/BigNumberIdTest.java
index d80168a..fb3c97b 100644
--- a/src/test/java/org/yaml/snakeyaml/issues/issue38/BigNumberIdTest.java
+++ b/src/test/java/org/yaml/snakeyaml/issues/issue38/BigNumberIdTest.java
@@ -20,6 +20,7 @@ import java.util.List;
 
 import junit.framework.TestCase;
 
+import org.yaml.snakeyaml.LoaderOptions;
 import org.yaml.snakeyaml.Yaml;
 
 /**
@@ -34,7 +35,9 @@ public class BigNumberIdTest extends TestCase {
             list.add(value);
             list.add(value);
         }
-        Yaml yaml = new Yaml();
+        LoaderOptions options = new LoaderOptions();
+        options.setMaxAliasesForCollections(1500);
+        Yaml yaml = new Yaml(options);
         String output = yaml.dump(list);
         // System.out.println(output);
         //
diff --git a/src/test/java/org/yaml/snakeyaml/recursive/HumanTest.java b/src/test/java/org/yaml/snakeyaml/recursive/HumanTest.java
index eab247e..205759c 100644
--- a/src/test/java/org/yaml/snakeyaml/recursive/HumanTest.java
+++ b/src/test/java/org/yaml/snakeyaml/recursive/HumanTest.java
@@ -28,11 +28,13 @@ import java.util.Set;
 import junit.framework.TestCase;
 
 import org.yaml.snakeyaml.DumperOptions;
+import org.yaml.snakeyaml.LoaderOptions;
 import org.yaml.snakeyaml.DumperOptions.FlowStyle;
 import org.yaml.snakeyaml.TypeDescription;
 import org.yaml.snakeyaml.Util;
 import org.yaml.snakeyaml.Yaml;
 import org.yaml.snakeyaml.constructor.Constructor;
+import org.yaml.snakeyaml.representer.Representer;
 
 public class HumanTest extends TestCase {
 
@@ -272,12 +274,14 @@ public class HumanTest extends TestCase {
         mother.setChildren(children);
         //
 
-        Constructor constructor = new Constructor(Human2.class);
+        LoaderOptions options = new LoaderOptions();
+        options.setAllowRecursiveKeys(true);
+        Constructor constructor = new Constructor(Human2.class, options);
         TypeDescription humanDescription = new TypeDescription(Human2.class);
         humanDescription.putMapPropertyType("children", Human2.class, String.class);
         constructor.addTypeDescription(humanDescription);
 
-        Yaml yaml = new Yaml(constructor);
+        Yaml yaml = new Yaml(constructor, new Representer(), new DumperOptions(), options);
         String output = yaml.dump(son);
         // System.out.println(output);
         String etalon = Util.getLocalResource("recursive/with-children-2.yaml");
diff --git a/src/test/java/org/yaml/snakeyaml/recursive/generics/HumanGenericsTest.java b/src/test/java/org/yaml/snakeyaml/recursive/generics/HumanGenericsTest.java
index 5bc1a0e..aab04dd 100644
--- a/src/test/java/org/yaml/snakeyaml/recursive/generics/HumanGenericsTest.java
+++ b/src/test/java/org/yaml/snakeyaml/recursive/generics/HumanGenericsTest.java
@@ -29,6 +29,7 @@ import java.util.Set;
 
 import junit.framework.TestCase;
 
+import org.yaml.snakeyaml.LoaderOptions;
 import org.yaml.snakeyaml.TypeDescription;
 import org.yaml.snakeyaml.Util;
 import org.yaml.snakeyaml.Yaml;
@@ -226,7 +227,10 @@ public class HumanGenericsTest extends TestCase {
         // load
         TypeDescription humanDescription = new TypeDescription(HumanGen2.class);
         humanDescription.putMapPropertyType("children", HumanGen2.class, String.class);
-        Yaml beanLoader = new Yaml(new Constructor(humanDescription));
+
+        LoaderOptions options = new LoaderOptions();
+        options.setAllowRecursiveKeys(true);
+        Yaml beanLoader = new Yaml(new Constructor(humanDescription, options));
         //
         HumanGen2 son2 = beanLoader.loadAs(output, HumanGen2.class);
         assertNotNull(son2);
