--- a/dist/immutable.js
+++ b/dist/immutable.js
@@ -4227,3 +4227,9 @@
   Iterable.Iterator = Iterator;
 
+  function isProtoKey(key) {
+    return (
+      typeof key === 'string' && (key === '__proto__' || key === 'constructor')
+    );
+  }
+
   mixin(Iterable, {
@@ -4265,7 +4271,13 @@
     toObject: function() {
       assertNotInfinite(this.size);
       var object = {};
-      this.__iterate(function(v, k)  { object[k] = v; });
+      this.__iterate(function(v, k)  {
+        if (isProtoKey(k)) {
+          return;
+        }
+
+        object[k] = v;
+      });
       return object;
     },
