--- a/lib/receiver.js
+++ b/lib/receiver.js
@@ -33,14 +33,27 @@
    * @param {Boolean} [isServer=false] Specifies whether to operate in client or
    *     server mode
    * @param {Number} [maxPayload=0] The maximum allowed message length
+   * @param {Number} [maxBufferedChunks=0] The maximum number of
+   *     buffered data chunks
+   * @param {Number} [maxFragments=0] The maximum number of message
+   *     fragments
    */
-  constructor(binaryType, extensions, isServer, maxPayload) {
+  constructor(
+    binaryType,
+    extensions,
+    isServer,
+    maxPayload,
+    maxBufferedChunks,
+    maxFragments
+  ) {
     super();
 
     this._binaryType = binaryType || BINARY_TYPES[0];
     this[kWebSocket] = undefined;
     this._extensions = extensions || {};
     this._isServer = !!isServer;
+    this._maxBufferedChunks = maxBufferedChunks | 0;
+    this._maxFragments = maxFragments | 0;
     this._maxPayload = maxPayload | 0;
 
     this._bufferedBytes = 0;
@@ -73,6 +86,21 @@
   _write(chunk, encoding, cb) {
     if (this._opcode === 0x08 && this._state == GET_INFO) return cb();
 
+    if (
+      this._maxBufferedChunks > 0 &&
+      this._buffers.length >= this._maxBufferedChunks
+    ) {
+      return cb(
+        error(
+          RangeError,
+          'Too many buffered chunks',
+          false,
+          1008,
+          'WS_ERR_TOO_MANY_BUFFERED_PARTS'
+        )
+      );
+    }
+
     this._bufferedBytes += chunk.length;
     this._buffers.push(chunk);
     this.startLoop(cb);
@@ -424,6 +452,20 @@
     }
 
     if (data.length) {
+      if (
+        this._maxFragments > 0 &&
+        this._fragments.length >= this._maxFragments
+      ) {
+        this._loop = false;
+        return error(
+          RangeError,
+          'Too many message fragments',
+          false,
+          1008,
+          'WS_ERR_TOO_MANY_BUFFERED_PARTS'
+        );
+      }
+
       //
       // This message is not compressed so its lenght is the sum of the payload
       // length of all fragments.
@@ -458,6 +500,21 @@
               false,
               1009,
               'WS_ERR_UNSUPPORTED_MESSAGE_LENGTH'
+            )
+          );
+        }
+
+        if (
+          this._maxFragments > 0 &&
+          this._fragments.length >= this._maxFragments
+        ) {
+          return cb(
+            error(
+              RangeError,
+              'Too many message fragments',
+              false,
+              1008,
+              'WS_ERR_TOO_MANY_BUFFERED_PARTS'
             )
           );
         }
--- a/lib/websocket-server.js
+++ b/lib/websocket-server.js
@@ -36,6 +36,10 @@
    *     track clients
    * @param {Function} [options.handleProtocols] A hook to handle protocols
    * @param {String} [options.host] The hostname where to bind the server
+   * @param {Number} [options.maxBufferedChunks=1048576] The maximum number of
+   *     buffered data chunks
+   * @param {Number} [options.maxFragments=131072] The maximum number of message
+   *     fragments
    * @param {Number} [options.maxPayload=104857600] The maximum allowed message
    *     size
    * @param {Boolean} [options.noServer=false] Enable no server mode
@@ -52,6 +56,8 @@
     super();
 
     options = {
+      maxBufferedChunks: 1024 * 1024,
+      maxFragments: 128 * 1024,
       maxPayload: 100 * 1024 * 1024,
       perMessageDeflate: false,
       handleProtocols: null,
@@ -350,7 +356,13 @@
     socket.write(headers.concat('\r\n').join('\r\n'));
     socket.removeListener('error', socketOnError);
 
-    ws.setSocket(socket, head, this.options.maxPayload);
+    ws.setSocket(
+      socket,
+      head,
+      this.options.maxPayload,
+      this.options.maxBufferedChunks,
+      this.options.maxFragments
+    );
 
     if (this.clients) {
       this.clients.add(ws);
--- a/lib/websocket.js
+++ b/lib/websocket.js
@@ -187,14 +187,20 @@
    *     server and client
    * @param {Buffer} head The first packet of the upgraded stream
    * @param {Number} [maxPayload=0] The maximum allowed message size
+   * @param {Number} [maxBufferedChunks=0] The maximum number of
+   *     buffered data chunks
+   * @param {Number} [maxFragments=0] The maximum number of message
+   *     fragments
    * @private
    */
-  setSocket(socket, head, maxPayload) {
+  setSocket(socket, head, maxPayload, maxBufferedChunks, maxFragments) {
     const receiver = new Receiver(
       this.binaryType,
       this._extensions,
       this._isServer,
-      maxPayload
+      maxPayload,
+      maxBufferedChunks,
+      maxFragments
     );
 
     this._sender = new Sender(socket, this._extensions);
@@ -571,6 +577,10 @@
  *     `Sec-WebSocket-Version` header
  * @param {String} [options.origin] Value of the `Origin` or
  *     `Sec-WebSocket-Origin` header
+ * @param {Number} [options.maxBufferedChunks=1048576] The maximum number of
+ *     buffered data chunks
+ * @param {Number} [options.maxFragments=131072] The maximum number of message
+ *     fragments
  * @param {Number} [options.maxPayload=104857600] The maximum allowed message
  *     size
  * @param {Boolean} [options.followRedirects=false] Whether or not to follow
@@ -582,6 +592,8 @@
 function initAsClient(websocket, address, protocols, options) {
   const opts = {
     protocolVersion: protocolVersions[1],
+    maxBufferedChunks: 1024 * 1024,
+    maxFragments: 128 * 1024,
     maxPayload: 100 * 1024 * 1024,
     perMessageDeflate: true,
     followRedirects: false,
@@ -881,7 +893,13 @@
       }
     }
 
-    websocket.setSocket(socket, head, opts.maxPayload);
+    websocket.setSocket(
+      socket,
+      head,
+      opts.maxPayload,
+      opts.maxBufferedChunks,
+      opts.maxFragments
+    );
   });
 }
 
