--- a/dist/shared/lib/router/utils/prepare-destination.js
+++ b/dist/shared/lib/router/utils/prepare-destination.js
@@ -137,7 +137,8 @@
         validate: false
     });
     const destHostnameCompiler = (0, _pathToRegexp).compile(destHostname, {
-        validate: false
+        validate: false,
+        encode: encodeURIComponent
     });
     // update any params in query values
     for (const [key1, strOrArray] of Object.entries(destQuery)){
--- a/dist/esm/shared/lib/router/utils/prepare-destination.js
+++ b/dist/esm/shared/lib/router/utils/prepare-destination.js
@@ -130,7 +130,8 @@
         validate: false
     });
     const destHostnameCompiler = compile(destHostname, {
-        validate: false
+        validate: false,
+        encode: encodeURIComponent
     });
     // update any params in query values
     for (const [key1, strOrArray] of Object.entries(destQuery)){
