From 560b2d8ef452bbba20ffed69dc155d63ac757b74 Mon Sep 17 00:00:00 2001
From: Kornel <kornel@geekhood.net>
Date: Fri, 27 Jan 2023 01:20:38 +0000
Subject: [PATCH] Don't use regex to trim whitespace

---
 package/node4/index.js | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/node4/index.js b/node4/index.js
--- a/node4/index.js
+++ b/node4/index.js
@@ -21,10 +21,10 @@ function parseCacheControl(header) {
     // TODO: When there is more than one value present for a given directive (e.g., two Expires header fields, multiple Cache-Control: max-age directives),
     // the directive's value is considered invalid. Caches are encouraged to consider responses that have invalid freshness information to be stale
-    var parts = header.trim().split(/\s*,\s*/); // TODO: lame parsing
+    var parts = header.trim().split(/,/);
     for (var _iterator = parts, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
         var _ref;
 
         if (_isArray) {
             if (_i >= _iterator.length) break;
             _ref = _iterator[_i++];
         } else {
             _i = _iterator.next();
             if (_i.done) break;
             _ref = _i.value;
         }
 
         var part = _ref;
 
-        var _part$split = part.split(/\s*=\s*/, 2),
+        var _part$split = part.split(/=/, 2),
             k = _part$split[0],
             v = _part$split[1];
 
-        cc[k] = v === undefined ? true : v.replace(/^"|"$/g, ''); // TODO: lame unquoting
+        cc[k.trim()] = v === undefined ? true : v.trim().replace(/^"|"$/g, '');
     }
 
     return cc;
