--- a/querydsl-core/src/main/java/com/querydsl/core/types/dsl/PathBuilderValidator.java
+++ b/querydsl-core/src/main/java/com/querydsl/core/types/dsl/PathBuilderValidator.java
@@ -41,6 +41,9 @@ public interface PathBuilderValidator extends Serializable {
     PathBuilderValidator DEFAULT = new PathBuilderValidator() {
         @Override
         public Class<?> validate(Class<?> parent, String property, Class<?> propertyType) {
+            if (property.contains(" ")) {
+                throw new IllegalStateException("Unsafe due to CVE-2024-49203");
+            }
             return propertyType;
         }
     };
--- a/querydsl-core/src/test/java/com/querydsl/core/types/dsl/PathBuilderTest.java
+++ b/querydsl-core/src/test/java/com/querydsl/core/types/dsl/PathBuilderTest.java
@@ -14,6 +14,9 @@
 package com.querydsl.core.types.dsl;
 
 import static org.junit.Assert.*;
+import com.querydsl.core.domain.Cat;
+import com.querydsl.core.types.Order;
+import com.querydsl.core.types.OrderSpecifier;
 
 import java.sql.Time;
 import java.util.Date;
@@ -130,4 +133,22 @@ public class PathBuilderTest {
             assertEquals(String.class, entity.get(pathName, Object.class).getType());
     }
 
+    @Test
+    public void order_HQL_injection() {
+        String orderBy = "breed";
+        PathBuilder<Cat> pathBuilder = new PathBuilder<Cat>(Cat.class, "entity");
+        new OrderSpecifier(Order.ASC, pathBuilder.get(orderBy));
+    }
+
+    @Test(expected = IllegalStateException.class)
+    // CVE-2024-49203
+    // https://github.com/OpenFeign/querydsl/security/advisories/GHSA-6q3q-6v5j-h6vg
+    public void unsafe_order_HQL_injection() {
+        String orderBy =
+            "test.name INTERSECT SELECT t FROM Test t WHERE (SELECT cast(pg_sleep(10) AS text))='2'"
+                + " ORDER BY t.id";
+        PathBuilder<Cat> pathBuilder = new PathBuilder<Cat>(Cat.class, "entity");
+        new OrderSpecifier(Order.ASC, pathBuilder.get(orderBy));
+    }
+
 }
