--- a/lib/index.js
+++ b/lib/index.js
@@ -3,6 +3,9 @@ var UnauthorizedError = require('./errors/UnauthorizedError');

 module.exports = function(options) {
   if (!options || !options.secret) throw new Error('secret should be set');
+
+  if (!options.algorithms) throw new Error('algorithms should be set');
+  if (!Array.isArray(options.algorithms)) throw new Error('algorithms must be an array');

   return function(req, res, next) {
     var token;
