--- a/lib/parser.js
+++ b/lib/parser.js
@@ -12,8 +12,7 @@
  * @typedef {import('./types').XastParent} XastParent
  */

-// @ts-ignore sax will be replaced with something else later
-const SAX = require('@trysound/sax');
+const SAX = require('sax');
 const JSAPI = require('./svgo/jsAPI.js');
 const { textElems } = require('../plugins/_collections.js');

@@ -82,6 +81,7 @@ const config = {
   lowercase: true,
   xmlns: true,
   position: true,
+  unparsedEntities: true,
 };

 /**
@@ -113,9 +113,6 @@ const parseSvg = (data, from) => {
     return wrapped;
   };

-  /**
-   * @type {(doctype: string) => void}
-   */
   sax.ondoctype = (doctype) => {
     /**
      * @type {XastDoctype}
@@ -183,9 +180,6 @@ const parseSvg = (data, from) => {
     pushToContent(node);
   };

-  /**
-   * @type {(data: { name: string, attributes: Record<string, { value: string }>}) => void}
-   */
   sax.onopentag = (data) => {
     /**
      * @type {XastElement}
@@ -204,9 +198,6 @@ const parseSvg = (data, from) => {
     stack.push(element);
   };

-  /**
-   * @type {(text: string) => void}
-   */
   sax.ontext = (text) => {
     if (current.type === 'element') {
       // prevent trimming of meaningful whitespace inside textual tags
@@ -237,14 +228,12 @@ const parseSvg = (data, from) => {
     current = stack[stack.length - 1];
   };

-  /**
-   * @type {(e: any) => void}
-   */
   sax.onerror = (e) => {
+    const reason = e.message.split('\n')[0];
     const error = new SvgoParserError(
-      e.reason,
-      e.line + 1,
-      e.column,
+      reason,
+      sax.line + 1,
+      sax.column,
       data,
       from
     );
--- a/package.json
+++ b/package.json
@@ -100,12 +100,12 @@
     ]
   },
   "dependencies": {
-    "@trysound/sax": "0.2.0",
     "commander": "^7.2.0",
     "css-select": "^4.1.3",
     "css-tree": "^1.1.3",
     "csso": "^4.2.0",
     "picocolors": "^1.0.0",
+    "sax": "^1.5.0",
     "stable": "^0.1.8"
   },
   "devDependencies": {
