diff --git a/dist-src/index.js b/dist-src/index.js
index 5eb1927..34b65fd 100644
--- a/dist-src/index.js
+++ b/dist-src/index.js
@@ -27,7 +27,7 @@ export class RequestError extends Error {
         const requestCopy = Object.assign({}, options.request);
         if (options.request.headers.authorization) {
             requestCopy.headers = Object.assign({}, options.request.headers, {
-                authorization: options.request.headers.authorization.replace(/ .*$/, " [REDACTED]"),
+                authorization: options.request.headers.authorization.replace(/(?<! ) .*$/, " [REDACTED]"),
             });
         }
         requestCopy.url = requestCopy.url
-- 
2.43.0

