--- a/src/main/java/com/nimbusds/jose/crypto/PasswordBasedDecrypter.java
+++ b/src/main/java/com/nimbusds/jose/crypto/PasswordBasedDecrypter.java
@@ -61,12 +61,18 @@ import net.jcip.annotations.ThreadSafe;
  * </ul>
  *
  * @author Vladimir Dzhuvinov
- * @version 2016-07-26
+ * @version 2023-12-03
  */
 @ThreadSafe
 public class PasswordBasedDecrypter extends PasswordBasedCryptoProvider implements JWEDecrypter, CriticalHeaderParamsAware {


+	/**
+	 * The maximum allowed iteration count (1 million).
+	 */
+	public static final int MAX_ALLOWED_ITERATION_COUNT = 1_000_000;
+
+
 	/**
 	 * The critical header policy.
 	 */
@@ -142,6 +148,10 @@ public class PasswordBasedDecrypter extends PasswordBasedCryptoProvider implemen

 		final int iterationCount = header.getPBES2Count();

+		if (iterationCount > MAX_ALLOWED_ITERATION_COUNT) {
+			throw new JOSEException("The JWE p2c header exceeds the maximum allowed " + MAX_ALLOWED_ITERATION_COUNT + " count");
+		}
+
 		critPolicy.ensureHeaderPasses(header);

 		final JWEAlgorithm alg = header.getAlgorithm();
