--- a/lib/types/json.js
+++ b/lib/types/json.js
@@ -56,12 +56,16 @@ function json (options) {
-  var limit = typeof opts.limit !== 'number'
-    ? bytes.parse(opts.limit || '100kb')
-    : opts.limit
+  var limit = typeof opts.limit === 'undefined' || opts.limit === null
+    ? 102400 // 100kb default
+    : bytes.parse(opts.limit)
   var inflate = opts.inflate !== false
   var reviver = opts.reviver
   var strict = opts.strict !== false
   var type = opts.type || 'application/json'
   var verify = opts.verify || false

+  if (limit === null) {
+    throw new TypeError('option limit "' + String(opts.limit) + '" is invalid')
+  }
+
   if (verify !== false && typeof verify !== 'function') {
     throw new TypeError('option verify must be function')
   }
--- a/lib/types/raw.js
+++ b/lib/types/raw.js
@@ -36,9 +36,13 @@ function raw (options) {
-  var limit = typeof opts.limit !== 'number'
-    ? bytes.parse(opts.limit || '100kb')
-    : opts.limit
+  var limit = typeof opts.limit === 'undefined' || opts.limit === null
+    ? 102400 // 100kb default
+    : bytes.parse(opts.limit)
   var type = opts.type || 'application/octet-stream'
   var verify = opts.verify || false

+  if (limit === null) {
+    throw new TypeError('option limit "' + String(opts.limit) + '" is invalid')
+  }
+
   if (verify !== false && typeof verify !== 'function') {
     throw new TypeError('option verify must be function')
   }
--- a/lib/types/text.js
+++ b/lib/types/text.js
@@ -38,9 +38,13 @@ function text (options) {
-  var limit = typeof opts.limit !== 'number'
-    ? bytes.parse(opts.limit || '100kb')
-    : opts.limit
+  var limit = typeof opts.limit === 'undefined' || opts.limit === null
+    ? 102400 // 100kb default
+    : bytes.parse(opts.limit)
   var type = opts.type || 'text/plain'
   var verify = opts.verify || false

+  if (limit === null) {
+    throw new TypeError('option limit "' + String(opts.limit) + '" is invalid')
+  }
+
   if (verify !== false && typeof verify !== 'function') {
     throw new TypeError('option verify must be function')
   }
--- a/lib/types/urlencoded.js
+++ b/lib/types/urlencoded.js
@@ -53,9 +53,13 @@ function urlencoded (options) {
-  var limit = typeof opts.limit !== 'number'
-    ? bytes.parse(opts.limit || '100kb')
-    : opts.limit
+  var limit = typeof opts.limit === 'undefined' || opts.limit === null
+    ? 102400 // 100kb default
+    : bytes.parse(opts.limit)
   var type = opts.type || 'application/x-www-form-urlencoded'
   var verify = opts.verify || false

+  if (limit === null) {
+    throw new TypeError('option limit "' + String(opts.limit) + '" is invalid')
+  }
+
   if (verify !== false && typeof verify !== 'function') {
     throw new TypeError('option verify must be function')
   }
